// @kitware/vtk.js/Rendering/Misc/GenericRenderWindow.js
import macros_default from "../../macros.js";
import "../../Common/Core/DataArray.js";
import "../../Common/Core/Points.js";
import URLExtract_default from "../../Common/Core/URLExtract.js";
import "../../Common/DataModel/PolyData.js";
import RenderWindowInteractor_default from "../Core/RenderWindowInteractor.js";
import "../Core/Actor.js";
import "../Core/Mapper.js";
import InteractorStyleTrackballCamera_default from "../../Interaction/Style/InteractorStyleTrackballCamera.js";
import Renderer_default from "../Core/Renderer.js";
import RenderWindow_default from "../Core/RenderWindow.js";
import "./RenderingAPIs.js";
//#region Sources/Rendering/Misc/GenericRenderWindow/index.js
var userParams = URLExtract_default.extractURLParameters();
function vtkGenericRenderWindow(publicAPI, model) {
	const invokeResize = publicAPI.invokeResize;
	delete publicAPI.invokeResize;
	model.renderWindow = RenderWindow_default.newInstance();
	model.renderer = Renderer_default.newInstance();
	model.renderWindow.addRenderer(model.renderer);
	model._apiSpecificRenderWindow = model.renderWindow.newAPISpecificView(userParams.viewAPI ?? model.defaultViewAPI);
	model.renderWindow.addView(model._apiSpecificRenderWindow);
	model.interactor = RenderWindowInteractor_default.newInstance();
	model.interactor.setInteractorStyle(InteractorStyleTrackballCamera_default.newInstance());
	model.interactor.setView(model._apiSpecificRenderWindow);
	model.interactor.initialize();
	publicAPI.setBackground = model.renderer.setBackground;
	publicAPI.setBackground(...model.background);
	publicAPI.resize = () => {
		if (model.container) {
			const dims = model.container.getBoundingClientRect();
			const devicePixelRatio = window.devicePixelRatio || 1;
			model._apiSpecificRenderWindow.setSize(Math.floor(dims.width * devicePixelRatio), Math.floor(dims.height * devicePixelRatio));
			invokeResize();
			model.renderWindow.render();
		}
	};
	publicAPI.setContainer = (el) => {
		if (model.container) model.interactor.unbindEvents(model.container);
		model.container = el;
		model._apiSpecificRenderWindow.setContainer(model.container);
		if (model.container) model.interactor.bindEvents(model.container);
	};
	publicAPI.delete = macros_default.chain(publicAPI.setContainer, model._apiSpecificRenderWindow.delete, () => {
		window.removeEventListener("resize", publicAPI.resize);
	}, publicAPI.delete);
	if (model.listenWindowResize) window.addEventListener("resize", publicAPI.resize);
	publicAPI.resize();
}
var DEFAULT_VALUES = {
	background: [
		.32,
		.34,
		.43
	],
	listenWindowResize: true,
	container: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.get(publicAPI, model, [
		"renderWindow",
		"renderer",
		"_apiSpecificRenderWindow",
		"interactor",
		"container"
	]);
	macros_default.moveToProtected(publicAPI, model, ["_apiSpecificRenderWindow"]);
	macros_default.event(publicAPI, model, "resize");
	vtkGenericRenderWindow(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend);
var GenericRenderWindow_default = {
	newInstance,
	extend
};
//#endregion
export { GenericRenderWindow_default as default, extend, newInstance };
