// @kitware/vtk.js/Rendering/Misc/RenderWindowWithControlBar.js
import macros_default from "../../macros.js";
import "../../Common/Core/DataArray.js";
import "../../Common/Core/Points.js";
import "../../Common/DataModel/PolyData.js";
import "../Core/Actor.js";
import "../Core/Mapper.js";
import GenericRenderWindow_default from "./GenericRenderWindow.js";
import RenderWindowWithControlBar_module_default from "./RenderWindowWithControlBar/RenderWindowWithControlBar.module.css.js";
//#region Sources/Rendering/Misc/RenderWindowWithControlBar/index.js
var CONTROL_STYLE = {
	left(size) {
		return {
			top: "0",
			left: "0",
			bottom: "0",
			right: "unset",
			height: "unset",
			width: `${size}px`
		};
	},
	right(size) {
		return {
			top: "0",
			right: "0",
			bottom: "0",
			left: "unset",
			height: "unset",
			width: `${size}px`
		};
	},
	top(size) {
		return {
			top: "0",
			left: "0",
			right: "0",
			bottom: "unset",
			width: "unset",
			height: `${size}px`
		};
	},
	bottom(size) {
		return {
			bottom: "0",
			left: "0",
			right: "0",
			top: "unset",
			width: "unset",
			height: `${size}px`
		};
	}
};
function applyControlStyle(el, position, size) {
	const styleToApply = CONTROL_STYLE[position](size);
	Object.keys(styleToApply).forEach((key) => {
		el.style[key] = styleToApply[key];
	});
}
function vtkRenderWindowWithControlBar(publicAPI, model) {
	const superClass = { ...publicAPI };
	function resetStyleToZero(key) {
		model.renderWindowContainer.style[key] = "0px";
	}
	function updateControlerStyle() {
		[
			"left",
			"right",
			"top",
			"bottom"
		].forEach(resetStyleToZero);
		model.renderWindowContainer.style[model.controlPosition] = `${model.controlSize}px`;
		applyControlStyle(model.controlContainer, model.controlPosition, model.controlSize);
	}
	model.renderWindowContainer = document.createElement("div");
	model.renderWindowContainer.classList.add(RenderWindowWithControlBar_module_default.renderWindow);
	superClass.setContainer(model.renderWindowContainer);
	model.controlContainer = document.createElement("div");
	model.controlContainer.classList.add(RenderWindowWithControlBar_module_default.control);
	publicAPI.setContainer = (el) => {
		if (model.rootContainer) {
			model.rootContainer.removeChild(model.container);
			model.rootContainer.removeChild(model.controlContainer);
			model.rootContainer.classList.remove(RenderWindowWithControlBar_module_default.rootContainer);
		}
		model.rootContainer = el;
		if (model.rootContainer) {
			model.rootContainer.appendChild(model.container);
			model.rootContainer.appendChild(model.controlContainer);
			model.rootContainer.classList.add(RenderWindowWithControlBar_module_default.rootContainer);
			updateControlerStyle();
			publicAPI.resize();
		}
	};
	publicAPI.setControlSize = (size) => {
		model.controlSize = size;
		updateControlerStyle();
		publicAPI.modified();
	};
	publicAPI.setControlPosition = (pos) => {
		model.controlPosition = pos;
		updateControlerStyle();
		publicAPI.modified();
	};
	if (model.listenWindowResize) window.addEventListener("resize", publicAPI.resize);
	updateControlerStyle();
	publicAPI.resize();
}
var DEFAULT_VALUES = {
	rootContainer: null,
	controlPosition: "left",
	controlSize: 10
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	GenericRenderWindow_default.extend(publicAPI, model);
	macros_default.get(publicAPI, model, [
		"rootContainer",
		"controlContainer",
		"renderWindowContainer"
	]);
	vtkRenderWindowWithControlBar(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkRenderWindowWithControlBar");
var RenderWindowWithControlBar_default = {
	newInstance,
	extend
};
//#endregion
export { RenderWindowWithControlBar_default as default, extend, newInstance };
