// @kitware/vtk.js/Rendering/Misc/TextureLODsDownloader.js
import macros_default from "../../macros.js";
//#region Sources/Rendering/Misc/TextureLODsDownloader/index.js
var { vtkErrorMacro } = macros_default;
var getRemoteFileSize = (url) => new Promise((resolve, reject) => {
	const xhr = new XMLHttpRequest();
	xhr.open("HEAD", url, true);
	xhr.onreadystatechange = () => {
		if (xhr.readyState === 4) if (xhr.status === 200) resolve(xhr.getResponseHeader("Content-Length"));
		else {
			console.log("Failed to get remote file size of", url);
			reject(xhr);
		}
	};
	xhr.send();
});
function vtkTextureLODsDownloader(publicAPI, model) {
	model.classHierarchy.push("vtkTextureLODsDownloader");
	const internal = { downloadStack: [] };
	publicAPI.startDownloads = () => {
		if (!model.texture) {
			vtkErrorMacro("Texture was not set.");
			return;
		}
		if (!model.files || model.files.length === 0) {
			vtkErrorMacro("No files set.");
			return;
		}
		let baseUrl = model.baseUrl;
		if (baseUrl && !baseUrl.endsWith("/")) baseUrl += "/";
		internal.downloadStack = [];
		model.files.forEach((file) => internal.downloadStack.push(`${baseUrl}${file}`));
		const downloadNextTexture = () => {
			if (internal.downloadStack.length === 0) return;
			const asyncDownloadNextTexture = () => {
				setTimeout(downloadNextTexture, model.waitTimeBetweenDownloads);
			};
			const img = new Image();
			if (model.crossOrigin) img.crossOrigin = model.crossOrigin;
			const url = internal.downloadStack.shift();
			getRemoteFileSize(url).then((size) => {
				if (!size || size / 1024 > model.maxTextureLODSize) {
					if (!size) console.log("Failed to get image size");
					else console.log("Skipping image", url, ", because it is larger", "than the max texture size:", model.maxTextureLODSize, "KiB");
					asyncDownloadNextTexture();
					return;
				}
				img.src = url;
				img.decode().then(() => {
					model.texture.setImage(img);
					if (model.stepFinishedCallback) model.stepFinishedCallback();
					asyncDownloadNextTexture();
				}).catch((encodingError) => {
					console.log("Failed to decode image:", img.src);
					console.log("Error is:", encodingError);
					asyncDownloadNextTexture();
				});
			}).catch((xhr) => {
				console.log("Failed to get size of:", url);
				console.log("status was:", xhr.status);
				console.log("statusText was:", xhr.statusText);
				asyncDownloadNextTexture();
			});
		};
		setTimeout(downloadNextTexture, model.waitTimeToStart);
	};
}
var DEFAULT_VALUES = {
	baseUrl: "",
	files: [],
	texture: null,
	crossOrigin: void 0,
	maxTextureLODSize: 5e4,
	stepFinishedCallback: null,
	waitTimeToStart: 4e3,
	waitTimeBetweenDownloads: 0
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, [
		"baseUrl",
		"files",
		"texture",
		"crossOrigin",
		"maxTextureLODSize",
		"stepFinishedCallback",
		"waitTimeToStart",
		"waitTimeBetweenDownloads"
	]);
	vtkTextureLODsDownloader(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkTextureLODsDownloader");
var TextureLODsDownloader_default = {
	newInstance,
	extend
};
//#endregion
export { TextureLODsDownloader_default as default, extend, newInstance };
