// @kitware/vtk.js/IO/Geometry/GLTFImporter/Constants.js
//#region Sources/IO/Geometry/GLTFImporter/Constants.js
var BINARY_HEADER_MAGIC = "glTF";
var BINARY_HEADER_LENGTH = 12;
var BINARY_CHUNK_TYPES = {
	JSON: 1313821514,
	BIN: 5130562
};
var BINARY_HEADER_INTS = 3;
var BINARY_CHUNK_HEADER_INTS = 2;
var MIN_LIGHT_ATTENUATION = .01;
var COMPONENTS = {
	SCALAR: 1,
	VEC2: 2,
	VEC3: 3,
	VEC4: 4,
	MAT2: 4,
	MAT3: 9,
	MAT4: 16
};
var BYTES = {
	5120: 1,
	5121: 1,
	5122: 2,
	5123: 2,
	5125: 4,
	5126: 4
};
var MODES = {
	GL_POINTS: 0,
	GL_LINES: 1,
	GL_LINE_LOOP: 2,
	GL_LINE_STRIP: 3,
	GL_TRIANGLES: 4,
	GL_TRIANGLE_STRIP: 5,
	GL_TRIANGLE_FAN: 6
};
var ARRAY_TYPES = {
	5120: Int8Array,
	5121: Uint8Array,
	5122: Int16Array,
	5123: Uint16Array,
	5125: Uint32Array,
	5126: Float32Array
};
var GL_SAMPLER = {
	NEAREST: 9728,
	LINEAR: 9729,
	NEAREST_MIPMAP_NEAREST: 9984,
	LINEAR_MIPMAP_NEAREST: 9985,
	NEAREST_MIPMAP_LINEAR: 9986,
	LINEAR_MIPMAP_LINEAR: 9987,
	REPEAT: 10497,
	CLAMP_TO_EDGE: 33071,
	MIRRORED_REPEAT: 33648,
	TEXTURE_MAG_FILTER: 10240,
	TEXTURE_MIN_FILTER: 10241,
	TEXTURE_WRAP_S: 10242,
	TEXTURE_WRAP_T: 10243
};
var DEFAULT_SAMPLER = {
	magFilter: GL_SAMPLER.NEAREST,
	minFilter: GL_SAMPLER.LINEAR_MIPMAP_LINEAR,
	wrapS: GL_SAMPLER.REPEAT,
	wrapT: GL_SAMPLER.REPEAT
};
var SEMANTIC_ATTRIBUTE_MAP = {
	NORMAL: "normal",
	POSITION: "position",
	TEXCOORD_0: "texcoord0",
	TEXCOORD_1: "texcoord1",
	WEIGHTS_0: "weight",
	JOINTS_0: "joint",
	COLOR_0: "color",
	TANGENT: "tangent"
};
var ALPHA_MODE = {
	OPAQUE: "OPAQUE",
	MASK: "MASK",
	BLEND: "BLEND"
};
//#endregion
export { ALPHA_MODE, ARRAY_TYPES, BINARY_CHUNK_HEADER_INTS, BINARY_CHUNK_TYPES, BINARY_HEADER_INTS, BINARY_HEADER_LENGTH, BINARY_HEADER_MAGIC, BYTES, COMPONENTS, DEFAULT_SAMPLER, GL_SAMPLER, MIN_LIGHT_ATTENUATION, MODES, SEMANTIC_ATTRIBUTE_MAP };
