// @kitware/vtk.js/IO/Geometry/GLTFImporter/Decoder.js
import BinaryHelper_default from "../../Core/BinaryHelper.js";
import { BINARY_CHUNK_TYPES } from "./Constants.js";
//#region Sources/IO/Geometry/GLTFImporter/Decoder.js
function getChunkInfo(headerStart, data) {
	const header = new Uint32Array(data, headerStart, 2);
	return {
		start: headerStart + 8,
		length: header[0],
		type: header[1]
	};
}
function getAllChunkInfos(data) {
	const maxChunks = Math.floor((data.byteLength - 12) / 8);
	const infos = new Array(maxChunks);
	let chunkStart = 12;
	let chunkCount = 0;
	while (chunkStart < data.byteLength) {
		const chunkInfo = getChunkInfo(chunkStart, data);
		infos[chunkCount++] = chunkInfo;
		chunkStart += chunkInfo.length + 8;
	}
	return infos.slice(0, chunkCount);
}
function getJsonFromChunk(chunkInfo, data) {
	const jsonView = new Uint8Array(data, 20, chunkInfo.length);
	const jsonString = new TextDecoder("utf-8").decode(jsonView);
	return JSON.parse(jsonString);
}
function getBufferFromChunk(chunkInfo, data) {
	return data.slice(chunkInfo.start, chunkInfo.start + chunkInfo.length);
}
function parseGLB(data) {
	if (data.byteLength < 12) throw new Error("Invalid GLB: File too small");
	const headerView = new DataView(data, 0, 12);
	const magic = new Uint8Array(data, 0, 4);
	const header = {
		magic: BinaryHelper_default.arrayBufferToString(magic),
		version: headerView.getUint32(4, true),
		length: headerView.getUint32(8, true)
	};
	if (header.magic !== "glTF") throw new Error("Unsupported glTF-Binary header.");
	if (header.version < 2) throw new Error("Unsupported legacy binary file detected.");
	if (header.length > data.byteLength) throw new Error("Invalid GLB: Declared length exceeds file size");
	const chunkInfos = getAllChunkInfos(data);
	let json = null;
	const buffers = [];
	for (let i = 0; i < chunkInfos.length; i++) {
		const chunkInfo = chunkInfos[i];
		if (chunkInfo.type === BINARY_CHUNK_TYPES.JSON && !json) json = getJsonFromChunk(chunkInfo, data);
		else if (chunkInfo.type === BINARY_CHUNK_TYPES.BIN) buffers.push(getBufferFromChunk(chunkInfo, data));
	}
	if (!json) throw new Error("glTF-Binary: JSON content not found.");
	return {
		json,
		buffers
	};
}
//#endregion
export { parseGLB as default };
