// @kitware/vtk.js/IO/Geometry/GLTFImporter/Utils.js
import macros_default from "../../../macros.js";
import Texture_default from "../../../Rendering/Core/Texture.js";
import { ARRAY_TYPES, BYTES, COMPONENTS, GL_SAMPLER } from "./Constants.js";
//#region Sources/IO/Geometry/GLTFImporter/Utils.js
var { vtkWarningMacro, vtkErrorMacro } = macros_default;
var imageBufferViewCache = /* @__PURE__ */ new WeakMap();
var imageUriCache = /* @__PURE__ */ new Map();
/**
* Get GL enum from sampler parameter
* @param {*} parameter The sampler parameter
* @returns The GL enum
*/
function getGLEnumFromSamplerParameter(parameter) {
	return {
		magFilter: 10240,
		minFilter: 10241,
		wrapS: 10242,
		wrapT: 10243
	}[parameter];
}
function getAccessorArrayTypeAndLength(accessor, bufferView) {
	const ArrayType = ARRAY_TYPES[accessor.componentType];
	const components = COMPONENTS[accessor.type];
	const bytesPerComponent = BYTES[accessor.componentType];
	return {
		ArrayType,
		length: accessor.count * components,
		byteLength: accessor.count * components * bytesPerComponent
	};
}
/**
* Resolves a URL based on the original path
* @param {*} url The URL to resolve
* @param {*} originalPath The original path to resolve the URL against
* @returns The resolved URL or an empty string if the URL is invalid
*/
function resolveUrl(url, originalPath) {
	if (typeof url !== "string" || url === "") return "";
	try {
		if (url.startsWith("data:")) return url;
		if (url.startsWith("blob:")) return url;
		const baseUrl = new URL(originalPath);
		if (!baseUrl.pathname.includes(".") && !baseUrl.pathname.endsWith("/")) baseUrl.pathname += "/";
		if (url.startsWith("http:") || url.startsWith("https:") || url.startsWith("//")) return new URL(url, baseUrl).href;
		if (url.startsWith("/")) return new URL(url, baseUrl).href;
		return new URL(url, baseUrl).href;
	} catch (error) {
		vtkErrorMacro("Error resolving URL:", error);
		return "";
	}
}
/**
* Loads image from buffer or URI
* @param {*} image
* @returns
*/
async function loadImage(image) {
	if (!image) return null;
	const cacheKey = image.bufferView || image.uri;
	const cache = image.bufferView ? imageBufferViewCache : imageUriCache;
	if (cacheKey) {
		const cached = cache.get(cacheKey);
		if (cached) {
			if (typeof cached?.then === "function") return cached;
			const value = cached?.deref?.();
			if (value) return value;
			cache.delete(cacheKey);
		}
	}
	const loadPromise = (async () => {
		if (image.bufferView) {
			const blob = new Blob([image.bufferView.data], { type: image.mimeType });
			return createImageBitmap(blob, {
				premultiplyAlpha: "none",
				colorSpaceConversion: "none",
				imageOrientation: "flipY"
			});
		}
		if (image.uri) {
			const img = new Image();
			img.crossOrigin = "Anonymous";
			await new Promise((resolve, reject) => {
				img.onload = resolve;
				img.onerror = reject;
				img.src = image.uri;
			});
			return createImageBitmap(img, {
				premultiplyAlpha: "none",
				colorSpaceConversion: "none",
				imageOrientation: "flipY"
			});
		}
		return null;
	})();
	if (cacheKey) cache.set(cacheKey, loadPromise);
	try {
		const result = await loadPromise;
		if (cacheKey && result) cache.set(cacheKey, new WeakRef(result));
		return result;
	} catch (err) {
		if (cacheKey) cache.delete(cacheKey);
		throw err;
	}
}
function clearImageCaches() {
	imageUriCache.clear();
	imageBufferViewCache.clear();
}
/**
*
* @param {*} image
* @param {*} sampler
* @param {*} extensions
* @returns
*/
function createVTKTextureFromGLTFTexture(image, sampler, extensions) {
	const texture = Texture_default.newInstance();
	if (sampler) if ("wrapS" in sampler && "wrapT" in sampler || "minFilter" in sampler && "magFilter" in sampler) {
		if (sampler.wrapS === GL_SAMPLER.CLAMP_TO_EDGE || sampler.wrapT === GL_SAMPLER.CLAMP_TO_EDGE) {
			texture.setRepeat(false);
			texture.setEdgeClamp(true);
		} else if (sampler.wrapS === GL_SAMPLER.REPEAT || sampler.wrapT === GL_SAMPLER.REPEAT) {
			texture.setRepeat(true);
			texture.setEdgeClamp(false);
		} else vtkWarningMacro("Mirrored texture wrapping is not supported!");
		const linearFilters = [
			GL_SAMPLER.LINEAR,
			GL_SAMPLER.LINEAR_MIPMAP_NEAREST,
			GL_SAMPLER.NEAREST_MIPMAP_LINEAR,
			GL_SAMPLER.LINEAR_MIPMAP_LINEAR
		];
		if (linearFilters.includes(sampler.minFilter) || linearFilters.includes(sampler.magFilter)) texture.setInterpolate(true);
	} else {
		texture.setMipLevel(8);
		texture.setInterpolate(true);
		texture.setEdgeClamp(true);
	}
	texture.setImageBitmap(image);
	return texture;
}
//#endregion
export { clearImageCaches, createVTKTextureFromGLTFTexture, getAccessorArrayTypeAndLength, getGLEnumFromSamplerParameter, loadImage, resolveUrl };
