// @kitware/vtk.js/Widgets/Representations/ArrowHandleRepresentation.js
import macros_default from "../../macros.js";
import MatrixBuilder_default from "../../Common/Core/MatrixBuilder.js";
import CubeSource_default from "../../Filters/Sources/CubeSource.js";
import ConeSource_default from "../../Filters/Sources/ConeSource.js";
import CircleSource_default from "../../Filters/Sources/CircleSource.js";
import SphereSource_default from "../../Filters/Sources/SphereSource.js";
import Arrow2DSource_default from "../../Filters/Sources/Arrow2DSource.js";
import ViewFinderSource_default from "../../Filters/Sources/ViewFinderSource.js";
import Actor_default from "../../Rendering/Core/Actor.js";
import { OrientationModes } from "../../Rendering/Core/Glyph3DMapper/Constants.js";
import PixelSpaceCallbackMapper_default from "../../Rendering/Core/PixelSpaceCallbackMapper.js";
import { RenderingTypes } from "../Core/WidgetManager/Constants.js";
import { allocateArray } from "./WidgetRepresentation.js";
import GlyphRepresentation_default from "./GlyphRepresentation.js";
import Constants_default from "../Widgets3D/LineWidget/Constants.js";
import { mat3, mat4, vec3 } from "gl-matrix";
//#region Sources/Widgets/Representations/ArrowHandleRepresentation/index.js
var { ShapeType, Shapes2D, ShapesOrientable } = Constants_default;
function vtkArrowHandleRepresentation(publicAPI, model) {
	model.classHierarchy.push("vtkArrowHandleRepresentation");
	const superClass = { ...publicAPI };
	/**
	* Set the shape for the glyph according to lineWidget state inputs
	*/
	function createGlyph(shape) {
		const rep = {
			[ShapeType.STAR]: {
				builder: Arrow2DSource_default,
				initialValues: {
					shape: "star",
					height: .6
				}
			},
			[ShapeType.ARROWHEAD3]: {
				builder: Arrow2DSource_default,
				initialValues: { shape: "triangle" }
			},
			[ShapeType.ARROWHEAD4]: {
				builder: Arrow2DSource_default,
				initialValues: { shape: "arrow4points" }
			},
			[ShapeType.ARROWHEAD6]: {
				builder: Arrow2DSource_default,
				initialValues: { shape: "arrow6points" }
			},
			[ShapeType.CONE]: {
				builder: ConeSource_default,
				initialValues: { direction: [
					0,
					1,
					0
				] }
			},
			[ShapeType.SPHERE]: { builder: SphereSource_default },
			[ShapeType.CUBE]: {
				builder: CubeSource_default,
				initialValues: {
					xLength: .8,
					yLength: .8,
					zLength: .8
				}
			},
			[ShapeType.DISK]: {
				builder: CircleSource_default,
				initialValues: {
					resolution: 30,
					radius: .5,
					direction: [
						0,
						0,
						1
					],
					lines: false,
					face: true
				}
			},
			[ShapeType.CIRCLE]: {
				builder: CircleSource_default,
				initialValues: {
					resolution: 30,
					radius: .5,
					direction: [
						0,
						0,
						1
					],
					lines: true,
					face: false
				}
			},
			[ShapeType.VIEWFINDER]: {
				builder: ViewFinderSource_default,
				initialValues: {
					radius: .1,
					spacing: .3,
					width: 1.4
				}
			},
			[ShapeType.NONE]: { builder: SphereSource_default }
		}[shape];
		return rep.builder.newInstance(rep.initialValues);
	}
	model.displayMapper = PixelSpaceCallbackMapper_default.newInstance();
	model.displayActor = Actor_default.newInstance({ parentProp: publicAPI });
	model.displayActor.setMapper(model.displayMapper);
	model.displayMapper.setInputConnection(publicAPI.getOutputPort());
	publicAPI.addActor(model.displayActor);
	model.alwaysVisibleActors = [model.displayActor];
	const superSetGlyphResolution = publicAPI.setGlyphResolution;
	publicAPI.setGlyphResolution = (r) => {
		const res = superSetGlyphResolution(r);
		model._pipeline.glyph.setPhiResolution(r);
		model._pipeline.glyph.setThetaResolution(r);
		return res;
	};
	function callbackProxy(coords) {
		if (model.displayCallback) {
			const filteredList = [];
			const states = publicAPI.getRepresentationStates();
			for (let i = 0; i < states.length; i++) if (states[i].getActive()) filteredList.push(coords[i]);
			if (filteredList.length) {
				model.displayCallback(filteredList);
				return;
			}
		}
		model.displayCallback();
	}
	publicAPI.setDisplayCallback = (callback) => {
		model.displayCallback = callback;
		model.displayMapper.setCallback(callback ? callbackProxy : null);
	};
	publicAPI.is2DShape = () => Shapes2D.includes(model.shape);
	publicAPI.isOrientableShape = () => ShapesOrientable.includes(model.shape);
	/**
	* Returns the orientation matrix to align glyph on model.orientation.
	* */
	function getOrientationRotation(viewMatrixInv) {
		const displayOrientation = new Float64Array(3);
		const baseDir = [
			0,
			1,
			0
		];
		vec3.transformMat3(displayOrientation, model.orientation, viewMatrixInv);
		displayOrientation[2] = 0;
		const displayMatrix = MatrixBuilder_default.buildFromDegree().rotateFromDirections(baseDir, displayOrientation).getMatrix();
		const displayRotation = new Float64Array(9);
		mat3.fromMat4(displayRotation, displayMatrix);
		return displayRotation;
	}
	function getCameraFacingRotation(scale3, displayRotation, viewMatrix) {
		const rotation = new Float64Array(9);
		mat3.multiply(rotation, viewMatrix, displayRotation);
		vec3.transformMat3(scale3, scale3, rotation);
		return rotation;
	}
	/**
	* Computes the rotation matrix of the glyph. There are 2 rotations:
	*  - a first rotation to be oriented along model.rotation
	*  - an optional second rotation to face the camera
	* @param {vec3} scale3 Scale of the glyph, rotated when glyph is rotated.
	*/
	function getGlyphRotation(scale3) {
		const shouldFaceCamera = model.faceCamera === true || model.faceCamera == null && publicAPI.is2DShape();
		const viewMatrix = new Float64Array(9);
		mat3.fromMat4(viewMatrix, model.viewMatrix);
		const viewMatrixInv = mat3.identity(new Float64Array(9));
		if (shouldFaceCamera) mat3.invert(viewMatrixInv, viewMatrix);
		let orientationRotation = null;
		if (publicAPI.isOrientableShape()) orientationRotation = getOrientationRotation(viewMatrixInv);
		else orientationRotation = mat3.identity(new Float64Array(9));
		if (shouldFaceCamera) orientationRotation = getCameraFacingRotation(scale3, orientationRotation, viewMatrix);
		return orientationRotation;
	}
	function applyOrientation(polyData, states) {
		model._pipeline.mapper.setOrientationArray("orientation");
		model._pipeline.mapper.setOrientationMode(OrientationModes.MATRIX);
		const orientation = allocateArray(polyData, "orientation", states.length, "Float32Array", 9).getData();
		const defaultScale3 = [
			1,
			1,
			1
		];
		for (let i = 0; i < states.length; ++i) {
			const rotation = getGlyphRotation(states[i].getScale3?.() ?? defaultScale3);
			orientation.set(rotation, 9 * i);
		}
	}
	publicAPI.setDirection(applyOrientation);
	publicAPI.setNoOrientation(applyOrientation);
	publicAPI.requestData = (inData, outData) => {
		const shape = publicAPI.getRepresentationStates(inData[0])[0]?.getShape();
		let shouldCreateGlyph = model._pipeline.glyph == null;
		if (model.shape !== shape && Object.values(ShapeType).includes(shape)) {
			model.shape = shape;
			shouldCreateGlyph = true;
		}
		if (shouldCreateGlyph && model.shape) {
			model._pipeline.glyph = createGlyph(model.shape);
			model._pipeline.mapper.setInputConnection(model._pipeline.glyph.getOutputPort(), 1);
		}
		return superClass.requestData(inData, outData);
	};
	publicAPI.updateActorVisibility = (renderingType = RenderingTypes.FRONT_BUFFER, ctxVisible = true, handleVisible = true) => {
		const hasValidState = publicAPI.getRepresentationStates().length > 0;
		superClass.updateActorVisibility(renderingType, ctxVisible, handleVisible && hasValidState);
	};
}
/**
*  'shape' default value is used first time 'shape' mixin is invalid.
*  'faceCamera' controls wether the glyph should face camera or not:
*    - null or undefined to leave it to shape type (i.e. 2D are facing camera,
*    3D are not)
*    - true to face camera
*    - false to not face camera
*/
function defaultValues(initialValues) {
	return {
		faceCamera: null,
		orientation: [
			1,
			0,
			0
		],
		shape: ShapeType.SPHERE,
		viewMatrix: mat4.identity(new Float64Array(16)),
		...initialValues
	};
}
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, defaultValues(initialValues));
	GlyphRepresentation_default.extend(publicAPI, model, initialValues);
	macros_default.setGetArray(publicAPI, model, ["visibilityFlagArray"], 2);
	macros_default.setGetArray(publicAPI, model, ["orientation"], 3);
	macros_default.setGetArray(publicAPI, model, ["viewMatrix"], 16);
	macros_default.setGet(publicAPI, model, ["faceCamera"]);
	vtkArrowHandleRepresentation(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkArrowHandleRepresentation");
var ArrowHandleRepresentation_default = {
	newInstance,
	extend
};
//#endregion
export { ArrowHandleRepresentation_default as default, extend, newInstance };
