// @kitware/vtk.js/Widgets/Representations/CircleContextRepresentation.js
import macros_default from "../../macros.js";
import CircleSource_default from "../../Filters/Sources/CircleSource.js";
import { Resolve } from "../../Rendering/Core/Mapper/Static.js";
import "../../Rendering/Core/Mapper/CoincidentTopologyHelper.js";
import { Behavior } from "./WidgetRepresentation/Constants.js";
import GlyphRepresentation_default from "./GlyphRepresentation.js";
//#region Sources/Widgets/Representations/CircleContextRepresentation/index.js
function vtkCircleContextRepresentation(publicAPI, model) {
	model.classHierarchy.push("vtkCircleContextRepresentation");
	const superClass = { ...publicAPI };
	model._pipeline.actor.getProperty().setOpacity(.2);
	model._pipeline.mapper.setResolveCoincidentTopology(Resolve.PolygonOffset);
	model._pipeline.mapper.setRelativeCoincidentTopologyPolygonOffsetParameters(-1, -1);
	publicAPI.setGlyphResolution = (r) => {
		const res = superClass.setGlyphResolution(r);
		model._pipeline.glyph.setResolution(r);
		return res;
	};
	publicAPI.setDrawBorder = (draw) => {
		const res = superClass.setDrawBorder(draw);
		model._pipeline.glyph.setLines(draw);
		return res;
	};
	publicAPI.setDrawFace = (draw) => {
		const res = superClass.setDrawFace(draw);
		model._pipeline.glyph.setFace(draw);
		return res;
	};
	publicAPI.setOpacity = (opacity) => {
		model._pipeline.actor.getProperty().setOpacity(opacity);
	};
}
function defaultValues(initialValues) {
	return {
		behavior: Behavior.CONTEXT,
		glyphResolution: 32,
		drawBorder: false,
		drawFace: true,
		...initialValues,
		_pipeline: {
			glyph: initialValues?.pipeline?.glyph ?? CircleSource_default.newInstance({
				resolution: initialValues.glyphResolution ?? 32,
				radius: 1,
				lines: initialValues.drawBorder ?? false,
				face: initialValues.drawFace ?? true,
				direction: [
					0,
					0,
					1
				]
			}),
			...initialValues?.pipeline
		}
	};
}
function extend(publicAPI, model, initialValues = {}) {
	GlyphRepresentation_default.extend(publicAPI, model, defaultValues(initialValues));
	macros_default.setGet(publicAPI, model, [
		"glyphResolution",
		"drawFace",
		"drawBorder"
	]);
	macros_default.get(publicAPI, model._pipeline, [
		"glyph",
		"mapper",
		"actor"
	]);
	vtkCircleContextRepresentation(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkCircleContextRepresentation");
var CircleContextRepresentation_default = {
	newInstance,
	extend
};
//#endregion
export { CircleContextRepresentation_default as default, extend, newInstance };
