// @kitware/vtk.js/Widgets/Representations/ConvexFaceContextRepresentation.js
import macros_default from "../../macros.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
import Actor_default from "../../Rendering/Core/Actor.js";
import Mapper_default from "../../Rendering/Core/Mapper.js";
import { RenderingTypes } from "../Core/WidgetManager/Constants.js";
import { Behavior } from "./WidgetRepresentation/Constants.js";
import { allocateArray } from "./WidgetRepresentation.js";
import ContextRepresentation_default from "./ContextRepresentation.js";
//#region Sources/Widgets/Representations/ConvexFaceContextRepresentation/index.js
function vtkConvexFaceContextRepresentation(publicAPI, model) {
	model.classHierarchy.push("vtkConvexFaceContextRepresentation");
	model.internalPolyData = PolyData_default.newInstance({ mtime: 0 });
	model.points = new Float32Array(12);
	model.cells = new Uint8Array([
		4,
		0,
		1,
		2,
		3
	]);
	model.internalPolyData.getPoints().setData(model.points, 3);
	model.internalPolyData.getPolys().setData(model.cells);
	function allocateSize(polyData, size) {
		const points = allocateArray(polyData, "points", size).getData();
		const oldCellsSize = polyData.getPolys().getNumberOfValues();
		const cells = allocateArray(polyData, "polys", size + 1).getData();
		if (oldCellsSize !== cells.length) {
			cells[0] = size;
			for (let i = 0; i < size; i++) cells[i + 1] = i;
		}
		return points;
	}
	model.mapper = Mapper_default.newInstance({ scalarVisibility: false });
	model.actor = Actor_default.newInstance({ parentProp: publicAPI });
	model.actor.getProperty().setOpacity(model.opacity);
	model.mapper.setInputConnection(publicAPI.getOutputPort());
	model.actor.setMapper(model.mapper);
	publicAPI.addActor(model.actor);
	publicAPI.requestData = (inData, outData) => {
		const validState = publicAPI.getRepresentationStates(inData[0]).filter((state) => state.getOrigin());
		const points = allocateSize(model.internalPolyData, validState.length);
		for (let i = 0; i < validState.length; i++) {
			const coords = validState[i].getOrigin();
			points[i * 3] = coords[0];
			points[i * 3 + 1] = coords[1];
			points[i * 3 + 2] = coords[2];
		}
		model.internalPolyData.modified();
		outData[0] = model.internalPolyData;
	};
	publicAPI.getSelectedState = (prop, compositeID) => {
		const state = model.inputData[0];
		const list = publicAPI.getRepresentationStates(state);
		if (state.updateFromOriginRightUp) state.updateFromOriginRightUp(list[0].getOrigin(), list[list.length - 1].getOrigin(), list[1].getOrigin());
		return state;
	};
	const superUpdateActorVisibility = publicAPI.updateActorVisibility;
	publicAPI.updateActorVisibility = (renderingType = RenderingTypes.FRONT_BUFFER, ctxVisible = true, handleVisible = true) => {
		switch (model.behavior) {
			case Behavior.HANDLE:
				if (renderingType === RenderingTypes.PICKING_BUFFER) model.actor.getProperty().setOpacity(1);
				else model.actor.getProperty().setOpacity(model.opacity);
				break;
			case Behavior.CONTEXT:
			default:
				model.actor.getProperty().setOpacity(model.opacity);
				break;
		}
		superUpdateActorVisibility(renderingType, ctxVisible, handleVisible);
	};
}
var DEFAULT_VALUES = {
	defaultColor: [
		1,
		0,
		.5
	],
	opacity: .2
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	ContextRepresentation_default.extend(publicAPI, model, initialValues);
	macros_default.setGetArray(publicAPI, model, ["defaultColor"], 3);
	macros_default.get(publicAPI, model, ["mapper", "actor"]);
	macros_default.setGet(publicAPI, model, ["opacity"]);
	vtkConvexFaceContextRepresentation(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkConvexFaceContextRepresentation");
var ConvexFaceContextRepresentation_default = {
	newInstance,
	extend
};
//#endregion
export { ConvexFaceContextRepresentation_default as default, extend, newInstance };
