// @kitware/vtk.js/Widgets/Representations/GlyphRepresentation.js
import macros_default from "../../macros.js";
import { ScalarMode } from "../../Rendering/Core/Mapper/Constants.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
import SphereSource_default from "../../Filters/Sources/SphereSource.js";
import Actor_default from "../../Rendering/Core/Actor.js";
import { OrientationModes } from "../../Rendering/Core/Glyph3DMapper/Constants.js";
import Glyph3DMapper_default from "../../Rendering/Core/Glyph3DMapper.js";
import { getPixelWorldHeightAtCoord } from "../Core/WidgetManager.js";
import { Behavior } from "./WidgetRepresentation/Constants.js";
import WidgetRepresentation_default, { allocateArray } from "./WidgetRepresentation.js";
import HandleRepresentation_default from "./HandleRepresentation.js";
import ContextRepresentation_default from "./ContextRepresentation.js";
//#region Sources/Widgets/Representations/GlyphRepresentation/index.js
function origin(publicAPI, model) {
	return (polyData, states) => {
		const points = allocateArray(polyData, "points", states.length).getData();
		let j = 0;
		for (let i = 0; i < states.length; ++i) {
			const coord = states[i].getOrigin(model.scaleInPixels && model.displayScaleParams);
			points[j++] = coord[0];
			points[j++] = coord[1];
			points[j++] = coord[2];
		}
	};
}
function noPosition(publicAPI, model) {
	return (polyData, states) => {
		allocateArray(polyData, "points", 0);
	};
}
function color3(publicAPI, model) {
	return (polyData, states) => {
		model._pipeline.mapper.setColorByArrayName("color");
		const colorArray = allocateArray(polyData, "color", states.length, "Uint8Array", 4);
		const colors = colorArray.getData();
		let j = 0;
		for (let i = 0; i < states.length; ++i) {
			let c3 = states[i].getColor3();
			if (states[i].getActive() && model.useActiveColor) c3 = model.activeColor;
			colors[j++] = c3[0];
			colors[j++] = c3[1];
			colors[j++] = c3[2];
			colors[j++] = states[i].getOpacity();
		}
		colorArray.dataChange();
	};
}
function color(publicAPI, model) {
	return (polyData, states) => {
		model._pipeline.mapper.setColorByArrayName("color");
		const colors = allocateArray(polyData, "color", states.length).getData();
		for (let i = 0; i < states.length; ++i) {
			let c = states[i].getColor();
			if (states[i].getActive() && model.useActiveColor) c = model.activeColor;
			colors[i] = c;
		}
	};
}
function noColor(publicAPI, model) {
	return (polyData, states) => {
		model._pipeline.mapper.setColorByArrayName(null);
	};
}
function scale3(publicAPI, model) {
	return (polyData, states) => {
		model._pipeline.mapper.setScaleArray("scale");
		model._pipeline.mapper.setScaleFactor(1);
		model._pipeline.mapper.setScaling(true);
		model._pipeline.mapper.setScaleMode(Glyph3DMapper_default.ScaleModes.SCALE_BY_COMPONENTS);
		const scales = allocateArray(polyData, "scale", states.length, "Float32Array", 3).getData();
		let j = 0;
		for (let i = 0; i < states.length; ++i) {
			const state = states[i];
			let scaleFactor = state.getActive() ? model.activeScaleFactor : 1;
			if (publicAPI.getScaleInPixels()) scaleFactor *= getPixelWorldHeightAtCoord(state.getOrigin(), model.displayScaleParams);
			const scale = state.getScale3?.() ?? model.defaultScale;
			scales[j++] = scaleFactor * scale[0];
			scales[j++] = scaleFactor * scale[1];
			scales[j++] = scaleFactor * scale[2];
		}
	};
}
function scale1(publicAPI, model) {
	return (polyData, states) => {
		model._pipeline.mapper.setScaleArray("scale");
		model._pipeline.mapper.setScaleFactor(1);
		model._pipeline.mapper.setScaling(true);
		const scales = allocateArray(polyData, "scale", states.length).getData();
		for (let i = 0; i < states.length; ++i) {
			const state = states[i];
			let scaleFactor = state.getActive() ? model.activeScaleFactor : 1;
			if (publicAPI.getScaleInPixels()) scaleFactor *= getPixelWorldHeightAtCoord(state.getOrigin(), model.displayScaleParams);
			const scale = state.getScale1?.() ?? model.defaultScale;
			scales[i] = scaleFactor * scale;
		}
	};
}
function noScale(publicAPI, model) {
	return (polyData, states) => {
		model._pipeline.mapper.setScaleArray(null);
		model._pipeline.mapper.setScaleFactor(model.defaultScale);
		model._pipeline.mapper.setScaling(model.defaultScale !== 1);
	};
}
function direction(publicAPI, model) {
	return (polyData, states) => {
		model._pipeline.mapper.setOrientationArray("orientation");
		model._pipeline.mapper.setOrientationMode(OrientationModes.MATRIX);
		const orientation = allocateArray(polyData, "orientation", states.length, "Float64Array", 9).getData();
		for (let i = 0; i < states.length; ++i) {
			const state = states[i];
			const right = state.getRight ? state.getRight() : [
				1,
				0,
				0
			];
			const up = state.getUp ? state.getUp() : [
				0,
				1,
				0
			];
			const dir = state.getDirection ? state.getDirection() : [
				0,
				0,
				1
			];
			orientation.set(right, 9 * i);
			orientation.set(up, 9 * i + 3);
			orientation.set(dir, 9 * i + 6);
		}
	};
}
function noOrientation(publicAPI, model) {
	return (polyData, states) => {
		model._pipeline.mapper.setOrientationArray(null);
	};
}
function vtkGlyphRepresentation(publicAPI, model) {
	model.classHierarchy.push("vtkGlyphRepresentation");
	const superClass = { ...publicAPI };
	const internalPolyData = PolyData_default.newInstance({ mtime: 0 });
	function hasMixin(states, ...requiredMixins) {
		return requiredMixins.every((requiredMixin) => states[0]?.[`get${macros_default.capitalize(requiredMixin)}`]?.() != null);
	}
	publicAPI.getRepresentationStates = (input = model.inputData[0]) => superClass.getRepresentationStates(input).filter((state) => state.getOrigin?.() && (state.isVisible?.() ?? true));
	publicAPI.getMixins = (states) => {
		const glyphProperties = {};
		if (hasMixin(states, "origin")) glyphProperties.position = model.applyMixin.origin;
		else glyphProperties.position = model.applyMixin.noPosition;
		if (hasMixin(states, "color3")) glyphProperties.color = model.applyMixin.color3;
		else if (hasMixin(states, "color")) glyphProperties.color = model.applyMixin.color;
		else glyphProperties.color = model.applyMixin.noColor;
		if (hasMixin(states, "scale3")) glyphProperties.scale = model.applyMixin.scale3;
		else if (hasMixin(states, "scale1")) glyphProperties.scale = model.applyMixin.scale1;
		else glyphProperties.scale = model.applyMixin.noScale;
		if (hasMixin(states, "direction")) glyphProperties.orientation = model.applyMixin.direction;
		else glyphProperties.orientation = model.applyMixin.noOrientation;
		return glyphProperties;
	};
	publicAPI.requestData = (inData, outData) => {
		const states = publicAPI.getRepresentationStates(inData[0]);
		outData[0] = internalPolyData;
		const glyphProperties = publicAPI.getMixins(states);
		Object.values(glyphProperties).forEach((property) => property(internalPolyData, states));
		internalPolyData.getPoints().modified();
		internalPolyData.modified();
	};
	WidgetRepresentation_default.connectPipeline(model._pipeline);
	publicAPI.addActor(model._pipeline.actor);
}
function defaultValues(publicAPI, model, initialValues) {
	return {
		defaultScale: 1,
		...initialValues,
		_pipeline: {
			source: initialValues._pipeline?.source ?? publicAPI,
			glyph: initialValues._pipeline?.glyph ?? SphereSource_default.newInstance({
				phiResolution: 8,
				thetaResolution: 8
			}),
			mapper: initialValues._pipeline?.mapper ?? Glyph3DMapper_default.newInstance({ scalarMode: ScalarMode.USE_POINT_FIELD_DATA }),
			actor: initialValues._pipeline?.actor ?? Actor_default.newInstance({ parentProp: publicAPI }),
			...initialValues._pipeline
		},
		applyMixin: {
			origin: initialValues.applyMixin?.origin ?? origin(publicAPI, model),
			noPosition: initialValues.applyMixin?.noPosition ?? noPosition(publicAPI, model),
			color3: initialValues.applyMixin?.color3 ?? color3(publicAPI, model),
			color: initialValues.applyMixin?.color ?? color(publicAPI, model),
			noColor: initialValues.applyMixin?.noColor ?? noColor(publicAPI, model),
			scale3: initialValues.applyMixin?.scale3 ?? scale3(publicAPI, model),
			scale1: initialValues.applyMixin?.scale1 ?? scale1(publicAPI, model),
			noScale: initialValues.applyMixin?.noScale ?? noScale(publicAPI, model),
			direction: initialValues.applyMixin?.direction ?? direction(publicAPI, model),
			noOrientation: initialValues.applyMixin?.noOrientation ?? noOrientation(publicAPI, model),
			...initialValues.applyMixin
		}
	};
}
function extend(publicAPI, model, initialValues = {}) {
	if (initialValues.behavior === Behavior.CONTEXT) ContextRepresentation_default.extend(publicAPI, model, defaultValues(publicAPI, model, initialValues));
	else HandleRepresentation_default.extend(publicAPI, model, defaultValues(publicAPI, model, initialValues));
	if ("lighting" in initialValues) model._pipeline.actor.getProperty().setLighting(initialValues.lighting);
	macros_default.setGet(publicAPI, model._pipeline, ["defaultScale"]);
	macros_default.get(publicAPI, model._pipeline, [
		"glyph",
		"mapper",
		"actor"
	]);
	macros_default.setGet(publicAPI, model.applyMixin, Object.keys(model.applyMixin));
	vtkGlyphRepresentation(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkGlyphRepresentation");
var GlyphRepresentation_default = {
	newInstance,
	extend
};
//#endregion
export { color, color3, GlyphRepresentation_default as default, direction, extend, newInstance, noColor, noOrientation, noPosition, noScale, origin, scale1, scale3 };
