// @kitware/vtk.js/Widgets/Representations/ImplicitPlaneRepresentation.js
import macros_default from "../../macros.js";
import Plane_default from "../../Common/DataModel/Plane.js";
import MatrixBuilder_default from "../../Common/Core/MatrixBuilder.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
import Cutter_default from "../../Filters/Core/Cutter.js";
import ClosedPolyLineToSurfaceFilter_default from "../../Filters/General/ClosedPolyLineToSurfaceFilter.js";
import CubeSource_default from "../../Filters/Sources/CubeSource.js";
import CylinderSource_default from "../../Filters/Sources/CylinderSource.js";
import SphereSource_default from "../../Filters/Sources/SphereSource.js";
import Constants_default from "../../Rendering/Core/Property/Constants.js";
import Actor_default from "../../Rendering/Core/Actor.js";
import Mapper_default from "../../Rendering/Core/Mapper.js";
import PixelSpaceCallbackMapper_default from "../../Rendering/Core/PixelSpaceCallbackMapper.js";
import Constants_default$1 from "../Core/WidgetManager/Constants.js";
import { getPixelWorldHeightAtCoord } from "../Core/WidgetManager.js";
import StateBuilder_default from "../Core/StateBuilder.js";
import WidgetRepresentation_default from "./WidgetRepresentation.js";
//#region Sources/Widgets/Representations/ImplicitPlaneRepresentation/index.js
var { RenderingTypes } = Constants_default$1;
var { Interpolation, Representation } = Constants_default;
function generateState() {
	return StateBuilder_default.createBuilder().addField({
		name: "origin",
		initialValue: [
			0,
			0,
			0
		]
	}).addField({
		name: "normal",
		initialValue: [
			0,
			0,
			1
		]
	}).addField({
		name: "activeHandle",
		initialValue: null
	}).addField({ name: "updateMethodName" }).build();
}
var STYLE_PIPELINE_NAMES = [
	"plane",
	"outline",
	"normal",
	"origin",
	"display2D"
];
var STYLE_DEFAULT = {
	active: {
		plane: {
			opacity: 1,
			color: [
				0,
				.9,
				0
			]
		},
		normal: {
			opacity: 1,
			color: [
				0,
				.9,
				0
			]
		},
		origin: {
			opacity: 1,
			color: [
				0,
				.9,
				0
			]
		}
	},
	inactive: {
		plane: {
			opacity: .6,
			color: [
				1,
				1,
				1
			]
		},
		normal: {
			opacity: 1,
			color: [
				.9,
				0,
				0
			]
		},
		origin: {
			opacity: 1,
			color: [
				1,
				0,
				0
			]
		}
	},
	static: {
		display2D: { representation: Representation.POINT },
		outline: {
			color: [
				1,
				1,
				1
			],
			opacity: 1,
			representation: Representation.WIREFRAME,
			interpolation: Interpolation.FLAT
		}
	}
};
function vtkImplicitPlaneRepresentation(publicAPI, model) {
	model.classHierarchy.push("vtkImplicitPlaneRepresentation");
	model.plane = Plane_default.newInstance();
	model.matrix = MatrixBuilder_default.buildFromDegree();
	model._pipelines = {};
	model._pipelines.outline = {
		source: CubeSource_default.newInstance(),
		mapper: Mapper_default.newInstance(),
		actor: Actor_default.newInstance({
			pickable: false,
			_parentProp: publicAPI
		})
	};
	model._pipelines.plane = {
		source: Cutter_default.newInstance({ cutFunction: model.plane }),
		filter: ClosedPolyLineToSurfaceFilter_default.newInstance(),
		mapper: Mapper_default.newInstance(),
		actor: Actor_default.newInstance({
			pickable: true,
			_parentProp: publicAPI
		})
	};
	model._pipelines.origin = {
		source: SphereSource_default.newInstance(),
		mapper: Mapper_default.newInstance(),
		actor: Actor_default.newInstance({
			pickable: true,
			_parentProp: publicAPI
		})
	};
	model._pipelines.normal = {
		source: CylinderSource_default.newInstance(),
		mapper: Mapper_default.newInstance(),
		actor: Actor_default.newInstance({
			pickable: true,
			_parentProp: publicAPI
		})
	};
	model._pipelines.display2D = {
		source: publicAPI,
		mapper: PixelSpaceCallbackMapper_default.newInstance(),
		actor: Actor_default.newInstance({
			pickable: false,
			_parentProp: publicAPI
		})
	};
	model._pipelines.plane.source.setInputConnection(model._pipelines.outline.source.getOutputPort());
	WidgetRepresentation_default.connectPipeline(model._pipelines.outline);
	WidgetRepresentation_default.connectPipeline(model._pipelines.plane);
	WidgetRepresentation_default.connectPipeline(model._pipelines.origin);
	WidgetRepresentation_default.connectPipeline(model._pipelines.normal);
	WidgetRepresentation_default.connectPipeline(model._pipelines.display2D);
	publicAPI.addActor(model._pipelines.outline.actor);
	publicAPI.addActor(model._pipelines.plane.actor);
	publicAPI.addActor(model._pipelines.origin.actor);
	publicAPI.addActor(model._pipelines.normal.actor);
	publicAPI.addActor(model._pipelines.display2D.actor);
	publicAPI.requestData = (inData, outData) => {
		const state = inData[0];
		const origin = state.getOrigin();
		if (!origin) return;
		const normal = state.getNormal();
		const bounds = state.getBounds();
		model.plane.setOrigin(origin);
		model.plane.setNormal(normal);
		model._pipelines.outline.source.setCenter((bounds[0] + bounds[1]) * .5, (bounds[2] + bounds[3]) * .5, (bounds[4] + bounds[5]) * .5);
		const xRange = bounds[1] - bounds[0];
		const yRange = bounds[3] - bounds[2];
		const zRange = bounds[5] - bounds[4];
		model._pipelines.outline.source.setXLength(xRange);
		model._pipelines.outline.source.setYLength(yRange);
		model._pipelines.outline.source.setZLength(zRange);
		let pixelScale = 1;
		if (model.scaleInPixels) pixelScale = getPixelWorldHeightAtCoord(origin, model.displayScaleParams);
		model._pipelines.normal.source.set({
			height: Math.max(xRange, yRange, zRange),
			radius: model.handleSizeRatio * Math.min(xRange, yRange, zRange) * model.axisScale * pixelScale,
			resolution: model.sphereResolution
		});
		const yAxis = model._pipelines.normal.source.getOutputData();
		const newAxis = PolyData_default.newInstance();
		newAxis.shallowCopy(yAxis);
		newAxis.getPoints().setData(Float32Array.from(yAxis.getPoints().getData()), 3);
		newAxis.getPointData().removeAllArrays();
		model.matrix.identity().translate(origin[0], origin[1], origin[2]).rotateFromDirections([
			0,
			1,
			0
		], normal).apply(newAxis.getPoints().getData());
		model._pipelines.normal.mapper.setInputData(newAxis);
		model._pipelines.origin.actor.setPosition(origin);
		const handleScale = model.handleSizeRatio * Math.min(xRange, yRange, zRange) * pixelScale;
		model._pipelines.origin.actor.setScale(handleScale, handleScale, handleScale);
		WidgetRepresentation_default.applyStyles(model._pipelines, model.representationStyle, state.getActive() && state.getActiveHandle());
		const output = outData[0]?.initialize() || PolyData_default.newInstance();
		output.shallowCopy(model._pipelines.plane.filter.getOutputData());
		outData[0] = output;
	};
	publicAPI.setSphereResolution = (res) => {
		model.sphereResolution = res;
		return model._pipelines.origin.source.setPhiResolution(res) && model._pipelines.origin.source.setThetaResolution(res);
	};
	publicAPI.setRepresentationStyle = (style) => {
		model.representationStyle = WidgetRepresentation_default.mergeStyles(STYLE_PIPELINE_NAMES, model.representationStyle, style);
		WidgetRepresentation_default.applyStyles(model._pipelines, model.representationStyle);
		publicAPI.modified();
	};
	publicAPI.updateActorVisibility = (renderingType, ctxVisible, hVisible) => {
		const { planeVisible, originVisible, normalVisible, outlineVisible } = model;
		if (renderingType === RenderingTypes.PICKING_BUFFER) {
			model._pipelines.plane.actor.setVisibility(planeVisible);
			model._pipelines.origin.actor.setVisibility(originVisible);
			model._pipelines.normal.actor.setVisibility(normalVisible);
			model._pipelines.plane.actor.getProperty().setOpacity(1);
		} else {
			model._pipelines.outline.actor.setVisibility(outlineVisible && ctxVisible);
			model._pipelines.plane.actor.setVisibility(planeVisible && hVisible);
			model._pipelines.origin.actor.setVisibility(originVisible && hVisible);
			model._pipelines.normal.actor.setVisibility(normalVisible && hVisible);
			const state = model.inputData[0];
			if (state) WidgetRepresentation_default.applyStyles(model._pipelines, model.representationStyle, state.getActive() && state.getActiveHandle());
		}
	};
	publicAPI.getSelectedState = (prop, compositeID) => {
		const state = model.inputData[0];
		state.setActiveHandle(prop);
		switch (prop) {
			case model._pipelines.plane.actor:
				state.setUpdateMethodName("updateFromPlane");
				break;
			case model._pipelines.origin.actor:
				state.setUpdateMethodName("updateFromOrigin");
				break;
			case model._pipelines.normal.actor:
				state.setUpdateMethodName("updateFromNormal");
				break;
			default:
				state.setUpdateMethodName("updateFromPlane");
				break;
		}
		return state;
	};
	publicAPI.setRepresentationStyle(STYLE_DEFAULT);
	publicAPI.setSphereResolution(model.sphereResolution);
}
var DEFAULT_VALUES = {
	sphereResolution: 24,
	handleSizeRatio: .05,
	axisScale: .1,
	normalVisible: true,
	originVisible: true,
	planeVisible: true,
	outlineVisible: true
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	WidgetRepresentation_default.extend(publicAPI, model, initialValues);
	macros_default.get(publicAPI, model, ["sphereResolution", "representationStyle"]);
	macros_default.setGet(publicAPI, model, [
		"handleSizeRatio",
		"axisScale",
		"normalVisible",
		"originVisible",
		"planeVisible",
		"outlineVisible"
	]);
	vtkImplicitPlaneRepresentation(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkImplicitPlaneRepresentation");
var ImplicitPlaneRepresentation_default = {
	newInstance,
	extend,
	generateState
};
//#endregion
export { ImplicitPlaneRepresentation_default as default, extend, newInstance };
