// @kitware/vtk.js/Widgets/Representations/PolyLineRepresentation.js
import macros_default from "../../macros.js";
import { areEquals } from "../../Common/Core/Math.js";
import BoundingBox_default from "../../Common/DataModel/BoundingBox.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
import TubeFilter_default from "../../Filters/General/TubeFilter.js";
import Actor_default from "../../Rendering/Core/Actor.js";
import Mapper_default from "../../Rendering/Core/Mapper.js";
import { RenderingTypes } from "../Core/WidgetManager/Constants.js";
import { getPixelWorldHeightAtCoord } from "../Core/WidgetManager.js";
import WidgetRepresentation_default, { allocateArray } from "./WidgetRepresentation.js";
//#region Sources/Widgets/Representations/PolyLineRepresentation/index.js
function vtkPolyLineRepresentation(publicAPI, model) {
	model.classHierarchy.push("vtkPolyLineRepresentation");
	const superClass = { ...publicAPI };
	const internalPolyData = PolyData_default.newInstance({ mtime: 0 });
	function allocateSize(polyData, size, closePolyLine = false) {
		let points = null;
		if (size < 2) {
			points = allocateArray(polyData, "points", 1).getData();
			points.set([
				0,
				0,
				0
			]);
			allocateArray(polyData, "lines", 0).getData();
		} else if (!polyData.getPoints() || polyData.getPoints().length !== size * 3) {
			points = allocateArray(polyData, "points", size).getData();
			const cellSize = size + (closePolyLine ? 1 : 0);
			if (polyData.getLines().getNumberOfCells() !== 1 || polyData.getLines().getCellSizes()[0] !== cellSize) {
				const lines = allocateArray(polyData, "lines", cellSize + 1);
				const cellData = lines.getData();
				cellData[0] = cellSize;
				for (let i = 1; i <= cellSize; i++) cellData[i] = i - 1;
				if (closePolyLine) cellData[cellSize] = 0;
				lines.setData(cellData);
			}
		}
		return points;
	}
	/**
	* Change the line/tube thickness.
	* @param {number} lineThickness
	*/
	function applyLineThickness(lineThickness) {
		let scaledLineThickness = lineThickness;
		if (publicAPI.getScaleInPixels() && internalPolyData) {
			const center = BoundingBox_default.getCenter(internalPolyData.getBounds());
			scaledLineThickness *= getPixelWorldHeightAtCoord(center, model.displayScaleParams);
		}
		model._pipelines.tubes.filter.setRadius(scaledLineThickness);
	}
	model._pipelines = { tubes: {
		source: publicAPI,
		filter: TubeFilter_default.newInstance({
			radius: model.lineThickness,
			numberOfSides: 12,
			capping: false
		}),
		mapper: Mapper_default.newInstance(),
		actor: Actor_default.newInstance({ parentProp: publicAPI })
	} };
	WidgetRepresentation_default.connectPipeline(model._pipelines.tubes);
	publicAPI.addActor(model._pipelines.tubes.actor);
	publicAPI.requestData = (inData, outData) => {
		const state = inData[0];
		outData[0] = internalPolyData;
		const list = publicAPI.getRepresentationStates(state).reduce((subStates, subState) => {
			const subStateOrigin = subState.getOrigin && subState.getOrigin() ? subState.getOrigin() : null;
			const previousSubStateOrigin = subStates.length && subStates[subStates.length - 1].getOrigin();
			if (!subStateOrigin || previousSubStateOrigin && areEquals(subStateOrigin, previousSubStateOrigin)) return subStates;
			subStates.push(subState);
			return subStates;
		}, []);
		const size = list.length;
		const points = allocateSize(outData[0], size, model.closePolyLine && size > 2);
		if (points) for (let i = 0; i < size; i++) {
			const coords = list[i].getOrigin();
			points[i * 3] = coords[0];
			points[i * 3 + 1] = coords[1];
			points[i * 3 + 2] = coords[2];
		}
		outData[0].getPoints().modified();
		outData[0].modified();
		applyLineThickness(state.getLineThickness?.() ?? model.lineThickness);
	};
	/**
	* When mousing over the line, if behavior != CONTEXT,
	* returns the parent state.
	* @param {object} prop
	* @param {number} compositeID
	* @returns {object}
	*/
	publicAPI.getSelectedState = (prop, compositeID) => model.inputData[0];
	publicAPI.updateActorVisibility = (renderingType, ctxVisible, hVisible) => {
		let lineThickness = model.inputData[0].getLineThickness?.() ?? model.lineThickness;
		if (renderingType === RenderingTypes.PICKING_BUFFER) lineThickness = Math.max(4, lineThickness);
		applyLineThickness(lineThickness);
		return superClass.updateActorVisibility(renderingType, ctxVisible, hVisible);
	};
}
var DEFAULT_VALUES = {
	threshold: Number.EPSILON,
	closePolyLine: false,
	lineThickness: 2,
	scaleInPixels: true
};
function extend(publicAPI, model, initialValues = {}) {
	const newDefault = {
		...DEFAULT_VALUES,
		...initialValues
	};
	WidgetRepresentation_default.extend(publicAPI, model, newDefault);
	macros_default.setGet(publicAPI, model, [
		"threshold",
		"closePolyLine",
		"lineThickness"
	]);
	vtkPolyLineRepresentation(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkPolyLineRepresentation");
var PolyLineRepresentation_default = {
	newInstance,
	extend
};
//#endregion
export { PolyLineRepresentation_default as default, extend, newInstance };
