// @kitware/vtk.js/Widgets/Representations/ScaleTransformHandleRepresentation.js
import macros_default from "../../macros.js";
import CubeSource_default from "../../Filters/Sources/CubeSource.js";
import GlyphRepresentation_default from "./GlyphRepresentation.js";
import TransformHandleSource_default from "./TranslateTransformHandleRepresentation/TransformHandleSource.js";
//#region Sources/Widgets/Representations/ScaleTransformHandleRepresentation/index.js
function vtkScaleTransformHandleRepresentation(publicAPI, model) {
	model.classHierarchy.push("vtkScaleTransformHandleRepresentation");
}
function defaultValues(initialValues) {
	const source = TransformHandleSource_default.newInstance({
		height: initialValues.height ?? 1,
		radius: initialValues.radius ?? 1,
		resolution: initialValues.glyphResolution ?? 12,
		direction: [
			0,
			0,
			1
		]
	});
	const cube1 = CubeSource_default.newInstance(initialValues.cubeSource);
	const cube2 = CubeSource_default.newInstance(initialValues.cubeSource);
	source.addInputConnection(cube1.getOutputPort());
	source.addInputConnection(cube2.getOutputPort());
	return {
		_pipeline: { glyph: source },
		...initialValues
	};
}
function extend(publicAPI, model, initialValues = {}) {
	GlyphRepresentation_default.extend(publicAPI, model, defaultValues(initialValues));
	vtkScaleTransformHandleRepresentation(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkScaleTransformHandleRepresentation");
var ScaleTransformHandleRepresentation_default = {
	newInstance,
	extend
};
//#endregion
export { ScaleTransformHandleRepresentation_default as default, extend, newInstance };
