// @kitware/vtk.js/Widgets/Representations/SphereContextRepresentation.js
import macros_default from "../../macros.js";
import SphereSource_default from "../../Filters/Sources/SphereSource.js";
import { Behavior } from "./WidgetRepresentation/Constants.js";
import GlyphRepresentation_default from "./GlyphRepresentation.js";
//#region Sources/Widgets/Representations/SphereContextRepresentation/index.js
function vtkSphereContextRepresentation(publicAPI, model) {
	model.classHierarchy.push("vtkSphereContextRepresentation");
	publicAPI.setGlyphResolution = macros_default.chain(publicAPI.setGlyphResolution, model._pipeline.glyph.setThetaResolution, model._pipeline.glyph.setPhiResolution);
	publicAPI.setDrawBorder = (draw) => {
		model._pipeline.glyph.setLines(draw);
	};
	publicAPI.setDrawFace = (draw) => {
		model._pipeline.glyph.setFace(draw);
	};
	publicAPI.setOpacity = (opacity) => {
		model._pipeline.actor.getProperty().setOpacity(opacity);
	};
	model._pipeline.actor.getProperty().setOpacity(.2);
}
function defaultValues(initialValues) {
	return {
		glyphResolution: 32,
		drawBorder: false,
		drawFace: true,
		behavior: Behavior.CONTEXT,
		_pipeline: { glyph: SphereSource_default.newInstance({
			phiResolution: initialValues.glyphResolution ?? 32,
			thetaResolution: initialValues.glyphResolution ?? 32
		}) },
		...initialValues
	};
}
function extend(publicAPI, model, initialValues = {}) {
	GlyphRepresentation_default.extend(publicAPI, model, defaultValues(initialValues));
	macros_default.setGet(publicAPI, model, ["glyphResolution"]);
	vtkSphereContextRepresentation(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkSphereContextRepresentation");
var SphereContextRepresentation_default = {
	newInstance,
	extend
};
//#endregion
export { SphereContextRepresentation_default as default, extend, newInstance };
