// @kitware/vtk.js/Widgets/Representations/SplineContextRepresentation.js
import macros_default from "../../macros.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
import Spline3D_default from "../../Common/DataModel/Spline3D.js";
import LineFilter_default from "../../Filters/General/LineFilter.js";
import TriangleFilter_default from "../../Filters/General/TriangleFilter.js";
import Actor_default from "../../Rendering/Core/Actor.js";
import Mapper_default from "../../Rendering/Core/Mapper.js";
import WidgetRepresentation_default, { allocateArray } from "./WidgetRepresentation.js";
import ContextRepresentation_default from "./ContextRepresentation.js";
//#region Sources/Widgets/Representations/SplineContextRepresentation/index.js
function vtkSplineContextRepresentation(publicAPI, model) {
	model.classHierarchy.push("vtkSplineContextRepresentation");
	model.internalPolyData = PolyData_default.newInstance({ mtime: 0 });
	model._pipelines = {
		area: {
			source: publicAPI,
			filter: TriangleFilter_default.newInstance(),
			mapper: Mapper_default.newInstance(),
			actor: Actor_default.newInstance({ parentProp: publicAPI })
		},
		border: {
			source: publicAPI,
			filter: LineFilter_default.newInstance(),
			mapper: Mapper_default.newInstance(),
			actor: Actor_default.newInstance({ parentProp: publicAPI })
		}
	};
	WidgetRepresentation_default.connectPipeline(model._pipelines.area);
	model._pipelines.area.actor.getProperty().setOpacity(.2);
	model._pipelines.area.actor.getProperty().setColor(0, 1, 0);
	publicAPI.addActor(model._pipelines.area.actor);
	WidgetRepresentation_default.connectPipeline(model._pipelines.border);
	model._pipelines.border.actor.getProperty().setOpacity(1);
	model._pipelines.border.actor.getProperty().setColor(.1, 1, .1);
	model._pipelines.border.actor.setVisibility(model.outputBorder);
	publicAPI.addActor(model._pipelines.border.actor);
	const superGetRepresentationStates = publicAPI.getRepresentationStates;
	publicAPI.getRepresentationStates = (input = model.inputData[0]) => superGetRepresentationStates(input).filter((state) => state.getOrigin?.() && state.isVisible?.());
	publicAPI.requestData = (inData, outData) => {
		const widgetState = inData[0];
		outData[0] = model.internalPolyData;
		const closed = widgetState.getSplineClosed();
		const inPoints = publicAPI.getRepresentationStates(widgetState).map((state) => state.getOrigin());
		if (inPoints.length <= 1) return;
		let numVertices = inPoints.length;
		if (!closed) --numVertices;
		else inPoints.push(inPoints[0]);
		const spline = Spline3D_default.newInstance({
			close: widgetState.getSplineClosed(),
			kind: widgetState.getSplineKind(),
			tension: widgetState.getSplineTension(),
			bias: widgetState.getSplineBias(),
			continuity: widgetState.getSplineContinuity(),
			boundaryCondition: widgetState.getSplineBoundaryCondition(),
			boundaryConditionValues: widgetState.getSplineBoundaryConditionValues()
		});
		spline.computeCoefficients(inPoints);
		const outPoints = allocateArray(model.internalPolyData, "points", (numVertices + !closed) * model.resolution).getData();
		const outCells = new Uint32Array(numVertices * model.resolution + 2);
		outCells[0] = numVertices * model.resolution + 1;
		outCells[numVertices * model.resolution + 1] = 0;
		for (let i = 0; i < numVertices; i++) for (let j = 0; j < model.resolution; j++) {
			const t = j / model.resolution;
			const point = spline.getPoint(i, t);
			outPoints[3 * (i * model.resolution + j) + 0] = point[0];
			outPoints[3 * (i * model.resolution + j) + 1] = point[1];
			outPoints[3 * (i * model.resolution + j) + 2] = point[2];
			outCells[i * model.resolution + j + 1] = i * model.resolution + j;
		}
		if (closed) outCells[numVertices * model.resolution + 1] = 0;
		else {
			const lastPointIndex = numVertices * model.resolution + 1;
			const lastPoint = spline.getPoint(numVertices, 0);
			outPoints[3 * lastPointIndex + 0] = lastPoint[0];
			outPoints[3 * lastPointIndex + 1] = lastPoint[1];
			outPoints[3 * lastPointIndex + 2] = lastPoint[2];
			outCells[numVertices * model.resolution + 1] = lastPointIndex;
		}
		if (model.fill) model.internalPolyData.getPolys().setData(outCells);
		model.internalPolyData.getLines().setData(model.outputBorder ? outCells : []);
		model.internalPolyData.modified();
		model._pipelines.area.filter.update();
		model._pipelines.border.actor.getProperty().setColor(...inPoints.length <= 3 || model._pipelines.area.filter.getErrorCount() === 0 ? model.borderColor : model.errorBorderColor);
	};
	publicAPI.getSelectedState = (prop, compositeID) => model.state;
	const superSetFill = publicAPI.setFill;
	publicAPI.setFill = (fill) => {
		const res = superSetFill(fill);
		model._pipelines.area.actor.setVisibility(fill);
		return res;
	};
	const superSetOutputBorder = publicAPI.setOutputBorder;
	publicAPI.setOutputBorder = (v) => {
		const res = superSetOutputBorder(v);
		model._pipelines.border.actor.setVisibility(v);
		return res;
	};
}
var DEFAULT_VALUES = {
	resolution: 16,
	fill: true,
	outputBorder: false,
	borderColor: [
		.1,
		1,
		.1
	],
	errorBorderColor: [
		1,
		0,
		0
	]
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	ContextRepresentation_default.extend(publicAPI, model, initialValues);
	macros_default.get(publicAPI, model, ["mapper"]);
	macros_default.setGet(publicAPI, model, [
		"resolution",
		"boundaryCondition",
		"fill",
		"outputBorder"
	]);
	macros_default.setGetArray(publicAPI, model, ["borderColor", "errorBorderColor"], 3);
	vtkSplineContextRepresentation(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkSplineContextRepresentation");
var SplineContextRepresentation_default = {
	newInstance,
	extend
};
//#endregion
export { SplineContextRepresentation_default as default, extend, newInstance };
