// @kitware/vtk.js/Widgets/Representations/TranslateTransformHandleRepresentation.js
import macros_default from "../../macros.js";
import ConeSource_default from "../../Filters/Sources/ConeSource.js";
import GlyphRepresentation_default from "./GlyphRepresentation.js";
import TransformHandleSource_default from "./TranslateTransformHandleRepresentation/TransformHandleSource.js";
//#region Sources/Widgets/Representations/TranslateTransformHandleRepresentation/index.js
function vtkTranslateTransformHandleRepresentation(publicAPI, model) {
	model.classHierarchy.push("vtkTranslateTransformHandleRepresentation");
}
function defaultValues(initialValues) {
	const source = TransformHandleSource_default.newInstance({
		height: initialValues.height ?? 1,
		radius: initialValues.radius ?? 1,
		resolution: initialValues.glyphResolution ?? 12,
		direction: [
			0,
			0,
			1
		]
	});
	const cone1 = ConeSource_default.newInstance(initialValues.coneSource);
	const cone2 = ConeSource_default.newInstance(initialValues.coneSource);
	source.addInputConnection(cone1.getOutputPort());
	source.addInputConnection(cone2.getOutputPort());
	return {
		_pipeline: { glyph: source },
		...initialValues
	};
}
function extend(publicAPI, model, initialValues = {}) {
	GlyphRepresentation_default.extend(publicAPI, model, defaultValues(initialValues));
	vtkTranslateTransformHandleRepresentation(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkTranslateTransformHandleRepresentation");
var TranslateTransformHandleRepresentation_default = {
	newInstance,
	extend
};
//#endregion
export { TranslateTransformHandleRepresentation_default as default, extend, newInstance };
