// @kitware/vtk.js/Widgets/Representations/WidgetRepresentation.js
import macros_default from "../../macros.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import CellArray_default from "../../Common/Core/CellArray.js";
import Points_default from "../../Common/Core/Points.js";
import { POLYDATA_FIELDS } from "../../Common/DataModel/PolyData/Constants.js";
import Prop_default from "../../Rendering/Core/Prop.js";
import { CATEGORIES } from "../../Rendering/Core/Mapper/CoincidentTopologyHelper.js";
import { RenderingTypes } from "../Core/WidgetManager/Constants.js";
import { Behavior } from "./WidgetRepresentation/Constants.js";
//#region Sources/Widgets/Representations/WidgetRepresentation/index.js
var { vtkErrorMacro, vtkWarningMacro } = macros_default;
var STYLE_CATEGORIES = [
	"active",
	"inactive",
	"static"
];
function applyCoincidentTopologyParametersToMapper(mapper, parameters) {
	if (mapper && mapper.setResolveCoincidentTopologyToPolygonOffset) {
		mapper.setResolveCoincidentTopologyToPolygonOffset();
		CATEGORIES.forEach((category) => {
			if (parameters[category]) {
				const methodName = `setRelativeCoincidentTopology${category}OffsetParameters`;
				if (mapper[methodName]) {
					const { factor, offset } = parameters[category];
					mapper[methodName](factor, offset);
				}
			}
		});
	}
}
function mergeStyles(elementNames, ...stylesToMerge) {
	const newStyleObject = {
		active: {},
		inactive: {},
		static: {}
	};
	STYLE_CATEGORIES.forEach((category) => {
		const cat = newStyleObject[category];
		elementNames.forEach((name) => {
			if (!cat[name]) cat[name] = {};
			stylesToMerge.filter((s) => s && s[category] && s[category][name]).forEach((s) => Object.assign(cat[name], s[category][name]));
		});
	});
	return newStyleObject;
}
function applyStyles(pipelines, styles, activeActor) {
	if (!activeActor) {
		Object.keys(styles.static).forEach((name) => {
			if (pipelines[name]) pipelines[name].actor.getProperty().set(styles.static[name]);
		});
		Object.keys(styles.inactive).forEach((name) => {
			if (pipelines[name]) pipelines[name].actor.getProperty().set(styles.inactive[name]);
		});
	} else Object.keys(pipelines).forEach((name) => {
		const style = pipelines[name].actor === activeActor ? styles.active[name] : styles.inactive[name];
		if (style) pipelines[name].actor.getProperty().set(style);
	});
}
function connectPipeline(pipeline) {
	let source = pipeline.source;
	if (pipeline.filter) {
		if (source.isA("vtkDataSet")) pipeline.filter.setInputData(source);
		else pipeline.filter.setInputConnection(source.getOutputPort());
		source = pipeline.filter;
	}
	if (source) if (source.isA("vtkDataSet")) pipeline.mapper.setInputData(source);
	else pipeline.mapper.setInputConnection(source.getOutputPort());
	if (pipeline.glyph) pipeline.mapper.setInputConnection(pipeline.glyph.getOutputPort(), 1);
	pipeline.actor.setMapper(pipeline.mapper);
}
function allocateArray(polyData, name, numberOfTuples, dataType, numberOfComponents) {
	let dataArray = polyData[`get${macros_default.capitalize(name)}`]?.() || polyData.getPointData().getArrayByName(name);
	if (!dataArray || dataType !== void 0 && dataArray.getDataType() !== dataType || numberOfComponents !== void 0 && dataArray.getNumberOfComponents() !== numberOfComponents) {
		let arrayType = DataArray_default;
		let arrayDataType = dataType;
		let arrayNumberOfComponents = numberOfComponents;
		if (name === "points") {
			arrayType = Points_default;
			arrayDataType = arrayDataType ?? "Float32Array";
			arrayNumberOfComponents = numberOfComponents ?? 3;
		} else if (POLYDATA_FIELDS.includes(name)) {
			arrayType = CellArray_default;
			arrayDataType = arrayDataType ?? "Uint16Array";
			arrayNumberOfComponents = numberOfComponents ?? 1;
		} else {
			arrayDataType = arrayDataType ?? "Float32Array";
			arrayNumberOfComponents = numberOfComponents ?? 1;
		}
		dataArray = arrayType.newInstance({
			name,
			dataType: arrayDataType,
			numberOfComponents: arrayNumberOfComponents,
			size: arrayNumberOfComponents * numberOfTuples,
			empty: numberOfTuples === 0
		});
		if (name === "points" || POLYDATA_FIELDS.includes(name)) polyData[`set${macros_default.capitalize(name)}`](dataArray);
		else polyData.getPointData().addArray(dataArray);
	} else if (dataArray.getNumberOfTuples() !== numberOfTuples) dataArray.resize(numberOfTuples);
	return dataArray;
}
function vtkWidgetRepresentation(publicAPI, model) {
	model.classHierarchy.push("vtkWidgetRepresentation");
	const cache = {
		mtimes: {},
		states: []
	};
	model._onCoincidentTopologyParametersChanged = () => {
		publicAPI.getActors().forEach((actor) => {
			applyCoincidentTopologyParametersToMapper(actor.getMapper(), model.coincidentTopologyParameters);
		});
	};
	publicAPI.getActors = () => model.actors;
	publicAPI.getNestedProps = publicAPI.getActors;
	publicAPI.setLabels = (...labels) => {
		if (labels.length === 1) model.labels = [].concat(labels[0]);
		else model.labels = labels;
		publicAPI.modified();
	};
	publicAPI.getRepresentationStates = (input = model.inputData[0]) => {
		if (cache.mtimes.representation === publicAPI.getMTime() && cache.mtimes.input === input.getMTime()) return cache.states;
		cache.mtimes.representation = publicAPI.getMTime();
		cache.mtimes.input = input.getMTime();
		cache.states = [];
		model.labels.forEach((name) => {
			cache.states = cache.states.concat(input.getStatesWithLabel(name) || []);
		});
		return cache.states;
	};
	publicAPI.getSelectedState = (prop, compositeID) => {
		const representationStates = publicAPI.getRepresentationStates();
		if (compositeID < representationStates.length) return representationStates[compositeID];
		vtkErrorMacro(`Representation ${publicAPI.getClassName()} should implement getSelectedState(prop, compositeID) method.`);
		return null;
	};
	publicAPI.updateActorVisibility = (renderingType = RenderingTypes.FRONT_BUFFER, ctxVisible = true, handleVisible = true) => {
		let otherFlag = true;
		switch (model.behavior) {
			case Behavior.HANDLE:
				otherFlag = renderingType === RenderingTypes.PICKING_BUFFER || handleVisible;
				break;
			case Behavior.CONTEXT:
				otherFlag = ctxVisible;
				break;
			default:
				otherFlag = true;
				break;
		}
		const visibilityFlag = otherFlag;
		for (let i = 0; i < model.actors.length; i++) if (model.visibilityFlagArray) model.actors[i].setVisibility(visibilityFlag && model.visibilityFlagArray[i]);
		else model.actors[i].setVisibility(visibilityFlag);
		if (model.alwaysVisibleActors) for (let i = 0; i < model.alwaysVisibleActors.length; i++) model.alwaysVisibleActors[i].setVisibility(true);
	};
	model.actors.push = (...args) => {
		vtkWarningMacro("You should use publicAPI.addActor() to initialize the actor properly");
		args.forEach((actor) => publicAPI.addActor(actor));
	};
	publicAPI.addActor = (actor) => {
		applyCoincidentTopologyParametersToMapper(actor.getMapper(), model.coincidentTopologyParameters);
		Array.prototype.push.apply(model.actors, [actor]);
	};
	publicAPI.setLabels(model.labels);
}
function defaultValues(initialValues) {
	return {
		activeScaleFactor: 1.2,
		activeColor: 1,
		useActiveColor: true,
		actors: [],
		labels: [],
		behavior: Behavior.CONTEXT,
		coincidentTopologyParameters: {
			Point: {
				factor: -1,
				offset: -1
			},
			Line: {
				factor: -1,
				offset: -1
			},
			Polygon: {
				factor: -1,
				offset: -1
			}
		},
		scaleInPixels: false,
		displayScaleParams: {
			dispHeightFactor: 1,
			cameraPosition: [
				0,
				0,
				0
			],
			cameraDir: [
				1,
				0,
				0
			],
			isParallel: false,
			rendererPixelDims: [1, 1]
		},
		_internalArrays: {},
		...initialValues
	};
}
function extend(publicAPI, model, initialValues = {}) {
	Prop_default.extend(publicAPI, model, defaultValues(initialValues));
	macros_default.algo(publicAPI, model, 1, 1);
	macros_default.get(publicAPI, model, [
		"labels",
		"displayScaleParams",
		"coincidentTopologyParameters"
	]);
	macros_default.set(publicAPI, model, [{
		type: "object",
		name: "displayScaleParams"
	}, {
		type: "object",
		name: "coincidentTopologyParameters"
	}]);
	macros_default.setGet(publicAPI, model, [
		"scaleInPixels",
		"activeScaleFactor",
		"activeColor",
		"useActiveColor"
	]);
	vtkWidgetRepresentation(publicAPI, model);
}
var WidgetRepresentation_default = {
	extend,
	mergeStyles,
	applyStyles,
	connectPipeline
};
//#endregion
export { allocateArray, applyStyles, connectPipeline, WidgetRepresentation_default as default, extend, mergeStyles };
