// @kitware/vtk.js/Widgets/Widgets3D/LineWidget/state.js
import StateBuilder_default from "../../Core/StateBuilder.js";
//#region Sources/Widgets/Widgets3D/LineWidget/state.js
var linePosState = StateBuilder_default.createBuilder().addField({
	name: "posOnLine",
	initialValue: .5
}).build();
function generateState() {
	return StateBuilder_default.createBuilder().addStateFromMixin({
		labels: ["moveHandle"],
		mixins: [
			"origin",
			"color",
			"scale1",
			"visible",
			"manipulator",
			"shape"
		],
		name: "moveHandle",
		initialValues: {
			scale1: 30,
			visible: true
		}
	}).addStateFromMixin({
		labels: ["handle1"],
		mixins: [
			"origin",
			"color",
			"scale1",
			"visible",
			"manipulator",
			"shape"
		],
		name: "handle1",
		initialValues: { scale1: 30 }
	}).addStateFromMixin({
		labels: ["handle2"],
		mixins: [
			"origin",
			"color",
			"scale1",
			"visible",
			"manipulator",
			"shape"
		],
		name: "handle2",
		initialValues: { scale1: 30 }
	}).addStateFromMixin({
		labels: ["SVGtext"],
		mixins: [
			"origin",
			"color",
			"text",
			"visible"
		],
		name: "text",
		initialValues: { text: "" }
	}).addStateFromInstance({
		name: "positionOnLine",
		instance: linePosState
	}).addField({ name: "lineThickness" }).build();
}
//#endregion
export { generateState as default };
