// @kitware/vtk.js/Widgets/Representations/TranslateTransformHandleRepresentation/TransformHandleSource.js
import macros_default from "../../../macros.js";
import MatrixBuilder_default from "../../../Common/Core/MatrixBuilder.js";
import AppendPolyData_default from "../../../Filters/General/AppendPolyData.js";
import CylinderSource_default from "../../../Filters/Sources/CylinderSource.js";
//#region Sources/Widgets/Representations/TranslateTransformHandleRepresentation/TransformHandleSource.js
function rotatePolyData(pd, direction) {
	const points = pd.getPoints().getData();
	MatrixBuilder_default.buildFromRadian().rotateFromDirections([
		0,
		1,
		0
	], direction).apply(points);
	pd.getPoints().modified();
	pd.modified();
}
function translatePolyData(pd, translation) {
	const points = pd.getPoints().getData();
	MatrixBuilder_default.buildFromRadian().translate(...translation).apply(points);
	pd.modified();
}
function vtkTransformHandleSource(publicAPI, model) {
	model.classHierarchy.push("vtkTransformHandleSource");
	function requestData(inData, outData) {
		const cylinderSource = CylinderSource_default.newInstance({
			height: model.height,
			initAngle: model.initAngle,
			radius: model.radius,
			resolution: model.resolution,
			capping: model.capping,
			pointType: model.pointType,
			center: [
				0,
				0,
				0
			],
			direction: [
				0,
				1,
				0
			]
		});
		const appendFilter = AppendPolyData_default.newInstance();
		appendFilter.setInputConnection(cylinderSource.getOutputPort(), 0);
		if (inData[0]) {
			translatePolyData(inData[0], [
				0,
				model.height / 2,
				0
			]);
			appendFilter.addInputData(inData[0]);
		}
		if (inData[1]) {
			rotatePolyData(inData[1], [
				0,
				-1,
				0
			]);
			translatePolyData(inData[1], [
				0,
				-model.height / 2,
				0
			]);
			appendFilter.addInputData(inData[1]);
		}
		const poly = appendFilter.getOutputData();
		const points = poly.getPoints().getData();
		MatrixBuilder_default.buildFromRadian().translate(...model.center).rotateFromDirections([
			0,
			1,
			0
		], model.direction).translate(...model.center.map((c) => c * -1)).apply(points);
		outData[0] = poly;
	}
	publicAPI.requestData = requestData;
}
var DEFAULT_VALUES = { capPolyData: null };
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	CylinderSource_default.extend(publicAPI, model, initialValues);
	macros_default.algo(publicAPI, model, 2, 1);
	vtkTransformHandleSource(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkTransformHandleSource");
var TransformHandleSource_default = {
	newInstance,
	extend
};
//#endregion
export { TransformHandleSource_default as default, extend, newInstance };
