// @kitware/vtk.js/Filters/Core/Cutter.js
import { algo, macros_exports, newInstance as newInstance$1, newTypedArrayFrom, obj, setGet } from "../../macros.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
//#region Sources/Filters/Core/Cutter/index.js
var { vtkErrorMacro } = macros_exports;
function initPolyIterator(pd) {
	const polys = pd.getPolys().getData();
	const strips = pd.getStrips().getData();
	const it = {
		cellSize: 0,
		cell: [],
		done: false,
		polyIdx: 0,
		stripIdx: 0,
		remainingStripLength: 0,
		next() {
			if (it.polyIdx < polys.length) {
				it.cellSize = polys[it.polyIdx];
				const start = it.polyIdx + 1;
				const end = start + it.cellSize;
				it.polyIdx = end;
				let p = 0;
				for (let i = start; i < end; ++i) it.cell[p++] = polys[i];
			} else if (it.stripIdx < strips.length) {
				it.cellSize = 3;
				if (it.remainingStripLength === 0) {
					it.remainingStripLength = strips[it.stripIdx] - 2;
					it.stripIdx += 3;
				}
				const start = it.stripIdx - 2;
				const end = it.stripIdx + 1;
				it.stripIdx++;
				it.remainingStripLength--;
				let p = 0;
				for (let i = start; i < end; ++i) it.cell[p++] = strips[i];
			} else if (!it.done) it.done = true;
			else throw new Error("Iterator is done");
		}
	};
	it.next();
	return it;
}
function vtkCutter(publicAPI, model) {
	model.classHierarchy.push("vtkCutter");
	const superClass = { ...publicAPI };
	publicAPI.getMTime = () => {
		let mTime = superClass.getMTime();
		if (!model.cutFunction) return mTime;
		mTime = Math.max(mTime, model.cutFunction.getMTime());
		return mTime;
	};
	function dataSetCutter(input, output) {
		const points = input.getPoints();
		const pointsData = points.getData();
		const pointData = input.getPointData();
		const numPts = points.getNumberOfPoints();
		const newPointsData = [];
		const newLinesData = [];
		const newPolysData = [];
		const newPointData = {};
		const numberOfArrays = pointData.getNumberOfArrays();
		for (let arrayIdx = 0; arrayIdx < numberOfArrays; arrayIdx++) newPointData[pointData.getArrayName(arrayIdx)] = [];
		if (!model.cutScalars || model.cutScalars.length < numPts) model.cutScalars = new Float32Array(numPts);
		let inOffset = 0;
		let outOffset = 0;
		while (inOffset < pointsData.length) model.cutScalars[outOffset++] = model.cutFunction.evaluateFunction(pointsData[inOffset++], pointsData[inOffset++], pointsData[inOffset++]);
		const crossedEdges = [];
		const x1 = new Array(3);
		const x2 = new Array(3);
		const cellPointsScalars = [];
		for (const it = initPolyIterator(input); !it.done; it.next()) {
			if (it.cellSize <= 2) continue;
			for (let i = 0; i < it.cellSize;) cellPointsScalars[i] = model.cutScalars[it.cell[i++]];
			const sideFirstPoint = cellPointsScalars[0] > 0;
			let allPointsSameSide = true;
			for (let i = 1; i < it.cell.length; i++) if (cellPointsScalars[i] > 0 !== sideFirstPoint) {
				allPointsSameSide = false;
				break;
			}
			if (allPointsSameSide) continue;
			const intersectedEdgesList = [];
			for (let i = 0; i < it.cellSize; i++) {
				const idNext = i + 1 === it.cellSize ? 0 : i + 1;
				const signPoint0 = cellPointsScalars[i] > 0;
				if (cellPointsScalars[idNext] > 0 === signPoint0) continue;
				let e1 = i;
				let e2 = idNext;
				let deltaScalar = cellPointsScalars[e2] - cellPointsScalars[e1];
				if (deltaScalar <= 0) {
					e1 = idNext;
					e2 = i;
					deltaScalar *= -1;
				}
				let t = 0;
				if (deltaScalar !== 0) t = (model.cutValue - cellPointsScalars[e1]) / deltaScalar;
				const pointID1 = it.cell[e1];
				const pointID2 = it.cell[e2];
				x1[0] = pointsData[pointID1 * 3];
				x1[1] = pointsData[pointID1 * 3 + 1];
				x1[2] = pointsData[pointID1 * 3 + 2];
				x2[0] = pointsData[pointID2 * 3];
				x2[1] = pointsData[pointID2 * 3 + 1];
				x2[2] = pointsData[pointID2 * 3 + 2];
				const computedIntersectedPoint = [
					x1[0] + t * (x2[0] - x1[0]),
					x1[1] + t * (x2[1] - x1[1]),
					x1[2] + t * (x2[2] - x1[2])
				];
				const computedIntersectedArrays = {};
				for (let arrayIdx = 0; arrayIdx < numberOfArrays; arrayIdx++) {
					const array = pointData.getArrayByIndex(arrayIdx);
					const name = pointData.getArrayName(arrayIdx);
					const data = array.getData();
					const n = array.getNumberOfComponents();
					const computedIntersectedArray = new Array(n);
					for (let j = 0; j < n; j++) {
						const scalar1 = data[n * pointID1 + j];
						const scalar2 = data[n * pointID2 + j];
						computedIntersectedArray.push(scalar1 + t * (scalar2 - scalar1));
					}
					computedIntersectedArrays[name] = computedIntersectedArray;
				}
				intersectedEdgesList.push({
					pointEdge1: pointID1,
					pointEdge2: pointID2,
					intersectedPoint: computedIntersectedPoint,
					intersectedArrays: computedIntersectedArrays,
					newPointID: -1
				});
			}
			for (let i = 0; i < intersectedEdgesList.length; i++) {
				const intersectedEdge = intersectedEdgesList[i];
				let alreadyAdded = false;
				for (let j = 0; j < crossedEdges.length; j++) {
					const crossedEdge = crossedEdges[j];
					const sameEdge = intersectedEdge.pointEdge1 === crossedEdge.pointEdge1 && intersectedEdge.pointEdge2 === crossedEdge.pointEdge2;
					const samePoint = intersectedEdge.intersectedPoint[0] === crossedEdge.intersectedPoint[0] && intersectedEdge.intersectedPoint[1] === crossedEdge.intersectedPoint[1] && intersectedEdge.intersectedPoint[2] === crossedEdge.intersectedPoint[2];
					if (sameEdge || samePoint) {
						alreadyAdded = true;
						intersectedEdgesList[i].newPointID = crossedEdges[j].newPointID;
						break;
					}
				}
				if (!alreadyAdded) {
					newPointsData.push(intersectedEdge.intersectedPoint[0]);
					newPointsData.push(intersectedEdge.intersectedPoint[1]);
					newPointsData.push(intersectedEdge.intersectedPoint[2]);
					Object.keys(intersectedEdge.intersectedArrays).forEach((name) => {
						newPointData[name].push(...intersectedEdge.intersectedArrays[name]);
					});
					intersectedEdgesList[i].newPointID = newPointsData.length / 3 - 1;
					crossedEdges.push(intersectedEdgesList[i]);
				}
			}
			const cellSize = intersectedEdgesList.length;
			if (cellSize === 2) newLinesData.push(cellSize, intersectedEdgesList[0].newPointID, intersectedEdgesList[1].newPointID);
			else if (cellSize > 2) {
				newPolysData.push(cellSize);
				intersectedEdgesList.forEach((edge) => {
					newPolysData.push(edge.newPointID);
				});
			}
		}
		output.getPoints().setData(newTypedArrayFrom(points.getDataType(), newPointsData), 3);
		const outputPointData = output.getPointData();
		for (let arrayIdx = 0; arrayIdx < numberOfArrays; arrayIdx++) {
			const name = pointData.getArrayName(arrayIdx);
			const array = DataArray_default.newInstance({
				name,
				dataType: pointData.getArrayByIndex(arrayIdx).getDataType(),
				values: newPointData[name],
				numberOfComponents: pointData.getArrayByIndex(arrayIdx).getNumberOfComponents()
			});
			outputPointData.addArray(array);
		}
		if (newLinesData.length !== 0) output.getLines().setData(Uint16Array.from(newLinesData));
		if (newPolysData.length !== 0) output.getPolys().setData(Uint16Array.from(newPolysData));
	}
	publicAPI.requestData = (inData, outData) => {
		const input = inData[0];
		if (!input) {
			vtkErrorMacro("Invalid or missing input");
			return;
		}
		if (!model.cutFunction) {
			vtkErrorMacro("Missing cut function");
			return;
		}
		const output = outData[0]?.initialize() || PolyData_default.newInstance();
		dataSetCutter(input, output);
		outData[0] = output;
	};
}
var DEFAULT_VALUES = {
	cutFunction: null,
	cutScalars: null,
	cutValue: 0
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	obj(publicAPI, model);
	algo(publicAPI, model, 1, 1);
	setGet(publicAPI, model, ["cutFunction", "cutValue"]);
	vtkCutter(publicAPI, model);
}
var newInstance = newInstance$1(extend, "vtkCutter");
var Cutter_default = {
	newInstance,
	extend
};
//#endregion
export { Cutter_default as default, extend, newInstance };
