// @kitware/vtk.js/Filters/Core/PolyDataNormals.js
import macros_default from "../../macros.js";
import Math_default from "../../Common/Core/Math.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import Triangle_default from "../../Common/DataModel/Triangle.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
//#region Sources/Filters/Core/PolyDataNormals/index.js
function vtkPolyDataNormals(publicAPI, model) {
	model.classHierarchy.push("vtkPolyDataNormals");
	publicAPI.vtkPolyDataNormalsExecute = (numberOfPolys, polysData, pointsData) => {
		if (!pointsData) return null;
		const pointNormals = new Float32Array(pointsData.length);
		const cellNormals = new Float32Array(3 * numberOfPolys);
		let cellNormalComponent = 0;
		let numberOfPoints = 0;
		const polysDataLength = polysData.length;
		const cellPointIds = [
			0,
			0,
			0
		];
		const cellNormal = [
			0,
			0,
			0
		];
		for (let c = 0; c < polysDataLength; c += numberOfPoints + 1) {
			numberOfPoints = polysData[c];
			if (numberOfPoints < 3) continue;
			for (let i = 1; i <= 3; ++i) cellPointIds[i - 1] = 3 * polysData[c + i];
			Triangle_default.computeNormal(pointsData.slice(cellPointIds[0], cellPointIds[0] + 3), pointsData.slice(cellPointIds[1], cellPointIds[1] + 3), pointsData.slice(cellPointIds[2], cellPointIds[2] + 3), cellNormal);
			cellNormals[cellNormalComponent++] = cellNormal[0];
			cellNormals[cellNormalComponent++] = cellNormal[1];
			cellNormals[cellNormalComponent++] = cellNormal[2];
			if (model.computePointNormals) for (let i = 1; i <= numberOfPoints; ++i) {
				let pointId = 3 * polysData[c + i];
				pointNormals[pointId] += cellNormal[0];
				pointNormals[++pointId] += cellNormal[1];
				pointNormals[++pointId] += cellNormal[2];
			}
		}
		if (model.computePointNormals) {
			const pointNormal = [
				0,
				0,
				0
			];
			for (let i = 0; i < pointsData.length;) {
				pointNormal[0] = pointNormals[i];
				pointNormal[1] = pointNormals[i + 1];
				pointNormal[2] = pointNormals[i + 2];
				Math_default.normalize(pointNormal);
				pointNormals[i++] = pointNormal[0];
				pointNormals[i++] = pointNormal[1];
				pointNormals[i++] = pointNormal[2];
			}
		}
		return [cellNormals, pointNormals];
	};
	publicAPI.requestData = (inData, outData) => {
		if (!publicAPI.getNumberOfInputPorts()) return;
		const input = inData[0];
		if (!input) return;
		const output = outData[0]?.initialize() || PolyData_default.newInstance();
		output.setPoints(input.getPoints());
		output.setVerts(input.getVerts());
		output.setLines(input.getLines());
		output.setPolys(input.getPolys());
		output.setStrips(input.getStrips());
		output.getPointData().passData(input.getPointData());
		output.getCellData().passData(input.getCellData());
		output.getFieldData().passData(input.getFieldData());
		const [cellNormals, pointNormals] = publicAPI.vtkPolyDataNormalsExecute(input.getNumberOfPolys(), input.getPolys().getData(), input.getPoints().getData());
		if (model.computePointNormals) {
			const outputPointNormals = DataArray_default.newInstance({
				numberOfComponents: 3,
				name: "Normals",
				values: pointNormals
			});
			output.getPointData().setNormals(outputPointNormals);
		}
		if (model.computeCellNormals) {
			const outputCellNormals = DataArray_default.newInstance({
				numberOfComponents: 3,
				name: "Normals",
				values: cellNormals
			});
			output.getCellData().setNormals(outputCellNormals);
		}
		outData[0] = output;
	};
}
function defaultValues(initialValues) {
	return {
		computeCellNormals: false,
		computePointNormals: true,
		...initialValues
	};
}
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, defaultValues(initialValues));
	macros_default.obj(publicAPI, model);
	macros_default.algo(publicAPI, model, 1, 1);
	macros_default.setGet(publicAPI, model, ["computeCellNormals", "computePointNormals"]);
	vtkPolyDataNormals(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkPolyDataNormals");
var PolyDataNormals_default = {
	newInstance,
	extend
};
//#endregion
export { PolyDataNormals_default as default, extend, newInstance };
