// @kitware/vtk.js/Filters/Core/ReverseSense.js
import macros_default from "../../macros.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import CellArray_default from "../../Common/Core/CellArray.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
//#region Sources/Filters/Core/ReverseSense/index.js
var { vtkErrorMacro } = macros_default;
function reverseCellArray(inputCells) {
	if (!inputCells) return null;
	const inputData = inputCells.getData();
	const outputData = inputData.slice();
	for (let offset = 0, len = inputData.length; offset < len;) {
		const cellSize = inputData[offset];
		let left = offset + 1;
		let right = offset + cellSize;
		while (left < right) {
			const l = inputData[left];
			const r = inputData[right];
			outputData[left++] = r;
			outputData[right--] = l;
		}
		offset += cellSize + 1;
	}
	return CellArray_default.newInstance({
		values: outputData,
		numberOfComponents: 1
	});
}
function reverseNormals(inputNormals) {
	if (!inputNormals) return null;
	const values = inputNormals.getData().slice();
	for (let i = 0; i < values.length; i++) {
		const v = -values[i];
		values[i] = v === 0 ? 0 : v;
	}
	return DataArray_default.newInstance({
		name: inputNormals.getName(),
		values,
		numberOfComponents: inputNormals.getNumberOfComponents()
	});
}
function vtkReverseSense(publicAPI, model) {
	model.classHierarchy.push("vtkReverseSense");
	publicAPI.requestData = (inData, outData) => {
		const input = inData[0];
		if (!input) {
			vtkErrorMacro("No input!");
			return;
		}
		const output = outData[0]?.initialize() || PolyData_default.newInstance();
		output.setPoints(input.getPoints());
		output.setVerts(model.reverseCells ? reverseCellArray(input.getVerts()) : input.getVerts());
		output.setLines(model.reverseCells ? reverseCellArray(input.getLines()) : input.getLines());
		output.setPolys(model.reverseCells ? reverseCellArray(input.getPolys()) : input.getPolys());
		output.setStrips(model.reverseCells ? reverseCellArray(input.getStrips()) : input.getStrips());
		const outPointData = output.getPointData();
		const outCellData = output.getCellData();
		outPointData.passData(input.getPointData());
		outCellData.passData(input.getCellData());
		output.getFieldData().passData(input.getFieldData());
		if (model.reverseNormals) {
			const pointNormals = reverseNormals(input.getPointData().getNormals());
			const cellNormals = reverseNormals(input.getCellData().getNormals());
			if (pointNormals) outPointData.setNormals(pointNormals);
			if (cellNormals) outCellData.setNormals(cellNormals);
		}
		outData[0] = output;
	};
}
var DEFAULT_VALUES = {
	reverseCells: true,
	reverseNormals: false
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.algo(publicAPI, model, 1, 1);
	macros_default.setGet(publicAPI, model, ["reverseCells", "reverseNormals"]);
	vtkReverseSense(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkReverseSense");
var ReverseSense_default = {
	newInstance,
	extend
};
//#endregion
export { ReverseSense_default as default, extend, newInstance };
