// @kitware/vtk.js/Common/DataModel/Cylinder.js
import macros_default from "../../macros.js";
import { dot } from "../Core/Math.js";
import ImplicitFunction_default from "./ImplicitFunction.js";
//#region Sources/Common/DataModel/Cylinder/index.js
function evaluate(radius, center, axis, x) {
	const x2C = new Float32Array(3);
	x2C[0] = x[0] - center[0];
	x2C[1] = x[1] - center[1];
	x2C[2] = x[2] - center[2];
	const proj = dot(axis, x2C);
	return dot(x2C, x2C) - proj * proj - radius * radius;
}
var STATIC = { evaluate };
function vtkCylinder(publicAPI, model) {
	model.classHierarchy.push("vtkCylinder");
	publicAPI.evaluateFunction = (xyz) => {
		const x2C = [
			xyz[0] - model.center[0],
			xyz[1] - model.center[1],
			xyz[2] - model.center[2]
		];
		const proj = dot(model.axis, x2C);
		return dot(x2C, x2C) - proj * proj - model.radius * model.radius;
	};
	publicAPI.evaluateGradient = (xyz) => {
		const t = model.axis[0] * (xyz[0] - model.center[0]) + model.axis[1] * (xyz[1] - model.center[1]) + model.axis[2] * (xyz[2] - model.center[2]);
		const cp = [
			model.center[0] + t * model.axis[0],
			model.center[1] + t * model.axis[1],
			model.center[2] + t * model.axis[2]
		];
		return [
			2 * (xyz[0] - cp[0]),
			2 * (xyz[1] - cp[1]),
			2 * (xyz[2] - cp[2])
		];
	};
}
var DEFAULT_VALUES = {
	radius: .5,
	center: [
		0,
		0,
		0
	],
	axis: [
		0,
		1,
		0
	]
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	ImplicitFunction_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, ["radius"]);
	macros_default.setGetArray(publicAPI, model, ["center", "axis"], 3);
	vtkCylinder(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkCylinder");
var Cylinder_default = {
	newInstance,
	extend,
	...STATIC
};
//#endregion
export { STATIC, Cylinder_default as default, extend, newInstance };
