// @kitware/vtk.js/Common/DataModel/DataSetAttributes.js
import macros_default from "../../macros.js";
import DataArray_default from "../Core/DataArray.js";
import FieldData_default from "./DataSetAttributes/FieldData.js";
import Constants_default from "./DataSetAttributes/Constants.js";
//#region Sources/Common/DataModel/DataSetAttributes/index.js
var { AttributeTypes, AttributeCopyOperations } = Constants_default;
var { vtkWarningMacro } = macros_default;
function vtkDataSetAttributes(publicAPI, model) {
	const attrTypes = [
		"Scalars",
		"Vectors",
		"Normals",
		"TCoords",
		"Tensors",
		"GlobalIds",
		"PedigreeIds"
	];
	function cleanAttributeType(attType) {
		let cleanAttType = attrTypes.find((ee) => AttributeTypes[ee.toUpperCase()] === attType || typeof attType !== "number" && ee.toLowerCase() === attType.toLowerCase());
		if (typeof cleanAttType === "undefined") cleanAttType = null;
		return cleanAttType;
	}
	model.classHierarchy.push("vtkDataSetAttributes");
	const superClass = { ...publicAPI };
	publicAPI.checkNumberOfComponents = (x) => true;
	publicAPI.setAttribute = (arr, uncleanAttType) => {
		const attType = cleanAttributeType(uncleanAttType);
		if (arr && attType.toUpperCase() === "PEDIGREEIDS" && !arr.isA("vtkDataArray")) {
			vtkWarningMacro(`Cannot set attribute ${attType}. The attribute must be a vtkDataArray.`);
			return -1;
		}
		if (arr && !publicAPI.checkNumberOfComponents(arr, attType)) {
			vtkWarningMacro(`Cannot set attribute ${attType}. Incorrect number of components.`);
			return -1;
		}
		if (arr) {
			const currentAttribute = publicAPI.addArray(arr);
			model[`active${attType}`] = currentAttribute;
		} else model[`active${attType}`] = -1;
		publicAPI.modified();
		return model[`active${attType}`];
	};
	publicAPI.getAttributes = (arr) => attrTypes.filter((attrType) => publicAPI[`get${attrType}`]() === arr);
	publicAPI.setActiveAttributeByName = (arrayName, attType) => publicAPI.setActiveAttributeByIndex(publicAPI.getArrayWithIndex(arrayName).index, attType);
	publicAPI.setActiveAttributeByIndex = (arrayIdx, uncleanAttType) => {
		const attType = cleanAttributeType(uncleanAttType);
		if (arrayIdx >= 0 && arrayIdx < model.arrays.length) {
			if (attType.toUpperCase() !== "PEDIGREEIDS") {
				const arr = publicAPI.getArrayByIndex(arrayIdx);
				if (!arr.isA("vtkDataArray")) {
					vtkWarningMacro(`Cannot set attribute ${attType}. Only vtkDataArray subclasses can be set as active attributes.`);
					return -1;
				}
				if (!publicAPI.checkNumberOfComponents(arr, attType)) {
					vtkWarningMacro(`Cannot set attribute ${attType}. Incorrect number of components.`);
					return -1;
				}
			}
			model[`active${attType}`] = arrayIdx;
			publicAPI.modified();
			return arrayIdx;
		}
		if (arrayIdx === -1) {
			model[`active${attType}`] = arrayIdx;
			publicAPI.modified();
		}
		return -1;
	};
	publicAPI.getActiveAttribute = (attType) => {
		return publicAPI[`get${cleanAttributeType(attType)}`]();
	};
	publicAPI.removeAllArrays = () => {
		attrTypes.forEach((attType) => {
			model[`active${attType}`] = -1;
		});
		superClass.removeAllArrays();
	};
	publicAPI.removeArrayByIndex = (arrayIdx) => {
		if (arrayIdx !== -1) attrTypes.forEach((attType) => {
			if (arrayIdx === model[`active${attType}`]) model[`active${attType}`] = -1;
			else if (arrayIdx < model[`active${attType}`]) model[`active${attType}`] -= 1;
		});
		return superClass.removeArrayByIndex(arrayIdx);
	};
	attrTypes.forEach((value) => {
		const activeVal = `active${value}`;
		publicAPI[`get${value}`] = () => publicAPI.getArrayByIndex(model[activeVal]);
		publicAPI[`set${value}`] = (da) => publicAPI.setAttribute(da, value);
		publicAPI[`setActive${value}`] = (arrayName) => publicAPI.setActiveAttributeByIndex(publicAPI.getArrayWithIndex(arrayName).index, value);
		publicAPI[`copy${value}Off`] = () => {
			const attType = value.toUpperCase();
			model.copyAttributeFlags[AttributeCopyOperations.PASSDATA][AttributeTypes[attType]] = false;
		};
		publicAPI[`copy${value}On`] = () => {
			const attType = value.toUpperCase();
			model.copyAttributeFlags[AttributeCopyOperations.PASSDATA][AttributeTypes[attType]] = true;
		};
	});
	publicAPI.initializeAttributeCopyFlags = () => {
		model.copyAttributeFlags = [];
		Object.keys(AttributeCopyOperations).filter((op) => op !== "ALLCOPY").forEach((attCopyOp) => {
			model.copyAttributeFlags[AttributeCopyOperations[attCopyOp]] = Object.keys(AttributeTypes).filter((ty) => ty !== "NUM_ATTRIBUTES").reduce((a, b) => {
				a[AttributeTypes[b]] = true;
				return a;
			}, []);
		});
		model.copyAttributeFlags[AttributeCopyOperations.COPYTUPLE][AttributeTypes.GLOBALIDS] = false;
		model.copyAttributeFlags[AttributeCopyOperations.INTERPOLATE][AttributeTypes.GLOBALIDS] = false;
		model.copyAttributeFlags[AttributeCopyOperations.COPYTUPLE][AttributeTypes.PEDIGREEIDS] = false;
	};
	publicAPI.initialize = macros_default.chain(publicAPI.initialize, publicAPI.initializeAttributeCopyFlags);
	if (model.dataArrays && Object.keys(model.dataArrays).length) Object.keys(model.dataArrays).forEach((name) => {
		if (!model.dataArrays[name].ref && model.dataArrays[name].type === "vtkDataArray") publicAPI.addArray(DataArray_default.newInstance(model.dataArrays[name]));
	});
	const superShallowCopy = publicAPI.shallowCopy;
	publicAPI.shallowCopy = (other, debug) => {
		superShallowCopy(other, debug);
		model.arrays = other.getArrays().map((arr) => {
			const arrNew = arr.newClone();
			arrNew.shallowCopy(arr, debug);
			return { data: arrNew };
		});
	};
	publicAPI.initializeAttributeCopyFlags();
}
var DEFAULT_VALUES = {
	activeScalars: -1,
	activeVectors: -1,
	activeTensors: -1,
	activeNormals: -1,
	activeTCoords: -1,
	activeGlobalIds: -1,
	activePedigreeIds: -1
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	FieldData_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, [
		"activeScalars",
		"activeNormals",
		"activeTCoords",
		"activeVectors",
		"activeTensors",
		"activeGlobalIds",
		"activePedigreeIds"
	]);
	if (!model.arrays) model.arrays = {};
	vtkDataSetAttributes(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkDataSetAttributes");
var DataSetAttributes_default = {
	newInstance,
	extend,
	...Constants_default
};
//#endregion
export { DataSetAttributes_default as default, extend, newInstance };
