// @kitware/vtk.js/Common/DataModel/EdgeLocator.js
//#region Sources/Common/DataModel/EdgeLocator/index.js
var EdgeLocator = class {
	constructor(oriented = false) {
		this.oriented = oriented;
		this.edgeMap = /* @__PURE__ */ new Map();
	}
	initialize() {
		this.edgeMap.clear();
	}
	computeEdgeKey(pointId0, pointId1) {
		return this.oriented || pointId0 < pointId1 ? .5 * (pointId0 * pointId1) * (pointId0 * pointId1 + 1) + pointId1 : .5 * (pointId1 * pointId0) * (pointId1 * pointId0 + 1) + pointId0;
	}
	insertUniqueEdge(pointId0, pointId1, newEdgeValue) {
		const key = this.computeEdgeKey(pointId0, pointId1);
		let node = this.edgeMap.get(key);
		if (!node) {
			node = {
				key,
				edgeId: this.edgeMap.size,
				value: newEdgeValue
			};
			this.edgeMap.set(key, node);
		}
		return node;
	}
	insertEdge(pointId0, pointId1, newEdgeValue) {
		const key = this.computeEdgeKey(pointId0, pointId1);
		const node = {
			key,
			edgeId: this.edgeMap.size,
			value: newEdgeValue
		};
		this.edgeMap.set(key, node);
		return node;
	}
	isInsertedEdge(pointId0, pointId1) {
		const key = this.computeEdgeKey(pointId0, pointId1);
		return this.edgeMap.get(key);
	}
	static getEdgePointIds(node) {
		const n = .5 * (-1 + Math.sqrt(8 * node.key + 1));
		const pointId0 = node.key - .5 * (n + 1) * n;
		return [pointId0, n - pointId0];
	}
};
function newInstance(initialValues = {}) {
	return new EdgeLocator(initialValues.oriented);
}
var EdgeLocator_default = { newInstance };
//#endregion
export { EdgeLocator_default as default };
