// @kitware/vtk.js/Common/DataModel/KochanekSpline1D.js
import macros_default from "../../macros.js";
import Spline1D_default from "./Spline1D.js";
import { BoundaryCondition } from "./Spline1D/Constants.js";
//#region Sources/Common/DataModel/KochanekSpline1D/index.js
var VTK_EPSILON = 1e-4;
function vtkKochanekSpline1D(publicAPI, model) {
	model.classHierarchy.push("vtkKochanekSpline1D");
	publicAPI.computeCloseCoefficients = (size, work, x, y) => {
		if (!model.coefficients || model.coefficients.length !== 4 * size) model.coefficients = new Float32Array(4 * size);
		const N = size - 1;
		for (let i = 1; i < N; i++) {
			const cs = y[i] - y[i - 1];
			const cd = y[i + 1] - y[i];
			let ds = cs * ((1 - model.tension) * (1 - model.continuity) * (1 + model.bias)) + cd * ((1 - model.tension) * (1 + model.continuity) * (1 - model.bias));
			let dd = cs * ((1 - model.tension) * (1 + model.continuity) * (1 + model.bias)) + cd * ((1 - model.tension) * (1 - model.continuity) * (1 - model.bias));
			const n1 = x[i + 1] - x[i];
			const n0 = x[i] - x[i - 1];
			ds *= n0 / (n0 + n1);
			dd *= n1 / (n0 + n1);
			model.coefficients[4 * i + 0] = y[i];
			model.coefficients[4 * i + 1] = dd;
			model.coefficients[4 * i + 2] = ds;
		}
		model.coefficients[0] = y[0];
		model.coefficients[4 * N + 0] = y[N];
		model.coefficients[4 * N + 1] = 0;
		model.coefficients[4 * N + 2] = 0;
		model.coefficients[4 * N + 3] = 0;
		const cs = y[N] - y[N - 1];
		const cd = y[1] - y[0];
		let ds = cs * ((1 - model.tension) * (1 - model.continuity) * (1 + model.bias)) + cd * ((1 - model.tension) * (1 + model.continuity) * (1 - model.bias));
		let dd = cs * ((1 - model.tension) * (1 + model.continuity) * (1 + model.bias)) + cd * ((1 - model.tension) * (1 - model.continuity) * (1 - model.bias));
		const n1 = x[1] - x[0];
		const n0 = x[N] - x[N - 1];
		ds *= n0 / (n0 + n1);
		dd *= n1 / (n0 + n1);
		model.coefficients[1] = dd;
		model.coefficients[2] = ds;
		model.coefficients[4 * N + 1] = dd;
		model.coefficients[4 * N + 2] = ds;
		for (let i = 0; i < N; i++) {
			model.coefficients[4 * i + 2] = -3 * y[i] + 3 * y[i + 1] + -2 * model.coefficients[4 * i + 1] + -1 * model.coefficients[4 * (i + 1) + 2];
			model.coefficients[4 * i + 3] = 2 * y[i] + -2 * y[i + 1] + 1 * model.coefficients[4 * i + 1] + 1 * model.coefficients[4 * (i + 1) + 2];
		}
	};
	publicAPI.computeOpenCoefficients = (size, work, x, y, options = {}) => {
		if (!model.coefficients || model.coefficients.length !== 4 * size) model.coefficients = new Float32Array(4 * size);
		const N = size - 1;
		for (let i = 1; i < N; i++) {
			const cs = y[i] - y[i - 1];
			const cd = y[i + 1] - y[i];
			let ds = cs * ((1 - model.tension) * (1 - model.continuity) * (1 + model.bias)) + cd * ((1 - model.tension) * (1 + model.continuity) * (1 - model.bias));
			let dd = cs * ((1 - model.tension) * (1 + model.continuity) * (1 + model.bias)) + cd * ((1 - model.tension) * (1 - model.continuity) * (1 - model.bias));
			const n1 = x[i + 1] - x[i];
			const n0 = x[i] - x[i - 1];
			ds *= n0 / (n0 + n1);
			dd *= n1 / (n0 + n1);
			model.coefficients[4 * i + 0] = y[i];
			model.coefficients[4 * i + 1] = dd;
			model.coefficients[4 * i + 2] = ds;
		}
		model.coefficients[0] = y[0];
		model.coefficients[4 * N + 0] = y[N];
		switch (options.leftConstraint) {
			case BoundaryCondition.DERIVATIVE:
				model.coefficients[1] = options.leftValue;
				break;
			case BoundaryCondition.SECOND_DERIVATIVE:
				model.coefficients[1] = (6 * (y[1] - y[0]) - 2 * model.coefficients[6] - options.leftValue) / 4;
				break;
			case BoundaryCondition.SECOND_DERIVATIVE_INTERIOR_POINT:
				if (Math.abs(options.leftValue + 2) > VTK_EPSILON) model.coefficients[1] = (3 * (1 + options.leftValue) * (y[1] - y[0]) - (1 + 2 * options.leftValue) * model.coefficients[6]) / (2 + options.leftValue);
				else model.coefficients[1] = 0;
				break;
			case BoundaryCondition.DEFAULT:
			default:
				model.coefficients[1] = y[2] - y[0];
				break;
		}
		switch (options.rightConstraint) {
			case BoundaryCondition.DERIVATIVE:
				model.coefficients[4 * N + 2] = options.leftValue;
				break;
			case BoundaryCondition.SECOND_DERIVATIVE:
				model.coefficients[4 * N + 2] = (6 * (y[N] - y[N - 1]) - 2 * model.coefficients[4 * (N - 1) + 1] + options.leftValue) / 4;
				break;
			case BoundaryCondition.SECOND_DERIVATIVE_INTERIOR_POINT:
				if (Math.abs(options.leftValue + 2) > VTK_EPSILON) model.coefficients[4 * N + 2] = (3 * (1 + options.leftValue) * (y[N] - y[N - 1]) - (1 + 2 * options.leftValue) * model.coefficients[4 * (N - 1) + 1]) / (2 + options.leftValue);
				else model.coefficients[4 * N + 2] = 0;
				break;
			case BoundaryCondition.DEFAULT:
			default:
				model.coefficients[4 * N + 2] = y[N] - y[N - 2];
				break;
		}
		for (let i = 0; i < N; i++) {
			model.coefficients[4 * i + 2] = -3 * y[i] + 3 * y[i + 1] + -2 * model.coefficients[4 * i + 1] + -1 * model.coefficients[4 * (i + 1) + 2];
			model.coefficients[4 * i + 3] = 2 * y[i] + -2 * y[i + 1] + 1 * model.coefficients[4 * i + 1] + 1 * model.coefficients[4 * (i + 1) + 2];
		}
	};
	publicAPI.getValue = (intervalIndex, t) => {
		const t2 = t * t;
		const t3 = t * t * t;
		return model.coefficients[4 * intervalIndex + 3] * t3 + model.coefficients[4 * intervalIndex + 2] * t2 + model.coefficients[4 * intervalIndex + 1] * t + model.coefficients[4 * intervalIndex + 0];
	};
}
var DEFAULT_VALUES = {
	tension: 0,
	bias: 0,
	continuity: 0
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	Spline1D_default.extend(publicAPI, model, initialValues);
	macros_default.obj(publicAPI, model);
	vtkKochanekSpline1D(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkKochanekSpline1D");
var KochanekSpline1D_default = {
	newInstance,
	extend
};
//#endregion
export { KochanekSpline1D_default as default, extend, newInstance };
