// @kitware/vtk.js/Common/DataModel/PiecewiseFunction.js
import macros_default from "../../macros.js";
//#region Sources/Common/DataModel/PiecewiseFunction/index.js
var { vtkErrorMacro } = macros_default;
function vtkPiecewiseFunction(publicAPI, model) {
	model.classHierarchy.push("vtkPiecewiseFunction");
	publicAPI.getSize = () => model.nodes.length;
	publicAPI.getType = () => {
		let value;
		let prevValue = 0;
		let functionType = 0;
		if (model.nodes.length > 0) prevValue = model.nodes[0].y;
		for (let i = 1; i < model.nodes.length; i++) {
			value = model.nodes[i].y;
			if (value !== prevValue) if (value > prevValue) switch (functionType) {
				case 0:
				case 1:
					functionType = 1;
					break;
				default:
					functionType = 3;
					break;
			}
			else switch (functionType) {
				case 0:
				case 2:
					functionType = 2;
					break;
				default:
					functionType = 3;
					break;
			}
			prevValue = value;
			if (functionType === 3) break;
		}
		switch (functionType) {
			case 0: return "Constant";
			case 1: return "NonDecreasing";
			case 2: return "NonIncreasing";
			default: return "Varied";
		}
	};
	publicAPI.getDataPointer = () => {
		const size = model.nodes.length;
		model.function = null;
		if (size > 0) {
			model.function = [];
			for (let i = 0; i < size; i++) {
				model.function[2 * i] = model.nodes[i].x;
				model.function[2 * i + 1] = model.nodes[i].y;
			}
		}
		return model.function;
	};
	publicAPI.getFirstNonZeroValue = () => {
		if (model.nodes.length === 0) return 0;
		let allZero = 1;
		let x = 0;
		let i = 0;
		for (; i < model.nodes.length; i++) if (model.nodes[i].y !== 0) {
			allZero = 0;
			break;
		}
		if (allZero) x = Number.MAX_VALUE;
		else if (i > 0) x = model.nodes[i - 1].x;
		else if (model.clamping) x = -Number.MAX_VALUE;
		else x = model.nodes[0].x;
		return x;
	};
	publicAPI.getNodeValue = (index, val) => {
		const size = model.nodes.length;
		if (index < 0 || index >= size) {
			vtkErrorMacro("Index out of range!");
			return -1;
		}
		val[0] = model.nodes[index].x;
		val[1] = model.nodes[index].y;
		val[2] = model.nodes[index].midpoint;
		val[3] = model.nodes[index].sharpness;
		return 1;
	};
	publicAPI.setNodeValue = (index, val) => {
		const size = model.nodes.length;
		if (index < 0 || index >= size) {
			vtkErrorMacro("Index out of range!");
			return -1;
		}
		const oldX = model.nodes[index].x;
		model.nodes[index].x = val[0];
		model.nodes[index].y = val[1];
		model.nodes[index].midpoint = val[2];
		model.nodes[index].sharpness = val[3];
		if (oldX !== val[0]) publicAPI.sortAndUpdateRange();
		else publicAPI.modified();
		return 1;
	};
	publicAPI.addPoint = (x, y) => publicAPI.addPointLong(x, y, .5, 0);
	publicAPI.addPointLong = (x, y, midpoint, sharpness) => {
		if (midpoint < 0 || midpoint > 1) {
			vtkErrorMacro("Midpoint outside range [0.0, 1.0]");
			return -1;
		}
		if (sharpness < 0 || sharpness > 1) {
			vtkErrorMacro("Sharpness outside range [0.0, 1.0]");
			return -1;
		}
		if (!model.allowDuplicateScalars) publicAPI.removePoint(x);
		const node = {
			x,
			y,
			midpoint,
			sharpness
		};
		model.nodes.push(node);
		publicAPI.sortAndUpdateRange();
		let i;
		for (i = 0; i < model.nodes.length; i++) if (model.nodes[i].x === x) break;
		if (i < model.nodes.length) return i;
		return -1;
	};
	publicAPI.setNodes = (nodes) => {
		if (model.nodes !== nodes) {
			model.nodes = nodes;
			publicAPI.sortAndUpdateRange();
		}
	};
	publicAPI.sortAndUpdateRange = () => {
		model.nodes.sort((a, b) => a.x - b.x);
		if (!publicAPI.updateRange()) publicAPI.modified();
	};
	publicAPI.updateRange = () => {
		const oldRange = model.range.slice();
		const size = model.nodes.length;
		if (size) {
			model.range[0] = model.nodes[0].x;
			model.range[1] = model.nodes[size - 1].x;
		} else {
			model.range[0] = 0;
			model.range[1] = 0;
		}
		if (oldRange[0] === model.range[0] && oldRange[1] === model.range[1]) return false;
		publicAPI.modified();
		return true;
	};
	publicAPI.removePoint = (x) => {
		let i;
		for (i = 0; i < model.nodes.length; i++) if (model.nodes[i].x === x) break;
		if (i >= model.nodes.length) return -1;
		const retVal = i;
		let modifiedInvoked = false;
		model.nodes.splice(i, 1);
		if (i === 0 || i === model.nodes.length) modifiedInvoked = publicAPI.updateRange();
		if (!modifiedInvoked) publicAPI.modified();
		return retVal;
	};
	publicAPI.removeAllPoints = () => {
		model.nodes = [];
		publicAPI.sortAndUpdateRange();
	};
	publicAPI.addSegment = (x1, y1, x2, y2) => {
		publicAPI.sortAndUpdateRange();
		for (let i = 0; i < model.nodes.length;) if (model.nodes[i].x >= x1 && model.nodes[i].x <= x2) model.nodes.splice(i, 1);
		else i++;
		publicAPI.addPoint(x1, y1, .5, 0);
		publicAPI.addPoint(x2, y2, .5, 0);
	};
	publicAPI.getValue = (x) => {
		const table = [];
		publicAPI.getTable(x, x, 1, table);
		return table[0];
	};
	publicAPI.adjustRange = (range) => {
		if (range.length < 2) return 0;
		const functionRange = publicAPI.getRange();
		if (functionRange[0] < range[0]) publicAPI.addPoint(range[0], publicAPI.getValue(range[0]));
		else publicAPI.addPoint(range[0], publicAPI.getValue(functionRange[0]));
		if (functionRange[1] > range[1]) publicAPI.addPoint(range[1], publicAPI.getValue(range[1]));
		else publicAPI.addPoint(range[1], publicAPI.getValue(functionRange[1]));
		publicAPI.sortAndUpdateRange();
		for (let i = 0; i < model.nodes.length;) if (model.nodes[i].x >= range[0] && model.nodes[i].x <= range[1]) model.nodes.splice(i, 1);
		else ++i;
		publicAPI.sortAndUpdateRange();
		return 1;
	};
	publicAPI.estimateMinNumberOfSamples = (x1, x2) => {
		const d = publicAPI.findMinimumXDistance();
		return Math.ceil((x2 - x1) / d);
	};
	publicAPI.findMinimumXDistance = () => {
		const size = model.nodes.length;
		if (size < 2) return -1;
		let distance = model.nodes[1].x - model.nodes[0].x;
		for (let i = 0; i < size - 1; i++) {
			const currentDist = model.nodes[i + 1].x - model.nodes[i].x;
			if (currentDist < distance) distance = currentDist;
		}
		return distance;
	};
	publicAPI.getTable = (xStart, xEnd, size, table, stride = 1) => {
		let i;
		let idx = 0;
		const numNodes = model.nodes.length;
		let lastValue = 0;
		if (numNodes !== 0) lastValue = model.nodes[numNodes - 1].y;
		let x = 0;
		let x1 = 0;
		let x2 = 0;
		let y1 = 0;
		let y2 = 0;
		let midpoint = 0;
		let sharpness = 0;
		for (i = 0; i < size; i++) {
			const tidx = stride * i;
			if (size > 1) x = xStart + i / (size - 1) * (xEnd - xStart);
			else x = .5 * (xStart + xEnd);
			while (idx < numNodes && x > model.nodes[idx].x) {
				idx++;
				if (idx < numNodes) {
					x1 = model.nodes[idx - 1].x;
					x2 = model.nodes[idx].x;
					y1 = model.nodes[idx - 1].y;
					y2 = model.nodes[idx].y;
					midpoint = model.nodes[idx - 1].midpoint;
					sharpness = model.nodes[idx - 1].sharpness;
					if (midpoint < 1e-5) midpoint = 1e-5;
					if (midpoint > .99999) midpoint = .99999;
				}
			}
			if (idx >= numNodes) table[tidx] = model.clamping ? lastValue : 0;
			else if (idx === 0) table[tidx] = model.clamping ? model.nodes[0].y : 0;
			else {
				let s = (x - x1) / (x2 - x1);
				if (s < midpoint) s = .5 * s / midpoint;
				else s = .5 + .5 * (s - midpoint) / (1 - midpoint);
				if (sharpness > .99) if (s < .5) {
					table[tidx] = y1;
					continue;
				} else {
					table[tidx] = y2;
					continue;
				}
				if (sharpness < .01) {
					table[tidx] = (1 - s) * y1 + s * y2;
					continue;
				}
				if (s < .5) s = .5 * (s * 2) ** (1 + 10 * sharpness);
				else if (s > .5) s = 1 - .5 * ((1 - s) * 2) ** (1 + 10 * sharpness);
				const ss = s * s;
				const sss = ss * s;
				const h1 = 2 * sss - 3 * ss + 1;
				const h2 = -2 * sss + 3 * ss;
				const h3 = sss - 2 * ss + s;
				const h4 = sss - ss;
				const slope = y2 - y1;
				const t = (1 - sharpness) * slope;
				table[tidx] = h1 * y1 + h2 * y2 + h3 * t + h4 * t;
				const min = y1 < y2 ? y1 : y2;
				const max = y1 > y2 ? y1 : y2;
				table[tidx] = table[tidx] < min ? min : table[tidx];
				table[tidx] = table[tidx] > max ? max : table[tidx];
			}
		}
	};
}
var DEFAULT_VALUES = {
	range: [0, 0],
	clamping: true,
	allowDuplicateScalars: false
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	model.nodes = [];
	macros_default.setGet(publicAPI, model, ["allowDuplicateScalars", "clamping"]);
	macros_default.setArray(publicAPI, model, ["range"], 2);
	macros_default.getArray(publicAPI, model, ["range"]);
	vtkPiecewiseFunction(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkPiecewiseFunction");
var PiecewiseFunction_default = {
	newInstance,
	extend
};
//#endregion
export { PiecewiseFunction_default as default, extend, newInstance };
