// @kitware/vtk.js/Common/DataModel/Polygon.js
import macros_default from "../../macros.js";
import { add, cross, distance2BetweenPoints, dot, norm, normalize, subtract } from "../Core/Math.js";
import Plane_default from "./Plane.js";
import BoundingBox_default from "./BoundingBox.js";
import PriorityQueue_default from "../Core/PriorityQueue.js";
import { IntersectionState } from "./Line/Constants.js";
import Line_default from "./Line.js";
import { FLOAT_EPSILON, PolygonWithPointIntersectionState, TOLERANCE } from "./Polygon/Constants.js";
//#region Sources/Common/DataModel/Polygon/index.js
function pointLocation(axis0, axis1, p0, p1, point) {
	return (p1[axis0] - p0[axis0]) * (point[axis1] - p0[axis1]) - (point[axis0] - p0[axis0]) * (p1[axis1] - p0[axis1]);
}
function pointInPolygon(point, vertices, bounds, normal) {
	if (point[0] < bounds[0] || point[0] > bounds[1] || point[1] < bounds[2] || point[1] > bounds[3] || point[2] < bounds[4] || point[2] > bounds[5]) return PolygonWithPointIntersectionState.OUTSIDE;
	if (normalize(normal) <= 1.1920929e-7) return PolygonWithPointIntersectionState.FAILURE;
	let tol2 = TOLERANCE * ((bounds[1] - bounds[0]) * (bounds[1] - bounds[0]) + (bounds[3] - bounds[2]) * (bounds[3] - bounds[2]) + (bounds[5] - bounds[4]) * (bounds[5] - bounds[4]));
	tol2 *= tol2;
	tol2 = tol2 === 0 ? FLOAT_EPSILON : tol2;
	const p0 = [];
	const p1 = [];
	for (let i = 0; i < vertices.length;) {
		p0[0] = vertices[i++];
		p0[1] = vertices[i++];
		p0[2] = vertices[i++];
		if (distance2BetweenPoints(point, p0) <= tol2) return PolygonWithPointIntersectionState.INSIDE;
		const { distance, t } = Line_default.distanceToLine(point, p0, p1);
		if (distance <= tol2 && t > 0 && t < 1) return PolygonWithPointIntersectionState.INSIDE;
	}
	let axis0;
	let axis1;
	if (Math.abs(normal[0]) > Math.abs(normal[1])) if (Math.abs(normal[0]) > Math.abs(normal[2])) {
		axis0 = 1;
		axis1 = 2;
	} else {
		axis0 = 0;
		axis1 = 1;
	}
	else if (Math.abs(normal[1]) > Math.abs(normal[2])) {
		axis0 = 0;
		axis1 = 2;
	} else {
		axis0 = 0;
		axis1 = 1;
	}
	let wn = 0;
	for (let i = 0; i < vertices.length;) {
		p0[0] = vertices[i++];
		p0[1] = vertices[i++];
		p0[2] = vertices[i++];
		if (i < vertices.length) {
			p1[0] = vertices[i];
			p1[1] = vertices[i + 1];
			p1[2] = vertices[i + 2];
		} else {
			p1[0] = vertices[0];
			p1[1] = vertices[1];
			p1[2] = vertices[2];
		}
		if (p0[axis1] <= point[axis1]) {
			if (p1[axis1] > point[axis1]) {
				if (pointLocation(axis0, axis1, p0, p1, point) > 0) ++wn;
			}
		} else if (p1[axis1] <= point[axis1]) {
			if (pointLocation(axis0, axis1, p0, p1, point) < 0) --wn;
		}
	}
	return wn === 0 ? PolygonWithPointIntersectionState.OUTSIDE : PolygonWithPointIntersectionState.INSIDE;
}
/**
* Simple utility method for computing polygon bounds.
* Returns the sum of the squares of the dimensions.
* Requires a poly with at least one point.
*
* @param {Array<Number>|TypedArray<Number>} poly
* @param {vtkPoints} points
* @param {Bound} bounds
*/
function getBounds(poly, points, bounds) {
	const n = poly.length;
	const p = [];
	points.getPoint(poly[0], p);
	bounds[0] = p[0];
	bounds[1] = p[0];
	bounds[2] = p[1];
	bounds[3] = p[1];
	bounds[4] = p[2];
	bounds[5] = p[2];
	for (let j = 1; j < n; j++) {
		points.getPoint(poly[j], p);
		BoundingBox_default.addPoint(bounds, ...p);
	}
	const length = BoundingBox_default.getLengths(bounds);
	return dot(length, length);
}
/**
* Compute the normal of a polygon and return its norm.
*
* TBD: This does the same thing as vtkPolygon.computeNormal,
* but in a more generic way. Maybe we can keep the public
* static method somehow and have the private method use it instead.
*
* @param {Array<Number>|TypedArray<Number>} poly
* @param {vtkPoints} points
* @param {Vector3} normal
* @returns {Number}
*/
function getNormal(poly, points, normal) {
	normal.length = 3;
	normal[0] = 0;
	normal[1] = 0;
	normal[2] = 0;
	const p0 = [];
	let p1 = [];
	let p2 = [];
	const v1 = [];
	const v2 = [];
	points.getPoint(poly[0], p0);
	points.getPoint(poly[1], p1);
	for (let j = 2; j < poly.length; j++) {
		points.getPoint(poly[j], p2);
		subtract(p2, p1, v1);
		subtract(p0, p1, v2);
		const n = [
			0,
			0,
			0
		];
		cross(v1, v2, n);
		add(normal, n, normal);
		[p1, p2] = [p2, p1];
	}
	return normalize(normal);
}
/**
* Compute the centroid of a polygon.
* @param {Array<number>} poly - Array of point indices for the polygon
* @param {vtkPoints} points - vtkPoints instance
* @param {Vector3} [centroid] - Optional output array (length 3)
* @returns {Vector3} The centroid as [x, y, z]
*/
function computeCentroid(poly, points, centroid = [
	0,
	0,
	0
]) {
	centroid[0] = 0;
	centroid[1] = 0;
	centroid[2] = 0;
	const n = poly.length;
	const p = [];
	for (let i = 0; i < n; i++) {
		points.getPoint(poly[i], p);
		centroid[0] += p[0];
		centroid[1] += p[1];
		centroid[2] += p[2];
	}
	centroid[0] /= n;
	centroid[1] /= n;
	centroid[2] /= n;
	return centroid;
}
var STATIC = {
	PolygonWithPointIntersectionState,
	pointInPolygon,
	getBounds,
	getNormal,
	computeCentroid
};
function vtkPolygon(publicAPI, model) {
	model.classHierarchy.push("vtkPolygon");
	function computeNormal() {
		const v1 = [
			0,
			0,
			0
		];
		const v2 = [
			0,
			0,
			0
		];
		model.normal = [
			0,
			0,
			0
		];
		const anchor = [...model.firstPoint.point];
		let point = model.firstPoint;
		for (let i = 0; i < model.pointCount; i++) {
			subtract(point.point, anchor, v1);
			subtract(point.next.point, anchor, v2);
			const n = [
				0,
				0,
				0
			];
			cross(v1, v2, n);
			add(model.normal, n, model.normal);
			point = point.next;
		}
		return normalize(model.normal);
	}
	function computeMeasure(point) {
		const v1 = [
			0,
			0,
			0
		];
		const v2 = [
			0,
			0,
			0
		];
		const v3 = [
			0,
			0,
			0
		];
		const v4 = [
			0,
			0,
			0
		];
		subtract(point.point, point.previous.point, v1);
		subtract(point.next.point, point.point, v2);
		subtract(point.previous.point, point.next.point, v3);
		cross(v1, v2, v4);
		const area = dot(v4, model.normal);
		if (area <= 0) return -1;
		const perimeter = norm(v1) + norm(v2) + norm(v3);
		return perimeter * perimeter / area;
	}
	function canRemoveVertex(point) {
		if (model.pointCount <= 3) return true;
		const previous = point.previous;
		const next = point.next;
		const v = [
			0,
			0,
			0
		];
		subtract(next.point, previous.point, v);
		const sN = [
			0,
			0,
			0
		];
		cross(v, model.normal, sN);
		normalize(sN);
		if (norm(sN) === 0) return false;
		let val = Plane_default.evaluate(sN, previous.point, next.next.point);
		let currentSign = val > 1e-6 ? 1 : val < -1e-6 ? -1 : 0;
		let oneNegative = currentSign < 0 ? 1 : 0;
		for (let vertex = next.next.next; vertex.id !== previous.id; vertex = vertex.next) {
			const previousVertex = vertex.previous;
			val = Plane_default.evaluate(sN, previous.point, vertex.point);
			const sign = val > 1e-6 ? 1 : val < -1e-6 ? -1 : 0;
			if (sign !== currentSign) {
				if (!oneNegative) oneNegative = sign <= 0 ? 1 : 0;
				if (Line_default.intersection(previous.point, next.point, vertex.point, previousVertex.point, [0], [0]) === IntersectionState.YES_INTERSECTION) return false;
				currentSign = sign;
			}
		}
		return oneNegative === 1;
	}
	function removePoint(point, queue) {
		model.pointCount -= 1;
		const previous = point.previous;
		const next = point.next;
		model.tris = model.tris.concat(point.point);
		model.tris = model.tris.concat(next.point);
		model.tris = model.tris.concat(previous.point);
		previous.next = next;
		next.previous = previous;
		queue.deleteById(previous.id);
		queue.deleteById(next.id);
		const previousMeasure = computeMeasure(previous);
		if (previousMeasure > 0) queue.push(previousMeasure, previous);
		const nextMeasure = computeMeasure(next);
		if (nextMeasure > 0) queue.push(nextMeasure, next);
		if (point.id === model.firstPoint.id) model.firstPoint = next;
	}
	function earCutTriangulation() {
		computeNormal();
		const vertexQueue = PriorityQueue_default.newInstance();
		let point = model.firstPoint;
		for (let i = 0; i < model.pointCount; i++) {
			const measure = computeMeasure(point);
			if (measure > 0) vertexQueue.push(measure, point);
			point = point.next;
		}
		while (model.pointCount > 2 && vertexQueue.length() > 0) if (model.pointCount === vertexQueue.length()) removePoint(vertexQueue.pop(), vertexQueue);
		else {
			const pointToRemove = vertexQueue.pop();
			if (canRemoveVertex(pointToRemove)) removePoint(pointToRemove, vertexQueue);
		}
		return model.pointCount <= 2;
	}
	publicAPI.triangulate = () => {
		if (!model.firstPoint) return null;
		return earCutTriangulation();
	};
	publicAPI.setPoints = (points) => {
		model.pointCount = points.length;
		model.firstPoint = {
			id: 0,
			point: points[0],
			next: null,
			previous: null
		};
		let currentPoint = model.firstPoint;
		for (let i = 1; i < model.pointCount; i++) {
			currentPoint.next = {
				id: i,
				point: points[i],
				next: null,
				previous: currentPoint
			};
			currentPoint = currentPoint.next;
		}
		model.firstPoint.previous = currentPoint;
		currentPoint.next = model.firstPoint;
	};
	publicAPI.getPointArray = () => model.tris;
}
var DEFAULT_VALUES = {
	firstPoint: null,
	pointCount: 0,
	tris: []
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	vtkPolygon(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkPolygon");
var Polygon_default = {
	newInstance,
	extend,
	...STATIC
};
//#endregion
export { computeCentroid, Polygon_default as default, extend, getBounds, getNormal, newInstance };
