// @kitware/vtk.js/Common/DataModel/Spline3D.js
import macros_default from "../../macros.js";
import CardinalSpline1D_default from "./CardinalSpline1D.js";
import KochanekSpline1D_default from "./KochanekSpline1D.js";
import { splineKind } from "./Spline3D/Constants.js";
//#region Sources/Common/DataModel/Spline3D/index.js
var { vtkErrorMacro } = macros_default;
function vtkSpline3D(publicAPI, model) {
	model.classHierarchy.push("vtkSpline3D");
	function computeCoefficients1D(spline, points, boundaryConditionValue) {
		if (points.length === 0) vtkErrorMacro("Splines require at least one points");
		if (points.length === 1) points.push(points[0]);
		const size = points.length;
		let work = null;
		let intervals = null;
		work = new Float32Array(size);
		if (model.intervals.length === 0) {
			intervals = new Float32Array(size);
			for (let i = 0; i < intervals.length; i++) intervals[i] = i;
		} else intervals = model.intervals;
		if (model.close) spline.computeCloseCoefficients(size, work, intervals, points);
		else spline.computeOpenCoefficients(size, work, intervals, points, {
			leftConstraint: model.boundaryCondition,
			leftValue: boundaryConditionValue,
			rightConstraint: model.boundaryCondition,
			rightValue: boundaryConditionValue
		});
	}
	publicAPI.computeCoefficients = (points) => {
		const x = points.map((pt) => pt[0]);
		const y = points.map((pt) => pt[1]);
		const z = points.map((pt) => pt[2]);
		computeCoefficients1D(model.splineX, x, model.boundaryConditionValues[0]);
		computeCoefficients1D(model.splineY, y, model.boundaryConditionValues[1]);
		computeCoefficients1D(model.splineZ, z, model.boundaryConditionValues[2]);
	};
	publicAPI.getPoint = (intervalIndex, t) => [
		model.splineX.getValue(intervalIndex, t),
		model.splineY.getValue(intervalIndex, t),
		model.splineZ.getValue(intervalIndex, t)
	];
	if (model.kind === splineKind.KOCHANEK_SPLINE) {
		model.splineX = KochanekSpline1D_default.newInstance({
			tension: model.tension,
			continuity: model.continuity,
			bias: model.bias
		});
		model.splineY = KochanekSpline1D_default.newInstance({
			tension: model.tension,
			continuity: model.continuity,
			bias: model.bias
		});
		model.splineZ = KochanekSpline1D_default.newInstance({
			tension: model.tension,
			continuity: model.continuity,
			bias: model.bias
		});
	} else if (model.kind === splineKind.CARDINAL_SPLINE) {
		model.splineX = CardinalSpline1D_default.newInstance();
		model.splineY = CardinalSpline1D_default.newInstance();
		model.splineZ = CardinalSpline1D_default.newInstance();
	} else vtkErrorMacro(`Unknown spline type ${model.kind}`);
}
var DEFAULT_VALUES = {
	close: false,
	intervals: [],
	kind: splineKind.KOCHANEK_SPLINE,
	boundaryConditionValues: [
		0,
		0,
		0
	],
	tension: 0,
	continuity: 0,
	bias: 0
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, ["close", "intervals"]);
	vtkSpline3D(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkSpline3D");
var Spline3D_default = {
	newInstance,
	extend
};
//#endregion
export { Spline3D_default as default, extend, newInstance };
