// @kitware/vtk.js/Widgets/Widgets3D/ResliceCursorWidget/behavior.js
import macros_default from "../../../macros.js";
import { add, cross, dot, multiplyAccumulate, multiplyScalar, normalize, signedAngleBetweenVectors, subtract } from "../../../Common/Core/Math.js";
import BoundingBox_default from "../../../Common/DataModel/BoundingBox.js";
import Line_default from "../../../Common/DataModel/Line.js";
import { InteractionMethodsName, ScrollingMethods, planeNameToViewType } from "./Constants.js";
import { boundPointOnPlane, getLineInPlaneName, getLineNames, getLinePlaneName, getOtherLineName, rotateVector, updateState } from "./helpers.js";
//#region Sources/Widgets/Widgets3D/ResliceCursorWidget/behavior.js
function widgetBehavior(publicAPI, model) {
	model._isDragging = false;
	let isScrolling = false;
	let previousPosition;
	macros_default.setGet(publicAPI, model, ["keepOrthogonality", {
		type: "object",
		name: "cursorStyles"
	}]);
	publicAPI.setCursorStyles({
		[InteractionMethodsName.TranslateCenter]: "move",
		[InteractionMethodsName.RotateLine]: "alias",
		[InteractionMethodsName.TranslateAxis]: "pointer",
		default: "default"
	});
	publicAPI.setEnableTranslation = (enable) => {
		model.representations[0].setPickable(enable);
		model.representations[2].setPickable(enable);
	};
	publicAPI.setEnableRotation = (enable) => {
		model.representations[1].setPickable(enable);
	};
	publicAPI.getActiveInteraction = () => {
		if (model.widgetState.getStatesWithLabel("rotation").includes(model.activeState)) return InteractionMethodsName.RotateLine;
		if (model.widgetState.getStatesWithLabel("line").includes(model.activeState)) return InteractionMethodsName.TranslateAxis;
		if (model.widgetState.getStatesWithLabel("center").includes(model.activeState)) return InteractionMethodsName.TranslateCenter;
		return null;
	};
	/**
	* ActiveState can be RotationHandle or a LineHandle
	* @returns 'YinX', 'ZinX', 'XinY', 'ZinY', 'XinZ' or 'YinZ'
	*/
	publicAPI.getActiveLineName = () => getLineNames(model.widgetState).find((lineName) => model.widgetState.getStatesWithLabel(lineName).includes(model.activeState));
	publicAPI.getActiveLineHandle = () => model.widgetState[`getAxis${publicAPI.getActiveLineName()}`]?.();
	/**
	* Return the line handle of the other line in the same view.
	* @param {string} lineName name of the line (YinX, ZinX, XinY, ZinY, XinZ, YinZ)
	* @returns ZinX if lineName == YinX, YinX if lineName == ZinX, ZinY if lineName == XinY...
	*/
	publicAPI.getOtherLineHandle = (lineName) => model.widgetState[`getAxis${getOtherLineName(model.widgetState, lineName)}`]?.();
	/**
	* There are 2 rotation handles per axis: 'point0' and 'point1'.
	* This function returns which rotation handle (point0 or point1) is currently active.
	* ActiveState must be a RotationHandle.
	* @returns 'point0', 'point1' or null if no point is active (e.g. line is being rotated)
	*/
	publicAPI.getActiveRotationPointName = () => {
		if (model.widgetState.getStatesWithLabel("point0").includes(model.activeState)) return "point0";
		if (model.widgetState.getStatesWithLabel("point1").includes(model.activeState)) return "point1";
		return null;
	};
	publicAPI.startScrolling = (newPosition) => {
		if (newPosition) previousPosition = newPosition;
		isScrolling = true;
		publicAPI.startInteraction();
	};
	publicAPI.endScrolling = () => {
		isScrolling = false;
		publicAPI.endInteraction();
	};
	publicAPI.updateCursor = () => {
		const cursorStyles = publicAPI.getCursorStyles();
		if (cursorStyles) switch (publicAPI.getActiveInteraction()) {
			case InteractionMethodsName.TranslateCenter:
				model._apiSpecificRenderWindow.setCursor(cursorStyles.translateCenter);
				break;
			case InteractionMethodsName.RotateLine:
				model._apiSpecificRenderWindow.setCursor(cursorStyles.rotateLine);
				break;
			case InteractionMethodsName.TranslateAxis:
				model._apiSpecificRenderWindow.setCursor(cursorStyles.translateAxis);
				break;
			default:
				model._apiSpecificRenderWindow.setCursor(cursorStyles.default);
				break;
		}
	};
	publicAPI.handleLeftButtonPress = (callData) => {
		if (model.activeState && model.activeState.getActive()) {
			model._isDragging = true;
			const viewType = model.viewType;
			const currentPlaneNormal = model.widgetState.getPlanes()[viewType].normal;
			const manipulator = model.activeState?.getManipulator?.() ?? model.manipulator;
			manipulator.setWidgetOrigin(model.widgetState.getCenter());
			manipulator.setWidgetNormal(currentPlaneNormal);
			const { worldCoords } = manipulator.handleEvent(callData, model._apiSpecificRenderWindow);
			previousPosition = worldCoords;
			publicAPI.startInteraction();
		} else if (model.widgetState.getScrollingMethod() === ScrollingMethods.LEFT_MOUSE_BUTTON) publicAPI.startScrolling(callData.position);
		else return macros_default.VOID;
		return macros_default.EVENT_ABORT;
	};
	publicAPI.handleMouseMove = (callData) => {
		if (model._isDragging) return publicAPI.handleEvent(callData);
		if (isScrolling) {
			if (previousPosition.y !== callData.position.y) {
				const step = previousPosition.y - callData.position.y;
				publicAPI.translateCenterOnPlaneDirection(step);
				previousPosition = callData.position;
				publicAPI.invokeInteractionEvent(publicAPI.getActiveInteraction());
			}
		}
		return macros_default.VOID;
	};
	publicAPI.handleLeftButtonRelease = () => {
		if (model._isDragging || isScrolling) publicAPI.endScrolling();
		model._isDragging = false;
		model.widgetState.deactivate();
	};
	publicAPI.handleRightButtonPress = (calldata) => {
		if (model.widgetState.getScrollingMethod() === ScrollingMethods.RIGHT_MOUSE_BUTTON) publicAPI.startScrolling(calldata.position);
	};
	publicAPI.handleRightButtonRelease = () => {
		if (model.widgetState.getScrollingMethod() === ScrollingMethods.RIGHT_MOUSE_BUTTON) publicAPI.endScrolling();
	};
	publicAPI.handleStartMouseWheel = () => {
		publicAPI.startInteraction();
	};
	publicAPI.handleMouseWheel = (calldata) => {
		const step = calldata.spinY;
		isScrolling = true;
		publicAPI.translateCenterOnPlaneDirection(step);
		publicAPI.invokeInteractionEvent(InteractionMethodsName.TranslateCenter);
		isScrolling = false;
		return macros_default.EVENT_ABORT;
	};
	publicAPI.handleEndMouseWheel = () => {
		publicAPI.endScrolling();
	};
	publicAPI.handleMiddleButtonPress = (calldata) => {
		if (model.widgetState.getScrollingMethod() === ScrollingMethods.MIDDLE_MOUSE_BUTTON) publicAPI.startScrolling(calldata.position);
	};
	publicAPI.handleMiddleButtonRelease = () => {
		if (model.widgetState.getScrollingMethod() === ScrollingMethods.MIDDLE_MOUSE_BUTTON) publicAPI.endScrolling();
	};
	publicAPI.handleEvent = (callData) => {
		if (model.activeState.getActive()) {
			const methodName = publicAPI.getActiveInteraction();
			publicAPI[methodName](callData);
			publicAPI.invokeInteractionEvent(methodName);
			return macros_default.EVENT_ABORT;
		}
		return macros_default.VOID;
	};
	publicAPI.startInteraction = () => {
		publicAPI.invokeStartInteractionEvent();
		publicAPI.getViewWidgets().forEach((viewWidget) => {
			viewWidget.getInteractor().requestAnimation(publicAPI);
		});
	};
	publicAPI.endInteraction = () => {
		publicAPI.invokeEndInteractionEvent();
		publicAPI.getViewWidgets().forEach((viewWidget) => {
			viewWidget.getInteractor().cancelAnimation(publicAPI);
		});
	};
	publicAPI.translateCenterOnPlaneDirection = (nbSteps) => {
		const dirProj = model.widgetState.getPlanes()[model.viewType].normal;
		const oldCenter = model.widgetState.getCenter();
		const imageSpacing = model.widgetState.getImage().getSpacing();
		const absDirProj = dirProj.map((value) => Math.abs(value));
		const index = absDirProj.indexOf(Math.max(...absDirProj));
		const movingFactor = nbSteps * imageSpacing[index] / Math.abs(dirProj[index]);
		let newCenter = [
			oldCenter[0] + movingFactor * dirProj[0],
			oldCenter[1] + movingFactor * dirProj[1],
			oldCenter[2] + movingFactor * dirProj[2]
		];
		newCenter = publicAPI.getBoundedCenter(newCenter);
		model.widgetState.setCenter(newCenter);
		updateState(model.widgetState, model._factory.getScaleInPixels(), model._factory.getRotationHandlePosition());
	};
	publicAPI[InteractionMethodsName.TranslateAxis] = (calldata) => {
		const lineHandle = publicAPI.getActiveLineHandle();
		const lineName = publicAPI.getActiveLineName();
		const pointOnLine = add(lineHandle.getOrigin(), lineHandle.getDirection(), []);
		const currentLineVector = lineHandle.getDirection();
		normalize(currentLineVector);
		const otherLineHandle = publicAPI.getOtherLineHandle(lineName);
		const center = model.widgetState.getCenter();
		const manipulator = model.activeState?.getManipulator?.() ?? model.manipulator;
		let worldCoords = null;
		let newOrigin = [];
		if (model.activeState?.getManipulator?.()) {
			worldCoords = manipulator.handleEvent(calldata, model._apiSpecificRenderWindow).worldCoords;
			add(center, subtract(worldCoords, previousPosition, []), newOrigin);
		} else if (otherLineHandle) {
			const otherLineVector = otherLineHandle.getDirection();
			normalize(otherLineVector);
			const axisTranslation = otherLineVector;
			const dot$1 = dot(currentLineVector, otherLineVector);
			if (dot$1 === 1 || dot$1 === -1) cross(currentLineVector, manipulator.getWidgetNormal(), axisTranslation);
			const closestPoint = [];
			worldCoords = manipulator.handleEvent(calldata, model._apiSpecificRenderWindow).worldCoords;
			Line_default.distanceToLine(worldCoords, lineHandle.getOrigin(), pointOnLine, closestPoint);
			newOrigin = multiplyAccumulate(center, axisTranslation, dot(subtract(worldCoords, closestPoint, []), axisTranslation), newOrigin);
		}
		newOrigin = publicAPI.getBoundedCenter(newOrigin);
		model.widgetState.setCenter(newOrigin);
		updateState(model.widgetState, model._factory.getScaleInPixels(), model._factory.getRotationHandlePosition());
		previousPosition = worldCoords;
	};
	publicAPI.getBoundedCenter = (newCenter) => {
		const oldCenter = model.widgetState.getCenter();
		const imageBounds = model.widgetState.getImage().getBounds();
		if (BoundingBox_default.containsPoint(imageBounds, ...newCenter)) return newCenter;
		return boundPointOnPlane(newCenter, oldCenter, imageBounds);
	};
	publicAPI[InteractionMethodsName.TranslateCenter] = (calldata) => {
		const { worldCoords } = (model.activeState?.getManipulator?.() ?? model.manipulator).handleEvent(calldata, model._apiSpecificRenderWindow);
		const translation = subtract(worldCoords, previousPosition, []);
		previousPosition = worldCoords;
		let newCenter = add(model.widgetState.getCenter(), translation, []);
		newCenter = publicAPI.getBoundedCenter(newCenter);
		model.widgetState.setCenter(newCenter);
		updateState(model.widgetState, model._factory.getScaleInPixels(), model._factory.getRotationHandlePosition());
	};
	publicAPI[InteractionMethodsName.RotateLine] = (calldata) => {
		const activeLineHandle = publicAPI.getActiveLineHandle();
		const manipulator = model.activeState?.getManipulator?.() ?? model.manipulator;
		const planeNormal = manipulator.getWidgetNormal();
		const { worldCoords } = manipulator.handleEvent(calldata, model._apiSpecificRenderWindow);
		if (!worldCoords || !worldCoords.length) return;
		const center = model.widgetState.getCenter();
		const currentVectorToOrigin = [
			0,
			0,
			0
		];
		subtract(worldCoords, center, currentVectorToOrigin);
		normalize(currentVectorToOrigin);
		const previousLineDirection = activeLineHandle.getDirection();
		normalize(previousLineDirection);
		const activePointName = publicAPI.getActiveRotationPointName();
		if (activePointName === "point1" || !activePointName && dot(currentVectorToOrigin, previousLineDirection) < 0) multiplyScalar(previousLineDirection, -1);
		const radianAngle = signedAngleBetweenVectors(previousLineDirection, currentVectorToOrigin, planeNormal);
		publicAPI.rotateLineInView(publicAPI.getActiveLineName(), radianAngle);
	};
	/**
	* Rotate a line by a specified angle
	* @param {string} lineName The line name to rotate (e.g. YinX, ZinX, XinY, ZinY, XinZ, YinZ)
	* @param {Number} radianAngle Applied angle in radian
	*/
	publicAPI.rotateLineInView = (lineName, radianAngle) => {
		const viewType = planeNameToViewType[getLinePlaneName(lineName)];
		const inViewType = planeNameToViewType[getLineInPlaneName(lineName)];
		const planeNormal = model.widgetState.getPlanes()[inViewType].normal;
		publicAPI.rotatePlane(viewType, radianAngle, planeNormal);
		if (publicAPI.getKeepOrthogonality()) {
			const otherPlaneName = getLinePlaneName(getOtherLineName(model.widgetState, lineName));
			publicAPI.rotatePlane(planeNameToViewType[otherPlaneName], radianAngle, planeNormal);
		}
		updateState(model.widgetState, model._factory.getScaleInPixels(), model._factory.getRotationHandlePosition());
	};
	/**
	* Rotate a specified plane around an other specified plane.
	* @param {ViewTypes} viewType Define which plane will be rotated
	* @param {Number} radianAngle Applied angle in radian
	* @param {vec3} planeNormal Define the axis to rotate around
	*/
	publicAPI.rotatePlane = (viewType, radianAngle, planeNormal) => {
		const { normal, viewUp } = model.widgetState.getPlanes()[viewType];
		const newNormal = rotateVector(normal, planeNormal, radianAngle);
		const newViewUp = rotateVector(viewUp, planeNormal, radianAngle);
		model.widgetState.getPlanes()[viewType] = {
			normal: newNormal,
			viewUp: newViewUp
		};
	};
	/**
	* Rotate a line by a specified angle
	* @param {string} lineName The line name to rotate (e.g. YinX, ZinX, XinY, ZinY, XinZ, YinZ)
	* @param {Number} radianAngle Applied angle in radian
	*/
	publicAPI.setViewPlane = (viewType, normal, viewUp) => {
		let newViewUp = viewUp;
		if (newViewUp == null) newViewUp = model.widgetState.getPlanes()[viewType].viewUp;
		model.widgetState.getPlanes()[viewType] = {
			normal,
			viewUp: newViewUp
		};
		updateState(model.widgetState, model._factory.getScaleInPixels(), model._factory.getRotationHandlePosition());
	};
}
//#endregion
export { widgetBehavior as default };
