// @kitware/vtk.js/Widgets/Widgets3D/ResliceCursorWidget/Constants.js
import { ViewTypes } from "../../Core/WidgetManager/Constants.js";
//#region Sources/Widgets/Widgets3D/ResliceCursorWidget/Constants.js
var ScrollingMethods = {
	MIDDLE_MOUSE_BUTTON: 0,
	LEFT_MOUSE_BUTTON: 1,
	RIGHT_MOUSE_BUTTON: 2
};
var InteractionMethodsName = {
	TranslateAxis: "translateAxis",
	RotateLine: "rotateLine",
	TranslateCenter: "translateCenter",
	TranslateCenterAndUpdatePlanes: "translateCenterAndUpdatePlanes"
};
var defaultViewUpFromViewType = {
	[ViewTypes.YZ_PLANE]: [
		0,
		0,
		1
	],
	[ViewTypes.XZ_PLANE]: [
		0,
		0,
		1
	],
	[ViewTypes.XY_PLANE]: [
		0,
		-1,
		0
	]
};
var xyzToViewType = [
	ViewTypes.YZ_PLANE,
	ViewTypes.XZ_PLANE,
	ViewTypes.XY_PLANE
];
var viewTypeToXYZ = {
	[ViewTypes.YZ_PLANE]: 0,
	[ViewTypes.XZ_PLANE]: 1,
	[ViewTypes.XY_PLANE]: 2
};
var planeNames = [
	"X",
	"Y",
	"Z"
];
var viewTypeToPlaneName = {
	[ViewTypes.YZ_PLANE]: "X",
	[ViewTypes.XZ_PLANE]: "Y",
	[ViewTypes.XY_PLANE]: "Z"
};
var planeNameToViewType = {
	X: ViewTypes.YZ_PLANE,
	Y: ViewTypes.XZ_PLANE,
	Z: ViewTypes.XY_PLANE
};
var lineNames = [
	"YinX",
	"ZinX",
	"XinY",
	"ZinY",
	"XinZ",
	"YinZ"
];
var Constants_default = {
	ScrollingMethods,
	InteractionMethodsName,
	xyzToViewType,
	viewTypeToXYZ,
	planeNames,
	viewTypeToPlaneName,
	planeNameToViewType,
	lineNames
};
//#endregion
export { InteractionMethodsName, ScrollingMethods, Constants_default as default, defaultViewUpFromViewType, lineNames, planeNameToViewType, planeNames, viewTypeToPlaneName, viewTypeToXYZ, xyzToViewType };
