// @kitware/vtk.js/Rendering/Core.js
import RenderWindowInteractor_default from "./Core/RenderWindowInteractor.js";
import InteractorObserver_default from "./Core/InteractorObserver.js";
import InteractorStyle_default from "./Core/InteractorStyle.js";
import Prop_default from "./Core/Prop.js";
import Prop3D_default from "./Core/Prop3D.js";
import Property_default from "./Core/Property.js";
import Actor_default from "./Core/Actor.js";
import AbstractMapper_default from "./Core/AbstractMapper.js";
import AbstractMapper3D_default from "./Core/AbstractMapper3D.js";
import Mapper_default from "./Core/Mapper.js";
import AbstractPicker_default from "./Core/AbstractPicker.js";
import Picker_default from "./Core/Picker.js";
import PointPicker_default from "./Core/PointPicker.js";
import Camera_default from "./Core/Camera.js";
import Light_default from "./Core/Light.js";
import Viewport_default from "./Core/Viewport.js";
import Renderer_default from "./Core/Renderer.js";
import ColorTransferFunction_default from "./Core/ColorTransferFunction.js";
import VolumeProperty_default from "./Core/VolumeProperty.js";
import Volume_default from "./Core/Volume.js";
import Texture_default from "./Core/Texture.js";
import VolumeMapper_default from "./Core/VolumeMapper.js";
import ColorMaps_default from "./Core/ColorTransferFunction/ColorMaps.js";
import Coordinate_default from "./Core/Coordinate.js";
import Property2D_default from "./Core/Property2D.js";
import Actor2D_default from "./Core/Actor2D.js";
import AnnotatedCubeActor_default from "./Core/AnnotatedCubeActor.js";
import AxesActor_default from "./Core/AxesActor.js";
import CellPicker_default from "./Core/CellPicker.js";
import CutterMapper_default from "./Core/CutterMapper.js";
import CubeAxesActor_default from "./Core/CubeAxesActor.js";
import Follower_default from "./Core/Follower.js";
import Glyph3DMapper_default from "./Core/Glyph3DMapper.js";
import HardwareSelector_default from "./Core/HardwareSelector.js";
import ImageMapper_default from "./Core/ImageMapper.js";
import ImageProperty_default from "./Core/ImageProperty.js";
import ImageResliceMapper_default from "./Core/ImageResliceMapper.js";
import ImageSlice_default from "./Core/ImageSlice.js";
import Mapper2D_default from "./Core/Mapper2D.js";
import PixelSpaceCallbackMapper_default from "./Core/PixelSpaceCallbackMapper.js";
import RenderWindow_default from "./Core/RenderWindow.js";
import ScalarBarActor_default from "./Core/ScalarBarActor.js";
import Skybox_default from "./Core/Skybox.js";
import SphereMapper_default from "./Core/SphereMapper.js";
import StickMapper_default from "./Core/StickMapper.js";
import TextProperty_default from "./Core/TextProperty.js";
import TextActor_default from "./Core/TextActor.js";
import VectorText_default from "./Core/VectorText.js";
//#region Sources/Rendering/Core/index.js
var Core_default = {
	vtkAbstractMapper: AbstractMapper_default,
	vtkAbstractMapper3D: AbstractMapper3D_default,
	vtkAbstractPicker: AbstractPicker_default,
	vtkActor: Actor_default,
	vtkActor2D: Actor2D_default,
	vtkAnnotatedCubeActor: AnnotatedCubeActor_default,
	vtkAxesActor: AxesActor_default,
	vtkCamera: Camera_default,
	vtkCellPicker: CellPicker_default,
	vtkColorTransferFunction: {
		vtkColorMaps: ColorMaps_default,
		...ColorTransferFunction_default
	},
	vtkCoordinate: Coordinate_default,
	vtkCutterMapper: CutterMapper_default,
	vtkCubeAxesActor: CubeAxesActor_default,
	vtkFollower: Follower_default,
	vtkGlyph3DMapper: Glyph3DMapper_default,
	vtkHardwareSelector: HardwareSelector_default,
	vtkImageMapper: ImageMapper_default,
	vtkImageProperty: ImageProperty_default,
	vtkImageResliceMapper: ImageResliceMapper_default,
	vtkImageSlice: ImageSlice_default,
	vtkInteractorObserver: InteractorObserver_default,
	vtkInteractorStyle: InteractorStyle_default,
	vtkLight: Light_default,
	vtkMapper: Mapper_default,
	vtkMapper2D: Mapper2D_default,
	vtkPicker: Picker_default,
	vtkPixelSpaceCallbackMapper: PixelSpaceCallbackMapper_default,
	vtkPointPicker: PointPicker_default,
	vtkProp: Prop_default,
	vtkProp3D: Prop3D_default,
	vtkProperty: Property_default,
	vtkProperty2D: Property2D_default,
	vtkRenderer: Renderer_default,
	vtkRenderWindow: RenderWindow_default,
	vtkRenderWindowInteractor: RenderWindowInteractor_default,
	vtkScalarBarActor: ScalarBarActor_default,
	vtkSkybox: Skybox_default,
	vtkSphereMapper: SphereMapper_default,
	vtkStickMapper: StickMapper_default,
	vtkTexture: Texture_default,
	vtkTextActor: TextActor_default,
	vtkTextProperty: TextProperty_default,
	vtkVector: VectorText_default,
	vtkViewport: Viewport_default,
	vtkVolume: Volume_default,
	vtkVolumeMapper: VolumeMapper_default,
	vtkVolumeProperty: VolumeProperty_default
};
//#endregion
export { Core_default as default };
