// @kitware/vtk.js/Rendering/OpenGL.js
import ViewNodeFactory_default from "./OpenGL/ViewNodeFactory.js";
import Texture_default from "./OpenGL/Texture.js";
import Framebuffer_default from "./OpenGL/Framebuffer.js";
import BufferObject_default from "./OpenGL/BufferObject.js";
import CellArrayBufferObject_default from "./OpenGL/CellArrayBufferObject.js";
import Shader_default from "./OpenGL/Shader.js";
import ShaderProgram_default from "./OpenGL/ShaderProgram.js";
import VertexArrayObject_default from "./OpenGL/VertexArrayObject.js";
import Helper_default from "./OpenGL/Helper.js";
import ForwardPass_default from "./OpenGL/ForwardPass.js";
import HardwareSelector_default from "./OpenGL/HardwareSelector.js";
import ShaderCache_default from "./OpenGL/ShaderCache.js";
import TextureUnitManager_default from "./OpenGL/TextureUnitManager.js";
import RenderWindow_default from "./OpenGL/RenderWindow.js";
import Actor_default from "./OpenGL/Actor.js";
import Actor2D_default from "./OpenGL/Actor2D.js";
import Camera_default from "./OpenGL/Camera.js";
import Convolution2DPass_default from "./OpenGL/Convolution2DPass.js";
import "./OpenGL/CubeAxesActor.js";
import PolyDataMapper_default from "./OpenGL/PolyDataMapper.js";
import CutterMapper_default from "./OpenGL/CutterMapper.js";
import Glyph3DMapper_default from "./OpenGL/Glyph3DMapper.js";
import ImageMapper_default from "./OpenGL/ImageMapper.js";
import ImageSlice_default from "./OpenGL/ImageSlice.js";
import PixelSpaceCallbackMapper_default from "./OpenGL/PixelSpaceCallbackMapper.js";
import PolyDataMapper2D_default from "./OpenGL/PolyDataMapper2D.js";
import Renderer_default from "./OpenGL/Renderer.js";
import ImageResliceMapper_default from "./OpenGL/ImageResliceMapper.js";
import "./OpenGL/ScalarBarActor.js";
import Skybox_default from "./OpenGL/Skybox.js";
import SphereMapper_default from "./OpenGL/SphereMapper.js";
import StickMapper_default from "./OpenGL/StickMapper.js";
import Volume_default from "./OpenGL/Volume.js";
import VolumeMapper_default from "./OpenGL/VolumeMapper.js";
//#region Sources/Rendering/OpenGL/index.js
var OpenGL_default = {
	vtkActor: Actor_default,
	vtkActor2D: Actor2D_default,
	vtkBufferObject: BufferObject_default,
	vtkCamera: Camera_default,
	vtkCellArrayBufferObject: CellArrayBufferObject_default,
	vtkCutterMapper: CutterMapper_default,
	vtkConvolution2DPass: Convolution2DPass_default,
	vtkForwardPass: ForwardPass_default,
	vtkFramebuffer: Framebuffer_default,
	vtkGlyph3DMapper: Glyph3DMapper_default,
	vtkHardwareSelector: HardwareSelector_default,
	vtkHelper: Helper_default,
	vtkImageMapper: ImageMapper_default,
	vtkImageSlice: ImageSlice_default,
	vtkPixelSpaceCallbackMapper: PixelSpaceCallbackMapper_default,
	vtkPolyDataMapper: PolyDataMapper_default,
	vtkPolyDataMapper2D: PolyDataMapper2D_default,
	vtkRenderWindow: RenderWindow_default,
	vtkRenderer: Renderer_default,
	vtkImageResliceMapper: ImageResliceMapper_default,
	vtkShader: Shader_default,
	vtkShaderCache: ShaderCache_default,
	vtkShaderProgram: ShaderProgram_default,
	vtkSkybox: Skybox_default,
	vtkSphereMapper: SphereMapper_default,
	vtkStickMapper: StickMapper_default,
	vtkTexture: Texture_default,
	vtkTextureUnitManager: TextureUnitManager_default,
	vtkVertexArrayObject: VertexArrayObject_default,
	vtkViewNodeFactory: ViewNodeFactory_default,
	vtkVolume: Volume_default,
	vtkVolumeMapper: VolumeMapper_default
};
//#endregion
export { OpenGL_default as default };
