// @kitware/vtk.js/Interaction/Manipulators/CompositeMouseManipulator.js
import macros_default from "../../macros.js";
//#region Sources/Interaction/Manipulators/CompositeMouseManipulator/index.js
function vtkCompositeMouseManipulator(publicAPI, model) {
	model.classHierarchy.push("vtkCompositeMouseManipulator");
	publicAPI.startInteraction = () => {};
	publicAPI.endInteraction = () => {};
	publicAPI.onButtonDown = (interactor, renderer, position) => {};
	publicAPI.onButtonUp = (interactor) => {};
	publicAPI.onMouseMove = (interactor, renderer, position) => {};
	publicAPI.onStartScroll = (interactor, renderer, delta) => {};
	publicAPI.onScroll = (interactor, renderer, delta) => {};
	publicAPI.onEndScroll = (interactor) => {};
	publicAPI.isDragEnabled = () => model.dragEnabled;
	publicAPI.isScrollEnabled = () => model.scrollEnabled;
}
var DEFAULT_VALUES = {
	button: 1,
	shift: false,
	control: false,
	alt: false,
	dragEnabled: true,
	scrollEnabled: false
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.setGet(publicAPI, model, [
		"button",
		"shift",
		"control",
		"alt"
	]);
	macros_default.set(publicAPI, model, ["dragEnabled", "scrollEnabled"]);
	vtkCompositeMouseManipulator(publicAPI, model);
}
var CompositeMouseManipulator_default = { extend };
//#endregion
export { CompositeMouseManipulator_default as default, extend };
