// @kitware/vtk.js/Interaction/Manipulators/KeyboardCameraManipulator.js
import macros_default from "../../macros.js";
import { add, cross, normalize } from "../../Common/Core/Math.js";
import BoundingBox_default from "../../Common/DataModel/BoundingBox.js";
import CompositeKeyboardManipulator_default from "./CompositeKeyboardManipulator.js";
//#region Sources/Interaction/Manipulators/KeyboardCameraManipulator/index.js
var { vtkErrorMacro } = macros_default;
var ANIMATION_REQUESTER = "vtkKeyboardCameraManipulator";
function vtkKeyboardCameraManipulator(publicAPI, model) {
	model.classHierarchy.push("vtkKeyboardCameraManipulator");
	const internal = {
		keysDown: [],
		direction: [
			0,
			0,
			0
		],
		skipUpdateDirection: false,
		animationSub: null,
		cameraModifiedSub: null
	};
	publicAPI.inMotion = () => internal.animationSub !== null;
	publicAPI.resetMovementSpeed = () => {
		const { renderer } = model;
		const bounds = renderer.computeVisiblePropBounds();
		model.movementSpeed = BoundingBox_default.getMaxLength(bounds) / 500;
	};
	publicAPI.startMovement = () => {
		if (publicAPI.inMotion()) {
			vtkErrorMacro("Camera is already in motion!");
			return;
		}
		if (model.movementSpeed === null) publicAPI.resetMovementSpeed();
		const { interactor, renderer } = model;
		const move = () => {
			if (internal.keysDown.length === 0) return;
			internal.skipUpdateDirection = true;
			publicAPI.moveCamera(renderer.getActiveCamera(), internal.direction, model.movementSpeed);
			renderer.resetCameraClippingRange();
			if (interactor.getLightFollowCamera()) renderer.updateLightsGeometryToFollowCamera();
			internal.skipUpdateDirection = false;
		};
		publicAPI.calculateCurrentDirection();
		internal.cameraModifiedSub = renderer.getActiveCamera().onModified(publicAPI.calculateCurrentDirection);
		interactor.requestAnimation(ANIMATION_REQUESTER);
		internal.animationSub = interactor.onAnimation(() => move());
	};
	publicAPI.endMovement = () => {
		if (internal.animationSub) {
			internal.animationSub.unsubscribe();
			internal.animationSub = null;
		}
		model._interactor.cancelAnimation(ANIMATION_REQUESTER);
		if (internal.cameraModifiedSub) {
			internal.cameraModifiedSub.unsubscribe();
			internal.cameraModifiedSub = null;
		}
	};
	publicAPI.calculateCurrentDirection = () => {
		if (internal.skipUpdateDirection) return;
		internal.direction = [
			0,
			0,
			0
		];
		const { renderer } = model;
		if (!renderer) return;
		const camera = renderer.getActiveCamera();
		if (!camera) return;
		if (internal.keysDown.length === 0) return;
		let directions = internal.keysDown.map((key) => publicAPI.getDirectionFromKey(key, camera));
		directions = directions.filter((item) => item);
		if (directions.length === 0) return;
		const netDirection = directions.reduce((a, b) => {
			add(a, b, b);
			return b;
		});
		normalize(netDirection);
		internal.direction = netDirection;
	};
	publicAPI.getDirectionFromKey = (key, camera) => {
		let direction;
		if (model.moveForwardKeys.includes(key)) direction = camera.getDirectionOfProjection();
		else if (model.moveLeftKeys.includes(key)) {
			const dirProj = camera.getDirectionOfProjection();
			direction = [
				0,
				0,
				0
			];
			cross(camera.getViewUp(), dirProj, direction);
		} else if (model.moveBackwardKeys.includes(key)) direction = camera.getDirectionOfProjection().map((e) => -e);
		else if (model.moveRightKeys.includes(key)) {
			const dirProj = camera.getDirectionOfProjection();
			direction = [
				0,
				0,
				0
			];
			cross(dirProj, camera.getViewUp(), direction);
		} else if (model.moveUpKeys.includes(key)) direction = camera.getViewUp();
		else if (model.moveDownKeys.includes(key)) direction = camera.getViewUp().map((e) => -e);
		else return;
		normalize(direction);
		return direction;
	};
	publicAPI.moveCamera = (camera, direction, speed) => {
		const position = camera.getPosition();
		const focalPoint = camera.getFocalPoint();
		camera.setFocalPoint(focalPoint[0] + direction[0] * speed, focalPoint[1] + direction[1] * speed, focalPoint[2] + direction[2] * speed);
		camera.setPosition(position[0] + direction[0] * speed, position[1] + direction[1] * speed, position[2] + direction[2] * speed);
	};
	publicAPI.onKeyPress = (interactor, renderer, key) => {};
	publicAPI.onKeyDown = (interactor, renderer, key) => {
		if (!internal.keysDown.includes(key)) {
			internal.keysDown.push(key);
			publicAPI.calculateCurrentDirection();
		}
		if (!publicAPI.inMotion()) {
			Object.assign(model, {
				interactor,
				renderer
			});
			publicAPI.startMovement();
		}
	};
	publicAPI.onKeyUp = (interactor, renderer, key) => {
		internal.keysDown = internal.keysDown.filter((item) => item.toUpperCase() !== key.toUpperCase());
		publicAPI.calculateCurrentDirection();
		if (internal.keysDown.length === 0) publicAPI.endMovement();
	};
}
var DEFAULT_VALUES = {
	movementSpeed: null,
	moveForwardKeys: [
		"w",
		"W",
		"ArrowUp"
	],
	moveLeftKeys: [
		"a",
		"A",
		"ArrowLeft"
	],
	moveBackwardKeys: [
		"s",
		"S",
		"ArrowDown"
	],
	moveRightKeys: [
		"d",
		"D",
		"ArrowRight"
	],
	moveUpKeys: [" "],
	moveDownKeys: ["Shift"],
	interactor: null,
	renderer: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	CompositeKeyboardManipulator_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, [
		"movementSpeed",
		"moveForwardKeys",
		"moveLeftKeys",
		"moveBackwardKeys",
		"moveRightKeys",
		"moveUpKeys",
		"moveDownKeys",
		"interactor",
		"renderer"
	]);
	vtkKeyboardCameraManipulator(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkKeyboardCameraManipulator");
var KeyboardCameraManipulator_default = {
	newInstance,
	extend
};
//#endregion
export { KeyboardCameraManipulator_default as default, extend, newInstance };
