// @kitware/vtk.js/Interaction/Manipulators/MouseCameraTrackballFirstPersonManipulator.js
import macros_default from "../../macros.js";
import CompositeCameraManipulator_default from "./CompositeCameraManipulator.js";
import CompositeMouseManipulator_default from "./CompositeMouseManipulator.js";
//#region Sources/Interaction/Manipulators/MouseCameraTrackballFirstPersonManipulator/index.js
var ANIMATION_REQUESTER = "vtkMouseCameraTrackballFirstPersonManipulator";
function vtkMouseCameraTrackballFirstPersonManipulator(publicAPI, model) {
	model.classHierarchy.push("vtkMouseCameraTrackballFirstPersonManipulator");
	const internal = {
		interactor: null,
		renderer: null
	};
	publicAPI.onButtonDown = (interactor, renderer, position) => {
		if (model.usePointerLock && !interactor.isPointerLocked()) {
			Object.assign(internal, {
				interactor,
				renderer
			});
			Promise.resolve(interactor.requestPointerLock()).then(() => {
				publicAPI.startPointerLockInteraction();
			});
		}
	};
	publicAPI.onMouseMove = (interactor, renderer, position) => {
		if (!position) return;
		const sensitivity = model.sensitivity;
		const yaw = -position.movementX * sensitivity;
		const pitch = -position.movementY * sensitivity;
		Object.assign(internal, {
			interactor,
			renderer
		});
		publicAPI.moveCamera(yaw, pitch);
	};
	publicAPI.moveCamera = (yaw, pitch) => {
		const { renderer, interactor } = internal;
		const camera = renderer.getActiveCamera();
		const numSteps = model.numAnimationSteps;
		const yawStep = yaw / numSteps;
		const pitchStep = pitch / numSteps;
		const animationRequester = `${ANIMATION_REQUESTER}.${performance.now().toString()}`;
		let curStep = 0;
		let animationSub = null;
		const performStep = () => {
			camera.yaw(yawStep);
			camera.pitch(pitchStep);
			camera.orthogonalizeViewUp();
			curStep += 1;
			if (curStep === numSteps) {
				animationSub.unsubscribe();
				renderer.resetCameraClippingRange();
				if (interactor.getLightFollowCamera()) renderer.updateLightsGeometryToFollowCamera();
				const cancelRequest = () => {
					internal.interactor.cancelAnimation(animationRequester);
				};
				setTimeout(cancelRequest, 0);
			}
		};
		interactor.requestAnimation(animationRequester);
		animationSub = interactor.onAnimation(() => performStep());
	};
}
var DEFAULT_VALUES = {
	numAnimationSteps: 5,
	sensitivity: .05,
	usePointerLock: true
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	CompositeCameraManipulator_default.extend(publicAPI, model, initialValues);
	CompositeMouseManipulator_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, [
		"numAnimationSteps",
		"sensitivity",
		"usePointerLock"
	]);
	vtkMouseCameraTrackballFirstPersonManipulator(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkMouseCameraTrackballFirstPersonManipulator");
var MouseCameraTrackballFirstPersonManipulator_default = {
	newInstance,
	extend
};
//#endregion
export { MouseCameraTrackballFirstPersonManipulator_default as default, extend, newInstance };
