// @kitware/vtk.js/Interaction/Manipulators/MouseCameraUnicamManipulator.js
import macros_default from "../../macros.js";
import { cross, dot, multiplyScalar, normalize, radiansFromDegrees, subtract } from "../../Common/Core/Math.js";
import CompositeCameraManipulator_default from "./CompositeCameraManipulator.js";
import CompositeMouseManipulator_default from "./CompositeMouseManipulator.js";
import Constants_default from "../../Rendering/Core/InteractorStyle/Constants.js";
import MouseCameraUnicamRotateManipulator_default from "./MouseCameraUnicamRotateManipulator.js";
//#region Sources/Interaction/Manipulators/MouseCameraUnicamManipulator/index.js
var { States } = Constants_default;
function vtkMouseCameraUnicamManipulator(publicAPI, model) {
	model.classHierarchy.push("vtkMouseCameraUnicamManipulator");
	model.state = States.IS_NONE;
	model.rotateManipulator = MouseCameraUnicamRotateManipulator_default.newInstance({
		button: model.button,
		shift: model.shift,
		control: model.control,
		alt: model.alt,
		dragEnabled: model.dragEnabled,
		scrollEnabled: model.scrollEnabled,
		displayFocusSphereOnButtonDown: false
	});
	const normalize$1 = (position, interactor) => {
		const renderer = interactor.findPokedRenderer();
		const [width, height] = interactor.getView().getViewportSize(renderer);
		return {
			x: -1 + 2 * position.x / width,
			y: -1 + 2 * position.y / height
		};
	};
	const getRightVAndUpV = (downPoint, interactor) => {
		const camera = interactor.findPokedRenderer().getActiveCamera();
		const cameraPosition = camera.getPosition();
		const cameraToPointVec = [
			0,
			0,
			0
		];
		subtract(downPoint, cameraPosition, cameraToPointVec);
		if (camera.getParallelProjection()) multiplyScalar(cameraToPointVec, camera.getParallelScale());
		const atV = camera.getViewPlaneNormal();
		normalize(atV);
		const l = dot(cameraToPointVec, atV);
		const viewAngle = radiansFromDegrees(camera.getViewAngle());
		const renderer = interactor.findPokedRenderer();
		const [width, height] = interactor.getView().getViewportSize(renderer);
		const scaleX = width / height * (2 * l * Math.tan(viewAngle / 2) / 2);
		const scaleY = 2 * l * Math.tan(viewAngle / 2) / 2;
		const upV = camera.getViewUp();
		const rightV = [];
		cross(upV, atV, rightV);
		cross(atV, rightV, upV);
		normalize(rightV);
		normalize(upV);
		multiplyScalar(rightV, scaleX);
		multiplyScalar(upV, scaleY);
		return {
			rightV,
			upV
		};
	};
	const choose = (interactor, position) => {
		const normalizedPosition = normalize$1(position, interactor);
		const normalizedPreviousPosition = normalize$1(model.previousPosition, interactor);
		const delta = {
			x: normalizedPosition.x - normalizedPreviousPosition.x,
			y: normalizedPosition.y - normalizedPreviousPosition.y
		};
		model.previousPosition = position;
		const deltaT = Date.now() / 1e3 - model.time;
		model.dist += Math.sqrt(delta.x ** 2 + delta.y ** 2);
		const sDelta = {
			x: position.x - model.startPosition.x,
			y: position.y - model.startPosition.y
		};
		const len = Math.sqrt(sDelta.x ** 2 + sDelta.y ** 2);
		if (Math.abs(sDelta.y) / len > .9 && deltaT > .05) model.state = States.IS_DOLLY;
		else if (deltaT >= .1 || model.dist >= .03) if (Math.abs(sDelta.x) / len > .6) model.state = States.IS_PAN;
		else model.state = States.IS_DOLLY;
	};
	const pan = (interactor, position) => {
		const renderer = interactor.findPokedRenderer();
		const normalizedPosition = normalize$1(position, interactor);
		const normalizedPreviousPosition = normalize$1(model.previousPosition, interactor);
		const delta = {
			x: normalizedPosition.x - normalizedPreviousPosition.x,
			y: normalizedPosition.y - normalizedPreviousPosition.y
		};
		const camera = renderer.getActiveCamera();
		model.previousPosition = position;
		const { rightV, upV } = getRightVAndUpV(model.downPoint, interactor);
		const offset = [];
		for (let index = 0; index < 3; index++) offset[index] = delta.x * rightV[index] + delta.y * upV[index];
		camera.translate(...offset);
		renderer.resetCameraClippingRange();
		interactor.render();
	};
	const dolly = (interactor, position) => {
		const renderer = interactor.findPokedRenderer();
		const normalizedPosition = normalize$1(position, interactor);
		const normalizedPreviousPosition = normalize$1(model.previousPosition, interactor);
		const delta = {
			x: normalizedPosition.x - normalizedPreviousPosition.x,
			y: normalizedPosition.y - normalizedPreviousPosition.y
		};
		const camera = renderer.getActiveCamera();
		const cameraPosition = camera.getPosition();
		if (camera.getParallelProjection()) camera.zoom(1 - delta.y);
		else {
			const offset1 = [];
			subtract(model.downPoint, cameraPosition, offset1);
			multiplyScalar(offset1, delta.y * -4);
			camera.translate(...offset1);
		}
		const { rightV: offset2 } = getRightVAndUpV(model.downPoint, interactor);
		multiplyScalar(offset2, delta.x);
		camera.translate(...offset2);
		renderer.resetCameraClippingRange();
		interactor.render();
	};
	publicAPI.onButtonDown = (interactor, renderer, position) => {
		model.buttonPressed = true;
		model.startPosition = position;
		model.previousPosition = position;
		model.time = Date.now() / 1e3;
		model.dist = 0;
		model.rotateManipulator.onButtonDown(interactor, renderer, position);
		model.downPoint = model.rotateManipulator.getDownPoint();
	};
	publicAPI.onMouseMove = (interactor, renderer, position) => {
		if (!model.buttonPressed) return;
		if (model.rotateManipulator.getState() === States.IS_ROTATE) model.rotateManipulator.onMouseMove(interactor, renderer, position);
		else switch (model.state) {
			case States.IS_NONE:
				choose(interactor, position);
				break;
			case States.IS_PAN:
				pan(interactor, position);
				break;
			case States.IS_DOLLY:
				dolly(interactor, position);
				break;
			default: break;
		}
		model.previousPosition = position;
	};
	publicAPI.onButtonUp = (interactor) => {
		model.buttonPressed = false;
		if (model.state === States.IS_NONE) model.rotateManipulator.onButtonUp(interactor);
		model.state = States.IS_NONE;
	};
	publicAPI.getUseWorldUpVec = () => model.rotateManipulator.getUseWorldUpVec();
	publicAPI.setUseWorldUpVec = (useWorldUpVec) => {
		model.rotateManipulator.setUseWorldUpVec(useWorldUpVec);
	};
	publicAPI.getWorldUpVec = () => model.rotateManipulator.getWorldUpVec();
	publicAPI.setWorldUpVec = (x, y, z) => {
		model.rotateManipulator.setWorldUpVec(x, y, z);
	};
	publicAPI.getUseHardwareSelector = () => model.rotateManipulator.getUseHardwareSelector();
	publicAPI.setUseHardwareSelector = (useHardwareSelector) => {
		model.rotateManipulator.setUseHardwareSelector(useHardwareSelector);
	};
	publicAPI.getFocusSphereColor = () => {
		model.rotateManipulator.getFocusSphereColor();
	};
	publicAPI.setFocusSphereColor = (r, g, b) => {
		model.rotateManipulator.setFocusSphereColor(r, g, b);
	};
	publicAPI.getFocusSphereRadiusFactor = () => model.rotateManipulator.getFocusSphereRadiusFactor();
	publicAPI.setFocusSphereRadiusFactor = (focusSphereRadiusFactor) => {
		model.rotateManipulator.setFocusSphereRadiusFactor(focusSphereRadiusFactor);
	};
}
var DEFAULT_VALUES = {};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	CompositeCameraManipulator_default.extend(publicAPI, model, initialValues);
	CompositeMouseManipulator_default.extend(publicAPI, model, initialValues);
	vtkMouseCameraUnicamManipulator(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkMouseCameraUnicamManipulator");
var MouseCameraUnicamManipulator_default = {
	newInstance,
	extend
};
//#endregion
export { MouseCameraUnicamManipulator_default as default, extend, newInstance };
