// @kitware/vtk.js/Interaction/Manipulators/MouseCameraTrackballRotateManipulator.js
import macros_default from "../../macros.js";
import { add, cross, dot, multiplyScalar, radiansFromDegrees } from "../../Common/Core/Math.js";
import CompositeCameraManipulator_default from "./CompositeCameraManipulator.js";
import CompositeMouseManipulator_default from "./CompositeMouseManipulator.js";
import { mat4, vec3 } from "gl-matrix";
//#region Sources/Interaction/Manipulators/MouseCameraTrackballRotateManipulator/index.js
function vtkMouseCameraTrackballRotateManipulator(publicAPI, model) {
	model.classHierarchy.push("vtkMouseCameraTrackballRotateManipulator");
	const newCamPos = new Float64Array(3);
	const newFp = new Float64Array(3);
	const newViewUp = new Float64Array(3);
	const trans = new Float64Array(16);
	const v2 = new Float64Array(3);
	const centerNeg = new Float64Array(3);
	const direction = new Float64Array(3);
	publicAPI.onButtonDown = (interactor, renderer, position) => {
		model.previousPosition = position;
	};
	publicAPI.onMouseMove = (interactor, renderer, position) => {
		if (!position) return;
		const camera = renderer.getActiveCamera();
		const cameraPos = camera.getPosition();
		const cameraFp = camera.getFocalPoint();
		mat4.identity(trans);
		const { center, rotationFactor } = model;
		if (model.useFocalPointAsCenterOfRotation) {
			center[0] = cameraFp[0];
			center[1] = cameraFp[1];
			center[2] = cameraFp[2];
		}
		const dx = model.previousPosition.x - position.x;
		const dy = model.previousPosition.y - position.y;
		const size = interactor.getView().getViewportSize(renderer);
		const viewUp = camera.getViewUp();
		if (model.useWorldUpVec) {
			const centerOfRotation = new Float64Array(3);
			vec3.copy(centerOfRotation, model.worldUpVec);
			multiplyScalar(centerOfRotation, dot(cameraPos, model.worldUpVec) / dot(model.worldUpVec, model.worldUpVec));
			add(center, centerOfRotation, centerOfRotation);
			mat4.translate(trans, trans, centerOfRotation);
			mat4.rotate(trans, trans, radiansFromDegrees(360 * dx / size[0] * rotationFactor), model.worldUpVec);
			centerOfRotation[0] = -centerOfRotation[0];
			centerOfRotation[1] = -centerOfRotation[1];
			centerOfRotation[2] = -centerOfRotation[2];
			mat4.translate(trans, trans, centerOfRotation);
			mat4.translate(trans, trans, center);
		} else {
			mat4.translate(trans, trans, center);
			mat4.rotate(trans, trans, radiansFromDegrees(360 * dx / size[0] * rotationFactor), viewUp);
		}
		cross(camera.getDirectionOfProjection(), viewUp, v2);
		mat4.rotate(trans, trans, radiansFromDegrees(-360 * dy / size[1] * rotationFactor), v2);
		centerNeg[0] = -center[0];
		centerNeg[1] = -center[1];
		centerNeg[2] = -center[2];
		mat4.translate(trans, trans, centerNeg);
		vec3.transformMat4(newCamPos, cameraPos, trans);
		vec3.transformMat4(newFp, cameraFp, trans);
		direction[0] = viewUp[0] + cameraPos[0];
		direction[1] = viewUp[1] + cameraPos[1];
		direction[2] = viewUp[2] + cameraPos[2];
		vec3.transformMat4(newViewUp, direction, trans);
		camera.setPosition(newCamPos[0], newCamPos[1], newCamPos[2]);
		camera.setFocalPoint(newFp[0], newFp[1], newFp[2]);
		camera.setViewUp(newViewUp[0] - newCamPos[0], newViewUp[1] - newCamPos[1], newViewUp[2] - newCamPos[2]);
		camera.orthogonalizeViewUp();
		renderer.resetCameraClippingRange();
		if (interactor.getLightFollowCamera()) renderer.updateLightsGeometryToFollowCamera();
		model.previousPosition = position;
	};
}
var DEFAULT_VALUES = {
	useWorldUpVec: false,
	worldUpVec: [
		0,
		1,
		0
	],
	useFocalPointAsCenterOfRotation: false
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	CompositeMouseManipulator_default.extend(publicAPI, model, initialValues);
	CompositeCameraManipulator_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, ["useWorldUpVec"]);
	macros_default.setGetArray(publicAPI, model, ["worldUpVec"], 3);
	macros_default.setGet(publicAPI, model, ["useFocalPointAsCenterOfRotation"]);
	vtkMouseCameraTrackballRotateManipulator(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkMouseCameraTrackballRotateManipulator");
var MouseCameraTrackballRotateManipulator_default = {
	newInstance,
	extend
};
//#endregion
export { MouseCameraTrackballRotateManipulator_default as default, extend, newInstance };
