// @kitware/vtk.js/Interaction/Manipulators/MouseBoxSelectorManipulator.js
import { event, newInstance as newInstance$1, obj, setGet } from "../../macros.js";
import CompositeMouseManipulator_default from "./CompositeMouseManipulator.js";
//#region Sources/Interaction/Manipulators/MouseBoxSelectorManipulator/index.js
var OUTSIDE_BOUNDS = [
	-2,
	-1,
	-2,
	-1
];
var DEFAULT_STYLE = {
	position: "absolute",
	zIndex: 1,
	border: "2px solid #F44336",
	backgroundColor: "rgba(0, 0, 0, 0.1)",
	borderRadius: "4px",
	boxSizing: "border-box"
};
function applyStyle(element, style) {
	Object.keys(style).forEach((name) => {
		element.style[name] = style[name];
	});
}
function vtkMouseBoxSelectionManipulator(publicAPI, model) {
	model.classHierarchy.push("vtkMouseBoxSelectionManipulator");
	let view = null;
	let container = null;
	let previousPosition = null;
	let currentPosition = null;
	let div = null;
	let inDOM = false;
	function getBounds() {
		if (!previousPosition || !currentPosition) return OUTSIDE_BOUNDS;
		return [
			Math.min(previousPosition.x, currentPosition.x),
			Math.max(previousPosition.x, currentPosition.x),
			Math.min(previousPosition.y, currentPosition.y),
			Math.max(previousPosition.y, currentPosition.y)
		];
	}
	function applyStyleToDiv() {
		if (!view || !container) return;
		const [viewWidth, viewHeight] = view.getSize();
		const { width, height } = container.getBoundingClientRect();
		const [xMin, xMax, yMin, yMax] = getBounds();
		div.style.left = `${width * xMin / viewWidth}px`;
		div.style.top = `${height - height * yMax / viewHeight}px`;
		div.style.width = `${width * (xMax - xMin) / viewWidth}px`;
		div.style.height = `${height * (yMax - yMin) / viewHeight}px`;
	}
	publicAPI.onButtonDown = (interactor, renderer, position) => {
		previousPosition = position;
		if (model.renderSelection) {
			if (!view) view = interactor.getView();
			if (!container && view?.getContainer) container = view.getContainer();
			if (!container) container = model.container;
			if (!div) {
				div = document.createElement("div");
				applyStyle(div, model.selectionStyle);
			}
			applyStyleToDiv();
			if (container && !inDOM) {
				inDOM = true;
				container.appendChild(div);
			}
		}
	};
	publicAPI.onMouseMove = (interactor, renderer, position) => {
		if (!previousPosition) return;
		if (!position) return;
		currentPosition = position;
		publicAPI.invokeBoxSelectInput({
			view,
			container,
			selection: getBounds()
		});
		if (model.renderSelection) applyStyleToDiv();
	};
	publicAPI.onButtonUp = (interactor, renderer) => {
		if (!previousPosition || !currentPosition && !model.boxChangeOnClick) return;
		if (!currentPosition) currentPosition = previousPosition;
		publicAPI.invokeBoxSelectChange({
			view,
			container,
			selection: getBounds()
		});
		if (inDOM) {
			div.parentElement.removeChild(div);
			inDOM = false;
		}
		view = null;
		container = null;
		previousPosition = null;
		currentPosition = null;
	};
}
function DEFAULT_VALUES(initialValues) {
	return {
		boxChangeOnClick: false,
		renderSelection: true,
		...initialValues,
		selectionStyle: {
			...DEFAULT_STYLE,
			...initialValues.selectionStyle
		}
	};
}
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES(initialValues));
	obj(publicAPI, model);
	CompositeMouseManipulator_default.extend(publicAPI, model, initialValues);
	event(publicAPI, model, "BoxSelectChange");
	event(publicAPI, model, "BoxSelectInput");
	setGet(publicAPI, model, [
		"renderSelection",
		"boxChangeOnClick",
		"selectionStyle",
		"container"
	]);
	vtkMouseBoxSelectionManipulator(publicAPI, model);
}
var newInstance = newInstance$1(extend, "vtkMouseBoxSelectionManipulator");
var MouseBoxSelectorManipulator_default = {
	newInstance,
	extend
};
//#endregion
export { MouseBoxSelectorManipulator_default as default, extend, newInstance };
