// @kitware/vtk.js/Interaction/Manipulators/MouseCameraUnicamRotateManipulator.js
import macros_default from "../../macros.js";
import { areEquals, clampValue, cross, distance2BetweenPoints, dot, multiplyScalar, normalize, subtract } from "../../Common/Core/Math.js";
import { FieldAssociations } from "../../Common/DataModel/DataSet/Constants.js";
import SphereSource_default from "../../Filters/Sources/SphereSource.js";
import CompositeCameraManipulator_default from "./CompositeCameraManipulator.js";
import CompositeMouseManipulator_default from "./CompositeMouseManipulator.js";
import Constants_default from "../../Rendering/Core/InteractorStyle/Constants.js";
import Actor_default from "../../Rendering/Core/Actor.js";
import Mapper_default from "../../Rendering/Core/Mapper.js";
import PointPicker_default from "../../Rendering/Core/PointPicker.js";
import { mat4, vec3 } from "gl-matrix";
//#region Sources/Interaction/Manipulators/MouseCameraUnicamRotateManipulator/index.js
var { States } = Constants_default;
function vtkMouseCameraUnicamRotateManipulator(publicAPI, model) {
	model.classHierarchy.push("vtkMouseCameraUnicamRotateManipulator");
	model.picker = PointPicker_default.newInstance();
	model.downPoint = [
		0,
		0,
		0
	];
	model.isDot = false;
	model.state = States.IS_NONE;
	const sphereSource = SphereSource_default.newInstance();
	sphereSource.setThetaResolution(6);
	sphereSource.setPhiResolution(6);
	const sphereMapper = Mapper_default.newInstance();
	sphereMapper.setInputConnection(sphereSource.getOutputPort());
	model.focusSphere = Actor_default.newInstance();
	model.focusSphere.setMapper(sphereMapper);
	model.focusSphere.getProperty().setColor(.89, .66, .41);
	model.focusSphere.getProperty().setAmbient(1);
	model.focusSphere.getProperty().setDiffuse(0);
	model.focusSphere.getProperty().setRepresentationToWireframe();
	const updateAndRender = (interactor) => {
		if (!interactor) return;
		if (model.useWorldUpVec) {
			const camera = interactor.findPokedRenderer().getActiveCamera();
			if (!areEquals(model.worldUpVec, camera.getViewPlaneNormal())) camera.setViewUp(model.worldUpVec);
		}
		interactor.render();
	};
	const normalize$1 = (position, interactor) => {
		const renderer = interactor.findPokedRenderer();
		const [width, height] = interactor.getView().getViewportSize(renderer);
		return {
			x: -1 + 2 * position.x / width,
			y: -1 + 2 * position.y / height
		};
	};
	const rotateCamera = (camera, cx, cy, cz, ax, ay, az, angle) => {
		const cameraPosition = camera.getPosition();
		const cameraFocalPoint = camera.getFocalPoint();
		const cameraViewUp = camera.getViewUp();
		cameraPosition[3] = 1;
		cameraFocalPoint[3] = 1;
		cameraViewUp[3] = 0;
		const transform = mat4.identity(new Float64Array(16));
		mat4.translate(transform, transform, [
			cx,
			cy,
			cz
		]);
		mat4.rotate(transform, transform, angle, [
			ax,
			ay,
			az
		]);
		mat4.translate(transform, transform, [
			-cx,
			-cy,
			-cz
		]);
		const newCameraPosition = [];
		const newCameraFocalPoint = [];
		vec3.transformMat4(newCameraPosition, cameraPosition, transform);
		vec3.transformMat4(newCameraFocalPoint, cameraFocalPoint, transform);
		mat4.identity(transform);
		mat4.rotate(transform, transform, angle, [
			ax,
			ay,
			az
		]);
		const newCameraViewUp = [];
		vec3.transformMat4(newCameraViewUp, cameraViewUp, transform);
		camera.setPosition(...newCameraPosition);
		camera.setFocalPoint(...newCameraFocalPoint);
		camera.setViewUp(...newCameraViewUp);
	};
	const rotate = (interactor, position) => {
		const renderer = interactor.findPokedRenderer();
		const normalizedPosition = normalize$1(position, interactor);
		const normalizedPreviousPosition = normalize$1(model.previousPosition, interactor);
		const center = model.focusSphere.getPosition();
		let normalizedCenter = interactor.getView().worldToDisplay(...center, renderer);
		normalizedCenter = normalize$1({
			x: center[0],
			y: center[1]
		}, interactor);
		normalizedCenter = [
			normalizedCenter.x,
			normalizedCenter.y,
			center[2]
		];
		const radsq = (1 + Math.abs(normalizedCenter[0])) ** 2;
		const op = [
			normalizedPreviousPosition.x,
			0,
			0
		];
		const oe = [
			normalizedPosition.x,
			0,
			0
		];
		const opsq = op[0] ** 2;
		const oesq = oe[0] ** 2;
		const lop = opsq > radsq ? 0 : Math.sqrt(radsq - opsq);
		const loe = oesq > radsq ? 0 : Math.sqrt(radsq - oesq);
		const nop = [
			op[0],
			0,
			lop
		];
		normalize(nop);
		const noe = [
			oe[0],
			0,
			loe
		];
		normalize(noe);
		const dot$1 = dot(nop, noe);
		if (Math.abs(dot$1) > 1e-4) {
			const angle = -2 * Math.acos(clampValue(dot$1, -1, 1)) * Math.sign(normalizedPosition.x - normalizedPreviousPosition.x) * publicAPI.getRotationFactor();
			const camera = renderer.getActiveCamera();
			const upVec = model.useWorldUpVec ? model.worldUpVec : camera.getViewUp();
			normalize(upVec);
			rotateCamera(camera, ...center, ...upVec, angle);
			const dVec = [];
			subtract(camera.getPosition(), position, dVec);
			let rDist = (normalizedPosition.y - normalizedPreviousPosition.y) * publicAPI.getRotationFactor();
			normalize(dVec);
			const atV = camera.getViewPlaneNormal();
			const upV = camera.getViewUp();
			const rightV = [];
			cross(upV, atV, rightV);
			normalize(rightV);
			if (model.useWorldUpVec) {
				const OVER_THE_TOP_THRESHOLD = .99;
				if (dot(upVec, atV) > OVER_THE_TOP_THRESHOLD && rDist < 0) rDist = 0;
				if (dot(upVec, atV) < -OVER_THE_TOP_THRESHOLD && rDist > 0) rDist = 0;
			}
			rotateCamera(camera, ...center, ...rightV, rDist);
			if (model.useWorldUpVec && !areEquals(upVec, camera.getViewPlaneNormal())) camera.setViewUp(...upVec);
			model.previousPosition = position;
			renderer.resetCameraClippingRange();
			updateAndRender(interactor);
		}
	};
	const placeFocusSphere = (interactor) => {
		const renderer = interactor.findPokedRenderer();
		model.focusSphere.setPosition(...model.downPoint);
		const camera = renderer.getActiveCamera();
		const cameraPosition = camera.getPosition();
		const cameraToPointVec = [];
		subtract(model.downPoint, cameraPosition, cameraToPointVec);
		if (camera.getParallelProjection()) multiplyScalar(cameraToPointVec, camera.getParallelScale());
		const atV = camera.getDirectionOfProjection();
		normalize(atV);
		const scale = .02 * dot(atV, cameraToPointVec) * model.focusSphereRadiusFactor;
		model.focusSphere.setScale(scale, scale, scale);
	};
	const placeAndDisplayFocusSphere = (interactor) => {
		placeFocusSphere(interactor);
		interactor.findPokedRenderer().addActor(model.focusSphere);
		model.isDot = true;
	};
	const hideFocusSphere = (interactor) => {
		interactor.findPokedRenderer().removeActor(model.focusSphere);
		model.isDot = false;
	};
	const pickWithPointPicker = (interactor, position) => {
		const renderer = interactor.findPokedRenderer();
		model.picker.pick([
			position.x,
			position.y,
			position.z
		], renderer);
		const pickedPositions = model.picker.getPickedPositions();
		if (pickedPositions.length === 0) return model.picker.getPickPosition();
		const cameraPosition = renderer.getActiveCamera().getPosition();
		pickedPositions.sort((pointA, pointB) => distance2BetweenPoints(pointA, cameraPosition) - distance2BetweenPoints(pointB, cameraPosition));
		return pickedPositions[0];
	};
	const pickPoint = (interactor, position) => {
		const renderer = interactor.findPokedRenderer();
		let selections = null;
		if (model.useHardwareSelector) {
			const selector = interactor.getView().getSelector();
			selector.setCaptureZValues(true);
			selector.setFieldAssociation(FieldAssociations.FIELD_ASSOCIATION_POINTS);
			selector.attach(interactor.getView(), renderer);
			selector.setArea(position.x, position.y, position.x, position.y);
			selections = selector.select();
		}
		if (selections && selections.length !== 0) return Array.from(selections[0].getProperties().worldPosition);
		return pickWithPointPicker(interactor, position);
	};
	publicAPI.onButtonDown = (interactor, renderer, position) => {
		model.buttonPressed = true;
		model.startPosition = position;
		model.previousPosition = position;
		const normalizedPosition = normalize$1(position, interactor);
		const borderRatio = .1;
		if (Math.abs(normalizedPosition.x) > 1 - borderRatio || Math.abs(normalizedPosition.y) > 1 - borderRatio) {
			model.state = States.IS_ROTATE;
			placeAndDisplayFocusSphere(interactor);
			return;
		}
		model.downPoint = pickPoint(interactor, position);
		if (model.isDot) model.state = States.IS_ROTATE;
		else {
			model.state = States.IS_NONE;
			if (model.displayFocusSphereOnButtonDown) placeAndDisplayFocusSphere(interactor);
		}
	};
	publicAPI.onMouseMove = (interactor, renderer, position) => {
		if (!model.buttonPressed) return;
		model.state = States.IS_ROTATE;
		rotate(interactor, position);
		model.previousPosition = position;
	};
	publicAPI.onButtonUp = (interactor) => {
		const renderer = interactor.findPokedRenderer();
		model.buttonPressed = false;
		if (model.state === States.IS_ROTATE && !model.isDot) return;
		if (model.state === States.IS_ROTATE) hideFocusSphere(interactor);
		else if (model.state === States.IS_NONE) placeAndDisplayFocusSphere(interactor);
		renderer.resetCameraClippingRange();
		updateAndRender(interactor);
	};
	publicAPI.getFocusSphereColor = () => {
		model.focusSphere.getProperty().getColor();
	};
	publicAPI.setFocusSphereColor = (r, g, b) => {
		model.focusSphere.getProperty().setColor(r, g, b);
	};
}
var DEFAULT_VALUES = {
	focusSphereRadiusFactor: 1,
	displayFocusSphereOnButtonDown: true,
	useHardwareSelector: true,
	useWorldUpVec: true,
	worldUpVec: [
		0,
		0,
		1
	]
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	CompositeCameraManipulator_default.extend(publicAPI, model, initialValues);
	CompositeMouseManipulator_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, [
		"focusSphereRadiusFactor",
		"displayFocusSphereOnButtonDown",
		"useHardwareSelector",
		"useWorldUpVec"
	]);
	macros_default.get(publicAPI, model, ["state"]);
	macros_default.getArray(publicAPI, model, ["downPoint"], 3);
	macros_default.setGetArray(publicAPI, model, ["worldUpVec"], 3);
	vtkMouseCameraUnicamRotateManipulator(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkMouseCameraUnicamRotateManipulator");
var MouseCameraUnicamRotateManipulator_default = {
	newInstance,
	extend
};
//#endregion
export { MouseCameraUnicamRotateManipulator_default as default, extend, newInstance };
