// @kitware/vtk.js/Interaction/Style/InteractorStyleMPRSlice.js
import macros_default from "../../macros.js";
import { areEquals, distance2BetweenPoints, normalize } from "../../Common/Core/Math.js";
import BoundingBox_default from "../../Common/DataModel/BoundingBox.js";
import MatrixBuilder_default from "../../Common/Core/MatrixBuilder.js";
import InteractorStyleManipulator_default from "./InteractorStyleManipulator.js";
import MouseCameraTrackballRotateManipulator_default from "../Manipulators/MouseCameraTrackballRotateManipulator.js";
import MouseCameraTrackballPanManipulator_default from "../Manipulators/MouseCameraTrackballPanManipulator.js";
import MouseCameraTrackballZoomManipulator_default from "../Manipulators/MouseCameraTrackballZoomManipulator.js";
import MouseRangeManipulator_default from "../Manipulators/MouseRangeManipulator.js";
import { mat4 } from "gl-matrix";
//#region Sources/Interaction/Style/InteractorStyleMPRSlice/index.js
function clamp(value, min, max) {
	if (value < min) return min;
	if (value > max) return max;
	return value;
}
function vtkInteractorStyleMPRSlice(publicAPI, model) {
	model.classHierarchy.push("vtkInteractorStyleMPRSlice");
	model.trackballManipulator = MouseCameraTrackballRotateManipulator_default.newInstance({ button: 1 });
	model.panManipulator = MouseCameraTrackballPanManipulator_default.newInstance({
		button: 1,
		shift: true
	});
	model.zoomManipulator = MouseCameraTrackballZoomManipulator_default.newInstance({ button: 3 });
	model.scrollManipulator = MouseRangeManipulator_default.newInstance({
		scrollEnabled: true,
		dragEnabled: false
	});
	const cache = {
		sliceNormal: [
			0,
			0,
			0
		],
		sliceRange: [0, 0]
	};
	let cameraSub = null;
	function updateScrollManipulator() {
		const range = publicAPI.getSliceRange();
		model.scrollManipulator.removeScrollListener();
		model.scrollManipulator.setScrollListener(range[0], range[1], 1, publicAPI.getSlice, publicAPI.setSlice);
	}
	function setManipulators() {
		publicAPI.removeAllMouseManipulators();
		publicAPI.addMouseManipulator(model.trackballManipulator);
		publicAPI.addMouseManipulator(model.panManipulator);
		publicAPI.addMouseManipulator(model.zoomManipulator);
		publicAPI.addMouseManipulator(model.scrollManipulator);
		updateScrollManipulator();
	}
	const superSetInteractor = publicAPI.setInteractor;
	publicAPI.setInteractor = (interactor) => {
		superSetInteractor(interactor);
		if (cameraSub) {
			cameraSub.unsubscribe();
			cameraSub = null;
		}
		if (interactor) cameraSub = interactor.getCurrentRenderer().getActiveCamera().onModified(() => {
			updateScrollManipulator();
			publicAPI.modified();
		});
	};
	publicAPI.handleMouseMove = macros_default.chain(publicAPI.handleMouseMove, () => {
		const camera = model._interactor.getCurrentRenderer().getActiveCamera();
		const dist = camera.getDistance();
		camera.setClippingRange(dist, dist + .1);
	});
	const superSetVolumeMapper = publicAPI.setVolumeMapper;
	publicAPI.setVolumeMapper = (mapper) => {
		if (superSetVolumeMapper(mapper)) {
			const camera = model._interactor.getCurrentRenderer().getActiveCamera();
			if (mapper) {
				camera.setFreezeFocalPoint(true);
				publicAPI.setSliceNormal(...publicAPI.getSliceNormal());
			} else camera.setFreezeFocalPoint(false);
		}
	};
	publicAPI.getSlice = () => {
		const camera = model._interactor.getCurrentRenderer().getActiveCamera();
		const sliceNormal = publicAPI.getSliceNormal();
		const transform = MatrixBuilder_default.buildFromDegree().identity().rotateFromDirections(sliceNormal, [
			1,
			0,
			0
		]);
		const fp = camera.getFocalPoint();
		transform.apply(fp);
		return fp[0];
	};
	publicAPI.setSlice = (slice) => {
		const camera = model._interactor.getCurrentRenderer().getActiveCamera();
		if (model.volumeMapper) {
			const range = publicAPI.getSliceRange();
			const clampedSlice = clamp(slice, ...range);
			const center = model.volumeMapper.getCenter();
			const distance = camera.getDistance();
			const dop = camera.getDirectionOfProjection();
			normalize(dop);
			const midPoint = (range[1] + range[0]) / 2;
			const zeroPoint = [
				center[0] - dop[0] * midPoint,
				center[1] - dop[1] * midPoint,
				center[2] - dop[2] * midPoint
			];
			const slicePoint = [
				zeroPoint[0] + dop[0] * clampedSlice,
				zeroPoint[1] + dop[1] * clampedSlice,
				zeroPoint[2] + dop[2] * clampedSlice
			];
			const newPos = [
				slicePoint[0] - dop[0] * distance,
				slicePoint[1] - dop[1] * distance,
				slicePoint[2] - dop[2] * distance
			];
			camera.setPosition(...newPos);
			camera.setFocalPoint(...slicePoint);
		}
	};
	publicAPI.getSliceRange = () => {
		if (model.volumeMapper) {
			const sliceNormal = publicAPI.getSliceNormal();
			if (areEquals(sliceNormal, cache.sliceNormal)) return cache.sliceRange;
			const sliceOrientation = MatrixBuilder_default.buildFromDegree().identity().rotateFromDirections(sliceNormal, [
				1,
				0,
				0
			]);
			const imageAlongSliceNormal = mat4.create();
			mat4.multiply(imageAlongSliceNormal, sliceOrientation.getMatrix(), model.volumeMapper.getInputData().getIndexToWorld());
			const transformedBounds = BoundingBox_default.transformBounds(model.volumeMapper.getInputData().getSpatialExtent(), imageAlongSliceNormal);
			const minX = transformedBounds[0];
			const maxX = transformedBounds[1];
			cache.sliceNormal = sliceNormal;
			cache.sliceRange = [minX, maxX];
			return cache.sliceRange;
		}
		return [0, 0];
	};
	publicAPI.getSliceNormal = () => {
		if (model.volumeMapper) return model._interactor.getCurrentRenderer().getActiveCamera().getDirectionOfProjection();
		return [
			0,
			0,
			0
		];
	};
	publicAPI.setSliceNormal = (...normal) => {
		const camera = model._interactor.getCurrentRenderer().getActiveCamera();
		normalize(normal);
		if (model.volumeMapper) {
			const bounds = model.volumeMapper.getBounds();
			const diagonal = Math.sqrt(distance2BetweenPoints([
				bounds[0],
				bounds[2],
				bounds[4]
			], [
				bounds[1],
				bounds[3],
				bounds[5]
			]));
			const center = [
				(bounds[0] + bounds[1]) / 2,
				(bounds[2] + bounds[3]) / 2,
				(bounds[4] + bounds[5]) / 2
			];
			const angle = 90;
			const dist = diagonal / (2 * Math.tan(angle / 360 * Math.PI));
			const cameraPos = [
				center[0] - normal[0] * dist,
				center[1] - normal[1] * dist,
				center[2] - normal[2] * dist
			];
			const oldDop = camera.getDirectionOfProjection();
			const transform = MatrixBuilder_default.buildFromDegree().identity().rotateFromDirections(oldDop, normal);
			const viewUp = [
				0,
				1,
				0
			];
			transform.apply(viewUp);
			camera.setPosition(...cameraPos);
			camera.setDistance(dist);
			camera.setDirectionOfProjection(...normal);
			camera.setViewUp(...viewUp);
			camera.setViewAngle(angle);
			camera.setClippingRange(dist, dist + .1);
			publicAPI.setCenterOfRotation(center);
		}
	};
	setManipulators();
}
var DEFAULT_VALUES = {};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	InteractorStyleManipulator_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, ["volumeMapper"]);
	vtkInteractorStyleMPRSlice(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkInteractorStyleMPRSlice");
var InteractorStyleMPRSlice_default = {
	newInstance,
	extend
};
//#endregion
export { InteractorStyleMPRSlice_default as default, extend, newInstance };
