// @kitware/vtk.js/Interaction/UI/FPSMonitor/FPSMonitor.module.css.js
if (typeof document !== 'undefined') {
  var __vtkStyle = document.createElement('style');
  __vtkStyle.textContent = "._verticalContainer_1ezol_1 {\n  display: flex;\n  flex-direction: column;\n  justify-content: flex-start;\n  align-items: stretch;\n}\n\n._horizontalContainer_1ezol_8 {\n  display: flex;\n  flex-direction: row;\n  justify-content: flex-start;\n}\n\n._leftPane_1ezol_14 {\n  flex: none;\n  display: flex;\n  flex-direction: column;\n  justify-content: flex-start;\n  align-items: stretch;\n}\n\n._rightPane_1ezol_22 {\n  flex: 1;\n  display: grid;\n  grid-template-columns: auto auto;\n  grid-auto-rows: 1.5em;\n  grid-column-gap: 5px;\n  grid-row-gap: 2px;\n  padding: 10px;\n}\n\n._title_1ezol_32 {\n  flex: 1;\n  font-weight: bold;\n  padding: 5px 10px 0 10px;\n}\n\n._graph_1ezol_38 {\n  flex: none;\n  border: solid 1px black;\n  margin: 10px ;\n  border-radius: 2px;\n  overflow: hidden;\n}\n\n._label_1ezol_46 {\n  font-weight: bold;\n  text-transform: capitalize;\n  text-align: right;\n  align-self: center;\n}\n\n._value_1ezol_53 {\n  font-style: italic;\n  text-align: center;\n  align-self: center;\n}\n";
  (document.head || document.getElementsByTagName('head')[0]).appendChild(__vtkStyle);
}
var FPSMonitor_module_default = {
	verticalContainer: "_verticalContainer_1ezol_1",
	horizontalContainer: "_horizontalContainer_1ezol_8",
	leftPane: "_leftPane_1ezol_14",
	rightPane: "_rightPane_1ezol_22",
	title: "_title_1ezol_32",
	graph: "_graph_1ezol_38",
	label: "_label_1ezol_46",
	value: "_value_1ezol_53"
};
//#endregion
export { FPSMonitor_module_default as default };
