// @kitware/vtk.js/Imaging/Core/AbstractImageInterpolator/InterpolationInfo.js
import { ImageBorderMode, InterpolationMode } from "./Constants.js";
//#region Sources/Imaging/Core/AbstractImageInterpolator/InterpolationInfo.js
var vtkInterpolationInfo = {
	pointer: null,
	extent: [
		0,
		-1,
		0,
		-1,
		0,
		-1
	],
	increments: [
		0,
		0,
		0
	],
	scalarType: null,
	dataTypeSize: 1,
	numberOfComponents: 1,
	borderMode: ImageBorderMode.CLAMP,
	interpolationMode: InterpolationMode.LINEAR,
	extraInfo: null
};
var vtkInterpolationWeights = {
	...vtkInterpolationInfo,
	positions: [
		0,
		0,
		0
	],
	weights: null,
	weightExtent: [
		0,
		-1,
		0,
		-1,
		0,
		-1
	],
	kernelSize: [
		1,
		1,
		1
	],
	workspace: null,
	lastY: null,
	lastZ: null
};
function vtkInterpolationMathFloor(x) {
	const integer = Math.floor(x);
	return {
		floored: integer,
		error: x - integer
	};
}
function vtkInterpolationMathRound(x) {
	return Math.round(x);
}
function vtkInterpolationMathClamp(a, b, c) {
	let clamp = a <= c ? a : c;
	clamp -= b;
	clamp = clamp >= 0 ? clamp : 0;
	return clamp;
}
function vtkInterpolationMathWrap(a, b, c) {
	const range = c - b + 1;
	let wrap = a - b;
	wrap %= range;
	wrap = wrap >= 0 ? wrap : wrap + range;
	return wrap;
}
function vtkInterpolationMathMirror(a, b, c) {
	const range = c - b;
	const ifzero = range === 0 ? 1 : 0;
	const range2 = 2 * range + ifzero;
	let mirror = a - b;
	mirror = mirror >= 0 ? mirror : -mirror;
	mirror %= range2;
	mirror = mirror <= range ? mirror : range2 - mirror;
	return mirror;
}
var InterpolationInfo_default = {
	vtkInterpolationInfo,
	vtkInterpolationWeights,
	vtkInterpolationMathFloor,
	vtkInterpolationMathRound,
	vtkInterpolationMathClamp,
	vtkInterpolationMathWrap,
	vtkInterpolationMathMirror
};
//#endregion
export { InterpolationInfo_default as default, vtkInterpolationInfo, vtkInterpolationMathClamp, vtkInterpolationMathFloor, vtkInterpolationMathMirror, vtkInterpolationMathRound, vtkInterpolationMathWrap, vtkInterpolationWeights };
