// @kitware/vtk.js/Common/DataModel/DataSetAttributes/Constants.js
//#region Sources/Common/DataModel/DataSetAttributes/Constants.js
var AttributeTypes = {
	SCALARS: 0,
	VECTORS: 1,
	NORMALS: 2,
	TCOORDS: 3,
	TENSORS: 4,
	GLOBALIDS: 5,
	PEDIGREEIDS: 6,
	EDGEFLAG: 7,
	NUM_ATTRIBUTES: 8
};
var AttributeLimitTypes = {
	MAX: 0,
	EXACT: 1,
	NOLIMIT: 2
};
var CellGhostTypes = {
	DUPLICATECELL: 1,
	HIGHCONNECTIVITYCELL: 2,
	LOWCONNECTIVITYCELL: 4,
	REFINEDCELL: 8,
	EXTERIORCELL: 16,
	HIDDENCELL: 32
};
var PointGhostTypes = {
	DUPLICATEPOINT: 1,
	HIDDENPOINT: 2
};
var AttributeCopyOperations = {
	COPYTUPLE: 0,
	INTERPOLATE: 1,
	PASSDATA: 2,
	ALLCOPY: 3
};
var ghostArrayName = "vtkGhostType";
var DesiredOutputPrecision = {
	DEFAULT: 0,
	SINGLE: 1,
	DOUBLE: 2
};
var Constants_default = {
	AttributeCopyOperations,
	AttributeLimitTypes,
	AttributeTypes,
	CellGhostTypes,
	DesiredOutputPrecision,
	PointGhostTypes,
	ghostArrayName
};
//#endregion
export { AttributeCopyOperations, AttributeLimitTypes, AttributeTypes, CellGhostTypes, DesiredOutputPrecision, PointGhostTypes, Constants_default as default, ghostArrayName };
