// @kitware/vtk.js/IO/Misc/GCodeReader.js
import macros_default from "../../macros.js";
import CellArray_default from "../../Common/Core/CellArray.js";
import Points_default from "../../Common/Core/Points.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
import DataAccessHelper_default from "../Core/DataAccessHelper.js";
import "../Core/DataAccessHelper/LiteHttpDataAccessHelper.js";
import BinaryHelper_default from "../Core/BinaryHelper.js";
//#region Sources/IO/Misc/GCodeReader/index.js
function vtkGCodeReader(publicAPI, model) {
	const state = {
		currentPosition: {
			x: 0,
			y: 0,
			z: 0
		},
		offset: {
			x: 0,
			y: 0,
			z: 0
		},
		currentLayer: 0,
		layers: /* @__PURE__ */ new Map(),
		isAbsolute: true,
		isMetric: true,
		lastZ: 0
	};
	model.classHierarchy.push("vtkGCodeReader");
	if (!model.dataAccessHelper) model.dataAccessHelper = DataAccessHelper_default.get("http");
	function fetchData(url, option = {}) {
		const { compression, progressCallback } = model;
		if (option.binary) return model.dataAccessHelper.fetchBinary(url, {
			compression,
			progressCallback
		});
		return model.dataAccessHelper.fetchText(publicAPI, url, {
			compression,
			progressCallback
		});
	}
	function detectLayerChange(newZ) {
		if (Math.abs(newZ - state.lastZ) > .001) {
			state.currentLayer++;
			state.lastZ = newZ;
			return true;
		}
		return false;
	}
	function initializeLayer() {
		if (!state.layers.has(state.currentLayer)) {
			const points = Points_default.newInstance();
			const lines = CellArray_default.newInstance();
			const polyData = PolyData_default.newInstance();
			polyData.setPoints(points);
			polyData.setLines(lines);
			state.layers.set(state.currentLayer, {
				polyData,
				points,
				lines,
				zHeight: state.lastZ
			});
		}
	}
	function addLineToLayer(startPoint, endPoint) {
		initializeLayer();
		const layer = state.layers.get(state.currentLayer);
		const startIndex = layer.points.insertNextPoint(startPoint[0], startPoint[1], startPoint[2]);
		const endIndex = layer.points.insertNextPoint(endPoint[0], endPoint[1], endPoint[2]);
		layer.lines.insertNextCell([startIndex, endIndex]);
	}
	function processMove(params) {
		const newPosition = { ...state.currentPosition };
		let positionChanged = false;
		[
			"X",
			"Y",
			"Z"
		].forEach((axis) => {
			if (axis in params) {
				const value = state.isMetric ? params[axis] : params[axis] * 25.4;
				newPosition[axis.toLowerCase()] = state.isAbsolute ? value + state.offset[axis.toLowerCase()] : state.currentPosition[axis.toLowerCase()] + value;
				positionChanged = true;
			}
		});
		if (positionChanged) {
			if ("Z" in params) detectLayerChange(newPosition.z);
			addLineToLayer([
				state.currentPosition.x,
				state.currentPosition.y,
				state.currentPosition.z
			], [
				newPosition.x,
				newPosition.y,
				newPosition.z
			]);
			state.currentPosition = newPosition;
		}
	}
	function processG92(params) {
		[
			"X",
			"Y",
			"Z"
		].forEach((axis) => {
			if (axis in params) state.offset[axis.toLowerCase()] = state.currentPosition[axis.toLowerCase()] - (state.isMetric ? params[axis] : params[axis] * 25.4);
		});
	}
	function processCommand(command, params) {
		switch (command) {
			case "G0":
			case "G1":
				processMove(params);
				break;
			case "G20":
				state.isMetric = false;
				break;
			case "G21":
				state.isMetric = true;
				break;
			case "G90":
				state.isAbsolute = true;
				break;
			case "G91":
				state.isAbsolute = false;
				break;
			case "G92":
				processG92(params);
				break;
			default: break;
		}
	}
	function parseGCode(gcodeText) {
		gcodeText.split("\n").forEach((line) => {
			const sline = line.split(";")[0].trim();
			if (!sline) return;
			const tokens = sline.split(" ");
			const command = tokens[0];
			const params = {};
			tokens.slice(1).forEach((token) => {
				const param = token[0];
				const value = parseFloat(token.slice(1));
				if (!Number.isNaN(value)) params[param] = value;
			});
			processCommand(command, params);
		});
	}
	publicAPI.setUrl = (url, option = { binary: true }) => {
		model.url = url;
		const path = url.split("/");
		path.pop();
		model.baseURL = path.join("/");
		model.compression = option.compression;
		return publicAPI.loadData({
			progressCallback: option.progressCallback,
			binary: !!option.binary
		});
	};
	publicAPI.loadData = (option = {}) => {
		const promise = fetchData(model.url, option);
		promise.then(publicAPI.parse);
		return promise;
	};
	publicAPI.parseAsText = (content) => {
		parseGCode(content);
	};
	publicAPI.parseAsArrayBuffer = (content) => {
		parseGCode(BinaryHelper_default.arrayBufferToString(content));
	};
	publicAPI.parse = (content) => {
		if (typeof content === "string") publicAPI.parseAsText(content);
		else publicAPI.parseAsArrayBuffer(content);
		state.layers.forEach((layer, i) => {
			model.output[i] = layer.polyData;
		});
	};
	publicAPI.requestData = (inData, outData) => {
		publicAPI.parse(model.parseData);
	};
	publicAPI.getNumberOfOutputPorts = () => state.layers.size;
}
var DEFAULT_VALUES = {};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.get(publicAPI, model, ["url", "baseURL"]);
	macros_default.setGet(publicAPI, model, ["dataAccessHelper"]);
	macros_default.algo(publicAPI, model, 0, 1);
	macros_default.event(publicAPI, model, "ready");
	vtkGCodeReader(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkGCodeReader");
var GCodeReader_default = {
	extend,
	newInstance
};
//#endregion
export { GCodeReader_default as default, extend, newInstance };
