// @kitware/vtk.js/IO/Misc/HttpDataSetLODsLoader.js
import macros_default from "../../macros.js";
import DataAccessHelper_default from "../Core/DataAccessHelper.js";
import "../Core/DataAccessHelper/LiteHttpDataAccessHelper.js";
import HttpDataSetReader_default from "../Core/HttpDataSetReader.js";
//#region Sources/IO/Misc/HttpDataSetLODsLoader/index.js
var { vtkErrorMacro } = macros_default;
function vtkHttpDataSetLODsLoader(publicAPI, model) {
	model.classHierarchy.push("vtkHttpDataSetLODsLoader");
	const internal = { downloadStack: [] };
	publicAPI.startDownloads = () => {
		if (!model.mapper) {
			vtkErrorMacro("Mapper was not set.");
			return;
		}
		if (!model.files || model.files.length === 0) {
			vtkErrorMacro("No files set.");
			return;
		}
		let baseUrl = model.baseUrl;
		if (baseUrl && !baseUrl.endsWith("/")) baseUrl += "/";
		internal.downloadStack = [];
		model.files.forEach((file) => internal.downloadStack.push(`${baseUrl}${file}`));
		const downloadNextSource = () => {
			const url = internal.downloadStack.shift();
			const nextSource = HttpDataSetReader_default.newInstance({ dataAccessHelper: DataAccessHelper_default.get("http") });
			model.currentSource = nextSource;
			nextSource.setUrl(url, {
				compression: "zip",
				loadData: true,
				fullpath: true
			}).then(() => {
				model.mapper.setInputConnection(nextSource.getOutputPort());
				if (model.sceneItem) {
					const settings = model.sceneItem.defaultSettings;
					if (settings.mapper) {
						if (settings.mapper.colorByArrayName) nextSource.enableArray(settings.mapper.colorByArrayName, settings.mapper.colorByArrayName);
					}
					model.sceneItem.source = nextSource;
				}
				if (model.stepFinishedCallback) model.stepFinishedCallback();
				if (internal.downloadStack.length !== 0) setTimeout(downloadNextSource, model.waitTimeBetweenDownloads);
			});
		};
		setTimeout(downloadNextSource, model.waitTimeToStart);
	};
}
var DEFAULT_VALUES = {
	baseUrl: "",
	currentSource: null,
	files: [],
	mapper: null,
	sceneItem: null,
	stepFinishedCallback: null,
	waitTimeToStart: 4e3,
	waitTimeBetweenDownloads: 0
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, [
		"baseUrl",
		"files",
		"mapper",
		"sceneItem",
		"stepFinishedCallback",
		"waitTimeToStart",
		"waitTimeBetweenDownloads"
	]);
	macros_default.get(publicAPI, model, ["currentSource"]);
	vtkHttpDataSetLODsLoader(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkHttpDataSetLODsLoader");
var HttpDataSetLODsLoader_default = {
	newInstance,
	extend
};
//#endregion
export { HttpDataSetLODsLoader_default as default, extend, newInstance };
