// @kitware/vtk.js/IO/Misc/ITKImageReader.js
import macros_default from "../../macros.js";
import ITKHelper_default from "../../Common/DataModel/ITKHelper.js";
//#region Sources/IO/Misc/ITKImageReader/index.js
var { convertItkToVtkImage } = ITKHelper_default;
var readImageArrayBuffer = null;
var resultPreprocessor = (result) => result;
function getArrayName(filename) {
	const idx = filename.lastIndexOf(".");
	return `Scalars ${idx > -1 ? filename.substring(0, idx) : filename}`;
}
function setReadImageArrayBufferFromITK(fn) {
	readImageArrayBuffer = fn;
	if (readImageArrayBuffer.length === 4) {
		readImageArrayBuffer = (...args) => fn(null, ...args);
		resultPreprocessor = ({ webWorker, image }) => {
			webWorker.terminate();
			return image;
		};
	}
}
function vtkITKImageReader(publicAPI, model) {
	model.classHierarchy.push("vtkITKImageReader");
	publicAPI.parseAsArrayBuffer = (arrayBuffer) => {
		if (!arrayBuffer || arrayBuffer === model.rawDataBuffer) return Promise.resolve();
		model.rawDataBuffer = arrayBuffer;
		return readImageArrayBuffer(arrayBuffer, model.fileName).then(resultPreprocessor).then((itkImage) => {
			const imageData = convertItkToVtkImage(itkImage, { scalarArrayName: model.arrayName || getArrayName(model.fileName) });
			model.output[0] = imageData;
			publicAPI.modified();
		});
	};
	publicAPI.requestData = (inData, outData) => {
		publicAPI.parseAsArrayBuffer(model.rawDataBuffer, model.fileName);
	};
}
var DEFAULT_VALUES = {
	fileName: "",
	arrayName: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.algo(publicAPI, model, 0, 1);
	macros_default.setGet(publicAPI, model, ["fileName", "arrayName"]);
	vtkITKImageReader(publicAPI, model);
	if (!readImageArrayBuffer) console.error(`
      // Dependency needs to be added inside your project
      import readImageArrayBuffer from 'itk/readImageArrayBuffer';
      vtkITKImageReader.setReadImageArrayBufferFromITK(readImageArrayBuffer);
      `);
}
var newInstance = macros_default.newInstance(extend, "vtkITKImageReader");
var ITKImageReader_default = {
	newInstance,
	extend,
	setReadImageArrayBufferFromITK
};
//#endregion
export { ITKImageReader_default as default, extend, newInstance };
