// @kitware/vtk.js/IO/Misc/ITKPolyDataReader.js
import vtk from "../../vtk.js";
import macros_default from "../../macros.js";
//#region Sources/IO/Misc/ITKPolyDataReader/index.js
var readPolyDataArrayBuffer = null;
var resultPreprocessor = (result) => result;
function setReadPolyDataArrayBufferFromITK(fn) {
	readPolyDataArrayBuffer = fn;
	readPolyDataArrayBuffer = (...args) => fn(null, ...args);
	resultPreprocessor = ({ webWorker, polyData }) => {
		webWorker.terminate();
		return polyData;
	};
}
function vtkITKPolyDataReader(publicAPI, model) {
	model.classHierarchy.push("vtkITKPolyDataReader");
	publicAPI.parseAsArrayBuffer = (arrayBuffer) => {
		if (!arrayBuffer || arrayBuffer === model.rawDataBuffer) return Promise.resolve();
		model.rawDataBuffer = arrayBuffer;
		return readPolyDataArrayBuffer(arrayBuffer, model.fileName).then(resultPreprocessor).then((polyData) => {
			model.output[0] = vtk(polyData);
			publicAPI.modified();
		});
	};
	publicAPI.requestData = (inData, outData) => {
		publicAPI.parseAsArrayBuffer(model.rawDataBuffer, model.fileName);
	};
}
var DEFAULT_VALUES = {
	fileName: "",
	arrayName: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.algo(publicAPI, model, 0, 1);
	macros_default.setGet(publicAPI, model, ["fileName", "arrayName"]);
	vtkITKPolyDataReader(publicAPI, model);
	if (!readPolyDataArrayBuffer) console.error(`
      // Dependency needs to be added inside your project
      import readPolyDataArrayBuffer from 'itk/readPolyDataArrayBuffer';
      vtkITKPolyDataReader.setReadPolyDataArrayBufferFromITK(readPolyDataArrayBuffer);
      `);
}
var newInstance = macros_default.newInstance(extend, "vtkITKPolyDataReader");
var ITKPolyDataReader_default = {
	newInstance,
	extend,
	setReadPolyDataArrayBufferFromITK
};
//#endregion
export { ITKPolyDataReader_default as default, extend, newInstance };
