// @kitware/vtk.js/IO/Misc/OBJWriter.js
import macros_default from "../../macros.js";
import CellArray_default from "../../Common/Core/CellArray.js";
import TriangleStrip_default from "../../Common/DataModel/TriangleStrip.js";
import { strToU8, zipSync } from "fflate";
//#region Sources/IO/Misc/OBJWriter/index.js
var { vtkErrorMacro } = macros_default;
var writeFaces = (faces, withNormals, withTCoords) => {
	let outputData = "";
	const fd = faces.getData();
	let offset = 0;
	while (offset < fd.length) {
		const faceSize = fd[offset++];
		outputData += "f";
		for (let i = 0; i < faceSize; i++) {
			outputData += ` ${fd[offset + i] + 1}`;
			if (withTCoords) {
				outputData += `/${fd[offset + i] + 1}`;
				if (withNormals) outputData += `//${fd[offset + i] + 1}`;
			} else if (withNormals) outputData += `//${fd[offset + i] + 1}`;
		}
		offset += faceSize;
		outputData += "\n";
	}
	return outputData;
};
var writeLines = (lines) => {
	let outputData = "";
	const ld = lines.getData();
	let offset = 0;
	while (offset < ld.length) {
		const lineSize = ld[offset++];
		outputData += "l";
		for (let i = 0; i < lineSize; i++) outputData += ` ${ld[offset + i] + 1}`;
		offset += lineSize;
		outputData += "\n";
	}
	return outputData;
};
var writePoints = (pts, normals, tcoords) => {
	const outputData = [];
	const nbPts = pts.getNumberOfPoints();
	let p;
	for (let i = 0; i < nbPts; i++) {
		p = pts.getPoint(i);
		outputData.push(`v ${p[0]} ${p[1]} ${p[2]}`);
	}
	if (normals) for (let i = 0; i < nbPts; i++) {
		p = normals.getTuple(i);
		outputData.push(`vn ${p[0]} ${p[1]} ${p[2]}`);
	}
	if (tcoords) for (let i = 0; i < nbPts; i++) {
		p = tcoords.getTuple(i);
		if (p[0] !== -1) outputData.push(`vt ${p[0]} ${p[1]}`);
	}
	return `${outputData.join("\n")}\n`;
};
var writeMTL = (materialName, textureFileName) => {
	const outputData = [];
	outputData.push(`newmtl ${materialName}`);
	outputData.push(`map_Kd ${textureFileName}`);
	return outputData.join("\n");
};
var writeOBJ = (polyData, materialFilename, materialName) => {
	let outputData = "# VTK.js generated OBJ File\n";
	const pts = polyData.getPoints();
	const polys = polyData.getPolys();
	const strips = polyData.getStrips() ? polyData.getStrips().getData() : null;
	const lines = polyData.getLines();
	const normals = polyData.getPointData().getNormals();
	const tcoords = polyData.getPointData().getTCoords();
	const hasPtNormals = normals !== null;
	const hasPtTCoords = tcoords !== null;
	if (!pts) {
		vtkErrorMacro("No data to write!");
		return outputData;
	}
	if (materialFilename) outputData += `mtllib ${materialFilename}\n`;
	if (materialName) outputData += `usemtl ${materialName}\n`;
	outputData += writePoints(pts, normals, tcoords);
	const polyStrips = CellArray_default.newInstance();
	if (strips && strips.length > 0) TriangleStrip_default.decomposeStrip(pts, polyStrips);
	if (polyStrips.getNumberOfCells() > 0) outputData += writeFaces(polyStrips, hasPtNormals, hasPtTCoords);
	if (polys) outputData += writeFaces(polys, hasPtNormals, hasPtTCoords);
	if (lines) outputData += writeLines(lines);
	return outputData;
};
var STATIC = { writeOBJ };
function vtkOBJWriter(publicAPI, model) {
	model.classHierarchy.push("vtkOBJWriter");
	publicAPI.exportAsZip = () => {
		publicAPI.update();
		const modelFilename = model.modelFilename;
		const materialFilename = model.materialFilename;
		const textureFileName = model.textureFileName;
		const imageData = model.texture.getInputAsJsImageData?.();
		const zipContent = {};
		zipContent[`${modelFilename}.obj`] = strToU8(model.output[0]);
		zipContent[`${materialFilename}.mtl`] = strToU8(model.mtl);
		const canvas = document.createElement("canvas");
		canvas.width = imageData.width;
		canvas.height = imageData.height;
		canvas.getContext("2d").putImageData(imageData, 0, 0);
		return new Promise((resolve) => {
			canvas.toBlob(async (blob) => {
				const arrayBuffer = await blob.arrayBuffer();
				zipContent[`${textureFileName}.png`] = new Uint8Array(arrayBuffer);
				resolve(zipSync(zipContent));
			}, "image/png");
		});
	};
	publicAPI.requestData = (inData, outData) => {
		const input = inData[0];
		if (!input || !input.isA("vtkPolyData")) {
			vtkErrorMacro("Invalid or missing vtkPolyData input");
			return;
		}
		const materialFilename = `${model.materialFilename}.mtl`;
		const textureFileName = `${model.textureFileName}.png`;
		outData[0] = writeOBJ(input, materialFilename, model.materialName);
		model.mtl = writeMTL(model.materialName, textureFileName);
	};
}
var DEFAULT_VALUES = {
	modelFilename: "model",
	materialName: "mat_01",
	materialFilename: "material",
	texture: null,
	textureFileName: "texture"
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.algo(publicAPI, model, 1, 1);
	macros_default.get(publicAPI, model, ["mtl"]);
	macros_default.set(publicAPI, model, [
		"modelFilename",
		"materialFilename",
		"texture",
		"textureFileName"
	]);
	vtkOBJWriter(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkOBJWriter");
var OBJWriter_default = {
	newInstance,
	extend,
	...STATIC
};
//#endregion
export { STATIC, OBJWriter_default as default, extend, newInstance };
