// @kitware/vtk.js/IO/Misc/PDBReader.js
import macros_default from "../../macros.js";
import Molecule_default from "../../Common/DataModel/Molecule.js";
import DataAccessHelper_default from "../Core/DataAccessHelper.js";
import "../Core/DataAccessHelper/LiteHttpDataAccessHelper.js";
import elements_default from "../../Utilities/XMLConverter/chemistry-mapper/elements.json.js";
//#region Sources/IO/Misc/PDBReader/index.js
function vtkPDBReader(publicAPI, model) {
	model.classHierarchy.push("vtkPDBReader");
	if (!model.dataAccessHelper) model.dataAccessHelper = DataAccessHelper_default.get("http");
	function fetchPDB(url, option) {
		return model.dataAccessHelper.fetchText(publicAPI, url, option);
	}
	publicAPI.setUrl = (url, option) => {
		if (url.indexOf(".pdb") === -1) {
			model.baseURL = url;
			model.url = `${url}`;
		} else {
			model.url = url;
			const path = url.split("/");
			path.pop();
			model.baseURL = path.join("/");
		}
		return publicAPI.loadData(option);
	};
	publicAPI.loadData = (option) => fetchPDB(model.url, option).then(publicAPI.parseAsText);
	publicAPI.parseAsText = (txt) => {
		model.pdb = txt;
		model.molecule = [];
		model.molecule = model.pdb.split("\n");
		publicAPI.modified();
		return true;
	};
	publicAPI.requestData = (inData, outData) => {
		const moleculedata = Molecule_default.newInstance();
		if (model.molecule) {
			const jSize = model.molecule.length;
			const pointValues = [];
			const atomicNumber = [];
			model.numberOfAtoms = 0;
			let j = 0;
			while (j < jSize && model.molecule[j] !== "END") {
				const iSize = model.molecule[j].length;
				const linebuf = model.molecule[j];
				const command = linebuf.substr(0, 6).replace(/\s+/g, "");
				command.toUpperCase();
				if (command === "ATOM" || command === "HETATM") {
					const dum1 = linebuf.substr(12, 4).replace(/\s+/g, "");
					const x = linebuf.substr(30, 8).replace(/\s+/g, "");
					const y = linebuf.substr(38, 8).replace(/\s+/g, "");
					const z = linebuf.substr(46, 8).replace(/\s+/g, "");
					let elem = "";
					if (iSize >= 78) elem = linebuf.substr(76, 2).replace(/\s+/g, "");
					if (elem === "") elem = dum1.substr(0, 2).replace(/\d/g, "");
					pointValues.push(x);
					pointValues.push(y);
					pointValues.push(z);
					const [atomicNumberData] = elements_default[elem];
					atomicNumber.push(atomicNumberData);
					model.numberOfAtoms++;
				}
				j++;
			}
			moleculedata.getAtoms().elements = {};
			moleculedata.getAtoms().elements.number = Int8Array.from(atomicNumber);
			moleculedata.getAtoms().coords = {};
			moleculedata.getAtoms().coords["3d"] = Float32Array.from(pointValues);
		}
		model.output[0] = moleculedata;
	};
	publicAPI.isBusy = () => !!model.requestCount;
}
var DEFAULT_VALUES = {
	numberOfAtoms: 0,
	requestCount: 0
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.get(publicAPI, model, [
		"url",
		"baseURL",
		"numberOfAtoms",
		"requestCount"
	]);
	macros_default.setGet(publicAPI, model, ["dataAccessHelper"]);
	macros_default.algo(publicAPI, model, 0, 1);
	macros_default.event(publicAPI, model, "busy");
	vtkPDBReader(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkPDBReader");
var PDBReader_default = {
	newInstance,
	extend
};
//#endregion
export { PDBReader_default as default, extend, newInstance };
