// @kitware/vtk.js/IO/Image/TGAReader.js
import macros_default from "../../macros.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import ImageData_default from "../../Common/DataModel/ImageData.js";
import DataAccessHelper_default from "../Core/DataAccessHelper.js";
import "../Core/DataAccessHelper/LiteHttpDataAccessHelper.js";
import Constants_default from "./TGAReader/Constants.js";
//#region Sources/IO/Image/TGAReader/index.js
var { vtkErrorMacro } = macros_default;
function vtkTGAReader(publicAPI, model) {
	model.classHierarchy.push("vtkTGAReader");
	if (!model.dataAccessHelper) model.dataAccessHelper = DataAccessHelper_default.get("http");
	/**
	* Gets the header of a TGA file
	* @param data defines the TGA data
	* @returns the header
	*/
	function parseHeader(data) {
		let offset = 0;
		return {
			idLength: data[offset++],
			colormap_type: data[offset++],
			imageType: data[offset++],
			colormapIndex: data[offset++] | data[offset++] << 8,
			colormapLength: data[offset++] | data[offset++] << 8,
			colormapSize: data[offset++],
			origin: [data[offset++] | data[offset++] << 8, data[offset++] | data[offset++] << 8],
			width: data[offset++] | data[offset++] << 8,
			height: data[offset++] | data[offset++] << 8,
			pixelSize: data[offset++],
			flags: data[offset++]
		};
	}
	const handlers = {
		getImageData8bits(header, palettes, pixeData, yStart, yStep, yEnd, xStart, xStep, xEnd) {
			const image = pixeData;
			const colormap = palettes;
			const width = header.width;
			const height = header.height;
			let color;
			let i = 0;
			let x;
			let y;
			const imageData = new Uint8Array(width * height * 4);
			for (y = yStart; y !== yEnd; y += yStep) for (x = xStart; x !== xEnd; x += xStep, i++) {
				color = image[i];
				imageData[(x + width * y) * 4 + 3] = 255;
				imageData[(x + width * y) * 4 + 2] = colormap[color * 3 + 0];
				imageData[(x + width * y) * 4 + 1] = colormap[color * 3 + 1];
				imageData[(x + width * y) * 4 + 0] = colormap[color * 3 + 2];
			}
			return imageData;
		},
		getImageData16bits(header, palettes, pixeData, yStart, yStep, yEnd, xStart, xStep, xEnd) {
			const image = pixeData;
			const width = header.width;
			const height = header.height;
			let color;
			let i = 0;
			let x;
			let y;
			const imageData = new Uint8Array(width * height * 4);
			for (y = yStart; y !== yEnd; y += yStep) for (x = xStart; x !== xEnd; x += xStep, i += 2) {
				color = image[i + 0] + (image[i + 1] << 8);
				const r = ((color & 31744) >> 10) * 255 / 31 | 0;
				const g = ((color & 992) >> 5) * 255 / 31 | 0;
				const b = (color & 31) * 255 / 31 | 0;
				imageData[(x + width * y) * 4 + 0] = r;
				imageData[(x + width * y) * 4 + 1] = g;
				imageData[(x + width * y) * 4 + 2] = b;
				imageData[(x + width * y) * 4 + 3] = color & 32768 ? 0 : 255;
			}
			return imageData;
		},
		getImageData24bits(header, palettes, pixeData, yStart, yStep, yEnd, xStart, xStep, xEnd) {
			const image = pixeData;
			const width = header.width;
			const height = header.height;
			let i = 0;
			let x;
			let y;
			const imageData = new Uint8Array(width * height * 4);
			for (y = yStart; y !== yEnd; y += yStep) for (x = xStart; x !== xEnd; x += xStep, i += 3) {
				imageData[(x + width * y) * 4 + 3] = 255;
				imageData[(x + width * y) * 4 + 2] = image[i + 0];
				imageData[(x + width * y) * 4 + 1] = image[i + 1];
				imageData[(x + width * y) * 4 + 0] = image[i + 2];
			}
			return imageData;
		},
		getImageData32bits(header, palettes, pixeData, yStart, yStep, yEnd, xStart, xStep, xEnd) {
			const image = pixeData;
			const width = header.width;
			const height = header.height;
			let i = 0;
			let x;
			let y;
			const imageData = new Uint8Array(width * height * 4);
			for (y = yStart; y !== yEnd; y += yStep) for (x = xStart; x !== xEnd; x += xStep, i += 4) {
				imageData[(x + width * y) * 4 + 2] = image[i + 0];
				imageData[(x + width * y) * 4 + 1] = image[i + 1];
				imageData[(x + width * y) * 4 + 0] = image[i + 2];
				imageData[(x + width * y) * 4 + 3] = image[i + 3];
			}
			return imageData;
		},
		getImageDataGrey8bits(header, palettes, pixeData, yStart, yStep, yEnd, xStart, xStep, xEnd) {
			const image = pixeData;
			const width = header.width;
			const height = header.height;
			let color;
			let i = 0;
			let x;
			let y;
			const imageData = new Uint8Array(width * height * 4);
			for (y = yStart; y !== yEnd; y += yStep) for (x = xStart; x !== xEnd; x += xStep, i++) {
				color = image[i];
				imageData[(x + width * y) * 4 + 0] = color;
				imageData[(x + width * y) * 4 + 1] = color;
				imageData[(x + width * y) * 4 + 2] = color;
				imageData[(x + width * y) * 4 + 3] = 255;
			}
			return imageData;
		},
		getImageDataGrey16bits(header, palettes, pixeData, yStart, yStep, yEnd, xStart, xStep, xEnd) {
			const image = pixeData;
			const width = header.width;
			const height = header.height;
			let i = 0;
			let x;
			let y;
			const imageData = new Uint8Array(width * height * 4);
			for (y = yStart; y !== yEnd; y += yStep) for (x = xStart; x !== xEnd; x += xStep, i += 2) {
				imageData[(x + width * y) * 4 + 0] = image[i + 0];
				imageData[(x + width * y) * 4 + 1] = image[i + 0];
				imageData[(x + width * y) * 4 + 2] = image[i + 0];
				imageData[(x + width * y) * 4 + 3] = image[i + 1];
			}
			return imageData;
		}
	};
	function fetchData(url, option = {}) {
		const { compression, progressCallback } = model;
		return model.dataAccessHelper.fetchBinary(url, {
			compression,
			progressCallback
		});
	}
	publicAPI.setUrl = (url, option = { binary: true }) => {
		model.url = url;
		const path = url.split("/");
		path.pop();
		model.baseURL = path.join("/");
		model.compression = option.compression;
		return publicAPI.loadData({ progressCallback: option.progressCallback });
	};
	publicAPI.loadData = (option = {}) => {
		const promise = fetchData(model.url, option);
		promise.then(publicAPI.parse);
		return promise;
	};
	publicAPI.parse = (content) => {
		publicAPI.parseAsArrayBuffer(content);
	};
	publicAPI.parseAsArrayBuffer = (content) => {
		if (!content) return;
		const data = new Uint8Array(content);
		if (data.length < 19) {
			vtkErrorMacro("Unable to load TGA file - Not enough data to contain header");
			return;
		}
		let offset = 18;
		const header = parseHeader(data);
		if (header.idLength + offset > data.length) {
			vtkErrorMacro("Unable to load TGA file - Not enough data");
			return;
		}
		offset += header.idLength;
		let useRle = false;
		let usePal = false;
		let useGrey = false;
		switch (header.imageType) {
			case Constants_default.TYPE_RLE_INDEXED: useRle = true;
			case Constants_default.TYPE_INDEXED:
				usePal = true;
				break;
			case Constants_default.TYPE_RLE_RGB: useRle = true;
			case Constants_default.TYPE_RGB: break;
			case Constants_default.TYPE_RLE_GREY: useRle = true;
			case Constants_default.TYPE_GREY:
				useGrey = true;
				break;
			default:
				vtkErrorMacro("TGA file has unknown image type");
				return;
		}
		let pixelData;
		const pixelSize = header.pixelSize >> 3;
		const pixelTotal = header.width * header.height * pixelSize;
		let palettes;
		if (usePal) palettes = data.subarray(offset, offset += header.colormapLength * (header.colormapSize >> 3));
		if (useRle) {
			pixelData = new Uint8Array(pixelTotal);
			let c;
			let count;
			let i;
			let localOffset = 0;
			const pixels = new Uint8Array(pixelSize);
			while (offset < pixelTotal && localOffset < pixelTotal) {
				c = data[offset++];
				count = (c & 127) + 1;
				if (c & 128) {
					for (i = 0; i < pixelSize; ++i) pixels[i] = data[offset++];
					for (i = 0; i < count; ++i) pixelData.set(pixels, localOffset + i * pixelSize);
					localOffset += pixelSize * count;
				} else {
					count *= pixelSize;
					for (i = 0; i < count; ++i) pixelData[localOffset + i] = data[offset++];
					localOffset += count;
				}
			}
		} else pixelData = data.subarray(offset, offset += usePal ? header.width * header.height : pixelTotal);
		let xStart;
		let yStart;
		let xStep;
		let yStep;
		let yEnd;
		let xEnd;
		switch ((header.flags & Constants_default.ORIGIN_MASK) >> Constants_default.ORIGIN_SHIFT) {
			case Constants_default.ORIGIN_UL:
				xStart = 0;
				xStep = 1;
				xEnd = header.width;
				yStart = 0;
				yStep = 1;
				yEnd = header.height;
				break;
			case Constants_default.ORIGIN_BL:
				xStart = 0;
				xStep = 1;
				xEnd = header.width;
				yStart = 0;
				yStep = 1;
				yEnd = header.height;
				break;
			case Constants_default.ORIGIN_UR:
				xStart = header.width - 1;
				xStep = -1;
				xEnd = -1;
				yStart = 0;
				yStep = 1;
				yEnd = header.height;
				break;
			case Constants_default.ORIGIN_BR:
				xStart = header.width - 1;
				xStep = -1;
				xEnd = -1;
				yStart = header.height - 1;
				yStep = -1;
				yEnd = -1;
				break;
			default:
				vtkErrorMacro("TGA file has unknown origin");
				return;
		}
		const output = handlers[`getImageData${useGrey ? "Grey" : ""}${header.pixelSize}bits`](header, palettes, pixelData, yStart, yStep, yEnd, xStart, xStep, xEnd);
		const dataExtent = [
			0,
			header.width - 1,
			0,
			header.height - 1
		];
		const dataSpacing = [
			1,
			1,
			1
		];
		const imageData = ImageData_default.newInstance();
		imageData.setDimensions(header.width, header.height, 1);
		imageData.setExtent(dataExtent);
		imageData.setSpacing(dataSpacing);
		const dataArray = DataArray_default.newInstance({
			name: "TGAImage",
			numberOfComponents: 4,
			values: output
		});
		imageData.getPointData().setScalars(dataArray);
		model.output[0] = imageData;
	};
	publicAPI.requestData = (inData, outData) => {
		publicAPI.parse(model.parseData);
	};
}
var DEFAULT_VALUES = {};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.algo(publicAPI, model, 0, 1);
	macros_default.get(publicAPI, model, ["url", "baseURL"]);
	macros_default.setGet(publicAPI, model, ["dataAccessHelper"]);
	vtkTGAReader(publicAPI, model);
	if (!model.compression) model.compression = null;
	if (!model.progressCallback) model.progressCallback = null;
}
var newInstance = macros_default.newInstance(extend, "vtkTGAReader");
var TGAReader_default = {
	newInstance,
	extend
};
//#endregion
export { TGAReader_default as default, extend, newInstance };
