// @kitware/vtk.js/IO/Image/TIFFReader.js
import macros_default from "../../macros.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import ImageData_default from "../../Common/DataModel/ImageData.js";
import DataAccessHelper_default from "../Core/DataAccessHelper.js";
import "../Core/DataAccessHelper/LiteHttpDataAccessHelper.js";
import UTIF from "utif";
//#region Sources/IO/Image/TIFFReader/index.js
function vtkTIFFReader(publicAPI, model) {
	model.classHierarchy.push("vtkTIFFReader");
	if (!model.dataAccessHelper) model.dataAccessHelper = DataAccessHelper_default.get("http");
	function fetchData(url, option = {}) {
		const { compression, progressCallback } = model;
		return model.dataAccessHelper.fetchBinary(url, {
			compression,
			progressCallback
		});
	}
	publicAPI.setUrl = (url, option = { binary: true }) => {
		model.url = url;
		const path = url.split("/");
		path.pop();
		model.baseURL = path.join("/");
		model.compression = option.compression;
		return publicAPI.loadData({ progressCallback: option.progressCallback });
	};
	publicAPI.loadData = (option = {}) => {
		const promise = fetchData(model.url, option);
		promise.then(publicAPI.parse);
		return promise;
	};
	publicAPI.parse = (content) => {
		publicAPI.parseAsArrayBuffer(content);
	};
	publicAPI.parseAsArrayBuffer = (content) => {
		if (!content) return;
		const ifds = UTIF.decode(content);
		UTIF.decodeImage(content, ifds[0]);
		const data = UTIF.toRGBA8(ifds[0]);
		const width = ifds[0].width;
		const height = ifds[0].height;
		const output = new Uint8Array(data.length);
		if (model.flipY) for (let y = 0; y < height; y++) for (let x = 0; x < width; x++) {
			const srcIndex = (y * width + x) * 4;
			const destIndex = ((height - y - 1) * width + x) * 4;
			output[destIndex] = data[srcIndex];
			output[destIndex + 1] = data[srcIndex + 1];
			output[destIndex + 2] = data[srcIndex + 2];
			output[destIndex + 3] = data[srcIndex + 3];
		}
		const dataExtent = [
			0,
			width - 1,
			0,
			height - 1
		];
		const dataSpacing = [
			1,
			1,
			1
		];
		const imageData = ImageData_default.newInstance();
		imageData.setDimensions(width, height, 1);
		imageData.setExtent(dataExtent);
		imageData.setSpacing(dataSpacing);
		const dataArray = DataArray_default.newInstance({
			name: "TIFFImage",
			numberOfComponents: 4,
			values: output
		});
		imageData.getPointData().setScalars(dataArray);
		model.output[0] = imageData;
	};
	publicAPI.requestData = (inData, outData) => {
		publicAPI.parse(model.parseData);
	};
}
var DEFAULT_VALUES = {
	flipY: true,
	compression: null,
	progressCallback: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.algo(publicAPI, model, 0, 1);
	macros_default.get(publicAPI, model, ["url", "baseURL"]);
	macros_default.setGet(publicAPI, model, ["dataAccessHelper", "flipY"]);
	vtkTIFFReader(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkTIFFReader");
var TIFFReader_default = {
	newInstance,
	extend
};
//#endregion
export { TIFFReader_default as default, extend, newInstance };
