// @kitware/vtk.js/Proxy/Representations.js
import GeometryRepresentationProxy_default from "./Representations/GeometryRepresentationProxy.js";
import GlyphRepresentationProxy_default from "./Representations/GlyphRepresentationProxy.js";
import MoleculeRepresentationProxy_default from "./Representations/MoleculeRepresentationProxy.js";
import SkyboxRepresentationProxy_default from "./Representations/SkyboxRepresentationProxy.js";
import SliceRepresentationProxy_default from "./Representations/SliceRepresentationProxy.js";
import SlicedGeometryRepresentationProxy_default from "./Representations/SlicedGeometryRepresentationProxy.js";
import VolumeRepresentationProxy_default from "./Representations/VolumeRepresentationProxy.js";
//#region Sources/Proxy/Representations/index.js
var Representations_default = {
	vtkGeometryRepresentationProxy: GeometryRepresentationProxy_default,
	vtkGlyphRepresentationProxy: GlyphRepresentationProxy_default,
	vtkMoleculeRepresentationProxy: MoleculeRepresentationProxy_default,
	vtkSkyboxRepresentationProxy: SkyboxRepresentationProxy_default,
	vtkSliceRepresentationProxy: SliceRepresentationProxy_default,
	vtkSlicedGeometryRepresentationProxy: SlicedGeometryRepresentationProxy_default,
	vtkVolumeRepresentationProxy: VolumeRepresentationProxy_default
};
//#endregion
export { Representations_default as default };
