// @kitware/vtk.js/Filters/General/OBBTree/OBBNode.js
import macros_default from "../../../macros.js";
//#region Sources/Filters/General/OBBTree/OBBNode.js
function vtkOBBNode(publicAPI, model) {
	model.classHierarchy.push("vtkOBBNode");
	/**
	* Copy a vtkOBBNode into an other one
	* @param {vtkOBBNode} nodeSource
	* @param {vtkOBBNode} nodeTarget
	*/
	publicAPI.deepCopy = (nodeSource) => {
		publicAPI.setCorner(nodeSource.getCorner());
		const axes = nodeSource.getAxes();
		publicAPI.setAxes([
			[...axes[0]],
			[...axes[1]],
			[...axes[2]]
		]);
		publicAPI.setCells([...nodeSource.getCells()]);
		if (nodeSource.getKids()) {
			const kids0 = vtkOBBNode.newInstance();
			kids0.setParent(publicAPI);
			const kids1 = vtkOBBNode.newInstance();
			kids1.setParent(publicAPI);
			kids0.deepCopy(nodeSource.getKids()[0]);
			kids1.deepCopy(nodeSource.getKids()[1], kids1);
			publicAPI.setKids(kids0, kids1);
		}
	};
	publicAPI.getAxis = (axis) => model.axes[axis];
}
function defaultValues(initialValues) {
	return {
		corner: [
			0,
			0,
			0
		],
		axes: [
			[
				0,
				0,
				0
			],
			[
				0,
				0,
				0
			],
			[
				0,
				0,
				0
			]
		],
		cells: [],
		...initialValues
	};
}
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, defaultValues(initialValues));
	model.corner = [
		0,
		0,
		0
	];
	model.axes = [
		[
			0,
			0,
			0
		],
		[
			0,
			0,
			0
		],
		[
			0,
			0,
			0
		]
	];
	model.parent = null;
	model.kids = null;
	model.cells = [];
	macros_default.setGet(publicAPI, model, [
		"parent",
		"cells",
		"kids"
	]);
	macros_default.setGetArray(publicAPI, model, ["corner", "axes"], 3);
	macros_default.obj(publicAPI, model);
	vtkOBBNode(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkOBBNode");
var OBBNode_default = {
	newInstance,
	extend
};
//#endregion
export { OBBNode_default as default, extend, newInstance };
