// @kitware/vtk.js/Proxy/Animation/AnimationProxyManager.js
import macros_default from "../../macros.js";
//#region Sources/Proxy/Animation/AnimationProxyManager/index.js
function vtkAnimationProxyManager(publicAPI, model) {
	model.classHierarchy.push("vtkAnimationProxyManager");
	publicAPI.addAnimation = (animation) => {
		if (!model.animations.includes(animation)) model.animations.push(animation);
		animation.onModified(publicAPI.updateFrames);
		publicAPI.updateFrames();
	};
	publicAPI.play = () => {
		const currentTime = model.frames[model.currentFrameIndex];
		const nextTime = model.frames[model.currentFrameIndex + 1];
		clearTimeout(model.timeOut);
		if (model.currentFrameIndex < model.frames.length - 1) model.timeOut = setTimeout(() => {
			publicAPI.nextFrame();
			publicAPI.play();
		}, (nextTime - currentTime) * 1e3);
		else publicAPI.invokeDonePlaying();
	};
	publicAPI.pause = () => {
		clearTimeout(model.timeOut);
		model.timeOut = null;
	};
	publicAPI.nextFrame = () => {
		if (model.currentFrameIndex < model.frames.length - 1) publicAPI.setFrameIndex(model.currentFrameIndex + 1);
	};
	publicAPI.previousFrame = () => {
		if (model.currentFrameIndex > 0) publicAPI.setFrameIndex(model.currentFrameIndex - 1);
	};
	publicAPI.firstFrame = () => {
		publicAPI.setFrameIndex(0);
	};
	publicAPI.lastFrame = () => {
		publicAPI.setFrameIndex(model.frames.length - 1);
	};
	publicAPI.setFrameIndex = (frameId) => {
		model.currentFrameIndex = frameId;
		model.animations.forEach((animationProxy) => {
			animationProxy.setTime(model.frames[model.currentFrameIndex]);
		});
		publicAPI.invokeCurrentFrameChanged();
	};
	publicAPI.updateFrames = () => {
		const frames = [];
		model.animations.forEach((animationProxy) => {
			frames.push(...animationProxy.getFrames());
		});
		model.frames = frames.sort((a, b) => a - b).filter((val, index, array) => array.indexOf(val) === index);
		publicAPI.invokeFramesChanged();
		publicAPI.setFrameIndex(0);
	};
	publicAPI.getCurrentFrame = () => model.frames[model.currentFrameIndex];
}
var DEFAULT_VALUES = {
	animations: [],
	currentFrameIndex: 0,
	frames: [],
	timeOut: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.get(publicAPI, model, ["frames", "currentFrameIndex"]);
	vtkAnimationProxyManager(publicAPI, model);
	macros_default.proxy(publicAPI, model);
	macros_default.event(publicAPI, model, "currentFrameChanged");
	macros_default.event(publicAPI, model, "framesChanged");
	macros_default.event(publicAPI, model, "donePlaying");
}
var newInstance = macros_default.newInstance(extend, "vtkAnimationProxyManager");
var AnimationProxyManager_default = {
	newInstance,
	extend
};
//#endregion
export { AnimationProxyManager_default as default, newInstance };
