// @kitware/vtk.js/Proxy/Core/LookupTableProxy.js
import macros_default from "../../macros.js";
import ColorTransferFunction_default from "../../Rendering/Core/ColorTransferFunction.js";
import ColorMaps_default from "../../Rendering/Core/ColorTransferFunction/ColorMaps.js";
import Constants_default from "./LookupTableProxy/Constants.js";
//#region Sources/Proxy/Core/LookupTableProxy/index.js
var { Mode, Defaults } = Constants_default;
function vtkLookupTableProxy(publicAPI, model) {
	model.classHierarchy.push("vtkLookupTableProxy");
	model.lookupTable = model.lookupTable || ColorTransferFunction_default.newInstance();
	model.lookupTable.setVectorModeToMagnitude();
	publicAPI.setPresetName = (presetName) => {
		if (model.presetName !== presetName) {
			model.presetName = presetName;
			model.mode = Mode.Preset;
			publicAPI.applyMode();
		}
	};
	publicAPI.setRGBPoints = (rgbPoints) => {
		if (model.rgbPoints !== rgbPoints) {
			model.rgbPoints = (rgbPoints || Defaults.RGBPoints).slice();
			publicAPI.applyMode();
		}
	};
	publicAPI.setHSVPoints = (hsvPoints) => {
		if (model.hsvPoints !== hsvPoints) {
			model.hsvPoints = (hsvPoints || Defaults.HSVPoints).slice();
			publicAPI.applyMode();
		}
	};
	publicAPI.setNodes = (nodes) => {
		if (model.nodes !== nodes) {
			model.nodes = (nodes || Defaults.Nodes).slice();
			publicAPI.applyMode();
		}
	};
	publicAPI.setMode = (mode) => {
		if (model.mode !== mode) {
			model.mode = mode;
			publicAPI.applyMode();
		}
	};
	publicAPI.applyMode = () => {
		switch (model.mode) {
			case Mode.Preset:
				{
					const preset = ColorMaps_default.getPresetByName(model.presetName);
					if (preset) model.lookupTable.applyColorMap(preset);
				}
				break;
			case Mode.RGBPoints:
				model.lookupTable.removeAllPoints();
				model.rgbPoints.forEach((point) => model.lookupTable.addRGBPointLong(...point));
				break;
			case Mode.HSVPoints:
				model.lookupTable.removeAllPoints();
				model.hsvPoints.forEach((point) => model.lookupTable.addHSVPointLong(...point));
				break;
			case Mode.Nodes:
				model.lookupTable.setNodes(model.nodes);
				break;
			default:
		}
		model.lookupTable.setMappingRange(model.dataRange[0], model.dataRange[1]);
		model.lookupTable.updateRange();
		publicAPI.modified();
	};
	publicAPI.setDataRange = (min, max) => {
		if (model.dataRange[0] !== min || model.dataRange[1] !== max) {
			model.dataRange[0] = min;
			model.dataRange[1] = max;
			model.lookupTable.setMappingRange(model.dataRange[0], model.dataRange[1]);
			model.lookupTable.updateRange();
			publicAPI.applyMode();
		}
	};
	publicAPI.applyMode();
}
var DEFAULT_VALUES = {
	mode: Mode.Preset,
	presetName: Defaults.Preset,
	rgbPoints: Defaults.RGBPoints,
	hsvPoints: Defaults.HSVPoints,
	nodes: Defaults.Nodes,
	arrayName: "No array associated",
	arrayLocation: "pointData",
	dataRange: [0, 1]
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, ["arrayName"]);
	macros_default.get(publicAPI, model, [
		"mode",
		"lookupTable",
		"presetName",
		"rgbPoints",
		"hsvPoints",
		"nodes",
		"dataRange"
	]);
	vtkLookupTableProxy(publicAPI, model);
	macros_default.proxy(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkLookupTableProxy");
var LookupTableProxy_default = {
	newInstance,
	extend,
	Mode,
	Defaults
};
//#endregion
export { Mode, LookupTableProxy_default as default, newInstance };
