// @kitware/vtk.js/Proxy/Core/PiecewiseFunctionProxy.js
import macros_default from "../../macros.js";
import PiecewiseFunction_default from "../../Common/DataModel/PiecewiseFunction.js";
import PiecewiseGaussianWidget_default from "../../Interaction/Widgets/PiecewiseGaussianWidget.js";
import Constants_default from "./PiecewiseFunctionProxy/Constants.js";
//#region Sources/Proxy/Core/PiecewiseFunctionProxy/index.js
var { Mode, Defaults } = Constants_default;
function applyPointsToPiecewiseFunction(points, range, pwf) {
	const width = range[1] - range[0];
	const rescaled = points.map(([x, y]) => [x * width + range[0], y]);
	pwf.removeAllPoints();
	rescaled.forEach(([x, y]) => pwf.addPoint(x, y));
}
function applyNodesToPiecewiseFunction(nodes, range, pwf) {
	const width = range[1] - range[0];
	const rescaled = nodes.map((n) => ({
		...n,
		x: n.x * width + range[0]
	}));
	pwf.setNodes(rescaled);
}
function copyGaussians(gaussians) {
	return gaussians.map((g) => ({ ...g }));
}
function vtkPiecewiseFunctionProxy(publicAPI, model) {
	model.classHierarchy.push("vtkPiecewiseFunctionProxy");
	model.piecewiseFunction = model.piecewiseFunction || PiecewiseFunction_default.newInstance();
	publicAPI.setGaussians = (gaussians) => {
		model.gaussians = copyGaussians(gaussians || []);
		if (model.gaussians.length === 0) model.gaussians = copyGaussians(Defaults.Gaussians);
		publicAPI.applyMode();
	};
	publicAPI.setPoints = (points) => {
		model.points = (points || []).slice();
		if (model.points.length === 0) model.points = Defaults.Points.slice();
		publicAPI.applyMode();
	};
	publicAPI.setNodes = (nodes) => {
		model.nodes = (nodes || []).slice();
		if (model.nodes.length === 0) model.nodes = Defaults.Nodes.slice();
		publicAPI.applyMode();
	};
	publicAPI.setMode = (mode) => {
		if (model.mode === mode) return;
		model.mode = mode;
		publicAPI.applyMode();
	};
	publicAPI.applyMode = () => {
		switch (model.mode) {
			case Mode.Gaussians:
				PiecewiseGaussianWidget_default.applyGaussianToPiecewiseFunction(model.gaussians, 255, model.dataRange, model.piecewiseFunction);
				publicAPI.modified();
				break;
			case Mode.Points:
				applyPointsToPiecewiseFunction(model.points, model.dataRange, model.piecewiseFunction);
				publicAPI.modified();
				break;
			case Mode.Nodes:
				applyNodesToPiecewiseFunction(model.nodes, model.dataRange, model.piecewiseFunction);
				publicAPI.modified();
				break;
			default:
		}
	};
	publicAPI.getLookupTableProxy = () => model.proxyManager.getLookupTable(model.arrayName);
	publicAPI.setDataRange = (min, max) => {
		if (model.dataRange[0] !== min || model.dataRange[1] !== max) {
			model.dataRange[0] = min;
			model.dataRange[1] = max;
			publicAPI.applyMode();
		}
	};
	publicAPI.applyMode();
}
var DEFAULT_VALUES = {
	mode: Mode.Gaussians,
	gaussians: Defaults.Gaussians,
	points: Defaults.Points,
	nodes: Defaults.Nodes,
	arrayName: "No array associated",
	arrayLocation: "pointData",
	dataRange: [0, 1]
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, ["arrayName"]);
	macros_default.get(publicAPI, model, [
		"piecewiseFunction",
		"gaussians",
		"nodes",
		"points",
		"mode",
		"dataRange"
	]);
	vtkPiecewiseFunctionProxy(publicAPI, model);
	macros_default.proxy(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkPiecewiseFunctionProxy");
var PiecewiseFunctionProxy_default = {
	newInstance,
	extend,
	Mode,
	Defaults
};
//#endregion
export { PiecewiseFunctionProxy_default as default, newInstance };
