// @kitware/vtk.js/Proxy/Core/ProxyManager.js
import macros_default from "../../macros.js";
import addRegistrationAPI from "./ProxyManager/core.js";
import addStateAPI from "./ProxyManager/state.js";
import addViewHandlingAPI from "./ProxyManager/view.js";
import addVPropertyHandlingAPI from "./ProxyManager/properties.js";
//#region Sources/Proxy/Core/ProxyManager/index.js
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, {
		proxyIdMapping: {},
		proxyByGroup: {},
		proxyConfiguration: {},
		sv2rMapping: {},
		r2svMapping: {},
		collapseState: {},
		lookupTables: {},
		piecewiseFunctions: {},
		animating: false
	}, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, [
		"proxyConfiguration",
		"activeSource",
		"activeView"
	]);
	macros_default.event(publicAPI, model, "ActiveSourceChange");
	macros_default.event(publicAPI, model, "ActiveViewChange");
	macros_default.event(publicAPI, model, "ProxyRegistrationChange");
	addRegistrationAPI(publicAPI, model);
	addStateAPI(publicAPI, model);
	addViewHandlingAPI(publicAPI, model);
	addVPropertyHandlingAPI(publicAPI, model);
	macros_default.proxy(publicAPI, model);
	model.classHierarchy.push("vtkProxyManager");
}
var newInstance = macros_default.newInstance(extend, "vtkProxyManager");
var ProxyManager_default = {
	newInstance,
	extend
};
//#endregion
export { ProxyManager_default as default, extend, newInstance };
