// @kitware/vtk.js/Rendering/Core/AbstractImageMapper.js
import macros_default from "../../macros.js";
import { createUninitializedBounds } from "../../Common/Core/Math.js";
import AbstractMapper3D_default from "./AbstractMapper3D.js";
//#region Sources/Rendering/Core/AbstractImageMapper/index.js
function vtkAbstractImageMapper(publicAPI, model) {
	model.classHierarchy.push("vtkAbstractImageMapper");
	publicAPI.getIsOpaque = () => true;
	publicAPI.getCurrentImage = () => null;
	publicAPI.getBoundsForSlice = () => {
		macros_default.vtkErrorMacro("vtkAbstractImageMapper.getBoundsForSlice - NOT IMPLEMENTED");
		return createUninitializedBounds();
	};
}
var DEFAULT_VALUES = {
	slice: 0,
	customDisplayExtent: [
		0,
		0,
		0,
		0,
		0,
		0
	],
	useCustomExtents: false,
	backgroundColor: [
		0,
		0,
		0,
		1
	],
	colorTextureWidth: 1024,
	opacityTextureWidth: 1024,
	labelOutlineTextureWidth: 1024
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	AbstractMapper3D_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, [
		"slice",
		"useCustomExtents",
		"colorTextureWidth",
		"opacityTextureWidth",
		"labelOutlineTextureWidth"
	]);
	macros_default.setGetArray(publicAPI, model, ["customDisplayExtent"], 6);
	macros_default.setGetArray(publicAPI, model, ["backgroundColor"], 4);
	vtkAbstractImageMapper(publicAPI, model);
}
var AbstractImageMapper_default = { extend };
//#endregion
export { AbstractImageMapper_default as default, extend };
