// @kitware/vtk.js/Rendering/Core/Actor2D.js
import macros_default from "../../macros.js";
import Prop_default from "./Prop.js";
import { Coordinate } from "./Coordinate/Constants.js";
import Coordinate_default from "./Coordinate.js";
import Property2D_default from "./Property2D.js";
//#region Sources/Rendering/Core/Actor2D/index.js
function vtkActor2D(publicAPI, model) {
	model.classHierarchy.push("vtkActor2D");
	publicAPI.getActors2D = () => publicAPI;
	publicAPI.getIsOpaque = () => {
		if (!model.property) publicAPI.getProperty();
		let isOpaque = model.property.getOpacity() >= 1;
		isOpaque = isOpaque && (!model.texture || !model.texture.isTranslucent());
		return isOpaque;
	};
	publicAPI.hasTranslucentPolygonalGeometry = () => {
		if (model.mapper === null) return false;
		if (model.property === null) publicAPI.setProperty(publicAPI.makeProperty());
		return !publicAPI.getIsOpaque();
	};
	publicAPI.makeProperty = Property2D_default.newInstance;
	publicAPI.ensureProperty = () => {
		if (model.property === null) publicAPI.setProperty(publicAPI.makeProperty());
		return model.property;
	};
	publicAPI.getProperty = () => publicAPI.ensureProperty();
	publicAPI.setDisplayPosition = (XPos, YPos) => {
		model.positionCoordinate.setCoordinateSystem(Coordinate.DISPLAY);
		model.positionCoordinate.setValue(XPos, YPos, 0);
	};
	publicAPI.setWidth = (w) => {
		const pos = model.position2Coordinate.getValue();
		model.position2Coordinate.setCoordinateSystemToNormalizedViewport();
		model.position2Coordinate.setValue(w, pos[1]);
	};
	publicAPI.setHeight = (w) => {
		const pos = model.position2Coordinate.getValue();
		model.position2Coordinate.setCoordinateSystemToNormalizedViewport();
		model.position2Coordinate.setValue(pos[0], w);
	};
	publicAPI.getWidth = () => model.position2Coordinate.getValue()[0];
	publicAPI.getHeight = () => model.position2Coordinate.getValue()[1];
	publicAPI.getMTime = () => {
		let mt = model.mtime;
		if (model.property !== null) {
			const time = model.property.getMTime();
			mt = time > mt ? time : mt;
		}
		mt = model.positionCoordinate.getMTime() > mt ? model.positionCoordinate.getMTime() : mt;
		mt = model.positionCoordinate2.getMTime() > mt ? model.positionCoordinate2.getMTime() : mt;
		return mt;
	};
	publicAPI.getRedrawMTime = () => {
		let mt = model.mtime;
		if (model.mapper !== null) {
			let time = model.mapper.getMTime();
			mt = time > mt ? time : mt;
			if (model.mapper.getInput() !== null) {
				model.mapper.getInputAlgorithm().update();
				time = model.mapper.getInput().getMTime();
				mt = time > mt ? time : mt;
			}
		}
		return mt;
	};
	publicAPI.getBounds = () => {
		if (typeof publicAPI.getMapper().getBounds === "function") {
			model.useBounds = true;
			return publicAPI.getMapper().getBounds();
		}
		model.useBounds = false;
		return [];
	};
	publicAPI.getActualPositionCoordinate = () => model.positionCoordinate;
	publicAPI.getActualPositionCoordinate2 = () => model.positionCoordinate2;
}
var DEFAULT_VALUES = {
	mapper: null,
	property: null,
	layerNumber: 0,
	positionCoordinate: null,
	positionCoordinate2: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	Prop_default.extend(publicAPI, model, initialValues);
	model.positionCoordinate = Coordinate_default.newInstance();
	model.positionCoordinate.setCoordinateSystemToViewport();
	model.positionCoordinate2 = Coordinate_default.newInstance();
	model.positionCoordinate2.setCoordinateSystemToNormalizedViewport();
	model.positionCoordinate2.setValue(.5, .5);
	model.positionCoordinate2.setReferenceCoordinate(model.positionCoordinate);
	macros_default.set(publicAPI, model, ["property"]);
	macros_default.setGet(publicAPI, model, ["mapper", "layerNumber"]);
	vtkActor2D(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkActor2D");
var Actor2D_default = {
	newInstance,
	extend
};
//#endregion
export { Actor2D_default as default, extend, newInstance };
