// @kitware/vtk.js/Rendering/Core/Actor.js
import macros_default from "../../macros.js";
import Prop3D_default from "./Prop3D.js";
import Property_default from "./Property.js";
//#region Sources/Rendering/Core/Actor/index.js
function vtkActor(publicAPI, model) {
	model.classHierarchy.push("vtkActor");
	const superClass = { ...publicAPI };
	publicAPI.getActors = () => [publicAPI];
	publicAPI.getIsOpaque = () => {
		if (model.forceOpaque) return true;
		if (model.forceTranslucent) return false;
		if (!model.properties[0]) publicAPI.getProperty();
		let isOpaque = model.properties[0].getOpacity() >= 1;
		isOpaque = isOpaque && (!model.texture || !model.texture.isTranslucent());
		isOpaque = isOpaque && (!model.mapper || model.mapper.getIsOpaque());
		return isOpaque;
	};
	publicAPI.hasTranslucentPolygonalGeometry = () => {
		if (model.mapper === null) return false;
		if (!model.properties[0]) publicAPI.getProperty();
		return !publicAPI.getIsOpaque();
	};
	publicAPI.makeProperty = Property_default.newInstance;
	publicAPI.getMTime = () => {
		let mt = superClass.getMTime();
		if (model.backfaceProperty !== null) {
			const time = model.backfaceProperty.getMTime();
			mt = time > mt ? time : mt;
		}
		return mt;
	};
	publicAPI.getRedrawMTime = () => {
		let mt = model.mtime;
		if (model.mapper !== null) {
			let time = model.mapper.getMTime();
			mt = time > mt ? time : mt;
			if (model.mapper.getInput() !== null) {
				model.mapper.getInputAlgorithm().update();
				time = model.mapper.getInput().getMTime();
				mt = time > mt ? time : mt;
			}
		}
		return mt;
	};
	publicAPI.getSupportsSelection = () => model.mapper ? model.mapper.getSupportsSelection() : false;
	publicAPI.processSelectorPixelBuffers = (selector, pixelOffsets) => {
		if (model.mapper && model.mapper.processSelectorPixelBuffers) model.mapper.processSelectorPixelBuffers(selector, pixelOffsets);
	};
}
var DEFAULT_VALUES = {
	mapper: null,
	backfaceProperty: null,
	forceOpaque: false,
	forceTranslucent: false
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	Prop3D_default.extend(publicAPI, model, initialValues);
	model.boundsMTime = {};
	macros_default.obj(model.boundsMTime);
	macros_default.setGet(publicAPI, model, [
		"backfaceProperty",
		"forceOpaque",
		"forceTranslucent",
		"mapper"
	]);
	vtkActor(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkActor");
var Actor_default = {
	newInstance,
	extend
};
//#endregion
export { Actor_default as default, extend, newInstance };
