// @kitware/vtk.js/Rendering/Core/AxesActor.js
import macros_default from "../../macros.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import MatrixBuilder_default from "../../Common/Core/MatrixBuilder.js";
import AppendPolyData_default from "../../Filters/General/AppendPolyData.js";
import ArrowSource_default from "../../Filters/Sources/ArrowSource.js";
import Actor_default from "./Actor.js";
import Mapper_default from "./Mapper.js";
//#region Sources/Rendering/Core/AxesActor/index.js
function centerDataSet(ds) {
	const bounds = ds.getPoints().getBounds();
	const center = [
		-(bounds[0] + bounds[1]) * .5,
		-(bounds[2] + bounds[3]) * .5,
		-(bounds[4] + bounds[5]) * .5
	];
	MatrixBuilder_default.buildFromDegree().translate(...center).apply(ds.getPoints().getData());
}
function shiftDataset(ds, axis, invert = false) {
	const bounds = ds.getPoints().getBounds();
	const center = [
		0,
		0,
		0
	];
	if (invert) center[axis] = -bounds[axis * 2 + 1];
	else center[axis] = -bounds[axis * 2];
	MatrixBuilder_default.buildFromDegree().translate(...center).apply(ds.getPoints().getData());
}
function addColor(ds, r, g, b) {
	const size = ds.getPoints().getData().length;
	const rgbArray = new Uint8ClampedArray(size);
	let offset = 0;
	while (offset < size) {
		rgbArray[offset++] = r;
		rgbArray[offset++] = g;
		rgbArray[offset++] = b;
	}
	ds.getPointData().setScalars(DataArray_default.newInstance({
		name: "color",
		numberOfComponents: 3,
		values: rgbArray
	}));
}
function vtkAxesActor(publicAPI, model) {
	model.classHierarchy.push("vtkAxesActor");
	const _mapper = Mapper_default.newInstance();
	publicAPI.setMapper(_mapper);
	publicAPI.update = () => {
		let currentConfig = {
			...model.config,
			...model.xConfig
		};
		const xAxis = ArrowSource_default.newInstance({
			direction: [
				1,
				0,
				0
			],
			...currentConfig
		}).getOutputData();
		if (model.config.recenter) centerDataSet(xAxis);
		else shiftDataset(xAxis, 0, currentConfig.invert);
		addColor(xAxis, ...currentConfig.color);
		currentConfig = {
			...model.config,
			...model.yConfig
		};
		const yAxis = ArrowSource_default.newInstance({
			direction: [
				0,
				1,
				0
			],
			...currentConfig
		}).getOutputData();
		if (model.config.recenter) centerDataSet(yAxis);
		else shiftDataset(yAxis, 1, currentConfig.invert);
		addColor(yAxis, ...currentConfig.color);
		currentConfig = {
			...model.config,
			...model.zConfig
		};
		const zAxis = ArrowSource_default.newInstance({
			direction: [
				0,
				0,
				1
			],
			...currentConfig
		}).getOutputData();
		if (model.config.recenter) centerDataSet(zAxis);
		else shiftDataset(zAxis, 2, currentConfig.invert);
		addColor(zAxis, ...currentConfig.color);
		const source = AppendPolyData_default.newInstance();
		source.setInputData(xAxis);
		source.addInputData(yAxis);
		source.addInputData(zAxis);
		_mapper.setInputConnection(source.getOutputPort());
	};
	publicAPI.update();
	const _debouncedUpdate = macros_default.debounce(publicAPI.update, 0);
	publicAPI.setXAxisColor = (color) => publicAPI.setXConfig({
		...publicAPI.getXConfig(),
		color
	});
	publicAPI.setYAxisColor = (color) => publicAPI.setYConfig({
		...publicAPI.getYConfig(),
		color
	});
	publicAPI.setZAxisColor = (color) => publicAPI.setZConfig({
		...publicAPI.getZConfig(),
		color
	});
	publicAPI.getXAxisColor = () => model.getXConfig().color;
	publicAPI.getYAxisColor = () => model.getYConfig().color;
	publicAPI.getZAxisColor = () => model.getZConfig().color;
	model._onConfigChanged = _debouncedUpdate;
	model._onXConfigChanged = _debouncedUpdate;
	model._onYConfigChanged = _debouncedUpdate;
	model._onZConfigChanged = _debouncedUpdate;
}
function defaultValues(initialValues) {
	return {
		config: {
			recenter: true,
			tipResolution: 60,
			tipRadius: .1,
			tipLength: .2,
			shaftResolution: 60,
			shaftRadius: .03,
			invert: false,
			...initialValues?.config
		},
		xConfig: {
			color: [
				255,
				0,
				0
			],
			invert: false,
			...initialValues?.xConfig
		},
		yConfig: {
			color: [
				255,
				255,
				0
			],
			invert: false,
			...initialValues?.yConfig
		},
		zConfig: {
			color: [
				0,
				128,
				0
			],
			invert: false,
			...initialValues?.zConfig
		}
	};
}
function extend(publicAPI, model, initialValues = {}) {
	Actor_default.extend(publicAPI, model, defaultValues(initialValues));
	macros_default.setGet(publicAPI, model, [
		"config",
		"xConfig",
		"yConfig",
		"zConfig"
	]);
	vtkAxesActor(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkAxesActor");
var AxesActor_default = {
	newInstance,
	extend
};
//#endregion
export { AxesActor_default as default, extend, newInstance };
