// @kitware/vtk.js/Rendering/Core/Camera.js
import macros_default from "../../macros.js";
import { add, cross, normalize4D, radiansFromDegrees } from "../../Common/Core/Math.js";
import { mat4, quat, vec3, vec4 } from "gl-matrix";
//#region Sources/Rendering/Core/Camera/index.js
var { vtkDebugMacro } = macros_default;
function vtkCamera(publicAPI, model) {
	model.classHierarchy.push("vtkCamera");
	const origin = new Float64Array(3);
	const dopbasis = new Float64Array([
		0,
		0,
		-1
	]);
	const upbasis = new Float64Array([
		0,
		1,
		0
	]);
	const tmpMatrix = mat4.identity(new Float64Array(16));
	const tmpMatrix2 = mat4.identity(new Float64Array(16));
	const tmpvec1 = new Float64Array(3);
	const tmpvec2 = new Float64Array(3);
	const tmpvec3 = new Float64Array(3);
	const rotateMatrix = mat4.identity(new Float64Array(16));
	const trans = mat4.identity(new Float64Array(16));
	const newPosition = new Float64Array(3);
	const newFocalPoint = new Float64Array(3);
	function computeViewPlaneNormal() {
		model.viewPlaneNormal[0] = -model.directionOfProjection[0];
		model.viewPlaneNormal[1] = -model.directionOfProjection[1];
		model.viewPlaneNormal[2] = -model.directionOfProjection[2];
	}
	publicAPI.orthogonalizeViewUp = () => {
		const vt = publicAPI.getViewMatrix();
		model.viewUp[0] = vt[4];
		model.viewUp[1] = vt[5];
		model.viewUp[2] = vt[6];
		publicAPI.modified();
	};
	publicAPI.setPosition = (x, y, z) => {
		if (x === model.position[0] && y === model.position[1] && z === model.position[2]) return;
		model.position[0] = x;
		model.position[1] = y;
		model.position[2] = z;
		publicAPI.computeDistance();
		publicAPI.modified();
	};
	publicAPI.setFocalPoint = (x, y, z) => {
		if (x === model.focalPoint[0] && y === model.focalPoint[1] && z === model.focalPoint[2]) return;
		model.focalPoint[0] = x;
		model.focalPoint[1] = y;
		model.focalPoint[2] = z;
		publicAPI.computeDistance();
		publicAPI.modified();
	};
	publicAPI.setDistance = (d) => {
		if (model.distance === d) return;
		model.distance = d;
		if (model.distance < 1e-20) {
			model.distance = 1e-20;
			vtkDebugMacro("Distance is set to minimum.");
		}
		const vec = model.directionOfProjection;
		model.focalPoint[0] = model.position[0] + vec[0] * model.distance;
		model.focalPoint[1] = model.position[1] + vec[1] * model.distance;
		model.focalPoint[2] = model.position[2] + vec[2] * model.distance;
		publicAPI.modified();
	};
	publicAPI.computeDistance = () => {
		const dx = model.focalPoint[0] - model.position[0];
		const dy = model.focalPoint[1] - model.position[1];
		const dz = model.focalPoint[2] - model.position[2];
		model.distance = Math.sqrt(dx * dx + dy * dy + dz * dz);
		if (model.distance < 1e-20) {
			model.distance = 1e-20;
			vtkDebugMacro("Distance is set to minimum.");
			const vec = model.directionOfProjection;
			model.focalPoint[0] = model.position[0] + vec[0] * model.distance;
			model.focalPoint[1] = model.position[1] + vec[1] * model.distance;
			model.focalPoint[2] = model.position[2] + vec[2] * model.distance;
		}
		model.directionOfProjection[0] = dx / model.distance;
		model.directionOfProjection[1] = dy / model.distance;
		model.directionOfProjection[2] = dz / model.distance;
		computeViewPlaneNormal();
	};
	publicAPI.dolly = (amount) => {
		if (amount <= 0) return;
		const d = model.distance / amount;
		publicAPI.setPosition(model.focalPoint[0] - d * model.directionOfProjection[0], model.focalPoint[1] - d * model.directionOfProjection[1], model.focalPoint[2] - d * model.directionOfProjection[2]);
	};
	publicAPI.roll = (angle) => {
		const eye = model.position;
		const at = model.focalPoint;
		const up = model.viewUp;
		const viewUpVec4 = new Float64Array([
			up[0],
			up[1],
			up[2],
			0
		]);
		mat4.identity(rotateMatrix);
		const viewDir = new Float64Array([
			at[0] - eye[0],
			at[1] - eye[1],
			at[2] - eye[2]
		]);
		mat4.rotate(rotateMatrix, rotateMatrix, radiansFromDegrees(angle), viewDir);
		vec4.transformMat4(viewUpVec4, viewUpVec4, rotateMatrix);
		model.viewUp[0] = viewUpVec4[0];
		model.viewUp[1] = viewUpVec4[1];
		model.viewUp[2] = viewUpVec4[2];
		publicAPI.modified();
	};
	publicAPI.azimuth = (angle) => {
		const fp = model.focalPoint;
		mat4.identity(trans);
		mat4.translate(trans, trans, fp);
		mat4.rotate(trans, trans, radiansFromDegrees(angle), model.viewUp);
		mat4.translate(trans, trans, [
			-fp[0],
			-fp[1],
			-fp[2]
		]);
		vec3.transformMat4(newPosition, model.position, trans);
		publicAPI.setPosition(newPosition[0], newPosition[1], newPosition[2]);
	};
	publicAPI.yaw = (angle) => {
		const position = model.position;
		mat4.identity(trans);
		mat4.translate(trans, trans, position);
		mat4.rotate(trans, trans, radiansFromDegrees(angle), model.viewUp);
		mat4.translate(trans, trans, [
			-position[0],
			-position[1],
			-position[2]
		]);
		vec3.transformMat4(newFocalPoint, model.focalPoint, trans);
		publicAPI.setFocalPoint(newFocalPoint[0], newFocalPoint[1], newFocalPoint[2]);
	};
	publicAPI.elevation = (angle) => {
		const fp = model.focalPoint;
		const vt = publicAPI.getViewMatrix();
		const axis = [
			-vt[0],
			-vt[1],
			-vt[2]
		];
		mat4.identity(trans);
		mat4.translate(trans, trans, fp);
		mat4.rotate(trans, trans, radiansFromDegrees(angle), axis);
		mat4.translate(trans, trans, [
			-fp[0],
			-fp[1],
			-fp[2]
		]);
		vec3.transformMat4(newPosition, model.position, trans);
		publicAPI.setPosition(newPosition[0], newPosition[1], newPosition[2]);
	};
	publicAPI.pitch = (angle) => {
		const position = model.position;
		const vt = publicAPI.getViewMatrix();
		const axis = [
			vt[0],
			vt[1],
			vt[2]
		];
		mat4.identity(trans);
		mat4.translate(trans, trans, position);
		mat4.rotate(trans, trans, radiansFromDegrees(angle), axis);
		mat4.translate(trans, trans, [
			-position[0],
			-position[1],
			-position[2]
		]);
		vec3.transformMat4(newFocalPoint, model.focalPoint, trans);
		publicAPI.setFocalPoint(...newFocalPoint);
	};
	publicAPI.zoom = (factor) => {
		if (factor <= 0) return;
		if (model.parallelProjection) model.parallelScale /= factor;
		else model.viewAngle /= factor;
		publicAPI.modified();
	};
	publicAPI.translate = (x, y, z) => {
		const offset = [
			x,
			y,
			z
		];
		add(model.position, offset, model.position);
		add(model.focalPoint, offset, model.focalPoint);
		publicAPI.computeDistance();
		publicAPI.modified();
	};
	publicAPI.applyTransform = (transformMat4) => {
		const vuOld = [...model.viewUp, 1];
		const posNew = [];
		const fpNew = [];
		const vuNew = [];
		vuOld[0] += model.position[0];
		vuOld[1] += model.position[1];
		vuOld[2] += model.position[2];
		vec4.transformMat4(posNew, [...model.position, 1], transformMat4);
		vec4.transformMat4(fpNew, [...model.focalPoint, 1], transformMat4);
		vec4.transformMat4(vuNew, vuOld, transformMat4);
		vuNew[0] -= posNew[0];
		vuNew[1] -= posNew[1];
		vuNew[2] -= posNew[2];
		publicAPI.setPosition(...posNew.slice(0, 3));
		publicAPI.setFocalPoint(...fpNew.slice(0, 3));
		publicAPI.setViewUp(...vuNew.slice(0, 3));
	};
	publicAPI.getThickness = () => model.clippingRange[1] - model.clippingRange[0];
	publicAPI.setThickness = (thickness) => {
		let t = thickness;
		if (t < 1e-20) {
			t = 1e-20;
			vtkDebugMacro("Thickness is set to minimum.");
		}
		publicAPI.setClippingRange(model.clippingRange[0], model.clippingRange[0] + t);
	};
	publicAPI.setThicknessFromFocalPoint = (thickness) => {
		let t = thickness;
		if (t < 1e-20) {
			t = 1e-20;
			vtkDebugMacro("Thickness is set to minimum.");
		}
		publicAPI.setClippingRange(model.distance - t / 2, model.distance + t / 2);
	};
	publicAPI.setRoll = (angle) => {};
	publicAPI.getRoll = () => {};
	publicAPI.setObliqueAngles = (alpha, beta) => {};
	publicAPI.getOrientation = () => {};
	publicAPI.getOrientationWXYZ = () => {};
	publicAPI.getFrustumPlanes = (aspect = 1, planes = new Float64Array(24)) => {
		const normals = [
			[
				1,
				0,
				0,
				1
			],
			[
				-1,
				0,
				0,
				1
			],
			[
				0,
				1,
				0,
				1
			],
			[
				0,
				-1,
				0,
				1
			],
			[
				0,
				0,
				1,
				1
			],
			[
				0,
				0,
				-1,
				1
			]
		];
		const matrix = publicAPI.getCompositeProjectionMatrix(aspect, -1, 1);
		for (let i = 0; i < 6; i++) {
			vec4.transformMat4(normals[i], normals[i], matrix);
			normalize4D(normals[i]);
			planes[4 * i + 0] = normals[i][0];
			planes[4 * i + 1] = normals[i][1];
			planes[4 * i + 2] = normals[i][2];
			planes[4 * i + 3] = normals[i][3];
		}
		return planes;
	};
	publicAPI.getCameraLightTransformMatrix = (matrix) => {
		mat4.copy(matrix, model.cameraLightTransform);
		return matrix;
	};
	publicAPI.computeCameraLightTransform = () => {
		mat4.copy(tmpMatrix, publicAPI.getViewMatrix());
		mat4.invert(tmpMatrix, tmpMatrix);
		mat4.fromScaling(tmpMatrix2, [
			model.distance,
			model.distance,
			model.distance
		]);
		mat4.multiply(tmpMatrix, tmpMatrix, tmpMatrix2);
		mat4.identity(model.cameraLightTransform);
		mat4.translate(model.cameraLightTransform, tmpMatrix, [
			0,
			0,
			-1
		]);
	};
	publicAPI.deepCopy = (sourceCamera) => {};
	publicAPI.physicalOrientationToWorldDirection = (ori) => {
		const oriq = quat.fromValues(ori[0], ori[1], ori[2], ori[3]);
		const coriq = quat.create();
		const qdir = quat.fromValues(0, 0, 1, 0);
		quat.conjugate(coriq, oriq);
		quat.multiply(qdir, oriq, qdir);
		quat.multiply(qdir, qdir, coriq);
		return [
			qdir[0],
			qdir[1],
			qdir[2]
		];
	};
	publicAPI.getPhysicalToWorldMatrix = (result) => {
		publicAPI.getWorldToPhysicalMatrix(result);
		mat4.invert(result, result);
	};
	publicAPI.getWorldToPhysicalMatrix = (result) => {
		mat4.identity(result);
		const physVRight = [3];
		cross(model.physicalViewNorth, model.physicalViewUp, physVRight);
		result[0] = physVRight[0];
		result[1] = physVRight[1];
		result[2] = physVRight[2];
		result[4] = model.physicalViewUp[0];
		result[5] = model.physicalViewUp[1];
		result[6] = model.physicalViewUp[2];
		result[8] = -model.physicalViewNorth[0];
		result[9] = -model.physicalViewNorth[1];
		result[10] = -model.physicalViewNorth[2];
		mat4.transpose(result, result);
		vec3.set(tmpvec1, 1 / model.physicalScale, 1 / model.physicalScale, 1 / model.physicalScale);
		mat4.scale(result, result, tmpvec1);
		mat4.translate(result, result, model.physicalTranslation);
	};
	publicAPI.computeViewParametersFromViewMatrix = (vmat) => {
		mat4.invert(tmpMatrix, vmat);
		vec3.transformMat4(tmpvec1, origin, tmpMatrix);
		publicAPI.computeDistance();
		const oldDist = model.distance;
		publicAPI.setPosition(tmpvec1[0], tmpvec1[1], tmpvec1[2]);
		vec3.transformMat4(tmpvec2, dopbasis, tmpMatrix);
		vec3.subtract(tmpvec2, tmpvec2, tmpvec1);
		vec3.normalize(tmpvec2, tmpvec2);
		publicAPI.setDirectionOfProjection(tmpvec2[0], tmpvec2[1], tmpvec2[2]);
		vec3.transformMat4(tmpvec3, upbasis, tmpMatrix);
		vec3.subtract(tmpvec3, tmpvec3, tmpvec1);
		vec3.normalize(tmpvec3, tmpvec3);
		publicAPI.setViewUp(tmpvec3[0], tmpvec3[1], tmpvec3[2]);
		publicAPI.setDistance(oldDist);
	};
	publicAPI.computeViewParametersFromPhysicalMatrix = (mat) => {
		publicAPI.getWorldToPhysicalMatrix(tmpMatrix);
		mat4.multiply(tmpMatrix, mat, tmpMatrix);
		publicAPI.computeViewParametersFromViewMatrix(tmpMatrix);
	};
	publicAPI.setModelTransformMatrix = (mat) => {
		model.modelTransformMatrix = mat;
	};
	publicAPI.getModelTransformMatrix = () => model.modelTransformMatrix;
	publicAPI.setViewMatrix = (mat) => {
		model.viewMatrix = mat;
		if (model.viewMatrix) {
			mat4.copy(tmpMatrix, model.viewMatrix);
			publicAPI.computeViewParametersFromViewMatrix(tmpMatrix);
			mat4.transpose(model.viewMatrix, model.viewMatrix);
		}
	};
	publicAPI.getViewMatrix = () => {
		if (model.viewMatrix) {
			if (model.modelTransformMatrix) {
				mat4.multiply(tmpMatrix, model.viewMatrix, model.modelTransformMatrix);
				return tmpMatrix;
			}
			return model.viewMatrix;
		}
		mat4.lookAt(tmpMatrix, model.position, model.focalPoint, model.viewUp);
		mat4.transpose(tmpMatrix, tmpMatrix);
		const result = new Float64Array(16);
		if (model.modelTransformMatrix) mat4.multiply(result, tmpMatrix, model.modelTransformMatrix);
		else mat4.copy(result, tmpMatrix);
		return result;
	};
	publicAPI.setProjectionMatrix = (mat) => {
		model.projectionMatrix = mat;
	};
	publicAPI.getProjectionMatrix = (aspect, nearz, farz) => {
		const result = new Float64Array(16);
		mat4.identity(result);
		if (model.projectionMatrix) {
			const scale = 1 / model.physicalScale;
			vec3.set(tmpvec1, scale, scale, scale);
			mat4.copy(result, model.projectionMatrix);
			mat4.scale(result, result, tmpvec1);
			mat4.transpose(result, result);
			return result;
		}
		mat4.identity(tmpMatrix);
		const cWidth = model.clippingRange[1] - model.clippingRange[0];
		const cRange = [model.clippingRange[0] + (nearz + 1) * cWidth / 2, model.clippingRange[0] + (farz + 1) * cWidth / 2];
		if (model.parallelProjection) {
			const width = model.parallelScale * aspect;
			const height = model.parallelScale;
			const xmin = (model.windowCenter[0] - 1) * width;
			const xmax = (model.windowCenter[0] + 1) * width;
			const ymin = (model.windowCenter[1] - 1) * height;
			const ymax = (model.windowCenter[1] + 1) * height;
			mat4.ortho(tmpMatrix, xmin, xmax, ymin, ymax, cRange[0], cRange[1]);
			mat4.transpose(tmpMatrix, tmpMatrix);
		} else if (model.useOffAxisProjection) throw new Error("Off-Axis projection is not supported at this time");
		else {
			const tmp = Math.tan(radiansFromDegrees(model.viewAngle) / 2);
			let width;
			let height;
			if (model.useHorizontalViewAngle === true) {
				width = model.clippingRange[0] * tmp;
				height = model.clippingRange[0] * tmp / aspect;
			} else {
				width = model.clippingRange[0] * tmp * aspect;
				height = model.clippingRange[0] * tmp;
			}
			const xmin = (model.windowCenter[0] - 1) * width;
			const xmax = (model.windowCenter[0] + 1) * width;
			const ymin = (model.windowCenter[1] - 1) * height;
			const ymax = (model.windowCenter[1] + 1) * height;
			const znear = cRange[0];
			const zfar = cRange[1];
			tmpMatrix[0] = 2 * znear / (xmax - xmin);
			tmpMatrix[5] = 2 * znear / (ymax - ymin);
			tmpMatrix[2] = (xmin + xmax) / (xmax - xmin);
			tmpMatrix[6] = (ymin + ymax) / (ymax - ymin);
			tmpMatrix[10] = -(znear + zfar) / (zfar - znear);
			tmpMatrix[14] = -1;
			tmpMatrix[11] = -2 * znear * zfar / (zfar - znear);
			tmpMatrix[15] = 0;
		}
		mat4.copy(result, tmpMatrix);
		return result;
	};
	publicAPI.getCompositeProjectionMatrix = (aspect, nearz, farz) => {
		const vMat = publicAPI.getViewMatrix();
		const pMat = publicAPI.getProjectionMatrix(aspect, nearz, farz);
		mat4.multiply(pMat, vMat, pMat);
		return pMat;
	};
	publicAPI.setDirectionOfProjection = (x, y, z) => {
		if (model.directionOfProjection[0] === x && model.directionOfProjection[1] === y && model.directionOfProjection[2] === z) return;
		model.directionOfProjection[0] = x;
		model.directionOfProjection[1] = y;
		model.directionOfProjection[2] = z;
		const vec = model.directionOfProjection;
		model.focalPoint[0] = model.position[0] + vec[0] * model.distance;
		model.focalPoint[1] = model.position[1] + vec[1] * model.distance;
		model.focalPoint[2] = model.position[2] + vec[2] * model.distance;
		computeViewPlaneNormal();
	};
	publicAPI.setDeviceAngles = (alpha, beta, gamma, screen) => {
		const physVRight = [3];
		cross(model.physicalViewNorth, model.physicalViewUp, physVRight);
		const rotmat = mat4.identity(new Float64Array(16));
		mat4.rotate(rotmat, rotmat, radiansFromDegrees(alpha), model.physicalViewUp);
		mat4.rotate(rotmat, rotmat, radiansFromDegrees(beta), physVRight);
		mat4.rotate(rotmat, rotmat, radiansFromDegrees(gamma), model.physicalViewNorth);
		mat4.rotate(rotmat, rotmat, radiansFromDegrees(-screen), model.physicalViewUp);
		const dop = new Float64Array([
			-model.physicalViewUp[0],
			-model.physicalViewUp[1],
			-model.physicalViewUp[2]
		]);
		const vup = new Float64Array(model.physicalViewNorth);
		vec3.transformMat4(dop, dop, rotmat);
		vec3.transformMat4(vup, vup, rotmat);
		publicAPI.setDirectionOfProjection(dop[0], dop[1], dop[2]);
		publicAPI.setViewUp(vup[0], vup[1], vup[2]);
		publicAPI.modified();
	};
	publicAPI.setOrientationWXYZ = (degrees, x, y, z) => {
		const quatMat = mat4.identity(new Float64Array(16));
		if (degrees !== 0 && (x !== 0 || y !== 0 || z !== 0)) {
			const angle = radiansFromDegrees(degrees);
			const q = quat.create();
			quat.setAxisAngle(q, [
				x,
				y,
				z
			], angle);
			mat4.fromQuat(quatMat, q);
		}
		const newdop = new Float64Array(3);
		vec3.transformMat4(newdop, [
			0,
			0,
			-1
		], quatMat);
		const newvup = new Float64Array(3);
		vec3.transformMat4(newvup, [
			0,
			1,
			0
		], quatMat);
		publicAPI.setDirectionOfProjection(...newdop);
		publicAPI.setViewUp(...newvup);
		publicAPI.modified();
	};
	publicAPI.computeClippingRange = (bounds) => {
		let vn = null;
		let position = null;
		vn = model.viewPlaneNormal;
		position = model.position;
		const a = -vn[0];
		const b = -vn[1];
		const c = -vn[2];
		const d = -(a * position[0] + b * position[1] + c * position[2]);
		const range = [a * bounds[0] + b * bounds[2] + c * bounds[4] + d, 1e-18];
		for (let k = 0; k < 2; k++) for (let j = 0; j < 2; j++) for (let i = 0; i < 2; i++) {
			const dist = a * bounds[i] + b * bounds[2 + j] + c * bounds[4 + k] + d;
			range[0] = dist < range[0] ? dist : range[0];
			range[1] = dist > range[1] ? dist : range[1];
		}
		return range;
	};
}
var DEFAULT_VALUES = {
	position: [
		0,
		0,
		1
	],
	focalPoint: [
		0,
		0,
		0
	],
	viewUp: [
		0,
		1,
		0
	],
	directionOfProjection: [
		0,
		0,
		-1
	],
	parallelProjection: false,
	useHorizontalViewAngle: false,
	viewAngle: 30,
	parallelScale: 1,
	clippingRange: [.01, 1000.01],
	windowCenter: [0, 0],
	viewPlaneNormal: [
		0,
		0,
		1
	],
	useOffAxisProjection: false,
	screenBottomLeft: [
		-.5,
		-.5,
		-.5
	],
	screenBottomRight: [
		.5,
		-.5,
		-.5
	],
	screenTopRight: [
		.5,
		.5,
		-.5
	],
	freezeFocalPoint: false,
	projectionMatrix: null,
	viewMatrix: null,
	modelTransformMatrix: null,
	cameraLightTransform: mat4.create(),
	physicalTranslation: [
		0,
		0,
		0
	],
	physicalScale: 1,
	physicalViewUp: [
		0,
		1,
		0
	],
	physicalViewNorth: [
		0,
		0,
		-1
	]
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.get(publicAPI, model, ["distance"]);
	macros_default.setGet(publicAPI, model, [
		"parallelProjection",
		"useHorizontalViewAngle",
		"viewAngle",
		"parallelScale",
		"useOffAxisProjection",
		"freezeFocalPoint",
		"physicalScale"
	]);
	macros_default.getArray(publicAPI, model, [
		"directionOfProjection",
		"viewPlaneNormal",
		"position",
		"focalPoint"
	]);
	macros_default.setGetArray(publicAPI, model, ["clippingRange", "windowCenter"], 2);
	macros_default.setGetArray(publicAPI, model, [
		"viewUp",
		"screenBottomLeft",
		"screenBottomRight",
		"screenTopRight",
		"physicalTranslation",
		"physicalViewUp",
		"physicalViewNorth"
	], 3);
	vtkCamera(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkCamera");
var Camera_default = {
	newInstance,
	extend
};
//#endregion
export { DEFAULT_VALUES, Camera_default as default, extend, newInstance };
