// @kitware/vtk.js/Rendering/Core/CutterMapper.js
import macros_default from "../../macros.js";
import Mapper_default from "./Mapper.js";
//#region Sources/Rendering/Core/CutterMapper/index.js
var SUPPORTED_IMPLICIT_FUNCTIONS = [
	"vtkPlane",
	"vtkSphere",
	"vtkBox",
	"vtkCylinder",
	"vtkCone"
];
function vtkCutterMapper(publicAPI, model) {
	model.classHierarchy.push("vtkCutterMapper");
	const superClass = { ...publicAPI };
	publicAPI.getMTime = () => {
		let mTime = superClass.getMTime();
		if (model.cutFunction) {
			mTime = Math.max(mTime, model.cutFunction.getMTime());
			const transform = model.cutFunction.getTransform?.();
			if (transform?.getMTime) mTime = Math.max(mTime, transform.getMTime());
		}
		return mTime;
	};
	publicAPI.getSupportedImplicitFunctionName = () => {
		const cutFunction = model.cutFunction;
		if (!cutFunction || !cutFunction.isA) return null;
		return SUPPORTED_IMPLICIT_FUNCTIONS.find((className) => cutFunction.isA(className)) || null;
	};
}
var DEFAULT_VALUES = {
	cutFunction: null,
	cutValue: 0,
	cutWidth: 1.5
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	Mapper_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, [
		"cutFunction",
		"cutValue",
		"cutWidth"
	]);
	vtkCutterMapper(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkCutterMapper");
var CutterMapper_default = {
	newInstance,
	extend
};
//#endregion
export { CutterMapper_default as default, extend, newInstance };
