// @kitware/vtk.js/Rendering/Core/ImageMapper.js
import macros_default from "../../macros.js";
import { clampValue, createUninitializedBounds, getSparseOrthogonalMatrix, multiply3x3_vect3 } from "../../Common/Core/Math.js";
import BoundingBox_default from "../../Common/DataModel/BoundingBox.js";
import CoincidentTopologyHelper_default from "./Mapper/CoincidentTopologyHelper.js";
import Constants_default from "./ImageMapper/Constants.js";
import AbstractImageMapper_default from "./AbstractImageMapper.js";
import { intersectWithLineForCellPicking, intersectWithLineForPointPicking } from "./AbstractImageMapper/helper.js";
//#region Sources/Rendering/Core/ImageMapper/index.js
var { staticOffsetAPI, otherStaticMethods } = CoincidentTopologyHelper_default;
var { SlicingMode } = Constants_default;
function vtkImageMapper(publicAPI, model) {
	model.classHierarchy.push("vtkImageMapper");
	publicAPI.getSliceAtPosition = (pos) => {
		const image = publicAPI.getCurrentImage();
		let pos3;
		if (pos.length === 3) pos3 = pos;
		else if (Number.isFinite(pos)) {
			const bds = image.getBounds();
			switch (model.slicingMode) {
				case SlicingMode.X:
					pos3 = [
						pos,
						(bds[3] + bds[2]) / 2,
						(bds[5] + bds[4]) / 2
					];
					break;
				case SlicingMode.Y:
					pos3 = [
						(bds[1] + bds[0]) / 2,
						pos,
						(bds[5] + bds[4]) / 2
					];
					break;
				case SlicingMode.Z:
					pos3 = [
						(bds[1] + bds[0]) / 2,
						(bds[3] + bds[2]) / 2,
						pos
					];
					break;
				default: break;
			}
		}
		const ijk = [
			0,
			0,
			0
		];
		image.worldToIndex(pos3, ijk);
		const ex = image.getExtent();
		const { ijkMode } = publicAPI.getClosestIJKAxis();
		let slice = 0;
		switch (ijkMode) {
			case SlicingMode.I:
				slice = clampValue(ijk[0], ex[0], ex[1]);
				break;
			case SlicingMode.J:
				slice = clampValue(ijk[1], ex[2], ex[3]);
				break;
			case SlicingMode.K:
				slice = clampValue(ijk[2], ex[4], ex[5]);
				break;
			default: return 0;
		}
		return slice;
	};
	publicAPI.setSliceFromCamera = (cam) => {
		const fp = cam.getFocalPoint();
		switch (model.slicingMode) {
			case SlicingMode.I:
			case SlicingMode.J:
			case SlicingMode.K:
				{
					const slice = publicAPI.getSliceAtPosition(fp);
					publicAPI.setSlice(slice);
				}
				break;
			case SlicingMode.X:
				publicAPI.setSlice(fp[0]);
				break;
			case SlicingMode.Y:
				publicAPI.setSlice(fp[1]);
				break;
			case SlicingMode.Z:
				publicAPI.setSlice(fp[2]);
				break;
			default: break;
		}
	};
	publicAPI.setXSlice = (id) => {
		publicAPI.setSlicingMode(SlicingMode.X);
		publicAPI.setSlice(id);
	};
	publicAPI.setYSlice = (id) => {
		publicAPI.setSlicingMode(SlicingMode.Y);
		publicAPI.setSlice(id);
	};
	publicAPI.setZSlice = (id) => {
		publicAPI.setSlicingMode(SlicingMode.Z);
		publicAPI.setSlice(id);
	};
	publicAPI.setISlice = (id) => {
		publicAPI.setSlicingMode(SlicingMode.I);
		publicAPI.setSlice(id);
	};
	publicAPI.setJSlice = (id) => {
		publicAPI.setSlicingMode(SlicingMode.J);
		publicAPI.setSlice(id);
	};
	publicAPI.setKSlice = (id) => {
		publicAPI.setSlicingMode(SlicingMode.K);
		publicAPI.setSlice(id);
	};
	publicAPI.getSlicingModeNormal = () => {
		const out = [
			0,
			0,
			0
		];
		const mat3 = publicAPI.getCurrentImage().getDirection();
		switch (model.slicingMode) {
			case SlicingMode.X:
				out[0] = 1;
				break;
			case SlicingMode.Y:
				out[1] = 1;
				break;
			case SlicingMode.Z:
				out[2] = 1;
				break;
			case SlicingMode.I:
				multiply3x3_vect3(mat3, [
					1,
					0,
					0
				], out);
				break;
			case SlicingMode.J:
				multiply3x3_vect3(mat3, [
					0,
					1,
					0
				], out);
				break;
			case SlicingMode.K:
				multiply3x3_vect3(mat3, [
					0,
					0,
					1
				], out);
				break;
			default: break;
		}
		return out;
	};
	function computeClosestIJKAxis() {
		let xyzMode;
		switch (model.slicingMode) {
			case SlicingMode.X:
				xyzMode = 0;
				break;
			case SlicingMode.Y:
				xyzMode = 1;
				break;
			case SlicingMode.Z:
				xyzMode = 2;
				break;
			default:
				model.closestIJKAxis = {
					ijkMode: model.slicingMode,
					flip: false
				};
				return;
		}
		const newMatrix = getSparseOrthogonalMatrix(publicAPI.getCurrentImage().getDirection());
		let ijkMode = 0;
		for (; ijkMode < 3; ++ijkMode) if (newMatrix[xyzMode + 3 * ijkMode] !== 0) break;
		const flip = newMatrix[xyzMode + 3 * ijkMode] < 0;
		model.closestIJKAxis = {
			ijkMode,
			flip
		};
	}
	publicAPI.setSlicingMode = (mode) => {
		if (model.slicingMode === mode) return;
		model.slicingMode = mode;
		if (publicAPI.getCurrentImage()) computeClosestIJKAxis();
		publicAPI.modified();
	};
	publicAPI.getClosestIJKAxis = () => {
		if ((model.closestIJKAxis === void 0 || model.closestIJKAxis.ijkMode === SlicingMode.NONE) && publicAPI.getCurrentImage()) computeClosestIJKAxis();
		return model.closestIJKAxis;
	};
	publicAPI.computeBounds = () => {
		const image = publicAPI.getCurrentImage();
		if (!image) {
			BoundingBox_default.reset(model.bounds);
			return;
		}
		if (!model.useCustomExtents) {
			BoundingBox_default.setBounds(model.bounds, image.getBounds());
			return;
		}
		const ex = model.customDisplayExtent.slice();
		const { ijkMode } = publicAPI.getClosestIJKAxis();
		let nSlice = model.slice;
		if (ijkMode !== model.slicingMode) nSlice = publicAPI.getSliceAtPosition(model.slice);
		switch (ijkMode) {
			case SlicingMode.I:
				ex[0] = nSlice;
				ex[1] = nSlice;
				break;
			case SlicingMode.J:
				ex[2] = nSlice;
				ex[3] = nSlice;
				break;
			case SlicingMode.K:
				ex[4] = nSlice;
				ex[5] = nSlice;
				break;
			default: break;
		}
		BoundingBox_default.setBounds(model.bounds, image.extentToBounds(ex));
	};
	publicAPI.getBoundsForSlice = (slice = model.slice, halfThickness = 0) => {
		const image = publicAPI.getCurrentImage();
		if (!image) return createUninitializedBounds();
		const extent = image.getSpatialExtent();
		const { ijkMode } = publicAPI.getClosestIJKAxis();
		let nSlice = slice;
		if (ijkMode !== model.slicingMode) nSlice = publicAPI.getSliceAtPosition(slice);
		switch (ijkMode) {
			case SlicingMode.I:
				extent[0] = nSlice - halfThickness;
				extent[1] = nSlice + halfThickness;
				break;
			case SlicingMode.J:
				extent[2] = nSlice - halfThickness;
				extent[3] = nSlice + halfThickness;
				break;
			case SlicingMode.K:
				extent[4] = nSlice - halfThickness;
				extent[5] = nSlice + halfThickness;
				break;
			default: break;
		}
		return image.extentToBounds(extent);
	};
	publicAPI.intersectWithLineForPointPicking = (p1, p2) => intersectWithLineForPointPicking(p1, p2, publicAPI);
	publicAPI.intersectWithLineForCellPicking = (p1, p2) => intersectWithLineForCellPicking(p1, p2, publicAPI);
	publicAPI.getCurrentImage = () => publicAPI.getInputData();
}
var DEFAULT_VALUES = {
	slicingMode: SlicingMode.NONE,
	closestIJKAxis: {
		ijkMode: SlicingMode.NONE,
		flip: false
	},
	renderToRectangle: false,
	sliceAtFocalPoint: false,
	preferSizeOverAccuracy: false
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	AbstractImageMapper_default.extend(publicAPI, model, initialValues);
	macros_default.get(publicAPI, model, ["slicingMode"]);
	macros_default.setGet(publicAPI, model, [
		"closestIJKAxis",
		"renderToRectangle",
		"sliceAtFocalPoint",
		"preferSizeOverAccuracy"
	]);
	CoincidentTopologyHelper_default.implementCoincidentTopologyMethods(publicAPI, model);
	vtkImageMapper(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkImageMapper");
var ImageMapper_default = {
	newInstance,
	extend,
	...staticOffsetAPI,
	...otherStaticMethods,
	...Constants_default
};
//#endregion
export { ImageMapper_default as default, extend, newInstance };
