// @kitware/vtk.js/Rendering/Core/ImageArrayMapper.js
import macros_default from "../../macros.js";
import { createUninitializedBounds } from "../../Common/Core/Math.js";
import BoundingBox_default from "../../Common/DataModel/BoundingBox.js";
import CoincidentTopologyHelper_default from "./Mapper/CoincidentTopologyHelper.js";
import AbstractImageMapper_default from "./AbstractImageMapper.js";
import { intersectWithLineForCellPicking, intersectWithLineForPointPicking } from "./AbstractImageMapper/helper.js";
import ImageMapper_default from "./ImageMapper.js";
//#region Sources/Rendering/Core/ImageArrayMapper/index.js
var { staticOffsetAPI, otherStaticMethods } = CoincidentTopologyHelper_default;
var { vtkErrorMacro, vtkWarningMacro } = macros_default;
var { SlicingMode } = ImageMapper_default;
function vtkImageArrayMapper(publicAPI, model) {
	model.classHierarchy.push("vtkImageArrayMapper");
	const _computeSliceToSubSliceMap = () => {
		const inputCollection = publicAPI.getInputData();
		if (!inputCollection || inputCollection.empty()) {
			if (model.sliceToSubSliceMap.length !== 0) {
				model.sliceToSubSliceMap.length = 0;
				publicAPI.modified();
			}
			return;
		}
		if (model.sliceToSubSliceMap.length === 0 || inputCollection.getMTime() > publicAPI.getMTime()) {
			model.sliceToSubSliceMap = inputCollection.map((image, index) => {
				const dim = image.getDimensions();
				const out = new Array(dim[model.slicingMode]);
				for (let i = 0; i < out.length; ++i) out[i] = {
					imageIndex: index,
					subSlice: i
				};
				return out;
			}).flat();
			publicAPI.modified();
		}
	};
	const _superSetInputData = publicAPI.setInputData;
	const _superSetInputConnection = publicAPI.setInputConnection;
	publicAPI.setInputData = (inputData) => {
		_superSetInputData(inputData);
		_computeSliceToSubSliceMap();
	};
	publicAPI.setInputConnection = (outputPort, port = 0) => {
		_superSetInputConnection(outputPort, port);
		_computeSliceToSubSliceMap();
	};
	publicAPI.getImage = (slice = publicAPI.getSlice()) => {
		const inputCollection = publicAPI.getInputData();
		if (!inputCollection) vtkWarningMacro("No input set.");
		else if (slice < 0 || slice >= publicAPI.getTotalSlices()) vtkWarningMacro("Invalid slice number.");
		else {
			_computeSliceToSubSliceMap();
			return inputCollection.getItem(model.sliceToSubSliceMap[slice].imageIndex);
		}
		return null;
	};
	publicAPI.computeBounds = () => {
		const image = publicAPI.getCurrentImage();
		if (!image) {
			BoundingBox_default.reset(model.bounds);
			return;
		}
		if (!model.useCustomExtents) {
			BoundingBox_default.setBounds(model.bounds, image.getBounds());
			return;
		}
		const ex = model.customDisplayExtent.slice();
		const nSlice = publicAPI.getSubSlice();
		ex[4] = nSlice;
		ex[5] = nSlice;
		BoundingBox_default.setBounds(model.bounds, image.extentToBounds(ex));
	};
	publicAPI.getBoundsForSlice = (slice = publicAPI.getSlice(), halfThickness = 0) => {
		const image = publicAPI.getImage(slice);
		if (!image) return createUninitializedBounds();
		const extent = image.getSpatialExtent();
		const nSlice = publicAPI.getSubSlice(slice);
		extent[4] = nSlice - halfThickness;
		extent[5] = nSlice + halfThickness;
		return image.extentToBounds(extent);
	};
	publicAPI.getClosestIJKAxis = () => ({
		ijkMode: model.slicingMode,
		flip: false
	});
	publicAPI.computeTotalSlices = () => {
		const inputCollection = publicAPI.getInputData();
		const collectionLength = inputCollection.getNumberOfItems();
		let slicesCount = 0;
		for (let i = 0; i < collectionLength; ++i) {
			const image = inputCollection.getItem(i);
			if (image) slicesCount += image.getDimensions()[model.slicingMode];
		}
		return slicesCount;
	};
	publicAPI.getTotalSlices = () => {
		_computeSliceToSubSliceMap();
		return model.sliceToSubSliceMap.length;
	};
	publicAPI.setSlice = (slice) => {
		if (!publicAPI.getInputData()) {
			vtkWarningMacro("No input set.");
			return;
		}
		const totalSlices = publicAPI.getTotalSlices();
		if (slice >= 0 && slice < totalSlices) {
			model.slice = slice;
			publicAPI.modified();
		} else vtkErrorMacro(`Slice number out of range. Acceptable range is: [0, ${totalSlices > 0 ? totalSlices - 1 : 0}]slice <= totalSlices`);
	};
	publicAPI.computeSlice = (imageIndex, subSlice) => {
		_computeSliceToSubSliceMap();
		return model.sliceToSubSliceMap.findIndex((x) => x.imageIndex === imageIndex && x.subSlice === subSlice);
	};
	publicAPI.getImageIndex = (slice = publicAPI.getSlice()) => {
		_computeSliceToSubSliceMap();
		return model.sliceToSubSliceMap[slice]?.imageIndex;
	};
	publicAPI.getSubSlice = (slice = publicAPI.getSlice()) => {
		_computeSliceToSubSliceMap();
		return model.sliceToSubSliceMap[slice]?.subSlice;
	};
	publicAPI.getCurrentImage = () => publicAPI.getImage(publicAPI.getSlice());
	publicAPI.intersectWithLineForPointPicking = (p1, p2) => intersectWithLineForPointPicking(p1, p2, publicAPI);
	publicAPI.intersectWithLineForCellPicking = (p1, p2) => intersectWithLineForCellPicking(p1, p2, publicAPI);
}
var DEFAULT_VALUES = {
	slicingMode: SlicingMode.K,
	sliceToSubSliceMap: []
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	AbstractImageMapper_default.extend(publicAPI, model, initialValues);
	macros_default.get(publicAPI, model, ["slicingMode"]);
	CoincidentTopologyHelper_default.implementCoincidentTopologyMethods(publicAPI, model);
	vtkImageArrayMapper(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkImageArrayMapper");
var ImageArrayMapper_default = {
	newInstance,
	extend,
	...staticOffsetAPI,
	...otherStaticMethods
};
//#endregion
export { ImageArrayMapper_default as default, extend, newInstance };
