// @kitware/vtk.js/Rendering/Core/ImageResliceMapper.js
import macros_default from "../../macros.js";
import BoundingBox_default from "../../Common/DataModel/BoundingBox.js";
import CoincidentTopologyHelper_default from "./Mapper/CoincidentTopologyHelper.js";
import AbstractImageMapper_default from "./AbstractImageMapper.js";
import Constants_default from "./ImageResliceMapper/Constants.js";
//#region Sources/Rendering/Core/ImageResliceMapper/index.js
var { SlabTypes } = Constants_default;
var { staticOffsetAPI, otherStaticMethods } = CoincidentTopologyHelper_default;
function vtkImageResliceMapper(publicAPI, model) {
	model.classHierarchy.push("vtkImageResliceMapper");
	publicAPI.computeBounds = () => {
		const image = publicAPI.getInputData();
		if (publicAPI.getSlicePolyData()) BoundingBox_default.setBounds(model.bounds, publicAPI.getSlicePolyData().getBounds());
		else if (image) {
			BoundingBox_default.setBounds(model.bounds, image.getBounds());
			if (publicAPI.getSlicePlane()) BoundingBox_default.cutWithPlane(model.bounds, publicAPI.getSlicePlane().getOrigin(), publicAPI.getSlicePlane().getNormal());
		} else BoundingBox_default.reset(model.bounds);
	};
}
var defaultValues = (initialValues) => ({
	slabThickness: 0,
	slabTrapezoidIntegration: 0,
	slabType: SlabTypes.MEAN,
	slicePlane: null,
	slicePolyData: null,
	...initialValues
});
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, defaultValues(initialValues));
	AbstractImageMapper_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, [
		"slabThickness",
		"slabTrapezoidIntegration",
		"slabType",
		"slicePlane",
		"slicePolyData"
	]);
	CoincidentTopologyHelper_default.implementCoincidentTopologyMethods(publicAPI, model);
	vtkImageResliceMapper(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkImageResliceMapper");
var ImageResliceMapper_default = {
	newInstance,
	extend,
	...staticOffsetAPI,
	...otherStaticMethods
};
//#endregion
export { ImageResliceMapper_default as default, extend, newInstance };
