// @kitware/vtk.js/Rendering/Core/ImageSlice.js
import macros_default from "../../macros.js";
import BoundingBox_default from "../../Common/DataModel/BoundingBox.js";
import Prop3D_default from "./Prop3D.js";
import ImageProperty_default from "./ImageProperty.js";
import { mat4 } from "gl-matrix";
//#region Sources/Rendering/Core/ImageSlice/index.js
function vtkImageSlice(publicAPI, model) {
	model.classHierarchy.push("vtkImageSlice");
	publicAPI.getActors = () => publicAPI;
	publicAPI.getImages = () => publicAPI;
	publicAPI.getIsOpaque = () => {
		if (model.forceOpaque) return true;
		if (model.forceTranslucent) return false;
		if (!model.properties[0]) publicAPI.getProperty();
		let isOpaque = model.properties[0].getOpacity() >= 1;
		isOpaque = isOpaque && (!model.mapper || model.mapper.getIsOpaque());
		return isOpaque;
	};
	publicAPI.hasTranslucentPolygonalGeometry = () => false;
	publicAPI.makeProperty = ImageProperty_default.newInstance;
	publicAPI.getBoundsForSlice = (slice, thickness) => {
		const bds = model.mapper.getBoundsForSlice(slice, thickness);
		if (!BoundingBox_default.isValid(bds)) return bds;
		publicAPI.computeMatrix();
		const tmp4 = new Float64Array(16);
		mat4.transpose(tmp4, model.matrix);
		return BoundingBox_default.transformBounds(bds, tmp4);
	};
	publicAPI.getMinXBound = () => publicAPI.getBoundsByReference()[0];
	publicAPI.getMaxXBound = () => publicAPI.getBoundsByReference()[1];
	publicAPI.getMinYBound = () => publicAPI.getBoundsByReference()[2];
	publicAPI.getMaxYBound = () => publicAPI.getBoundsByReference()[3];
	publicAPI.getMinZBound = () => publicAPI.getBoundsByReference()[4];
	publicAPI.getMaxZBound = () => publicAPI.getBoundsByReference()[5];
	publicAPI.getRedrawMTime = () => {
		let mt = model.mtime;
		if (model.mapper !== null) {
			let time = model.mapper.getMTime();
			mt = time > mt ? time : mt;
			if (model.mapper.getInput() !== null) {
				model.mapper.getInputAlgorithm().update();
				time = model.mapper.getInput().getMTime();
				mt = time > mt ? time : mt;
			}
		}
		model.properties.forEach((property) => {
			mt = Math.max(mt, property.getMTime());
			const rgbFunc = property.getRGBTransferFunction();
			if (rgbFunc !== null) mt = Math.max(mt, rgbFunc.getMTime());
		});
		return mt;
	};
	publicAPI.getSupportsSelection = () => model.mapper ? model.mapper.getSupportsSelection() : false;
}
var DEFAULT_VALUES = {
	mapper: null,
	forceOpaque: false,
	forceTranslucent: false
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	Prop3D_default.extend(publicAPI, model, initialValues);
	model.boundsMTime = {};
	macros_default.obj(model.boundsMTime);
	macros_default.setGet(publicAPI, model, [
		"mapper",
		"forceOpaque",
		"forceTranslucent"
	]);
	vtkImageSlice(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkImageSlice");
var ImageSlice_default = {
	newInstance,
	extend
};
//#endregion
export { ImageSlice_default as default, extend, newInstance };
