// @kitware/vtk.js/Rendering/Core/ImageProperty.js
import macros_default from "../../macros.js";
import Constants_default from "./ImageProperty/Constants.js";
//#region Sources/Rendering/Core/ImageProperty/index.js
var { InterpolationType } = Constants_default;
var { vtkErrorMacro } = macros_default;
var VTK_MAX_VRCOMP = 4;
function vtkImageProperty(publicAPI, model) {
	model.classHierarchy.push("vtkImageProperty");
	publicAPI.getMTime = () => {
		let mTime = model.mtime;
		let time;
		for (let index = 0; index < VTK_MAX_VRCOMP; index++) {
			if (model.componentData[index].rGBTransferFunction) {
				time = model.componentData[index].rGBTransferFunction.getMTime();
				mTime = mTime > time ? mTime : time;
			}
			if (model.componentData[index].piecewiseFunction) {
				time = model.componentData[index].piecewiseFunction.getMTime();
				mTime = mTime > time ? mTime : time;
			}
		}
		return mTime;
	};
	publicAPI.setRGBTransferFunction = (index = 0, func = null) => {
		let idx = index;
		let transferFunc = func;
		if (!Number.isInteger(index)) {
			transferFunc = index;
			idx = 0;
		}
		if (model.componentData[idx].rGBTransferFunction !== transferFunc) {
			model.componentData[idx].rGBTransferFunction = transferFunc;
			publicAPI.modified();
			return true;
		}
		return false;
	};
	publicAPI.getRGBTransferFunction = (idx = 0) => model.componentData[idx].rGBTransferFunction;
	publicAPI.setPiecewiseFunction = (index = 0, func = null) => {
		let idx = index;
		let transferFunc = func;
		if (!Number.isInteger(index)) {
			transferFunc = index;
			idx = 0;
		}
		if (model.componentData[idx].piecewiseFunction !== transferFunc) {
			model.componentData[idx].piecewiseFunction = transferFunc;
			publicAPI.modified();
			return true;
		}
		return false;
	};
	publicAPI.getPiecewiseFunction = (idx = 0) => model.componentData[idx].piecewiseFunction;
	publicAPI.setScalarOpacity = (index = 0, func = null) => {
		let idx = index;
		let transferFunc = func;
		if (!Number.isInteger(index)) {
			transferFunc = index;
			idx = 0;
		}
		return publicAPI.setPiecewiseFunction(idx, transferFunc);
	};
	publicAPI.getScalarOpacity = (idx = 0) => publicAPI.getPiecewiseFunction(idx);
	publicAPI.setComponentWeight = (index = 0, value = 1) => {
		if (index < 0 || index >= VTK_MAX_VRCOMP) {
			vtkErrorMacro("Invalid index");
			return false;
		}
		const val = Math.min(1, Math.max(0, value));
		if (model.componentData[index].componentWeight !== val) {
			model.componentData[index].componentWeight = val;
			publicAPI.modified();
			return true;
		}
		return false;
	};
	publicAPI.getComponentWeight = (index = 0) => {
		if (index < 0 || index >= VTK_MAX_VRCOMP) {
			vtkErrorMacro("Invalid index");
			return 0;
		}
		return model.componentData[index].componentWeight;
	};
	publicAPI.setInterpolationTypeToNearest = () => publicAPI.setInterpolationType(InterpolationType.NEAREST);
	publicAPI.setInterpolationTypeToLinear = () => publicAPI.setInterpolationType(InterpolationType.LINEAR);
	publicAPI.getInterpolationTypeAsString = () => macros_default.enumToString(InterpolationType, model.interpolationType);
}
var defaultValues = (initialValues) => ({
	independentComponents: false,
	interpolationType: InterpolationType.LINEAR,
	colorWindow: 255,
	colorLevel: 127.5,
	ambient: 1,
	diffuse: 0,
	opacity: 1,
	useLookupTableScalarRange: false,
	useLabelOutline: false,
	labelOutlineThickness: [1],
	labelOutlineOpacity: 1,
	updatedExtents: [],
	...initialValues
});
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, defaultValues(initialValues));
	macros_default.obj(publicAPI, model);
	if (!model.componentData) {
		model.componentData = [];
		for (let i = 0; i < VTK_MAX_VRCOMP; i++) model.componentData.push({
			rGBTransferFunction: null,
			piecewiseFunction: null,
			componentWeight: 1
		});
	}
	macros_default.setGet(publicAPI, model, [
		"independentComponents",
		"interpolationType",
		"colorWindow",
		"colorLevel",
		"ambient",
		"diffuse",
		"opacity",
		"useLookupTableScalarRange",
		"useLabelOutline",
		"labelOutlineOpacity",
		"updatedExtents"
	]);
	macros_default.setGetArray(publicAPI, model, ["labelOutlineThickness"]);
	vtkImageProperty(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkImageProperty");
var ImageProperty_default = {
	newInstance,
	extend
};
//#endregion
export { ImageProperty_default as default, extend, newInstance };
