// @kitware/vtk.js/Rendering/Core/InteractorStyle.js
import macros_default from "../../macros.js";
import InteractorObserver_default from "./InteractorObserver.js";
import Constants_default from "./InteractorStyle/Constants.js";
//#region Sources/Rendering/Core/InteractorStyle/index.js
var { States } = Constants_default;
var stateNames = {
	Rotate: States.IS_ROTATE,
	Pan: States.IS_PAN,
	Spin: States.IS_SPIN,
	Dolly: States.IS_DOLLY,
	CameraPose: States.IS_CAMERA_POSE,
	WindowLevel: States.IS_WINDOW_LEVEL,
	Slice: States.IS_SLICE
};
function vtkInteractorStyle(publicAPI, model) {
	model.classHierarchy.push("vtkInteractorStyle");
	Object.keys(stateNames).forEach((key) => {
		macros_default.event(publicAPI, model, `Start${key}Event`);
		publicAPI[`start${key}`] = () => {
			if (model.state !== States.IS_NONE) return;
			model.state = stateNames[key];
			model._interactor.requestAnimation(publicAPI);
			publicAPI.invokeStartInteractionEvent({ type: "StartInteractionEvent" });
			publicAPI[`invokeStart${key}Event`]({ type: `Start${key}Event` });
		};
		macros_default.event(publicAPI, model, `End${key}Event`);
		publicAPI[`end${key}`] = () => {
			if (model.state !== stateNames[key]) return;
			model.state = States.IS_NONE;
			model._interactor.cancelAnimation(publicAPI);
			publicAPI.invokeEndInteractionEvent({ type: "EndInteractionEvent" });
			publicAPI[`invokeEnd${key}Event`]({ type: `End${key}Event` });
			model._interactor.render();
		};
	});
	model.getRenderer = (callData) => model.focusedRenderer || callData.pokedRenderer;
	publicAPI.handleKeyPress = (callData) => {
		const rwi = model._interactor;
		let ac = null;
		switch (callData.key) {
			case "r":
			case "R":
				model.getRenderer(callData).resetCamera();
				rwi.render();
				break;
			case "w":
			case "W":
				ac = model.getRenderer(callData).getActors();
				ac.forEach((anActor) => {
					const prop = anActor.getProperty();
					if (prop.setRepresentationToWireframe) prop.setRepresentationToWireframe();
				});
				rwi.render();
				break;
			case "s":
			case "S":
				ac = model.getRenderer(callData).getActors();
				ac.forEach((anActor) => {
					const prop = anActor.getProperty();
					if (prop.setRepresentationToSurface) prop.setRepresentationToSurface();
				});
				rwi.render();
				break;
			case "v":
			case "V":
				ac = model.getRenderer(callData).getActors();
				ac.forEach((anActor) => {
					const prop = anActor.getProperty();
					if (prop.setRepresentationToPoints) prop.setRepresentationToPoints();
				});
				rwi.render();
				break;
			default: break;
		}
	};
}
var DEFAULT_VALUES = {
	state: States.IS_NONE,
	handleObservers: 1,
	autoAdjustCameraClippingRange: 1
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	InteractorObserver_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, ["focusedRenderer"]);
	vtkInteractorStyle(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkInteractorStyle");
var InteractorStyle_default = {
	newInstance,
	extend,
	...Constants_default
};
//#endregion
export { InteractorStyle_default as default, extend, newInstance };
