// @kitware/vtk.js/Rendering/Core/InteractorObserver.js
import macros_default from "../../macros.js";
import RenderWindowInteractor_default from "./RenderWindowInteractor.js";
//#region Sources/Rendering/Core/InteractorObserver/index.js
var { vtkErrorMacro, VOID } = macros_default;
function computeWorldToDisplay(renderer, x, y, z) {
	return renderer.getRenderWindow().getViews()[0].worldToDisplay(x, y, z, renderer);
}
function computeDisplayToWorld(renderer, x, y, z) {
	return renderer.getRenderWindow().getViews()[0].displayToWorld(x, y, z, renderer);
}
var STATIC = {
	computeWorldToDisplay,
	computeDisplayToWorld
};
function vtkInteractorObserver(publicAPI, model) {
	model.classHierarchy.push("vtkInteractorObserver");
	const superClass = { ...publicAPI };
	function unsubscribeFromEvents() {
		while (model.subscribedEvents.length) model.subscribedEvents.pop().unsubscribe();
	}
	function subscribeToEvents() {
		RenderWindowInteractor_default.handledEvents.forEach((eventName) => {
			if (publicAPI[`handle${eventName}`]) model.subscribedEvents.push(model._interactor[`on${eventName}`]((callData) => {
				if (model.processEvents) return publicAPI[`handle${eventName}`](callData);
				return VOID;
			}, model.priority));
		});
	}
	publicAPI.setInteractor = (i) => {
		if (i === model._interactor) return;
		unsubscribeFromEvents();
		model._interactor = i;
		if (i && model.enabled) subscribeToEvents();
		publicAPI.modified();
	};
	publicAPI.setEnabled = (enable) => {
		if (enable === model.enabled) return;
		unsubscribeFromEvents();
		if (enable) if (model._interactor) subscribeToEvents();
		else vtkErrorMacro(`
          The interactor must be set before subscribing to events
        `);
		model.enabled = enable;
		publicAPI.modified();
	};
	publicAPI.computeDisplayToWorld = (renderer, x, y, z) => {
		if (!renderer) return null;
		return model._interactor.getView().displayToWorld(x, y, z, renderer);
	};
	publicAPI.computeWorldToDisplay = (renderer, x, y, z) => {
		if (!renderer) return null;
		return model._interactor.getView().worldToDisplay(x, y, z, renderer);
	};
	publicAPI.setPriority = (priority) => {
		if (superClass.setPriority(priority) && model._interactor) {
			unsubscribeFromEvents();
			subscribeToEvents();
		}
	};
}
var DEFAULT_VALUES = {
	enabled: true,
	priority: 0,
	processEvents: true,
	subscribedEvents: []
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.event(publicAPI, model, "InteractionEvent");
	macros_default.event(publicAPI, model, "StartInteractionEvent");
	macros_default.event(publicAPI, model, "EndInteractionEvent");
	macros_default.get(publicAPI, model, ["_interactor", "enabled"]);
	macros_default.setGet(publicAPI, model, ["priority", "processEvents"]);
	macros_default.moveToProtected(publicAPI, model, ["interactor"]);
	vtkInteractorObserver(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkInteractorObserver");
var InteractorObserver_default = {
	newInstance,
	extend,
	...STATIC
};
//#endregion
export { STATIC, InteractorObserver_default as default, extend, newInstance };
