// @kitware/vtk.js/Rendering/Core/Light.js
import macros_default from "../../macros.js";
import { normalize, radiansFromDegrees } from "../../Common/Core/Math.js";
import { vec3 } from "gl-matrix";
//#region Sources/Rendering/Core/Light/index.js
var LIGHT_TYPES = [
	"HeadLight",
	"CameraLight",
	"SceneLight"
];
function vtkLight(publicAPI, model) {
	model.classHierarchy.push("vtkLight");
	const tmpVec = new Float64Array(3);
	publicAPI.getTransformedPosition = () => {
		if (model.transformMatrix) vec3.transformMat4(tmpVec, model.position, model.transformMatrix);
		else vec3.set(tmpVec, model.position[0], model.position[1], model.position[2]);
		return tmpVec;
	};
	publicAPI.getTransformedFocalPoint = () => {
		if (model.transformMatrix) vec3.transformMat4(tmpVec, model.focalPoint, model.transformMatrix);
		else vec3.set(tmpVec, model.focalPoint[0], model.focalPoint[1], model.focalPoint[2]);
		return tmpVec;
	};
	publicAPI.getDirection = () => {
		if (model.directionMTime < model.mtime) {
			vec3.sub(model.direction, model.focalPoint, model.position);
			normalize(model.direction);
			model.directionMTime = model.mtime;
		}
		return model.direction;
	};
	publicAPI.setDirection = (directionVector) => {
		const newFocalPoint = new Float64Array(3);
		vec3.sub(newFocalPoint, model.position, directionVector);
		model.focalPoint = newFocalPoint;
	};
	publicAPI.setDirectionAngle = (elevation, azimuth) => {
		const elevationRadians = radiansFromDegrees(elevation);
		const azimuthRadians = radiansFromDegrees(azimuth);
		publicAPI.setPosition(Math.cos(elevationRadians) * Math.sin(azimuthRadians), Math.sin(elevationRadians), Math.cos(elevationRadians) * Math.cos(azimuthRadians));
		publicAPI.setFocalPoint(0, 0, 0);
		publicAPI.setPositional(0);
	};
	publicAPI.setLightTypeToHeadLight = () => {
		publicAPI.setLightType("HeadLight");
	};
	publicAPI.setLightTypeToCameraLight = () => {
		publicAPI.setLightType("CameraLight");
	};
	publicAPI.setLightTypeToSceneLight = () => {
		publicAPI.setTransformMatrix(null);
		publicAPI.setLightType("SceneLight");
	};
	publicAPI.lightTypeIsHeadLight = () => model.lightType === "HeadLight";
	publicAPI.lightTypeIsSceneLight = () => model.lightType === "SceneLight";
	publicAPI.lightTypeIsCameraLight = () => model.lightType === "CameraLight";
}
var DEFAULT_VALUES = {
	switch: true,
	intensity: 1,
	color: [
		1,
		1,
		1
	],
	position: [
		0,
		0,
		1
	],
	focalPoint: [
		0,
		0,
		0
	],
	positional: false,
	exponent: 1,
	coneAngle: 30,
	coneFalloff: 5,
	attenuationValues: [
		1,
		0,
		0
	],
	transformMatrix: null,
	lightType: "SceneLight",
	shadowAttenuation: 1,
	direction: [
		0,
		0,
		0
	],
	directionMTime: 0
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, [
		"intensity",
		"switch",
		"positional",
		"exponent",
		"coneAngle",
		"coneFalloff",
		"transformMatrix",
		"lightType",
		"shadowAttenuation",
		"attenuationValues"
	]);
	macros_default.setGetArray(publicAPI, model, [
		"color",
		"position",
		"focalPoint",
		"attenuationValues"
	], 3);
	vtkLight(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkLight");
var Light_default = {
	newInstance,
	extend,
	LIGHT_TYPES
};
//#endregion
export { LIGHT_TYPES, Light_default as default, extend, newInstance };
