// @kitware/vtk.js/Rendering/Core/Mapper2D.js
import macros_default from "../../macros.js";
import { isNan } from "../../Common/Core/Math.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import ImageData_default from "../../Common/DataModel/ImageData.js";
import Constants_default from "../../Common/Core/ScalarsToColors/Constants.js";
import Constants_default$1 from "./Mapper/Constants.js";
import LookupTable_default from "../../Common/Core/LookupTable.js";
import AbstractMapper_default from "./AbstractMapper.js";
//#region Sources/Rendering/Core/Mapper2D/index.js
var { ColorMode, ScalarMode, GetArray } = Constants_default$1;
var { VectorMode } = Constants_default;
var { VtkDataTypes } = DataArray_default;
/**
* Increase by one the 3D coordinates
* It will follow a zigzag pattern so that each coordinate is the neighbor of the next coordinate
* This enables interpolation between two texels without issues
* Note: texture coordinates can't be interpolated using this pattern
* @param {vec3} coordinates The 3D coordinates using integers for each coordinate
* @param {vec3} dimensions The 3D dimensions of the volume
*/
function updateZigzaggingCoordinates(coordinates, dimensions) {
	const directionX = coordinates[1] % 2 === 0 ? 1 : -1;
	coordinates[0] += directionX;
	if (coordinates[0] >= dimensions[0] || coordinates[0] < 0) {
		const directionY = coordinates[2] % 2 === 0 ? 1 : -1;
		coordinates[0] -= directionX;
		coordinates[1] += directionY;
		if (coordinates[1] >= dimensions[1] || coordinates[1] < 0) {
			coordinates[1] -= directionY;
			coordinates[2]++;
		}
	}
}
/**
* Returns the index in the array representing the volume from a 3D coordinate
* @param {vec3} coordinates The 3D integer coordinates
* @param {vec3} dimensions The 3D dimensions of the volume
* @returns The index in a flat array representing the volume
*/
function getIndexFromCoordinates(coordinates, dimensions) {
	return coordinates[0] + dimensions[0] * (coordinates[1] + dimensions[1] * coordinates[2]);
}
/**
* Write texture coordinates for the given `texelIndexPosition` in `textureCoordinate`.
* The `texelIndexPosition` is a floating point number that represents the distance in index space
* from the center of the first texel to the final output position.
* The output is given in texture coordinates and not in index coordinates (this is done at the very end of the function)
* @param {vec3} textureCoordinate The output texture coordinates (to avoid allocating a new Array)
* @param {Number} texelIndexPosition The floating point distance from the center of the first texel, following a zigzag pattern
* @param {vec3} dimensions The 3D dimensions of the volume
*/
function getZigZagTextureCoordinatesFromTexelPosition(textureCoordinate, texelIndexPosition, dimensions) {
	const intTexelIndex = Math.floor(texelIndexPosition);
	const xCoordBeforeWrap = intTexelIndex % (2 * dimensions[0]);
	let xDirection;
	let xEndFlag;
	if (xCoordBeforeWrap < dimensions[0]) {
		textureCoordinate[0] = xCoordBeforeWrap;
		xDirection = 1;
		xEndFlag = textureCoordinate[0] === dimensions[0] - 1;
	} else {
		textureCoordinate[0] = 2 * dimensions[0] - 1 - xCoordBeforeWrap;
		xDirection = -1;
		xEndFlag = textureCoordinate[0] === 0;
	}
	const intRowIndex = Math.floor(intTexelIndex / dimensions[0]);
	const yCoordBeforeWrap = intRowIndex % (2 * dimensions[1]);
	let yDirection;
	let yEndFlag;
	if (yCoordBeforeWrap < dimensions[1]) {
		textureCoordinate[1] = yCoordBeforeWrap;
		yDirection = 1;
		yEndFlag = textureCoordinate[1] === dimensions[1] - 1;
	} else {
		textureCoordinate[1] = 2 * dimensions[1] - 1 - yCoordBeforeWrap;
		yDirection = -1;
		yEndFlag = textureCoordinate[1] === 0;
	}
	textureCoordinate[2] = Math.floor(intRowIndex / dimensions[1]);
	const remainder = texelIndexPosition - intTexelIndex;
	if (xEndFlag) if (yEndFlag) textureCoordinate[2] += remainder;
	else textureCoordinate[1] += yDirection * remainder;
	else textureCoordinate[0] += xDirection * remainder;
	textureCoordinate[0] = (textureCoordinate[0] + .5) / dimensions[0];
	textureCoordinate[1] = (textureCoordinate[1] + .5) / dimensions[1];
	textureCoordinate[2] = (textureCoordinate[2] + .5) / dimensions[2];
}
var colorTextureCoordinatesCache = /* @__PURE__ */ new WeakMap();
/**
* The minimum of the range is mapped to the center of the first texel excluding min texel (texel at index distance 1)
* The maximum of the range is mapped to the center of the last texel excluding max and NaN texels (texel at index distance numberOfColorsInRange)
* The result is cached, and is reused if the arguments are the same and the input doesn't change
* @param {vtkDataArray} input The input data array used for coloring
* @param {Number} component The component of the input data array that is used for coloring (-1 for magnitude of the vectors)
* @param {Range} range The range of the scalars
* @param {boolean} useLogScale Should the values be transformed to logarithmic scale. When true, the range must already be in logarithmic scale.
* @param {Number} numberOfColorsInRange The number of colors that are used in the range
* @param {vec3} dimensions The dimensions of the texture
* @param {boolean} useZigzagPattern If a zigzag pattern should be used. Otherwise 1 row for colors (including min and max) and 1 row for NaN are used.
* @returns A vtkDataArray containing the texture coordinates (2D or 3D)
*/
function getOrCreateColorTextureCoordinates(input, component, range, useLogScale, numberOfColorsInRange, dimensions, useZigzagPattern) {
	const argStrings = new Array(arguments.length);
	for (let argIndex = 0; argIndex < arguments.length; ++argIndex) {
		const arg = arguments[argIndex];
		argStrings[argIndex] = arg.getMTime?.() ?? arg;
	}
	const stringHash = argStrings.join("/");
	const cachedResult = colorTextureCoordinatesCache.get(input);
	if (cachedResult && cachedResult.stringHash === stringHash) return cachedResult.textureCoordinates;
	const scalarTexelWidth = (range[1] - range[0]) / (numberOfColorsInRange - 1);
	const [paddedRangeMin, paddedRangeMax] = [range[0] - scalarTexelWidth, range[1] + scalarTexelWidth];
	const textureSOrigin = paddedRangeMin - .5 * scalarTexelWidth;
	const textureSCoeff = 1 / (paddedRangeMax - paddedRangeMin + scalarTexelWidth);
	const texelIndexOrigin = paddedRangeMin;
	const texelIndexCoeff = (numberOfColorsInRange + 1) / (paddedRangeMax - paddedRangeMin);
	const inputV = input.getData();
	const numScalars = input.getNumberOfTuples();
	const numComps = input.getNumberOfComponents();
	const useMagnitude = component < 0 || component >= numComps;
	const numberOfOutputComponents = dimensions[2] <= 1 ? 2 : 3;
	const output = DataArray_default.newInstance({
		numberOfComponents: numberOfOutputComponents,
		values: new Float32Array(numScalars * numberOfOutputComponents)
	});
	const outputV = output.getData();
	const nanTextureCoordinate = [
		0,
		0,
		0
	];
	getZigZagTextureCoordinatesFromTexelPosition(nanTextureCoordinate, numberOfColorsInRange + 2, dimensions);
	let inputIdx = 0;
	let outputIdx = 0;
	const textureCoordinate = [
		.5,
		.5,
		.5
	];
	for (let scalarIdx = 0; scalarIdx < numScalars; ++scalarIdx) {
		let scalarValue;
		if (useMagnitude) {
			let sum = 0;
			for (let compIdx = 0; compIdx < numComps; ++compIdx) {
				const compValue = Number(inputV[inputIdx + compIdx]);
				sum += compValue * compValue;
			}
			scalarValue = Math.sqrt(sum);
		} else scalarValue = Number(inputV[inputIdx + component]);
		if (useLogScale) scalarValue = Math.log10(scalarValue);
		inputIdx += numComps;
		if (isNan(scalarValue)) {
			textureCoordinate[0] = nanTextureCoordinate[0];
			textureCoordinate[1] = nanTextureCoordinate[1];
			textureCoordinate[2] = nanTextureCoordinate[2];
		} else if (useZigzagPattern) {
			let texelIndexPosition = (scalarValue - texelIndexOrigin) * texelIndexCoeff;
			if (texelIndexPosition < 1) texelIndexPosition = 0;
			else if (texelIndexPosition > numberOfColorsInRange) texelIndexPosition = numberOfColorsInRange + 1;
			getZigZagTextureCoordinatesFromTexelPosition(textureCoordinate, texelIndexPosition, dimensions);
		} else {
			textureCoordinate[1] = .49;
			const textureS = (scalarValue - textureSOrigin) * textureSCoeff;
			if (textureS > 1e3) textureCoordinate[0] = 1e3;
			else if (textureS < -1e3) textureCoordinate[0] = -1e3;
			else textureCoordinate[0] = textureS;
		}
		for (let i = 0; i < numberOfOutputComponents; ++i) outputV[outputIdx++] = textureCoordinate[i];
	}
	colorTextureCoordinatesCache.set(input, {
		stringHash,
		textureCoordinates: output
	});
	return output;
}
function vtkMapper2D(publicAPI, model) {
	model.classHierarchy.push("vtkMapper2D");
	publicAPI.createDefaultLookupTable = () => {
		model.lookupTable = LookupTable_default.newInstance();
	};
	publicAPI.getColorModeAsString = () => macros_default.enumToString(ColorMode, model.colorMode);
	publicAPI.setColorModeToDefault = () => publicAPI.setColorMode(0);
	publicAPI.setColorModeToMapScalars = () => publicAPI.setColorMode(1);
	publicAPI.setColorModeToDirectScalars = () => publicAPI.setColorMode(2);
	publicAPI.getScalarModeAsString = () => macros_default.enumToString(ScalarMode, model.scalarMode);
	publicAPI.setScalarModeToDefault = () => publicAPI.setScalarMode(0);
	publicAPI.setScalarModeToUsePointData = () => publicAPI.setScalarMode(1);
	publicAPI.setScalarModeToUseCellData = () => publicAPI.setScalarMode(2);
	publicAPI.setScalarModeToUsePointFieldData = () => publicAPI.setScalarMode(3);
	publicAPI.setScalarModeToUseCellFieldData = () => publicAPI.setScalarMode(4);
	publicAPI.setScalarModeToUseFieldData = () => publicAPI.setScalarMode(5);
	publicAPI.getAbstractScalars = (input, scalarMode, arrayAccessMode, arrayId, arrayName) => {
		if (!input || !model.scalarVisibility) return {
			scalars: null,
			cellFlag: false
		};
		let scalars = null;
		let cellFlag = false;
		if (scalarMode === ScalarMode.DEFAULT) {
			scalars = input.getPointData().getScalars();
			if (!scalars) {
				scalars = input.getCellData().getScalars();
				cellFlag = true;
			}
		} else if (scalarMode === ScalarMode.USE_POINT_DATA) scalars = input.getPointData().getScalars();
		else if (scalarMode === ScalarMode.USE_CELL_DATA) {
			scalars = input.getCellData().getScalars();
			cellFlag = true;
		} else if (scalarMode === ScalarMode.USE_POINT_FIELD_DATA) {
			const pd = input.getPointData();
			if (arrayAccessMode === GetArray.BY_ID) scalars = pd.getArrayByIndex(arrayId);
			else scalars = pd.getArrayByName(arrayName);
		} else if (scalarMode === ScalarMode.USE_CELL_FIELD_DATA) {
			const cd = input.getCellData();
			cellFlag = true;
			if (arrayAccessMode === GetArray.BY_ID) scalars = cd.getArrayByIndex(arrayId);
			else scalars = cd.getArrayByName(arrayName);
		} else if (scalarMode === ScalarMode.USE_FIELD_DATA) {
			const fd = input.getFieldData();
			if (arrayAccessMode === GetArray.BY_ID) scalars = fd.getArrayByIndex(arrayId);
			else scalars = fd.getArrayByName(arrayName);
		}
		return {
			scalars,
			cellFlag
		};
	};
	publicAPI.getLookupTable = () => {
		if (!model.lookupTable) publicAPI.createDefaultLookupTable();
		return model.lookupTable;
	};
	publicAPI.getMTime = () => {
		let mt = model.mtime;
		if (model.lookupTable !== null) {
			const time = model.lookupTable.getMTime();
			mt = time > mt ? time : mt;
		}
		return mt;
	};
	publicAPI.mapScalars = (input, alpha) => {
		const { scalars, cellFlag } = publicAPI.getAbstractScalars(input, model.scalarMode, model.arrayAccessMode, model.arrayId, model.colorByArrayName);
		model.areScalarsMappedFromCells = cellFlag;
		if (!scalars) {
			model.colorCoordinates = null;
			model.colorTextureMap = null;
			model.colorMapColors = null;
			return;
		}
		const toString = `${publicAPI.getMTime()}${scalars.getMTime()}${alpha}`;
		if (model.colorBuildString === toString) return;
		if (!model.useLookupTableScalarRange) publicAPI.getLookupTable().setRange(model.scalarRange[0], model.scalarRange[1]);
		if (publicAPI.canUseTextureMapForColoring(scalars, cellFlag)) model.mapScalarsToTexture(scalars, cellFlag, alpha);
		else {
			model.colorCoordinates = null;
			model.colorTextureMap = null;
			const lut = publicAPI.getLookupTable();
			if (lut) {
				lut.build();
				model.colorMapColors = lut.mapScalars(scalars, model.colorMode, model.fieldDataTupleId);
			}
		}
		model.colorBuildString = `${publicAPI.getMTime()}${scalars.getMTime()}${alpha}`;
	};
	publicAPI.canUseTextureMapForColoring = (scalars, cellFlag) => {
		if (cellFlag && !(model.colorMode === ColorMode.DIRECT_SCALARS)) return true;
		if (model.lookupTable && model.lookupTable.getIndexedLookup()) return false;
		if (!scalars) return false;
		if (model.colorMode === ColorMode.DEFAULT && scalars.getDataType() === VtkDataTypes.UNSIGNED_CHAR || model.colorMode === ColorMode.DIRECT_SCALARS) return false;
		return true;
	};
	model.mapScalarsToTexture = (scalars, cellFlag, alpha) => {
		const range = model.lookupTable.getRange();
		const useLogScale = model.lookupTable.usingLogScale();
		const origAlpha = model.lookupTable.getAlpha();
		const scaledRange = useLogScale ? [Math.log10(range[0]), Math.log10(range[1])] : range;
		model.colorMapColors = null;
		if (model.colorTextureMap == null || publicAPI.getMTime() > model.colorTextureMap.getMTime() || model.lookupTable.getMTime() > model.colorTextureMap.getMTime() || model.lookupTable.getAlpha() !== alpha) {
			model.lookupTable.setAlpha(alpha);
			model.colorTextureMap = null;
			model.lookupTable.build();
			const numberOfAvailableColors = model.lookupTable.getNumberOfAvailableColors();
			const maxTextureWidthForCells = 2048;
			const maxColorsInRangeForCells = maxTextureWidthForCells ** 3 - 3;
			model.numberOfColorsInRange = Math.min(Math.max(numberOfAvailableColors, 2), cellFlag ? maxColorsInRangeForCells : 4094);
			const numberOfColorsForCells = model.numberOfColorsInRange + 3;
			const numberOfColorsInUpperRowForPoints = model.numberOfColorsInRange + 2;
			const textureDimensions = cellFlag ? [
				Math.min(Math.ceil(numberOfColorsForCells / maxTextureWidthForCells ** 0), maxTextureWidthForCells),
				Math.min(Math.ceil(numberOfColorsForCells / maxTextureWidthForCells ** 1), maxTextureWidthForCells),
				Math.min(Math.ceil(numberOfColorsForCells / maxTextureWidthForCells ** 2), maxTextureWidthForCells)
			] : [
				numberOfColorsInUpperRowForPoints,
				2,
				1
			];
			const textureSize = textureDimensions[0] * textureDimensions[1] * textureDimensions[2];
			const scalarsArray = new Float64Array(textureSize);
			scalarsArray.fill(NaN);
			const numberOfNonSpecialColors = model.numberOfColorsInRange;
			const numberOfNonNaNColors = numberOfNonSpecialColors + 2;
			const textureCoordinates = [
				0,
				0,
				0
			];
			const rangeMin = scaledRange[0];
			const rangeDifference = scaledRange[1] - scaledRange[0];
			for (let i = 0; i < numberOfNonNaNColors; ++i) {
				const scalarsArrayIndex = getIndexFromCoordinates(textureCoordinates, textureDimensions);
				const intermediateValue = rangeMin + rangeDifference * (i - 1) / (numberOfNonSpecialColors - 1);
				scalarsArray[scalarsArrayIndex] = useLogScale ? 10 ** intermediateValue : intermediateValue;
				updateZigzaggingCoordinates(textureCoordinates, textureDimensions);
			}
			const scalarsDataArray = DataArray_default.newInstance({
				numberOfComponents: 1,
				values: scalarsArray
			});
			const colorsDataArray = model.lookupTable.mapScalars(scalarsDataArray, model.colorMode, 0);
			model.colorTextureMap = ImageData_default.newInstance();
			model.colorTextureMap.setDimensions(textureDimensions);
			model.colorTextureMap.getPointData().setScalars(colorsDataArray);
			model.lookupTable.setAlpha(origAlpha);
		}
		model.colorCoordinates = getOrCreateColorTextureCoordinates(scalars, model.lookupTable.getVectorMode() === VectorMode.MAGNITUDE && scalars.getNumberOfComponents() > 1 ? -1 : model.lookupTable.getVectorComponent(), scaledRange, useLogScale, model.numberOfColorsInRange, model.colorTextureMap.getDimensions(), cellFlag);
	};
	publicAPI.getPrimitiveCount = () => {
		const input = publicAPI.getInputData();
		return {
			points: input.getPoints().getNumberOfValues() / 3,
			verts: input.getVerts().getNumberOfValues() - input.getVerts().getNumberOfCells(),
			lines: input.getLines().getNumberOfValues() - 2 * input.getLines().getNumberOfCells(),
			triangles: input.getPolys().getNumberOfValues() - 3 * input.getPolys().getNumberOfCells()
		};
	};
}
var DEFAULT_VALUES = {
	static: false,
	lookupTable: null,
	scalarVisibility: false,
	scalarRange: [0, 1],
	useLookupTableScalarRange: false,
	colorMode: 0,
	scalarMode: 0,
	arrayAccessMode: 1,
	colorMapColors: null,
	areScalarsMappedFromCells: false,
	renderTime: 0,
	colorByArrayName: null,
	transformCoordinate: null,
	viewSpecificProperties: null,
	customShaderAttributes: []
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	AbstractMapper_default.extend(publicAPI, model, initialValues);
	macros_default.get(publicAPI, model, [
		"areScalarsMappedFromCells",
		"colorCoordinates",
		"colorTextureMap",
		"colorMapColors"
	]);
	macros_default.setGet(publicAPI, model, [
		"arrayAccessMode",
		"colorByArrayName",
		"colorMode",
		"lookupTable",
		"renderTime",
		"scalarMode",
		"scalarVisibility",
		"static",
		"transformCoordinate",
		"useLookupTableScalarRange",
		"viewSpecificProperties",
		"customShaderAttributes"
	]);
	macros_default.setGetArray(publicAPI, model, ["scalarRange"], 2);
	if (!model.viewSpecificProperties) model.viewSpecificProperties = {};
	vtkMapper2D(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkMapper2D");
var Mapper2D_default = {
	newInstance,
	extend
};
//#endregion
export { Mapper2D_default as default, extend, newInstance };
