// @kitware/vtk.js/Rendering/Core/PointPicker.js
import macros_default from "../../macros.js";
import { dot } from "../../Common/Core/Math.js";
import Picker_default from "./Picker.js";
//#region Sources/Rendering/Core/PointPicker/index.js
var { vtkErrorMacro } = macros_default;
function vtkPointPicker(publicAPI, model) {
	model.classHierarchy.push("vtkPointPicker");
	model.intersectWithLine = (p1, p2, tolerance, prop, mapper) => {
		let tMin = Number.MAX_VALUE;
		if (mapper.isA("vtkImageMapper") || mapper.isA("vtkImageArrayMapper")) {
			const pickData = mapper.intersectWithLineForPointPicking(p1, p2);
			if (pickData) {
				tMin = pickData.t;
				model.pointIJK = pickData.ijk;
			}
		} else if (mapper.isA("vtkMapper")) tMin = model.intersectActorWithLine(p1, p2, tolerance, mapper);
		return tMin;
	};
	model.intersectActorWithLine = (p1, p2, tolerance, mapper) => {
		const input = mapper.getInputData();
		let ptId = 0;
		const numPts = input.getPoints().getNumberOfPoints();
		if (numPts <= ptId) return 2;
		const ray = [];
		for (let i = 0; i < 3; i++) ray[i] = p2[i] - p1[i];
		const rayFactor = dot(ray, ray);
		if (rayFactor === 0) {
			vtkErrorMacro("Cannot process points");
			return 2;
		}
		let t;
		let minPtId = -1;
		let tMin = Number.MAX_VALUE;
		let minPtDist = Number.MAX_VALUE;
		const projXYZ = [];
		const minXYZ = [];
		const x = [];
		const points = input.getPoints();
		if (model.useCells) {
			const cellData = input.getPolys().getData();
			const nbPointsPerCell = cellData[0];
			const nbCells = input.getPolys().getNumberOfCells();
			for (let cellID = 0; cellID < nbCells; cellID++) {
				const firstPointIndex = cellID * nbPointsPerCell + 1;
				const lastPointIndex = firstPointIndex + nbPointsPerCell;
				for (let pointIndex = firstPointIndex; pointIndex < lastPointIndex; pointIndex++) {
					const pointDataIndex = cellData[pointIndex];
					points.getPoint(pointDataIndex, x);
					t = (ray[0] * (x[0] - p1[0]) + ray[1] * (x[1] - p1[1]) + ray[2] * (x[2] - p1[2])) / rayFactor;
					if (t >= 0 && t <= 1 && t <= tMin + model.tolerance) {
						let maxDist = 0;
						for (let i = 0; i < 3; i++) {
							projXYZ[i] = p1[i] + t * ray[i];
							const dist = Math.abs(x[i] - projXYZ[i]);
							if (dist > maxDist) maxDist = dist;
						}
						if (maxDist <= tolerance && maxDist < minPtDist) {
							minPtId = ptId;
							minXYZ[0] = x[0];
							minXYZ[1] = x[1];
							minXYZ[2] = x[2];
							minPtDist = maxDist;
							tMin = t;
						}
					}
				}
			}
		} else for (ptId = 0; ptId < numPts; ptId++) {
			points.getPoint(ptId, x);
			t = (ray[0] * (x[0] - p1[0]) + ray[1] * (x[1] - p1[1]) + ray[2] * (x[2] - p1[2])) / rayFactor;
			if (t >= 0 && t <= 1 && t <= tMin + model.tolerance) {
				let maxDist = 0;
				for (let i = 0; i < 3; i++) {
					projXYZ[i] = p1[i] + t * ray[i];
					const dist = Math.abs(x[i] - projXYZ[i]);
					if (dist > maxDist) maxDist = dist;
				}
				if (maxDist <= tolerance && maxDist < minPtDist) {
					minPtId = ptId;
					minXYZ[0] = x[0];
					minXYZ[1] = x[1];
					minXYZ[2] = x[2];
					minPtDist = maxDist;
					tMin = t;
				}
			}
		}
		if (minPtId > -1 && tMin < model.globalTMin) {
			model.globalTMin = tMin;
			model.pointId = minPtId;
		}
		return tMin;
	};
}
var DEFAULT_VALUES = {
	pointId: -1,
	pointIJK: [],
	useCells: false
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	Picker_default.extend(publicAPI, model, initialValues);
	macros_default.getArray(publicAPI, model, ["pointIJK"]);
	macros_default.get(publicAPI, model, ["pointId"]);
	macros_default.setGet(publicAPI, model, ["useCells"]);
	vtkPointPicker(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkPointPicker");
var PointPicker_default = {
	newInstance,
	extend
};
//#endregion
export { PointPicker_default as default, extend, newInstance };
