// @kitware/vtk.js/Rendering/Core/Prop.js
import macros_default from "../../macros.js";
import Constants_default from "./Prop/Constants.js";
//#region Sources/Rendering/Core/Prop/index.js
var { CoordinateSystem } = Constants_default;
function notImplemented(method) {
	return () => macros_default.vtkErrorMacro(`vtkProp::${method} - NOT IMPLEMENTED`);
}
function vtkProp(publicAPI, model) {
	model.classHierarchy.push("vtkProp");
	publicAPI.getMTime = () => {
		let m1 = model.mtime;
		for (let index = 0; index < model.textures.length; ++index) {
			const m2 = model.textures[index].getMTime();
			if (m2 > m1) m1 = m2;
		}
		return m1;
	};
	publicAPI.processSelectorPixelBuffers = (selector, pixeloffsets) => {};
	publicAPI.getNestedProps = () => null;
	publicAPI.getActors = () => [];
	publicAPI.getActors2D = () => [];
	publicAPI.getVolumes = () => [];
	publicAPI.pick = notImplemented("pick");
	publicAPI.hasKey = notImplemented("hasKey");
	publicAPI.getNestedVisibility = () => model.visibility && (!model._parentProp || model._parentProp.getNestedVisibility());
	publicAPI.getNestedPickable = () => model.pickable && (!model._parentProp || model._parentProp.getNestedPickable());
	publicAPI.getNestedDragable = () => model.dragable && (!model._parentProp || model._parentProp.getNestedDragable());
	publicAPI.getRedrawMTime = () => model.mtime;
	publicAPI.setEstimatedRenderTime = (t) => {
		model.estimatedRenderTime = t;
		model.savedEstimatedRenderTime = t;
	};
	publicAPI.restoreEstimatedRenderTime = () => {
		model.estimatedRenderTime = model.savedEstimatedRenderTime;
	};
	publicAPI.addEstimatedRenderTime = (t) => {
		model.estimatedRenderTime += t;
	};
	publicAPI.setAllocatedRenderTime = (t) => {
		model.allocatedRenderTime = t;
		model.savedEstimatedRenderTime = model.estimatedRenderTime;
		model.estimatedRenderTime = 0;
	};
	publicAPI.getSupportsSelection = () => false;
	publicAPI.getTextures = () => model.textures;
	publicAPI.hasTexture = (texture) => model.textures.indexOf(texture) !== -1;
	publicAPI.addTexture = (texture) => {
		if (texture && !publicAPI.hasTexture(texture)) {
			model.textures = model.textures.concat(texture);
			publicAPI.modified();
		}
	};
	publicAPI.removeTexture = (texture) => {
		const newTextureList = model.textures.filter((item) => item !== texture);
		if (model.textures.length !== newTextureList.length) {
			model.textures = newTextureList;
			publicAPI.modified();
		}
	};
	publicAPI.removeAllTextures = () => {
		model.textures = [];
		publicAPI.modified();
	};
	publicAPI.setCoordinateSystemToWorld = () => publicAPI.setCoordinateSystem(CoordinateSystem.WORLD);
	publicAPI.setCoordinateSystemToDisplay = () => publicAPI.setCoordinateSystem(CoordinateSystem.DISPLAY);
}
var DEFAULT_VALUES = {
	allocatedRenderTime: 10,
	coordinateSystem: CoordinateSystem.WORLD,
	dragable: true,
	estimatedRenderTime: 0,
	paths: null,
	pickable: true,
	renderTimeMultiplier: 1,
	savedEstimatedRenderTime: 0,
	textures: [],
	useBounds: true,
	visibility: true
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.get(publicAPI, model, ["estimatedRenderTime", "allocatedRenderTime"]);
	macros_default.setGet(publicAPI, model, [
		"_parentProp",
		"coordinateSystem",
		"dragable",
		"pickable",
		"renderTimeMultiplier",
		"useBounds",
		"visibility"
	]);
	macros_default.moveToProtected(publicAPI, model, ["parentProp"]);
	vtkProp(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkProp");
var Prop_default = {
	newInstance,
	extend,
	...Constants_default
};
//#endregion
export { Prop_default as default, extend, newInstance };
