// @kitware/vtk.js/Rendering/Core/Prop3D.js
import macros_default from "../../macros.js";
import { areMatricesEqual, degreesFromRadians, radiansFromDegrees } from "../../Common/Core/Math.js";
import BoundingBox_default from "../../Common/DataModel/BoundingBox.js";
import Prop_default from "./Prop.js";
import { mat4, quat } from "gl-matrix";
//#region Sources/Rendering/Core/Prop3D/index.js
var VTK_EPSILON = 1e-6;
function vtkProp3D(publicAPI, model) {
	model.classHierarchy.push("vtkProp3D");
	publicAPI.addPosition = (deltaXYZ) => {
		model.position = model.position.map((value, index) => value + deltaXYZ[index]);
		publicAPI.modified();
	};
	publicAPI.getOrientationWXYZ = () => {
		const q = quat.create();
		mat4.getRotation(q, model.rotation);
		const oaxis = new Float64Array(3);
		return [
			degreesFromRadians(quat.getAxisAngle(oaxis, q)),
			oaxis[0],
			oaxis[1],
			oaxis[2]
		];
	};
	publicAPI.getOrientationQuaternion = (out = []) => mat4.getRotation(out, model.rotation);
	publicAPI.rotateX = (val) => {
		if (val === 0) return;
		mat4.rotateX(model.rotation, model.rotation, radiansFromDegrees(val));
		publicAPI.modified();
	};
	publicAPI.rotateY = (val) => {
		if (val === 0) return;
		mat4.rotateY(model.rotation, model.rotation, radiansFromDegrees(val));
		publicAPI.modified();
	};
	publicAPI.rotateZ = (val) => {
		if (val === 0) return;
		mat4.rotateZ(model.rotation, model.rotation, radiansFromDegrees(val));
		publicAPI.modified();
	};
	publicAPI.rotateWXYZ = (degrees, x, y, z) => {
		if (degrees === 0 || x === 0 && y === 0 && z === 0) return;
		const angle = radiansFromDegrees(degrees);
		const q = quat.create();
		quat.setAxisAngle(q, [
			x,
			y,
			z
		], angle);
		const quatMat = new Float64Array(16);
		mat4.fromQuat(quatMat, q);
		mat4.multiply(model.rotation, model.rotation, quatMat);
		publicAPI.modified();
	};
	publicAPI.rotateQuaternion = (orientationQuaternion) => {
		if (Math.abs(orientationQuaternion[3]) >= 1 - VTK_EPSILON) return;
		const oriQuatMat = mat4.fromQuat(new Float64Array(16), orientationQuaternion);
		mat4.multiply(model.rotation, model.rotation, oriQuatMat);
		publicAPI.modified();
	};
	publicAPI.setOrientation = (x, y, z) => {
		if (x === model.orientation[0] && y === model.orientation[1] && z === model.orientation[2]) return false;
		model.orientation = [
			x,
			y,
			z
		];
		mat4.identity(model.rotation);
		publicAPI.rotateZ(z);
		publicAPI.rotateX(x);
		publicAPI.rotateY(y);
		publicAPI.modified();
		return true;
	};
	publicAPI.setOrientationFromQuaternion = (q) => {
		const rotation = mat4.create();
		mat4.fromQuat(rotation, q);
		if (!areMatricesEqual(rotation, model.rotation)) {
			model.rotation = rotation;
			publicAPI.modified();
			return true;
		}
		return false;
	};
	publicAPI.setUserMatrix = (matrix) => {
		if (areMatricesEqual(model.userMatrix, matrix)) return false;
		mat4.copy(model.userMatrix, matrix);
		publicAPI.modified();
		return true;
	};
	publicAPI.getMatrix = () => {
		publicAPI.computeMatrix();
		return model.matrix;
	};
	publicAPI.computeMatrix = () => {
		if (publicAPI.getMTime() > model.matrixMTime.getMTime()) {
			mat4.identity(model.matrix);
			if (model.userMatrix) mat4.multiply(model.matrix, model.matrix, model.userMatrix);
			mat4.translate(model.matrix, model.matrix, model.origin);
			mat4.translate(model.matrix, model.matrix, model.position);
			mat4.multiply(model.matrix, model.matrix, model.rotation);
			mat4.scale(model.matrix, model.matrix, model.scale);
			mat4.translate(model.matrix, model.matrix, [
				-model.origin[0],
				-model.origin[1],
				-model.origin[2]
			]);
			mat4.transpose(model.matrix, model.matrix);
			model.isIdentity = true;
			for (let i = 0; i < 4; ++i) for (let j = 0; j < 4; ++j) if ((i === j ? 1 : 0) !== model.matrix[i + j * 4]) model.isIdentity = false;
			model.matrixMTime.modified();
		}
	};
	publicAPI.computeBounds = () => {
		if (model.mapper === null) {
			BoundingBox_default.reset(model.bounds);
			return;
		}
		const bds = model.mapper.getBounds();
		if (!bds || bds.length !== 6 || !BoundingBox_default.isValid(bds)) {
			model.mapperBounds = bds;
			BoundingBox_default.reset(model.bounds);
			model.boundsMTime.modified();
			return;
		}
		if (!model.mapperBounds || !bds.every((_, i) => bds[i] === model.mapperBounds[i]) || publicAPI.getMTime() > model.boundsMTime.getMTime()) {
			macros_default.vtkDebugMacro("Recomputing bounds...");
			model.mapperBounds = bds;
			publicAPI.computeMatrix();
			const transposedMatrix = new Float64Array(16);
			mat4.transpose(transposedMatrix, model.matrix);
			BoundingBox_default.transformBounds(bds, transposedMatrix, model.bounds);
			model.boundsMTime.modified();
		}
	};
	const superGetBounds = publicAPI.getBounds;
	publicAPI.getBounds = () => {
		publicAPI.computeBounds();
		return superGetBounds();
	};
	const superGetBoundsByReference = publicAPI.getBoundsByReference;
	publicAPI.getBoundsByReference = () => {
		publicAPI.computeBounds();
		return superGetBoundsByReference();
	};
	publicAPI.getCenter = () => BoundingBox_default.getCenter(publicAPI.getBoundsByReference());
	publicAPI.getLength = () => BoundingBox_default.getLength(publicAPI.getBoundsByReference());
	publicAPI.getXRange = () => BoundingBox_default.getXRange(publicAPI.getBoundsByReference());
	publicAPI.getYRange = () => BoundingBox_default.getYRange(publicAPI.getBoundsByReference());
	publicAPI.getZRange = () => BoundingBox_default.getZRange(publicAPI.getBoundsByReference());
	publicAPI.getUserMatrix = () => model.userMatrix;
	function updateIdentityFlag() {
		publicAPI.computeMatrix();
	}
	publicAPI.onModified(updateIdentityFlag);
	publicAPI.getProperty = (mapperInputPort = 0) => {
		if (model.properties[mapperInputPort] == null) model.properties[mapperInputPort] = publicAPI.makeProperty?.();
		return model.properties[mapperInputPort];
	};
	publicAPI.getProperties = () => {
		if (model.properties.length === 0) model.properties[0] = publicAPI.makeProperty?.();
		return model.properties;
	};
	publicAPI.setProperty = (firstArg, secondArg) => {
		const [mapperInputPort, property] = Number.isInteger(firstArg) ? [firstArg, secondArg] : [0, firstArg];
		if (model.properties[mapperInputPort] === property) return false;
		model.properties[mapperInputPort] = property;
		return true;
	};
	publicAPI.getMTime = () => {
		let mt = model.mtime;
		model.properties.forEach((property) => {
			if (property !== null) {
				const time = property.getMTime();
				mt = time > mt ? time : mt;
			}
		});
		return mt;
	};
}
var DEFAULT_VALUES = {
	bounds: [...BoundingBox_default.INIT_BOUNDS],
	origin: [
		0,
		0,
		0
	],
	position: [
		0,
		0,
		0
	],
	orientation: [
		0,
		0,
		0
	],
	rotation: null,
	scale: [
		1,
		1,
		1
	],
	properties: [],
	userMatrix: null,
	userMatrixMTime: null,
	cachedProp3D: null,
	isIdentity: true,
	matrixMTime: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	Prop_default.extend(publicAPI, model, initialValues);
	model.matrixMTime = {};
	macros_default.obj(model.matrixMTime);
	macros_default.get(publicAPI, model, ["isIdentity"]);
	macros_default.getArray(publicAPI, model, ["orientation"]);
	macros_default.setGetArray(publicAPI, model, [
		"origin",
		"position",
		"scale"
	], 3);
	macros_default.setGet(publicAPI, model, ["properties"]);
	macros_default.getArray(publicAPI, model, ["bounds"], 6);
	model.matrix = mat4.identity(new Float64Array(16));
	model.rotation = mat4.identity(new Float64Array(16));
	model.userMatrix = mat4.identity(new Float64Array(16));
	model.transform = null;
	vtkProp3D(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkProp3D");
var Prop3D_default = {
	newInstance,
	extend
};
//#endregion
export { Prop3D_default as default, extend, newInstance };
