// @kitware/vtk.js/Rendering/Core/Property2D.js
import macros_default from "../../macros.js";
import { Representation } from "./Property/Constants.js";
import Constants_default from "./Property2D/Constants.js";
//#region Sources/Rendering/Core/Property2D/index.js
var { DisplayLocation } = Constants_default;
function vtkProperty2D(publicAPI, model) {
	model.classHierarchy.push("vtkProperty2D");
	publicAPI.setDisplayLocationToBackground = () => publicAPI.setDisplayLocation(DisplayLocation.BACKGROUND);
	publicAPI.setDisplayLocationToForeground = () => publicAPI.setDisplayLocation(DisplayLocation.FOREGROUND);
	publicAPI.setRepresentationToWireframe = () => publicAPI.setRepresentation(Representation.WIREFRAME);
	publicAPI.setRepresentationToSurface = () => publicAPI.setRepresentation(Representation.SURFACE);
	publicAPI.setRepresentationToPoints = () => publicAPI.setRepresentation(Representation.POINTS);
	publicAPI.getRepresentationAsString = () => macros_default.enumToString(Representation, model.representation);
}
var DEFAULT_VALUES = {
	color: [
		1,
		1,
		1
	],
	opacity: 1,
	pointSize: 1,
	lineWidth: 1,
	representation: Representation.SURFACE,
	displayLocation: DisplayLocation.FOREGROUND
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, [
		"opacity",
		"lineWidth",
		"pointSize",
		"displayLocation",
		"representation"
	]);
	macros_default.setGetArray(publicAPI, model, ["color"], 3);
	vtkProperty2D(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkProperty2D");
var Property2D_default = {
	newInstance,
	extend,
	...Constants_default
};
//#endregion
export { Property2D_default as default, extend, newInstance };
