// @kitware/vtk.js/Rendering/Core/Property.js
import macros_default from "../../macros.js";
import Constants_default from "./Property/Constants.js";
//#region Sources/Rendering/Core/Property/index.js
var { Representation, Interpolation } = Constants_default;
function notImplemented(method) {
	return () => macros_default.vtkErrorMacro(`vtkProperty::${method} - NOT IMPLEMENTED`);
}
function vtkProperty(publicAPI, model) {
	model.classHierarchy.push("vtkProperty");
	publicAPI.setColor = (r, g, b) => {
		if (Array.isArray(r)) {
			if (model.color[0] !== r[0] || model.color[1] !== r[1] || model.color[2] !== r[2]) {
				model.color[0] = r[0];
				model.color[1] = r[1];
				model.color[2] = r[2];
				publicAPI.modified();
			}
		} else if (model.color[0] !== r || model.color[1] !== g || model.color[2] !== b) {
			model.color[0] = r;
			model.color[1] = g;
			model.color[2] = b;
			publicAPI.modified();
		}
		publicAPI.setDiffuseColor(model.color);
		publicAPI.setAmbientColor(model.color);
		publicAPI.setSpecularColor(model.color);
	};
	publicAPI.computeCompositeColor = notImplemented("ComputeCompositeColor");
	publicAPI.getColor = () => {
		let norm = 0;
		if (model.ambient + model.diffuse + model.specular > 0) norm = 1 / (model.ambient + model.diffuse + model.specular);
		for (let i = 0; i < 3; i++) model.color[i] = norm * (model.ambient * model.ambientColor[i] + model.diffuse * model.diffuseColor[i] + model.specular * model.specularColor[i]);
		return [].concat(model.color);
	};
	publicAPI.addShaderVariable = notImplemented("AddShaderVariable");
	publicAPI.setInterpolationToFlat = () => publicAPI.setInterpolation(Interpolation.FLAT);
	publicAPI.setInterpolationToGouraud = () => publicAPI.setInterpolation(Interpolation.GOURAUD);
	publicAPI.setInterpolationToPhong = () => publicAPI.setInterpolation(Interpolation.PHONG);
	publicAPI.getInterpolationAsString = () => macros_default.enumToString(Interpolation, model.interpolation);
	publicAPI.setRepresentationToWireframe = () => publicAPI.setRepresentation(Representation.WIREFRAME);
	publicAPI.setRepresentationToSurface = () => publicAPI.setRepresentation(Representation.SURFACE);
	publicAPI.setRepresentationToPoints = () => publicAPI.setRepresentation(Representation.POINTS);
	publicAPI.getRepresentationAsString = () => macros_default.enumToString(Representation, model.representation);
}
var DEFAULT_VALUES = {
	color: [
		1,
		1,
		1
	],
	ambientColor: [
		1,
		1,
		1
	],
	diffuseColor: [
		1,
		1,
		1
	],
	specularColor: [
		1,
		1,
		1
	],
	edgeColor: [
		0,
		0,
		0
	],
	ambient: 0,
	diffuse: 1,
	metallic: 0,
	roughness: .6,
	normalStrength: 1,
	emission: 1,
	baseIOR: 1.45,
	specular: 0,
	specularPower: 1,
	opacity: 1,
	edgeOpacity: 1,
	interpolation: Interpolation.GOURAUD,
	representation: Representation.SURFACE,
	edgeVisibility: false,
	backfaceCulling: false,
	frontfaceCulling: false,
	pointSize: 1,
	lineWidth: 1,
	lighting: true,
	shading: false,
	materialName: null,
	ORMTexture: null,
	RMTexture: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, [
		"lighting",
		"interpolation",
		"ambient",
		"diffuse",
		"metallic",
		"roughness",
		"normalStrength",
		"emission",
		"baseIOR",
		"specular",
		"specularPower",
		"opacity",
		"edgeOpacity",
		"edgeVisibility",
		"lineWidth",
		"pointSize",
		"backfaceCulling",
		"frontfaceCulling",
		"representation",
		"diffuseTexture",
		"metallicTexture",
		"roughnessTexture",
		"normalTexture",
		"ambientOcclusionTexture",
		"emissionTexture",
		"ORMTexture",
		"RMTexture"
	]);
	macros_default.setGetArray(publicAPI, model, [
		"ambientColor",
		"specularColor",
		"diffuseColor",
		"edgeColor"
	], 3);
	vtkProperty(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkProperty");
var Property_default = {
	newInstance,
	extend,
	...Constants_default
};
//#endregion
export { Property_default as default, extend, newInstance };
