// @kitware/vtk.js/Rendering/Core/SurfaceLICInterface.js
import macros_default from "../../macros.js";
import { ColorMode, ContrastEnhanceMode, NoiseType } from "./SurfaceLICInterface/Constants.js";
//#region Sources/Rendering/Core/SurfaceLICInterface/index.js
function vtkSurfaceLICInterface(publicAPI, model) {
	model.classHierarchy.push("vtkSurfaceLICInterface");
}
var DEFAULT_VALUES = {
	enableLIC: false,
	nuberOfSteps: 40,
	stepSize: .25,
	transformVectors: true,
	normalizeVectors: true,
	maskOnSurface: false,
	maskThreshold: 0,
	maskColor: [
		0,
		0,
		0
	],
	maskIntensity: 0,
	enhancedLIC: true,
	enhanceContrast: ContrastEnhanceMode.NONE,
	lowLICContrastEnhancementFactor: 0,
	highLICContrastEnhancementFactor: 0,
	lowColorContrastEnhancementFactor: 0,
	highColorContrastEnhancementFactor: 0,
	antiAlias: 0,
	colorMode: ColorMode.BLEND,
	LICIntensity: 1,
	mapModeBias: 0,
	noiseTextureSize: 200,
	noiseTextureType: NoiseType.GAUSSIAN,
	noiseGrainSize: 8,
	noiseImpulseProbability: .1,
	noiseImpulseBackgroundValue: 0,
	noiseGeneratorSeed: 0,
	minNoiseValue: 0,
	maxNoiseValue: 1,
	numberOfNoiseLevels: 2,
	shadersNeedBuilding: true,
	reallocateTextures: true,
	rebuildNoiseTexture: false,
	viewPortScale: 1
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, [
		"enableLIC",
		"numberOfSteps",
		"stepSize",
		"normalizeVectors",
		"transformVectors",
		"maskOnSurface",
		"maskThreshold",
		"maskColor",
		"maskIntensity",
		"enhancedLIC",
		"enhanceContrast",
		"lowLICContrastEnhancementFactor",
		"highLICContrastEnhancementFactor",
		"lowColorContrastEnhancementFactor",
		"highColorContrastEnhancementFactor",
		"antiAlias",
		"colorMode",
		"LICIntensity",
		"mapModeBias",
		"noiseTextureSize",
		"noiseTextureType",
		"noiseGrainSize",
		"minNoiseValue",
		"maxNoiseValue",
		"numberOfNoiseLevels",
		"noiseImpulseProbability",
		"noiseImpulseBackgroundValue",
		"noiseGeneratorSeed",
		"viewPortScale",
		"rebuildNoiseTexture"
	]);
	vtkSurfaceLICInterface(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkSurfaceLICInterface");
var SurfaceLICInterface_default = {
	newInstance,
	extend
};
//#endregion
export { SurfaceLICInterface_default as default, extend, newInstance };
