// @kitware/vtk.js/Rendering/Core/TextProperty.js
import macros_default from "../../macros.js";
import Property2D_default from "./Property2D.js";
//#region Sources/Rendering/Core/TextProperty/index.js
function vtkTextProperty(publicAPI, model) {
	model.classHierarchy.push("vtkTextProperty");
}
var DEFAULT_VALUES = {
	fontFamily: "Arial",
	fontColor: [
		0,
		0,
		0
	],
	fontStyle: "normal",
	fontSizeScale: (resolution) => resolution / 1.8,
	resolution: 200
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	Property2D_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, [
		"fontFamily",
		"fontStyle",
		"fillStyle",
		"fontSizeScale",
		"resolution",
		"shadowBlur"
	]);
	macros_default.setGetArray(publicAPI, model, ["shadowOffset"], 2);
	macros_default.setGetArray(publicAPI, model, [
		"backgroundColor",
		"fontColor",
		"shadowColor"
	], 3);
	vtkTextProperty(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkTextProperty");
var TextProperty_default = {
	newInstance,
	extend
};
//#endregion
export { TextProperty_default as default, extend, newInstance };
