// @kitware/vtk.js/Rendering/Core/TextActor.js
import macros_default from "../../macros.js";
import { floatRGB2HexCode } from "../../Common/Core/Math.js";
import ImageHelper_default from "../../Common/Core/ImageHelper.js";
import PlaneSource_default from "../../Filters/Sources/PlaneSource.js";
import Texture_default from "./Texture.js";
import Actor2D_default from "./Actor2D.js";
import Mapper2D_default from "./Mapper2D.js";
import TextProperty_default from "./TextProperty.js";
//#region Sources/Rendering/Core/TextActor/index.js
function vtkTextActor(publicAPI, model) {
	model.classHierarchy.push("vtkTextActor");
	const superClass = { ...publicAPI };
	publicAPI.makeProperty = TextProperty_default.newInstance;
	const texture = Texture_default.newInstance({ resizable: true });
	const canvas = new OffscreenCanvas(1, 1);
	const mapper = Mapper2D_default.newInstance();
	const plane = PlaneSource_default.newInstance({
		xResolution: 1,
		yResolution: 1
	});
	/**
	* Normalize escaped and platform specific line endings, then split into lines.
	*/
	function splitLines(text) {
		return text.replace(/\\r\\n/g, "\n").replace(/\\n/g, "\n").replace(/\\r/g, "\n").replace(/\r\n/g, "\n").replace(/\r/g, "\n").split("\n");
	}
	function createImageData(text) {
		const fontSizeScale = publicAPI.getProperty().getFontSizeScale();
		const fontStyle = publicAPI.getProperty().getFontStyle();
		const fontFamily = publicAPI.getProperty().getFontFamily();
		const fontColor = publicAPI.getProperty().getFontColor();
		const shadowColor = publicAPI.getProperty().getShadowColor();
		const shadowOffset = publicAPI.getProperty().getShadowOffset();
		const shadowBlur = publicAPI.getProperty().getShadowBlur();
		const resolution = publicAPI.getProperty().getResolution();
		const backgroundColor = publicAPI.getProperty().getBackgroundColor();
		const dpr = Math.max(window.devicePixelRatio || 1, 1);
		const ctx = canvas.getContext("2d", { willReadFrequently: true });
		const textSize = fontSizeScale(resolution);
		const lines = splitLines(text);
		ctx.font = `${fontStyle} ${textSize}px "${fontFamily}"`;
		ctx.textBaseline = "alphabetic";
		ctx.textAlign = "left";
		const lineMetrics = lines.map((line) => {
			const metrics = ctx.measureText(line);
			return {
				left: metrics.actualBoundingBoxLeft || 0,
				right: metrics.actualBoundingBoxRight || metrics.width || 0,
				actualAscent: metrics.actualBoundingBoxAscent || 0,
				actualDescent: metrics.actualBoundingBoxDescent || 0
			};
		});
		const maxActualAscent = lineMetrics.reduce((value, metrics) => Math.max(value, metrics.actualAscent), 0);
		const maxActualDescent = lineMetrics.reduce((value, metrics) => Math.max(value, metrics.actualDescent), 0);
		const maxLeft = lineMetrics.reduce((value, metrics) => Math.max(value, metrics.left), 0);
		const maxRight = lineMetrics.reduce((value, metrics) => Math.max(value, metrics.right), 0);
		const lineAscent = maxActualAscent;
		const lineDescent = maxActualDescent;
		const baselineOffset = lineAscent;
		const lineHeight = Math.max(lineAscent + lineDescent, 1);
		const textWidth = maxLeft + maxRight;
		const textHeight = Math.max(lines.length * lineHeight, lineHeight);
		const width = Math.max(Math.round(textWidth * dpr), 1);
		const height = Math.max(Math.round(textHeight * dpr), 1);
		canvas.width = width;
		canvas.height = height;
		ctx.setTransform(dpr, 0, 0, -dpr, 0, height);
		ctx.clearRect(0, 0, width, height);
		if (backgroundColor) {
			ctx.fillStyle = floatRGB2HexCode(backgroundColor);
			ctx.fillRect(0, 0, width, height);
		}
		ctx.imageSmoothingEnabled = true;
		ctx.imageSmoothingQuality = "high";
		ctx.font = `${fontStyle} ${textSize}px "${fontFamily}"`;
		ctx.fillStyle = floatRGB2HexCode(fontColor);
		ctx.textBaseline = "alphabetic";
		ctx.textAlign = "left";
		if (shadowColor) {
			ctx.shadowColor = floatRGB2HexCode(shadowColor);
			ctx.shadowOffsetX = shadowOffset[0];
			ctx.shadowOffsetY = shadowOffset[1];
			ctx.shadowBlur = shadowBlur;
		}
		const x = maxLeft;
		const baseline = baselineOffset;
		const lineHeightPx = lineHeight;
		lines.forEach((line, index) => {
			ctx.fillText(line, x, baseline + index * lineHeightPx);
		});
		plane.set({
			point1: [
				width,
				0,
				0
			],
			point2: [
				0,
				height,
				0
			]
		});
		return ImageHelper_default.canvasToImageData(canvas);
	}
	function updateTexture() {
		const image = createImageData(model.input);
		texture.setInputData(image, 0);
		model.textureBuildTime.modified();
	}
	function updateTextureIfNeeded() {
		if (model.input === void 0) return;
		if (model.textureBuildTime.getMTime() < model.mtime || model.textureBuildTime.getMTime() < publicAPI.getProperty().getMTime()) updateTexture();
	}
	mapper.setInputConnection(plane.getOutputPort());
	publicAPI.setMapper(mapper);
	publicAPI.addTexture(texture);
	publicAPI.getTextures = () => {
		updateTextureIfNeeded();
		return superClass.getTextures();
	};
}
var DEFAULT_VALUES = { textureBuildTime: null };
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	Actor2D_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, ["input"]);
	model.textureBuildTime = {};
	macros_default.obj(model.textureBuildTime, { mtime: 0 });
	vtkTextActor(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkTextActor");
var TextActor_default = {
	newInstance,
	extend
};
//#endregion
export { TextActor_default as default, extend, newInstance };
