// @kitware/vtk.js/Rendering/Core/VectorText.js
import macros_default from "../../macros.js";
import { subtract } from "../../Common/Core/Math.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import CellArray_default from "../../Common/Core/CellArray.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
import { buildLidFaces, buildSideFaces, computeBevelVector, createShapePath, getBoundingSize, isClockWise, scalePoint, triangulateShape } from "./VectorText/Utils.js";
//#region Sources/Rendering/Core/VectorText/index.js
var { vtkErrorMacro, vtkWarningMacro } = macros_default;
function vtkVectorText(publicAPI, model) {
	model.classHierarchy.push("vtkVectorText");
	/**
	* Process a shape into 3D geometry
	* @param {Object} shape - The shape to process
	* @param {Array} offsetSize - The offset size for positioning the shape
	* @param {Array} letterColor - The color for the shape
	*/
	function addShape(shape, offsetSize, letterColor) {
		const curveSegments = model.curveSegments;
		const steps = model.steps;
		const depth = model.depth;
		const bevelEnabled = model.bevelEnabled;
		let bevelThickness = model.bevelThickness;
		let bevelSize = bevelThickness - .1;
		let bevelOffset = model.bevelOffset;
		let bevelSegments = model.bevelSegments;
		if (!bevelEnabled) {
			bevelSegments = 0;
			bevelThickness = 0;
			bevelSize = 0;
			bevelOffset = 0;
		}
		const shapePoints = shape.extractPoints(curveSegments);
		let vertices = shapePoints.shape;
		const holes = shapePoints.holes;
		vertices.forEach((p) => {
			p[0] += offsetSize[0];
			p[1] += offsetSize[1];
		});
		holes.forEach((hole) => {
			hole.forEach((p) => {
				p[0] += offsetSize[0];
				p[1] += offsetSize[1];
			});
		});
		if (vertices.length < 3) {
			vtkWarningMacro("Not enough points to create a shape");
			return;
		}
		const faces = triangulateShape(model.earcut, vertices, holes);
		const contour = vertices;
		vertices = [...vertices, ...holes.flat()];
		const vlen = vertices.length;
		const contourMovements = [];
		for (let i = 0, j = contour.length - 1, k = i + 1; i < contour.length; i++, j++, k++) {
			if (j === contour.length) j = 0;
			if (k === contour.length) k = 0;
			contourMovements[i] = computeBevelVector(contour[i], contour[j], contour[k]);
		}
		const holesMovements = [];
		let oneHoleMovements;
		let verticesMovements = [...contourMovements];
		for (let h = 0, hl = holes.length; h < hl; h++) {
			const ahole = holes[h];
			oneHoleMovements = [];
			for (let i = 0, j = ahole.length - 1, k = i + 1; i < ahole.length; i++, j++, k++) {
				if (j === ahole.length) j = 0;
				if (k === ahole.length) k = 0;
				oneHoleMovements[i] = computeBevelVector(ahole[i], ahole[j], ahole[k]);
			}
			holesMovements.push(oneHoleMovements);
			verticesMovements = [...verticesMovements, ...oneHoleMovements];
		}
		const layers = [];
		for (let b = 0; b < bevelSegments; b++) {
			const t = b / bevelSegments;
			const z = bevelThickness * Math.cos(t * Math.PI / 2);
			const bs = bevelSize * Math.sin(t * Math.PI / 2) + bevelOffset;
			for (let i = 0; i < contour.length; i++) {
				const vert = scalePoint(contour[i], contourMovements[i], bs);
				layers.push(vert[0], vert[1], -z + offsetSize[2]);
			}
			for (let h = 0, hl = holes.length; h < hl; h++) {
				const ahole = holes[h];
				oneHoleMovements = holesMovements[h];
				for (let i = 0; i < ahole.length; i++) {
					const vert = scalePoint(ahole[i], oneHoleMovements[i], bs);
					layers.push(vert[0], vert[1], -z + offsetSize[2]);
				}
			}
		}
		const bs = bevelSize + bevelOffset;
		for (let i = 0; i < vlen; i++) {
			const vert = bevelEnabled ? scalePoint(vertices[i], verticesMovements[i], bs) : vertices[i];
			layers.push(vert[0], vert[1], 0 + offsetSize[2]);
		}
		for (let s = 1; s <= steps; s++) for (let i = 0; i < vlen; i++) {
			const vert = bevelEnabled ? scalePoint(vertices[i], verticesMovements[i], bs) : vertices[i];
			layers.push(vert[0], vert[1], depth / steps * s + offsetSize[2]);
		}
		for (let b = bevelSegments - 1; b >= 0; b--) {
			const t = b / bevelSegments;
			const z = bevelThickness * Math.cos(t * Math.PI / 2);
			const topBevelSize = bevelSize * Math.sin(t * Math.PI / 2) + bevelOffset;
			for (let i = 0, il = contour.length; i < il; i++) {
				const vert = scalePoint(contour[i], contourMovements[i], topBevelSize);
				layers.push(vert[0], vert[1], depth + z + offsetSize[2]);
			}
			for (let h = 0, hl = holes.length; h < hl; h++) {
				const ahole = holes[h];
				oneHoleMovements = holesMovements[h];
				for (let i = 0, il = ahole.length; i < il; i++) {
					const vert = scalePoint(ahole[i], oneHoleMovements[i], topBevelSize);
					layers.push(vert[0], vert[1], depth + z + offsetSize[2]);
				}
			}
		}
		buildLidFaces(layers, faces, vlen, steps, bevelEnabled, bevelSegments, model.verticesArray, model.uvArray, model.colorArray, letterColor);
		buildSideFaces(layers, contour, holes, vlen, steps, bevelSegments, model.verticesArray, model.uvArray, model.colorArray, letterColor);
	}
	/**
	* Creates shape paths from the font and text
	*/
	function buildShape() {
		model.shapes = [];
		if (!model.font || !model.text) return;
		const path = model.font.getPath(model.text, 0, 0, model.fontSize);
		if (!path || !path.commands || !path.commands.length) return;
		let first;
		let shapePath = createShapePath();
		const commands = path.commands;
		for (let i = 0; i < commands.length; i++) {
			const command = commands[i];
			shapePath = shapePath || createShapePath();
			switch (command.type) {
				case "M":
					shapePath.moveTo(command.x, -command.y);
					first = command;
					break;
				case "L":
					shapePath.lineTo(command.x, -command.y);
					break;
				case "C":
					shapePath.bezierCurveTo(command.x1, -command.y1, command.x2, -command.y2, command.x, -command.y);
					break;
				case "Q":
					shapePath.quadraticCurveTo(command.x1, -command.y1, command.x, -command.y);
					break;
				case "Z":
					shapePath.lineTo(first.x, -first.y);
					if (isClockWise(shapePath.getPoints(1))) model.shapes.push(shapePath);
					else for (let j = 0; j < model.shapes.length; j++) {
						const shape = model.shapes[j];
						if (shape.isIntersect(shapePath)) {
							shape.holes.push(shapePath);
							break;
						}
					}
					shapePath = null;
					break;
				default:
					console.warn(`Unknown path command: ${command.type}`);
					break;
			}
		}
		if (shapePath) model.shapes.push(shapePath);
	}
	/**
	* Creates a vtkPolyData from the processed shapes
	* @returns {Object} vtkPolyData instance
	*/
	function buildPolyData(polyData) {
		model.verticesArray = [];
		model.uvArray = [];
		model.colorArray = [];
		const cells = CellArray_default.newInstance();
		const pointData = polyData.getPointData();
		const boundingSize = getBoundingSize(model.shapes, model.depth, model.curveSegments);
		const offsetSize = [
			0,
			0,
			0
		];
		subtract(boundingSize.min, boundingSize.max, offsetSize);
		let letterIndex = 0;
		model.shapes.forEach((shape) => {
			let color = null;
			if (model.perLetterFaceColors) color = model.perLetterFaceColors(letterIndex) || [
				1,
				1,
				1
			];
			addShape(shape, offsetSize, color);
			letterIndex++;
		});
		const vertexCount = model.verticesArray.length / 3;
		const indices = [];
		for (let i = 0; i < vertexCount; i += 3) indices.push(i, i + 2, i + 1);
		const cellSize = indices.length;
		cells.resize(cellSize + cellSize / 3);
		for (let i = 0; i < indices.length; i += 3) cells.insertNextCell([
			indices[i],
			indices[i + 1],
			indices[i + 2]
		]);
		polyData.setPolys(cells);
		polyData.getPoints().setData(Float32Array.from(model.verticesArray), 3);
		const da = DataArray_default.newInstance({
			numberOfComponents: 2,
			values: Float32Array.from(model.uvArray),
			name: "TEXCOORD_0"
		});
		pointData.addArray(da);
		pointData.setActiveTCoords(da.getName());
		if (model.colorArray && model.colorArray.length) {
			const ca = DataArray_default.newInstance({
				numberOfComponents: 3,
				values: Uint8Array.from(model.colorArray),
				name: "Colors"
			});
			pointData.addArray(ca);
			pointData.setActiveScalars(ca.getName());
		}
		return polyData;
	}
	/**
	* Handles the request to generate vector text data
	* @param {Object} inData - Input data (not used)
	* @param {Object} outData - Output data target
	*/
	publicAPI.requestData = (inData, outData) => {
		if (!model.font) {
			vtkErrorMacro("Font object not set, make sure the TTF file is parsed using opentype.js.");
			return;
		}
		if (!model.text) {
			vtkErrorMacro("Text not set. Cannot generate vector text.");
			return;
		}
		buildShape();
		const polyData = outData[0]?.initialize() || PolyData_default.newInstance();
		buildPolyData(polyData);
		outData[0] = polyData;
	};
}
/**
* Default values for the VectorText model
* shapes: Array to store shape paths
* verticesArray: Array of vertex coordinates
* uvArray: Array of texture coordinates
* font: Font object (from opentype.js)
* earcut: Earcut module for triangulation
* fontSize: Font size in points
* depth: Depth of the extruded text
* steps: Number of steps in extrusion (for curved surfaces)
* bevelEnabled: Whether to add beveled edges
* curveSegments: Number of segments for curved paths
* bevelThickness: Thickness of the bevel
* bevelSize: Size of the bevel
* bevelOffset: Offset of the bevel
* bevelSegments: Number of segments in the bevel
* text: The text to render
* perLetterFaceColors: Function to get per-letter face colors
*/
var DEFAULT_VALUES = {
	shapes: [],
	verticesArray: [],
	uvArray: [],
	font: null,
	earcut: null,
	fontSize: 10,
	depth: 1,
	steps: 1,
	bevelEnabled: false,
	curveSegments: 12,
	bevelThickness: .2,
	bevelSize: .1,
	bevelOffset: 0,
	bevelSegments: 1,
	text: null,
	perLetterFaceColors: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.algo(publicAPI, model, 0, 1);
	macros_default.setGet(publicAPI, model, [
		"fontSize",
		"text",
		"depth",
		"steps",
		"bevelEnabled",
		"curveSegments",
		"bevelThickness",
		"bevelSize",
		"bevelOffset",
		"bevelSegments",
		"perLetterFaceColors"
	]);
	macros_default.set(publicAPI, model, ["font"]);
	vtkVectorText(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkVectorText");
var VectorText_default = {
	newInstance,
	extend
};
//#endregion
export { VectorText_default as default, extend, newInstance };
