// @kitware/vtk.js/Rendering/Core/Viewport.js
import macros_default from "../../macros.js";
//#region Sources/Rendering/Core/Viewport/index.js
var { vtkErrorMacro } = macros_default;
function notImplemented(method) {
	return () => vtkErrorMacro(`vtkViewport::${method} - NOT IMPLEMENTED`);
}
function vtkViewport(publicAPI, model) {
	model.classHierarchy.push("vtkViewport");
	publicAPI.getViewProps = () => model.props;
	publicAPI.hasViewProp = (prop) => model.props.includes(prop);
	publicAPI.addViewProp = (prop) => {
		if (prop && !publicAPI.hasViewProp(prop)) model.props.push(prop);
	};
	publicAPI.removeViewProp = (prop) => {
		const newPropList = model.props.filter((item) => item !== prop);
		if (model.props.length !== newPropList.length) model.props = newPropList;
	};
	publicAPI.removeAllViewProps = () => {
		model.props = [];
	};
	function gatherProps(prop, allProps = []) {
		allProps.push(prop);
		const children = prop.getNestedProps();
		if (children && children.length) for (let i = 0; i < children.length; i++) gatherProps(children[i], allProps);
		return allProps;
	}
	publicAPI.getViewPropsWithNestedProps = () => {
		let allPropsArray = [];
		model.actors2D = [];
		for (let i = 0; i < model.props.length; i++) {
			const isActor2D = model.props[i].getActors2D();
			if (isActor2D && (!Array.isArray(isActor2D) || isActor2D.length > 0)) model.actors2D = model.actors2D.concat(isActor2D);
			else gatherProps(model.props[i], allPropsArray);
		}
		model.actors2D.sort((a, b) => a.getLayerNumber() - b.getLayerNumber());
		allPropsArray = allPropsArray.concat(model.actors2D);
		return allPropsArray;
	};
	publicAPI.addActor2D = publicAPI.addViewProp;
	publicAPI.removeActor2D = (prop) => {
		publicAPI.removeViewProp(prop);
	};
	publicAPI.getActors2D = () => {
		model.actors2D = [];
		model.props.forEach((prop) => {
			model.actors2D = model.actors2D.concat(prop.getActors2D());
		});
		return model.actors2D;
	};
	publicAPI.displayToView = () => vtkErrorMacro("call displayToView on your view instead");
	publicAPI.viewToDisplay = () => vtkErrorMacro("callviewtodisplay on your view instead");
	publicAPI.getSize = () => vtkErrorMacro("call getSize on your View instead");
	publicAPI.normalizedDisplayToProjection = (x, y, z) => {
		const nvp = publicAPI.normalizedDisplayToNormalizedViewport(x, y, z);
		return publicAPI.normalizedViewportToProjection(nvp[0], nvp[1], nvp[2]);
	};
	publicAPI.normalizedDisplayToNormalizedViewport = (x, y, z) => {
		const scale = [model.viewport[2] - model.viewport[0], model.viewport[3] - model.viewport[1]];
		return [
			(x - model.viewport[0]) / scale[0],
			(y - model.viewport[1]) / scale[1],
			z
		];
	};
	publicAPI.normalizedViewportToProjection = (x, y, z) => [
		x * 2 - 1,
		y * 2 - 1,
		z * 2 - 1
	];
	publicAPI.projectionToNormalizedDisplay = (x, y, z) => {
		const nvp = publicAPI.projectionToNormalizedViewport(x, y, z);
		return publicAPI.normalizedViewportToNormalizedDisplay(nvp[0], nvp[1], nvp[2]);
	};
	publicAPI.normalizedViewportToNormalizedDisplay = (x, y, z) => {
		const scale = [model.viewport[2] - model.viewport[0], model.viewport[3] - model.viewport[1]];
		return [
			x * scale[0] + model.viewport[0],
			y * scale[1] + model.viewport[1],
			z
		];
	};
	publicAPI.projectionToNormalizedViewport = (x, y, z) => [
		(x + 1) * .5,
		(y + 1) * .5,
		(z + 1) * .5
	];
	publicAPI.PickPropFrom = notImplemented("PickPropFrom");
}
var DEFAULT_VALUES = {
	background: [
		0,
		0,
		0
	],
	background2: [
		.2,
		.2,
		.2
	],
	gradientBackground: false,
	viewport: [
		0,
		0,
		1,
		1
	],
	aspect: [1, 1],
	pixelAspect: [1, 1],
	props: [],
	actors2D: []
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.event(publicAPI, model, "event");
	macros_default.setGet(publicAPI, model, ["gradientBackground"]);
	macros_default.setGetArray(publicAPI, model, ["viewport"], 4);
	macros_default.setGetArray(publicAPI, model, ["background", "background2"], 3);
	vtkViewport(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkViewport");
var Viewport_default = {
	newInstance,
	extend
};
//#endregion
export { Viewport_default as default, extend, newInstance };
