// @kitware/vtk.js/Rendering/Core/VolumeMapper.js
import macros_default from "../../macros.js";
import BoundingBox_default from "../../Common/DataModel/BoundingBox.js";
import PiecewiseFunction_default from "../../Common/DataModel/PiecewiseFunction.js";
import AbstractMapper3D_default from "./AbstractMapper3D.js";
import Constants_default from "./VolumeMapper/Constants.js";
//#region Sources/Rendering/Core/VolumeMapper/index.js
var { BlendMode } = Constants_default;
function createRadonTransferFunction(firstAbsorbentMaterialHounsfieldValue, firstAbsorbentMaterialAbsorption, maxAbsorbentMaterialHounsfieldValue, maxAbsorbentMaterialAbsorption, outputTransferFunction) {
	let ofun = null;
	if (outputTransferFunction) {
		ofun = outputTransferFunction;
		ofun.removeAllPoints();
	} else ofun = PiecewiseFunction_default.newInstance();
	ofun.addPointLong(-1024, 0, 1, 1);
	ofun.addPoint(firstAbsorbentMaterialHounsfieldValue, firstAbsorbentMaterialAbsorption);
	ofun.addPoint(maxAbsorbentMaterialHounsfieldValue, maxAbsorbentMaterialAbsorption);
	return ofun;
}
var methodNamesMovedToVolumeProperties = [
	"getAnisotropy",
	"getComputeNormalFromOpacity",
	"getFilterMode",
	"getFilterModeAsString",
	"getGlobalIlluminationReach",
	"getIpScalarRange",
	"getIpScalarRangeByReference",
	"getLAOKernelRadius",
	"getLAOKernelSize",
	"getLocalAmbientOcclusion",
	"getPreferSizeOverAccuracy",
	"getVolumetricScatteringBlending",
	"setAnisotropy",
	"setAverageIPScalarRange",
	"setComputeNormalFromOpacity",
	"setFilterMode",
	"setFilterModeToNormalized",
	"setFilterModeToOff",
	"setFilterModeToRaw",
	"setGlobalIlluminationReach",
	"setIpScalarRange",
	"setIpScalarRangeFrom",
	"setLAOKernelRadius",
	"setLAOKernelSize",
	"setLocalAmbientOcclusion",
	"setPreferSizeOverAccuracy",
	"setVolumetricScatteringBlending"
];
var STATIC = { createRadonTransferFunction };
function vtkVolumeMapper(publicAPI, model) {
	model.classHierarchy.push("vtkVolumeMapper");
	const superClass = { ...publicAPI };
	publicAPI.computeBounds = () => {
		const input = publicAPI.getInputData();
		if (!input) {
			BoundingBox_default.reset(model.bounds);
			return;
		}
		if (!model.static) publicAPI.update();
		BoundingBox_default.setBounds(model.bounds, input.getBounds());
	};
	publicAPI.setBlendModeToComposite = () => {
		publicAPI.setBlendMode(BlendMode.COMPOSITE_BLEND);
	};
	publicAPI.setBlendModeToMaximumIntensity = () => {
		publicAPI.setBlendMode(BlendMode.MAXIMUM_INTENSITY_BLEND);
	};
	publicAPI.setBlendModeToMinimumIntensity = () => {
		publicAPI.setBlendMode(BlendMode.MINIMUM_INTENSITY_BLEND);
	};
	publicAPI.setBlendModeToAverageIntensity = () => {
		publicAPI.setBlendMode(BlendMode.AVERAGE_INTENSITY_BLEND);
	};
	publicAPI.setBlendModeToAdditiveIntensity = () => {
		publicAPI.setBlendMode(BlendMode.ADDITIVE_INTENSITY_BLEND);
	};
	publicAPI.setBlendModeToRadonTransform = () => {
		publicAPI.setBlendMode(BlendMode.RADON_TRANSFORM_BLEND);
	};
	publicAPI.getBlendModeAsString = () => macros_default.enumToString(BlendMode, model.blendMode);
	publicAPI.setVolumeShadowSamplingDistFactor = (vsdf) => superClass.setVolumeShadowSamplingDistFactor(vsdf >= 1 ? vsdf : 1);
	methodNamesMovedToVolumeProperties.forEach((removedMethodName) => {
		const removedMethod = () => {
			throw new Error(`The method "volumeMapper.${removedMethodName}()" doesn't exist anymore. It is a rendering property that has been moved to the volume property. Replace your code with:\nvolumeActor.getProperty().${removedMethodName}()\n`);
		};
		publicAPI[removedMethodName] = removedMethod;
	});
}
var defaultValues = (initialValues) => ({
	bounds: [...BoundingBox_default.INIT_BOUNDS],
	sampleDistance: 1,
	imageSampleDistance: 1,
	maximumSamplesPerRay: 1e3,
	autoAdjustSampleDistances: true,
	initialInteractionScale: 1,
	interactionSampleDistanceFactor: 1,
	blendMode: BlendMode.COMPOSITE_BLEND,
	volumeShadowSamplingDistFactor: 5,
	colorTextureWidth: 1024,
	opacityTextureWidth: 1024,
	labelOutlineTextureWidth: 1024,
	...initialValues
});
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, defaultValues(initialValues));
	AbstractMapper3D_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, [
		"sampleDistance",
		"imageSampleDistance",
		"maximumSamplesPerRay",
		"autoAdjustSampleDistances",
		"initialInteractionScale",
		"interactionSampleDistanceFactor",
		"blendMode",
		"volumeShadowSamplingDistFactor",
		"colorTextureWidth",
		"opacityTextureWidth",
		"labelOutlineTextureWidth"
	]);
	macros_default.event(publicAPI, model, "lightingActivated");
	vtkVolumeMapper(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkVolumeMapper");
var VolumeMapper_default = {
	newInstance,
	extend,
	...STATIC
};
//#endregion
export { STATIC, VolumeMapper_default as default, extend, newInstance };
