// @kitware/vtk.js/Widgets/Widgets3D/RectangleWidget/state.js
import StateBuilder_default from "../../Core/StateBuilder.js";
import { TextPosition } from "../ShapeWidget/Constants.js";
//#region Sources/Widgets/Widgets3D/RectangleWidget/state.js
function generateState() {
	return StateBuilder_default.createBuilder().addStateFromMixin({
		labels: ["moveHandle"],
		mixins: [
			"origin",
			"color",
			"scale1",
			"visible",
			"manipulator"
		],
		name: "point1Handle",
		initialValues: {
			scale1: 10,
			visible: false
		}
	}).addStateFromMixin({
		labels: ["moveHandle"],
		mixins: [
			"origin",
			"color",
			"scale1",
			"visible",
			"manipulator"
		],
		name: "point2Handle",
		initialValues: {
			scale1: 10,
			visible: false
		}
	}).addStateFromMixin({
		labels: ["rectangleHandle"],
		mixins: [
			"origin",
			"corner",
			"color",
			"visible",
			"orientation"
		],
		name: "rectangleHandle",
		initialValues: { visible: false }
	}).addStateFromMixin({
		labels: ["SVGtext"],
		mixins: [
			"origin",
			"color",
			"text",
			"visible"
		],
		name: "text",
		initialValues: { text: "" }
	}).addField({
		name: "textPosition",
		initialValue: [
			TextPosition.CENTER,
			TextPosition.CENTER,
			TextPosition.CENTER
		]
	}).addField({
		name: "textWorldMargin",
		initialValue: 0
	}).build();
}
//#endregion
export { generateState as default };
