// @kitware/vtk.js/macros.js
import { __exportAll } from "./_virtual/_rolldown/runtime.js";
import vtk, { vtkGlobal } from "./vtk.js";
import ClassHierarchy from "./Common/Core/ClassHierarchy.js";
import DeepEqual from "fast-deep-equal";
//#region Sources/macros.js
/**
* macros.js is the old macro.js.
* The name change is so we do not get eaten by babel-plugin-macros.
*/
var macros_exports = /* @__PURE__ */ __exportAll({
	EVENT_ABORT: () => EVENT_ABORT,
	TYPED_ARRAYS: () => TYPED_ARRAYS,
	VOID: () => VOID,
	_capitalize: () => _capitalize,
	algo: () => algo,
	capitalize: () => capitalize,
	chain: () => chain,
	debounce: () => debounce,
	default: () => macros_default,
	event: () => event,
	formatBytesToProperUnit: () => formatBytesToProperUnit,
	formatNumbersWithThousandSeparator: () => formatNumbersWithThousandSeparator,
	get: () => get,
	getArray: () => getArray,
	isVtkObject: () => isVtkObject,
	keystore: () => keystore,
	measurePromiseExecution: () => measurePromiseExecution,
	moveToProtected: () => moveToProtected,
	newInstance: () => newInstance,
	newTypedArray: () => newTypedArray,
	newTypedArrayFrom: () => newTypedArrayFrom,
	normalizeWheel: () => normalizeWheel,
	obj: () => obj,
	proxy: () => proxy,
	proxyPropertyMapping: () => proxyPropertyMapping,
	proxyPropertyState: () => proxyPropertyState,
	requiredParam: () => requiredParam,
	set: () => set,
	setArray: () => setArray,
	setGet: () => setGet,
	setGetArray: () => setGetArray,
	setImmediateVTK: () => setImmediateVTK,
	setLoggerFunction: () => setLoggerFunction,
	throttle: () => throttle,
	traverseInstanceTree: () => traverseInstanceTree,
	uncapitalize: () => uncapitalize,
	vtkDebugMacro: () => vtkDebugMacro,
	vtkErrorMacro: () => vtkErrorMacro,
	vtkInfoMacro: () => vtkInfoMacro,
	vtkLogMacro: () => vtkLogMacro,
	vtkOnceErrorMacro: () => vtkOnceErrorMacro,
	vtkWarningMacro: () => vtkWarningMacro
});
var globalMTime = 0;
var requiredParam = (name) => {
	throw new Error(`Named parameter '${name}' is missing`);
};
var VOID = Symbol("void");
function getCurrentGlobalMTime() {
	return globalMTime;
}
var fakeConsole = {};
function noOp() {}
[
	"log",
	"debug",
	"info",
	"warn",
	"error",
	"time",
	"timeEnd",
	"group",
	"groupEnd"
].forEach((methodName) => {
	fakeConsole[methodName] = noOp;
});
vtkGlobal.console = console.hasOwnProperty("log") ? console : fakeConsole;
var loggerFunctions = {
	debug: noOp,
	error: vtkGlobal.console.error || noOp,
	info: vtkGlobal.console.info || noOp,
	log: vtkGlobal.console.log || noOp,
	warn: vtkGlobal.console.warn || noOp
};
function setLoggerFunction(name, fn) {
	if (loggerFunctions[name]) loggerFunctions[name] = fn || noOp;
}
function vtkLogMacro(...args) {
	loggerFunctions.log(...args);
}
function vtkInfoMacro(...args) {
	loggerFunctions.info(...args);
}
function vtkDebugMacro(...args) {
	loggerFunctions.debug(...args);
}
function vtkErrorMacro(...args) {
	loggerFunctions.error(...args);
}
function vtkWarningMacro(...args) {
	loggerFunctions.warn(...args);
}
var ERROR_ONCE_MAP = {};
function vtkOnceErrorMacro(str) {
	if (!ERROR_ONCE_MAP[str]) {
		loggerFunctions.error(str);
		ERROR_ONCE_MAP[str] = true;
	}
}
var TYPED_ARRAYS = Object.create(null);
TYPED_ARRAYS.Float32Array = Float32Array;
TYPED_ARRAYS.Float64Array = Float64Array;
TYPED_ARRAYS.Uint8Array = Uint8Array;
TYPED_ARRAYS.Int8Array = Int8Array;
TYPED_ARRAYS.Uint16Array = Uint16Array;
TYPED_ARRAYS.Int16Array = Int16Array;
TYPED_ARRAYS.Uint32Array = Uint32Array;
TYPED_ARRAYS.Int32Array = Int32Array;
TYPED_ARRAYS.Uint8ClampedArray = Uint8ClampedArray;
try {
	TYPED_ARRAYS.BigInt64Array = BigInt64Array;
	TYPED_ARRAYS.BigUint64Array = BigUint64Array;
} catch {}
function newTypedArray(type, ...args) {
	return new (TYPED_ARRAYS[type] || Float64Array)(...args);
}
function newTypedArrayFrom(type, ...args) {
	return (TYPED_ARRAYS[type] || Float64Array).from(...args);
}
function capitalize(str) {
	return str.charAt(0).toUpperCase() + str.slice(1);
}
function _capitalize(str) {
	return capitalize(str[0] === "_" ? str.slice(1) : str);
}
function uncapitalize(str) {
	return str.charAt(0).toLowerCase() + str.slice(1);
}
function formatBytesToProperUnit(size, precision = 2, chunkSize = 1e3) {
	const units = [
		"TB",
		"GB",
		"MB",
		"KB"
	];
	let value = Number(size);
	let currentUnit = "B";
	while (value > chunkSize) {
		value /= chunkSize;
		currentUnit = units.pop();
	}
	return `${value.toFixed(precision)} ${currentUnit}`;
}
function formatNumbersWithThousandSeparator(n, separator = " ") {
	const sections = [];
	let size = n;
	while (size > 1e3) {
		sections.push(`000${size % 1e3}`.slice(-3));
		size = Math.floor(size / 1e3);
	}
	if (size > 0) sections.push(size);
	sections.reverse();
	return sections.join(separator);
}
function safeArrays(model) {
	Object.keys(model).forEach((key) => {
		if (Array.isArray(model[key])) model[key] = [].concat(model[key]);
	});
}
function isTypedArray(value) {
	return Object.values(TYPED_ARRAYS).some((ctor) => value instanceof ctor);
}
function shallowEquals(a, b) {
	if (a === b) return true;
	if (Array.isArray(a) && Array.isArray(b)) {
		if (a.length !== b.length) return false;
		for (let i = 0; i < a.length; i++) if (a[i] !== b[i]) return false;
		return true;
	}
	return false;
}
function enumToString(e, value) {
	return Object.keys(e).find((key) => e[key] === value);
}
function setImmediateVTK(fn) {
	setTimeout(fn, 0);
}
function measurePromiseExecution(promise, callback) {
	const start = performance.now();
	promise.finally(() => {
		callback(performance.now() - start);
	});
}
function obj(publicAPI = {}, model = {}) {
	safeArrays(model);
	const callbacks = [];
	if (!Number.isInteger(model.mtime)) model.mtime = ++globalMTime;
	if (!("classHierarchy" in model)) model.classHierarchy = new ClassHierarchy("vtkObject");
	else if (!(model.classHierarchy instanceof ClassHierarchy)) {
		const hierarchy = new ClassHierarchy();
		for (let i = 0; i < model.classHierarchy.length; i++) hierarchy.push(model.classHierarchy[i]);
		model.classHierarchy = hierarchy;
	}
	function off(index) {
		callbacks[index] = null;
	}
	function on(index) {
		function unsubscribe() {
			off(index);
		}
		return Object.freeze({ unsubscribe });
	}
	publicAPI.isDeleted = () => !!model.deleted;
	publicAPI.modified = (otherMTime) => {
		if (model.deleted) {
			vtkErrorMacro("instance deleted - cannot call any method");
			return;
		}
		if (otherMTime && otherMTime < publicAPI.getMTime()) return;
		model.mtime = ++globalMTime;
		callbacks.forEach((callback) => callback && callback(publicAPI));
	};
	publicAPI.onModified = (callback) => {
		if (model.deleted) {
			vtkErrorMacro("instance deleted - cannot call any method");
			return null;
		}
		const index = callbacks.length;
		callbacks.push(callback);
		return on(index);
	};
	publicAPI.getMTime = () => model.mtime;
	publicAPI.isA = (className) => {
		let count = model.classHierarchy.length;
		while (count--) if (model.classHierarchy[count] === className) return true;
		return false;
	};
	publicAPI.getClassName = (depth = 0) => model.classHierarchy[model.classHierarchy.length - 1 - depth];
	publicAPI.set = (map = {}, noWarning = false, noFunction = false) => {
		let ret = false;
		Object.keys(map).forEach((name) => {
			const fn = noFunction ? null : publicAPI[`set${capitalize(name)}`];
			if (fn && Array.isArray(map[name]) && fn.length > 1) ret = fn(...map[name]) || ret;
			else if (fn) ret = fn(map[name]) || ret;
			else {
				if (["mtime"].indexOf(name) === -1 && !noWarning) vtkWarningMacro(`Warning: Set value to model directly ${name}, ${map[name]}`);
				ret = model[name] !== map[name] || ret;
				model[name] = map[name];
			}
		});
		return ret;
	};
	publicAPI.get = (...list) => {
		if (!list.length) return model;
		const subset = {};
		list.forEach((name) => {
			subset[name] = model[name];
		});
		return subset;
	};
	publicAPI.getReferenceByName = (val) => model[val];
	publicAPI.delete = () => {
		Object.keys(model).forEach((field) => delete model[field]);
		callbacks.forEach((el, index) => off(index));
		model.deleted = true;
	};
	publicAPI.getState = ({ preserveTypedArrays = false } = {}) => {
		if (model.deleted) return null;
		const options = { preserveTypedArrays };
		const jsonArchive = {
			...model,
			vtkClass: publicAPI.getClassName()
		};
		Object.keys(jsonArchive).forEach((keyName) => {
			if (jsonArchive[keyName] === null || jsonArchive[keyName] === void 0 || keyName[0] === "_") delete jsonArchive[keyName];
			else if (jsonArchive[keyName].isA) jsonArchive[keyName] = jsonArchive[keyName].getState(options);
			else if (Array.isArray(jsonArchive[keyName])) jsonArchive[keyName] = jsonArchive[keyName].map((item) => item && item.isA ? item.getState(options) : item);
			else if (isTypedArray(jsonArchive[keyName])) {
				if (!preserveTypedArrays) jsonArchive[keyName] = Array.from(jsonArchive[keyName]);
			}
		});
		const sortedObj = {};
		Object.keys(jsonArchive).sort().forEach((name) => {
			sortedObj[name] = jsonArchive[name];
		});
		if (sortedObj.mtime) delete sortedObj.mtime;
		return sortedObj;
	};
	publicAPI.shallowCopy = (other, debug = false) => {
		if (other.getClassName() !== publicAPI.getClassName()) throw new Error(`Cannot ShallowCopy ${other.getClassName()} into ${publicAPI.getClassName()}`);
		const otherModel = other.get();
		const keyList = Object.keys(model).sort();
		Object.keys(otherModel).sort().forEach((key) => {
			const keyIdx = keyList.indexOf(key);
			if (keyIdx === -1) {
				if (debug) vtkDebugMacro(`add ${key} in shallowCopy`);
			} else keyList.splice(keyIdx, 1);
			model[key] = otherModel[key];
		});
		if (keyList.length && debug) vtkDebugMacro(`Untouched keys: ${keyList.join(", ")}`);
		publicAPI.modified();
	};
	publicAPI.toJSON = function vtkObjToJSON() {
		return publicAPI.getState();
	};
	return publicAPI;
}
var objectGetterMap = { object(publicAPI, model, field) {
	return function getter() {
		return { ...model[field.name] };
	};
} };
function get(publicAPI, model, fieldNames) {
	fieldNames.forEach((field) => {
		if (typeof field === "object") {
			const getter = objectGetterMap[field.type];
			if (getter) publicAPI[`get${_capitalize(field.name)}`] = getter(publicAPI, model, field);
			else publicAPI[`get${_capitalize(field.name)}`] = () => model[field.name];
		} else publicAPI[`get${_capitalize(field)}`] = () => model[field];
	});
}
var objectSetterMap = {
	enum(publicAPI, model, field) {
		const onChanged = `_on${_capitalize(field.name)}Changed`;
		return (value) => {
			if (typeof value === "string") {
				if (field.enum[value] !== void 0) {
					if (model[field.name] !== field.enum[value]) {
						model[field.name] = field.enum[value];
						publicAPI.modified();
						return true;
					}
					return false;
				}
				vtkErrorMacro(`Set Enum with invalid argument ${field}, ${value}`);
				throw new RangeError("Set Enum with invalid string argument");
			}
			if (typeof value === "number") {
				if (model[field.name] !== value) {
					if (Object.keys(field.enum).map((key) => field.enum[key]).indexOf(value) !== -1) {
						const previousValue = model[field.name];
						model[field.name] = value;
						model[onChanged]?.(publicAPI, model, value, previousValue);
						publicAPI.modified();
						return true;
					}
					vtkErrorMacro(`Set Enum outside numeric range ${field}, ${value}`);
					throw new RangeError("Set Enum outside numeric range");
				}
				return false;
			}
			vtkErrorMacro(`Set Enum with invalid argument (String/Number) ${field}, ${value}`);
			throw new TypeError("Set Enum with invalid argument (String/Number)");
		};
	},
	object(publicAPI, model, field) {
		if (field.params?.length === 1) vtkWarningMacro("Setter of type \"object\" with a single \"param\" field is not supported");
		const onChanged = `_on${_capitalize(field.name)}Changed`;
		return (...args) => {
			let value;
			if (args.length > 1 && field.params?.length) value = field.params.reduce((acc, prop, idx) => Object.assign(acc, { [prop]: args[idx] }), {});
			else value = args[0];
			if (!DeepEqual(model[field.name], value)) {
				const previousValue = model[field.name];
				model[field.name] = value;
				model[onChanged]?.(publicAPI, model, value, previousValue);
				publicAPI.modified();
				return true;
			}
			return false;
		};
	}
};
function findSetter(field) {
	if (typeof field === "object") {
		const fn = objectSetterMap[field.type];
		if (fn) return (publicAPI, model) => fn(publicAPI, model, field);
		vtkErrorMacro(`No setter for field ${field}`);
		throw new TypeError("No setter for field");
	}
	return function getSetter(publicAPI, model) {
		const onChanged = `_on${_capitalize(field)}Changed`;
		return function setter(value) {
			if (model.deleted) {
				vtkErrorMacro("instance deleted - cannot call any method");
				return false;
			}
			if (model[field] !== value) {
				const previousValue = model[field.name];
				model[field] = value;
				model[onChanged]?.(publicAPI, model, value, previousValue);
				publicAPI.modified();
				return true;
			}
			return false;
		};
	};
}
function set(publicAPI, model, fields) {
	fields.forEach((field) => {
		if (typeof field === "object") publicAPI[`set${_capitalize(field.name)}`] = findSetter(field)(publicAPI, model);
		else publicAPI[`set${_capitalize(field)}`] = findSetter(field)(publicAPI, model);
	});
}
function setGet(publicAPI, model, fieldNames) {
	get(publicAPI, model, fieldNames);
	set(publicAPI, model, fieldNames);
}
function getArray(publicAPI, model, fieldNames) {
	fieldNames.forEach((field) => {
		publicAPI[`get${_capitalize(field)}`] = () => model[field] ? Array.from(model[field]) : model[field];
		publicAPI[`get${_capitalize(field)}ByReference`] = () => model[field];
	});
}
function setArray(publicAPI, model, fieldNames, size, defaultVal = void 0) {
	fieldNames.forEach((field) => {
		if (model[field] && size && model[field].length !== size) throw new RangeError(`Invalid initial number of values for array (${field})`);
		const onChanged = `_on${_capitalize(field)}Changed`;
		publicAPI[`set${_capitalize(field)}`] = (...args) => {
			if (model.deleted) {
				vtkErrorMacro("instance deleted - cannot call any method");
				return false;
			}
			let array = args;
			let changeDetected;
			let needCopy = false;
			if (array.length === 1 && (array[0] == null || array[0].length >= 0)) {
				array = array[0];
				needCopy = true;
			}
			if (array == null) changeDetected = model[field] !== array;
			else {
				if (size && array.length !== size) if (array.length < size && defaultVal !== void 0) {
					array = Array.from(array);
					needCopy = false;
					while (array.length < size) array.push(defaultVal);
				} else throw new RangeError(`Invalid number of values for array setter (${field})`);
				changeDetected = model[field] == null || model[field].length !== array.length;
				for (let i = 0; !changeDetected && i < array.length; ++i) changeDetected = model[field][i] !== array[i];
				if (changeDetected && needCopy) array = Array.from(array);
			}
			if (changeDetected) {
				const previousValue = model[field.name];
				model[field] = array;
				model[onChanged]?.(publicAPI, model, array, previousValue);
				publicAPI.modified();
			}
			return changeDetected;
		};
		publicAPI[`set${_capitalize(field)}From`] = (otherArray) => {
			const target = model[field];
			otherArray.forEach((v, i) => {
				target[i] = v;
			});
		};
	});
}
function setGetArray(publicAPI, model, fieldNames, size, defaultVal = void 0) {
	getArray(publicAPI, model, fieldNames);
	setArray(publicAPI, model, fieldNames, size, defaultVal);
}
function moveToProtected(publicAPI, model, fieldNames) {
	for (let i = 0; i < fieldNames.length; i++) {
		const fieldName = fieldNames[i];
		if (model[fieldName] !== void 0) {
			model[`_${fieldName}`] = model[fieldName];
			delete model[fieldName];
		}
	}
}
function algo(publicAPI, model, numberOfInputs, numberOfOutputs) {
	if (model.inputData) model.inputData = model.inputData.map(vtk);
	else model.inputData = [];
	if (model.inputConnection) model.inputConnection = model.inputConnection.map(vtk);
	else model.inputConnection = [];
	if (model.output) model.output = model.output.map(vtk);
	else model.output = [];
	if (model.inputArrayToProcess) model.inputArrayToProcess = model.inputArrayToProcess.map(vtk);
	else model.inputArrayToProcess = [];
	model.numberOfInputs = numberOfInputs;
	function setInputData(dataset, port = 0) {
		if (model.deleted) {
			vtkErrorMacro("instance deleted - cannot call any method");
			return;
		}
		if (port >= model.numberOfInputs) {
			vtkErrorMacro(`algorithm ${publicAPI.getClassName()} only has ${model.numberOfInputs} input ports. To add more input ports, use addInputData()`);
			return;
		}
		if (model.inputData[port] !== dataset || model.inputConnection[port]) {
			model.inputData[port] = dataset;
			model.inputConnection[port] = null;
			if (publicAPI.modified) publicAPI.modified();
		}
	}
	function getInputData(port = 0) {
		if (model.inputConnection[port]) model.inputData[port] = model.inputConnection[port]();
		return model.inputData[port];
	}
	function setInputConnection(outputPort, port = 0) {
		if (model.deleted) {
			vtkErrorMacro("instance deleted - cannot call any method");
			return;
		}
		if (port >= model.numberOfInputs) {
			let msg = `algorithm ${publicAPI.getClassName()} only has `;
			msg += `${model.numberOfInputs}`;
			msg += " input ports. To add more input ports, use addInputConnection()";
			vtkErrorMacro(msg);
			return;
		}
		model.inputData[port] = null;
		model.inputConnection[port] = outputPort;
	}
	function getInputConnection(port = 0) {
		return model.inputConnection[port];
	}
	function getPortToFill() {
		let portToFill = model.numberOfInputs;
		while (portToFill && !model.inputData[portToFill - 1] && !model.inputConnection[portToFill - 1]) portToFill--;
		if (portToFill === model.numberOfInputs) model.numberOfInputs++;
		return portToFill;
	}
	function addInputConnection(outputPort) {
		if (model.deleted) {
			vtkErrorMacro("instance deleted - cannot call any method");
			return;
		}
		setInputConnection(outputPort, getPortToFill());
	}
	function addInputData(dataset) {
		if (model.deleted) {
			vtkErrorMacro("instance deleted - cannot call any method");
			return;
		}
		setInputData(dataset, getPortToFill());
	}
	function getOutputData(port = 0) {
		if (model.deleted) {
			vtkErrorMacro("instance deleted - cannot call any method");
			return null;
		}
		if (publicAPI.shouldUpdate()) publicAPI.update();
		return model.output[port];
	}
	publicAPI.shouldUpdate = () => {
		const localMTime = publicAPI.getMTime();
		let minOutputMTime = Infinity;
		let count = numberOfOutputs;
		while (count--) {
			if (!model.output[count] || model.output[count].isDeleted()) return true;
			const mt = model.output[count].getMTime();
			if (mt < localMTime) return true;
			if (mt < minOutputMTime) minOutputMTime = mt;
		}
		count = model.numberOfInputs;
		while (count--) if (model.inputConnection[count]?.filter.shouldUpdate() || publicAPI.getInputData(count)?.getMTime() > minOutputMTime) return true;
		return false;
	};
	function getOutputPort(port = 0) {
		const outputPortAccess = () => getOutputData(port);
		outputPortAccess.filter = publicAPI;
		return outputPortAccess;
	}
	if (model.numberOfInputs) {
		let count = model.numberOfInputs;
		while (count--) {
			model.inputData.push(null);
			model.inputConnection.push(null);
		}
		publicAPI.setInputData = setInputData;
		publicAPI.setInputConnection = setInputConnection;
		publicAPI.addInputData = addInputData;
		publicAPI.addInputConnection = addInputConnection;
		publicAPI.getInputData = getInputData;
		publicAPI.getInputConnection = getInputConnection;
	}
	if (numberOfOutputs) {
		publicAPI.getOutputData = getOutputData;
		publicAPI.getOutputPort = getOutputPort;
	}
	publicAPI.update = () => {
		const ins = [];
		if (model.numberOfInputs) {
			let count = 0;
			while (count < model.numberOfInputs) {
				ins[count] = publicAPI.getInputData(count);
				count++;
			}
		}
		if (publicAPI.requestData && !publicAPI.isDeleted() && publicAPI.shouldUpdate()) publicAPI.requestData(ins, model.output);
	};
	publicAPI.getNumberOfInputPorts = () => model.numberOfInputs;
	publicAPI.getNumberOfOutputPorts = () => numberOfOutputs || model.output.length;
	publicAPI.getInputArrayToProcess = (inputPort) => {
		const arrayDesc = model.inputArrayToProcess[inputPort];
		const ds = model.inputData[inputPort];
		if (arrayDesc && ds) return ds[`get${arrayDesc.fieldAssociation}`]().getArray(arrayDesc.arrayName);
		return null;
	};
	publicAPI.setInputArrayToProcess = (inputPort, arrayName, fieldAssociation, attributeType = "Scalars") => {
		while (model.inputArrayToProcess.length < inputPort) model.inputArrayToProcess.push(null);
		model.inputArrayToProcess[inputPort] = {
			arrayName,
			fieldAssociation,
			attributeType
		};
	};
}
var EVENT_ABORT = Symbol("Event abort");
function event(publicAPI, model, eventName) {
	const callbacks = [];
	const previousDelete = publicAPI.delete;
	let curCallbackID = 1;
	function off(callbackID) {
		for (let i = 0; i < callbacks.length; ++i) {
			const [cbID] = callbacks[i];
			if (cbID === callbackID) {
				callbacks.splice(i, 1);
				return;
			}
		}
	}
	function on(callbackID) {
		function unsubscribe() {
			off(callbackID);
		}
		return Object.freeze({ unsubscribe });
	}
	function invoke() {
		if (model.deleted) {
			vtkErrorMacro("instance deleted - cannot call any method");
			return;
		}
		const currentCallbacks = callbacks.slice();
		for (let index = 0; index < currentCallbacks.length; ++index) {
			const [, cb, priority] = currentCallbacks[index];
			if (!cb) continue;
			if (priority < 0) setTimeout(() => cb.apply(publicAPI, arguments), 1 - priority);
			else if (cb.apply(publicAPI, arguments) === EVENT_ABORT) break;
		}
	}
	publicAPI[`invoke${_capitalize(eventName)}`] = invoke;
	publicAPI[`on${_capitalize(eventName)}`] = (callback, priority = 0) => {
		if (!callback.apply) {
			console.error(`Invalid callback for event ${eventName}`);
			return null;
		}
		if (model.deleted) {
			vtkErrorMacro("instance deleted - cannot call any method");
			return null;
		}
		const callbackID = curCallbackID++;
		callbacks.push([
			callbackID,
			callback,
			priority
		]);
		callbacks.sort((cb1, cb2) => cb2[2] - cb1[2]);
		return on(callbackID);
	};
	publicAPI.delete = () => {
		previousDelete();
		callbacks.forEach(([cbID]) => off(cbID));
	};
}
function newInstance(extend, className) {
	const constructor = (initialValues = {}) => {
		const model = {};
		const publicAPI = {};
		extend(publicAPI, model, initialValues);
		return Object.freeze(publicAPI);
	};
	if (className) vtk.register(className, constructor);
	return constructor;
}
function chain(...fn) {
	return (...args) => fn.filter((i) => !!i).map((i) => i(...args));
}
function isVtkObject(instance) {
	return instance && instance.isA && instance.isA("vtkObject");
}
function traverseInstanceTree(instance, extractFunction, accumulator = [], visitedInstances = []) {
	if (isVtkObject(instance)) {
		if (visitedInstances.indexOf(instance) >= 0) return accumulator;
		visitedInstances.push(instance);
		const result = extractFunction(instance);
		if (result !== void 0) accumulator.push(result);
		const model = instance.get();
		Object.keys(model).forEach((key) => {
			const modelObj = model[key];
			if (Array.isArray(modelObj)) modelObj.forEach((subObj) => {
				traverseInstanceTree(subObj, extractFunction, accumulator, visitedInstances);
			});
			else traverseInstanceTree(modelObj, extractFunction, accumulator, visitedInstances);
		});
	}
	return accumulator;
}
function debounce(func, wait, immediate) {
	let timeout;
	const debounced = (...args) => {
		const context = this;
		const later = () => {
			timeout = null;
			if (!immediate) func.apply(context, args);
		};
		const callNow = immediate && !timeout;
		clearTimeout(timeout);
		timeout = setTimeout(later, wait);
		if (callNow) func.apply(context, args);
	};
	debounced.cancel = () => clearTimeout(timeout);
	return debounced;
}
function throttle(callback, delay) {
	let isThrottled = false;
	let argsToUse = null;
	function next() {
		isThrottled = false;
		if (argsToUse !== null) {
			wrapper(...argsToUse);
			argsToUse = null;
		}
	}
	function wrapper(...args) {
		if (isThrottled) {
			argsToUse = args;
			return;
		}
		isThrottled = true;
		callback(...args);
		setTimeout(next, delay);
	}
	return wrapper;
}
function keystore(publicAPI, model, initialKeystore = {}) {
	model.keystore = Object.assign(model.keystore || {}, initialKeystore);
	publicAPI.setKey = (key, value) => {
		model.keystore[key] = value;
	};
	publicAPI.getKey = (key) => model.keystore[key];
	publicAPI.getAllKeys = () => Object.keys(model.keystore);
	publicAPI.deleteKey = (key) => delete model.keystore[key];
	publicAPI.clearKeystore = () => publicAPI.getAllKeys().forEach((key) => delete model.keystore[key]);
}
var nextProxyId = 1;
var ROOT_GROUP_NAME = "__root__";
function proxy(publicAPI, model) {
	keystore(publicAPI, model);
	const parentDelete = publicAPI.delete;
	model.proxyId = `${nextProxyId++}`;
	model.ui = JSON.parse(JSON.stringify(model.ui || []));
	get(publicAPI, model, [
		"proxyId",
		"proxyGroup",
		"proxyName"
	]);
	setGet(publicAPI, model, ["proxyManager"]);
	const propertyMap = {};
	const groupChildrenNames = {};
	function registerProperties(descriptionList, currentGroupName) {
		if (!groupChildrenNames[currentGroupName]) groupChildrenNames[currentGroupName] = [];
		const childrenNames = groupChildrenNames[currentGroupName];
		for (let i = 0; i < descriptionList.length; i++) {
			childrenNames.push(descriptionList[i].name);
			propertyMap[descriptionList[i].name] = descriptionList[i];
			if (descriptionList[i].children && descriptionList[i].children.length) registerProperties(descriptionList[i].children, descriptionList[i].name);
		}
	}
	registerProperties(model.ui, ROOT_GROUP_NAME);
	publicAPI.updateUI = (ui) => {
		model.ui = JSON.parse(JSON.stringify(ui || []));
		Object.keys(propertyMap).forEach((k) => delete propertyMap[k]);
		Object.keys(groupChildrenNames).forEach((k) => delete groupChildrenNames[k]);
		registerProperties(model.ui, ROOT_GROUP_NAME);
		publicAPI.modified();
	};
	function listProxyProperties(gName = ROOT_GROUP_NAME) {
		return groupChildrenNames[gName];
	}
	publicAPI.updateProxyProperty = (propertyName, propUI) => {
		const prop = propertyMap[propertyName];
		if (prop) Object.assign(prop, propUI);
		else propertyMap[propertyName] = { ...propUI };
	};
	publicAPI.activate = () => {
		if (model.proxyManager) {
			const setActiveMethod = `setActive${_capitalize(publicAPI.getProxyGroup().slice(0, -1))}`;
			if (model.proxyManager[setActiveMethod]) model.proxyManager[setActiveMethod](publicAPI);
		}
	};
	model.propertyLinkSubscribers = {};
	publicAPI.registerPropertyLinkForGC = (otherLink, type) => {
		if (!(type in model.propertyLinkSubscribers)) model.propertyLinkSubscribers[type] = [];
		model.propertyLinkSubscribers[type].push(otherLink);
	};
	publicAPI.gcPropertyLinks = (type) => {
		const subscribers = model.propertyLinkSubscribers[type] || [];
		while (subscribers.length) subscribers.pop().unbind(publicAPI);
	};
	model.propertyLinkMap = {};
	publicAPI.getPropertyLink = (id, persistent = false) => {
		if (model.propertyLinkMap[id]) return model.propertyLinkMap[id];
		let value = null;
		const links = [];
		let count = 0;
		let updateInProgress = false;
		function update(source, force = false) {
			if (updateInProgress) return null;
			const needUpdate = [];
			let sourceLink = null;
			count = links.length;
			while (count--) {
				const link = links[count];
				if (link.instance === source) sourceLink = link;
				else needUpdate.push(link);
			}
			if (!sourceLink) return null;
			const newValue = sourceLink.instance[`get${_capitalize(sourceLink.propertyName)}`]();
			if (!shallowEquals(newValue, value) || force) {
				value = newValue;
				updateInProgress = true;
				while (needUpdate.length) {
					const linkToUpdate = needUpdate.pop();
					linkToUpdate.instance.set({ [linkToUpdate.propertyName]: value });
				}
				updateInProgress = false;
			}
			if (model.propertyLinkMap[id].persistent) model.propertyLinkMap[id].value = newValue;
			return newValue;
		}
		function unbind(instance, propertyName) {
			const indexToDelete = [];
			count = links.length;
			while (count--) {
				const link = links[count];
				if (link.instance === instance && (link.propertyName === propertyName || propertyName === void 0)) {
					link.subscription.unsubscribe();
					indexToDelete.push(count);
				}
			}
			while (indexToDelete.length) links.splice(indexToDelete.pop(), 1);
		}
		function bind(instance, propertyName, updateMe = false) {
			const subscription = instance.onModified(update);
			const other = links[0];
			links.push({
				instance,
				propertyName,
				subscription
			});
			if (updateMe) {
				if (model.propertyLinkMap[id].persistent && model.propertyLinkMap[id].value !== void 0) instance.set({ [propertyName]: model.propertyLinkMap[id].value });
				else if (other) update(other.instance, true);
			}
			return { unsubscribe: () => unbind(instance, propertyName) };
		}
		function unsubscribe() {
			while (links.length) links.pop().subscription.unsubscribe();
		}
		const linkHandler = {
			bind,
			unbind,
			unsubscribe,
			persistent
		};
		model.propertyLinkMap[id] = linkHandler;
		return linkHandler;
	};
	function getProperties(groupName = ROOT_GROUP_NAME) {
		const values = [];
		const id = model.proxyId;
		const propertyNames = listProxyProperties(groupName) || [];
		for (let i = 0; i < propertyNames.length; i++) {
			const name = propertyNames[i];
			const method = publicAPI[`get${_capitalize(name)}`];
			const prop = {
				id,
				name,
				value: method ? method() : void 0
			};
			const children = getProperties(name);
			if (children.length) prop.children = children;
			values.push(prop);
		}
		return values;
	}
	publicAPI.listPropertyNames = () => getProperties().map((p) => p.name);
	publicAPI.getPropertyByName = (name) => getProperties().find((p) => p.name === name);
	publicAPI.getPropertyDomainByName = (name) => (propertyMap[name] || {}).domain;
	publicAPI.getProxySection = () => ({
		id: model.proxyId,
		name: model.proxyGroup,
		ui: model.ui,
		properties: getProperties()
	});
	publicAPI.delete = () => {
		const list = Object.keys(model.propertyLinkMap);
		let count = list.length;
		while (count--) model.propertyLinkMap[list[count]].unsubscribe();
		Object.keys(model.propertyLinkSubscribers).forEach(publicAPI.gcPropertyLinks);
		parentDelete();
	};
	publicAPI.getState = () => null;
	function registerLinks() {
		if (model.links) for (let i = 0; i < model.links.length; i++) {
			const { link, property, persistent, updateOnBind, type } = model.links[i];
			if (type === "application") {
				const sLink = model.proxyManager.getPropertyLink(link, persistent);
				publicAPI.registerPropertyLinkForGC(sLink, "application");
				sLink.bind(publicAPI, property, updateOnBind);
			}
		}
	}
	setImmediateVTK(registerLinks);
}
function proxyPropertyMapping(publicAPI, model, map) {
	const parentDelete = publicAPI.delete;
	const subscriptions = [];
	const propertyNames = Object.keys(map);
	let count = propertyNames.length;
	while (count--) {
		const propertyName = propertyNames[count];
		const { modelKey, property, modified = true } = map[propertyName];
		const methodSrc = _capitalize(property);
		const methodDst = _capitalize(propertyName);
		publicAPI[`get${methodDst}`] = model[modelKey][`get${methodSrc}`];
		publicAPI[`set${methodDst}`] = model[modelKey][`set${methodSrc}`];
		if (modified) subscriptions.push(model[modelKey].onModified(publicAPI.modified));
	}
	publicAPI.delete = () => {
		while (subscriptions.length) subscriptions.pop().unsubscribe();
		parentDelete();
	};
}
function proxyPropertyState(publicAPI, model, state = {}, defaults = {}) {
	model.this = publicAPI;
	function applyState(map) {
		const modelKeys = Object.keys(map);
		let count = modelKeys.length;
		while (count--) {
			const modelKey = modelKeys[count];
			model[modelKey].set(map[modelKey]);
		}
	}
	const modelKeys = Object.keys(defaults);
	let count = modelKeys.length;
	while (count--) {
		const key = modelKeys[count];
		model[key] = defaults[key];
		const mapping = state[key];
		publicAPI[`set${_capitalize(key)}`] = (value) => {
			if (value !== model[key]) {
				model[key] = value;
				const propValues = mapping[value];
				applyState(propValues);
				publicAPI.modified();
			}
		};
	}
	if (modelKeys.length) get(publicAPI, model, modelKeys);
}
var PIXEL_STEP = 10;
var LINE_HEIGHT = 40;
var PAGE_HEIGHT = 800;
function normalizeWheel(wheelEvent) {
	let sX = 0;
	let sY = 0;
	let pX = 0;
	let pY = 0;
	if ("detail" in wheelEvent) sY = wheelEvent.detail;
	if ("wheelDelta" in wheelEvent) sY = -wheelEvent.wheelDelta / 120;
	if ("wheelDeltaY" in wheelEvent) sY = -wheelEvent.wheelDeltaY / 120;
	if ("wheelDeltaX" in wheelEvent) sX = -wheelEvent.wheelDeltaX / 120;
	if ("axis" in wheelEvent && wheelEvent.axis === wheelEvent.HORIZONTAL_AXIS) {
		sX = sY;
		sY = 0;
	}
	pX = sX * PIXEL_STEP;
	pY = sY * PIXEL_STEP;
	if ("deltaY" in wheelEvent) pY = wheelEvent.deltaY;
	if ("deltaX" in wheelEvent) pX = wheelEvent.deltaX;
	if ((pX || pY) && wheelEvent.deltaMode) if (wheelEvent.deltaMode === 1) {
		pX *= LINE_HEIGHT;
		pY *= LINE_HEIGHT;
	} else {
		pX *= PAGE_HEIGHT;
		pY *= PAGE_HEIGHT;
	}
	if (pX && !sX) sX = pX < 1 ? -1 : 1;
	if (pY && !sY) sY = pY < 1 ? -1 : 1;
	return {
		spinX: sX,
		spinY: sY || sX,
		pixelX: pX,
		pixelY: pY || pX
	};
}
var macros_default = {
	algo,
	capitalize,
	chain,
	debounce,
	enumToString,
	event,
	EVENT_ABORT,
	formatBytesToProperUnit,
	formatNumbersWithThousandSeparator,
	get,
	getArray,
	getCurrentGlobalMTime,
	isVtkObject,
	keystore,
	measurePromiseExecution,
	moveToProtected,
	newInstance,
	newTypedArray,
	newTypedArrayFrom,
	normalizeWheel,
	obj,
	proxy,
	proxyPropertyMapping,
	proxyPropertyState,
	safeArrays,
	set,
	setArray,
	setGet,
	setGetArray,
	setImmediate: setImmediateVTK,
	setLoggerFunction,
	throttle,
	traverseInstanceTree,
	TYPED_ARRAYS,
	uncapitalize,
	VOID,
	vtkDebugMacro,
	vtkErrorMacro,
	vtkInfoMacro,
	vtkLogMacro,
	vtkOnceErrorMacro,
	vtkWarningMacro,
	objectSetterMap,
	requiredParam
};
//#endregion
export { EVENT_ABORT, TYPED_ARRAYS, VOID, _capitalize, algo, capitalize, chain, debounce, macros_default as default, event, formatBytesToProperUnit, formatNumbersWithThousandSeparator, get, getArray, isVtkObject, keystore, macros_exports, measurePromiseExecution, moveToProtected, newInstance, newTypedArray, newTypedArrayFrom, normalizeWheel, obj, proxy, proxyPropertyMapping, proxyPropertyState, requiredParam, set, setArray, setGet, setGetArray, setImmediateVTK, setLoggerFunction, throttle, traverseInstanceTree, uncapitalize, vtkDebugMacro, vtkErrorMacro, vtkInfoMacro, vtkLogMacro, vtkOnceErrorMacro, vtkWarningMacro };
