// @kitware/vtk.js/Widgets/Core/AbstractWidgetFactory.js
import macros_default from "../../macros.js";
import AbstractWidget_default from "./AbstractWidget.js";
import { extractRenderingComponents } from "./WidgetManager.js";
//#region Sources/Widgets/Core/AbstractWidgetFactory/index.js
function NoOp() {}
function vtkAbstractWidgetFactory(publicAPI, model) {
	model.classHierarchy.push("vtkAbstractWidgetFactory");
	const viewToWidget = {};
	publicAPI.getWidgetForView = ({ viewId, renderer, viewType, initialValues }) => {
		if (!viewToWidget[viewId]) {
			if (!renderer) return null;
			const { interactor, apiSpecificRenderWindow, camera } = extractRenderingComponents(renderer);
			const widgetModel = {};
			const widgetPublicAPI = {};
			macros_default.obj(widgetPublicAPI, widgetModel);
			Object.assign(widgetPublicAPI, { onWidgetChange: publicAPI.onWidgetChange });
			Object.assign(widgetModel, {
				widgetState: model.widgetState,
				manipulator: model.manipulator,
				viewType,
				renderer,
				camera,
				apiSpecificRenderWindow,
				factory: publicAPI
			});
			macros_default.moveToProtected(widgetPublicAPI, widgetModel, [
				"renderer",
				"camera",
				"apiSpecificRenderWindow",
				"factory"
			]);
			macros_default.get(widgetPublicAPI, widgetModel, ["viewType"]);
			macros_default.safeArrays(widgetModel);
			AbstractWidget_default.extend(widgetPublicAPI, widgetModel, initialValues);
			const widgetInitialValues = initialValues;
			widgetModel.representations = publicAPI.getRepresentationsForViewType(viewType).map(({ builder, labels, initialValues }) => builder.newInstance({
				_parentProp: widgetPublicAPI,
				labels,
				...initialValues,
				...widgetInitialValues
			}));
			widgetModel.representations.forEach((r) => {
				r.setInputData(widgetModel.widgetState);
				r.getActors().forEach((actor) => {
					widgetModel.actorToRepresentationMap.set(actor, r);
				});
			});
			model.behavior(widgetPublicAPI, widgetModel);
			["coincidentTopologyParameters", ...model.methodsToLink || []].forEach((methodName) => {
				const set = `set${macros_default.capitalize(methodName)}`;
				const get = `get${macros_default.capitalize(methodName)}`;
				const methods = {
					[methodName]: [],
					[set]: [],
					[get]: []
				};
				widgetModel.representations.forEach((representation) => {
					if (representation[methodName]) methods[methodName].push(representation[methodName]);
					if (representation[set]) methods[set].push(representation[set]);
					if (representation[get]) methods[get].push(representation[get]);
				});
				Object.keys(methods).forEach((name) => {
					const calls = methods[name];
					if (calls.length === 1) widgetPublicAPI[name] = calls[0];
					else if (calls.length > 1) widgetPublicAPI[name] = macros_default.chain(...calls);
				});
			});
			widgetPublicAPI.delete = macros_default.chain(() => {
				delete viewToWidget[viewId];
			}, widgetPublicAPI.delete);
			widgetPublicAPI.setInteractor(interactor);
			const viewWidget = Object.freeze(widgetPublicAPI);
			viewToWidget[viewId] = viewWidget;
			return viewWidget;
		}
		return viewToWidget[viewId];
	};
	publicAPI.getViewIds = () => Object.keys(viewToWidget);
	publicAPI.getViewWidgets = () => Object.values(viewToWidget);
	publicAPI.setVisibility = (value) => {
		const viewIds = Object.keys(viewToWidget);
		for (let i = 0; i < viewIds.length; i++) viewToWidget[viewIds[i]].setVisibility(value);
	};
	publicAPI.setPickable = (value) => {
		const viewIds = Object.keys(viewToWidget);
		for (let i = 0; i < viewIds.length; i++) viewToWidget[viewIds[i]].setPickable(value);
	};
	publicAPI.setDragable = (value) => {
		const viewIds = Object.keys(viewToWidget);
		for (let i = 0; i < viewIds.length; i++) viewToWidget[viewIds[i]].setDragable(value);
	};
	publicAPI.setContextVisibility = (value) => {
		const viewIds = Object.keys(viewToWidget);
		for (let i = 0; i < viewIds.length; i++) viewToWidget[viewIds[i]].setContextVisibility(value);
	};
	publicAPI.setHandleVisibility = (value) => {
		const viewIds = Object.keys(viewToWidget);
		for (let i = 0; i < viewIds.length; i++) viewToWidget[viewIds[i]].setHandleVisibility(value);
	};
	publicAPI.placeWidget = (bounds) => model.widgetState.placeWidget(bounds);
	publicAPI.getPlaceFactor = () => model.widgetState.getPlaceFactor();
	publicAPI.setPlaceFactor = (factor) => model.widgetState.setPlaceFactor(factor);
	let unsubscribe = NoOp;
	publicAPI.delete = macros_default.chain(publicAPI.delete, () => unsubscribe());
	if (model.widgetState) unsubscribe = model.widgetState.onModified(() => publicAPI.invokeWidgetChange(model.widgetState)).unsubscribe;
}
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.get(publicAPI, model, ["widgetState"]);
	macros_default.event(publicAPI, model, "WidgetChange");
	vtkAbstractWidgetFactory(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkAbstractWidget");
var AbstractWidgetFactory_default = {
	newInstance,
	extend
};
//#endregion
export { AbstractWidgetFactory_default as default, extend, newInstance };
