// @kitware/vtk.js/Widgets/Core/AbstractWidget.js
import macros_default from "../../macros.js";
import InteractorObserver_default from "../../Rendering/Core/InteractorObserver.js";
import Prop_default from "../../Rendering/Core/Prop.js";
import { RenderingTypes } from "./WidgetManager/Constants.js";
import { WIDGET_PRIORITY } from "./AbstractWidget/Constants.js";
//#region Sources/Widgets/Core/AbstractWidget/index.js
function vtkAbstractWidget(publicAPI, model) {
	model.classHierarchy.push("vtkAbstractWidget");
	model.actorToRepresentationMap = /* @__PURE__ */ new WeakMap();
	publicAPI.getBounds = model.widgetState.getBounds;
	publicAPI.getNestedProps = () => model.representations;
	publicAPI.activateHandle = ({ selectedState, representation }) => {
		model.widgetState.activateOnly(selectedState);
		model.activeState = selectedState;
		if (selectedState && selectedState.updateManipulator) selectedState.updateManipulator();
		publicAPI.invokeActivateHandle({
			selectedState,
			representation
		});
		if (publicAPI.updateCursor) publicAPI.updateCursor();
	};
	publicAPI.deactivateAllHandles = () => {
		model.widgetState.deactivate();
	};
	publicAPI.hasActor = (actor) => model.actorToRepresentationMap.has(actor);
	publicAPI.grabFocus = () => {
		model.hasFocus = true;
	};
	publicAPI.loseFocus = () => {
		model.hasFocus = false;
	};
	publicAPI.hasFocus = () => model.hasFocus;
	publicAPI.placeWidget = (bounds) => model.widgetState.placeWidget(bounds);
	publicAPI.getPlaceFactor = () => model.widgetState.getPlaceFactor();
	publicAPI.setPlaceFactor = (factor) => model.widgetState.setPlaceFactor(factor);
	publicAPI.getRepresentationFromActor = (actor) => model.actorToRepresentationMap.get(actor);
	publicAPI.updateRepresentationForRender = (renderingType = RenderingTypes.FRONT_BUFFER) => {
		for (let i = 0; i < model.representations.length; i++) model.representations[i].updateActorVisibility(renderingType, model.contextVisibility, model.handleVisibility);
	};
	publicAPI.getViewWidgets = () => model._factory.getViewWidgets();
	publicAPI.setPriority(WIDGET_PRIORITY);
}
var DEFAULT_VALUES = {
	contextVisibility: true,
	handleVisibility: true,
	hasFocus: false
};
/**
* @param {*} publicAPI public methods to populate
* @param {*} model internal values to populate
* @param {object} initialValues Contains at least
*   {viewType, _renderer, _camera, _openGLRenderWindow, _factory}
*/
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	Prop_default.extend(publicAPI, model, initialValues);
	InteractorObserver_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, [
		"contextVisibility",
		"handleVisibility",
		"_widgetManager"
	]);
	macros_default.get(publicAPI, model, [
		"representations",
		"widgetState",
		"activeState"
	]);
	macros_default.moveToProtected(publicAPI, model, ["widgetManager"]);
	macros_default.event(publicAPI, model, "ActivateHandle");
	vtkAbstractWidget(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkAbstractWidget");
var AbstractWidget_default = {
	newInstance,
	extend
};
//#endregion
export { AbstractWidget_default as default, extend, newInstance };
