// @kitware/vtk.js/Widgets/Core/StateBuilder.js
import macros_default from "../../macros.js";
import WidgetState_default from "./WidgetState.js";
import boundsMixin_default from "./StateBuilder/boundsMixin.js";
import colorMixin_default from "./StateBuilder/colorMixin.js";
import color3Mixin_default from "./StateBuilder/color3Mixin.js";
import cornerMixin_default from "./StateBuilder/cornerMixin.js";
import directionMixin_default from "./StateBuilder/directionMixin.js";
import manipulatorMixin_default from "./StateBuilder/manipulatorMixin.js";
import nameMixin_default from "./StateBuilder/nameMixin.js";
import orientationMixin_default from "./StateBuilder/orientationMixin.js";
import originMixin_default from "./StateBuilder/originMixin.js";
import scale1Mixin_default from "./StateBuilder/scale1Mixin.js";
import scale3Mixin_default from "./StateBuilder/scale3Mixin.js";
import textMixin_default from "./StateBuilder/textMixin.js";
import visibleMixin_default from "./StateBuilder/visibleMixin.js";
import shapeMixin_default from "./StateBuilder/shapeMixin.js";
//#region Sources/Widgets/Core/StateBuilder/index.js
var { vtkErrorMacro } = macros_default;
var MIXINS = {
	bounds: boundsMixin_default,
	color: colorMixin_default,
	color3: color3Mixin_default,
	corner: cornerMixin_default,
	direction: directionMixin_default,
	manipulator: manipulatorMixin_default,
	name: nameMixin_default,
	orientation: orientationMixin_default,
	origin: originMixin_default,
	scale1: scale1Mixin_default,
	scale3: scale3Mixin_default,
	text: textMixin_default,
	visible: visibleMixin_default,
	shape: shapeMixin_default
};
function newInstance(mixins, initialValues, publicAPI = {}, model = {}, skipWidgetState = false) {
	if (!skipWidgetState) WidgetState_default.extend(publicAPI, model, initialValues);
	for (let i = 0; i < mixins.length; i++) {
		const mixin = MIXINS[mixins[i]];
		if (mixin) mixin.extend(publicAPI, model, initialValues);
		else vtkErrorMacro("Invalid mixin name:", mixins[i]);
	}
	macros_default.safeArrays(model);
	return Object.freeze(publicAPI);
}
var Builder = class {
	constructor() {
		this.publicAPI = {};
		this.model = {};
		WidgetState_default.extend(this.publicAPI, this.model);
		boundsMixin_default.extend(this.publicAPI, this.model);
	}
	addDynamicMixinState({ labels, mixins, name, initialValues }) {
		const listName = `${name}List`;
		this.model[listName] = [];
		this.publicAPI[`add${macros_default.capitalize(name)}`] = (values) => {
			const instance = newInstance(mixins, {
				...initialValues,
				...values
			});
			this.publicAPI.bindState(instance, labels);
			this.model[listName].push(instance);
			this.publicAPI.modified();
			return instance;
		};
		this.publicAPI[`remove${macros_default.capitalize(name)}`] = (instanceOrIndex) => {
			let removeIndex = this.model[listName].indexOf(instanceOrIndex);
			if (removeIndex === -1 && instanceOrIndex < this.model[listName].length) removeIndex = instanceOrIndex;
			const instance = this.model[listName][removeIndex];
			if (instance) this.publicAPI.unbindState(instance);
			this.model[listName].splice(removeIndex, 1);
			this.publicAPI.modified();
		};
		this.publicAPI[`get${macros_default.capitalize(name)}List`] = () => this.model[listName].slice();
		this.publicAPI[`clear${macros_default.capitalize(name)}List`] = () => {
			while (this.model[listName].length) {
				const instance = this.model[listName].pop();
				if (instance) this.publicAPI.unbindState(instance);
			}
			this.publicAPI.modified();
		};
		return this;
	}
	addStateFromMixin({ labels, mixins, name, initialValues }) {
		const instance = newInstance(mixins, initialValues);
		this.model[name] = instance;
		this.publicAPI.bindState(instance, labels);
		macros_default.setGet(this.publicAPI, this.model, [name]);
		return this;
	}
	addStateFromInstance({ labels, name, instance }) {
		this.model[name] = instance;
		this.publicAPI.bindState(instance, labels);
		macros_default.setGet(this.publicAPI, this.model, [name]);
		return this;
	}
	addField({ name, initialValue }) {
		if (Array.isArray(initialValue)) macros_default.setGetArray(this.publicAPI, this.model, [name], initialValue.length);
		else macros_default.setGet(this.publicAPI, this.model, [name]);
		this.model[name] = initialValue;
		return this;
	}
	build(...mixins) {
		return newInstance(mixins, {}, this.publicAPI, this.model, true);
	}
};
function createBuilder() {
	return new Builder();
}
var StateBuilder_default = { createBuilder };
//#endregion
export { createBuilder, StateBuilder_default as default };
