// @kitware/vtk.js/Filters/General/ClipClosedSurface.js
import macros_default from "../../macros.js";
import { VtkDataTypes } from "../../Common/Core/DataArray/Constants.js";
import { distance2BetweenPoints, dot } from "../../Common/Core/Math.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import CellArray_default from "../../Common/Core/CellArray.js";
import DataSetAttributes_default from "../../Common/DataModel/DataSetAttributes.js";
import Points_default from "../../Common/Core/Points.js";
import EdgeLocator_default from "../../Common/DataModel/EdgeLocator.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
import Constants_default from "./ClipClosedSurface/Constants.js";
import ContourTriangulator_default from "./ContourTriangulator.js";
//#region Sources/Filters/General/ClipClosedSurface/index.js
var { vtkErrorMacro, capitalize } = macros_default;
var { ScalarMode } = Constants_default;
function vtkClipClosedSurface(publicAPI, model) {
	model.classHierarchy.push("vtkClipClosedSurface");
	publicAPI.getMTime = () => model.clippingPlanes.reduce((a, b) => b.getMTime() > a ? b.getMTime() : a, model.mtime);
	/**
	* Take three colors as doubles, and convert to unsigned char.
	*
	* @param {Number} color1
	* @param {Number} color2
	* @param {Number} color3
	* @param {Number[3][3]} colors
	*/
	function createColorValues(color1, color2, color3, colors) {
		const dcolors = [
			color1,
			color2,
			color3
		];
		const clamp = (n, min, max) => Math.min(Math.max(n, min), max);
		for (let i = 0; i < 3; i++) for (let j = 0; j < 3; j++) colors[i][j] = Math.round(clamp(dcolors[i][j], 0, 1) * 255);
	}
	/**
	* Point interpolation for clipping and contouring, given the scalar
	* values (v0, v1) for the two endpoints (p0, p1).  The use of this
	* function guarantees perfect consistency in the results.
	*
	* @param {vtkPoints} points
	* @param {vtkDataArray} pointData
	* @param {CCSEdgeLocator} locator
	* @param {Number} tol
	* @param {Number} i0
	* @param {Number} i1
	* @param {Number} v0
	* @param {Number} v1
	* @param {Number} i
	* @returns {Number}
	*/
	function interpolateEdge(points, pointData, locator, tol, i0, i1, v0, v1) {
		if (v1 > 0) {
			[i0, i1] = [i1, i0];
			[v0, v1] = [v1, v0];
		}
		const edge = locator.insertUniqueEdge(i0, i1);
		if (edge.value != null) return edge.value;
		const p0 = points.getPoint(i0);
		const p1 = points.getPoint(i1);
		const s = 1 - v0 / (v0 - v1);
		const t = 1 - s;
		const p = [
			s * p0[0] + t * p1[0],
			s * p0[1] + t * p1[1],
			s * p0[2] + t * p1[2]
		];
		const tol2 = tol * tol;
		if (distance2BetweenPoints(p, p0) < tol2) {
			edge.value = i0;
			return i0;
		}
		if (distance2BetweenPoints(p, p1) < tol2) {
			edge.value = i1;
			return i1;
		}
		edge.value = points.insertNextTuple(p);
		pointData.interpolateData(pointData, i0, i1, edge.value, t);
		return edge.value;
	}
	/**
	* Method for clipping lines and copying the scalar data.
	*
	* @param {vtkPoints} points
	* @param {vtkDataArray} pointScalars
	* @param {vtkDataSetAttributesk} pointData
	* @param {vtkEdgeLocator} edgeLocator
	* @param {vtkCellArray} inputLines
	* @param {vtkCellArray} outputLines
	* @param {vtkDataSetAttributes} inLineData
	* @param {vtkDataSetAttributes} outLineData
	*/
	function clipLines(points, pointScalars, pointData, edgeLocator, inputLines, outputLines, inLineData, outLineData) {
		let numPts;
		let i0;
		let i1;
		let v0;
		let v1;
		let c0;
		let c1;
		const linePts = [];
		const values = inputLines.getData();
		let cellId = 0;
		for (let i = 0; i < values.length; i += numPts + 1, cellId++) {
			numPts = values[i];
			i1 = values[i + 1];
			v1 = pointScalars.getData()[i1];
			c1 = v1 > 0;
			for (let j = 2; j <= numPts; j++) {
				i0 = i1;
				v0 = v1;
				c0 = c1;
				i1 = values[i + j];
				v1 = pointScalars.getData()[i1];
				c1 = v1 > 0;
				if (c0 || c1) {
					if (c0 ? !c1 : c1) linePts[c0 ? 1 : 0] = interpolateEdge(points, pointData, edgeLocator, model.tolerance, i0, i1, v0, v1);
					if (i0 !== i1) {
						linePts[0] = i0;
						linePts[1] = i1;
						const newCellId = outputLines.insertNextCell(linePts);
						outLineData.passData(inLineData, cellId, newCellId);
					}
				}
			}
		}
	}
	/**
	* Break polylines into individual lines, copying scalar values from
	* inputScalars starting at firstLineScalar.  If inputScalars is zero,
	* then scalars will be set to color.  If scalars is zero, then no
	* scalars will be generated.
	*
	* @param {vtkCellArray} inputLines
	* @param {vtkCellArray} outputLines
	* @param {vtkDataArray} inputScalars
	* @param {Number} firstLineScalar
	* @param {vtkDataArray} scalars
	* @param {Vector3} color
	*/
	function breakPolylines(inputLines, outputLines, inputScalars, firstLineScalar, scalars, color) {
		const cellColor = [...color];
		let cellId = 0;
		const values = inputLines.getData();
		let numPts;
		for (let i = 0; i < values.length; i += numPts + 1) {
			numPts = values[i];
			if (inputScalars) inputScalars.getTuple(firstLineScalar + cellId++, cellColor);
			for (let j = 1; j < numPts; j++) {
				outputLines.insertNextCell([values[i + j], values[i + j + 1]]);
				if (scalars) scalars.insertNextTuple(cellColor);
			}
		}
	}
	/**
	* Copy polygons and their associated scalars to a new array.
	* If inputScalars is set to zero, set polyScalars to color instead.
	* If polyScalars is set to zero, don't generate scalars.
	*
	* @param {vtkCellArray} inputPolys
	* @param {vtkCellArray} outputPolys
	* @param {vtkDataArray} inputScalars
	* @param {Number} firstPolyScalar
	* @param {vtkDataArray} polyScalars
	* @param {Vector3} color
	*/
	function copyPolygons(inputPolys, outputPolys, inputScalars, firstPolyScalar, polyScalars, color) {
		if (!inputPolys) return;
		outputPolys.deepCopy(inputPolys);
		if (polyScalars) {
			const scalarValue = [...color];
			const n = outputPolys.getNumberOfCells();
			polyScalars.insertTuple(n - 1, scalarValue);
			if (inputScalars) for (let i = 0; i < n; i++) {
				inputScalars.getTuple(i + firstPolyScalar, scalarValue);
				polyScalars.setTuple(i, scalarValue);
			}
			else for (let i = 0; i < n; i++) polyScalars.setTuple(i, scalarValue);
		}
	}
	function breakTriangleStrips(inputStrips, polys, inputScalars, firstStripScalar, polyScalars, color) {
		if (inputStrips.getNumberOfCells() === 0) return;
		const values = inputStrips.getData();
		let cellId = firstStripScalar;
		let numPts;
		for (let i = 0; i < values.length; i += numPts + 1, cellId++) {
			numPts = values[i];
			let p1 = values[i + 1];
			let p2 = values[i + 2];
			for (let j = 0; j < numPts - 2; j++) {
				const p3 = values[i + j + 3];
				if (j % 2) polys.insertNextCell([
					p2,
					p1,
					p3
				]);
				else polys.insertNextCell([
					p1,
					p2,
					p3
				]);
				p1 = p2;
				p2 = p3;
			}
			if (polyScalars) {
				const scalarValue = [...color];
				if (inputScalars) inputScalars.getTuple(cellId, scalarValue);
				const n = numPts - 3;
				const m = polyScalars.getNumberOfTuples();
				if (n >= 0) {
					polyScalars.insertTuple(m + n, scalarValue);
					for (let k = 0; k < n; k++) polyScalars.setTuple(m + k, scalarValue);
				}
			}
		}
	}
	/**
	* Given some closed contour lines, create a triangle mesh that
	* fills those lines.  The input lines must be single-segment lines,
	* not polylines.  The input lines do not have to be in order.
	* Only lines from firstLine to will be used.  Specify the normal
	* of the clip plane, which will be opposite the normals
	* of the polys that will be produced.  If outCD has scalars, then color
	* scalars will be added for each poly that is created.
	*
	* @param {vtkPolyData} polyData
	* @param {Number} firstLine
	* @param {Number} numLines
	* @param {vtkCellArray} outputPolys
	* @param {Vector3} normal
	*/
	function triangulateContours(polyData, firstLine, numLines, outputPolys, normal) {
		if (numLines <= 0) return;
		if (!ContourTriangulator_default.triangulateContours(polyData, firstLine, numLines, outputPolys, [
			-normal[0],
			-normal[1],
			-normal[2]
		]) && model.triangulationErrorDisplay) vtkErrorMacro("Triangulation failed, polyData may not be watertight.");
	}
	/**
	* Break polylines into individual lines, copying scalar values from
	* inputScalars starting at firstLineScalar. If inputScalars is zero,
	* then scalars will be set to color. If scalars is zero, then no
	* scalars will be generated.
	*
	* @param {Number[]} polygon
	* @param {vtkPoints} points
	* @param {vtkCellArray} triangles
	* @returns {Boolean}
	*/
	function triangulatePolygon(polygon, points, triangles) {
		return ContourTriangulator_default.triangulatePolygon(polygon, points, triangles);
	}
	/**
	* Clip and contour polys in one step, in order to guarantee
	* that the contour lines exactly match the new free edges of
	* the clipped polygons.  This exact correspondence is necessary
	* in order to guarantee that the surface remains closed.
	*
	* @param {vtkPoints} points
	* @param {vtkDataArray} pointScalars
	* @param {vtkDataSetAttributes} pointData
	* @param {vtkEdgeLocator} edgeLocator
	* @param {Number} triangulate
	* @param {vtkCellArray} inputPolys
	* @param {vtkCellArray} outputPolys
	* @param {vtkCellArray} outputLines
	* @param {vtkDataSetAttributes} inCellData
	* @param {vtkDataSetAttributes} outPolyData
	* @param {vtkDataSetAttributes} outLineData
	*/
	function clipAndContourPolys(points, pointScalars, pointData, edgeLocator, triangulate, inputPolys, outputPolys, outputLines, inCellData, outPolyData, outLineData) {
		const idList = model._idList;
		let polyMax = Number.MAX_VALUE;
		if (triangulate) {
			if (triangulate < 4) polyMax = 3;
			else if (triangulate === 4) polyMax = 4;
		}
		let triangulationFailure = false;
		const values = inputPolys.getData();
		const linePts = [];
		let cellId = 0;
		let numPts;
		for (let i = 0; i < values.length; i += numPts + 1, cellId++) {
			numPts = values[i];
			let i1 = values[i + numPts];
			let v1 = pointScalars.getData()[i1];
			let c1 = v1 > 0;
			let j0 = c1 ? i1 : -1;
			let j1 = 0;
			linePts[0] = 0;
			linePts[1] = 0;
			let idListIdx = 0;
			for (let j = 1; j <= numPts; j++) {
				const i0 = i1;
				const v0 = v1;
				const c0 = c1;
				i1 = values[i + j];
				v1 = pointScalars.getData()[i1];
				c1 = v1 > 0;
				if (c0 || c1) {
					if (c0 ? !c1 : c1) {
						j1 = interpolateEdge(points, pointData, edgeLocator, model.tolerance, i0, i1, v0, v1);
						if (j1 !== j0) {
							idList[idListIdx++] = j1;
							j0 = j1;
						}
						linePts[c0 ? 1 : 0] = j1;
					}
					if (c1) {
						j1 = i1;
						if (j1 !== j0) {
							idList[idListIdx++] = j1;
							j0 = j1;
						}
					}
				}
			}
			const numPoints = idListIdx;
			idList.length = numPoints;
			if (model.triangulatePolys && numPoints > polyMax) {
				let newCellId = outputPolys.getNumberOfCells();
				if (!triangulatePolygon(idList, points, outputPolys)) triangulationFailure = true;
				const ncells = outputPolys.getNumberOfCells();
				for (; newCellId < ncells; newCellId++) outPolyData.passData(inCellData, cellId, newCellId);
			} else if (numPoints > 2) {
				const newCellId = outputPolys.insertNextCell(idList);
				outPolyData.passData(inCellData, cellId, newCellId);
			}
			if (linePts[0] !== linePts[1]) {
				const newCellId = outputLines.insertNextCell(linePts);
				outLineData.passData(inCellData, cellId, newCellId);
			}
		}
		if (triangulationFailure && model.triangulationErrorDisplay) vtkErrorMacro("Triangulation failed, output may not be watertight");
	}
	/**
	* Squeeze the points and store them in the output.  Only the points that
	* are used by the cells will be saved, and the pointIds of the cells will
	* be modified.
	*
	* @param {vtkPolyData} output
	* @param {vtkPoints} points
	* @param {vtkDataSetAttributes} pointData
	* @param {String} outputPointDataType
	*/
	function squeezeOutputPoints(output, points, pointData, outputPointDataType) {
		const n = points.getNumberOfPoints();
		let numNewPoints = 0;
		const outPointData = output.getPointData();
		const pointMap = [];
		pointMap.length = n;
		const cellArrays = [
			output.getVerts(),
			output.getLines(),
			output.getPolys(),
			output.getStrips()
		];
		cellArrays.forEach((cellArray) => {
			if (!cellArray) return;
			const values = cellArray.getData();
			let numPts;
			let pointId;
			for (let i = 0; i < values.length; i += numPts + 1) {
				numPts = values[i];
				for (let j = 1; j <= numPts; j++) {
					pointId = values[i + j];
					if (pointMap[pointId] === void 0) pointMap[pointId] = numNewPoints++;
				}
			}
		});
		const newPoints = Points_default.newInstance({
			size: numNewPoints * 3,
			dataType: outputPointDataType
		});
		const p = [];
		let newPointId;
		for (let pointId = 0; pointId < n; pointId++) {
			newPointId = pointMap[pointId];
			if (newPointId !== void 0) {
				points.getPoint(pointId, p);
				newPoints.setTuple(newPointId, p);
				outPointData.passData(pointData, pointId, newPointId);
			}
		}
		cellArrays.forEach((cellArray) => {
			if (!cellArray) return;
			const values = cellArray.getData();
			let numPts;
			let pointId;
			for (let i = 0; i < values.length; i += numPts + 1) {
				numPts = values[i];
				for (let j = 1; j <= numPts; j++) {
					pointId = values[i + j];
					values[i + j] = pointMap[pointId];
				}
			}
		});
		output.setPoints(newPoints);
	}
	publicAPI.requestData = (inData, outData) => {
		const input = inData[0];
		const output = outData[0]?.initialize() || PolyData_default.newInstance();
		outData[0] = output;
		if (!input) {
			vtkErrorMacro("Invalid or missing input");
			return;
		}
		if (model._idList == null) model._idList = [];
		else model._idList.length = 0;
		const inputPoints = input.getPoints();
		let numPts = 0;
		let inputPointsType = VtkDataTypes.FLOAT;
		if (inputPoints) {
			numPts = inputPoints.getNumberOfPoints();
			inputPointsType = inputPoints.getDataType();
		}
		const points = Points_default.newInstance({
			size: numPts * 3,
			dataType: VtkDataTypes.DOUBLE
		});
		const pointData = DataSetAttributes_default.newInstance();
		let inPointData = null;
		if (model.passPointData) inPointData = input.getPointData();
		const point = [];
		for (let ptId = 0; ptId < numPts; ptId++) {
			inputPoints.getPoint(ptId, point);
			points.setTuple(ptId, point);
			if (inPointData) pointData.passData(inPointData, ptId, ptId);
		}
		const edgeLocator = EdgeLocator_default.newInstance();
		const tmpContourData = PolyData_default.newInstance();
		let lineScalars;
		let polyScalars;
		let inputScalars;
		let firstLineScalar = 0;
		let firstPolyScalar = 0;
		let firstStripScalar = 0;
		let numberOfScalarComponents = 1;
		const colors = [
			[
				0,
				0,
				0
			],
			[
				0,
				0,
				0
			],
			[
				0,
				0,
				0
			]
		];
		if (model.scalarMode === ScalarMode.COLORS) {
			numberOfScalarComponents = 3;
			createColorValues(model.baseColor, model.clipColor, model.activePlaneColor, colors);
		} else if (model.scalarMode === ScalarMode.LABELS) {
			colors[0][0] = 0;
			colors[1][0] = 1;
			colors[2][0] = 2;
		}
		const numVerts = input.getVerts()?.getNumberOfCells() || 0;
		const inputLines = input.getLines();
		const numLines = inputLines?.getNumberOfCells() || 0;
		const inputPolys = input.getPolys();
		const numPolys = inputPolys?.getNumberOfCells() || 0;
		const numStrips = input.getStrips()?.getNumberOfCells() || 0;
		if (model.scalarMode !== ScalarMode.NONE) {
			lineScalars = DataArray_default.newInstance({
				dataType: VtkDataTypes.UNSIGNED_CHAR,
				empty: true,
				numberOfComponents: numberOfScalarComponents
			});
			const tryInputScalars = input.getCellData().getScalars();
			if (tryInputScalars && tryInputScalars.getDataType() === VtkDataTypes.UNSIGNED_CHAR && numberOfScalarComponents === 3 && tryInputScalars.getNumberOfComponents() === 3) {
				inputScalars = input.getCellData().getScalars();
				firstLineScalar = numVerts;
				firstPolyScalar = numVerts + numLines;
				firstStripScalar = numVerts + numLines + numPolys;
			}
		}
		let lines;
		if (numLines > 0) {
			lines = CellArray_default.newInstance({
				dataType: inputLines.getDataType(),
				values: new Uint8Array(numLines * 3),
				size: 0
			});
			breakPolylines(inputLines, lines, inputScalars, firstLineScalar, lineScalars, colors[0]);
		} else lines = CellArray_default.newInstance({ empty: true });
		let polys = null;
		let polyMax = 3;
		if (numPolys > 0 || numStrips > 0) {
			if (lineScalars) polyScalars = DataArray_default.newInstance({
				dataType: VtkDataTypes.UNSIGNED_CHAR,
				empty: true,
				numberOfComponents: numberOfScalarComponents
			});
			polys = CellArray_default.newInstance();
			copyPolygons(inputPolys, polys, inputScalars, firstPolyScalar, polyScalars, colors[0]);
			breakTriangleStrips(input.getStrips(), polys, inputScalars, firstStripScalar, polyScalars, colors[0]);
			polyMax = inputPolys.getCellSizes().reduce((a, b) => a > b ? a : b, 0);
		}
		let newLines = CellArray_default.newInstance({
			dataType: lines.getDataType(),
			empty: true
		});
		let newPolys = null;
		if (polys) newPolys = CellArray_default.newInstance({
			dataType: polys.getDataType(),
			empty: true
		});
		let inLineData = DataSetAttributes_default.newInstance();
		inLineData.copyScalarsOn();
		inLineData.setScalars(lineScalars);
		let inPolyData = DataSetAttributes_default.newInstance();
		inPolyData.copyScalarsOn();
		inPolyData.setScalars(polyScalars);
		let outLineData = DataSetAttributes_default.newInstance();
		outLineData.copyScalarsOn();
		let outPolyData = DataSetAttributes_default.newInstance();
		outPolyData.copyScalarsOn();
		const planes = model.clippingPlanes;
		for (let planeId = 0; planeId < planes.length; planeId++) {
			const plane = planes[planeId];
			let triangulate = 5;
			if (planeId === planes.length - 1) triangulate = polyMax;
			const active = planeId === model.activePlaneId;
			const pc = plane.getNormal();
			pc[3] = -dot(pc, plane.getOrigin());
			const numPoints = points.getNumberOfPoints();
			const pointScalars = DataArray_default.newInstance({
				dataType: VtkDataTypes.DOUBLE,
				size: numPoints
			});
			const pointScalarsData = pointScalars.getData();
			const pointsData = points.getData();
			let i = 0;
			for (let pointId = 0; pointId < numPoints;) pointScalarsData[pointId++] = pointsData[i++] * pc[0] + pointsData[i++] * pc[1] + pointsData[i++] * pc[2] + pc[3];
			edgeLocator.initialize();
			clipLines(points, pointScalars, pointData, edgeLocator, lines, newLines, inLineData, outLineData);
			if (polys) {
				const numClipLines = newLines.getNumberOfCells();
				clipAndContourPolys(points, pointScalars, pointData, edgeLocator, triangulate, polys, newPolys, newLines, inPolyData, outPolyData, outLineData);
				let scalars = outLineData.getScalars();
				if (scalars) {
					const color = colors[1 + (active ? 1 : 0)];
					const activeColor = colors[2];
					const numNewLines = newLines.getNumberOfCells();
					const oldColor = [];
					for (let lineId = numClipLines; lineId < numNewLines; lineId++) {
						scalars.getTuple(lineId, oldColor);
						if (numberOfScalarComponents !== 3 || oldColor[0] !== activeColor[0] || oldColor[1] !== activeColor[1] || oldColor[2] !== activeColor[2]) scalars.setTuple(lineId, color);
					}
				}
				let cellId = newPolys.getNumberOfCells();
				const numClipAndContourLines = newLines.getNumberOfCells();
				tmpContourData.setPoints(points);
				tmpContourData.setLines(newLines);
				tmpContourData.buildCells();
				triangulateContours(tmpContourData, numClipLines, numClipAndContourLines - numClipLines, newPolys, pc);
				scalars = outPolyData.getScalars();
				if (scalars) {
					const color = colors[1 + (active ? 1 : 0)];
					const numCells = newPolys.getNumberOfCells();
					if (numCells > cellId) {
						scalars.insertTuple(numCells - 1, color);
						for (; cellId < numCells; cellId++) scalars.setTuple(cellId, color);
					}
				}
				scalars = outLineData.getScalars();
				if (scalars) {
					const color = [
						0,
						255,
						255
					];
					const numCells = newLines.getNumberOfCells();
					if (numCells > numClipAndContourLines) {
						scalars.insertTuple(numCells - 1, color);
						for (let lineCellId = numClipAndContourLines; lineCellId < numCells; lineCellId++) scalars.setTuple(lineCellId, color);
					}
				}
			}
			[lines, newLines] = [newLines, lines];
			newLines.initialize();
			if (polys) {
				[polys, newPolys] = [newPolys, polys];
				newPolys.initialize();
			}
			[inLineData, outLineData] = [outLineData, inLineData];
			outLineData.initialize();
			[inPolyData, outPolyData] = [outPolyData, inPolyData];
			outPolyData.initialize();
		}
		const scalars = inLineData.getScalars();
		if (model.generateOutline) output.setLines(lines);
		else if (scalars) scalars.initialize();
		if (model.generateFaces) {
			output.setPolys(polys);
			if (polys && scalars) {
				const pScalars = inPolyData.getScalars();
				const m = scalars.getNumberOfTuples();
				const n = pScalars.getNumberOfTuples();
				if (n > 0) {
					const color = [
						0,
						0,
						0
					];
					scalars.insertTuple(n + m - 1, color);
					for (let i = 0; i < n; i++) {
						pScalars.getTuple(i, color);
						scalars.setTuple(i + m, color);
					}
				}
			}
		}
		if (scalars && model.scalarMode === ScalarMode.COLORS) {
			scalars.setName("Colors");
			output.getCellData().setScalars(scalars);
		} else if (model.scalarMode === ScalarMode.LABELS) {
			const categories = scalars.newClone();
			categories.setData(scalars.getData().slice());
			categories.setName("Labels");
			output.getCellData().setScalars(categories);
		} else output.getCellData().setScalars(null);
		squeezeOutputPoints(output, points, pointData, inputPointsType);
	};
	Object.keys(ScalarMode).forEach((key) => {
		const name = capitalize(key.toLowerCase());
		publicAPI[`setScalarModeTo${name}`] = () => {
			model.scalarMode = ScalarMode[key];
		};
	});
}
var DEFAULT_VALUES = {
	clippingPlanes: null,
	tolerance: 1e-6,
	passPointData: false,
	triangulatePolys: false,
	scalarMode: ScalarMode.NONE,
	generateOutline: false,
	generateFaces: true,
	activePlaneId: -1,
	baseColor: [
		255 / 255,
		99 / 255,
		71 / 255
	],
	clipColor: [
		244 / 255,
		164 / 255,
		96 / 255
	],
	activePlaneColor: [
		227 / 255,
		207 / 255,
		87 / 255
	],
	triangulationErrorDisplay: false
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.algo(publicAPI, model, 1, 1);
	macros_default.setGet(publicAPI, model, [
		"clippingPlanes",
		"tolerance",
		"passPointData",
		"triangulatePolys",
		"scalarMode",
		"generateOutline",
		"generateFaces",
		"activePlaneId",
		"triangulationErrorDisplay"
	]);
	macros_default.setGetArray(publicAPI, model, [
		"baseColor",
		"clipColor",
		"activePlaneColor"
	], 3);
	vtkClipClosedSurface(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkClipClosedSurface");
var ClipClosedSurface_default = {
	newInstance,
	extend,
	...Constants_default
};
//#endregion
export { ClipClosedSurface_default as default, extend, newInstance };
