// @kitware/vtk.js/Filters/General/ContourLoopExtraction.js
import macros_default from "../../macros.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
//#region Sources/Filters/General/ContourLoopExtraction/index.js
var Dir = {
	Forward: 1,
	Backward: -1
};
var visited = /* @__PURE__ */ new Set();
function vtkContourLoopExtraction(publicAPI, model) {
	publicAPI.requestData = (inData, outData) => {
		const [input] = inData;
		const output = outData[0]?.initialize() || PolyData_default.newInstance();
		outData[0] = output;
		publicAPI.extractContours(input, output);
		output.modified();
	};
	publicAPI.traverseLoop = (pd, dir, startLineId, startPtId, loopPoints) => {
		let lineId = startLineId;
		let lastPtId = startPtId;
		let terminated = false;
		let numInserted = 0;
		while (!terminated) {
			const { cellPointIds } = pd.getCellPoints(lineId);
			if (!cellPointIds) continue;
			lastPtId = cellPointIds[0] !== lastPtId ? cellPointIds[0] : cellPointIds[1];
			numInserted++;
			const t = dir * numInserted;
			loopPoints.push({
				t,
				ptId: lastPtId
			});
			const lineCell = pd.getPointCells(lastPtId);
			if (lineCell.length !== 2 || lastPtId === startPtId) return lastPtId;
			if (lineCell.length === 2) {
				lineId = lineCell[0] !== lineId ? lineCell[0] : lineCell[1];
				visited.add(lineId);
			} else terminated = true;
		}
		return lastPtId;
	};
	publicAPI.extractContours = (input, output) => {
		const loops = [];
		visited.clear();
		const inLines = input.getLines();
		output.getPoints().setData(Float32Array.from(input.getPoints().getData()));
		for (let li = 0; li < inLines.getNumberOfCells(); li++) {
			if (visited.has(li)) continue;
			const { cellPointIds } = input.getCellPoints(li);
			if (!cellPointIds) continue;
			visited.add(li);
			const startPtId = cellPointIds[0];
			const loopPoints = [];
			loopPoints.push({
				t: 0,
				ptId: startPtId
			});
			if (startPtId !== publicAPI.traverseLoop(input, Dir.Forward, li, startPtId, loopPoints)) {
				publicAPI.traverseLoop(input, Dir.Backward, li, startPtId, loopPoints);
				loopPoints.sort((a, b) => a.t < b.t ? -1 : 1);
				if (loopPoints.length && loopPoints[0].ptId !== loopPoints[loopPoints.length - 1]?.ptId) loopPoints.push({ ...loopPoints[loopPoints.length - 1] });
			}
			if (loopPoints.length) loops.push(loopPoints);
		}
		const outLines = output.getLines();
		outLines.resize(0);
		loops.forEach((loop) => {
			outLines.insertNextCell(loop.map((pt) => pt.ptId));
		});
	};
}
var DEFAULT_VALUES = {};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.algo(publicAPI, model, 1, 1);
	vtkContourLoopExtraction(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkContourLoopExtraction");
var ContourLoopExtraction_default = {
	newInstance,
	extend
};
//#endregion
export { ContourLoopExtraction_default as default, extend, newInstance };
