// @kitware/vtk.js/Filters/General/ContourTriangulator.js
import macros_default from "../../macros.js";
import { VtkDataTypes } from "../../Common/Core/DataArray/Constants.js";
import CellArray_default from "../../Common/Core/CellArray.js";
import Polygon_default from "../../Common/DataModel/Polygon.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
import { vtkCCSCutHoleyPolys, vtkCCSFindTrueEdges, vtkCCSJoinLooseEnds, vtkCCSMakeHoleyPolys, vtkCCSMakePolysFromLines, vtkCCSSplitAtPinchPoints, vtkCCSTriangulate } from "./ContourTriangulator/helper.js";
//#region Sources/Filters/General/ContourTriangulator/index.js
var { vtkErrorMacro } = macros_default;
function triangulateContours(polyData, firstLine, numLines, polys, normal, triangulatePolys = true) {
	let triangulationFailure = false;
	if (numLines <= 0) return false;
	const points = polyData.getPoints();
	const newPolys = [];
	const incompletePolys = [];
	const oriented = normal?.length >= 3;
	vtkCCSMakePolysFromLines(polyData, firstLine, firstLine + numLines, oriented, newPolys, incompletePolys);
	let computedNormal = normal;
	if (!oriented) {
		computedNormal = [
			0,
			0,
			1
		];
		let maxnorm = 0;
		const n = [];
		for (let i = 0; i < newPolys.length; i++) {
			const norm = Polygon_default.getNormal(newPolys[i], points, n);
			if (norm > maxnorm) {
				maxnorm = norm;
				computedNormal[0] = n[0];
				computedNormal[1] = n[1];
				computedNormal[2] = n[2];
			}
		}
	}
	vtkCCSJoinLooseEnds(newPolys, incompletePolys, points, computedNormal);
	const polyEdges = [];
	const originalEdges = [];
	vtkCCSFindTrueEdges(newPolys, points, polyEdges, originalEdges);
	const numNewPolys = newPolys.length;
	const polyGroups = new Array(numNewPolys);
	for (let i = 0; i < numNewPolys; i++) polyGroups[i] = [i];
	vtkCCSMakeHoleyPolys(newPolys, points, polyGroups, polyEdges, originalEdges, computedNormal, oriented);
	if (!vtkCCSCutHoleyPolys(newPolys, points, polyGroups, polyEdges, computedNormal)) triangulationFailure = true;
	vtkCCSSplitAtPinchPoints(newPolys, points, polyGroups, polyEdges, computedNormal, oriented);
	for (let polyId = 0; polyId < polyGroups.length; polyId++) {
		if (polyGroups[polyId].length === 0) continue;
		if (!triangulatePolys) polys.insertNextCell(originalEdges.slice(1, originalEdges.length));
		else if (!vtkCCSTriangulate(newPolys[polyId], points, polyEdges[polyId], originalEdges, polys, computedNormal)) triangulationFailure = false;
	}
	return !triangulationFailure;
}
function triangulatePolygon(polygon, points, triangles) {
	const poly = [...polygon];
	const polys = [poly];
	const originalEdges = [];
	const polyEdges = [];
	vtkCCSFindTrueEdges(polys, points, polyEdges, originalEdges);
	const edges = polyEdges[0];
	let success = true;
	const normal = [];
	if (Polygon_default.getNormal(poly, points, normal) !== 0) success = vtkCCSTriangulate(poly, points, edges, originalEdges, triangles, normal);
	return success;
}
var STATIC = {
	triangulateContours,
	triangulatePolygon
};
function vtkContourTriangulator(publicAPI, model) {
	model.classHierarchy.push("vtkContourTriangulator");
	publicAPI.requestData = (inData, outData) => {
		const input = inData[0];
		const output = outData[0]?.initialize() || PolyData_default.newInstance();
		outData[0] = output;
		if (!input) {
			vtkErrorMacro("Invalid or missing input");
			return false;
		}
		const lines = input.getLines();
		if (lines == null || lines.getNumberOfCells() === 0) return true;
		input.buildCells();
		const polysArray = CellArray_default.newInstance({
			dataType: VtkDataTypes.DOUBLE,
			empty: true
		});
		output.setPolys(polysArray);
		output.setPoints(input.getPoints());
		output.getPointData().passData(input.getPointData());
		triangulateContours(input, input.getNumberOfVerts(), lines.getNumberOfCells(), polysArray, null, model.triangulatePolys);
		return true;
	};
}
var DEFAULT_VALUES = { triangulatePolys: true };
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.algo(publicAPI, model, 1, 1);
	macros_default.setGet(publicAPI, model, ["triangulatePolys"]);
	vtkContourTriangulator(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkContourTriangulator");
var ContourTriangulator_default = {
	newInstance,
	extend,
	...STATIC
};
//#endregion
export { STATIC, ContourTriangulator_default as default, extend, newInstance };
