// @kitware/vtk.js/Filters/General/ImageOutlineFilter.js
import macros_default from "../../macros.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import ImageData_default from "../../Common/DataModel/ImageData.js";
//#region Sources/Filters/General/ImageOutlineFilter/index.js
var { vtkErrorMacro } = macros_default;
function vtkImageOutlineFilter(publicAPI, model) {
	model.classHierarchy.push("vtkImageOutlineFilter");
	publicAPI.requestData = (inData, outData) => {
		const input = inData[0];
		if (!input || input.getClassName() !== "vtkImageData") {
			vtkErrorMacro("Invalid or missing input");
			return;
		}
		const output = outData[0]?.initialize() || ImageData_default.newInstance();
		output.set(input.get("spacing", "origin", "direction"));
		const getIndex = (point, dims) => point[0] + point[1] * dims[0] + point[2] * dims[0] * dims[1];
		const getIJK = (index, dims) => {
			const ijk = [
				0,
				0,
				0
			];
			ijk[0] = index % dims[0];
			ijk[1] = Math.floor(index / dims[0]) % dims[1];
			ijk[2] = Math.floor(index / (dims[0] * dims[1]));
			return ijk;
		};
		const dims = input.getDimensions();
		output.setDimensions(dims);
		output.computeTransforms();
		const values = new Uint8Array(input.getNumberOfPoints());
		const inputDataArray = input.getPointData().getScalars().getData();
		let kernelX = 0;
		let kernelY = 1;
		if (model.slicingMode === 1) {
			kernelX = 0;
			kernelY = 2;
		} else if (model.slicingMode === 0) {
			kernelX = 1;
			kernelY = 2;
		}
		inputDataArray.forEach((el, index) => {
			if (el !== model.background) {
				const ijk = getIJK(index, dims);
				let isBorder = false;
				for (let x = -1; x <= 1 && !isBorder; x++) for (let y = -1; y <= 1 && !isBorder; y++) {
					let dx = x;
					let dy = y;
					let dz = 0;
					if (model.slicingMode === 1) {
						dx = x;
						dy = 0;
						dz = y;
					} else if (model.slicingMode === 0) {
						dx = 0;
						dy = y;
						dz = x;
					}
					const evalX = ijk[kernelX] + dx;
					const evalY = ijk[kernelY] + dy;
					if (evalX >= 0 && evalX < dims[kernelX] && evalY >= 0 && evalY < dims[kernelY]) {
						if (inputDataArray[getIndex([
							ijk[0] + dx,
							ijk[1] + dy,
							ijk[2] + dz
						], dims)] !== el) isBorder = true;
					}
				}
				if (isBorder) values[index] = el;
				else values[index] = model.background;
			} else values[index] = model.background;
		});
		const dataArray = DataArray_default.newInstance({
			numberOfComponents: 1,
			values
		});
		output.getPointData().setScalars(dataArray);
		outData[0] = output;
	};
}
var DEFAULT_VALUES = {
	slicingMode: 2,
	background: 0
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.algo(publicAPI, model, 1, 1);
	macros_default.setGet(publicAPI, model, ["slicingMode", "background"]);
	vtkImageOutlineFilter(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkImageOutlineFilter");
var ImageOutlineFilter_default = {
	newInstance,
	extend
};
//#endregion
export { ImageOutlineFilter_default as default, extend, newInstance };
