// @kitware/vtk.js/Filters/General/ImageMarchingSquares.js
import macros_default from "../../macros.js";
import EdgeLocator_default from "../../Common/DataModel/EdgeLocator.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
import caseTable_default from "./ImageMarchingSquares/caseTable.js";
//#region Sources/Filters/General/ImageMarchingSquares/index.js
var { vtkErrorMacro, vtkDebugMacro } = macros_default;
function vtkImageMarchingSquares(publicAPI, model) {
	/**
	* Get the X,Y kernels based on the set slicing mode.
	* @returns {[number, number]}
	*/
	function getKernels() {
		let kernelX = 0;
		let kernelY = 1;
		if (model.slicingMode === 1) {
			kernelX = 0;
			kernelY = 2;
		} else if (model.slicingMode === 0) {
			kernelX = 1;
			kernelY = 2;
		}
		return [kernelX, kernelY];
	}
	model.classHierarchy.push("vtkImageMarchingSquares");
	/**
	* Get the list of contour values.
	* @returns {number[]}
	*/
	publicAPI.getContourValues = () => model.contourValues;
	/**
	* Set the list contour values.
	* @param {number[]} cValues
	*/
	publicAPI.setContourValues = (cValues) => {
		model.contourValues = cValues;
		publicAPI.modified();
	};
	const ids = [];
	const pixelScalars = [];
	const pixelPts = [];
	const edgeLocator = EdgeLocator_default.newInstance();
	/**
	* Retrieve scalars and pixel coordinates.
	* @param {Vector3} ijk origin of the pixel
	* @param {Vector3} dims dimensions of the image
	* @param {TypedArray} scalars list of scalar values
	* @param {Vector3} increments IJK slice increments
	* @param {number} kernelX index of the X element
	* @param {number} kernelY index of the Y element
	*/
	publicAPI.getPixelScalars = (ijk, dims, scalars, increments, kernelX, kernelY) => {
		const [i, j, k] = ijk;
		ids[0] = k * dims[1] * dims[0] + j * dims[0] + i;
		ids[1] = ids[0] + increments[kernelX];
		ids[2] = ids[0] + increments[kernelY];
		ids[3] = ids[2] + increments[kernelX];
		for (let ii = 0; ii < 4; ++ii) pixelScalars[ii] = scalars[ids[ii]];
	};
	/**
	* Retrieve pixel coordinates.
	* @param {Vector3} ijk origin of the pixel
	* @param {number} kernelX index of the X element
	* @param {number} kernelY index of the Y element
	* @param {Function} indexToWorld function to convert index to world coordinates
	*/
	publicAPI.getPixelPoints = (ijk, kernelX, kernelY, indexToWorld) => {
		const neighborIJK = [...ijk];
		indexToWorld(neighborIJK, pixelPts);
		neighborIJK[kernelX] += 1;
		const temp = indexToWorld(neighborIJK, []);
		pixelPts[3] = temp[0];
		pixelPts[4] = temp[1];
		pixelPts[5] = temp[2];
		neighborIJK[kernelY] += 1;
		indexToWorld(neighborIJK, temp);
		pixelPts[9] = temp[0];
		pixelPts[10] = temp[1];
		pixelPts[11] = temp[2];
		neighborIJK[kernelX] -= 1;
		indexToWorld(neighborIJK, temp);
		pixelPts[6] = temp[0];
		pixelPts[7] = temp[1];
		pixelPts[8] = temp[2];
	};
	/**
	* Produce points and lines for the polydata.
	* @param {number[]} cVal list of contour values
	* @param {Vector3} ijk origin of the pixel
	* @param {Vector3} dims dimensions of the image
	* @param {Vector3} spacing sapcing of the image
	* @param {TypedArray} scalars list of scalar values
	* @param {number[]} points list of points
	* @param {number[]} lines list of lines
	* @param {Vector3} increments IJK slice increments
	* @param {number} kernelX index of the X element
	* @param {number} kernelY index of the Y element
	* @param {Function} indexToWorld function to convert index to world coordinates
	*/
	publicAPI.produceLines = (cVal, ijk, dims, scalars, points, lines, increments, kernelX, kernelY, indexToWorld) => {
		const CASE_MASK = [
			1,
			2,
			8,
			4
		];
		const xyz = [];
		let pId;
		publicAPI.getPixelScalars(ijk, dims, scalars, increments, kernelX, kernelY);
		let index = 0;
		for (let idx = 0; idx < 4; idx++) if (pixelScalars[idx] >= cVal) index |= CASE_MASK[idx];
		const pixelLines = caseTable_default.getCase(index);
		if (pixelLines[0] < 0) return;
		publicAPI.getPixelPoints(ijk, kernelX, kernelY, indexToWorld);
		for (let idx = 0; pixelLines[idx] >= 0; idx += 2) {
			lines.push(2);
			for (let eid = 0; eid < 2; eid++) {
				const edgeVerts = caseTable_default.getEdge(pixelLines[idx + eid]);
				pId = void 0;
				if (model.mergePoints) pId = edgeLocator.isInsertedEdge(ids[edgeVerts[0]], ids[edgeVerts[1]])?.value;
				if (pId === void 0) {
					const t = (cVal - pixelScalars[edgeVerts[0]]) / (pixelScalars[edgeVerts[1]] - pixelScalars[edgeVerts[0]]);
					const x0 = pixelPts.slice(edgeVerts[0] * 3, (edgeVerts[0] + 1) * 3);
					const x1 = pixelPts.slice(edgeVerts[1] * 3, (edgeVerts[1] + 1) * 3);
					xyz[0] = x0[0] + t * (x1[0] - x0[0]);
					xyz[1] = x0[1] + t * (x1[1] - x0[1]);
					xyz[2] = x0[2] + t * (x1[2] - x0[2]);
					pId = points.length / 3;
					points.push(xyz[0], xyz[1], xyz[2]);
					if (model.mergePoints) edgeLocator.insertEdge(ids[edgeVerts[0]], ids[edgeVerts[1]], pId);
				}
				lines.push(pId);
			}
		}
	};
	publicAPI.requestData = (inData, outData) => {
		const input = inData[0];
		if (!input) {
			vtkErrorMacro("Invalid or missing input");
			return;
		}
		if (model.slicingMode == null || model.slicingMode < 0 || model.slicingMode > 2) {
			vtkErrorMacro("Invalid or missing slicing mode");
			return;
		}
		console.time("msquares");
		const dims = input.getDimensions();
		const extent = input.getExtent();
		const increments = input.computeIncrements(extent);
		const scalars = input.getPointData().getScalars().getData();
		const [kernelX, kernelY] = getKernels();
		const indexToWorld = input.indexToWorld;
		const points = [];
		const lines = [];
		let k = Math.round(model.slice);
		if (k >= dims[model.slicingMode]) k = 0;
		const ijk = [
			0,
			0,
			0
		];
		ijk[model.slicingMode] = k;
		for (let cv = 0; cv < model.contourValues.length; ++cv) {
			for (let j = 0; j < dims[kernelY] - 1; ++j) {
				ijk[kernelY] = j;
				for (let i = 0; i < dims[kernelX] - 1; ++i) {
					ijk[kernelX] = i;
					publicAPI.produceLines(model.contourValues[cv], ijk, dims, scalars, points, lines, increments, kernelX, kernelY, indexToWorld);
				}
			}
			edgeLocator.initialize();
		}
		const polydata = outData[0]?.initialize() || PolyData_default.newInstance();
		polydata.getPoints().setData(new Float32Array(points), 3);
		polydata.getLines().setData(new Uint32Array(lines));
		outData[0] = polydata;
		vtkDebugMacro("Produced output");
		console.timeEnd("msquares");
	};
}
var DEFAULT_VALUES = {
	contourValues: [],
	slicingMode: 2,
	slice: 0,
	mergePoints: false
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.algo(publicAPI, model, 1, 1);
	macros_default.setGet(publicAPI, model, [
		"slicingMode",
		"slice",
		"mergePoints"
	]);
	macros_default.algo(publicAPI, model, 1, 1);
	vtkImageMarchingSquares(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkImageMarchingSquares");
var ImageMarchingSquares_default = {
	newInstance,
	extend
};
//#endregion
export { ImageMarchingSquares_default as default, extend, newInstance };
