// @kitware/vtk.js/Filters/General/ScalarToRGBA.js
import macros_default from "../../macros.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import ImageData_default from "../../Common/DataModel/ImageData.js";
//#region Sources/Filters/General/ScalarToRGBA/index.js
var { vtkErrorMacro } = macros_default;
function vtkScalarToRGBA(publicAPI, model) {
	model.classHierarchy.push("vtkScalarToRGBA");
	publicAPI.requestData = (inData, outData) => {
		const input = inData[0];
		if (!input) {
			vtkErrorMacro("Invalid or missing input");
			return;
		}
		const scalars = input.getPointData().getScalars();
		if (!scalars) {
			vtkErrorMacro("No scalars from input");
			return;
		}
		if (!model.lookupTable) {
			vtkErrorMacro("No lookupTable available");
			return;
		}
		if (!model.piecewiseFunction) {
			vtkErrorMacro("No piecewiseFunction available");
			return;
		}
		const rgba = [
			0,
			0,
			0,
			0
		];
		const data = scalars.getData();
		const rgbaArray = new Uint8ClampedArray(data.length * 4);
		let offset = 0;
		for (let idx = 0; idx < data.length; idx++) {
			const x = data[idx];
			model.lookupTable.getColor(x, rgba);
			rgba[3] = model.piecewiseFunction.getValue(x);
			rgbaArray[offset++] = 255 * rgba[0];
			rgbaArray[offset++] = 255 * rgba[1];
			rgbaArray[offset++] = 255 * rgba[2];
			rgbaArray[offset++] = 255 * rgba[3];
		}
		const colorArray = DataArray_default.newInstance({
			name: "rgba",
			numberOfComponents: 4,
			values: rgbaArray
		});
		const datasetDefinition = input.get("extent", "spacing", "origin", "direction");
		const output = outData[0]?.initialize() || ImageData_default.newInstance();
		output.set(datasetDefinition);
		output.getPointData().setScalars(colorArray);
		outData[0] = output;
	};
}
var DEFAULT_VALUES = {};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.algo(publicAPI, model, 1, 1);
	macros_default.setGet(publicAPI, model, ["lookupTable", "piecewiseFunction"]);
	vtkScalarToRGBA(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkScalarToRGBA");
var ScalarToRGBA_default = {
	newInstance,
	extend
};
//#endregion
export { ScalarToRGBA_default as default, extend, newInstance };
