// @kitware/vtk.js/Filters/General/TransformPolyDataFilter.js
import macros_default from "../../macros.js";
import { VtkDataTypes } from "../../Common/Core/DataArray/Constants.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import { DesiredOutputPrecision } from "../../Common/DataModel/DataSetAttributes/Constants.js";
import Points_default from "../../Common/Core/Points.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
//#region Sources/Filters/General/TransformPolyDataFilter/index.js
var { vtkErrorMacro } = macros_default;
function vtkTransformPolyDataFilter(publicAPI, model) {
	model.classHierarchy.push("vtkTransformPolyDataFilter");
	function transformPolyData(input, output) {
		if (!model.transform) {
			vtkErrorMacro("No transform defined!");
			return false;
		}
		const inPts = input.getPoints();
		const inPtData = input.getPointData();
		const outPD = output.getPointData();
		const inCellData = input.getCellData();
		const outCD = output.getCellData();
		if (!inPts) return true;
		const numPoints = inPts.getNumberOfPoints();
		const inVectors = inPtData.getVectors();
		const inNormals = inPtData.getNormals();
		const inCellVectors = inCellData.getVectors();
		const inCellNormals = inCellData.getNormals();
		let pointType = inPts.getDataType();
		if (model.outputPointsPrecision === DesiredOutputPrecision.SINGLE) pointType = VtkDataTypes.FLOAT;
		else if (model.outputPointsPrecision === DesiredOutputPrecision.DOUBLE) pointType = VtkDataTypes.DOUBLE;
		const outPts = Points_default.newInstance({ dataType: pointType });
		const inPtsData = inPts.getData();
		outPts.setNumberOfPoints(numPoints);
		const outPtsData = outPts.getData();
		let outVectors = null;
		if (inVectors) outVectors = DataArray_default.newInstance({
			name: inVectors.getName(),
			numberOfComponents: 3,
			size: numPoints * 3
		});
		let outNormals = null;
		if (inNormals) outNormals = DataArray_default.newInstance({
			name: inNormals.getName(),
			numberOfComponents: 3,
			size: numPoints * 3
		});
		if (inVectors || inNormals) model.transform.transformPointsNormalsVectors(inPts, outPts, inNormals, outNormals, inVectors, outVectors);
		else model.transform.transformPoints(inPtsData, outPtsData);
		let outCellVectors = null;
		if (inCellVectors) {
			const numCells = inCellVectors.getNumberOfTuples();
			outCellVectors = DataArray_default.newInstance({
				name: inCellVectors.getName(),
				numberOfComponents: 3,
				size: numCells * 3
			});
			model.transform.transformVectors(inCellVectors, outCellVectors);
		}
		let outCellNormals = null;
		if (inCellNormals) {
			const numCells = inCellNormals.getNumberOfTuples();
			outCellNormals = DataArray_default.newInstance({
				name: inCellNormals.getName(),
				numberOfComponents: 3,
				size: numCells * 3
			});
			model.transform.transformNormals(inCellNormals, outCellNormals);
		}
		output.setPoints(outPts);
		output.setVerts(input.getVerts());
		output.setLines(input.getLines());
		output.setPolys(input.getPolys());
		output.setStrips(input.getStrips());
		if (outNormals) {
			outPD.setNormals(outNormals);
			outPD.copyFieldOff(outNormals.getName());
		}
		if (outVectors) {
			outPD.setVectors(outVectors);
			outPD.copyFieldOff(outVectors.getName());
		}
		if (outCellNormals) {
			outCD.setNormals(outCellNormals);
			outCD.copyFieldOff(outCellNormals.getName());
		}
		if (outCellVectors) {
			outCD.setVectors(outCellVectors);
			outCD.copyFieldOff(outCellVectors.getName());
		}
		outPD.passData(inPtData);
		outCD.passData(inCellData);
		return true;
	}
	publicAPI.requestData = (inData, outData) => {
		const input = inData[0];
		const output = outData[0]?.initialize() || PolyData_default.newInstance();
		if (transformPolyData(input, output)) outData[0] = output;
		else vtkErrorMacro("TransformPolyDataFilter failed to transform input data.");
	};
}
var DEFAULT_VALUES = {
	transform: null,
	outputPointsPrecision: DesiredOutputPrecision.DEFAULT
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.algo(publicAPI, model, 1, 1);
	macros_default.setGet(publicAPI, model, ["transform", "outputPointsPrecision"]);
	vtkTransformPolyDataFilter(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkTransformPolyDataFilter");
var TransformPolyDataFilter_default = {
	newInstance,
	extend
};
//#endregion
export { TransformPolyDataFilter_default as default, extend, newInstance };
