// @kitware/vtk.js/Filters/General.js
import AppendPolyData_default from "./General/AppendPolyData.js";
import BFSConnectivityFilter_default from "./General/BFSConnectivityFilter.js";
import Calculator_default from "./General/Calculator.js";
import ContourTriangulator_default from "./General/ContourTriangulator.js";
import ClipClosedSurface_default from "./General/ClipClosedSurface.js";
import ClosedPolyLineToSurfaceFilter_default from "./General/ClosedPolyLineToSurfaceFilter.js";
import ImageCropFilter_default from "./General/ImageCropFilter.js";
import ImageMarchingCubes_default from "./General/ImageMarchingCubes.js";
import ImageMarchingSquares_default from "./General/ImageMarchingSquares.js";
import ImageOutlineFilter_default from "./General/ImageOutlineFilter.js";
import ImageSliceFilter_default from "./General/ImageSliceFilter.js";
import ImageStreamline_default from "./General/ImageStreamline.js";
import LineFilter_default from "./General/LineFilter.js";
import MoleculeToRepresentation_default from "./General/MoleculeToRepresentation.js";
import OBBTree_default from "./General/OBBTree.js";
import OutlineFilter_default from "./General/OutlineFilter.js";
import PaintFilter_default from "./General/PaintFilter.js";
import ScalarToRGBA_default from "./General/ScalarToRGBA.js";
import ShrinkPolyData_default from "./General/ShrinkPolyData.js";
import TransformPolyDataFilter_default from "./General/TransformPolyDataFilter.js";
import TriangleFilter_default from "./General/TriangleFilter.js";
import TubeFilter_default from "./General/TubeFilter.js";
import WarpScalar_default from "./General/WarpScalar.js";
import WindowedSincPolyDataFilter_default from "./General/WindowedSincPolyDataFilter.js";
//#region Sources/Filters/General/index.js
var General_default = {
	vtkAppendPolyData: AppendPolyData_default,
	vtkBFSConnectivityFilter: BFSConnectivityFilter_default,
	vtkCalculator: Calculator_default,
	vtkClipClosedSurface: ClipClosedSurface_default,
	vtkClosedPolyLineToSurfaceFilter: ClosedPolyLineToSurfaceFilter_default,
	vtkContourTriangulator: ContourTriangulator_default,
	vtkImageCropFilter: ImageCropFilter_default,
	vtkImageMarchingCubes: ImageMarchingCubes_default,
	vtkImageMarchingSquares: ImageMarchingSquares_default,
	vtkImageOutlineFilter: ImageOutlineFilter_default,
	vtkImageSliceFilter: ImageSliceFilter_default,
	vtkImageStreamline: ImageStreamline_default,
	vtkLineFilter: LineFilter_default,
	vtkMoleculeToRepresentation: MoleculeToRepresentation_default,
	vtkOBBTree: OBBTree_default,
	vtkOutlineFilter: OutlineFilter_default,
	vtkPaintFilter: PaintFilter_default,
	vtkScalarToRGBA: ScalarToRGBA_default,
	vtkShrinkPolyData: ShrinkPolyData_default,
	vtkTransformPolyDataFilter: TransformPolyDataFilter_default,
	vtkTriangleFilter: TriangleFilter_default,
	vtkTubeFilter: TubeFilter_default,
	vtkWarpScalar: WarpScalar_default,
	vtkWindowedSincPolyDataFilter: WindowedSincPolyDataFilter_default
};
//#endregion
export { General_default as default };
