// @kitware/vtk.js/Filters/Sources.js
import CubeSource_default from "./Sources/CubeSource.js";
import ArcSource_default from "./Sources/ArcSource.js";
import ConeSource_default from "./Sources/ConeSource.js";
import CylinderSource_default from "./Sources/CylinderSource.js";
import ArrowSource_default from "./Sources/ArrowSource.js";
import CircleSource_default from "./Sources/CircleSource.js";
import ConcentricCylinderSource_default from "./Sources/ConcentricCylinderSource.js";
import Cursor3D_default from "./Sources/Cursor3D.js";
import DiskSource_default from "./Sources/DiskSource.js";
import EllipseArcSource_default from "./Sources/EllipseArcSource.js";
import ImageGridSource_default from "./Sources/ImageGridSource.js";
import LineSource_default from "./Sources/LineSource.js";
import PlaneSource_default from "./Sources/PlaneSource.js";
import PlatonicSolidSource_default from "./Sources/PlatonicSolidSource.js";
import PointSource_default from "./Sources/PointSource.js";
import RegularPolygonSource_default from "./Sources/RegularPolygonSource.js";
import RTAnalyticSource_default from "./Sources/RTAnalyticSource.js";
import SLICSource_default from "./Sources/SLICSource.js";
import SphereSource_default from "./Sources/SphereSource.js";
import TorusSource_default from "./Sources/TorusSource.js";
//#region Sources/Filters/Sources/index.js
var Sources_default = {
	vtkArcSource: ArcSource_default,
	vtkArrowSource: ArrowSource_default,
	vtkCircleSource: CircleSource_default,
	vtkConcentricCylinderSource: ConcentricCylinderSource_default,
	vtkConeSource: ConeSource_default,
	vtkCubeSource: CubeSource_default,
	vtkCursor3D: Cursor3D_default,
	vtkCylinderSource: CylinderSource_default,
	vtkDiskSource: DiskSource_default,
	vtkEllipseArcSource: EllipseArcSource_default,
	vtkImageGridSource: ImageGridSource_default,
	vtkLineSource: LineSource_default,
	vtkPlaneSource: PlaneSource_default,
	vtkPlatonicSolidSource: PlatonicSolidSource_default,
	vtkPointSource: PointSource_default,
	vtkRegularPolygonSource: RegularPolygonSource_default,
	vtkRTAnalyticSource: RTAnalyticSource_default,
	vtkSLICSource: SLICSource_default,
	vtkSphereSource: SphereSource_default,
	vtkTorusSource: TorusSource_default
};
//#endregion
export { Sources_default as default };
