// @kitware/vtk.js/Rendering/Core/ColorTransferFunction/CssFilters.js
import { identity, multiplyMatrix } from "../../../Common/Core/Math.js";
//#region Sources/Rendering/Core/ColorTransferFunction/CssFilters.js
/**
* A helper file to transform RGBA points using CSS filters equivalent
* The equivalents of CSS filters using SVG filters can be found here:
* https://www.w3.org/TR/filter-effects-1/#ShorthandEquivalents
* For each SVG filter, you can look for the maths behind it on the same page:
* https://www.w3.org/TR/filter-effects-1/#FilterPrimitivesOverview
*
* For example, the saturate filter equivalent is here:
* https://www.w3.org/TR/filter-effects-1/#saturateEquivalent
* And the maths behind the feColorMatrix of type saturate is here:
* https://www.w3.org/TR/filter-effects-1/#ref-for-attr-valuedef-type-saturate
*
* The transforms are done using matrices of size 5 by 5. They are row major
* as in vtkMath. The vectors representing the RGBA points uses
* [R, G, B, A, 1] vectors, with each channel between 0 and 1.
*/
var luminanceWeights = [
	.213,
	.715,
	.072
];
function createCSSFilter() {
	return new Array(25);
}
function createIdentityFilter(outFilter = createCSSFilter()) {
	return identity(5, outFilter);
}
function combineFilters(baseFilter, newFilter, outFilter = createCSSFilter()) {
	multiplyMatrix(newFilter, baseFilter, 5, 5, 5, 5, outFilter);
	return outFilter;
}
function applyFilter(filter, r, g, b, a = 1) {
	const vec = [
		r,
		g,
		b,
		a,
		1
	];
	multiplyMatrix(filter, vec, 5, 5, 5, 1, vec);
	const output = new Array(4);
	for (let i = 0; i < 4; ++i) {
		const value = vec[i];
		if (value < 0) output[i] = 0;
		else if (value > 1) output[i] = 1;
		else output[i] = value;
	}
	return output;
}
function createLinearFilter(slope, intercept, outFilter = createCSSFilter()) {
	createIdentityFilter(outFilter);
	for (let row = 0; row < 3; ++row) {
		outFilter[row * 5 + row] = slope;
		outFilter[row * 5 + 4] = intercept;
	}
	return outFilter;
}
function createContrastFilter(contrast, outFilter = createCSSFilter()) {
	return createLinearFilter(contrast, -(.5 * contrast) + .5, outFilter);
}
function createSaturateFilter(saturate, outFilter = createCSSFilter()) {
	createIdentityFilter(outFilter);
	for (let col = 0; col < 3; ++col) {
		const columnLuminance = luminanceWeights[col];
		const diagonalValue = columnLuminance + (1 - columnLuminance) * saturate;
		const nonDiagonalValue = columnLuminance - columnLuminance * saturate;
		for (let row = 0; row < 3; ++row) outFilter[row * 5 + col] = row === col ? diagonalValue : nonDiagonalValue;
	}
	return outFilter;
}
function createBrightnessFilter(brightness, outFilter = createCSSFilter()) {
	return createLinearFilter(brightness, 0, outFilter);
}
function createInvertFilter(invert, outFilter = createCSSFilter()) {
	return createLinearFilter(1 - 2 * invert, invert, outFilter);
}
//#endregion
export { applyFilter, combineFilters, createBrightnessFilter, createCSSFilter, createContrastFilter, createIdentityFilter, createInvertFilter, createLinearFilter, createSaturateFilter, luminanceWeights };
