// @kitware/vtk.js/IO/Core/ImageStream.js
import macros_default from "../../macros.js";
import createMethods from "./ImageStream/DefaultProtocol.js";
import ViewStream_default from "./ImageStream/ViewStream.js";
//#region Sources/IO/Core/ImageStream/index.js
function vtkImageStream(publicAPI, model) {
	model.classHierarchy.push("vtkImageStream");
	function onImage(data) {
		const message = data[0];
		if (!message || !message.image) return;
		for (let i = 0; i < model.viewStreams.length; i++) model.viewStreams[i].processMessage(message);
	}
	publicAPI.setServerAnimationFPS = (maxFPS = 30) => {
		let changeDetected = false;
		if (model.serverAnimationFPS !== maxFPS) {
			model.serverAnimationFPS = maxFPS;
			changeDetected = true;
		}
		if (!model.protocol) return Promise.resolve(true);
		if (changeDetected) publicAPI.modified();
		return model.protocol.setServerAnimationFPS(maxFPS);
	};
	publicAPI.connect = (session, protocol = createMethods) => {
		if (model.connected || !session || !protocol) return;
		model.protocol = protocol(session);
		model.protocol.subscribeToImageStream(onImage).promise.then((subscription) => {
			model.renderTopicSubscription = subscription;
			model.connected = true;
		}).catch((e) => {
			model.connected = false;
			console.error(e);
		});
	};
	publicAPI.disconnect = () => {
		if (model.protocol && model.connected && model.renderTopicSubscription) {
			model.protocol.unsubscribeToImageStream(model.renderTopicSubscription);
			model.renderTopicSubscription = null;
		}
		model.connected = false;
	};
	publicAPI.registerViewStream = (view) => {
		model.viewStreams.push(view);
	};
	publicAPI.unregisterViewStream = (view) => {
		model.viewStreams = model.viewStreams.filter((v) => v !== view);
	};
	publicAPI.createViewStream = (viewId = "-1", size = [400, 400]) => {
		const { setServerAnimationFPS, getServerAnimationFPS, unregisterViewStream } = publicAPI;
		const viewStream = ViewStream_default.newInstance({
			protocol: model.protocol,
			unregisterViewStream,
			sharedAPI: {
				setServerAnimationFPS,
				getServerAnimationFPS
			}
		});
		viewStream.setViewId(viewId);
		viewStream.setSize(size[0], size[1]);
		publicAPI.registerViewStream(viewStream);
		return viewStream;
	};
	publicAPI.delete = macros_default.chain(() => {
		while (model.viewStreams.length) model.viewStreams.pop().delete();
		publicAPI.disconnect();
	}, publicAPI.delete);
}
var DEFAULT_VALUES = {
	viewStreams: [],
	serverAnimationFPS: -1
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.get(publicAPI, model, ["serverAnimationFPS", "protocol"]);
	vtkImageStream(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkImageStream");
var ImageStream_default = {
	newInstance,
	extend
};
//#endregion
export { ImageStream_default as default, extend, newInstance };
