// @kitware/vtk.js/IO/Core/WSLinkClient.js
import macros_default from "../../macros.js";
import ImageStream_default from "./ImageStream.js";
//#region Sources/IO/Core/WSLinkClient/index.js
var SMART_CONNECT_CLASS = null;
function setSmartConnectClass(klass) {
	SMART_CONNECT_CLASS = klass;
}
function busy(fn, update) {
	return (...args) => new Promise((resolve, reject) => {
		update(1);
		fn(...args).then((response) => {
			update(-1);
			resolve(response);
		}, (error) => {
			update(-1);
			reject(error);
		});
	});
}
function busyWrap(methodMap, update, skipList = []) {
	const busyContainer = {};
	Object.keys(methodMap).forEach((methodName) => {
		if (skipList.indexOf(methodName) === -1) busyContainer[methodName] = busy(methodMap[methodName], update);
		else busyContainer[methodName] = methodMap[methodName];
	});
	return busyContainer;
}
function vtkWSLinkClient(publicAPI, model) {
	model.classHierarchy.push("vtkWSLinkClient");
	function notifyBusy() {
		publicAPI.invokeBusyChange(model.busyCount);
	}
	function updateBusy(delta = 0) {
		model.busyCount += delta;
		if (model.timeoutId) {
			clearTimeout(model.timeoutId);
			model.timeoutId = 0;
		}
		if (model.busyCount) notifyBusy();
		else model.timeoutId = setTimeout(notifyBusy, model.notificationTimeout);
	}
	publicAPI.beginBusy = () => updateBusy(1);
	publicAPI.endBusy = () => updateBusy(-1);
	publicAPI.isBusy = () => !!model.busyCount;
	publicAPI.isConnected = () => !!model.connection;
	publicAPI.connect = (config = {}, configDecorator = null) => {
		if (!SMART_CONNECT_CLASS) return Promise.reject(/* @__PURE__ */ new Error("Need to provide SmartConnect"));
		if (model.connection) return Promise.reject(/* @__PURE__ */ new Error("Need to disconnect first"));
		model.config = config;
		model.configDecorator = configDecorator || model.configDecorator;
		return new Promise((resolve, reject) => {
			model.smartConnect = SMART_CONNECT_CLASS.newInstance({
				config,
				configDecorator: model.configDecorator
			});
			model.smartConnect.onConnectionReady((connection) => {
				model.connection = connection;
				model.remote = {};
				model.config = model.smartConnect.getConfig();
				const session = connection.getSession();
				model.protocols = model.protocols || {};
				Object.keys(model.protocols).forEach((name) => {
					model.remote[name] = busyWrap(model.protocols[name](session), updateBusy, model.notBusyList);
				});
				if (model.createImageStream) {
					model.imageStream = ImageStream_default.newInstance();
					model.imageStream.connect(session);
				}
				publicAPI.invokeConnectionReady(publicAPI);
				resolve(publicAPI);
			});
			model.smartConnect.onConnectionError((error) => {
				publicAPI.invokeConnectionError(error);
				reject(error);
			});
			model.smartConnect.onConnectionClose((close) => {
				publicAPI.invokeConnectionClose(close);
				reject(close);
			});
			model.smartConnect.connect();
		});
	};
	publicAPI.disconnect = (timeout = 60) => {
		if (model.connection) {
			model.connection.destroy(timeout);
			model.connection = null;
		}
	};
	publicAPI.registerProtocol = (name, protocol) => {
		model.remote[name] = busyWrap(protocol(model.connection.getSession()), updateBusy, model.notBusyList);
	};
	publicAPI.unregisterProtocol = (name) => {
		delete model.remote[name];
	};
}
var DEFAULT_VALUES = {
	notBusyList: [],
	busyCount: 0,
	timeoutId: 0,
	notificationTimeout: 50,
	createImageStream: true
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, [
		"protocols",
		"notBusyList",
		"createImageStream",
		"configDecorator"
	]);
	macros_default.get(publicAPI, model, [
		"connection",
		"config",
		"remote",
		"imageStream"
	]);
	macros_default.event(publicAPI, model, "BusyChange");
	macros_default.event(publicAPI, model, "ConnectionReady");
	macros_default.event(publicAPI, model, "ConnectionError");
	macros_default.event(publicAPI, model, "ConnectionClose");
	vtkWSLinkClient(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkWSLinkClient");
var WSLinkClient_default = {
	newInstance,
	extend,
	setSmartConnectClass
};
//#endregion
export { WSLinkClient_default as default, extend, newInstance };
