// @kitware/vtk.js/Rendering/OpenGL/SurfaceLIC/LineIntegralConvolution2D.js
import macros_default from "../../../macros.js";
import Framebuffer_default from "../Framebuffer.js";
import ShaderProgram_default from "../ShaderProgram.js";
import { ContrastEnhanceMode } from "../../Core/SurfaceLICInterface/Constants.js";
import vtkLineIntegralConvolution2D_LIC0_default from "./glsl/vtkLineIntegralConvolution2D_LIC0.glsl.js";
import vtkLineIntegralConvolution2D_LICI_default from "./glsl/vtkLineIntegralConvolution2D_LICI.glsl.js";
import vtkLineIntegralConvolution2D_LICN_default from "./glsl/vtkLineIntegralConvolution2D_LICN.glsl.js";
import vtkLineIntegralConvolution2D_CE_default from "./glsl/vtkLineIntegralConvolution2D_CE.glsl.js";
import vtkLineIntegralConvolution2D_EE_default from "./glsl/vtkLineIntegralConvolution2D_EE.glsl.js";
import vtkLineIntegralConvolution2D_AAH_default from "./glsl/vtkLineIntegralConvolution2D_AAH.glsl.js";
import vtkLineIntegralConvolution2D_VT_default from "./glsl/vtkLineIntegralConvolution2D_VT.glsl.js";
import vtkLineIntegralConvolution2D_AAV_default from "./glsl/vtkLineIntegralConvolution2D_AAV.glsl.js";
import vtkLineIntegralConvolution2D_quadVS_default from "./glsl/vtkLineIntegralConvolution2D_quadVS.glsl.js";
import pingpong_default from "./LineIntegralConvolution2D/pingpong.js";
//#region Sources/Rendering/OpenGL/SurfaceLIC/LineIntegralConvolution2D/index.js
function getVectorLookupProgram(normalize = true) {
	const getNormVecSrc = `
    vec2 getVector( vec2 vectc )\n
      {\n
      vec2 V = texture2D( texVectors, vectc ).xy;\n
      // normalize if |V| not 0\n
      float lenV = length( V );\n
      if ( lenV > 1.0e-8 )\n
        {\n
        return V/lenV;\n
        }\n
      else\n
        {\n
        return vec2( 0.0, 0.0 );\n
        }\n
      }\n
    `;
	const getVecSrc = `
    vec2 getVector( vec2 vectc )\n
      {\n
      return texture2D( texVectors, vectc ).xy;\n
      }\n
    `;
	if (normalize) return getNormVecSrc;
	return getVecSrc;
}
function vtkLineIntegralConvolution2D(publicAPI, model) {
	model.classHierarchy.push("vtkLineIntegralConvolution2D");
	publicAPI.buildAShader = (fSource) => model._openGLRenderWindow.getShaderCache().readyShaderProgramArray(vtkLineIntegralConvolution2D_quadVS_default, fSource, "");
	publicAPI.dumpTextureValues = (texture, [width, height], context = model.context, openGLRenderWindow = model._openGLRenderWindow, nComp = 4) => {
		const fb = Framebuffer_default.newInstance();
		const gl = context;
		let pixels = null;
		fb.setOpenGLRenderWindow(openGLRenderWindow);
		fb.saveCurrentBindingsAndBuffers();
		fb.create(width, height);
		fb.populateFramebuffer();
		fb.setColorBuffer(texture);
		pixels = new Float32Array(width * height * nComp);
		gl.readPixels(0, 0, width, height, nComp === 4 ? gl.RGBA : gl.RGB, gl.FLOAT, pixels);
		fb.restorePreviousBindingsAndBuffers();
		return pixels;
	};
	publicAPI.getTextureMinMax = (texture, size, context = model.context, openGLRenderWindow = model._openGLRenderWindow) => {
		const values = publicAPI.dumpTextureValues(texture, size, context, openGLRenderWindow, 4);
		let min = Number.MAX_VALUE;
		let max = Number.MIN_VALUE;
		for (let i = 0; i < values.length; i += 4) if (values[i + 1] === 0) {
			const val = values[i];
			if (val < min) min = val;
			if (val > max) max = val;
		}
		return {
			min,
			max
		};
	};
	publicAPI.getComponentSelectionProgram = (ids) => {
		const compNames = "xyzw";
		return `.${compNames[ids[0]]}${compNames[ids[1]]}`;
	};
	publicAPI.buildShaders = () => {
		model.LIC0ShaderProgram = publicAPI.buildAShader(vtkLineIntegralConvolution2D_LIC0_default);
		const VTFSource = ShaderProgram_default.substitute(vtkLineIntegralConvolution2D_VT_default, "//VTK::LICComponentSelection::Impl", `vec2 V = texture2D(texVectors, tcoordVC.st)${publicAPI.getComponentSelectionProgram(model.componentIds)};`).result;
		model.VTProgram = publicAPI.buildAShader(VTFSource);
		const LICISource = ShaderProgram_default.substitute(vtkLineIntegralConvolution2D_LICI_default, "//VTK::LICVectorLookup::Impl", getVectorLookupProgram(model.normalizeVectors), true).result;
		model.LICIShaderProgram = publicAPI.buildAShader(LICISource);
		model.LICNShaderProgram = publicAPI.buildAShader(vtkLineIntegralConvolution2D_LICN_default);
		model.CEProgram = publicAPI.buildAShader(vtkLineIntegralConvolution2D_CE_default);
		model.EEProgram = publicAPI.buildAShader(vtkLineIntegralConvolution2D_EE_default);
		model.AAHProgram = publicAPI.buildAShader(vtkLineIntegralConvolution2D_AAH_default);
		model.AAVProgram = publicAPI.buildAShader(vtkLineIntegralConvolution2D_AAV_default);
	};
	function setLICUniforms(program, bufs) {
		program.setUniformi("texLIC", bufs.getLICTextureUnit());
		program.setUniformi("texSeedPts", bufs.getSeedTextureUnit());
	}
	function renderPingPong(bufs, size, program) {
		bufs.attachLICBuffers();
		bufs.renderQuad(size, program);
		bufs.detachLICBuffers();
		bufs.swap();
	}
	publicAPI.executeLIC = (size, vectorTexture, maskVectorTexture, noiseTexture, openGLRenderWindow, options) => {
		model._openGLRenderWindow = openGLRenderWindow;
		model.context = openGLRenderWindow.getContext();
		Object.assign(model, options);
		if (size[0] <= 0 || size[1] <= 0) return null;
		const tcScale = [1 / size[0], 1 / size[1]];
		let stepSize = model.stepSize * Math.sqrt(tcScale[0] * tcScale[0] + tcScale[1] * tcScale[1]);
		if (stepSize <= 0) stepSize = 1e-10;
		const gl = model.context;
		let fb = model.framebuffer;
		const fbSize = fb?.getSize();
		if (!fb || !fbSize || size[0] !== fbSize || size[1] !== fbSize) {
			fb = Framebuffer_default.newInstance();
			fb.setOpenGLRenderWindow(model._openGLRenderWindow);
			fb.saveCurrentBindingsAndBuffers();
			fb.create(...size);
			fb.populateFramebuffer();
			fb.restorePreviousBindingsAndBuffers();
			model.framebuffer = fb;
		}
		fb.saveCurrentBindingsAndBuffers();
		fb.bind();
		gl.viewport(0, 0, ...size);
		gl.scissor(0, 0, ...size);
		if (model.shadersNeedBuild) {
			publicAPI.buildShaders();
			model.shadersNeedBuild = false;
		}
		if (!model.bufs) model.bufs = pingpong_default.newInstance({
			openGLRenderWindow,
			doEEPass: model.enhancedLIC,
			doVTPass: model.transformVectors,
			vectorTexture,
			maskVectorTexture,
			noiseTexture,
			framebuffer: fb,
			size
		});
		else {
			model.bufs.setVectorTexture(vectorTexture);
			model.bufs.setMaskVectorTexture(maskVectorTexture);
			model.bufs.setNoiseTexture(noiseTexture);
		}
		const noiseBoundsPt1 = [(noiseTexture.getWidth() + 1) / size[0], (noiseTexture.getHeight() + 1) / size[1]];
		const dx = 1 / size[0];
		const dy = 1 / size[1];
		const shaderCache = model._openGLRenderWindow.getShaderCache();
		if (model.transformVectors) {
			const VTShaderProgram = model.VTProgram;
			shaderCache.readyShaderProgram(VTShaderProgram);
			model.bufs.attachImageVectorBuffer();
			VTShaderProgram.setUniform2f("uTexSize", ...size);
			VTShaderProgram.setUniformi("texVectors", model.bufs.getVectorTextureUnit());
			gl.clearColor(0, 0, 0, 0);
			gl.clear(gl.COLOR_BUFFER_BIT);
			model.bufs.renderQuad(size, VTShaderProgram);
			model.bufs.detachImageVectorBuffer();
		}
		model.bufs.clearBuffers(model.enhancedLIC);
		model.bufs.activateVectorTextures();
		model.bufs.activateNoiseTexture(0);
		const { LIC0ShaderProgram } = model;
		shaderCache.readyShaderProgram(LIC0ShaderProgram);
		LIC0ShaderProgram.setUniformi("uStepNo", 0);
		LIC0ShaderProgram.setUniformi("uPassNo", 0);
		LIC0ShaderProgram.setUniformf("uMaskThreshold", model.maskThreshold);
		LIC0ShaderProgram.setUniform2f("uNoiseBoundsPt1", ...noiseBoundsPt1);
		LIC0ShaderProgram.setUniformi("texMaskVectors", model.bufs.getMaskVectorTextureUnit());
		LIC0ShaderProgram.setUniformi("texLIC", model.bufs.getLICTextureUnit());
		LIC0ShaderProgram.setUniformi("texNoise", model.bufs.getNoiseTextureUnit(0));
		renderPingPong(model.bufs, size, LIC0ShaderProgram);
		const { LICIShaderProgram } = model;
		shaderCache.readyShaderProgram(LICIShaderProgram);
		LICIShaderProgram.setUniformi("uPassNo", 0);
		LICIShaderProgram.setUniformf("uStepSize", -stepSize);
		LICIShaderProgram.setUniform2f("uNoiseBoundsPt1", ...noiseBoundsPt1);
		LICIShaderProgram.setUniformi("texVectors", model.bufs.getImageVectorTextureUnit());
		LICIShaderProgram.setUniformi("texNoise", model.bufs.getNoiseTextureUnit(0));
		for (let stepIdx = 0; stepIdx < model.numberOfSteps; ++stepIdx) {
			setLICUniforms(LICIShaderProgram, model.bufs);
			renderPingPong(model.bufs, size, LICIShaderProgram);
		}
		shaderCache.readyShaderProgram(LIC0ShaderProgram);
		LIC0ShaderProgram.setUniformi("uStepNo", 1);
		setLICUniforms(LIC0ShaderProgram, model.bufs);
		renderPingPong(model.bufs, size, LIC0ShaderProgram);
		shaderCache.readyShaderProgram(LICIShaderProgram);
		LICIShaderProgram.setUniformf("uStepSize", stepSize);
		for (let stepIdx = 0; stepIdx < model.numberOfSteps; ++stepIdx) {
			setLICUniforms(LICIShaderProgram, model.bufs);
			renderPingPong(model.bufs, size, LICIShaderProgram);
		}
		model.bufs.deactivateNoiseTexture(0);
		model.bufs.deactivateVectorTextures();
		const { LICNShaderProgram } = model;
		shaderCache.readyShaderProgram(LICNShaderProgram);
		LICNShaderProgram.setUniformi("texLIC", model.bufs.getLICTextureUnit());
		renderPingPong(model.bufs, size, LICNShaderProgram);
		if (model.enhancedLIC) {
			if (model.enhanceContrast === ContrastEnhanceMode.LIC || model.enhanceContrast === ContrastEnhanceMode.BOTH) publicAPI.contrastEnhance(false, size);
			model.bufs.attachEEBuffer();
			const { EEProgram } = model;
			shaderCache.readyShaderProgram(EEProgram);
			EEProgram.setUniformi("texLIC", model.bufs.getLICTextureUnit());
			EEProgram.setUniformf("uDx", dx);
			EEProgram.setUniformf("uDy", dy);
			model.bufs.renderQuad(size, EEProgram);
			model.bufs.detachEEBuffer();
			model.bufs.detachBuffers();
			model.bufs.clearBuffers(false);
			model.bufs.activateVectorTextures();
			model.bufs.activateNoiseTexture(1);
			shaderCache.readyShaderProgram(LIC0ShaderProgram);
			LIC0ShaderProgram.setUniformi("uStepNo", 0);
			LIC0ShaderProgram.setUniformi("uPassNo", 1);
			setLICUniforms(LIC0ShaderProgram, model.bufs);
			LIC0ShaderProgram.setUniformi("texNoise", model.bufs.getNoiseTextureUnit(1));
			renderPingPong(model.bufs, size, LIC0ShaderProgram);
			shaderCache.readyShaderProgram(LICIShaderProgram);
			LICIShaderProgram.setUniformi("uPassNo", 1);
			LICIShaderProgram.setUniformf("uStepSize", -stepSize);
			LICIShaderProgram.setUniformi("texNoise", model.bufs.getNoiseTextureUnit(1));
			const nSteps = model.numberOfSteps / 2;
			for (let stepIdx = 0; stepIdx < nSteps; ++stepIdx) {
				setLICUniforms(LICIShaderProgram, model.bufs);
				renderPingPong(model.bufs, size, LICIShaderProgram);
			}
			shaderCache.readyShaderProgram(LIC0ShaderProgram);
			LIC0ShaderProgram.setUniformi("uStepNo", 1);
			setLICUniforms(LIC0ShaderProgram, model.bufs);
			renderPingPong(model.bufs, size, LIC0ShaderProgram);
			shaderCache.readyShaderProgram(LICIShaderProgram);
			LICIShaderProgram.setUniformf("uStepSize", stepSize);
			for (let stepIdx = 0; stepIdx < nSteps; ++stepIdx) {
				setLICUniforms(LICIShaderProgram, model.bufs);
				renderPingPong(model.bufs, size, LICIShaderProgram);
			}
			model.bufs.deactivateNoiseTexture(1);
			model.bufs.deactivateVectorTextures();
			shaderCache.readyShaderProgram(LICNShaderProgram);
			LICNShaderProgram.setUniformi("texLIC", model.bufs.getLICTextureUnit());
			LICNShaderProgram.setUniformi("texSeedPts", model.bufs.getSeedTextureUnit());
			renderPingPong(model.bufs, size, LICNShaderProgram);
		}
		if (model.antiAlias) {
			const AAHShaderProgram = model.AAHProgram;
			shaderCache.readyShaderProgram(AAHShaderProgram);
			AAHShaderProgram.setUniformi("texLIC", model.bufs.getLICTextureUnit());
			AAHShaderProgram.setUniformf("uDx", dx);
			const AAVShaderProgram = model.AAVProgram;
			shaderCache.readyShaderProgram(AAVShaderProgram);
			AAVShaderProgram.setUniformi("texLIC", model.bufs.getLICTextureUnit());
			AAVShaderProgram.setUniformf("uDy", dy);
			for (let i = 0; i < model.antiAlias; ++i) {
				shaderCache.readyShaderProgram(AAHShaderProgram);
				setLICUniforms(AAHShaderProgram, model.bufs);
				renderPingPong(model.bufs, size, AAHShaderProgram);
				shaderCache.readyShaderProgram(AAVShaderProgram);
				setLICUniforms(AAVShaderProgram, model.bufs);
				renderPingPong(model.bufs, size, AAVShaderProgram);
			}
		}
		if (model.enhanceContrast === ContrastEnhanceMode.LIC || model.enhanceContrast === ContrastEnhanceMode.BOTH) publicAPI.contrastEnhance(true, size);
		model.bufs.detachBuffers();
		fb.restorePreviousBindingsAndBuffers();
		return model.bufs.getLastLICBuffer();
	};
	publicAPI.contrastEnhance = (isSecondStage, size) => {
		const shaderCache = model._openGLRenderWindow.getShaderCache();
		let { min, max } = publicAPI.getTextureMinMax(model.bufs.getLastLICBuffer(), size, model.context, model._openGLRenderWindow);
		if (max <= min || max > 1 || min < 0) {
			console.error("Invalid color range: ", min, max);
			min = 0;
			max = 1;
		}
		let diff = max - min;
		if (isSecondStage) {
			min += diff * model.lowLICContrastEnhancementFactor;
			max -= diff * model.highLICContrastEnhancementFactor;
			diff = max - min;
		}
		const { CEProgram } = model;
		shaderCache.readyShaderProgram(CEProgram);
		CEProgram.setUniformi("texLIC", model.bufs.getLICTextureUnit());
		CEProgram.setUniformf("uMin", min);
		CEProgram.setUniformf("uMaxMinDiff", diff);
		renderPingPong(model.bufs, size, CEProgram);
	};
}
var DEFAULT_VALUES = {
	shadersNeedBuild: true,
	stepSize: 1,
	numberOfSteps: 10,
	enhancedLIC: true,
	enhanceContrast: false,
	lowContrastEnhancementFactor: 0,
	highContrastEnhancementFactor: 0,
	antiAlias: 0,
	componentIds: [0, 1],
	normalizeVectors: true,
	maskThreshold: 0,
	transformVectors: true,
	bufs: null,
	isComposite: true
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, [
		"context",
		"_openGLRenderWindow",
		"nuberOfSteps",
		"stepSize",
		"normalizeVectors",
		"maskThreshold",
		"enhancedLIC",
		"enhanceContrast",
		"lowLICContrastEnhancementFactor",
		"highLICContrastEnhancementFactor",
		"antiAlias",
		"componentIds",
		"isComposite"
	]);
	macros_default.moveToProtected(publicAPI, model, ["openGLRenderWindow"]);
	vtkLineIntegralConvolution2D(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkLineIntegralConvolution2D");
var LineIntegralConvolution2D_default = {
	newInstance,
	extend
};
//#endregion
export { LineIntegralConvolution2D_default as default, extend, newInstance };
