// @kitware/vtk.js/Rendering/OpenGL/SurfaceLIC/SurfaceLICMapper.js
import { macros_exports, newInstance as newInstance$1, setGet } from "../../../macros.js";
import { registerOverride } from "../ViewNodeFactory.js";
import ShaderProgram_default from "../ShaderProgram.js";
import PolyDataMapper_default from "../PolyDataMapper.js";
import SurfaceLICInterface_default from "../../Core/SurfaceLICInterface.js";
import SurfaceLICInterface_default$1 from "./SurfaceLICInterface.js";
//#region Sources/Rendering/OpenGL/SurfaceLIC/SurfaceLICMapper/index.js
var { vtkErrorMacro } = macros_exports;
function vtkOpenGLSurfaceLICMapper(publicAPI, model) {
	model.classHierarchy.push("vtkOpenGLSurfaceLICMapper");
	const superClass = { ...publicAPI };
	publicAPI.getNeedToRebuildShaders = (cellBO, ren, actor) => model.rebuildLICShaders || superClass.getNeedToRebuildShaders(cellBO, ren, actor);
	publicAPI.replaceShaderValues = (shaders, ren, actor) => {
		const prevComplexity = model.lastBoundBO.getReferenceByName("lastLightComplexity");
		let VSSource = shaders.Vertex;
		let FSSource = shaders.Fragment;
		const array = model.renderable.getInputArrayToProcess(0);
		if (array && model.canDrawLIC) {
			FSSource = ShaderProgram_default.substitute(FSSource, "//VTK::Output::Dec", [
				"//VTK::Output::Dec",
				"layout(location = 2) out vec4 vectorTexture;",
				"layout(location = 3) out vec4 maskVectorTexture;"
			]).result;
			const attributeName = `${array.getName()}MC`;
			if (prevComplexity === 0) model.lastBoundBO.set({ lastLightComplexity: 1 }, true);
			VSSource = ShaderProgram_default.substitute(VSSource, "//VTK::TCoord::Dec", [
				`attribute vec3 ${attributeName};`,
				"out vec3 licOutput;",
				"//VTK::TCoord::Dec"
			]).result;
			VSSource = ShaderProgram_default.substitute(VSSource, "//VTK::TCoord::Impl", [`licOutput = ${attributeName};`, "//VTK::TCoord::Impl"]).result;
			FSSource = ShaderProgram_default.substitute(FSSource, "//VTK::TCoord::Dec", [
				"uniform int uMaskOnSurface;",
				"uniform mat3 normalMatrix;",
				"in vec3 licOutput;",
				"//VTK::TCoord::Dec"
			]).result;
			FSSource = ShaderProgram_default.substitute(FSSource, "//VTK::TCoord::Impl", [
				"// projected vectors",
				"  vec3 tcoordLIC = normalMatrix * licOutput;",
				"  vec3 normN = normalize(normalVCVSOutput);",
				"  float k = dot(tcoordLIC, normN);",
				"  vec3 projected = (tcoordLIC - k*normN);",
				"  vectorTexture = vec4(projected.x, projected.y, 0.0 , 1.0);",
				"// vectors for fragment masking",
				"  if (uMaskOnSurface == 0)",
				"    {",
				"    maskVectorTexture = vec4(licOutput, 1.0);",
				"    }",
				"  else",
				"    {",
				"    maskVectorTexture = vec4(projected.x, projected.y, 0.0 , 1.0);",
				"    }",
				"//VTK::TCoord::Impl"
			], false).result;
			shaders.Vertex = VSSource;
		}
		model.rebuildLICShaders = false;
		shaders.Fragment = FSSource;
		superClass.replaceShaderValues(shaders, ren, actor);
		if (prevComplexity > 0) model.lastBoundBO.set({ lastLightComplexity: prevComplexity }, true);
	};
	publicAPI.setMapperShaderParameters = (cellBO, ren, actor) => {
		superClass.setMapperShaderParameters(cellBO, ren, actor);
		if (model.canDrawLIC) cellBO.getProgram().setUniformi("uMaskOnSurface", model.maskOnSurface);
	};
	publicAPI.getNeedToRebuildBufferObjects = (ren, actor) => model.rebuildLICBuffers || superClass.getNeedToRebuildBufferObjects(ren, actor);
	publicAPI.buildBufferObjects = (ren, actor) => {
		if (model.canDrawLIC) {
			const array = model.renderable.getInputArrayToProcess(0);
			if (array && array.getNumberOfComponents() > 1) model.renderable.setCustomShaderAttributes([array.getName()]);
		}
		model.rebuildLICBuffers = false;
		superClass.buildBufferObjects(ren, actor);
	};
	publicAPI.pushState = (gl) => {
		model.stateCache = {
			[gl.BLEND]: gl.isEnabled(gl.BLEND),
			[gl.DEPTH_TEST]: gl.isEnabled(gl.DEPTH_TEST),
			[gl.SCISSOR_TEST]: gl.isEnabled(gl.SCISSOR_TEST),
			[gl.CULL_FACE]: gl.isEnabled(gl.CULL_FACE)
		};
	};
	publicAPI.popState = (gl) => {
		const apply = (param) => model.stateCache[param] ? gl.enable(param) : gl.disable(param);
		apply(gl.BLEND);
		apply(gl.DEPTH_TEST);
		apply(gl.SCISSOR_TEST);
		if (model.stateCache[gl.CULL_FACE]) model._openGLRenderWindow.enableCullFace();
		else model._openGLRenderWindow.disableCullFace();
	};
	publicAPI.renderPiece = (ren, actor) => {
		let canDrawLIC = true;
		if (!model._openGLRenderWindow.getWebgl2()) {
			vtkErrorMacro("SurfaceLICMapper Requires WebGL 2");
			canDrawLIC = false;
		}
		if (!model.context.getExtension("EXT_color_buffer_float") || !model.context.getExtension("OES_texture_float_linear")) {
			vtkErrorMacro("SurfaceLICMapper requires the EXT_color_buffer_float and OES_texture_float_linear WebGL2 extensions.");
			canDrawLIC = false;
		}
		model.currentInput = model.renderable.getInputData();
		if (!model.currentInput) {
			vtkErrorMacro("No input");
			canDrawLIC = false;
		}
		let licInterface = model.renderable.getLicInterface();
		if (!licInterface) {
			licInterface = SurfaceLICInterface_default.newInstance();
			model.renderable.setLicInterface(licInterface);
		}
		if (!model.openGLLicInterface) model.openGLLicInterface = SurfaceLICInterface_default$1.newInstance();
		if (licInterface !== model.openGLLicInterface.getLicInterface()) model.openGLLicInterface.setLicInterface(licInterface);
		const array = model.renderable.getInputArrayToProcess(0);
		if (licInterface.getEnableLIC() && (!array || array.getNumberOfComponents() < 2)) {
			vtkErrorMacro("No vector input array");
			canDrawLIC = false;
		}
		if (!licInterface.getEnableLIC()) canDrawLIC = false;
		if (model.canDrawLIC !== canDrawLIC) {
			model.rebuildLICShaders = true;
			model.rebuildLICBuffers = true;
		}
		model.canDrawLIC = canDrawLIC;
		if (!canDrawLIC || !licInterface.getEnableLIC()) {
			superClass.renderPiece(ren, actor);
			return;
		}
		const gl = model.context;
		const backfaceCulling = actor.getProperty().getBackfaceCulling();
		const frontfaceCulling = actor.getProperty().getFrontfaceCulling();
		if (!backfaceCulling && !frontfaceCulling) model._openGLRenderWindow.disableCullFace();
		else if (frontfaceCulling) {
			model._openGLRenderWindow.enableCullFace();
			model._openGLRenderWindow.setCullFaceMode(gl.FRONT);
		} else {
			model._openGLRenderWindow.enableCullFace();
			model._openGLRenderWindow.setCullFaceMode(gl.BACK);
		}
		const windowSize = model._openGLRenderWindow.getSize();
		const size = windowSize.map((i) => Math.round(i * licInterface.getViewPortScale()));
		model.openGLLicInterface.setSize(size);
		model.openGLLicInterface.setOpenGLRenderWindow(model._openGLRenderWindow);
		model.openGLLicInterface.setContext(model.context);
		publicAPI.pushState(model.context);
		model.openGLLicInterface.initializeResources();
		model.openGLLicInterface.prepareForGeometry();
		publicAPI.popState(model.context);
		superClass.renderPieceStart(ren, actor);
		superClass.renderPieceDraw(ren, actor);
		superClass.renderPieceFinish(ren, actor);
		publicAPI.pushState(model.context);
		model.VBOBuildTime.modified();
		model.openGLLicInterface.completedGeometry();
		model._openGLRenderWindow.disableCullFace();
		model.openGLLicInterface.applyLIC();
		model.openGLLicInterface.combineColorsAndLIC();
		model.openGLLicInterface.copyToScreen(windowSize);
		publicAPI.popState(model.context);
	};
}
var DEFAULT_VALUES = {
	canDrawLIC: false,
	rebuildLICShaders: false,
	rebuildLICBuffers: false,
	openGLLicInterface: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	PolyDataMapper_default.extend(publicAPI, model, initialValues);
	vtkOpenGLSurfaceLICMapper(publicAPI, model);
	setGet(publicAPI, model, ["openGLLicInterface"]);
}
var newInstance = newInstance$1(extend, "vtkOpenGLSurfaceLICMapper");
var SurfaceLICMapper_default = {
	newInstance,
	extend
};
registerOverride("vtkSurfaceLICMapper", newInstance);
//#endregion
export { SurfaceLICMapper_default as default, extend, newInstance };
