// @kitware/vtk.js/IO/Core/Serializer/ArraySerializer.js
import macros_default from "../../../macros.js";
import DataArray_default from "../../../Common/Core/DataArray.js";
import CellArray_default from "../../../Common/Core/CellArray.js";
import Points_default from "../../../Common/Core/Points.js";
//#region Sources/IO/Core/Serializer/ArraySerializer.js
var FACTORY = {
	vtkDataArray: DataArray_default,
	vtkCellArray: CellArray_default,
	vtkPoints: Points_default
};
function createDefaultTypedArrayHandler() {
	const arrays = [];
	function write(array) {
		const id = arrays.length;
		arrays.push(array);
		return id;
	}
	function read(arrayId) {
		return arrays[arrayId];
	}
	return {
		write,
		read,
		arrays
	};
}
function vtkArraySerializer(publicAPI, model) {
	model.classHierarchy.push("vtkArraySerializer");
	if (!model.typedArrayHandler) model.typedArrayHandler = createDefaultTypedArrayHandler();
	publicAPI.serialize = (obj) => {
		const name = obj.getName();
		const numberOfTuples = obj.getNumberOfTuples();
		const vtkClass = obj.getClassName();
		const rawData = obj.getData();
		return {
			id: model.typedArrayHandler.write(rawData),
			name,
			numberOfTuples,
			vtkClass
		};
	};
	publicAPI.deserialize = (obj) => {
		const values = model.typedArrayHandler.read(obj.id);
		const { name, numberOfTuples } = obj;
		return FACTORY[obj.vtkClass].newInstance({
			name,
			numberOfTuples,
			values
		});
	};
}
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, ["typedArrayHandler"]);
	vtkArraySerializer(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkArraySerializer");
var ArraySerializer_default = {
	newInstance,
	extend
};
//#endregion
export { ArraySerializer_default as default, extend, newInstance };
