// @kitware/vtk.js/IO/Core.js
import DataAccessHelper_default from "./Core/DataAccessHelper.js";
import "./Core/DataAccessHelper/HtmlDataAccessHelper.js";
import "./Core/DataAccessHelper/HttpDataAccessHelper.js";
import "./Core/DataAccessHelper/JSZipDataAccessHelper.js";
import BinaryHelper_default from "./Core/BinaryHelper.js";
import HttpDataSetReader_default from "./Core/HttpDataSetReader.js";
import HttpSceneLoader_default from "./Core/HttpSceneLoader.js";
import ImageStream_default from "./Core/ImageStream.js";
import ResourceLoader_default from "./Core/ResourceLoader.js";
import WSLinkClient_default from "./Core/WSLinkClient.js";
//#region Sources/IO/Core/index.js
var Core_default = {
	BinaryHelper: BinaryHelper_default,
	DataAccessHelper: DataAccessHelper_default,
	vtkHttpDataSetReader: HttpDataSetReader_default,
	vtkHttpSceneLoader: HttpSceneLoader_default,
	vtkImageStream: ImageStream_default,
	vtkResourceLoader: ResourceLoader_default,
	vtkWSLinkClient: WSLinkClient_default
};
//#endregion
export { Core_default as default };
