// @kitware/vtk.js/IO/Misc.js
import OBJReader_default from "./Misc/OBJReader.js";
import ElevationReader_default from "./Misc/ElevationReader.js";
import ITKImageReader_default from "./Misc/ITKImageReader.js";
import ITKPolyDataReader_default from "./Misc/ITKPolyDataReader.js";
import JSONNucleoReader_default from "./Misc/JSONNucleoReader.js";
import JSONReader_default from "./Misc/JSONReader.js";
import MTLReader_default from "./Misc/MTLReader.js";
import OBJWriter_default from "./Misc/OBJWriter.js";
import PDBReader_default from "./Misc/PDBReader.js";
import SkyboxReader_default from "./Misc/SkyboxReader.js";
import GCodeReader_default from "./Misc/GCodeReader.js";
//#region Sources/IO/Misc/index.js
var Misc_default = {
	vtkElevationReader: ElevationReader_default,
	vtkITKImageReader: ITKImageReader_default,
	vtkITKPolyDataReader: ITKPolyDataReader_default,
	vtkJSONNucleoReader: JSONNucleoReader_default,
	vtkJSONReader: JSONReader_default,
	vtkMTLReader: MTLReader_default,
	vtkOBJReader: OBJReader_default,
	vtkOBJWriter: OBJWriter_default,
	vtkPDBReader: PDBReader_default,
	vtkSkyboxReader: SkyboxReader_default,
	vtkGCodeReader: GCodeReader_default
};
//#endregion
export { Misc_default as default };
