// @kitware/vtk.js/Widgets/Widgets3D/InteractiveOrientationWidget/state.js
import StateBuilder_default from "../../Core/StateBuilder.js";
//#region Sources/Widgets/Widgets3D/InteractiveOrientationWidget/state.js
var INITIAL_POINTS = [
	[
		-1,
		-1,
		-1
	],
	[
		-1,
		1,
		-1
	],
	[
		1,
		-1,
		-1
	],
	[
		1,
		1,
		-1
	],
	[
		-1,
		-1,
		1
	],
	[
		-1,
		1,
		1
	],
	[
		1,
		-1,
		1
	],
	[
		1,
		1,
		1
	]
];
function generateState() {
	return StateBuilder_default.createBuilder().addStateFromMixin({
		labels: ["handles", "---"],
		mixins: ["origin"],
		name: "handle",
		initialValues: { origin: INITIAL_POINTS[0] }
	}).addStateFromMixin({
		labels: ["handles", "-+-"],
		mixins: ["origin"],
		name: "handle",
		initialValues: { origin: INITIAL_POINTS[1] }
	}).addStateFromMixin({
		labels: ["handles", "+--"],
		mixins: ["origin"],
		name: "handle",
		initialValues: { origin: INITIAL_POINTS[2] }
	}).addStateFromMixin({
		labels: ["handles", "++-"],
		mixins: ["origin"],
		name: "handle",
		initialValues: { origin: INITIAL_POINTS[3] }
	}).addStateFromMixin({
		labels: ["handles", "--+"],
		mixins: ["origin"],
		name: "handle",
		initialValues: { origin: INITIAL_POINTS[4] }
	}).addStateFromMixin({
		labels: ["handles", "-++"],
		mixins: ["origin"],
		name: "handle",
		initialValues: { origin: INITIAL_POINTS[5] }
	}).addStateFromMixin({
		labels: ["handles", "+-+"],
		mixins: ["origin"],
		name: "handle",
		initialValues: { origin: INITIAL_POINTS[6] }
	}).addStateFromMixin({
		labels: ["handles", "+++"],
		mixins: ["origin"],
		name: "handle",
		initialValues: { origin: INITIAL_POINTS[7] }
	}).build("orientation", "name");
}
//#endregion
export { INITIAL_POINTS, generateState as default, generateState };
