// @kitware/vtk.js/Interaction/UI/CornerAnnotation/CornerAnnotation.module.css.js
if (typeof document !== 'undefined') {
  var __vtkStyle = document.createElement('style');
  __vtkStyle.textContent = "._container_1x228_1 {\n  display: flex;\n  align-items: stretch;\n  justify-content: space-between;\n  flex-direction: column;\n  -webkit-user-select: none;\n     -moz-user-select: none;\n          user-select: none;\n  z-index: 100;\n  position: absolute;\n  top: 0;\n  left: 0;\n  width: 100%;\n  height: 100%;\n}\n\n._row_1x228_15 {\n  flex: 1;\n  display: flex;\n  justify-content: space-between;\n  flex-direction: row;\n}\n\n._topRow_1x228_22 {\n  align-items: flex-start;\n}\n\n._middleRow_1x228_27 {\n  align-items: center;\n}\n\n._bottomRow_1x228_32 {\n  align-items: flex-end;\n}\n\n._item_1x228_37 {\n  flex: 1;\n  padding: 5px;\n}\n\n._east_1x228_42 {\n  text-align: right;\n}\n\n._west_1x228_47 {\n  text-align: left;\n}\n\n._north_1x228_52 {\n  text-align: center;\n}\n\n._south_1x228_57 {\n  text-align: center;\n}\n\n._northEast_1x228_62 {\n}\n\n._northWest_1x228_66 {\n}\n\n._southEast_1x228_70 {\n}\n\n._southWest_1x228_74 {\n}\n\n";
  (document.head || document.getElementsByTagName('head')[0]).appendChild(__vtkStyle);
}
var CornerAnnotation_module_default = {
	container: "_container_1x228_1",
	row: "_row_1x228_15",
	topRow: "_topRow_1x228_22 _row_1x228_15",
	middleRow: "_middleRow_1x228_27 _row_1x228_15",
	bottomRow: "_bottomRow_1x228_32 _row_1x228_15",
	item: "_item_1x228_37",
	east: "_east_1x228_42 _item_1x228_37",
	west: "_west_1x228_47 _item_1x228_37",
	north: "_north_1x228_52 _item_1x228_37",
	south: "_south_1x228_57 _item_1x228_37",
	northEast: "_northEast_1x228_62 _east_1x228_42 _item_1x228_37",
	northWest: "_northWest_1x228_66 _west_1x228_47 _item_1x228_37",
	southEast: "_southEast_1x228_70 _east_1x228_42 _item_1x228_37",
	southWest: "_southWest_1x228_74 _west_1x228_47 _item_1x228_37"
};
//#endregion
export { CornerAnnotation_module_default as default };
