// @kitware/vtk.js/Interaction/Manipulators.js
import CompositeCameraManipulator_default from "./Manipulators/CompositeCameraManipulator.js";
import CompositeGestureManipulator_default from "./Manipulators/CompositeGestureManipulator.js";
import CompositeKeyboardManipulator_default from "./Manipulators/CompositeKeyboardManipulator.js";
import CompositeMouseManipulator_default from "./Manipulators/CompositeMouseManipulator.js";
import CompositeVRManipulator_default from "./Manipulators/CompositeVRManipulator.js";
import GestureCameraManipulator_default from "./Manipulators/GestureCameraManipulator.js";
import KeyboardCameraManipulator_default from "./Manipulators/KeyboardCameraManipulator.js";
import MouseBoxSelectorManipulator_default from "./Manipulators/MouseBoxSelectorManipulator.js";
import MouseCameraAxisRotateManipulator_default from "./Manipulators/MouseCameraAxisRotateManipulator.js";
import MouseCameraSliceManipulator_default from "./Manipulators/MouseCameraSliceManipulator.js";
import MouseCameraUnicamRotateManipulator_default from "./Manipulators/MouseCameraUnicamRotateManipulator.js";
import MouseCameraUnicamManipulator_default from "./Manipulators/MouseCameraUnicamManipulator.js";
import MouseCameraTrackballFirstPersonManipulator_default from "./Manipulators/MouseCameraTrackballFirstPersonManipulator.js";
import MouseCameraTrackballRotateManipulator_default from "./Manipulators/MouseCameraTrackballRotateManipulator.js";
import MouseCameraTrackballRollManipulator_default from "./Manipulators/MouseCameraTrackballRollManipulator.js";
import MouseCameraTrackballMultiRotateManipulator_default from "./Manipulators/MouseCameraTrackballMultiRotateManipulator.js";
import MouseCameraTrackballPanManipulator_default from "./Manipulators/MouseCameraTrackballPanManipulator.js";
import MouseCameraTrackballZoomManipulator_default from "./Manipulators/MouseCameraTrackballZoomManipulator.js";
import MouseCameraTrackballZoomToMouseManipulator_default from "./Manipulators/MouseCameraTrackballZoomToMouseManipulator.js";
import MouseRangeManipulator_default from "./Manipulators/MouseRangeManipulator.js";
import VRButtonPanManipulator_default from "./Manipulators/VRButtonPanManipulator.js";
//#region Sources/Interaction/Manipulators/index.js
var Manipulators_default = {
	vtkCompositeCameraManipulator: CompositeCameraManipulator_default,
	vtkCompositeGestureManipulator: CompositeGestureManipulator_default,
	vtkCompositeKeyboardManipulator: CompositeKeyboardManipulator_default,
	vtkCompositeMouseManipulator: CompositeMouseManipulator_default,
	vtkCompositeVRManipulator: CompositeVRManipulator_default,
	vtkGestureCameraManipulator: GestureCameraManipulator_default,
	vtkKeyboardCameraManipulator: KeyboardCameraManipulator_default,
	vtkMouseBoxSelectorManipulator: MouseBoxSelectorManipulator_default,
	vtkMouseCameraAxisRotateManipulator: MouseCameraAxisRotateManipulator_default,
	vtkMouseCameraSliceManipulator: MouseCameraSliceManipulator_default,
	vtkMouseCameraUnicamManipulator: MouseCameraUnicamManipulator_default,
	vtkMouseCameraUnicamRotateManipulator: MouseCameraUnicamRotateManipulator_default,
	vtkMouseCameraTrackballFirstPersonManipulator: MouseCameraTrackballFirstPersonManipulator_default,
	vtkMouseCameraTrackballMultiRotateManipulator: MouseCameraTrackballMultiRotateManipulator_default,
	vtkMouseCameraTrackballPanManipulator: MouseCameraTrackballPanManipulator_default,
	vtkMouseCameraTrackballRollManipulator: MouseCameraTrackballRollManipulator_default,
	vtkMouseCameraTrackballRotateManipulator: MouseCameraTrackballRotateManipulator_default,
	vtkMouseCameraTrackballZoomManipulator: MouseCameraTrackballZoomManipulator_default,
	vtkMouseCameraTrackballZoomToMouseManipulator: MouseCameraTrackballZoomToMouseManipulator_default,
	vtkMouseRangeManipulator: MouseRangeManipulator_default,
	vtkVRButtonPanManipulator: VRButtonPanManipulator_default
};
//#endregion
export { Manipulators_default as default };
