// @kitware/vtk.js/IO/XML/XMLImageDataWriter.js
import macros_default from "../../macros.js";
import XMLWriter_default from "./XMLWriter.js";
//#region Sources/IO/XML/XMLImageDataWriter/index.js
function vtkXMLImageDataWriter(publicAPI, model) {
	model.classHierarchy.push("vtkXMLImageDataWriter");
	const superClass = { ...publicAPI };
	publicAPI.create = (dataObject) => {
		const parent = superClass.create(dataObject);
		const piece = parent.ele("ImageData", {
			WholeExtent: dataObject.getExtent().join(" "),
			Origin: dataObject.getOrigin().join(" "),
			Spacing: dataObject.getSpacing().join(" "),
			Direction: dataObject.getDirection().join(" ")
		}).ele("Piece", { Extent: dataObject.getExtent().join(" ") });
		publicAPI.processDataSetAttributes(piece, "PointData", dataObject.getPointData());
		publicAPI.processDataSetAttributes(piece, "CellData", dataObject.getCellData());
		return parent;
	};
}
var DEFAULT_VALUES = { dataType: "ImageData" };
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	XMLWriter_default.extend(publicAPI, model, initialValues);
	vtkXMLImageDataWriter(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkXMLImageDataWriter");
var XMLImageDataWriter_default = {
	newInstance,
	extend
};
//#endregion
export { XMLImageDataWriter_default as default, extend, newInstance };
