// @kitware/vtk.js/IO/XML/XMLPolyDataWriter.js
import macros_default from "../../macros.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import { POLYDATA_FIELDS } from "../../Common/DataModel/PolyData/Constants.js";
import XMLWriter_default from "./XMLWriter.js";
//#region Sources/IO/XML/XMLPolyDataWriter/index.js
function vtkXMLPolyDataWriter(publicAPI, model) {
	model.classHierarchy.push("vtkXMLPolyDataWriter");
	const superClass = { ...publicAPI };
	function camelize(str) {
		return str.replace(/(?:^\w|[A-Z]|\b\w)/g, (letter) => letter.toUpperCase()).replace(/\s+/g, "");
	}
	publicAPI.create = (dataObject) => {
		const parent = superClass.create(dataObject);
		const piece = parent.ele("PolyData", {}).ele("Piece", {
			NumberOfPoints: dataObject.getPoints().getNumberOfPoints(),
			NumberOfVerts: dataObject.getNumberOfVerts(),
			NumberOfLines: dataObject.getNumberOfLines(),
			NumberOfStrips: dataObject.getNumberOfStrips(),
			NumberOfPolys: dataObject.getNumberOfPolys()
		});
		publicAPI.processDataSetAttributes(piece, "PointData", dataObject.getPointData());
		publicAPI.processDataSetAttributes(piece, "CellData", dataObject.getCellData());
		publicAPI.processDataSetAttributes(piece, "FieldData", dataObject.getFieldData());
		publicAPI.processDataArray(piece.ele("Points"), dataObject.getPoints());
		POLYDATA_FIELDS.forEach((cellType) => {
			const cellTypeName = camelize(cellType);
			const cells = dataObject[`get${cellTypeName}`]();
			const connectivity = [];
			const offsets = [];
			const cellsData = cells.getData();
			let npts = cellsData[0];
			let offset = 0;
			for (let i = 0; i < cellsData.length;) {
				npts = cellsData[i++];
				for (let j = 0; j < npts; ++j) connectivity.push(cellsData[i++]);
				offset += npts;
				offsets.push(offset);
			}
			const connectivityDataArray = DataArray_default.newInstance({
				numberOfComponents: 1,
				name: "connectivity",
				values: Int32Array.from(connectivity)
			});
			const offsetsDataArray = DataArray_default.newInstance({
				numberOfComponents: 1,
				name: "offsets",
				values: Int32Array.from(offsets)
			});
			const cellEle = piece.ele(cellTypeName);
			publicAPI.processDataArray(cellEle, connectivityDataArray);
			publicAPI.processDataArray(cellEle, offsetsDataArray);
		});
		return parent;
	};
}
var DEFAULT_VALUES = { dataType: "PolyData" };
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	XMLWriter_default.extend(publicAPI, model, initialValues);
	vtkXMLPolyDataWriter(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkXMLPolyDataWriter");
var XMLPolyDataWriter_default = {
	newInstance,
	extend
};
//#endregion
export { XMLPolyDataWriter_default as default, extend, newInstance };
