// @kitware/vtk.js/IO/Core/DataAccessHelper/HtmlDataAccessHelper.js
import macros_default from "../../../macros.js";
import Base64_default from "../../../Common/Core/Base64.js";
import { DataTypeByteSize } from "../../../Common/Core/DataArray/Constants.js";
import Endian_default from "../../../Common/Core/Endian.js";
import { registerType } from "../DataAccessHelper.js";
import { decompressSync, strFromU8 } from "fflate";
//#region Sources/IO/Core/DataAccessHelper/HtmlDataAccessHelper.js
var { vtkErrorMacro, vtkDebugMacro } = macros_default;
var requestCount = 0;
function getContent(url) {
	const el = document.querySelector(`.webResource[data-url="${url}"]`);
	return el ? el.innerHTML : null;
}
function getElement(url) {
	return document.querySelector(`.webResource[data-url="${url}"]`);
}
function removeLeadingSlash(str) {
	return str[0] === "/" ? str.substr(1) : str;
}
function fetchText(instance, url, options = {}) {
	return new Promise((resolve, reject) => {
		const txt = getContent(url);
		if (txt === null) reject(/* @__PURE__ */ new Error(`No such text ${url}`));
		else resolve(txt);
	});
}
function fetchJSON(instance, url, options = {}) {
	return new Promise((resolve, reject) => {
		const txt = getContent(removeLeadingSlash(url));
		if (txt === null) reject(/* @__PURE__ */ new Error(`No such JSON ${url}`));
		else resolve(JSON.parse(txt));
	});
}
function fetchArray(instance, baseURL, array, options = {}) {
	return new Promise((resolve, reject) => {
		const url = removeLeadingSlash([
			baseURL,
			array.ref.basepath,
			options.compression ? `${array.ref.id}.gz` : array.ref.id
		].join("/"));
		const txt = getContent(url);
		if (txt === null) reject(/* @__PURE__ */ new Error(`No such array ${url}`));
		else {
			if (array.dataType === "string") {
				let bText = atob(txt);
				if (options.compression) bText = strFromU8(decompressSync(bText));
				array.values = JSON.parse(bText);
			} else {
				const uint8array = new Uint8Array(Base64_default.toArrayBuffer(txt));
				array.buffer = new ArrayBuffer(uint8array.length);
				new Uint8Array(array.buffer).set(uint8array);
				if (options.compression) if (array.dataType === "string" || array.dataType === "JSON") array.buffer = strFromU8(decompressSync(new Uint8Array(array.buffer)));
				else array.buffer = decompressSync(new Uint8Array(array.buffer)).buffer;
				if (array.ref.encode === "JSON") array.values = JSON.parse(array.buffer);
				else {
					if (Endian_default.ENDIANNESS !== array.ref.encode && Endian_default.ENDIANNESS) {
						vtkDebugMacro(`Swap bytes of ${array.name}`);
						Endian_default.swapBytes(array.buffer, DataTypeByteSize[array.dataType]);
					}
					array.values = macros_default.newTypedArray(array.dataType, array.buffer);
				}
				if (array.values.length !== array.size) vtkErrorMacro(`Error in FetchArray: ${array.name} does not have the proper array size. Got ${array.values.length}, instead of ${array.size}`);
			}
			delete array.ref;
			if (--requestCount === 0 && instance?.invokeBusy) instance.invokeBusy(false);
			if (instance?.modified) instance.modified();
			resolve(array);
		}
	});
}
function fetchImage(instance, url, options = {}) {
	return new Promise((resolve, reject) => {
		const img = getElement(url);
		if (img) resolve(img);
		else reject(/* @__PURE__ */ new Error(`No such image ${url}`));
	});
}
var HtmlDataAccessHelper = {
	fetchJSON,
	fetchText,
	fetchArray,
	fetchImage
};
registerType("html", (options) => HtmlDataAccessHelper);
//#endregion
export { HtmlDataAccessHelper as default };
