// @kitware/vtk.js/IO/Core/DataAccessHelper/LiteHttpDataAccessHelper.js
import macros_default from "../../../macros.js";
import { DataTypeByteSize } from "../../../Common/Core/DataArray/Constants.js";
import Endian_default from "../../../Common/Core/Endian.js";
import { has, registerType } from "../DataAccessHelper.js";
//#region Sources/IO/Core/DataAccessHelper/LiteHttpDataAccessHelper.js
var { vtkErrorMacro, vtkDebugMacro } = macros_default;
var REJECT_COMPRESSION = () => {
	vtkErrorMacro("LiteHttpDataAccessHelper does not support compression. Need to register HttpDataAccessHelper instead.");
	return Promise.reject(/* @__PURE__ */ new Error("LiteHttpDataAccessHelper does not support compression. Need to register HttpDataAccessHelper instead."));
};
var requestCount = 0;
function openAsyncXHR(method, url, options = {}) {
	const xhr = new XMLHttpRequest();
	xhr.open(method, url, true);
	if (options.headers) Object.entries(options.headers).forEach(([key, value]) => xhr.setRequestHeader(key, value));
	if (options.progressCallback) xhr.addEventListener("progress", options.progressCallback);
	return xhr;
}
function fetchBinary(url, options = {}) {
	return new Promise((resolve, reject) => {
		const xhr = openAsyncXHR("GET", url, options);
		xhr.onreadystatechange = (e) => {
			if (xhr.readyState === 4) if (xhr.status === 200 || xhr.status === 0) resolve(xhr.response);
			else reject({
				xhr,
				e
			});
		};
		xhr.responseType = "arraybuffer";
		xhr.send();
	});
}
function fetchArray(instance, baseURL, array, options = {}) {
	if (options && options.compression) return REJECT_COMPRESSION();
	if (array.ref && !array.ref.pending) return new Promise((resolve, reject) => {
		const xhr = openAsyncXHR("GET", [
			baseURL,
			array.ref.basepath,
			array.ref.id
		].join("/"), options);
		xhr.onreadystatechange = (e) => {
			if (xhr.readyState === 1) {
				array.ref.pending = true;
				if (++requestCount === 1 && instance?.invokeBusy) instance.invokeBusy(true);
			}
			if (xhr.readyState === 4) {
				array.ref.pending = false;
				if (xhr.status === 200 || xhr.status === 0) {
					array.buffer = xhr.response;
					if (array.ref.encode === "JSON") array.values = JSON.parse(array.buffer);
					else {
						if (Endian_default.ENDIANNESS !== array.ref.encode && Endian_default.ENDIANNESS) {
							vtkDebugMacro(`Swap bytes of ${array.name}`);
							Endian_default.swapBytes(array.buffer, DataTypeByteSize[array.dataType]);
						}
						array.values = macros_default.newTypedArray(array.dataType, array.buffer);
					}
					if (array.values.length !== array.size) vtkErrorMacro(`Error in FetchArray: ${array.name}, does not have the proper array size. Got ${array.values.length}, instead of ${array.size}`);
					delete array.ref;
					if (--requestCount === 0 && instance?.invokeBusy) instance.invokeBusy(false);
					if (instance?.modified) instance.modified();
					resolve(array);
				} else reject({
					xhr,
					e
				});
			}
		};
		xhr.responseType = array.dataType !== "string" ? "arraybuffer" : "text";
		xhr.send();
	});
	return Promise.resolve(array);
}
function fetchJSON(instance, url, options = {}) {
	if (options && options.compression) return REJECT_COMPRESSION();
	return new Promise((resolve, reject) => {
		const xhr = openAsyncXHR("GET", url, options);
		xhr.onreadystatechange = (e) => {
			if (xhr.readyState === 1) {
				if (++requestCount === 1 && instance?.invokeBusy) instance.invokeBusy(true);
			}
			if (xhr.readyState === 4) {
				if (--requestCount === 0 && instance?.invokeBusy) instance.invokeBusy(false);
				if (xhr.status === 200 || xhr.status === 0) resolve(JSON.parse(xhr.responseText));
				else reject({
					xhr,
					e
				});
			}
		};
		xhr.responseType = "text";
		xhr.send();
	});
}
function fetchText(instance, url, options = {}) {
	if (options && options.compression) return REJECT_COMPRESSION();
	return new Promise((resolve, reject) => {
		const xhr = openAsyncXHR("GET", url, options);
		xhr.onreadystatechange = (e) => {
			if (xhr.readyState === 1) {
				if (++requestCount === 1 && instance?.invokeBusy) instance.invokeBusy(true);
			}
			if (xhr.readyState === 4) {
				if (--requestCount === 0 && instance?.invokeBusy) instance.invokeBusy(false);
				if (xhr.status === 200 || xhr.status === 0) resolve(xhr.responseText);
				else reject({
					xhr,
					e
				});
			}
		};
		xhr.responseType = "text";
		xhr.send();
	});
}
function fetchImage(instance, url, options = {}) {
	return new Promise((resolve, reject) => {
		const img = new Image();
		if (options.crossOrigin) img.crossOrigin = options.crossOrigin;
		img.onload = () => resolve(img);
		img.onerror = reject;
		img.src = url;
	});
}
var LiteHttpDataAccessHelper = {
	fetchArray,
	fetchJSON,
	fetchText,
	fetchBinary,
	fetchImage
};
if (!has("http")) registerType("http", (options) => LiteHttpDataAccessHelper);
//#endregion
export { LiteHttpDataAccessHelper as default };
