// @kitware/vtk.js/Common/Core.js
import Base64_default from "./Core/Base64.js";
import { Math_exports } from "./Core/Math.js";
import DataArray_default from "./Core/DataArray.js";
import CellArray_default from "./Core/CellArray.js";
import Endian_default from "./Core/Endian.js";
import HalfFloat_default from "./Core/HalfFloat.js";
import ImageHelper_default from "./Core/ImageHelper.js";
import ScalarsToColors_default from "./Core/ScalarsToColors.js";
import LookupTable_default from "./Core/LookupTable.js";
import MatrixBuilder_default from "./Core/MatrixBuilder.js";
import Points_default from "./Core/Points.js";
import ProgressHandler_default from "./Core/ProgressHandler.js";
import StringArray_default from "./Core/StringArray.js";
import URLExtract_default from "./Core/URLExtract.js";
import VariantArray_default from "./Core/VariantArray.js";
//#region Sources/Common/Core/index.js
var Core_default = {
	vtkBase64: Base64_default,
	vtkCellArray: CellArray_default,
	vtkDataArray: DataArray_default,
	vtkEndian: Endian_default,
	vtkHalfFloat: HalfFloat_default,
	vtkImageHelper: ImageHelper_default,
	vtkLookupTable: LookupTable_default,
	vtkMath: Math_exports,
	vtkMatrixBuilder: MatrixBuilder_default,
	vtkPoints: Points_default,
	vtkProgressHandler: ProgressHandler_default,
	vtkScalarsToColors: ScalarsToColors_default,
	vtkStringArray: StringArray_default,
	vtkURLExtract: URLExtract_default,
	vtkVariantArray: VariantArray_default
};
//#endregion
export { Core_default as default };
