// @kitware/vtk.js/Common/DataModel.js
import Plane_default from "./DataModel/Plane.js";
import BoundingBox_default from "./DataModel/BoundingBox.js";
import DataSetAttributes_default from "./DataModel/DataSetAttributes.js";
import DataSet_default from "./DataModel/DataSet.js";
import StructuredData_default from "./DataModel/StructuredData.js";
import ImageData_default from "./DataModel/ImageData.js";
import Box_default from "./DataModel/Box.js";
import Cell_default from "./DataModel/Cell.js";
import Cone_default from "./DataModel/Cone.js";
import Cylinder_default from "./DataModel/Cylinder.js";
import ImplicitBoolean_default from "./DataModel/ImplicitBoolean.js";
import IncrementalPointLocator_default from "./DataModel/IncrementalPointLocator.js";
import Line_default from "./DataModel/Line.js";
import PointSet_default from "./DataModel/PointSet.js";
import Triangle_default from "./DataModel/Triangle.js";
import TriangleStrip_default from "./DataModel/TriangleStrip.js";
import PolyData_default from "./DataModel/PolyData.js";
import ITKHelper_default from "./DataModel/ITKHelper.js";
import Molecule_default from "./DataModel/Molecule.js";
import PiecewiseFunction_default from "./DataModel/PiecewiseFunction.js";
import Planes_default from "./DataModel/Planes.js";
import SelectionNode_default from "./DataModel/SelectionNode.js";
import Sphere_default from "./DataModel/Sphere.js";
//#region Sources/Common/DataModel/index.js
var DataModel_default = {
	vtkBoundingBox: BoundingBox_default,
	vtkBox: Box_default,
	vtkCell: Cell_default,
	vtkCone: Cone_default,
	vtkCylinder: Cylinder_default,
	vtkDataSet: DataSet_default,
	vtkDataSetAttributes: DataSetAttributes_default,
	vtkITKHelper: ITKHelper_default,
	vtkImageData: ImageData_default,
	vtkImplicitBoolean: ImplicitBoolean_default,
	vtkIncrementalPointLocator: IncrementalPointLocator_default,
	vtkLine: Line_default,
	vtkMolecule: Molecule_default,
	vtkPiecewiseFunction: PiecewiseFunction_default,
	vtkPlane: Plane_default,
	vtkPlanes: Planes_default,
	vtkPointSet: PointSet_default,
	vtkPolyData: PolyData_default,
	vtkSelectionNode: SelectionNode_default,
	vtkSphere: Sphere_default,
	vtkStructuredData: StructuredData_default,
	vtkTriangle: Triangle_default,
	vtkTriangleStrip: TriangleStrip_default
};
//#endregion
export { DataModel_default as default };
