// @kitware/vtk.js/Widgets/Widgets3D/AngleWidget.js
import macros_default from "../../macros.js";
import { angleBetweenVectors, subtract } from "../../Common/Core/Math.js";
import { ViewTypes } from "../Core/WidgetManager/Constants.js";
import AbstractWidgetFactory_default from "../Core/AbstractWidgetFactory.js";
import PlaneManipulator_default from "../Manipulators/PlaneManipulator.js";
import PolyLineRepresentation_default from "../Representations/PolyLineRepresentation.js";
import SphereHandleRepresentation_default from "../Representations/SphereHandleRepresentation.js";
import widgetBehavior from "./AngleWidget/behavior.js";
import generateState from "./AngleWidget/state.js";
//#region Sources/Widgets/Widgets3D/AngleWidget/index.js
function vtkAngleWidget(publicAPI, model) {
	model.classHierarchy.push("vtkAngleWidget");
	model.methodsToLink = [
		"activeScaleFactor",
		"activeColor",
		"useActiveColor",
		"glyphResolution",
		"defaultScale",
		"scaleInPixels"
	];
	model._onManipulatorChanged = () => {
		model.widgetState.getMoveHandle().setManipulator(model.manipulator);
		model.widgetState.getHandleList().forEach((handle) => {
			handle.setManipulator(model.manipulator);
		});
	};
	publicAPI.getRepresentationsForViewType = (viewType) => {
		switch (viewType) {
			case ViewTypes.DEFAULT:
			case ViewTypes.GEOMETRY:
			case ViewTypes.SLICE:
			case ViewTypes.VOLUME:
			default: return [{
				builder: SphereHandleRepresentation_default,
				labels: ["handles", "moveHandle"]
			}, {
				builder: PolyLineRepresentation_default,
				labels: ["handles", "moveHandle"]
			}];
		}
	};
	publicAPI.getAngle = () => {
		const handles = model.widgetState.getHandleList();
		if (handles.length !== 3) return 0;
		if (!handles[0].getOrigin() || !handles[1].getOrigin() || !handles[2].getOrigin()) return 0;
		const vec1 = [
			0,
			0,
			0
		];
		const vec2 = [
			0,
			0,
			0
		];
		subtract(handles[0].getOrigin(), handles[1].getOrigin(), vec1);
		subtract(handles[2].getOrigin(), handles[1].getOrigin(), vec2);
		return angleBetweenVectors(vec1, vec2);
	};
	model.widgetState.onBoundsChange((bounds) => {
		const center = [
			(bounds[0] + bounds[1]) * .5,
			(bounds[2] + bounds[3]) * .5,
			(bounds[4] + bounds[5]) * .5
		];
		model.widgetState.getMoveHandle().setOrigin(center);
	});
	publicAPI.setManipulator(model.manipulator || PlaneManipulator_default.newInstance({ useCameraNormal: true }));
}
var defaultValues = (initialValues) => ({
	behavior: widgetBehavior,
	widgetState: generateState(),
	...initialValues
});
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, defaultValues(initialValues));
	AbstractWidgetFactory_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, ["manipulator"]);
	vtkAngleWidget(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkAngleWidget");
var AngleWidget_default = {
	newInstance,
	extend
};
//#endregion
export { AngleWidget_default as default, extend, newInstance };
