// @kitware/vtk.js/Widgets/Widgets3D/EllipseWidget.js
import macros_default from "../../macros.js";
import { ViewTypes } from "../Core/WidgetManager/Constants.js";
import PlaneManipulator_default from "../Manipulators/PlaneManipulator.js";
import CircleContextRepresentation_default from "../Representations/CircleContextRepresentation.js";
import SphereHandleRepresentation_default from "../Representations/SphereHandleRepresentation.js";
import { BehaviorCategory, ShapeBehavior } from "./ShapeWidget/Constants.js";
import ShapeWidget_default from "./ShapeWidget.js";
import widgetBehavior from "./EllipseWidget/behavior.js";
import generateState from "./EllipseWidget/state.js";
//#region Sources/Widgets/Widgets3D/EllipseWidget/index.js
function vtkEllipseWidget(publicAPI, model) {
	model.classHierarchy.push("vtkEllipseWidget");
	model.methodsToLink = [
		...model.methodsToLink,
		"activeScaleFactor",
		"activeColor",
		"useActiveColor",
		"glyphResolution",
		"defaultScale",
		"drawBorder",
		"drawFace",
		"opacity"
	];
	publicAPI.getRepresentationsForViewType = (viewType) => {
		switch (viewType) {
			case ViewTypes.DEFAULT:
			case ViewTypes.GEOMETRY:
			case ViewTypes.SLICE:
			case ViewTypes.VOLUME:
			default: return [{
				builder: SphereHandleRepresentation_default,
				labels: ["moveHandle"]
			}, {
				builder: CircleContextRepresentation_default,
				labels: ["ellipseHandle"]
			}];
		}
	};
	publicAPI.setManipulator(model.manipulator || PlaneManipulator_default.newInstance({ useCameraNormal: true }));
}
function defaultValues(initialValues) {
	return {
		behavior: widgetBehavior,
		widgetState: generateState(),
		modifierBehavior: {
			None: {
				[BehaviorCategory.PLACEMENT]: ShapeBehavior[BehaviorCategory.PLACEMENT].CLICK_AND_DRAG,
				[BehaviorCategory.POINTS]: ShapeBehavior[BehaviorCategory.POINTS].CENTER_TO_CORNER,
				[BehaviorCategory.RATIO]: ShapeBehavior[BehaviorCategory.RATIO].FREE
			},
			Shift: { [BehaviorCategory.RATIO]: ShapeBehavior[BehaviorCategory.RATIO].FIXED },
			Control: { [BehaviorCategory.POINTS]: ShapeBehavior[BehaviorCategory.POINTS].CORNER_TO_CORNER }
		},
		...initialValues
	};
}
function extend(publicAPI, model, initialValues = {}) {
	ShapeWidget_default.extend(publicAPI, model, defaultValues(initialValues));
	macros_default.setGet(publicAPI, model, ["widgetState"]);
	vtkEllipseWidget(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkEllipseWidget");
var EllipseWidget_default = {
	newInstance,
	extend
};
//#endregion
export { EllipseWidget_default as default, extend, newInstance };
