// @kitware/vtk.js/Widgets/Widgets3D/ImageCroppingWidget.js
import macros_default from "../../macros.js";
import { ViewTypes } from "../Core/WidgetManager/Constants.js";
import AbstractWidgetFactory_default from "../Core/AbstractWidgetFactory.js";
import LineManipulator_default from "../Manipulators/LineManipulator.js";
import PlaneManipulator_default from "../Manipulators/PlaneManipulator.js";
import SphereHandleRepresentation_default from "../Representations/SphereHandleRepresentation.js";
import CroppingOutlineRepresentation_default from "../Representations/CroppingOutlineRepresentation.js";
import { AXES, transformVec3 } from "./ImageCroppingWidget/helpers.js";
import widgetBehavior from "./ImageCroppingWidget/behavior.js";
import build from "./ImageCroppingWidget/state.js";
//#region Sources/Widgets/Widgets3D/ImageCroppingWidget/index.js
function vtkImageCroppingWidget(publicAPI, model) {
	model.classHierarchy.push("vtkImageCroppingWidget");
	const superClass = { ...publicAPI };
	let stateSub = null;
	function setHandlesEnabled(label, flag) {
		model.widgetState.getStatesWithLabel(label).forEach((handle) => {
			handle.setVisible(flag);
		});
	}
	publicAPI.setFaceHandlesEnabled = (flag) => setHandlesEnabled("faces", flag);
	publicAPI.setEdgeHandlesEnabled = (flag) => setHandlesEnabled("edges", flag);
	publicAPI.setCornerHandlesEnabled = (flag) => setHandlesEnabled("corners", flag);
	publicAPI.copyImageDataDescription = (im) => {
		model.widgetState.setIndexToWorldT(...im.getIndexToWorld());
		model.widgetState.setWorldToIndexT(...im.getWorldToIndex());
		const dims = im.getDimensions();
		model.widgetState.getCroppingPlanes().setPlanes([
			0,
			dims[0],
			0,
			dims[1],
			0,
			dims[2]
		]);
		publicAPI.modified();
	};
	publicAPI.updateHandles = () => {
		const planes = model.widgetState.getCroppingPlanes().getPlanes();
		const midpts = [
			(planes[0] + planes[1]) / 2,
			(planes[2] + planes[3]) / 2,
			(planes[4] + planes[5]) / 2
		];
		const iAxis = [
			planes[0],
			midpts[0],
			planes[1]
		];
		const jAxis = [
			planes[2],
			midpts[1],
			planes[3]
		];
		const kAxis = [
			planes[4],
			midpts[2],
			planes[5]
		];
		const indexToWorldT = model.widgetState.getIndexToWorldT();
		const getAxis = (a) => AXES[a];
		for (let i = 0; i < 3; i++) for (let j = 0; j < 3; j++) for (let k = 0; k < 3; k++) if (i !== 1 || j !== 1 || k !== 1) {
			const name = [
				i,
				j,
				k
			].map(getAxis).join("");
			const coord = transformVec3([
				iAxis[i],
				jAxis[j],
				kAxis[k]
			], indexToWorldT);
			const [handle] = model.widgetState.getStatesWithLabel(name);
			handle.setOrigin(...coord);
		}
	};
	publicAPI.delete = macros_default.chain(publicAPI.delete, () => {
		if (stateSub) stateSub.unsubscribe();
	});
	model.methodsToLink = ["scaleInPixels"];
	publicAPI.getRepresentationsForViewType = (viewType) => {
		switch (viewType) {
			case ViewTypes.DEFAULT:
			case ViewTypes.GEOMETRY:
			case ViewTypes.SLICE:
			case ViewTypes.VOLUME:
			default: return [{
				builder: SphereHandleRepresentation_default,
				labels: ["handles"]
			}, {
				builder: CroppingOutlineRepresentation_default,
				labels: ["corners"]
			}];
		}
	};
	stateSub = model.widgetState.getCroppingPlanes().onModified(publicAPI.updateHandles);
	publicAPI.setCornerManipulator = (manipulator) => {
		superClass.setCornerManipulator(manipulator);
		model.widgetState.getStatesWithLabel("corners").forEach((handle) => handle.setManipulator(manipulator));
	};
	publicAPI.setEdgeManipulator = (manipulator) => {
		superClass.setEdgeManipulator(manipulator);
		model.widgetState.getStatesWithLabel("edges").forEach((handle) => handle.setManipulator(manipulator));
	};
	publicAPI.setFaceManipulator = (manipulator) => {
		superClass.setFaceManipulator(manipulator);
		model.widgetState.getStatesWithLabel("faces").forEach((handle) => handle.setManipulator(manipulator));
	};
	publicAPI.setCornerManipulator(PlaneManipulator_default.newInstance({ useCameraNormal: true }));
	publicAPI.setEdgeManipulator(PlaneManipulator_default.newInstance());
	publicAPI.setFaceManipulator(LineManipulator_default.newInstance());
}
var defaultValues = (initialValues) => ({
	behavior: widgetBehavior,
	widgetState: build(),
	...initialValues
});
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, defaultValues(initialValues));
	AbstractWidgetFactory_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, [
		"cornerManipulator",
		"edgeManipulator",
		"faceManipulator"
	]);
	vtkImageCroppingWidget(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkImageCroppingWidget");
var ImageCroppingWidget_default = {
	newInstance,
	extend
};
//#endregion
export { ImageCroppingWidget_default as default, extend, newInstance };
