// @kitware/vtk.js/Widgets/Widgets3D/ImplicitPlaneWidget.js
import macros_default from "../../macros.js";
import { add } from "../../Common/Core/Math.js";
import { ViewTypes } from "../Core/WidgetManager/Constants.js";
import AbstractWidgetFactory_default from "../Core/AbstractWidgetFactory.js";
import LineManipulator_default from "../Manipulators/LineManipulator.js";
import PlaneManipulator_default from "../Manipulators/PlaneManipulator.js";
import TrackballManipulator_default from "../Manipulators/TrackballManipulator.js";
import ImplicitPlaneRepresentation_default from "../Representations/ImplicitPlaneRepresentation.js";
//#region Sources/Widgets/Widgets3D/ImplicitPlaneWidget/index.js
function widgetBehavior(publicAPI, model) {
	model.classHierarchy.push("vtkPlaneWidget");
	model._isDragging = false;
	model._draggingWidgetOrigin = [
		0,
		0,
		0
	];
	publicAPI.setDisplayCallback = (callback) => model.representations[0].setDisplayCallback(callback);
	publicAPI.updateCursor = () => {
		switch (model.activeState.getUpdateMethodName()) {
			case "updateFromOrigin":
				model._apiSpecificRenderWindow.setCursor("crosshair");
				break;
			case "updateFromPlane":
				model._apiSpecificRenderWindow.setCursor("move");
				break;
			case "updateFromNormal":
				model._apiSpecificRenderWindow.setCursor("alias");
				break;
			default:
				model._apiSpecificRenderWindow.setCursor("grabbing");
				break;
		}
	};
	publicAPI.handleLeftButtonPress = (callData) => {
		if (!model.activeState || !model.activeState.getActive() || !model.pickable) return macros_default.VOID;
		model.lineManipulator.setWidgetOrigin(model.activeState.getOrigin());
		model.lineManipulator.setWidgetNormal(model.activeState.getNormal());
		model.planeManipulator.setWidgetOrigin(model.activeState.getOrigin());
		model.planeManipulator.setWidgetNormal(model.activeState.getNormal());
		model.trackballManipulator.reset(callData);
		model.trackballManipulator.setWidgetNormal(model.activeState.getNormal());
		let activeManipulator = null;
		switch (model.activeState.getUpdateMethodName()) {
			case "updateFromOrigin":
				activeManipulator = model.planeManipulator;
				break;
			case "updateFromPlane":
				activeManipulator = model.lineManipulator;
				break;
			case "updateFromNormal":
				activeManipulator = model.trackballManipulator;
				break;
			default:
		}
		if (activeManipulator) activeManipulator.handleEvent(callData, model._apiSpecificRenderWindow);
		if (model.dragable) {
			model._draggingWidgetOrigin = model.widgetState.getOrigin();
			model._isDragging = true;
			model._apiSpecificRenderWindow.setCursor("grabbing");
			model._interactor.requestAnimation(publicAPI);
		}
		publicAPI.invokeStartInteractionEvent();
		return macros_default.EVENT_ABORT;
	};
	publicAPI.handleMouseMove = (callData) => {
		if (model._isDragging) return publicAPI.handleEvent(callData);
		return macros_default.VOID;
	};
	publicAPI.handleLeftButtonRelease = () => {
		if (!model.activeState || !model.activeState.getActive() || !model.pickable) return macros_default.VOID;
		if (model._isDragging) {
			model._interactor.cancelAnimation(publicAPI);
			model._isDragging = false;
		}
		model.widgetState.deactivate();
		publicAPI.invokeEndInteractionEvent();
		return macros_default.EVENT_ABORT;
	};
	publicAPI.handleEvent = (callData) => {
		if (model.pickable && model.activeState && model.activeState.getActive()) {
			publicAPI[model.activeState.getUpdateMethodName()](callData);
			publicAPI.invokeInteractionEvent();
			return macros_default.EVENT_ABORT;
		}
		return macros_default.VOID;
	};
	publicAPI.updateFromOrigin = (callData) => {
		model.planeManipulator.setWidgetNormal(model.widgetState.getNormal());
		const { worldCoords, worldDelta } = model.planeManipulator.handleEvent(callData, model._apiSpecificRenderWindow);
		add(model._draggingWidgetOrigin, worldDelta, model._draggingWidgetOrigin);
		if (model.widgetState.containsPoint(...worldCoords)) model.activeState.setOrigin(model._draggingWidgetOrigin);
	};
	publicAPI.updateFromPlane = (callData) => {
		model.lineManipulator.setWidgetNormal(model.activeState.getNormal());
		const { worldDelta } = model.lineManipulator.handleEvent(callData, model._apiSpecificRenderWindow);
		add(model._draggingWidgetOrigin, worldDelta, model._draggingWidgetOrigin);
		if (model.widgetState.containsPoint(...model._draggingWidgetOrigin)) model.activeState.setOrigin(model._draggingWidgetOrigin);
	};
	publicAPI.updateFromNormal = (callData) => {
		model.trackballManipulator.setWidgetNormal(model.activeState.getNormal());
		const { worldCoords: newNormal } = model.trackballManipulator.handleEvent(callData, model._apiSpecificRenderWindow);
		model.activeState.setNormal(newNormal);
	};
	model.lineManipulator = LineManipulator_default.newInstance();
	model.planeManipulator = PlaneManipulator_default.newInstance();
	model.trackballManipulator = TrackballManipulator_default.newInstance();
}
function vtkImplicitPlaneWidget(publicAPI, model) {
	model.classHierarchy.push("vtkPlaneWidget");
	model.methodsToLink = [
		"representationStyle",
		"sphereResolution",
		"handleSizeRatio",
		"axisScale",
		"normalVisible",
		"originVisible",
		"planeVisible",
		"outlineVisible"
	];
	publicAPI.getRepresentationsForViewType = (viewType) => {
		switch (viewType) {
			case ViewTypes.DEFAULT:
			case ViewTypes.GEOMETRY:
			case ViewTypes.SLICE:
			case ViewTypes.VOLUME:
			default: return [{ builder: ImplicitPlaneRepresentation_default }];
		}
	};
}
var defaultValues = (initialValues) => ({
	behavior: widgetBehavior,
	widgetState: ImplicitPlaneRepresentation_default.generateState(),
	...initialValues
});
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, defaultValues(initialValues));
	AbstractWidgetFactory_default.extend(publicAPI, model, initialValues);
	vtkImplicitPlaneWidget(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkImplicitPlaneWidget");
var ImplicitPlaneWidget_default = {
	newInstance,
	extend
};
//#endregion
export { ImplicitPlaneWidget_default as default, extend, newInstance };
