// @kitware/vtk.js/Widgets/Widgets3D/InteractiveOrientationWidget.js
import macros_default from "../../macros.js";
import { ViewTypes } from "../Core/WidgetManager/Constants.js";
import AbstractWidgetFactory_default from "../Core/AbstractWidgetFactory.js";
import { Behavior } from "../Representations/WidgetRepresentation/Constants.js";
import ConvexFaceContextRepresentation_default from "../Representations/ConvexFaceContextRepresentation.js";
import widgetBehavior from "./InteractiveOrientationWidget/behavior.js";
import generateState, { INITIAL_POINTS } from "./InteractiveOrientationWidget/state.js";
//#region Sources/Widgets/Widgets3D/InteractiveOrientationWidget/index.js
function vtkInteractiveOrientationWidget(publicAPI, model) {
	model.classHierarchy.push("vtkInteractiveOrientationWidget");
	model.methodsToLink = [
		"closePolyLine",
		"activeScaleFactor",
		"activeColor",
		"useActiveColor",
		"glyphResolution",
		"defaultScale"
	];
	publicAPI.setBounds = (bounds) => {
		const handles = model.widgetState.getStatesWithLabel("handles");
		for (let i = 0; i < handles.length; i++) {
			const xyz = INITIAL_POINTS[i];
			const x = xyz[0] > 0 ? bounds[1] : bounds[0];
			const y = xyz[1] > 0 ? bounds[3] : bounds[2];
			const z = xyz[2] > 0 ? bounds[5] : bounds[4];
			handles[i].setOrigin(x, y, z);
		}
	};
	publicAPI.getRepresentationsForViewType = (viewType) => {
		switch (viewType) {
			case ViewTypes.DEFAULT:
			case ViewTypes.GEOMETRY:
			case ViewTypes.SLICE:
			case ViewTypes.VOLUME:
			default: return [
				{
					builder: ConvexFaceContextRepresentation_default,
					labels: [
						"---",
						"--+",
						"-++",
						"-+-"
					],
					initialValues: {
						behavior: Behavior.HANDLE,
						pickable: true,
						activeScaleFactor: 1.2,
						activeColor: 1,
						useActiveColor: true,
						name: "Face 1"
					}
				},
				{
					builder: ConvexFaceContextRepresentation_default,
					labels: [
						"---",
						"+--",
						"+-+",
						"--+"
					],
					initialValues: {
						behavior: Behavior.HANDLE,
						pickable: true,
						activeScaleFactor: 1.2,
						activeColor: 1,
						useActiveColor: true,
						name: "Face 2"
					}
				},
				{
					builder: ConvexFaceContextRepresentation_default,
					labels: [
						"+--",
						"++-",
						"+++",
						"+-+"
					],
					initialValues: {
						behavior: Behavior.HANDLE,
						pickable: true,
						activeScaleFactor: 1.2,
						activeColor: 1,
						useActiveColor: true,
						name: "Face 3"
					}
				},
				{
					builder: ConvexFaceContextRepresentation_default,
					labels: [
						"++-",
						"-+-",
						"-++",
						"+++"
					],
					initialValues: {
						behavior: Behavior.HANDLE,
						pickable: true,
						activeScaleFactor: 1.2,
						activeColor: 1,
						useActiveColor: true,
						name: "Face 4"
					}
				},
				{
					builder: ConvexFaceContextRepresentation_default,
					labels: [
						"-++",
						"--+",
						"+-+",
						"+++"
					],
					initialValues: {
						behavior: Behavior.HANDLE,
						pickable: true,
						activeScaleFactor: 1.2,
						activeColor: 1,
						useActiveColor: true,
						name: "Face 5"
					}
				},
				{
					builder: ConvexFaceContextRepresentation_default,
					labels: [
						"-+-",
						"++-",
						"+--",
						"---"
					],
					initialValues: {
						behavior: Behavior.HANDLE,
						pickable: true,
						activeScaleFactor: 1.2,
						activeColor: 1,
						useActiveColor: true,
						name: "Face 6"
					}
				}
			];
		}
	};
}
var defaultValues = (initialValues) => ({
	behavior: widgetBehavior,
	widgetState: generateState(),
	...initialValues
});
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, defaultValues(initialValues));
	AbstractWidgetFactory_default.extend(publicAPI, model, initialValues);
	vtkInteractiveOrientationWidget(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkInteractiveOrientationWidget");
var InteractiveOrientationWidget_default = {
	newInstance,
	extend
};
//#endregion
export { InteractiveOrientationWidget_default as default, extend, newInstance };
