// @kitware/vtk.js/Widgets/Widgets3D/LabelWidget.js
import macros_default from "../../macros.js";
import { ViewTypes } from "../Core/WidgetManager/Constants.js";
import AbstractWidgetFactory_default from "../Core/AbstractWidgetFactory.js";
import PlaneManipulator_default from "../Manipulators/PlaneManipulator.js";
import SphereHandleRepresentation_default from "../Representations/SphereHandleRepresentation.js";
import widgetBehavior from "./LabelWidget/behavior.js";
import generateState from "./LabelWidget/state.js";
//#region Sources/Widgets/Widgets3D/LabelWidget/index.js
function vtkLabelWidget(publicAPI, model) {
	model.classHierarchy.push("vtkLabelWidget");
	const superClass = { ...publicAPI };
	model.methodsToLink = ["scaleInPixels"];
	publicAPI.getRepresentationsForViewType = (viewType) => {
		switch (viewType) {
			case ViewTypes.DEFAULT:
			case ViewTypes.GEOMETRY:
			case ViewTypes.SLICE:
			case ViewTypes.VOLUME:
			default: return [{
				builder: SphereHandleRepresentation_default,
				labels: ["moveHandle"]
			}];
		}
	};
	publicAPI.setManipulator = (manipulator) => {
		superClass.setManipulator(manipulator);
		model.widgetState.getMoveHandle().setManipulator(manipulator);
		model.widgetState.getText().setManipulator(manipulator);
	};
	publicAPI.setManipulator(model.manipulator || PlaneManipulator_default.newInstance({ useCameraNormal: true }));
}
function defaultValues(initialValues) {
	return {
		behavior: widgetBehavior,
		widgetState: generateState(),
		...initialValues
	};
}
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, defaultValues(initialValues));
	AbstractWidgetFactory_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, ["manipulator"]);
	vtkLabelWidget(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkLabelWidget");
var LabelWidget_default = {
	newInstance,
	extend
};
//#endregion
export { LabelWidget_default as default, extend, newInstance };
