// @kitware/vtk.js/Widgets/Widgets3D/LineWidget.js
import macros_default from "../../macros.js";
import { distance2BetweenPoints } from "../../Common/Core/Math.js";
import { ViewTypes } from "../Core/WidgetManager/Constants.js";
import AbstractWidgetFactory_default from "../Core/AbstractWidgetFactory.js";
import PlaneManipulator_default from "../Manipulators/PlaneManipulator.js";
import { Behavior } from "../Representations/WidgetRepresentation/Constants.js";
import PolyLineRepresentation_default from "../Representations/PolyLineRepresentation.js";
import generateState from "./LineWidget/state.js";
import ArrowHandleRepresentation_default from "../Representations/ArrowHandleRepresentation.js";
import { getPoint, updateTextPosition } from "./LineWidget/helpers.js";
import widgetBehavior from "./LineWidget/behavior.js";
//#region Sources/Widgets/Widgets3D/LineWidget/index.js
function vtkLineWidget(publicAPI, model) {
	model.classHierarchy.push("vtkLineWidget");
	const superClass = { ...publicAPI };
	model.methodsToLink = [
		"activeScaleFactor",
		"activeColor",
		"useActiveColor",
		"glyphResolution",
		"defaultScale",
		"scaleInPixels"
	];
	publicAPI.getRepresentationsForViewType = (viewType) => {
		switch (viewType) {
			case ViewTypes.DEFAULT:
			case ViewTypes.GEOMETRY:
			case ViewTypes.SLICE:
			case ViewTypes.VOLUME:
			default: return [
				{
					builder: ArrowHandleRepresentation_default,
					labels: ["handle1"],
					initialValues: {
						visibilityFlagArray: [false, false],
						coincidentTopologyParameters: {
							Point: {
								factor: -1,
								offset: -1
							},
							Line: {
								factor: -1,
								offset: -1
							},
							Polygon: {
								factor: -3,
								offset: -3
							}
						}
					}
				},
				{
					builder: ArrowHandleRepresentation_default,
					labels: ["handle2"],
					initialValues: {
						visibilityFlagArray: [false, false],
						coincidentTopologyParameters: {
							Point: {
								factor: -1,
								offset: -1
							},
							Line: {
								factor: -1,
								offset: -1
							},
							Polygon: {
								factor: -3,
								offset: -3
							}
						}
					}
				},
				{
					builder: ArrowHandleRepresentation_default,
					labels: ["moveHandle"],
					initialValues: {
						visibilityFlagArray: [false, false],
						coincidentTopologyParameters: {
							Point: {
								factor: -1,
								offset: -1
							},
							Line: {
								factor: -1,
								offset: -1
							},
							Polygon: {
								factor: -3,
								offset: -3
							}
						}
					}
				},
				{
					builder: PolyLineRepresentation_default,
					labels: [
						"handle1",
						"handle2",
						"moveHandle"
					],
					initialValues: {
						behavior: Behavior.HANDLE,
						pickable: true
					}
				}
			];
		}
	};
	publicAPI.getDistance = () => {
		const p1 = getPoint(0, model.widgetState);
		const p2 = getPoint(1, model.widgetState);
		return p1 && p2 ? Math.sqrt(distance2BetweenPoints(p1, p2)) : 0;
	};
	publicAPI.setManipulator = (manipulator) => {
		superClass.setManipulator(manipulator);
		model.widgetState.getMoveHandle().setManipulator(manipulator);
		model.widgetState.getHandle1().setManipulator(manipulator);
		model.widgetState.getHandle2().setManipulator(manipulator);
	};
	/**
	* TBD: Why setting the move handle ?
	*/
	model.widgetState.onBoundsChange((bounds) => {
		const center = [
			(bounds[0] + bounds[1]) * .5,
			(bounds[2] + bounds[3]) * .5,
			(bounds[4] + bounds[5]) * .5
		];
		model.widgetState.getMoveHandle().setOrigin(center);
	});
	let linePosSub = model.widgetState.getPositionOnLine().onModified(() => {
		updateTextPosition(model);
	});
	publicAPI.setManipulator(model.manipulator || PlaneManipulator_default.newInstance({ useCameraNormal: true }));
	publicAPI.delete = macros_default.chain(publicAPI.delete, () => {
		linePosSub.unsubscribe();
		linePosSub = null;
	});
}
var defaultValues = (initialValues) => ({
	behavior: widgetBehavior,
	widgetState: generateState(),
	...initialValues
});
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, defaultValues(initialValues));
	AbstractWidgetFactory_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, ["manipulator"]);
	vtkLineWidget(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkLineWidget");
var LineWidget_default = {
	newInstance,
	extend
};
//#endregion
export { LineWidget_default as default, extend, newInstance };
