// @kitware/vtk.js/Widgets/Widgets3D/PaintWidget.js
import macros_default from "../../macros.js";
import { ViewTypes } from "../Core/WidgetManager/Constants.js";
import AbstractWidgetFactory_default from "../Core/AbstractWidgetFactory.js";
import PlaneManipulator_default from "../Manipulators/PlaneManipulator.js";
import CircleContextRepresentation_default from "../Representations/CircleContextRepresentation.js";
import SphereHandleRepresentation_default from "../Representations/SphereHandleRepresentation.js";
import widgetBehavior from "./PaintWidget/behavior.js";
import generateState from "./PaintWidget/state.js";
//#region Sources/Widgets/Widgets3D/PaintWidget/index.js
function vtkPaintWidget(publicAPI, model) {
	model.classHierarchy.push("vtkPaintWidget");
	const superClass = { ...publicAPI };
	publicAPI.getRepresentationsForViewType = (viewType) => {
		switch (viewType) {
			case ViewTypes.DEFAULT:
			case ViewTypes.GEOMETRY:
			case ViewTypes.SLICE: return [{
				builder: CircleContextRepresentation_default,
				labels: ["handle", "trail"]
			}];
			case ViewTypes.VOLUME:
			default: return [{
				builder: SphereHandleRepresentation_default,
				labels: ["handle"]
			}];
		}
	};
	publicAPI.setManipulator = (manipulator) => {
		superClass.setManipulator(manipulator);
		model.widgetState.getHandle().setManipulator(manipulator);
	};
	const superSetRadius = publicAPI.setRadius;
	publicAPI.setRadius = (r) => {
		if (superSetRadius(r)) model.widgetState.getHandle().setScale1(r);
	};
	publicAPI.setManipulator(model.manipulator || PlaneManipulator_default.newInstance({ useCameraNormal: true }));
}
var defaultValues = (initialValues) => ({
	radius: 1,
	painting: false,
	color: [1],
	behavior: widgetBehavior,
	widgetState: generateState(initialValues?.radius ?? 1),
	...initialValues
});
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, defaultValues(initialValues));
	AbstractWidgetFactory_default.extend(publicAPI, model, initialValues);
	macros_default.get(publicAPI, model, ["painting"]);
	macros_default.setGet(publicAPI, model, [
		"manipulator",
		"radius",
		"color"
	]);
	vtkPaintWidget(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkPaintWidget");
var PaintWidget_default = {
	newInstance,
	extend
};
//#endregion
export { PaintWidget_default as default, extend, newInstance };
