// @kitware/vtk.js/Widgets/Widgets3D/PolyLineWidget.js
import macros_default from "../../macros.js";
import { ViewTypes } from "../Core/WidgetManager/Constants.js";
import AbstractWidgetFactory_default from "../Core/AbstractWidgetFactory.js";
import PlaneManipulator_default from "../Manipulators/PlaneManipulator.js";
import PolyLineRepresentation_default from "../Representations/PolyLineRepresentation.js";
import SphereHandleRepresentation_default from "../Representations/SphereHandleRepresentation.js";
import widgetBehavior from "./PolyLineWidget/behavior.js";
import generateState from "./PolyLineWidget/state.js";
//#region Sources/Widgets/Widgets3D/PolyLineWidget/index.js
function vtkPolyLineWidget(publicAPI, model) {
	model.classHierarchy.push("vtkPolyLineWidget");
	const superClass = { ...publicAPI };
	model.methodsToLink = [
		"activeColor",
		"activeScaleFactor",
		"closePolyLine",
		"defaultScale",
		"glyphResolution",
		"lineThickness",
		"useActiveColor",
		"scaleInPixels"
	];
	publicAPI.getRepresentationsForViewType = (viewType) => {
		switch (viewType) {
			case ViewTypes.DEFAULT:
			case ViewTypes.GEOMETRY:
			case ViewTypes.SLICE:
			case ViewTypes.VOLUME:
			default: return [
				{
					builder: SphereHandleRepresentation_default,
					labels: ["handles"]
				},
				{
					builder: SphereHandleRepresentation_default,
					labels: ["moveHandle"]
				},
				{
					builder: PolyLineRepresentation_default,
					labels: ["handles", "moveHandle"]
				}
			];
		}
	};
	publicAPI.setManipulator = (manipulator) => {
		superClass.setManipulator(manipulator);
		model.widgetState.getMoveHandle().setManipulator(manipulator);
		model.widgetState.getHandleList().forEach((handle) => {
			handle.setManipulator(manipulator);
		});
	};
	publicAPI.setManipulator(model.manipulator || PlaneManipulator_default.newInstance({
		useCameraFocalPoint: true,
		useCameraNormal: true
	}));
}
var defaultValues = (initialValues) => ({
	manipulator: null,
	behavior: widgetBehavior,
	widgetState: generateState(),
	...initialValues
});
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, defaultValues(initialValues));
	AbstractWidgetFactory_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, ["manipulator"]);
	vtkPolyLineWidget(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkPolyLineWidget");
var PolyLineWidget_default = {
	newInstance,
	extend
};
//#endregion
export { PolyLineWidget_default as default, extend, newInstance };
