// @kitware/vtk.js/Widgets/Widgets3D/ResliceCursorWidget.js
import macros_default from "../../macros.js";
import { add, distance2BetweenPoints, multiplyAccumulate, multiplyScalar, normalize, subtract } from "../../Common/Core/Math.js";
import Plane_default from "../../Common/DataModel/Plane.js";
import BoundingBox_default from "../../Common/DataModel/BoundingBox.js";
import MatrixBuilder_default from "../../Common/Core/MatrixBuilder.js";
import Box_default from "../../Common/DataModel/Box.js";
import PlaneSource_default from "../../Filters/Sources/PlaneSource.js";
import { ViewTypes } from "../Core/WidgetManager/Constants.js";
import AbstractWidgetFactory_default from "../Core/AbstractWidgetFactory.js";
import PlaneManipulator_default from "../Manipulators/PlaneManipulator.js";
import LineHandleRepresentation_default from "../Representations/LineHandleRepresentation.js";
import SphereHandleRepresentation_default from "../Representations/SphereHandleRepresentation.js";
import { viewTypeToPlaneName } from "./ResliceCursorWidget/Constants.js";
import { boundPlane, transformPlane, updateState } from "./ResliceCursorWidget/helpers.js";
import widgetBehavior from "./ResliceCursorWidget/behavior.js";
import generateState from "./ResliceCursorWidget/state.js";
import { mat4 } from "gl-matrix";
//#region Sources/Widgets/Widgets3D/ResliceCursorWidget/index.js
var VTK_INT_MAX = 2147483647;
var { vtkErrorMacro } = macros_default;
function vtkResliceCursorWidget(publicAPI, model) {
	model.classHierarchy.push("vtkResliceCursorWidget");
	model.methodsToLink = [
		"scaleInPixels",
		"holeWidth",
		"infiniteLine"
	];
	/**
	* Compute the origin of the reslice plane prior to transformations
	* It does not take into account the current view normal. (always axis aligned)
	* @param {*} viewType axial, coronal or sagittal
	*/
	function computeReslicePlaneOrigin(viewType) {
		const bounds = model.widgetState.getImage().getBounds();
		const center = publicAPI.getWidgetState().getCenter();
		const imageCenter = model.widgetState.getImage().getCenter();
		const offset = [];
		for (let i = 0; i < 3; i++) {
			offset[i] = -Math.abs(center[i] - imageCenter[i]);
			offset[i] *= 2;
		}
		const planeSource = PlaneSource_default.newInstance();
		if (viewType === ViewTypes.XZ_PLANE) {
			planeSource.setOrigin(bounds[0] + offset[0], center[1], bounds[4] + offset[2]);
			planeSource.setPoint1(bounds[1] - offset[0], center[1], bounds[4] + offset[2]);
			planeSource.setPoint2(bounds[0] + offset[0], center[1], bounds[5] - offset[2]);
		} else if (viewType === ViewTypes.XY_PLANE) {
			planeSource.setOrigin(bounds[0] + offset[0], bounds[2] + offset[1], center[2]);
			planeSource.setPoint1(bounds[1] - offset[0], bounds[2] + offset[1], center[2]);
			planeSource.setPoint2(bounds[0] + offset[0], bounds[3] - offset[1], center[2]);
		} else if (viewType === ViewTypes.YZ_PLANE) {
			planeSource.setOrigin(center[0], bounds[2] + offset[1], bounds[4] + offset[2]);
			planeSource.setPoint1(center[0], bounds[3] - offset[1], bounds[4] + offset[2]);
			planeSource.setPoint2(center[0], bounds[2] + offset[1], bounds[5] - offset[2]);
		}
		return planeSource;
	}
	/**
	* Compute the offset between display reslice cursor position and
	* display focal point position
	* This will be used to keep the same offset between reslice cursor
	* center and focal point when needed.
	*/
	function computeFocalPointOffsetFromResliceCursorCenter(viewType, renderer) {
		const worldFocalPoint = renderer.getActiveCamera().getFocalPoint();
		const worldResliceCenter = model.widgetState.getCenter();
		const dims = renderer.getRenderWindow().getViews()[0].getViewportSize(renderer);
		const aspect = dims[0] / dims[1];
		const newOffset = subtract(renderer.worldToNormalizedDisplay(...worldFocalPoint, aspect), renderer.worldToNormalizedDisplay(...worldResliceCenter, aspect), [
			0,
			0,
			0
		]);
		const cameraOffsets = model.widgetState.getCameraOffsets();
		cameraOffsets[viewType] = newOffset;
		model.widgetState.setCameraOffsets(cameraOffsets);
	}
	function updateCamera(renderer, normal, viewType, resetFocalPoint, keepCenterFocalDistance) {
		const focalPoint = renderer.getActiveCamera().getFocalPoint();
		const distance = renderer.getActiveCamera().getDistance();
		const estimatedCameraPosition = multiplyAccumulate(focalPoint, normal, distance, [
			0,
			0,
			0
		]);
		let newFocalPoint = focalPoint;
		if (resetFocalPoint) newFocalPoint = Plane_default.intersectWithLine(focalPoint, estimatedCameraPosition, model.widgetState.getCenter(), normal).x;
		if (keepCenterFocalDistance) {
			const worldResliceCenter = model.widgetState.getCenter();
			const dims = renderer.getRenderWindow().getViews()[0].getViewportSize(renderer);
			const aspect = dims[0] / dims[1];
			const displayResliceCenter = renderer.worldToNormalizedDisplay(...worldResliceCenter, aspect);
			const realOffset = model.widgetState.getCameraOffsets()[viewType];
			const displayFocal = add(displayResliceCenter, realOffset, [
				0,
				0,
				0
			]);
			const worldFocal = renderer.normalizedDisplayToWorld(...displayFocal, aspect);
			const intersection2 = Plane_default.intersectWithLine(worldFocal, estimatedCameraPosition, worldResliceCenter, normal);
			newFocalPoint[0] = intersection2.x[0];
			newFocalPoint[1] = intersection2.x[1];
			newFocalPoint[2] = intersection2.x[2];
		}
		renderer.getActiveCamera().setFocalPoint(newFocalPoint[0], newFocalPoint[1], newFocalPoint[2]);
		const newCameraPosition = multiplyAccumulate(newFocalPoint, normal, distance, [
			0,
			0,
			0
		]);
		renderer.getActiveCamera().setPosition(newCameraPosition[0], newCameraPosition[1], newCameraPosition[2]);
		const bounds = model.widgetState.getImage().getBounds();
		if (resetFocalPoint) renderer.resetCamera(bounds);
		renderer.resetCameraClippingRange(bounds);
	}
	/**
	* Convenient function to return the widget for a given viewType
	* @param {string} viewType
	* @returns the widget instanced in the given viewType.
	*/
	function findWidgetForViewType(viewType) {
		return publicAPI.getViewIds().map((viewId) => publicAPI.getWidgetForView({ viewId })).find((widget) => widget.getViewType() === viewType);
	}
	/**
	* Convenient function to return the ResliceCursorRepresentation for a given viewType
	* @param {string} viewType
	* @returns an array of 3 representations (for line handles, rotation handles, center handle)
	* or an empty array if the widget has not yet been added to the view type.
	*/
	function findRepresentationsForViewType(viewType) {
		const widgetForViewType = findWidgetForViewType(viewType);
		return widgetForViewType ? widgetForViewType.getRepresentations() : [];
	}
	publicAPI.getRepresentationsForViewType = (viewType) => {
		switch (viewType) {
			case ViewTypes.XY_PLANE:
			case ViewTypes.XZ_PLANE:
			case ViewTypes.YZ_PLANE: return [
				{
					builder: LineHandleRepresentation_default,
					labels: [`lineIn${viewTypeToPlaneName[viewType]}`],
					initialValues: {
						useActiveColor: false,
						scaleInPixels: model.scaleInPixels
					}
				},
				{
					builder: SphereHandleRepresentation_default,
					labels: [`rotationIn${viewTypeToPlaneName[viewType]}`],
					initialValues: {
						useActiveColor: false,
						scaleInPixels: model.scaleInPixels,
						lighting: false
					}
				},
				{
					builder: SphereHandleRepresentation_default,
					labels: ["center"],
					initialValues: {
						useActiveColor: false,
						scaleInPixels: model.scaleInPixels,
						lighting: false
					}
				}
			];
			case ViewTypes.DEFAULT:
			case ViewTypes.GEOMETRY:
			case ViewTypes.SLICE:
			case ViewTypes.VOLUME:
			default: return [];
		}
	};
	publicAPI.setImage = (image) => {
		model.widgetState.setImage(image);
		const center = image.getCenter();
		publicAPI.setCenter(center);
	};
	publicAPI.setCenter = (center) => {
		model.widgetState.setCenter(center);
		updateState(model.widgetState, model.scaleInPixels, model.rotationHandlePosition);
		publicAPI.modified();
	};
	publicAPI.updateCameraPoints = (renderer, viewType, resetFocalPoint, computeFocalPointOffset) => {
		publicAPI.resetCamera(renderer, viewType, resetFocalPoint, !computeFocalPointOffset);
		if (computeFocalPointOffset) computeFocalPointOffsetFromResliceCursorCenter(viewType, renderer);
	};
	/**
	*
	* @param {*} renderer
	* @param {*} viewType
	* @param {*} resetFocalPoint Defines if the focal point is reset to the image center
	* @param {*} keepCenterFocalDistance Defines if the estimated focal point has to be updated
	* in order to keep the same distance to the center (according to the computed focal point
	* shift)
	*/
	publicAPI.resetCamera = (renderer, viewType, resetFocalPoint, keepCenterFocalDistance) => {
		const center = model.widgetState.getImage().getCenter();
		const focalPoint = renderer.getActiveCamera().getFocalPoint();
		const position = renderer.getActiveCamera().getPosition();
		const distance = Math.sqrt(distance2BetweenPoints(position, focalPoint));
		const normal = publicAPI.getPlaneNormalFromViewType(viewType);
		const estimatedFocalPoint = resetFocalPoint ? center : focalPoint;
		const estimatedCameraPosition = multiplyAccumulate(estimatedFocalPoint, normal, distance, [
			0,
			0,
			0
		]);
		renderer.getActiveCamera().setFocalPoint(...estimatedFocalPoint);
		renderer.getActiveCamera().setPosition(...estimatedCameraPosition);
		renderer.getActiveCamera().setViewUp(model.widgetState.getPlanes()[viewType].viewUp);
		updateCamera(renderer, normal, viewType, resetFocalPoint, keepCenterFocalDistance);
	};
	publicAPI.getPlaneSource = (viewType) => {
		const planeSource = computeReslicePlaneOrigin(viewType);
		const { normal, viewUp } = model.widgetState.getPlanes()[viewType];
		transformPlane(planeSource, model.widgetState.getCenter(), normal, viewUp);
		const boundedOrigin = [...planeSource.getOrigin()];
		const boundedP1 = [...planeSource.getPoint1()];
		const boundedP2 = [...planeSource.getPoint2()];
		boundPlane(model.widgetState.getImage().getBounds(), boundedOrigin, boundedP1, boundedP2);
		planeSource.setOrigin(...boundedOrigin);
		planeSource.setPoint1(...boundedP1);
		planeSource.setPoint2(...boundedP2);
		return planeSource;
	};
	publicAPI.getResliceAxes = (viewType) => {
		const planeSource = publicAPI.getPlaneSource(viewType);
		const { normal } = model.widgetState.getPlanes()[viewType];
		const planeOrigin = planeSource.getOrigin();
		const p1 = planeSource.getPoint1();
		const planeAxis1 = [];
		subtract(p1, planeOrigin, planeAxis1);
		normalize(planeAxis1);
		const p2 = planeSource.getPoint2();
		const planeAxis2 = [];
		subtract(p2, planeOrigin, planeAxis2);
		normalize(planeAxis2);
		const newResliceAxes = mat4.identity(new Float64Array(16));
		for (let i = 0; i < 3; i++) {
			newResliceAxes[i] = planeAxis1[i];
			newResliceAxes[4 + i] = planeAxis2[i];
			newResliceAxes[8 + i] = normal[i];
			newResliceAxes[12 + i] = planeOrigin[i];
		}
		return newResliceAxes;
	};
	publicAPI.updateReslicePlane = (imageReslice, viewType) => {
		const spacing = model.widgetState.getImage().getSpacing();
		const planeSource = publicAPI.getPlaneSource(viewType);
		const newResliceAxes = publicAPI.getResliceAxes(viewType);
		const planeOrigin = planeSource.getOrigin();
		const planeAxis1 = subtract(planeSource.getPoint1(), planeOrigin, []);
		const planeSizeX = normalize(planeAxis1);
		const planeAxis2 = subtract(planeSource.getPoint2(), planeOrigin, []);
		const planeSizeY = normalize(planeAxis2);
		const spacingX = Math.abs(planeAxis1[0] * spacing[0]) + Math.abs(planeAxis1[1] * spacing[1]) + Math.abs(planeAxis1[2] * spacing[2]);
		const spacingY = Math.abs(planeAxis2[0] * spacing[0]) + Math.abs(planeAxis2[1] * spacing[1]) + Math.abs(planeAxis2[2] * spacing[2]);
		let extentX = 0;
		let extentY = 0;
		const realExtentX = spacingX === 0 ? Number.MAX_SAFE_INTEGER : planeSizeX / spacingX;
		const value = VTK_INT_MAX >> 1;
		if (realExtentX > value) {
			vtkErrorMacro("Invalid X extent: ", realExtentX, " on view type : ", viewType);
			extentX = 0;
		} else {
			extentX = 1;
			while (extentX < realExtentX) extentX <<= 1;
		}
		const realExtentY = spacingY === 0 ? Number.MAX_SAFE_INTEGER : planeSizeY / spacingY;
		if (realExtentY > value) {
			vtkErrorMacro("Invalid Y extent:", realExtentY, " on view type : ", viewType);
			extentY = 0;
		} else {
			extentY = 1;
			while (extentY < realExtentY) extentY <<= 1;
		}
		const outputSpacingX = extentX === 0 ? 1 : planeSizeX / extentX;
		const outputSpacingY = extentY === 0 ? 1 : planeSizeY / extentY;
		let modified = imageReslice.setResliceAxes(newResliceAxes);
		modified = imageReslice.setOutputSpacing([
			outputSpacingX,
			outputSpacingY,
			1
		]) || modified;
		modified = imageReslice.setOutputOrigin([
			.5 * outputSpacingX,
			.5 * outputSpacingY,
			0
		]) || modified;
		modified = imageReslice.setOutputExtent([
			0,
			extentX - 1,
			0,
			extentY - 1,
			0,
			0
		]) || modified;
		return modified;
	};
	/**
	* Returns a plane source with origin at cursor center and
	* normal from the view.
	* @param {ViewType} type: Axial, Coronal or Sagittal
	*/
	publicAPI.getPlaneSourceFromViewType = (type) => {
		const planeSource = PlaneSource_default.newInstance();
		const origin = publicAPI.getWidgetState().getCenter();
		const planeNormal = publicAPI.getPlaneNormalFromViewType(type);
		planeSource.setNormal(planeNormal);
		planeSource.setOrigin(origin);
		return planeSource;
	};
	publicAPI.getPlaneNormalFromViewType = (viewType) => publicAPI.getWidgetState().getPlanes()[viewType].normal;
	/**
	* Returns the normals of the planes that are not viewType.
	* @param {ViewType} viewType ViewType to extract other normals
	*/
	publicAPI.getOtherPlaneNormals = (viewType) => [
		ViewTypes.YZ_PLANE,
		ViewTypes.XZ_PLANE,
		ViewTypes.XY_PLANE
	].filter((vt) => vt !== viewType).map((vt) => publicAPI.getPlaneNormalFromViewType(vt));
	/**
	* Return the reslice cursor matrix built as such: [YZ, XZ, XY, center]
	*/
	publicAPI.getResliceMatrix = () => {
		const resliceMatrix = mat4.identity(new Float64Array(16));
		for (let i = 0; i < 3; i++) {
			resliceMatrix[4 * i + 0] = publicAPI.getPlaneNormalFromViewType(ViewTypes.YZ_PLANE)[i];
			resliceMatrix[4 * i + 1] = publicAPI.getPlaneNormalFromViewType(ViewTypes.XZ_PLANE)[i];
			resliceMatrix[4 * i + 2] = publicAPI.getPlaneNormalFromViewType(ViewTypes.XY_PLANE)[i];
		}
		const origin = publicAPI.getWidgetState().getCenter();
		return MatrixBuilder_default.buildFromRadian().translate(...origin).multiply(resliceMatrix).translate(...multiplyScalar([...origin], -1)).getMatrix();
	};
	publicAPI.getDisplayScaleParams = () => [
		ViewTypes.YZ_PLANE,
		ViewTypes.XZ_PLANE,
		ViewTypes.XY_PLANE
	].reduce((res, viewType) => {
		res[viewType] = findRepresentationsForViewType(viewType)[0]?.getDisplayScaleParams?.();
		return res;
	}, {});
	const superSetScaleInPixels = publicAPI.setScaleInPixels;
	publicAPI.setScaleInPixels = (scale) => {
		const res = superSetScaleInPixels(scale);
		publicAPI.getViewWidgets().forEach((w) => w.setScaleInPixels(scale));
		updateState(model.widgetState, model.scaleInPixels, model.rotationHandlePosition);
		return res;
	};
	publicAPI.getPlaneExtremities = (viewType) => {
		const dirProj = publicAPI.getWidgetState().getPlanes()[viewType].normal;
		const length = BoundingBox_default.getDiagonalLength(publicAPI.getWidgetState().getImage().getBounds());
		const p1 = multiplyAccumulate(publicAPI.getWidgetState().getCenter(), dirProj, -length, []);
		const p2 = multiplyAccumulate(publicAPI.getWidgetState().getCenter(), dirProj, length, []);
		const intersectionPoints = Box_default.intersectWithLine(publicAPI.getWidgetState().getImage().getBounds(), p1, p2);
		return [intersectionPoints.x1, intersectionPoints.x2];
	};
}
/**
* Initializes the model.
* @param {*} initialValues optional object of member variables. initialValues.planes is an optional list of axis names (e.g. ['X', 'Y'])
* @returns the initial model object
*/
var defaultValues = (initialValues) => ({
	behavior: widgetBehavior,
	widgetState: generateState(initialValues.planes),
	rotationHandlePosition: .5,
	scaleInPixels: true,
	manipulator: PlaneManipulator_default.newInstance(),
	...initialValues
});
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, defaultValues(initialValues));
	AbstractWidgetFactory_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, [
		"scaleInPixels",
		"rotationHandlePosition",
		"manipulator"
	]);
	vtkResliceCursorWidget(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkResliceCursorWidget");
var ResliceCursorWidget_default = {
	newInstance,
	extend
};
//#endregion
export { ResliceCursorWidget_default as default, extend, newInstance };
