// @kitware/vtk.js/Widgets/Widgets3D/SeedWidget.js
import macros_default from "../../macros.js";
import AbstractWidgetFactory_default from "../Core/AbstractWidgetFactory.js";
import CubeHandleRepresentation_default from "../Representations/CubeHandleRepresentation.js";
import widgetBehavior from "./SeedWidget/behavior.js";
import stateGenerator from "./SeedWidget/state.js";
//#region Sources/Widgets/Widgets3D/SeedWidget/index.js
function vtkSeedWidget(publicAPI, model) {
	model.classHierarchy.push("vtkSeedWidget");
	const superClass = { ...publicAPI };
	model.methodsToLink = ["scaleInPixels"];
	publicAPI.getRepresentationsForViewType = (viewType) => [{
		builder: CubeHandleRepresentation_default,
		labels: ["moveHandle"],
		initialValues: { useActiveColor: false }
	}];
	publicAPI.setManipulator = (manipulator) => {
		superClass.setManipulator(manipulator);
		model.widgetState.getMoveHandle().setManipulator(manipulator);
	};
}
var defaultValues = (initialValues) => ({
	behavior: widgetBehavior,
	widgetState: stateGenerator(),
	...initialValues
});
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, defaultValues(initialValues));
	AbstractWidgetFactory_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, ["manipulator", "widgetState"]);
	vtkSeedWidget(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkSeedWidget");
var SeedWidget_default = {
	newInstance,
	extend
};
//#endregion
export { SeedWidget_default as default, extend, newInstance };
