// @kitware/vtk.js/Widgets/Widgets3D/ShapeWidget.js
import macros_default from "../../macros.js";
import AbstractWidgetFactory_default from "../Core/AbstractWidgetFactory.js";
import { BehaviorCategory, ShapeBehavior } from "./ShapeWidget/Constants.js";
//#region Sources/Widgets/Widgets3D/ShapeWidget/index.js
function vtkShapeWidget(publicAPI, model) {
	model.classHierarchy.push("vtkShapeWidget");
	const superClass = { ...publicAPI };
	model.methodsToLink = ["scaleInPixels"];
	publicAPI.setManipulator = (manipulator) => {
		superClass.setManipulator(manipulator);
		model.widgetState.getStatesWithLabel("moveHandle").forEach((handle) => handle.setManipulator(manipulator));
	};
}
function defaultValues(initialValues) {
	return {
		modifierBehavior: { None: {
			[BehaviorCategory.PLACEMENT]: ShapeBehavior[BehaviorCategory.PLACEMENT].CLICK_AND_DRAG,
			[BehaviorCategory.POINTS]: ShapeBehavior[BehaviorCategory.POINTS].CORNER_TO_CORNER,
			[BehaviorCategory.RATIO]: ShapeBehavior[BehaviorCategory.RATIO].FREE
		} },
		resetAfterPointPlacement: false,
		...initialValues
	};
}
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, defaultValues(initialValues));
	AbstractWidgetFactory_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, [
		"manipulator",
		"modifierBehavior",
		"resetAfterPointPlacement"
	]);
	vtkShapeWidget(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkShapeWidget");
var ShapeWidget_default = {
	newInstance,
	extend
};
//#endregion
export { ShapeWidget_default as default, extend, newInstance };
