// @kitware/vtk.js/Widgets/Widgets3D/SphereWidget.js
import macros_default from "../../macros.js";
import { distance2BetweenPoints } from "../../Common/Core/Math.js";
import AbstractWidgetFactory_default from "../Core/AbstractWidgetFactory.js";
import PlaneManipulator_default from "../Manipulators/PlaneManipulator.js";
import SphereHandleRepresentation_default from "../Representations/SphereHandleRepresentation.js";
import SphereContextRepresentation_default from "../Representations/SphereContextRepresentation.js";
import widgetBehavior from "./SphereWidget/behavior.js";
import stateGenerator from "./SphereWidget/state.js";
//#region Sources/Widgets/Widgets3D/SphereWidget/index.js
function vtkSphereWidget(publicAPI, model) {
	model.classHierarchy.push("vtkSphereWidget");
	const superClass = { ...publicAPI };
	model.methodsToLink = ["scaleInPixels"];
	publicAPI.getRepresentationsForViewType = (viewType) => [
		{
			builder: SphereHandleRepresentation_default,
			labels: ["moveHandle"]
		},
		{
			builder: SphereHandleRepresentation_default,
			labels: ["centerHandle"]
		},
		{
			builder: SphereHandleRepresentation_default,
			labels: ["borderHandle"]
		},
		{
			builder: SphereContextRepresentation_default,
			labels: ["sphereHandle"]
		}
	];
	publicAPI.getRadius = () => {
		const h1 = model.widgetState.getCenterHandle();
		const h2 = model.widgetState.getBorderHandle();
		return Math.sqrt(distance2BetweenPoints(h1.getOrigin(), h2.getOrigin()));
	};
	publicAPI.setManipulator = (manipulator) => {
		superClass.setManipulator(manipulator);
		model.widgetState.getMoveHandle().setManipulator(manipulator);
		model.widgetState.getCenterHandle().setManipulator(manipulator);
		model.widgetState.getBorderHandle().setManipulator(manipulator);
	};
	publicAPI.setManipulator(model.manipulator || PlaneManipulator_default.newInstance({ useCameraNormal: true }));
}
var defaultValues = (initialValues) => ({
	behavior: widgetBehavior,
	widgetState: stateGenerator(),
	...initialValues
});
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, defaultValues(initialValues));
	AbstractWidgetFactory_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, ["manipulator", "widgetState"]);
	vtkSphereWidget(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkSphereWidget");
var SphereWidget_default = {
	newInstance,
	extend
};
//#endregion
export { SphereWidget_default as default, extend, newInstance };
