// @kitware/vtk.js/Widgets/Widgets3D/SplineWidget.js
import macros_default from "../../macros.js";
import { ViewTypes } from "../Core/WidgetManager/Constants.js";
import AbstractWidgetFactory_default from "../Core/AbstractWidgetFactory.js";
import PlaneManipulator_default from "../Manipulators/PlaneManipulator.js";
import SphereHandleRepresentation_default from "../Representations/SphereHandleRepresentation.js";
import SplineContextRepresentation_default from "../Representations/SplineContextRepresentation.js";
import widgetBehavior from "./SplineWidget/behavior.js";
import generateState from "./SplineWidget/state.js";
//#region Sources/Widgets/Widgets3D/SplineWidget/index.js
function vtkSplineWidget(publicAPI, model) {
	model.classHierarchy.push("vtkSplineWidget");
	const superClass = { ...publicAPI };
	model.methodsToLink = [
		"boundaryCondition",
		"close",
		"outputBorder",
		"fill",
		"borderColor",
		"errorBorderColor",
		"scaleInPixels"
	];
	publicAPI.getRepresentationsForViewType = (viewType) => {
		switch (viewType) {
			case ViewTypes.DEFAULT:
			case ViewTypes.GEOMETRY:
			case ViewTypes.SLICE:
			case ViewTypes.VOLUME:
			default: return [{
				builder: SphereHandleRepresentation_default,
				labels: ["handles", "moveHandle"]
			}, {
				builder: SplineContextRepresentation_default,
				labels: ["handles", "moveHandle"]
			}];
		}
	};
	publicAPI.setManipulator = (manipulator) => {
		superClass.setManipulator(manipulator);
		model.widgetState.getMoveHandle().setManipulator(manipulator);
		model.widgetState.getHandleList().forEach((handle) => {
			handle.setManipulator(manipulator);
		});
	};
	publicAPI.setManipulator(model.manipulator || model.manipulator || PlaneManipulator_default.newInstance({ useCameraNormal: true }));
}
var defaultValues = (initialValues) => ({
	freehandMinDistance: .1,
	allowFreehand: true,
	resolution: 32,
	defaultCursor: "pointer",
	handleSizeInPixels: 10,
	resetAfterPointPlacement: false,
	behavior: widgetBehavior,
	widgetState: generateState(),
	...initialValues
});
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, defaultValues(initialValues));
	AbstractWidgetFactory_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, [
		"manipulator",
		"freehandMinDistance",
		"allowFreehand",
		"resolution",
		"defaultCursor",
		"handleSizeInPixels",
		"resetAfterPointPlacement"
	]);
	vtkSplineWidget(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkSplineWidget");
var SplineWidget_default = {
	newInstance,
	extend
};
//#endregion
export { SplineWidget_default as default, extend, newInstance };
