// @kitware/vtk.js/Widgets/Widgets3D/TransformControlsWidget.js
import macros_default from "../../macros.js";
import AbstractWidgetFactory_default from "../Core/AbstractWidgetFactory.js";
import RotateTransformHandleRepresentation_default from "../Representations/RotateTransformHandleRepresentation.js";
import ScaleTransformHandleRepresentation_default from "../Representations/ScaleTransformHandleRepresentation.js";
import TranslateTransformHandleRepresentation_default from "../Representations/TranslateTransformHandleRepresentation.js";
import { TransformMode } from "./TransformControlsWidget/constants.js";
import widgetBehavior from "./TransformControlsWidget/behavior.js";
import stateGenerator from "./TransformControlsWidget/state.js";
import { mat3 } from "gl-matrix";
//#region Sources/Widgets/Widgets3D/TransformControlsWidget/index.js
function updateHandleTransforms(widgetState) {
	const transformState = widgetState.getTransform();
	const sx = widgetState.getScaleHandleX();
	const sy = widgetState.getScaleHandleY();
	const sz = widgetState.getScaleHandleZ();
	const hx = widgetState.getRotateHandleX();
	const hy = widgetState.getRotateHandleY();
	const hz = widgetState.getRotateHandleZ();
	widgetState.getStatesWithLabel("handles").forEach((state) => {
		state.setOrigin(transformState.getTranslation());
	});
	const m3 = mat3.create();
	mat3.fromQuat(m3, transformState.getRotation());
	[sx, hx].forEach((state) => {
		state.setDirection(m3.slice(0, 3));
		state.setUp(m3.slice(3, 6).map((c) => -c));
		state.setRight(m3.slice(6, 9));
	});
	[sy, hy].forEach((state) => {
		state.setDirection(m3.slice(3, 6));
		state.setUp(m3.slice(6, 9));
		state.setRight(m3.slice(0, 3));
	});
	[sz, hz].forEach((state) => {
		state.setDirection(m3.slice(6, 9));
		state.setUp(m3.slice(3, 6));
		state.setRight(m3.slice(0, 3));
	});
}
function vtkTransformControlsWidget(publicAPI, model) {
	model.classHierarchy.push("vtkTransformControlsWidget");
	model.behavior = widgetBehavior;
	model.widgetState = stateGenerator();
	model.methodsToLink = [
		"scaleInPixels",
		"activeScaleFactor",
		"useActiveColor",
		"activeColor"
	];
	publicAPI.getRepresentationsForViewType = (viewType) => {
		switch (viewType) {
			default: return [
				{
					builder: TranslateTransformHandleRepresentation_default,
					labels: ["translateHandles"],
					initialValues: {
						radius: 3,
						glyphResolution: 12,
						coneSource: {
							radius: 8,
							height: .05,
							direction: [
								0,
								1,
								0
							]
						}
					}
				},
				{
					builder: ScaleTransformHandleRepresentation_default,
					labels: ["scaleHandles"],
					initialValues: {
						radius: 3,
						glyphResolution: 12,
						cubeSource: {
							xLength: 20,
							yLength: 20 / 320,
							zLength: 20
						}
					}
				},
				{
					builder: RotateTransformHandleRepresentation_default,
					labels: ["rotateHandles"]
				}
			];
		}
	};
	publicAPI.updateHandleVisibility = () => {
		model.widgetState.getStatesWithLabel("translateHandles").forEach((state) => {
			state.setVisible(model.mode === "translate");
		});
		model.widgetState.getStatesWithLabel("scaleHandles").forEach((state) => {
			state.setVisible(model.mode === "scale");
		});
		model.widgetState.getStatesWithLabel("rotateHandles").forEach((state) => {
			state.setVisible(model.mode === "rotate");
		});
	};
	model._onModeChanged = () => {
		publicAPI.updateHandleVisibility();
	};
	const transformSubscription = model.widgetState.getTransform().onModified((state) => {
		updateHandleTransforms(model.widgetState);
	});
	publicAPI.delete = macros_default.chain(publicAPI.delete, () => {
		transformSubscription.unsubscribe();
	});
	updateHandleTransforms(model.widgetState);
	publicAPI.updateHandleVisibility();
}
var DEFAULT_VALUES = { mode: TransformMode.TRANSLATE };
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	AbstractWidgetFactory_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, ["mode"]);
	macros_default.get(publicAPI, model, ["lineManipulator", "rotateManipulator"]);
	vtkTransformControlsWidget(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkTransformControlsWidget");
var TransformControlsWidget_default = {
	newInstance,
	extend,
	TransformMode
};
//#endregion
export { TransformControlsWidget_default as default, extend, newInstance };
