// @kitware/vtk.js/Rendering/WebGPU/BindGroup.js
import macros_default from "../../macros.js";
//#region Sources/Rendering/WebGPU/BindGroup/index.js
var { vtkErrorMacro } = macros_default;
function vtkWebGPUBindGroup(publicAPI, model) {
	model.classHierarchy.push("vtkWebGPUBindGroup");
	publicAPI.setBindables = (bindables) => {
		if (model.bindables.length === bindables.length) {
			let allMatch = true;
			for (let i = 0; i < model.bindables.length; i++) if (model.bindables[i] !== bindables[i]) allMatch = false;
			if (allMatch) return;
		}
		model.bindables = bindables;
		publicAPI.modified();
	};
	publicAPI.getBindGroupLayout = (device) => {
		const entries = [];
		for (let i = 0; i < model.bindables.length; i++) {
			model.bindables[i].setDevice?.(device);
			const entry = model.bindables[i].getBindGroupLayoutEntry();
			entry.binding = i;
			entries.push(entry);
		}
		return device.getBindGroupLayout({ entries });
	};
	publicAPI.getBindGroup = (device) => {
		const deviceChanged = model.bindGroupDevice !== device;
		for (let i = 0; i < model.bindables.length; i++) model.bindables[i].setDevice?.(device);
		let mtime = publicAPI.getMTime();
		for (let i = 0; i < model.bindables.length; i++) {
			const tm = model.bindables[i].getBindGroupTime().getMTime();
			mtime = tm > mtime ? tm : mtime;
		}
		if (!deviceChanged && mtime < model.bindGroupTime.getMTime()) return model.bindGroup;
		const entries = [];
		for (let i = 0; i < model.bindables.length; i++) {
			const entry = model.bindables[i].getBindGroupEntry();
			entry.binding = i;
			entries.push(entry);
		}
		model.bindGroup = device.getHandle().createBindGroup({
			layout: publicAPI.getBindGroupLayout(device),
			entries,
			label: model.label
		});
		model.bindGroupDevice = device;
		model.bindGroupTime.modified();
		return model.bindGroup;
	};
	publicAPI.getShaderCode = (pipeline) => {
		const lines = [];
		const bgroup = pipeline.getBindGroupLayoutIndex(model.label);
		if (bgroup < 0) {
			vtkErrorMacro(`vtkWebGPUBindGroup: bind group layout ${model.label} was not found in pipeline`);
			return "";
		}
		for (let i = 0; i < model.bindables.length; i++) lines.push(model.bindables[i].getShaderCode(i, bgroup));
		return lines.join("\n");
	};
	publicAPI.releaseGraphicsResources = () => {
		model.bindGroup = null;
		model.bindGroupDevice = null;
		model.bindGroupTime.modified();
		publicAPI.modified();
	};
}
var DEFAULT_VALUES = {
	device: null,
	handle: null,
	bindGroupDevice: null,
	label: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	model.bindables = [];
	model.bindGroupTime = {};
	macros_default.obj(model.bindGroupTime, { mtime: 0 });
	macros_default.get(publicAPI, model, [
		"bindGroupTime",
		"handle",
		"sizeInBytes",
		"usage"
	]);
	macros_default.setGet(publicAPI, model, [
		"label",
		"device",
		"arrayInformation"
	]);
	vtkWebGPUBindGroup(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend);
var BindGroup_default = {
	newInstance,
	extend
};
//#endregion
export { BindGroup_default as default, extend, newInstance };
