// @kitware/vtk.js/Rendering/WebGPU/Buffer.js
import macros_default from "../../macros.js";
import Constants_default from "./BufferManager/Constants.js";
//#region Sources/Rendering/WebGPU/Buffer/index.js
var forwarded = [
	"getMappedRange",
	"mapAsync",
	"unmap"
];
function bufferSubData(device, destBuffer, destOffset, srcArrayBuffer) {
	const byteCount = srcArrayBuffer.byteLength;
	const srcBuffer = device.createBuffer({
		size: byteCount,
		usage: GPUBufferUsage.COPY_SRC,
		mappedAtCreation: true
	});
	const arrayBuffer = srcBuffer.getMappedRange(0, byteCount);
	new Uint8Array(arrayBuffer).set(new Uint8Array(srcArrayBuffer));
	srcBuffer.unmap();
	const encoder = device.createCommandEncoder();
	encoder.copyBufferToBuffer(srcBuffer, 0, destBuffer, destOffset, byteCount);
	const commandBuffer = encoder.finish();
	device.queue.submit([commandBuffer]);
	srcBuffer.destroy();
}
function vtkWebGPUBuffer(publicAPI, model) {
	model.classHierarchy.push("vtkWebGPUBuffer");
	publicAPI.create = (sizeInBytes, usage) => {
		model.handle = model.device.getHandle().createBuffer({
			size: sizeInBytes,
			usage,
			label: model.label
		});
		model.sizeInBytes = sizeInBytes;
		model.usage = usage;
	};
	publicAPI.write = (data) => {
		bufferSubData(model.device.getHandle(), model.handle, 0, data.buffer);
	};
	publicAPI.createAndWrite = (data, usage) => {
		const paddedSize = Math.ceil(data.byteLength / 4) * 4;
		model.handle = model.device.getHandle().createBuffer({
			size: paddedSize,
			usage,
			mappedAtCreation: true,
			label: model.label
		});
		model.sizeInBytes = paddedSize;
		model.usage = usage;
		new Uint8Array(model.handle.getMappedRange()).set(new Uint8Array(data.buffer));
		model.handle.unmap();
	};
	for (let i = 0; i < forwarded.length; i++) publicAPI[forwarded[i]] = (...args) => model.handle[forwarded[i]](...args);
}
var DEFAULT_VALUES = {
	device: null,
	handle: null,
	sizeInBytes: 0,
	strideInBytes: 0,
	arrayInformation: null,
	usage: null,
	label: null,
	sourceTime: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.get(publicAPI, model, [
		"handle",
		"sizeInBytes",
		"usage"
	]);
	macros_default.setGet(publicAPI, model, [
		"strideInBytes",
		"device",
		"arrayInformation",
		"label",
		"sourceTime"
	]);
	vtkWebGPUBuffer(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend);
var Buffer_default = {
	newInstance,
	extend,
	...Constants_default
};
//#endregion
export { Buffer_default as default, extend, newInstance };
