// @kitware/vtk.js/Rendering/WebGPU/Device.js
import { get, newInstance as newInstance$1, obj, setGet } from "../../macros.js";
import ShaderCache_default from "./ShaderCache.js";
import BufferManager_default from "./BufferManager.js";
import TextureManager_default from "./TextureManager.js";
//#region Sources/Rendering/WebGPU/Device/index.js
/**
* provide a simple WeakRefMap class to share device objects based on
* hash values so that buffers/textures etc can be shared betwen mappers.
* This is roughly based on WeakLRUCache but without the actual caching
* behavior. This is just a map of key -> WeakRef(value)
*/
var WeakRefMap = class extends Map {
	constructor() {
		super();
		this.registry = new FinalizationRegistry((key) => {
			const entry = super.get(key);
			if (entry && entry.deref && entry.deref() === void 0) super.delete(key);
		});
	}
	getValue(key) {
		const entry = super.get(key);
		if (entry) {
			const value = entry.deref();
			if (value !== void 0) return value;
			super.delete(key);
		}
	}
	setValue(key, value) {
		let entry;
		if (value && typeof value === "object") {
			entry = new WeakRef(value);
			this.registry.register(value, key);
			super.set(key, entry);
		}
		return entry;
	}
};
function vtkWebGPUDevice(publicAPI, model) {
	model.classHierarchy.push("vtkWebGPUDevice");
	publicAPI.initialize = (handle) => {
		model.handle = handle;
	};
	publicAPI.createCommandEncoder = () => model.handle.createCommandEncoder();
	publicAPI.submitCommandEncoder = (commandEncoder) => {
		model.handle.queue.submit([commandEncoder.finish()]);
	};
	publicAPI.getShaderModule = (sd) => model.shaderCache.getShaderModule(sd);
	publicAPI.getBindGroupLayout = (val) => {
		if (!val.entries) return null;
		for (let i = 0; i < val.entries.length; i++) {
			const ent = val.entries[i];
			ent.binding = ent.binding || 0;
			ent.visibility = ent.visibility || GPUShaderStage.VERTEX | GPUShaderStage.FRAGMENT;
		}
		const sval = JSON.stringify(val);
		for (let i = 0; i < model.bindGroupLayouts.length; i++) if (model.bindGroupLayouts[i].sval === sval) return model.bindGroupLayouts[i].layout;
		const layout = model.handle.createBindGroupLayout(val);
		model.bindGroupLayouts.push({
			sval,
			layout
		});
		return layout;
	};
	publicAPI.getBindGroupLayoutDescription = (layout) => {
		for (let i = 0; i < model.bindGroupLayouts.length; i++) if (model.bindGroupLayouts[i].layout === layout) return model.bindGroupLayouts[i].sval;
		vtkErrorMacro("layout not found");
		console.trace();
		return null;
	};
	publicAPI.getPipeline = (hash) => {
		if (hash in model.pipelines) return model.pipelines[hash];
		return null;
	};
	publicAPI.createPipeline = (hash, pipeline) => {
		pipeline.initialize(publicAPI, hash);
		model.pipelines[hash] = pipeline;
	};
	publicAPI.onSubmittedWorkDone = () => model.handle.queue.onSubmittedWorkDone();
	publicAPI.hasCachedObject = (hash) => model.objectCache.getValue(hash);
	publicAPI.getCachedObject = (hash, creator, ...args) => {
		if (!hash) {
			vtkErrorMacro("attempt to cache an object without a hash");
			return null;
		}
		const existingValue = model.objectCache.getValue(hash);
		if (existingValue) return existingValue;
		const createdObject = creator(...args);
		model.objectCache.setValue(hash, createdObject);
		return createdObject;
	};
}
var DEFAULT_VALUES = {
	handle: null,
	pipelines: null,
	shaderCache: null,
	bindGroupLayouts: null,
	bufferManager: null,
	textureManager: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	obj(publicAPI, model);
	setGet(publicAPI, model, ["handle"]);
	get(publicAPI, model, [
		"bufferManager",
		"shaderCache",
		"textureManager"
	]);
	model.objectCache = new WeakRefMap();
	model.shaderCache = ShaderCache_default.newInstance();
	model.shaderCache.setDevice(publicAPI);
	model.bindGroupLayouts = [];
	model.bufferManager = BufferManager_default.newInstance();
	model.bufferManager.setDevice(publicAPI);
	model.textureManager = TextureManager_default.newInstance();
	model.textureManager.setDevice(publicAPI);
	model.pipelines = {};
	vtkWebGPUDevice(publicAPI, model);
}
var newInstance = newInstance$1(extend, "vtkWebGPUDevice");
var Device_default = {
	newInstance,
	extend
};
//#endregion
export { WeakRefMap, Device_default as default, extend, newInstance };
