// @kitware/vtk.js/Rendering/WebGPU/ForwardPass.js
import macros_default from "../../macros.js";
import RenderPass_default from "../SceneGraph/RenderPass.js";
import FullScreenQuad_default from "./FullScreenQuad.js";
import RenderEncoder_default from "./RenderEncoder.js";
import Sampler_default from "./Sampler.js";
import TextureView_default from "./TextureView.js";
import OpaquePass_default from "./OpaquePass.js";
import OrderIndependentTranslucentPass_default from "./OrderIndependentTranslucentPass.js";
import VolumePass_default from "./VolumePass.js";
//#region Sources/Rendering/WebGPU/ForwardPass/index.js
var finalBlitFragTemplate = `
//VTK::Mapper::Dec

//VTK::TCoord::Dec

//VTK::RenderEncoder::Dec

//VTK::IOStructs::Dec

@fragment
fn main(
//VTK::IOStructs::Input
)
//VTK::IOStructs::Output
{
  var output: fragmentOutput;

  var texCoord: vec2<i32> =
    vec2<i32>(i32(input.fragPos.x), i32(input.fragPos.y));
  var computedColor: vec4<f32> = clamp(
    textureLoad(opaquePassColorTexture, texCoord, 0),
    vec4<f32>(0.0),
    vec4<f32>(1.0)
  );

  //VTK::RenderEncoder::Impl
  return output;
}
`;
function vtkForwardPass(publicAPI, model) {
	model.classHierarchy.push("vtkForwardPass");
	publicAPI.traverse = (viewNode, parent = null) => {
		if (model.deleted) return;
		model._currentParent = parent;
		publicAPI.setCurrentOperation("buildPass");
		viewNode.traverse(publicAPI);
		if (!model.opaquePass) model.opaquePass = OpaquePass_default.newInstance();
		const numlayers = viewNode.getRenderable().getNumberOfLayers();
		const renderers = viewNode.getChildren();
		for (let i = 0; i < numlayers; i++) for (let index = 0; index < renderers.length; index++) {
			const renNode = renderers[index];
			const ren = viewNode.getRenderable().getRenderers()[index];
			if (ren.getDraw() && ren.getLayer() === i) {
				model.opaqueActorCount = 0;
				model.translucentActorCount = 0;
				model.volumes = [];
				publicAPI.setCurrentOperation("queryPass");
				renNode.traverse(publicAPI);
				publicAPI.setCurrentOperation("cameraPass");
				renNode.traverse(publicAPI);
				model.opaquePass.traverse(renNode, viewNode);
				if (model.translucentActorCount > 0) {
					if (!model.translucentPass) model.translucentPass = OrderIndependentTranslucentPass_default.newInstance();
					model.translucentPass.setColorTextureView(model.opaquePass.getColorTextureView());
					model.translucentPass.setDepthTextureView(model.opaquePass.getDepthTextureView());
					model.translucentPass.traverse(renNode, viewNode);
				}
				if (model.volumes.length > 0) {
					if (!model.volumePass) model.volumePass = VolumePass_default.newInstance();
					model.volumePass.setColorTextureView(model.opaquePass.getColorTextureView());
					model.volumePass.setDepthTextureView(model.opaquePass.getDepthTextureView());
					model.volumePass.setVolumes(model.volumes);
					model.volumePass.traverse(renNode, viewNode);
				}
				publicAPI.finalPass(viewNode, renNode);
			}
		}
	};
	publicAPI.finalPass = (viewNode, renNode) => {
		if (!model._finalBlitEncoder) publicAPI.createFinalBlitEncoder(viewNode);
		model._finalBlitOutputTextureView.createFromTextureHandle(viewNode.getCurrentTexture(), {
			depth: 1,
			format: viewNode.getPresentationFormat()
		});
		model._finalBlitEncoder.attachTextureViews();
		model._finalBlitEncoder.begin(viewNode.getCommandEncoder());
		renNode.scissorAndViewport(model._finalBlitEncoder);
		model._fullScreenQuad.prepareAndDraw(model._finalBlitEncoder);
		model._finalBlitEncoder.end();
	};
	publicAPI.createFinalBlitEncoder = (viewNode) => {
		model._finalBlitEncoder = RenderEncoder_default.newInstance({ label: "forwardPassBlit" });
		model._finalBlitEncoder.setDescription({ colorAttachments: [{
			view: null,
			loadOp: "load",
			storeOp: "store"
		}] });
		model._finalBlitEncoder.setPipelineHash("fpf");
		model._finalBlitEncoder.setPipelineSettings({
			primitive: { cullMode: "none" },
			fragment: { targets: [{
				format: viewNode.getPresentationFormat(),
				blend: {
					color: {
						srcFactor: "src-alpha",
						dstFactor: "one-minus-src-alpha"
					},
					alpha: {
						srcFactor: "one",
						dstFactor: "one-minus-src-alpha"
					}
				}
			}] }
		});
		model._fsqSampler = Sampler_default.newInstance({ label: "finalPassSampler" });
		model._fsqSampler.create(viewNode.getDevice(), {
			minFilter: "linear",
			magFilter: "linear"
		});
		model._fullScreenQuad = FullScreenQuad_default.newInstance();
		model._fullScreenQuad.setDevice(viewNode.getDevice());
		model._fullScreenQuad.setPipelineHash("fpfsq");
		model._fullScreenQuad.setTextureViews([model.opaquePass.getColorTextureView()]);
		model._fullScreenQuad.setAdditionalBindables([model._fsqSampler]);
		model._fullScreenQuad.setFragmentShaderTemplate(finalBlitFragTemplate);
		model._fullScreenQuad.getShaderReplacements().set("replaceShaderTCoord", (hash, pipeline) => {
			const vDesc = pipeline.getShaderDescription("vertex");
			if (!vDesc.hasOutput("tcoordVS")) vDesc.addOutput("vec2<f32>", "tcoordVS");
			pipeline.getShaderDescription("fragment").addBuiltinInput("vec4<f32>", "@builtin(position) fragPos");
		});
		model._finalBlitOutputTextureView = TextureView_default.newInstance();
		model._finalBlitEncoder.setColorTextureView(0, model._finalBlitOutputTextureView);
	};
	publicAPI.incrementOpaqueActorCount = () => model.opaqueActorCount++;
	publicAPI.incrementTranslucentActorCount = () => model.translucentActorCount++;
	publicAPI.addVolume = (volume) => {
		model.volumes.push(volume);
	};
	publicAPI.releaseGraphicsResources = () => {
		model.opaquePass?.releaseGraphicsResources?.();
		model.translucentPass?.releaseGraphicsResources?.();
		model.volumePass?.releaseGraphicsResources?.();
		model.opaquePass = null;
		model.translucentPass = null;
		model.volumePass = null;
		model._finalBlitEncoder = null;
		model._finalBlitOutputTextureView = null;
		model._fullScreenQuad = null;
		model._fsqSampler = null;
	};
}
var DEFAULT_VALUES = {
	opaqueActorCount: 0,
	translucentActorCount: 0,
	volumes: null,
	opaqueRenderEncoder: null,
	translucentPass: null,
	volumePass: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	RenderPass_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, [
		"opaquePass",
		"translucentPass",
		"volumePass"
	]);
	vtkForwardPass(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkForwardPass");
var ForwardPass_default = {
	newInstance,
	extend
};
//#endregion
export { ForwardPass_default as default, extend, newInstance };
