// @kitware/vtk.js/Rendering/WebGPU/HardwareSelectionPass.js
import macros_default from "../../macros.js";
import RenderPass_default from "../SceneGraph/RenderPass.js";
import ShaderCache_default from "./ShaderCache.js";
import RenderEncoder_default from "./RenderEncoder.js";
import Texture_default from "./Texture.js";
//#region Sources/Rendering/WebGPU/HardwareSelectionPass/index.js
function vtkWebGPUHardwareSelectionPass(publicAPI, model) {
	model.classHierarchy.push("vtkWebGPUHardwareSelectionPass");
	publicAPI.traverse = (viewNode, renNode) => {
		if (model.deleted) return;
		model._currentParent = null;
		publicAPI.setCurrentOperation("buildPass");
		viewNode.traverse(publicAPI);
		const device = viewNode.getDevice();
		if (!model.selectionRenderEncoder) {
			publicAPI.createRenderEncoder();
			model.colorTexture = Texture_default.newInstance({ label: "hardwareSelectorColor" });
			model.colorTexture.create(device, {
				width: viewNode.getCanvas().width,
				height: viewNode.getCanvas().height,
				format: "rgba32uint",
				usage: GPUTextureUsage.RENDER_ATTACHMENT | GPUTextureUsage.COPY_SRC
			});
			const v1 = model.colorTexture.createView("hardwareSelectColorTexture");
			model.selectionRenderEncoder.setColorTextureView(0, v1);
			model.depthTexture = Texture_default.newInstance({ label: "hardwareSelectorDepth" });
			model.depthTexture.create(device, {
				width: viewNode.getCanvas().width,
				height: viewNode.getCanvas().height,
				format: "depth32float",
				usage: GPUTextureUsage.RENDER_ATTACHMENT | GPUTextureUsage.COPY_SRC
			});
			const v2 = model.depthTexture.createView("hardwareSelectDepthTexture");
			model.selectionRenderEncoder.setDepthTextureView(v2);
		} else {
			model.colorTexture.resize(viewNode.getCanvas().width, viewNode.getCanvas().height);
			model.depthTexture.resizeToMatch(model.colorTexture);
		}
		model.selectionRenderEncoder.attachTextureViews();
		renNode.setRenderEncoder(model.selectionRenderEncoder);
		publicAPI.setCurrentOperation("cameraPass");
		renNode.traverse(publicAPI);
		publicAPI.setCurrentOperation("opaquePass");
		renNode.traverse(publicAPI);
	};
	publicAPI.createRenderEncoder = () => {
		model.selectionRenderEncoder = RenderEncoder_default.newInstance({ label: "HardwareSelectionPass" });
		model.selectionRenderEncoder.setPipelineHash("sel");
		model.selectionRenderEncoder.setReplaceShaderCodeFunction((pipeline) => {
			const fDesc = pipeline.getShaderDescription("fragment");
			fDesc.addOutput("vec4<u32>", "outColor");
			let code = fDesc.getCode();
			code = ShaderCache_default.substitute(code, "//VTK::Select::Impl", ["  var compositeID: u32 = 0u;", "  var attributeID: u32 = compositeID;"]).result;
			code = ShaderCache_default.substitute(code, "//VTK::RenderEncoder::Impl", ["output.outColor = vec4<u32>(mapperUBO.PropID, compositeID, attributeID + 1u, 0u);"]).result;
			fDesc.setCode(code);
		});
		const renDesc = model.selectionRenderEncoder.getDescription();
		renDesc.colorAttachments[0].clearValue = [
			0,
			0,
			0,
			0
		];
		model.selectionRenderEncoder.setPipelineSettings({
			primitive: { cullMode: "none" },
			depthStencil: {
				depthWriteEnabled: true,
				depthCompare: "greater",
				format: "depth32float"
			},
			fragment: { targets: [{
				format: "rgba32uint",
				blend: void 0
			}] }
		});
	};
}
var DEFAULT_VALUES = {
	selectionRenderEncoder: null,
	colorTexture: null,
	depthTexture: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	RenderPass_default.extend(publicAPI, model, initialValues);
	macros_default.get(publicAPI, model, ["colorTexture", "depthTexture"]);
	vtkWebGPUHardwareSelectionPass(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkWebGPUHardwareSelectionPass");
var HardwareSelectionPass_default = {
	newInstance,
	extend
};
//#endregion
export { HardwareSelectionPass_default as default, extend, newInstance };
