// @kitware/vtk.js/Rendering/WebGPU/OpaquePass.js
import macros_default from "../../macros.js";
import RenderPass_default from "../SceneGraph/RenderPass.js";
import RenderEncoder_default from "./RenderEncoder.js";
import Texture_default from "./Texture.js";
//#region Sources/Rendering/WebGPU/OpaquePass/index.js
function vtkWebGPUOpaquePass(publicAPI, model) {
	model.classHierarchy.push("vtkWebGPUOpaquePass");
	publicAPI.traverse = (renNode, viewNode) => {
		if (model.deleted) return;
		model._currentParent = viewNode;
		const device = viewNode.getDevice();
		if (!model.renderEncoder) {
			publicAPI.createRenderEncoder();
			model.colorTexture = Texture_default.newInstance({ label: "opaquePassColor" });
			model.colorTexture.create(device, {
				width: viewNode.getCanvas().width,
				height: viewNode.getCanvas().height,
				format: "rgba16float",
				usage: GPUTextureUsage.RENDER_ATTACHMENT | GPUTextureUsage.TEXTURE_BINDING | GPUTextureUsage.COPY_SRC
			});
			const ctView = model.colorTexture.createView("opaquePassColorTexture");
			model.renderEncoder.setColorTextureView(0, ctView);
			model.depthFormat = "depth32float";
			model.depthTexture = Texture_default.newInstance({ label: "opaquePassDepth" });
			model.depthTexture.create(device, {
				width: viewNode.getCanvas().width,
				height: viewNode.getCanvas().height,
				format: model.depthFormat,
				usage: GPUTextureUsage.RENDER_ATTACHMENT | GPUTextureUsage.TEXTURE_BINDING | GPUTextureUsage.COPY_SRC
			});
			const dView = model.depthTexture.createView("opaquePassDepthTexture");
			model.renderEncoder.setDepthTextureView(dView);
		} else {
			model.colorTexture.resize(viewNode.getCanvas().width, viewNode.getCanvas().height);
			model.depthTexture.resize(viewNode.getCanvas().width, viewNode.getCanvas().height);
		}
		model.renderEncoder.attachTextureViews();
		publicAPI.setCurrentOperation("opaquePass");
		renNode.setRenderEncoder(model.renderEncoder);
		renNode.traverse(publicAPI);
	};
	publicAPI.getColorTextureView = () => model.renderEncoder.getColorTextureViews()[0];
	publicAPI.getDepthTextureView = () => model.renderEncoder.getDepthTextureView();
	publicAPI.createRenderEncoder = () => {
		model.renderEncoder = RenderEncoder_default.newInstance({ label: "OpaquePass" });
		model.renderEncoder.setPipelineHash("op");
	};
	publicAPI.releaseGraphicsResources = () => {
		model.renderEncoder = null;
		model.colorTexture = null;
		model.depthTexture = null;
	};
}
var DEFAULT_VALUES = {
	renderEncoder: null,
	colorTexture: null,
	depthTexture: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	RenderPass_default.extend(publicAPI, model, initialValues);
	macros_default.get(publicAPI, model, ["colorTexture", "depthTexture"]);
	vtkWebGPUOpaquePass(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkWebGPUOpaquePass");
var OpaquePass_default = {
	newInstance,
	extend
};
//#endregion
export { OpaquePass_default as default, extend, newInstance };
