// @kitware/vtk.js/Rendering/WebGPU/OrderIndependentTranslucentPass.js
import macros_default from "../../macros.js";
import RenderPass_default from "../SceneGraph/RenderPass.js";
import ShaderCache_default from "./ShaderCache.js";
import FullScreenQuad_default from "./FullScreenQuad.js";
import RenderEncoder_default from "./RenderEncoder.js";
import Texture_default from "./Texture.js";
//#region Sources/Rendering/WebGPU/OrderIndependentTranslucentPass/index.js
var oitpFragTemplate = `
//VTK::Mapper::Dec

//VTK::TCoord::Dec

//VTK::RenderEncoder::Dec

//VTK::IOStructs::Dec

@fragment
fn main(
//VTK::IOStructs::Input
)
//VTK::IOStructs::Output
{
  var output: fragmentOutput;

  var tcoord: vec2<i32> = vec2<i32>(i32(input.fragPos.x), i32(input.fragPos.y));
  var reveal: f32 = textureLoad(oitpAccumTexture, tcoord, 0).r;
  if (reveal == 1.0) { discard; }
  var tcolor: vec4<f32> = textureLoad(oitpColorTexture, tcoord, 0);
  var total: f32 = max(tcolor.a, 0.01);
  var computedColor: vec4<f32> = vec4<f32>(tcolor.r/total, tcolor.g/total, tcolor.b/total, 1.0 - reveal);

  //VTK::RenderEncoder::Impl
  return output;
}
`;
function vtkWebGPUOrderIndependentTranslucentPass(publicAPI, model) {
	model.classHierarchy.push("vtkWebGPUOrderIndependentTranslucentPass");
	publicAPI.traverse = (renNode, viewNode) => {
		if (model.deleted) return;
		model._currentParent = viewNode;
		const device = viewNode.getDevice();
		if (!model.translucentRenderEncoder) {
			publicAPI.createRenderEncoder();
			publicAPI.createFinalEncoder();
			model.translucentColorTexture = Texture_default.newInstance({ label: "translucentPassColor" });
			model.translucentColorTexture.create(device, {
				width: viewNode.getCanvas().width,
				height: viewNode.getCanvas().height,
				format: "rgba16float",
				usage: GPUTextureUsage.RENDER_ATTACHMENT | GPUTextureUsage.TEXTURE_BINDING
			});
			const v1 = model.translucentColorTexture.createView("oitpColorTexture");
			model.translucentRenderEncoder.setColorTextureView(0, v1);
			model.translucentAccumulateTexture = Texture_default.newInstance({ label: "translucentPassAccumulate" });
			model.translucentAccumulateTexture.create(device, {
				width: viewNode.getCanvas().width,
				height: viewNode.getCanvas().height,
				format: "r16float",
				usage: GPUTextureUsage.RENDER_ATTACHMENT | GPUTextureUsage.TEXTURE_BINDING
			});
			const v2 = model.translucentAccumulateTexture.createView("oitpAccumTexture");
			model.translucentRenderEncoder.setColorTextureView(1, v2);
			model.fullScreenQuad = FullScreenQuad_default.newInstance();
			model.fullScreenQuad.setDevice(viewNode.getDevice());
			model.fullScreenQuad.setPipelineHash("oitpfsq");
			model.fullScreenQuad.setTextureViews(model.translucentRenderEncoder.getColorTextureViews());
			model.fullScreenQuad.setFragmentShaderTemplate(oitpFragTemplate);
		} else {
			model.translucentColorTexture.resizeToMatch(model.colorTextureView.getTexture());
			model.translucentAccumulateTexture.resizeToMatch(model.colorTextureView.getTexture());
		}
		model.translucentRenderEncoder.setDepthTextureView(model.depthTextureView);
		model.translucentRenderEncoder.attachTextureViews();
		publicAPI.setCurrentOperation("translucentPass");
		renNode.setRenderEncoder(model.translucentRenderEncoder);
		renNode.traverse(publicAPI);
		publicAPI.finalPass(viewNode, renNode);
	};
	publicAPI.finalPass = (viewNode, renNode) => {
		model.translucentFinalEncoder.setColorTextureView(0, model.colorTextureView);
		model.translucentFinalEncoder.attachTextureViews();
		model.translucentFinalEncoder.begin(viewNode.getCommandEncoder());
		renNode.scissorAndViewport(model.translucentFinalEncoder);
		model.fullScreenQuad.prepareAndDraw(model.translucentFinalEncoder);
		model.translucentFinalEncoder.end();
	};
	publicAPI.getTextures = () => [model.translucentColorTexture, model.translucentAccumulateTexture];
	publicAPI.createRenderEncoder = () => {
		model.translucentRenderEncoder = RenderEncoder_default.newInstance({ label: "translucentRender" });
		const rDesc = model.translucentRenderEncoder.getDescription();
		rDesc.colorAttachments = [{
			view: void 0,
			clearValue: [
				0,
				0,
				0,
				0
			],
			loadOp: "clear",
			storeOp: "store"
		}, {
			view: void 0,
			clearValue: [
				1,
				0,
				0,
				0
			],
			loadOp: "clear",
			storeOp: "store"
		}];
		rDesc.depthStencilAttachment = {
			view: void 0,
			depthLoadOp: "load",
			depthStoreOp: "store"
		};
		model.translucentRenderEncoder.setReplaceShaderCodeFunction((pipeline) => {
			const fDesc = pipeline.getShaderDescription("fragment");
			fDesc.addOutput("vec4<f32>", "outColor");
			fDesc.addOutput("f32", "outAccum");
			fDesc.addBuiltinInput("vec4<f32>", "@builtin(position) fragPos");
			let code = fDesc.getCode();
			code = ShaderCache_default.substitute(code, "//VTK::RenderEncoder::Impl", [
				"var w: f32 = computedColor.a * pow(0.1 + input.fragPos.z, 2.0);",
				"output.outColor = vec4<f32>(computedColor.rgb*w, w);",
				"output.outAccum = computedColor.a;"
			]).result;
			fDesc.setCode(code);
		});
		model.translucentRenderEncoder.setPipelineHash("oitpr");
		model.translucentRenderEncoder.setPipelineSettings({
			primitive: { cullMode: "none" },
			depthStencil: {
				depthWriteEnabled: false,
				depthCompare: "greater",
				format: "depth32float"
			},
			fragment: { targets: [{
				format: "rgba16float",
				blend: {
					color: {
						srcFactor: "one",
						dstFactor: "one"
					},
					alpha: {
						srcFactor: "one",
						dstFactor: "one"
					}
				}
			}, {
				format: "r16float",
				blend: {
					color: {
						srcFactor: "zero",
						dstFactor: "one-minus-src"
					},
					alpha: {
						srcFactor: "one",
						dstFactor: "one-minus-src-alpha"
					}
				}
			}] }
		});
	};
	publicAPI.createFinalEncoder = () => {
		model.translucentFinalEncoder = RenderEncoder_default.newInstance({ label: "translucentFinal" });
		model.translucentFinalEncoder.setDescription({ colorAttachments: [{
			view: null,
			loadOp: "load",
			storeOp: "store"
		}] });
		model.translucentFinalEncoder.setReplaceShaderCodeFunction((pipeline) => {
			const fDesc = pipeline.getShaderDescription("fragment");
			fDesc.addOutput("vec4<f32>", "outColor");
			fDesc.addBuiltinInput("vec4<f32>", "@builtin(position) fragPos");
			let code = fDesc.getCode();
			code = ShaderCache_default.substitute(code, "//VTK::RenderEncoder::Impl", ["output.outColor = vec4<f32>(computedColor.rgb, computedColor.a);"]).result;
			fDesc.setCode(code);
		});
		model.translucentFinalEncoder.setPipelineHash("oitpf");
		model.translucentFinalEncoder.setPipelineSettings({
			primitive: { cullMode: "none" },
			fragment: { targets: [{
				format: "rgba16float",
				blend: {
					color: {
						srcFactor: "src-alpha",
						dstFactor: "one-minus-src-alpha"
					},
					alpha: {
						srcFactor: "one",
						dstFactor: "one-minus-src-alpha"
					}
				}
			}] }
		});
	};
	publicAPI.releaseGraphicsResources = () => {
		model.translucentRenderEncoder = null;
		model.translucentFinalEncoder = null;
		model.translucentColorTexture = null;
		model.translucentAccumulateTexture = null;
		model.fullScreenQuad = null;
	};
}
var DEFAULT_VALUES = {
	colorTextureView: null,
	depthTextureView: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	RenderPass_default.extend(publicAPI, model, initialValues);
	macros_default.setGet(publicAPI, model, ["colorTextureView", "depthTextureView"]);
	vtkWebGPUOrderIndependentTranslucentPass(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkWebGPUOrderIndependentTranslucentPass");
var OrderIndependentTranslucentPass_default = {
	newInstance,
	extend
};
//#endregion
export { OrderIndependentTranslucentPass_default as default, extend, newInstance };
