// @kitware/vtk.js/Rendering/WebGPU/Pipeline.js
import { get, newInstance as newInstance$1, obj, setGet } from "../../macros.js";
//#region Sources/Rendering/WebGPU/Pipeline/index.js
function vtkWebGPUPipeline(publicAPI, model) {
	model.classHierarchy.push("vtkWebGPUPipeline");
	publicAPI.getShaderDescriptions = () => model.shaderDescriptions;
	publicAPI.applyPipelineSettings = (baseSettings, extraSettings) => {
		if (!extraSettings) return baseSettings;
		return {
			...baseSettings,
			...extraSettings,
			primitive: {
				...baseSettings.primitive,
				...extraSettings.primitive
			},
			depthStencil: {
				...baseSettings.depthStencil,
				...extraSettings.depthStencil
			},
			fragment: {
				...baseSettings.fragment,
				...extraSettings.fragment
			}
		};
	};
	publicAPI.initialize = (device, hash) => {
		model.pipelineDescription = publicAPI.applyPipelineSettings(model.renderEncoder.getPipelineSettings(), model.extraPipelineSettings);
		model.pipelineDescription.primitive.topology = model.topology;
		model.pipelineDescription.vertex = model.vertexState;
		model.pipelineDescription.label = hash;
		const bindGroupLayouts = [];
		for (let i = 0; i < model.layouts.length; i++) bindGroupLayouts.push(model.layouts[i].layout);
		model.pipelineLayout = device.getHandle().createPipelineLayout({ bindGroupLayouts });
		model.pipelineDescription.layout = model.pipelineLayout;
		for (let i = 0; i < model.shaderDescriptions.length; i++) {
			const sd = model.shaderDescriptions[i];
			const sm = device.getShaderModule(sd);
			if (sd.getType() === "vertex") {
				model.pipelineDescription.vertex.module = sm.getHandle();
				model.pipelineDescription.vertex.entryPoint = "main";
			}
			if (sd.getType() === "fragment") {
				model.pipelineDescription.fragment.module = sm.getHandle();
				model.pipelineDescription.fragment.entryPoint = "main";
			}
		}
		model.handle = device.getHandle().createRenderPipeline(model.pipelineDescription);
	};
	publicAPI.getShaderDescription = (stype) => {
		for (let i = 0; i < model.shaderDescriptions.length; i++) if (model.shaderDescriptions[i].getType() === stype) return model.shaderDescriptions[i];
		return null;
	};
	publicAPI.addBindGroupLayout = (bindGroup) => {
		if (!bindGroup) return;
		model.layouts.push({
			layout: bindGroup.getBindGroupLayout(model.device),
			label: bindGroup.getLabel()
		});
	};
	publicAPI.getBindGroupLayout = (idx) => model.layouts[idx].layout;
	publicAPI.getBindGroupLayoutIndex = (llabel) => {
		for (let i = 0; i < model.layouts.length; i++) if (model.layouts[i].label === llabel) return i;
		return -1;
	};
	publicAPI.bindVertexInput = (renderEncoder, vInput) => {
		vInput.bindBuffers(renderEncoder);
	};
}
var DEFAULT_VALUES = {
	extraPipelineSettings: null,
	handle: null,
	layouts: null,
	renderEncoder: null,
	shaderDescriptions: null,
	vertexState: null,
	topology: null,
	pipelineDescription: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	obj(publicAPI, model);
	model.layouts = [];
	model.shaderDescriptions = [];
	get(publicAPI, model, ["handle", "pipelineDescription"]);
	setGet(publicAPI, model, [
		"device",
		"extraPipelineSettings",
		"renderEncoder",
		"topology",
		"vertexState"
	]);
	vtkWebGPUPipeline(publicAPI, model);
}
var newInstance = newInstance$1(extend, "vtkWebGPUPipeline");
var Pipeline_default = {
	newInstance,
	extend
};
//#endregion
export { Pipeline_default as default, extend, newInstance };
