// @kitware/vtk.js/Rendering/WebGPU/RenderWindow.js
import macros_default from "../../macros.js";
import HalfFloat_default from "../../Common/Core/HalfFloat.js";
import { registerViewConstructor } from "../Core/RenderWindow.js";
import RenderPass_default from "../SceneGraph/RenderPass.js";
import RenderWindowViewNode_default from "../SceneGraph/RenderWindowViewNode.js";
import Buffer_default from "./Buffer.js";
import ForwardPass_default from "./ForwardPass.js";
import Device_default from "./Device.js";
import HardwareSelector_default from "./HardwareSelector.js";
import ViewNodeFactory_default, { registerOverride } from "./ViewNodeFactory.js";
//#region Sources/Rendering/WebGPU/RenderWindow/index.js
var { vtkErrorMacro, vtkWarningMacro } = macros_default;
var SCREENSHOT_PLACEHOLDER = {
	position: "absolute",
	top: 0,
	left: 0,
	width: "100%",
	height: "100%"
};
function vtkWebGPURenderWindow(publicAPI, model) {
	model.classHierarchy.push("vtkWebGPURenderWindow");
	publicAPI.getViewNodeFactory = () => model.myFactory;
	function releaseViewNodeResources(viewNode, visited = /* @__PURE__ */ new Set()) {
		if (!viewNode || visited.has(viewNode)) return;
		visited.add(viewNode);
		const children = viewNode.getChildren?.() || [];
		for (let i = 0; i < children.length; i++) {
			const child = children[i];
			child?.releaseGraphicsResources?.();
			releaseViewNodeResources(child, visited);
		}
	}
	function queueRenderAfterInitialization() {
		const subscription = publicAPI.onInitialized(() => {
			subscription.unsubscribe();
			if (!model.deleted) publicAPI.traverseAllPasses();
		});
	}
	function handleDeviceLost(info, deviceGeneration) {
		if (model.deleted || deviceGeneration !== model.deviceGeneration || model.handlingDeviceLost) return;
		model.handlingDeviceLost = true;
		model.deviceLostInfo = info;
		const reason = info?.reason ?? "unknown";
		const message = info?.message || "WebGPU device was lost.";
		vtkWarningMacro(`WebGPU device lost (${reason}): ${message}`);
		publicAPI.releaseGraphicsResources();
		publicAPI.invokeDeviceLost({
			reason,
			message,
			recoverable: reason !== "destroyed"
		});
		if (reason !== "destroyed") {
			queueRenderAfterInitialization();
			publicAPI.initialize();
		}
		model.handlingDeviceLost = false;
	}
	function watchForDeviceLoss(deviceHandle, deviceGeneration) {
		deviceHandle.lost.then((info) => {
			handleDeviceLost(info, deviceGeneration);
		});
	}
	const previousSize = [0, 0];
	function updateWindow() {
		if (model.renderable) {
			if (model.size[0] !== previousSize[0] || model.size[1] !== previousSize[1]) {
				previousSize[0] = model.size[0];
				previousSize[1] = model.size[1];
				model.canvas.setAttribute("width", model.size[0]);
				model.canvas.setAttribute("height", model.size[1]);
				publicAPI.recreateSwapChain();
			}
		}
		if (model.viewStream) model.viewStream.setSize(model.size[0], model.size[1]);
		model.canvas.style.display = model.useOffScreen ? "none" : "block";
		if (model.el) model.el.style.cursor = model.cursorVisibility ? model.cursor : "none";
		model.containerSize = null;
	}
	publicAPI.onModified(updateWindow);
	publicAPI.recreateSwapChain = () => {
		if (model.context) {
			model.context.unconfigure();
			model.presentationFormat = navigator.gpu.getPreferredCanvasFormat(model.adapter);
			model.context.configure({
				device: model.device.getHandle(),
				format: model.presentationFormat,
				alphaMode: "premultiplied",
				usage: GPUTextureUsage.RENDER_ATTACHMENT | GPUTextureUsage.COPY_DST,
				width: model.size[0],
				height: model.size[1]
			});
			model._configured = true;
		}
	};
	publicAPI.getCurrentTexture = () => model.context.getCurrentTexture();
	publicAPI.buildPass = (prepass) => {
		if (prepass) {
			if (!model.renderable) return;
			publicAPI.prepareNodes();
			publicAPI.addMissingNodes(model.renderable.getRenderersByReference());
			publicAPI.removeUnusedNodes();
			publicAPI.initialize();
		} else if (model.initialized) {
			if (!model._configured) publicAPI.recreateSwapChain();
			model.commandEncoder = model.device.createCommandEncoder();
		}
	};
	publicAPI.initialize = () => {
		if (!model.initializing) {
			model.initializing = true;
			model.deviceLostInfo = null;
			if (!navigator.gpu) {
				vtkErrorMacro("WebGPU is not enabled.");
				return;
			}
			publicAPI.create3DContextAsync().then(() => {
				model.initialized = true;
				if (model.deleted) return;
				publicAPI.invokeInitialized();
			});
		}
	};
	publicAPI.setContainer = (el) => {
		if (model.el && model.el !== el) {
			if (model.canvas.parentNode !== model.el) vtkErrorMacro("Error: canvas parent node does not match container");
			model.el.removeChild(model.canvas);
			if (model.el.contains(model.bgImage)) model.el.removeChild(model.bgImage);
		}
		if (model.el !== el) {
			model.el = el;
			if (model.el) {
				model.el.appendChild(model.canvas);
				if (model.useBackgroundImage) model.el.appendChild(model.bgImage);
			}
			publicAPI.modified();
		}
	};
	publicAPI.getContainer = () => model.el;
	publicAPI.getContainerSize = () => {
		if (!model.containerSize && model.el) {
			const { width, height } = model.el.getBoundingClientRect();
			model.containerSize = [width, height];
		}
		return model.containerSize || model.size;
	};
	publicAPI.getFramebufferSize = () => model.size;
	publicAPI.create3DContextAsync = async () => {
		model.adapter = await navigator.gpu.requestAdapter({ powerPreference: "high-performance" });
		if (model.deleted) return;
		model.device = Device_default.newInstance();
		model.device.initialize(await model.adapter.requestDevice({ requiredLimits: {
			maxBufferSize: model.adapter.limits.maxBufferSize,
			maxStorageBufferBindingSize: model.adapter.limits.maxStorageBufferBindingSize,
			maxUniformBufferBindingSize: model.adapter.limits.maxUniformBufferBindingSize
		} }));
		if (model.deleted) {
			model.device = null;
			return;
		}
		model.deviceGeneration += 1;
		watchForDeviceLoss(model.device.getHandle(), model.deviceGeneration);
		model.context = model.canvas.getContext("webgpu");
	};
	publicAPI.releaseGraphicsResources = () => {
		if (model.renderPasses) for (let i = 0; i < model.renderPasses.length; i++) model.renderPasses[i]?.releaseGraphicsResources?.();
		releaseViewNodeResources(publicAPI);
		const rp = RenderPass_default.newInstance();
		rp.setCurrentOperation("Release");
		rp.traverse(publicAPI, null);
		if (model.context) model.context.unconfigure();
		model.adapter = null;
		model.device = null;
		model.context = null;
		model.commandEncoder = null;
		model._configured = false;
		model.initialized = false;
		model.initializing = false;
	};
	publicAPI.setBackgroundImage = (img) => {
		model.bgImage.src = img.src;
	};
	publicAPI.setUseBackgroundImage = (value) => {
		model.useBackgroundImage = value;
		if (model.useBackgroundImage && !model.el.contains(model.bgImage)) model.el.appendChild(model.bgImage);
		else if (!model.useBackgroundImage && model.el.contains(model.bgImage)) model.el.removeChild(model.bgImage);
	};
	async function getCanvasDataURL(format = model.imageFormat) {
		const temporaryCanvas = document.createElement("canvas");
		const temporaryContext = temporaryCanvas.getContext("2d");
		temporaryCanvas.width = model.canvas.width;
		temporaryCanvas.height = model.canvas.height;
		const result = await publicAPI.getPixelsAsync();
		const imageData = new ImageData(result.colorValues, result.width, result.height);
		temporaryContext.putImageData(imageData, 0, 0);
		const mainBoundingClientRect = model.canvas.getBoundingClientRect();
		model.renderable.getRenderers().forEach((renderer) => {
			renderer.getViewProps().forEach((viewProp) => {
				if (viewProp.getContainer) {
					const canvasList = viewProp.getContainer().getElementsByTagName("canvas");
					for (let i = 0; i < canvasList.length; i++) {
						const currentCanvas = canvasList[i];
						const boundingClientRect = currentCanvas.getBoundingClientRect();
						const newXPosition = boundingClientRect.x - mainBoundingClientRect.x;
						const newYPosition = boundingClientRect.y - mainBoundingClientRect.y;
						temporaryContext.drawImage(currentCanvas, newXPosition, newYPosition);
					}
				}
			});
		});
		const screenshot = temporaryCanvas.toDataURL(format);
		temporaryCanvas.remove();
		publicAPI.invokeImageReady(screenshot);
	}
	publicAPI.captureNextImage = (format = "image/png", { resetCamera = false, size = null, scale = 1 } = {}) => {
		if (model.deleted) return null;
		model.imageFormat = format;
		const previous = model.notifyStartCaptureImage;
		model.notifyStartCaptureImage = true;
		model._screenshot = { size: !!size || scale !== 1 ? size || model.size.map((val) => val * scale) : null };
		return new Promise((resolve, reject) => {
			const subscription = publicAPI.onImageReady((imageURL) => {
				if (model._screenshot.size === null) {
					model.notifyStartCaptureImage = previous;
					subscription.unsubscribe();
					if (model._screenshot.placeHolder) {
						model.size = model._screenshot.originalSize;
						publicAPI.modified();
						if (model._screenshot.cameras) model._screenshot.cameras.forEach(({ restoreParamsFn, arg }) => restoreParamsFn(arg));
						publicAPI.traverseAllPasses();
						model.el.removeChild(model._screenshot.placeHolder);
						model._screenshot.placeHolder.remove();
						model._screenshot = null;
					}
					resolve(imageURL);
				} else {
					const tmpImg = document.createElement("img");
					tmpImg.style = SCREENSHOT_PLACEHOLDER;
					tmpImg.src = imageURL;
					model._screenshot.placeHolder = model.el.appendChild(tmpImg);
					model.canvas.style.display = "none";
					model._screenshot.originalSize = model.size;
					model.size = model._screenshot.size;
					model._screenshot.size = null;
					publicAPI.modified();
					if (resetCamera) {
						const isUserResetCamera = resetCamera !== true;
						model._screenshot.cameras = model.renderable.getRenderers().map((renderer) => {
							const camera = renderer.getActiveCamera();
							const params = camera.get("focalPoint", "position", "parallelScale");
							return {
								resetCameraArgs: isUserResetCamera ? { renderer } : void 0,
								resetCameraFn: isUserResetCamera ? resetCamera : renderer.resetCamera,
								restoreParamsFn: camera.set,
								arg: JSON.parse(JSON.stringify(params))
							};
						});
						model._screenshot.cameras.forEach(({ resetCameraFn, resetCameraArgs }) => resetCameraFn(resetCameraArgs));
					}
					publicAPI.traverseAllPasses();
				}
			});
		});
	};
	publicAPI.traverseAllPasses = () => {
		if (model.deleted) return;
		if (!model.initialized) {
			publicAPI.initialize();
			const subscription = publicAPI.onInitialized(() => {
				subscription.unsubscribe();
				publicAPI.traverseAllPasses();
			});
		} else {
			if (model.renderPasses) for (let index = 0; index < model.renderPasses.length; ++index) model.renderPasses[index].traverse(publicAPI, null);
			if (model.commandEncoder) {
				model.device.submitCommandEncoder(model.commandEncoder);
				model.commandEncoder = null;
				if (model.notifyStartCaptureImage) model.device.onSubmittedWorkDone().then(() => {
					getCanvasDataURL();
				});
			}
		}
	};
	publicAPI.setViewStream = (stream) => {
		if (model.viewStream === stream) return false;
		if (model.subscription) {
			model.subscription.unsubscribe();
			model.subscription = null;
		}
		model.viewStream = stream;
		if (model.viewStream) {
			const mainRenderer = model.renderable.getRenderers()[0];
			mainRenderer.getBackgroundByReference()[3] = 0;
			publicAPI.setUseBackgroundImage(true);
			model.subscription = model.viewStream.onImageReady((e) => publicAPI.setBackgroundImage(e.image));
			model.viewStream.setSize(model.size[0], model.size[1]);
			model.viewStream.invalidateCache();
			model.viewStream.render();
			publicAPI.modified();
		}
		return true;
	};
	publicAPI.getUniquePropID = () => model.nextPropID++;
	publicAPI.getPropFromID = (id) => {
		for (let i = 0; i < model.children.length; i++) {
			const res = model.children[i].getPropFromID(id);
			if (res !== null) return res;
		}
		return null;
	};
	publicAPI.getPixelsAsync = async () => {
		const device = model.device;
		const texture = model.renderPasses[0].getOpaquePass().getColorTexture();
		const result = {
			width: texture.getWidth(),
			height: texture.getHeight()
		};
		result.colorBufferWidth = 32 * Math.floor((result.width + 31) / 32);
		result.colorBufferSizeInBytes = result.colorBufferWidth * result.height * 8;
		const colorBuffer = Buffer_default.newInstance();
		colorBuffer.setDevice(device);
		colorBuffer.create(result.colorBufferSizeInBytes, GPUBufferUsage.MAP_READ | GPUBufferUsage.COPY_DST);
		const cmdEnc = model.device.createCommandEncoder();
		cmdEnc.copyTextureToBuffer({ texture: texture.getHandle() }, {
			buffer: colorBuffer.getHandle(),
			bytesPerRow: 8 * result.colorBufferWidth,
			rowsPerImage: result.height
		}, {
			width: result.width,
			height: result.height,
			depthOrArrayLayers: 1
		});
		device.submitCommandEncoder(cmdEnc);
		await colorBuffer.mapAsync(GPUMapMode.READ);
		result.colorValues = new Uint16Array(colorBuffer.getMappedRange().slice());
		colorBuffer.unmap();
		const tmparray = new Uint8ClampedArray(result.height * result.width * 4);
		for (let y = 0; y < result.height; y++) for (let x = 0; x < result.width; x++) {
			const doffset = (y * result.width + x) * 4;
			const soffset = (y * result.colorBufferWidth + x) * 4;
			tmparray[doffset] = 255 * HalfFloat_default.fromHalf(result.colorValues[soffset]);
			tmparray[doffset + 1] = 255 * HalfFloat_default.fromHalf(result.colorValues[soffset + 1]);
			tmparray[doffset + 2] = 255 * HalfFloat_default.fromHalf(result.colorValues[soffset + 2]);
			tmparray[doffset + 3] = 255 * HalfFloat_default.fromHalf(result.colorValues[soffset + 3]);
		}
		result.colorValues = tmparray;
		return result;
	};
	publicAPI.createSelector = () => {
		const ret = HardwareSelector_default.newInstance();
		ret.setWebGPURenderWindow(publicAPI);
		return ret;
	};
	const superSetSize = publicAPI.setSize;
	publicAPI.setSize = (width, height) => {
		const modified = superSetSize(width, height);
		if (modified) publicAPI.invokeWindowResizeEvent({
			width,
			height
		});
		return modified;
	};
	publicAPI.delete = macros_default.chain(publicAPI.delete, publicAPI.setViewStream);
}
var DEFAULT_VALUES = {
	initialized: false,
	initializing: false,
	handlingDeviceLost: false,
	deviceGeneration: 0,
	deviceLostInfo: null,
	context: null,
	adapter: null,
	device: null,
	canvas: null,
	cursorVisibility: true,
	cursor: "pointer",
	containerSize: null,
	renderPasses: [],
	notifyStartCaptureImage: false,
	imageFormat: "image/png",
	useOffScreen: false,
	useBackgroundImage: false,
	nextPropID: 1,
	xrSupported: false,
	presentationFormat: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	model.canvas = document.createElement("canvas");
	model.canvas.style.width = "100%";
	model.bgImage = new Image();
	model.bgImage.style.position = "absolute";
	model.bgImage.style.left = "0";
	model.bgImage.style.top = "0";
	model.bgImage.style.width = "100%";
	model.bgImage.style.height = "100%";
	model.bgImage.style.zIndex = "-1";
	RenderWindowViewNode_default.extend(publicAPI, model, initialValues);
	model.myFactory = ViewNodeFactory_default.newInstance();
	model.renderPasses[0] = ForwardPass_default.newInstance();
	if (!model.selector) {
		model.selector = HardwareSelector_default.newInstance();
		model.selector.setWebGPURenderWindow(publicAPI);
	}
	macros_default.event(publicAPI, model, "imageReady");
	macros_default.event(publicAPI, model, "initialized");
	macros_default.event(publicAPI, model, "deviceLost");
	macros_default.get(publicAPI, model, [
		"commandEncoder",
		"deviceLostInfo",
		"device",
		"presentationFormat",
		"useBackgroundImage",
		"xrSupported"
	]);
	macros_default.setGet(publicAPI, model, [
		"initialized",
		"context",
		"canvas",
		"device",
		"renderPasses",
		"notifyStartCaptureImage",
		"cursor",
		"useOffScreen"
	]);
	macros_default.setGetArray(publicAPI, model, ["size"], 2);
	macros_default.event(publicAPI, model, "windowResizeEvent");
	vtkWebGPURenderWindow(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkWebGPURenderWindow");
registerViewConstructor("WebGPU", newInstance);
var RenderWindow_default = {
	newInstance,
	extend
};
registerOverride("vtkRenderWindow", newInstance);
//#endregion
export { RenderWindow_default as default, extend, newInstance };
