// @kitware/vtk.js/Rendering/WebGPU/Sampler.js
import macros_default from "../../macros.js";
//#region Sources/Rendering/WebGPU/Sampler/index.js
function vtkWebGPUSampler(publicAPI, model) {
	model.classHierarchy.push("vtkWebGPUSampler");
	publicAPI.create = (device, options = {}) => {
		model.device = device;
		model.options.addressModeU = options.addressModeU ? options.addressModeU : "clamp-to-edge";
		model.options.addressModeV = options.addressModeV ? options.addressModeV : "clamp-to-edge";
		model.options.addressModeW = options.addressModeW ? options.addressModeW : "clamp-to-edge";
		model.options.magFilter = options.magFilter ? options.magFilter : "nearest";
		model.options.minFilter = options.minFilter ? options.minFilter : "nearest";
		model.options.mipmapFilter = options.mipmapFilter ? options.mipmapFilter : "nearest";
		model.options.label = model.label;
		model.handle = model.device.getHandle().createSampler(model.options);
		model.bindGroupTime.modified();
	};
	publicAPI.getShaderCode = (binding, group) => {
		return `@binding(${binding}) @group(${group}) var ${model.label}: sampler;`;
	};
	publicAPI.getBindGroupEntry = () => {
		return { resource: model.handle };
	};
}
var DEFAULT_VALUES = {
	device: null,
	handle: null,
	label: null,
	options: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	model.options = {};
	model.bindGroupLayoutEntry = {
		visibility: GPUShaderStage.VERTEX | GPUShaderStage.FRAGMENT,
		sampler: {}
	};
	model.bindGroupTime = {};
	macros_default.obj(model.bindGroupTime, { mtime: 0 });
	macros_default.get(publicAPI, model, [
		"bindGroupTime",
		"handle",
		"options"
	]);
	macros_default.setGet(publicAPI, model, [
		"bindGroupLayoutEntry",
		"device",
		"label"
	]);
	vtkWebGPUSampler(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend);
var Sampler_default = {
	newInstance,
	extend
};
//#endregion
export { Sampler_default as default, extend, newInstance };
