// @kitware/vtk.js/Rendering/WebGPU/ShaderCache.js
import macros_default from "../../macros.js";
import ShaderModule_default from "./ShaderModule.js";
//#region Sources/Rendering/WebGPU/ShaderCache/index.js
function substitute(source, search, replace, all = true) {
	const replaceStr = Array.isArray(replace) ? replace.join("\n") : replace;
	let replaced = false;
	if (source.search(search) !== -1) replaced = true;
	let gflag = "";
	if (all) gflag = "g";
	const regex = new RegExp(search, gflag);
	const resultstr = source.replace(regex, replaceStr);
	return {
		replace: replaced,
		result: resultstr
	};
}
function vtkWebGPUShaderCache(publicAPI, model) {
	model.classHierarchy.push("vtkWebGPUShaderCache");
	publicAPI.getShaderModule = (shaderDesc) => {
		const sType = shaderDesc.getType();
		const sHash = shaderDesc.getHash();
		const keys = model._shaderModules.keys();
		for (let i = 0; i < keys.length; i++) {
			const key = keys[i];
			if (key.getHash() === sHash && key.getType() === sType) return model._shaderModules.get(key);
		}
		const sm = ShaderModule_default.newInstance();
		sm.initialize(model.device, shaderDesc);
		model._shaderModules.set(shaderDesc, sm);
		return sm;
	};
}
var DEFAULT_VALUES = {
	shaderModules: null,
	device: null,
	window: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	model._shaderModules = /* @__PURE__ */ new Map();
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, ["device", "window"]);
	vtkWebGPUShaderCache(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkWebGPUShaderCache");
var ShaderCache_default = {
	newInstance,
	extend,
	substitute
};
//#endregion
export { ShaderCache_default as default, extend, newInstance };
