// @kitware/vtk.js/Rendering/WebGPU/ShaderDescription.js
import macros_default from "../../macros.js";
import ShaderCache_default from "./ShaderCache.js";
//#region Sources/Rendering/WebGPU/ShaderDescription/index.js
function vtkWebGPUShaderDescription(publicAPI, model) {
	model.classHierarchy.push("vtkWebGPUShaderDescription");
	publicAPI.hasOutput = (name) => model.outputNames.includes(name);
	publicAPI.addOutput = (type, name, interpolation = void 0) => {
		model.outputTypes.push(type);
		model.outputNames.push(name);
		model.outputInterpolations.push(interpolation);
	};
	publicAPI.addBuiltinOutput = (type, name) => {
		model.builtinOutputTypes.push(type);
		model.builtinOutputNames.push(name);
	};
	publicAPI.addBuiltinInput = (type, name) => {
		model.builtinInputTypes.push(type);
		model.builtinInputNames.push(name);
	};
	publicAPI.replaceShaderCode = (priorStage, vertexInput) => {
		const inputImpl = [];
		let iodec = [];
		if (vertexInput) inputImpl.push(vertexInput.getShaderCode());
		if (priorStage || model.builtinInputNames.length) {
			const inputStruct = [];
			inputStruct.push(`struct ${model.type}Input\n{`);
			if (priorStage) {
				const inputNames = priorStage.getOutputNamesByReference();
				const inputTypes = priorStage.getOutputTypesByReference();
				const inputInterpolations = priorStage.getOutputInterpolationsByReference();
				for (let i = 0; i < inputNames.length; i++) if (inputInterpolations[i] !== void 0) inputStruct.push(`  @location(${i}) @interpolate(${inputInterpolations[i]}) ${inputNames[i]} : ${inputTypes[i]},`);
				else inputStruct.push(`  @location(${i}) ${inputNames[i]} : ${inputTypes[i]},`);
			}
			for (let i = 0; i < model.builtinInputNames.length; i++) inputStruct.push(`  ${model.builtinInputNames[i]} : ${model.builtinInputTypes[i]},`);
			if (inputStruct.length > 1) {
				inputStruct.push("};");
				iodec = inputStruct;
				inputImpl[inputImpl.length - 1] += ",";
				inputImpl.push(`input: ${model.type}Input`);
			}
		}
		if (inputImpl.length) model.code = ShaderCache_default.substitute(model.code, "//VTK::IOStructs::Input", inputImpl).result;
		if (model.outputNames.length + model.builtinOutputNames.length || model.code.includes("//VTK::IOStructs::Output")) {
			const outputStruct = [`struct ${model.type}Output\n{`];
			for (let i = 0; i < model.outputNames.length; i++) if (model.outputInterpolations[i] !== void 0) outputStruct.push(`  @location(${i}) @interpolate(${model.outputInterpolations[i]}) ${model.outputNames[i]} : ${model.outputTypes[i]},`);
			else outputStruct.push(`  @location(${i}) ${model.outputNames[i]} : ${model.outputTypes[i]},`);
			for (let i = 0; i < model.builtinOutputNames.length; i++) outputStruct.push(`  ${model.builtinOutputNames[i]} : ${model.builtinOutputTypes[i]},`);
			outputStruct.push("};");
			iodec = iodec.concat(outputStruct);
			model.code = ShaderCache_default.substitute(model.code, "//VTK::IOStructs::Output", [`-> ${model.type}Output`]).result;
		}
		model.code = ShaderCache_default.substitute(model.code, "//VTK::IOStructs::Dec", iodec).result;
	};
}
var DEFAULT_VALUES = {
	type: null,
	hash: null,
	code: null,
	outputNames: null,
	outputTypes: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	model.outputNames = [];
	model.outputTypes = [];
	model.outputInterpolations = [];
	model.builtinOutputNames = [];
	model.builtinOutputTypes = [];
	model.builtinInputNames = [];
	model.builtinInputTypes = [];
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, [
		"type",
		"hash",
		"code"
	]);
	macros_default.getArray(publicAPI, model, [
		"outputTypes",
		"outputNames",
		"outputInterpolations"
	]);
	vtkWebGPUShaderDescription(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkWebGPUShaderDescription");
var ShaderDescription_default = {
	newInstance,
	extend
};
//#endregion
export { ShaderDescription_default as default, extend, newInstance };
