// @kitware/vtk.js/Rendering/WebGPU/SimpleMapper.js
import macros_default from "../../macros.js";
import ViewNode_default from "../SceneGraph/ViewNode.js";
import ShaderCache_default from "./ShaderCache.js";
import BindGroup_default from "./BindGroup.js";
import Pipeline_default from "./Pipeline.js";
import ShaderDescription_default from "./ShaderDescription.js";
import VertexInput_default from "./VertexInput.js";
//#region Sources/Rendering/WebGPU/SimpleMapper/index.js
var vtkWebGPUSimpleMapperVS = `
//VTK::Renderer::Dec

//VTK::Color::Dec

//VTK::Normal::Dec

//VTK::TCoord::Dec

//VTK::Select::Dec

//VTK::Mapper::Dec

//VTK::IOStructs::Dec

@vertex
fn main(
//VTK::IOStructs::Input
)
//VTK::IOStructs::Output
{
  var output : vertexOutput;

  // var vertex: vec4<f32> = vertexBC;

  //VTK::Color::Impl

  //VTK::Normal::Impl

  //VTK::TCoord::Impl

  //VTK::Select::Impl

  //VTK::Position::Impl

  return output;
}
`;
var vtkWebGPUSimpleMapperFS = `
//VTK::Renderer::Dec

//VTK::Color::Dec

//VTK::Normal::Dec

//VTK::TCoord::Dec

//VTK::Select::Dec

//VTK::RenderEncoder::Dec

//VTK::Mapper::Dec

//VTK::IOStructs::Dec

@fragment
fn main(
//VTK::IOStructs::Input
)
//VTK::IOStructs::Output
{
  var output : fragmentOutput;

  //VTK::Color::Impl

  //VTK::Normal::Impl

  //VTK::Light::Impl

  //VTK::TCoord::Impl

  //VTK::Select::Impl

  // var computedColor:vec4<f32> = vec4<f32>(1.0,0.5,0.5,1.0);

  //VTK::RenderEncoder::Impl
  return output;
}
`;
function vtkWebGPUSimpleMapper(publicAPI, model) {
	model.classHierarchy.push("vtkWebGPUSimpleMapper");
	publicAPI.generateShaderDescriptions = (hash, pipeline, vertexInput) => {
		const vDesc = ShaderDescription_default.newInstance({
			type: "vertex",
			hash,
			code: model.vertexShaderTemplate
		});
		const fDesc = ShaderDescription_default.newInstance({
			type: "fragment",
			hash,
			code: model.fragmentShaderTemplate
		});
		const sdrs = pipeline.getShaderDescriptions();
		sdrs.push(vDesc);
		sdrs.push(fDesc);
		const scode = model.vertexShaderTemplate + model.fragmentShaderTemplate;
		const re = /* @__PURE__ */ new RegExp("//VTK::[^:]*::", "g");
		const fnames = scode.match(re).filter((v, i, a) => a.indexOf(v) === i).map((v) => `replaceShader${v.substring(7, v.length - 2)}`);
		for (let i = 0; i < fnames.length; i++) {
			const fname = fnames[i];
			if (fname !== "replaceShaderIOStructs" && model.shaderReplacements.has(fname)) model.shaderReplacements.get(fname)(hash, pipeline, vertexInput);
		}
		publicAPI.replaceShaderIOStructs(hash, pipeline, vertexInput);
	};
	publicAPI.replaceShaderIOStructs = (hash, pipeline, vertexInput) => {
		const vDesc = pipeline.getShaderDescription("vertex");
		vDesc.replaceShaderCode(null, vertexInput);
		pipeline.getShaderDescription("fragment").replaceShaderCode(vDesc);
	};
	publicAPI.replaceShaderRenderEncoder = (hash, pipeline, vertexInput) => {
		model.renderEncoder.replaceShaderCode(pipeline);
	};
	model.shaderReplacements.set("replaceShaderRenderEncoder", publicAPI.replaceShaderRenderEncoder);
	publicAPI.replaceShaderRenderer = (hash, pipeline, vertexInput) => {
		if (!model.WebGPURenderer) return;
		const ubocode = model.WebGPURenderer.getBindGroup().getShaderCode(pipeline);
		const vDesc = pipeline.getShaderDescription("vertex");
		let code = vDesc.getCode();
		code = ShaderCache_default.substitute(code, "//VTK::Renderer::Dec", [ubocode]).result;
		vDesc.setCode(code);
		const fDesc = pipeline.getShaderDescription("fragment");
		code = fDesc.getCode();
		code = ShaderCache_default.substitute(code, "//VTK::Renderer::Dec", [ubocode]).result;
		fDesc.setCode(code);
	};
	model.shaderReplacements.set("replaceShaderRenderer", publicAPI.replaceShaderRenderer);
	publicAPI.replaceShaderMapper = (hash, pipeline, vertexInput) => {
		const ubocode = model.bindGroup.getShaderCode(pipeline);
		const vDesc = pipeline.getShaderDescription("vertex");
		let code = vDesc.getCode();
		code = ShaderCache_default.substitute(code, "//VTK::Mapper::Dec", [ubocode]).result;
		vDesc.setCode(code);
		const fDesc = pipeline.getShaderDescription("fragment");
		fDesc.addBuiltinInput("bool", "@builtin(front_facing) frontFacing");
		code = fDesc.getCode();
		code = ShaderCache_default.substitute(code, "//VTK::Mapper::Dec", [ubocode]).result;
		fDesc.setCode(code);
	};
	model.shaderReplacements.set("replaceShaderMapper", publicAPI.replaceShaderMapper);
	publicAPI.replaceShaderPosition = (hash, pipeline, vertexInput) => {
		const vDesc = pipeline.getShaderDescription("vertex");
		vDesc.addBuiltinOutput("vec4<f32>", "@builtin(position) Position");
		let code = vDesc.getCode();
		code = ShaderCache_default.substitute(code, "//VTK::Position::Impl", ["    output.Position = rendererUBO.SCPCMatrix*vertexBC;"]).result;
		vDesc.setCode(code);
	};
	model.shaderReplacements.set("replaceShaderPosition", publicAPI.replaceShaderPosition);
	publicAPI.replaceShaderTCoord = (hash, pipeline, vertexInput) => {
		pipeline.getShaderDescription("vertex").addOutput("vec2<f32>", "tcoordVS");
	};
	model.shaderReplacements.set("replaceShaderTCoord", publicAPI.replaceShaderTCoord);
	publicAPI.addTextureView = (view) => {
		if (model.textureViews.includes(view)) return;
		model.textureViews.push(view);
	};
	publicAPI.prepareToDraw = (renderEncoder) => {
		model.renderEncoder = renderEncoder;
		publicAPI.updateInput();
		publicAPI.updateBuffers();
		publicAPI.updateBindings();
		publicAPI.updatePipeline();
	};
	publicAPI.updateInput = () => {};
	publicAPI.updateBuffers = () => {};
	publicAPI.updateBindings = () => {
		model.bindGroup.setBindables(publicAPI.getBindables());
	};
	publicAPI.computePipelineHash = () => {};
	publicAPI.getPipelineSettings = () => null;
	publicAPI.registerDrawCallback = (encoder) => {
		encoder.registerDrawCallback(model.pipeline, publicAPI.draw);
	};
	publicAPI.prepareAndDraw = (encoder) => {
		publicAPI.prepareToDraw(encoder);
		encoder.setPipeline(model.pipeline);
		publicAPI.draw(encoder);
	};
	publicAPI.draw = (renderEncoder) => {
		if (model.numberOfVertices === 0 || model.numberOfInstances === 0) return;
		const pipeline = renderEncoder.getBoundPipeline();
		renderEncoder.activateBindGroup(model.bindGroup);
		if (model.WebGPURenderer) model.WebGPURenderer.bindUBO(renderEncoder);
		pipeline.bindVertexInput(renderEncoder, model.vertexInput);
		const indexBuffer = model.vertexInput.getIndexBuffer();
		if (indexBuffer) renderEncoder.drawIndexed(indexBuffer.getIndexCount(), model.numberOfInstances, 0, 0, 0);
		else renderEncoder.draw(model.numberOfVertices, model.numberOfInstances, 0, 0);
	};
	publicAPI.getBindables = () => {
		const bindables = [...model.additionalBindables];
		if (model.UBO) bindables.push(model.UBO);
		if (model.SSBO) bindables.push(model.SSBO);
		for (let t = 0; t < model.textureViews.length; t++) {
			bindables.push(model.textureViews[t]);
			const samp = model.textureViews[t].getSampler();
			if (samp) bindables.push(samp);
		}
		return bindables;
	};
	publicAPI.updatePipeline = () => {
		publicAPI.computePipelineHash();
		model.pipeline = model.device.getPipeline(model.pipelineHash);
		if (!model.pipeline) {
			model.pipeline = Pipeline_default.newInstance();
			model.pipeline.setDevice(model.device);
			if (model.WebGPURenderer) model.pipeline.addBindGroupLayout(model.WebGPURenderer.getBindGroup());
			model.pipeline.addBindGroupLayout(model.bindGroup);
			publicAPI.generateShaderDescriptions(model.pipelineHash, model.pipeline, model.vertexInput);
			model.pipeline.setExtraPipelineSettings(publicAPI.getPipelineSettings());
			model.pipeline.setTopology(model.topology);
			model.pipeline.setRenderEncoder(model.renderEncoder);
			model.pipeline.setVertexState(model.vertexInput.getVertexInputInformation());
			model.device.createPipeline(model.pipelineHash, model.pipeline);
		}
	};
	publicAPI.releaseGraphicsResources = () => {
		model.vertexInput?.releaseGraphicsResources?.();
		model.bindGroup?.releaseGraphicsResources?.();
		model.UBO?.releaseGraphicsResources?.();
		model.SSBO?.releaseGraphicsResources?.();
		model.pipeline = null;
		model.renderEncoder = null;
		model.textureViews.length = 0;
	};
}
var DEFAULT_VALUES = {
	additionalBindables: void 0,
	bindGroup: null,
	device: null,
	fragmentShaderTemplate: null,
	numberOfInstances: 1,
	numberOfVertices: 0,
	pipelineHash: null,
	shaderReplacements: null,
	SSBO: null,
	textureViews: null,
	topology: "triangle-list",
	UBO: null,
	vertexShaderTemplate: null,
	WebGPURenderer: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	ViewNode_default.extend(publicAPI, model, initialValues);
	model.textureViews = [];
	model.vertexInput = VertexInput_default.newInstance();
	model.bindGroup = BindGroup_default.newInstance({ label: "mapperBG" });
	model.additionalBindables = [];
	model.fragmentShaderTemplate = model.fragmentShaderTemplate || vtkWebGPUSimpleMapperFS;
	model.vertexShaderTemplate = model.vertexShaderTemplate || vtkWebGPUSimpleMapperVS;
	model.shaderReplacements = /* @__PURE__ */ new Map();
	macros_default.get(publicAPI, model, ["pipeline", "vertexInput"]);
	macros_default.setGet(publicAPI, model, [
		"additionalBindables",
		"device",
		"fragmentShaderTemplate",
		"interpolate",
		"numberOfInstances",
		"numberOfVertices",
		"pipelineHash",
		"shaderReplacements",
		"SSBO",
		"textureViews",
		"topology",
		"UBO",
		"vertexShaderTemplate",
		"WebGPURenderer"
	]);
	vtkWebGPUSimpleMapper(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkWebGPUSimpleMapper");
var SimpleMapper_default = {
	newInstance,
	extend
};
//#endregion
export { SimpleMapper_default as default, extend, newInstance };
