// @kitware/vtk.js/Rendering/WebGPU/Texture.js
import macros_default from "../../macros.js";
import HalfFloat_default from "../../Common/Core/HalfFloat.js";
import Texture_default$1 from "../Core/Texture.js";
import Types_default from "./Types.js";
import TextureView_default from "./TextureView.js";
//#region Sources/Rendering/WebGPU/Texture/index.js
var { vtkErrorMacro } = macros_default;
function vtkWebGPUTexture(publicAPI, model) {
	model.classHierarchy.push("vtkWebGPUTexture");
	const getUploadArrayType = (tDetails, fallbackType) => {
		if (tDetails.elementSize === 2 && tDetails.sampleType === "float") return "Uint16Array";
		if (tDetails.sampleType === "sint") {
			if (tDetails.elementSize === 1) return "Int8Array";
			if (tDetails.elementSize === 2) return "Int16Array";
			if (tDetails.elementSize === 4) return "Int32Array";
		} else if (tDetails.sampleType === "unfilterable-float") {
			if (tDetails.elementSize === 4) return "Float32Array";
		} else {
			if (tDetails.elementSize === 1) return "Uint8Array";
			if (tDetails.elementSize === 2) return "Uint16Array";
			if (tDetails.elementSize === 4) return "Uint32Array";
		}
		return fallbackType;
	};
	const prepareTextureUploadData = (arr, width, height, depth) => {
		const tDetails = Types_default.getDetailsFromTextureFormat(model.format);
		const expectedRowElements = width * tDetails.numComponents;
		const expectedElementCount = expectedRowElements * height * depth;
		const halfFloat = tDetails.elementSize === 2 && tDetails.sampleType === "float";
		if (!arr?.length && expectedElementCount > 0) {
			vtkErrorMacro("Texture upload failed: missing nativeArray data.");
			return null;
		}
		if (arr.length < expectedElementCount) {
			vtkErrorMacro(`Texture upload failed: expected ${expectedElementCount} values but received ${arr.length}.`);
			return null;
		}
		const inputArray = arr.length > expectedElementCount ? arr.subarray(0, expectedElementCount) : arr;
		const sourceBytesPerElement = inputArray.BYTES_PER_ELEMENT || tDetails.elementSize;
		const expectedBytesPerRow = width * tDetails.stride;
		const alignedBytesPerRow = 256 * Math.floor((expectedBytesPerRow + 255) / 256);
		const outputArrayType = getUploadArrayType(tDetails, inputArray.constructor.name);
		const alignedRowElements = alignedBytesPerRow / (halfFloat ? 2 : macros_default.newTypedArray(outputArrayType, 0).BYTES_PER_ELEMENT);
		const inputRowBytes = expectedRowElements * sourceBytesPerElement;
		if (!(halfFloat || inputArray.constructor.name !== outputArrayType || inputRowBytes !== alignedBytesPerRow)) return {
			data: inputArray,
			bytesPerRow: alignedBytesPerRow
		};
		const totalRows = height * depth;
		const outArray = macros_default.newTypedArray(outputArrayType, alignedRowElements * totalRows);
		if (halfFloat) for (let row = 0; row < totalRows; row++) {
			const inOffset = row * expectedRowElements;
			const outOffset = row * alignedRowElements;
			for (let i = 0; i < expectedRowElements; i++) outArray[outOffset + i] = HalfFloat_default.toHalf(inputArray[inOffset + i]);
		}
		else if (alignedRowElements === expectedRowElements) outArray.set(inputArray);
		else for (let row = 0; row < totalRows; row++) outArray.set(inputArray.subarray(row * expectedRowElements, (row + 1) * expectedRowElements), row * alignedRowElements);
		return {
			data: outArray,
			bytesPerRow: alignedBytesPerRow
		};
	};
	const validateTextureWriteBounds = (x, y, z, width, height, depth) => {
		if (x < 0 || y < 0 || z < 0 || width <= 0 || height <= 0 || depth <= 0) {
			vtkErrorMacro(`Texture upload failed: invalid write region origin=(${x}, ${y}, ${z}) size=(${width}, ${height}, ${depth}).`);
			return false;
		}
		if (x + width > model.width || y + height > model.height || z + depth > model.depth) {
			vtkErrorMacro(`Texture upload failed: write region origin=(${x}, ${y}, ${z}) size=(${width}, ${height}, ${depth}) exceeds texture extent=(${model.width}, ${model.height}, ${model.depth}).`);
			return false;
		}
		return true;
	};
	publicAPI.create = (device, options) => {
		model.device = device;
		model.width = options.width;
		model.height = options.height;
		model.depth = options.depth ? options.depth : 1;
		if (options.dimension) model.dimension = options.dimension;
		else model.dimension = model.depth === 1 ? "2d" : "3d";
		model.format = options.format ? options.format : "rgba8unorm";
		model.mipLevel = options.mipLevel ? options.mipLevel : 0;
		model.usage = options.usage ? options.usage : GPUTextureUsage.TEXTURE_BINDING | GPUTextureUsage.COPY_DST;
		model.handle = model.device.getHandle().createTexture({
			size: [
				model.width,
				model.height,
				model.depth
			],
			format: model.format,
			usage: model.usage,
			label: model.label,
			dimension: model.dimension,
			mipLevelCount: model.mipLevel + 1
		});
	};
	publicAPI.assignFromHandle = (device, handle, options) => {
		model.device = device;
		model.handle = handle;
		model.width = options.width;
		model.height = options.height;
		model.depth = options.depth ? options.depth : 1;
		if (options.dimension) model.dimension = options.dimension;
		else model.dimension = model.depth === 1 ? "2d" : "3d";
		model.format = options.format ? options.format : "rgba8unorm";
		model.usage = options.usage ? options.usage : GPUTextureUsage.TEXTURE_BINDING | GPUTextureUsage.COPY_DST;
	};
	publicAPI.writeImageData = (req) => {
		let nativeArray = [];
		const _copyImageToTexture = (source) => {
			const originZ = req.originZ ?? 0;
			const depth = req.depth ?? 1;
			model.device.getHandle().queue.copyExternalImageToTexture({
				source,
				flipY: req.flip
			}, {
				texture: model.handle,
				premultipliedAlpha: true,
				mipLevel: 0,
				origin: {
					x: 0,
					y: 0,
					z: originZ
				}
			}, [
				source.width,
				source.height,
				depth
			]);
			if (model.dimension === "2d" && depth === 1 && model.mipLevel > 0) Texture_default$1.generateMipmaps(model.device.getHandle(), model.handle, model.mipLevel + 1);
			model.ready = true;
		};
		if (req.canvas) {
			_copyImageToTexture(req.canvas);
			return;
		}
		if (req.imageBitmap) {
			req.width = req.imageBitmap.width;
			req.height = req.imageBitmap.height;
			req.depth = 1;
			req.format = "rgba8unorm";
			req.flip = true;
			_copyImageToTexture(req.imageBitmap);
			return;
		}
		if (req.jsImageData) {
			req.width = req.jsImageData.width;
			req.height = req.jsImageData.height;
			req.depth = 1;
			req.format = "rgba8unorm";
			req.flip = true;
			_copyImageToTexture(req.jsImageData);
			return;
		}
		if (req.image) {
			req.width = req.image.width;
			req.height = req.image.height;
			req.depth = 1;
			req.format = "rgba8unorm";
			req.flip = true;
			_copyImageToTexture(req.image);
			return;
		}
		if (req.nativeArray) nativeArray = req.nativeArray;
		const width = req.width ?? model.width;
		const height = req.height ?? model.height;
		const depth = req.depth ?? 1;
		const preparedData = prepareTextureUploadData(nativeArray, width, height, depth);
		if (!preparedData) return;
		const data = preparedData.data;
		model.device.getHandle().queue.writeTexture({
			texture: model.handle,
			mipLevel: 0,
			origin: {
				x: 0,
				y: 0,
				z: req.originZ ?? 0
			}
		}, data, {
			offset: 0,
			bytesPerRow: preparedData.bytesPerRow,
			rowsPerImage: height
		}, {
			width,
			height,
			depthOrArrayLayers: depth
		});
		if (model.dimension === "2d" && depth === 1 && model.mipLevel > 0) Texture_default$1.generateMipmaps(model.device.getHandle(), model.handle, model.mipLevel + 1);
		model.ready = true;
	};
	publicAPI.writeSubImageData = (req) => {
		const x = req.x ?? 0;
		const y = req.y ?? 0;
		const z = req.z ?? 0;
		const width = req.width ?? model.width - x;
		const height = req.height ?? model.height - y;
		const depth = req.depth ?? model.depth - z;
		const nativeArray = req.nativeArray || [];
		if (!validateTextureWriteBounds(x, y, z, width, height, depth)) return;
		const preparedData = prepareTextureUploadData(nativeArray, width, height, depth);
		if (!preparedData) return;
		model.device.getHandle().queue.writeTexture({
			texture: model.handle,
			mipLevel: 0,
			origin: {
				x,
				y,
				z
			}
		}, preparedData.data, {
			offset: 0,
			bytesPerRow: preparedData.bytesPerRow,
			rowsPerImage: height
		}, {
			width,
			height,
			depthOrArrayLayers: depth
		});
		if (publicAPI.getDimensionality() !== 3 && model.mipLevel > 0) Texture_default$1.generateMipmaps(model.device.getHandle(), model.handle, model.mipLevel + 1);
		model.ready = true;
	};
	publicAPI.getScale = () => {
		const tDetails = Types_default.getDetailsFromTextureFormat(model.format);
		return tDetails.elementSize === 2 && tDetails.sampleType === "float" ? 1 : 255;
	};
	publicAPI.getNumberOfComponents = () => {
		return Types_default.getDetailsFromTextureFormat(model.format).numComponents;
	};
	publicAPI.getDimensionality = () => {
		let dims = 0;
		if (model.width > 1) dims++;
		if (model.height > 1) dims++;
		if (model.depth > 1) dims++;
		return dims;
	};
	publicAPI.resizeToMatch = (tex) => {
		if (tex.getWidth() !== model.width || tex.getHeight() !== model.height || tex.getDepth() !== model.depth) {
			model.width = tex.getWidth();
			model.height = tex.getHeight();
			model.depth = tex.getDepth();
			model.handle = model.device.getHandle().createTexture({
				size: [
					model.width,
					model.height,
					model.depth
				],
				dimension: model.dimension,
				format: model.format,
				usage: model.usage,
				label: model.label
			});
		}
	};
	publicAPI.resize = (width, height, depth = 1) => {
		if (width !== model.width || height !== model.height || depth !== model.depth) {
			model.width = width;
			model.height = height;
			model.depth = depth;
			model.handle = model.device.getHandle().createTexture({
				size: [
					model.width,
					model.height,
					model.depth
				],
				dimension: model.dimension,
				format: model.format,
				usage: model.usage,
				label: model.label
			});
		}
	};
	publicAPI.createView = (label, options = {}) => {
		if (!options.dimension) if (model.dimension === "3d") options.dimension = "3d";
		else if (model.depth === 1) options.dimension = "2d";
		else options.dimension = "2d-array";
		const view = TextureView_default.newInstance({ label });
		view.create(publicAPI, options);
		return view;
	};
}
var DEFAULT_VALUES = {
	device: null,
	dimension: "2d",
	handle: null,
	buffer: null,
	ready: false,
	label: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.get(publicAPI, model, [
		"dimension",
		"handle",
		"ready",
		"width",
		"height",
		"depth",
		"format",
		"usage"
	]);
	macros_default.setGet(publicAPI, model, ["device", "label"]);
	vtkWebGPUTexture(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend);
var Texture_default = {
	newInstance,
	extend
};
//#endregion
export { Texture_default as default, extend, newInstance };
