// @kitware/vtk.js/Rendering/WebGPU/TextureManager.js
import macros_default from "../../macros.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import Texture_default from "./Texture.js";
//#region Sources/Rendering/WebGPU/TextureManager/index.js
var { VtkDataTypes } = DataArray_default;
function vtkWebGPUTextureManager(publicAPI, model) {
	model.classHierarchy.push("vtkWebGPUTextureManager");
	function _fillRequest(req) {
		if (req.imageData) {
			req.dataArray = req.imageData.getPointData().getScalars();
			req.time = req.dataArray.getMTime();
			req.nativeArray = req.dataArray.getData();
			const dims = req.imageData.getDimensions();
			req.width = dims[0];
			req.height = dims[1];
			req.depth = dims[2];
			switch (req.dataArray.getNumberOfComponents()) {
				case 1:
					req.format = "r";
					break;
				case 2:
					req.format = "rg";
					break;
				default:
					req.format = "rgba";
					break;
			}
			switch (req.dataArray.getDataType()) {
				case VtkDataTypes.UNSIGNED_CHAR:
					req.format += "8unorm";
					break;
				case VtkDataTypes.FLOAT:
				case VtkDataTypes.UNSIGNED_INT:
				case VtkDataTypes.INT:
				case VtkDataTypes.DOUBLE:
				case VtkDataTypes.UNSIGNED_SHORT:
				case VtkDataTypes.SHORT:
				default:
					req.format += "16float";
					break;
			}
		}
		if (req.image) {
			req.width = req.image.width;
			req.height = req.image.height;
			req.depth = 1;
			req.format = "rgba8unorm";
			req.flip = true;
			req.usage = GPUTextureUsage.STORAGE_BINDING | GPUTextureUsage.COPY_DST | GPUTextureUsage.TEXTURE_BINDING | GPUTextureUsage.RENDER_ATTACHMENT;
		}
		if (req.jsImageData) {
			req.width = req.jsImageData.width;
			req.height = req.jsImageData.height;
			req.depth = 1;
			req.format = "rgba8unorm";
			req.flip = true;
			req.nativeArray = req.jsImageData.data;
			req.usage = GPUTextureUsage.STORAGE_BINDING | GPUTextureUsage.COPY_DST | GPUTextureUsage.TEXTURE_BINDING | GPUTextureUsage.RENDER_ATTACHMENT;
		}
		if (req.imageBitmap) {
			req.width = req.imageBitmap.width;
			req.height = req.imageBitmap.height;
			req.depth = 1;
			req.format = "rgba8unorm";
			req.flip = true;
			req.usage = GPUTextureUsage.STORAGE_BINDING | GPUTextureUsage.COPY_DST | GPUTextureUsage.TEXTURE_BINDING | GPUTextureUsage.RENDER_ATTACHMENT;
		}
		if (req.canvas) {
			req.width = req.canvas.width;
			req.height = req.canvas.height;
			req.depth = 1;
			req.format = "rgba8unorm";
			req.flip = true;
			req.usage = GPUTextureUsage.STORAGE_BINDING | GPUTextureUsage.COPY_DST | GPUTextureUsage.TEXTURE_BINDING | GPUTextureUsage.RENDER_ATTACHMENT;
		}
	}
	function _createTexture(req) {
		const newTex = Texture_default.newInstance({ label: req.label });
		newTex.create(model.device, {
			width: req.width,
			height: req.height,
			depth: req.depth,
			format: req.format,
			usage: req.usage,
			mipLevel: req.mipLevel
		});
		if (req.nativeArray || req.image || req.canvas || req.imageBitmap) newTex.writeImageData(req);
		return newTex;
	}
	publicAPI.getTexture = (req) => {
		if (req.hash) return model.device.getCachedObject(req.hash, _createTexture, req);
		return _createTexture(req);
	};
	publicAPI.getTextureForImageData = (imgData) => {
		const treq = { time: imgData.getMTime() };
		treq.imageData = imgData;
		_fillRequest(treq);
		treq.hash = treq.time + treq.format + treq.mipLevel;
		return model.device.getTextureManager().getTexture(treq);
	};
	publicAPI.getTextureForVTKTexture = (srcTexture, label = void 0) => {
		const treq = {
			time: srcTexture.getMTime(),
			label
		};
		if (srcTexture.getInputData()) treq.imageData = srcTexture.getInputData();
		else if (srcTexture.getImage()) treq.image = srcTexture.getImage();
		else if (srcTexture.getJsImageData()) treq.jsImageData = srcTexture.getJsImageData();
		else if (srcTexture.getImageBitmap()) treq.imageBitmap = srcTexture.getImageBitmap();
		else if (srcTexture.getCanvas()) treq.canvas = srcTexture.getCanvas();
		_fillRequest(treq);
		treq.mipLevel = srcTexture.getMipLevel();
		treq.hash = treq.time + treq.format + treq.mipLevel;
		return model.device.getTextureManager().getTexture(treq);
	};
}
var DEFAULT_VALUES = {
	handle: null,
	device: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, ["device"]);
	vtkWebGPUTextureManager(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend);
var TextureManager_default = {
	newInstance,
	extend
};
//#endregion
export { TextureManager_default as default, extend, newInstance };
