// @kitware/vtk.js/Interaction/UI/CornerAnnotation.js
import macros_default from "../../macros.js";
import CornerAnnotation_module_default from "./CornerAnnotation/CornerAnnotation.module.css.js";
//#region Sources/Interaction/UI/CornerAnnotation/index.js
function noOp() {}
var KEY_MAPPING = {
	nw: "northWestContainer",
	n: "northContainer",
	ne: "northEastContainer",
	w: "westContainer",
	e: "eastContainer",
	sw: "southWestContainer",
	s: "southContainer",
	se: "southEastContainer"
};
function get(path, obj, fb = `$\{${path}}`) {
	return path.split(".").reduce((res, key) => res[key] !== void 0 ? res[key] : fb, obj);
}
function applyTemplate(template, map, fallback) {
	return template.replace(/\${([^{]+)}/g, (match) => {
		return get(match.substr(2, match.length - 3).trim(), map, fallback);
	});
}
function vtkCornerAnnotation(publicAPI, model) {
	model.classHierarchy.push("vtkCornerAnnotation");
	if (!model.templates) model.templates = {};
	if (!model.metadata) model.metadata = {};
	model.annotationContainer = document.createElement("div");
	model.annotationContainer.setAttribute("class", CornerAnnotation_module_default.container);
	model.annotationContainer.innerHTML = `
    <div class="${CornerAnnotation_module_default.topRow}">
      <div class="js-nw ${CornerAnnotation_module_default.northWest}"></div>
      <div class="js-n ${CornerAnnotation_module_default.north}"></div>
      <div class="js-ne ${CornerAnnotation_module_default.northEast}"></div>
    </div>
    <div class="${CornerAnnotation_module_default.middleRow}">
      <div class="js-w ${CornerAnnotation_module_default.west}"></div>
      <div class="js-e ${CornerAnnotation_module_default.east}"></div>
    </div>
    <div class="${CornerAnnotation_module_default.bottomRow}">
      <div class="js-sw ${CornerAnnotation_module_default.southWest}"></div>
      <div class="js-s ${CornerAnnotation_module_default.south}"></div>
      <div class="js-se ${CornerAnnotation_module_default.southEast}"></div>
    </div>`;
	model.northWestContainer = model.annotationContainer.querySelector(".js-nw");
	model.northContainer = model.annotationContainer.querySelector(".js-n");
	model.northEastContainer = model.annotationContainer.querySelector(".js-ne");
	model.westContainer = model.annotationContainer.querySelector(".js-w");
	model.eastContainer = model.annotationContainer.querySelector(".js-e");
	model.southWestContainer = model.annotationContainer.querySelector(".js-sw");
	model.southContainer = model.annotationContainer.querySelector(".js-s");
	model.southEastContainer = model.annotationContainer.querySelector(".js-se");
	function updateAnnotations() {
		const keys = Object.keys(model.templates);
		let count = keys.length;
		while (count--) {
			const el = model[KEY_MAPPING[keys[count]]];
			const fn = model.templates[keys[count]];
			if (el && fn) el.innerHTML = fn(model.metadata);
		}
	}
	publicAPI.setContainer = (el) => {
		if (model.container && model.container !== el) model.container.removeChild(model.annotationContainer);
		if (model.container !== el) {
			model.container = el;
			if (model.container) {
				model.container.appendChild(model.annotationContainer);
				publicAPI.resize();
			}
			publicAPI.modified();
		}
	};
	publicAPI.resize = noOp;
	publicAPI.updateTemplates = (templates) => {
		model.templates = Object.assign(model.templates, templates);
		updateAnnotations();
		publicAPI.modified();
	};
	publicAPI.updateMetadata = (metadata) => {
		model.metadata = Object.assign(model.metadata, metadata);
		updateAnnotations();
		publicAPI.modified();
	};
}
var DEFAULT_VALUES = {
	templates: null,
	metadata: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.get(publicAPI, model, [
		"annotationContainer",
		"northWestContainer",
		"northContainer",
		"northEastContainer",
		"westContainer",
		"eastContainer",
		"southWestContainer",
		"southContainer",
		"southEastContainer",
		"metadata"
	]);
	vtkCornerAnnotation(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkCornerAnnotation");
var CornerAnnotation_default = {
	newInstance,
	extend,
	applyTemplate
};
//#endregion
export { CornerAnnotation_default as default, extend, newInstance };
