// @kitware/vtk.js/Rendering/OpenGL/BufferObject.js
import macros_default from "../../macros.js";
import Constants_default from "./BufferObject/Constants.js";
//#region Sources/Rendering/OpenGL/BufferObject/index.js
var { ObjectType } = Constants_default;
var STATIC = {};
function vtkOpenGLBufferObject(publicAPI, model) {
	model.classHierarchy.push("vtkOpenGLBufferObject");
	function convertType(type) {
		switch (type) {
			case ObjectType.ELEMENT_ARRAY_BUFFER: return model.context.ELEMENT_ARRAY_BUFFER;
			case ObjectType.TEXTURE_BUFFER: if ("TEXTURE_BUFFER" in model.context) return model.context.TEXTURE_BUFFER;
			case ObjectType.ARRAY_BUFFER:
			default: return model.context.ARRAY_BUFFER;
		}
	}
	let internalType = null;
	let internalHandle = null;
	let dirty = true;
	let error = "";
	publicAPI.getType = () => internalType;
	publicAPI.setType = (value) => {
		internalType = value;
	};
	publicAPI.getHandle = () => internalHandle;
	publicAPI.isReady = () => dirty === false;
	publicAPI.generateBuffer = (type) => {
		const objectTypeGL = convertType(type);
		if (internalHandle === null) {
			internalHandle = model.context.createBuffer();
			internalType = type;
		}
		return convertType(internalType) === objectTypeGL;
	};
	publicAPI.upload = (data, type) => {
		if (!publicAPI.generateBuffer(type)) {
			error = "Trying to upload array buffer to incompatible buffer.";
			return false;
		}
		model.context.bindBuffer(convertType(internalType), internalHandle);
		model.context.bufferData(convertType(internalType), data, model.context.STATIC_DRAW);
		model.allocatedGPUMemoryInBytes = data.length * data.BYTES_PER_ELEMENT;
		dirty = false;
		return true;
	};
	publicAPI.bind = () => {
		if (!internalHandle) return false;
		model.context.bindBuffer(convertType(internalType), internalHandle);
		return true;
	};
	publicAPI.release = () => {
		if (!internalHandle) return false;
		model.context.bindBuffer(convertType(internalType), null);
		return true;
	};
	publicAPI.releaseGraphicsResources = () => {
		if (internalHandle !== null) {
			model.context.bindBuffer(convertType(internalType), null);
			model.context.deleteBuffer(internalHandle);
			internalHandle = null;
			model.allocatedGPUMemoryInBytes = 0;
		}
	};
	publicAPI.setOpenGLRenderWindow = (rw) => {
		if (model._openGLRenderWindow === rw) return;
		publicAPI.releaseGraphicsResources();
		model._openGLRenderWindow = rw;
		model.context = null;
		if (rw) model.context = model._openGLRenderWindow.getContext();
	};
	publicAPI.getError = () => error;
}
var DEFAULT_VALUES = {
	objectType: ObjectType.ARRAY_BUFFER,
	context: null,
	allocatedGPUMemoryInBytes: 0
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.get(publicAPI, model, ["_openGLRenderWindow", "allocatedGPUMemoryInBytes"]);
	macros_default.moveToProtected(publicAPI, model, ["openGLRenderWindow"]);
	vtkOpenGLBufferObject(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend);
var BufferObject_default = {
	newInstance,
	extend,
	...STATIC,
	...Constants_default
};
//#endregion
export { STATIC, BufferObject_default as default, extend, newInstance };
