// @kitware/vtk.js/Rendering/OpenGL/Camera.js
import { newInstance as newInstance$1, obj, setGet } from "../../macros.js";
import ViewNode_default from "../SceneGraph/ViewNode.js";
import { registerOverride } from "./ViewNodeFactory.js";
import { mat3, mat4 } from "gl-matrix";
//#region Sources/Rendering/OpenGL/Camera/index.js
function vtkOpenGLCamera(publicAPI, model) {
	model.classHierarchy.push("vtkOpenGLCamera");
	publicAPI.buildPass = (prepass) => {
		if (prepass) {
			model._openGLRenderer = publicAPI.getFirstAncestorOfType("vtkOpenGLRenderer");
			model._openGLRenderWindow = model._openGLRenderer.getParent();
			model.context = model._openGLRenderWindow.getContext();
		}
	};
	publicAPI.opaquePass = (prepass) => {
		if (prepass) {
			const tsize = model._openGLRenderer.getTiledSizeAndOrigin();
			model.context.viewport(tsize.lowerLeftU, tsize.lowerLeftV, tsize.usize, tsize.vsize);
			model.context.scissor(tsize.lowerLeftU, tsize.lowerLeftV, tsize.usize, tsize.vsize);
		}
	};
	publicAPI.translucentPass = publicAPI.opaquePass;
	publicAPI.zBufferPass = publicAPI.opaquePass;
	publicAPI.opaqueZBufferPass = publicAPI.opaquePass;
	publicAPI.volumePass = publicAPI.opaquePass;
	publicAPI.getKeyMatrices = (ren) => {
		if (ren !== model.lastRenderer || model._openGLRenderWindow.getMTime() > model.keyMatrixTime.getMTime() || publicAPI.getMTime() > model.keyMatrixTime.getMTime() || ren.getMTime() > model.keyMatrixTime.getMTime() || model.renderable.getMTime() > model.keyMatrixTime.getMTime()) {
			mat4.copy(model.keyMatrices.wcvc, model.renderable.getViewMatrix());
			mat3.fromMat4(model.keyMatrices.normalMatrix, model.keyMatrices.wcvc);
			mat3.invert(model.keyMatrices.normalMatrix, model.keyMatrices.normalMatrix);
			mat4.transpose(model.keyMatrices.wcvc, model.keyMatrices.wcvc);
			const aspectRatio = model._openGLRenderer.getAspectRatio();
			mat4.copy(model.keyMatrices.vcpc, model.renderable.getProjectionMatrix(aspectRatio, -1, 1));
			mat4.transpose(model.keyMatrices.vcpc, model.keyMatrices.vcpc);
			mat4.multiply(model.keyMatrices.wcpc, model.keyMatrices.vcpc, model.keyMatrices.wcvc);
			model.keyMatrixTime.modified();
			model.lastRenderer = ren;
		}
		return model.keyMatrices;
	};
}
var DEFAULT_VALUES = {
	context: null,
	lastRenderer: null,
	keyMatrixTime: null,
	keyMatrices: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	ViewNode_default.extend(publicAPI, model, initialValues);
	model.keyMatrixTime = {};
	obj(model.keyMatrixTime);
	model.keyMatrices = {
		normalMatrix: new Float64Array(9),
		vcpc: new Float64Array(16),
		wcvc: new Float64Array(16),
		wcpc: new Float64Array(16)
	};
	setGet(publicAPI, model, ["context", "keyMatrixTime"]);
	vtkOpenGLCamera(publicAPI, model);
}
var newInstance = newInstance$1(extend);
var Camera_default = {
	newInstance,
	extend
};
registerOverride("vtkCamera", newInstance);
//#endregion
export { Camera_default as default, extend, newInstance };
