// @kitware/vtk.js/Rendering/OpenGL/ImageResliceMapper.js
import { chain, get, macros_exports, newInstance as newInstance$1, obj } from "../../macros.js";
import { VtkDataTypes } from "../../Common/Core/DataArray/Constants.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
import Transform_default from "../../Common/Transform/Transform.js";
import Cutter_default from "../../Filters/Core/Cutter.js";
import ClosedPolyLineToSurfaceFilter_default from "../../Filters/General/ClosedPolyLineToSurfaceFilter.js";
import ImageDataOutlineFilter_default from "../../Filters/General/ImageDataOutlineFilter.js";
import { Representation } from "../Core/Property/Constants.js";
import { Resolve } from "../Core/Mapper/Static.js";
import "../Core/Mapper/CoincidentTopologyHelper.js";
import { InterpolationType } from "../Core/ImageProperty/Constants.js";
import { Filter } from "./Texture/Constants.js";
import ViewNode_default from "../SceneGraph/ViewNode.js";
import { registerOverride } from "./ViewNodeFactory.js";
import Texture_default from "./Texture.js";
import ShaderProgram_default from "./ShaderProgram.js";
import Helper_default from "./Helper.js";
import ReplacementShaderMapper_default from "./ReplacementShaderMapper.js";
import { getImageDataHash, getTransferFunctionsHash } from "./RenderWindow/resourceSharingHelper.js";
import { fillLabelOutlineTextureTable, findLabelOutlineProperties, getLabelOutlineTextureParameters } from "../Core/ImageResliceMapper/LabelOutlineHelper.js";
import { computeColorShiftScale, computeOpacityShiftScale } from "../Core/ImageResliceMapper/TransferFunctionHelper.js";
import { computeInputOutlineBasis, computeSliceTangents } from "../Core/ImageResliceMapper/OutlineBasisHelper.js";
import { computeObliqueSliceGeometryData, computeOrthoSliceGeometryData, computeResliceGeometryState } from "../Core/ImageResliceMapper/ResliceGeometryHelper.js";
import vtkImageResliceMapperVS_default from "./glsl/vtkImageResliceMapperVS.glsl.js";
import vtkImageResliceMapperFS_default from "./glsl/vtkImageResliceMapperFS.glsl.js";
import { mat4 } from "gl-matrix";
//#region Sources/Rendering/OpenGL/ImageResliceMapper/index.js
var { vtkErrorMacro } = macros_exports;
var splitStringOnEnter = (str) => str.split("\n").map((s) => s.trim()).filter(Boolean);
function safeMatrixMultiply(matrixArray, matrixType, tmpMat) {
	matrixType.identity(tmpMat);
	return matrixArray.reduce((res, matrix, index) => {
		if (index === 0) return matrix ? matrixType.copy(res, matrix) : matrixType.identity(res);
		return matrix ? matrixType.multiply(res, res, matrix) : res;
	}, tmpMat);
}
function vtkOpenGLImageResliceMapper(publicAPI, model) {
	model.classHierarchy.push("vtkOpenGLImageResliceMapper");
	const graphicsResourceReferenceCount = /* @__PURE__ */ new Map();
	function decreaseGraphicsResourceCount(openGLRenderWindow, coreObject) {
		if (!coreObject) return;
		const newCount = (graphicsResourceReferenceCount.get(coreObject) ?? 0) - 1;
		if (newCount <= 0) {
			openGLRenderWindow.unregisterGraphicsResourceUser(coreObject, publicAPI);
			graphicsResourceReferenceCount.delete(coreObject);
		} else graphicsResourceReferenceCount.set(coreObject, newCount);
	}
	function increaseGraphicsResourceCount(openGLRenderWindow, coreObject) {
		if (!coreObject) return;
		const oldCount = graphicsResourceReferenceCount.get(coreObject) ?? 0;
		const newCount = oldCount + 1;
		graphicsResourceReferenceCount.set(coreObject, newCount);
		if (oldCount <= 0) openGLRenderWindow.registerGraphicsResourceUser(coreObject, publicAPI);
	}
	function replaceGraphicsResource(openGLRenderWindow, oldResourceCoreObject, newResourceCoreObject) {
		if (oldResourceCoreObject === newResourceCoreObject) return;
		decreaseGraphicsResourceCount(openGLRenderWindow, oldResourceCoreObject);
		increaseGraphicsResourceCount(openGLRenderWindow, newResourceCoreObject);
	}
	function unregisterGraphicsResources(renderWindow) {
		[...graphicsResourceReferenceCount.keys()].forEach((coreObject) => renderWindow.unregisterGraphicsResourceUser(coreObject, publicAPI));
	}
	publicAPI.buildPass = (prepass) => {
		if (prepass) {
			model.currentRenderPass = null;
			model._openGLImageSlice = publicAPI.getFirstAncestorOfType("vtkOpenGLImageSlice");
			model._openGLRenderer = publicAPI.getFirstAncestorOfType("vtkOpenGLRenderer");
			const ren = model._openGLRenderer.getRenderable();
			model._openGLCamera = model._openGLRenderer.getViewNodeFor(ren.getActiveCamera(), model.openGLCamera);
			const oldOglRenderWindow = model._openGLRenderWindow;
			model._openGLRenderWindow = model._openGLRenderer.getLastAncestorOfType("vtkOpenGLRenderWindow");
			if (oldOglRenderWindow && !oldOglRenderWindow.isDeleted() && oldOglRenderWindow !== model._openGLRenderWindow) unregisterGraphicsResources(oldOglRenderWindow);
			model.context = model._openGLRenderWindow.getContext();
			model.tris.setOpenGLRenderWindow(model._openGLRenderWindow);
		}
	};
	publicAPI.translucentPass = (prepass, renderPass) => {
		if (prepass) {
			model.currentRenderPass = renderPass;
			publicAPI.render();
		}
	};
	publicAPI.zBufferPass = (prepass) => {
		if (prepass) {
			model.haveSeenDepthRequest = true;
			model.renderDepth = true;
			publicAPI.render();
			model.renderDepth = false;
		}
	};
	publicAPI.opaqueZBufferPass = (prepass) => publicAPI.zBufferPass(prepass);
	publicAPI.opaquePass = (prepass) => {
		if (prepass) publicAPI.render();
	};
	publicAPI.getCoincidentParameters = (ren, actor) => {
		if (model.renderable.getResolveCoincidentTopology() == Resolve.PolygonOffset) return model.renderable.getCoincidentTopologyPolygonOffsetParameters();
		return null;
	};
	publicAPI.render = () => {
		const actor = model._openGLImageSlice.getRenderable();
		const ren = model._openGLRenderer.getRenderable();
		publicAPI.renderPiece(ren, actor);
	};
	publicAPI.renderPiece = (ren, actor) => {
		publicAPI.invokeEvent({ type: "StartEvent" });
		model.renderable.update();
		const numberOfInputs = model.renderable.getNumberOfInputPorts();
		model.currentValidInputs = [];
		for (let inputIndex = 0; inputIndex < numberOfInputs; ++inputIndex) {
			const imageData = model.renderable.getInputData(inputIndex);
			if (imageData && !imageData.isDeleted()) model.currentValidInputs.push({
				imageData,
				inputIndex
			});
		}
		const numberOfValidInputs = model.currentValidInputs.length;
		if (numberOfValidInputs <= 0) {
			vtkErrorMacro("No input!");
			return;
		}
		model.labelOutlineProperties = findLabelOutlineProperties(actor, model.currentValidInputs);
		const firstScalars = model.currentValidInputs[0].imageData.getPointData().getScalars();
		model.multiTexturePerVolumeEnabled = numberOfValidInputs > 1;
		model.numberOfComponents = model.multiTexturePerVolumeEnabled ? numberOfValidInputs : firstScalars.getNumberOfComponents();
		publicAPI.updateResliceGeometry();
		publicAPI.renderPieceStart(ren, actor);
		publicAPI.renderPieceDraw(ren, actor);
		publicAPI.renderPieceFinish(ren, actor);
		publicAPI.invokeEvent({ type: "EndEvent" });
	};
	publicAPI.renderPieceStart = (ren, actor) => {
		publicAPI.updateBufferObjects(ren, actor);
		const actorProperties = actor.getProperties();
		model.currentValidInputs.forEach(({ inputIndex }, component) => {
			const actorProperty = actorProperties[inputIndex];
			const scalarTexture = model.scalarTextures[component];
			if (!actorProperty || !scalarTexture) return;
			if (actorProperty.getInterpolationType() === InterpolationType.NEAREST) {
				scalarTexture.setMinificationFilter(Filter.NEAREST);
				scalarTexture.setMagnificationFilter(Filter.NEAREST);
			} else {
				scalarTexture.setMinificationFilter(Filter.LINEAR);
				scalarTexture.setMagnificationFilter(Filter.LINEAR);
			}
		});
		if (actorProperties[model.currentValidInputs[0].inputIndex]?.getInterpolationType() === InterpolationType.NEAREST) {
			model.colorTexture.setMinificationFilter(Filter.NEAREST);
			model.colorTexture.setMagnificationFilter(Filter.NEAREST);
			model.pwfTexture.setMinificationFilter(Filter.NEAREST);
			model.pwfTexture.setMagnificationFilter(Filter.NEAREST);
		} else {
			model.colorTexture.setMinificationFilter(Filter.LINEAR);
			model.colorTexture.setMagnificationFilter(Filter.LINEAR);
			model.pwfTexture.setMinificationFilter(Filter.LINEAR);
			model.pwfTexture.setMagnificationFilter(Filter.LINEAR);
		}
		model.lastBoundBO = null;
	};
	publicAPI.renderPieceDraw = (ren, actor) => {
		const gl = model.context;
		const useLabelOutline = model.labelOutlineProperties.length > 0;
		const allTextures = [
			...model.scalarTextures,
			model.colorTexture,
			model.pwfTexture
		];
		if (useLabelOutline) {
			allTextures.push(model.labelOutlineThicknessTexture);
			allTextures.push(model.labelOutlineOpacityTexture);
		}
		allTextures.forEach((texture) => texture.activate());
		publicAPI.updateShaders(model.tris, ren, actor);
		gl.drawArrays(gl.TRIANGLES, 0, model.tris.getCABO().getElementCount());
		model.tris.getVAO().release();
		allTextures.forEach((texture) => texture.deactivate());
	};
	publicAPI.renderPieceFinish = (ren, actor) => {};
	publicAPI.updateBufferObjects = (ren, actor) => {
		if (publicAPI.getNeedToRebuildBufferObjects(ren, actor)) publicAPI.buildBufferObjects(ren, actor);
	};
	publicAPI.getNeedToRebuildBufferObjects = (ren, actor) => {
		const firstActorProperty = actor.getProperty(model.currentValidInputs[0].inputIndex);
		const useLabelOutline = model.labelOutlineProperties.length > 0;
		return model.VBOBuildTime.getMTime() < publicAPI.getMTime() || model.VBOBuildTime.getMTime() < actor.getMTime() || model.VBOBuildTime.getMTime() < model.renderable.getMTime() || model.VBOBuildTime.getMTime() < firstActorProperty?.getMTime() || model.currentValidInputs.some(({ imageData }) => model.VBOBuildTime.getMTime() < imageData.getMTime()) || model.VBOBuildTime.getMTime() < model.resliceGeom.getMTime() || model.scalarTextures.length !== model.currentValidInputs.length || !model.scalarTextures.every((texture) => !!texture?.getHandle()) || !model.colorTexture?.getHandle() || !model.pwfTexture?.getHandle() || useLabelOutline && (!model.labelOutlineThicknessTexture?.getHandle() || !model.labelOutlineOpacityTexture?.getHandle());
	};
	publicAPI.buildBufferObjects = (ren, actor) => {
		const actorProperties = actor.getProperties();
		model.currentValidInputs.forEach(({ imageData, inputIndex }, component) => {
			const scalars = imageData.getPointData().getScalars();
			const tex = model._openGLRenderWindow.getGraphicsResourceForObject(scalars);
			const scalarsHash = getImageDataHash(imageData, scalars);
			const reBuildTex = !tex?.oglObject?.getHandle() || tex?.hash !== scalarsHash;
			const actorProperty = actorProperties[inputIndex];
			const updatedExtents = actorProperty?.getUpdatedExtents() ?? [];
			const hasUpdatedExtents = !!updatedExtents.length;
			if (reBuildTex && !hasUpdatedExtents) {
				const newScalarTexture = Texture_default.newInstance();
				newScalarTexture.setOpenGLRenderWindow(model._openGLRenderWindow);
				const dims = imageData.getDimensions();
				newScalarTexture.setOglNorm16Ext(model.context.getExtension("EXT_texture_norm16"));
				newScalarTexture.resetFormatAndType();
				newScalarTexture.create3DFilterableFromDataArray({
					width: dims[0],
					height: dims[1],
					depth: dims[2],
					dataArray: scalars
				});
				model._openGLRenderWindow.setGraphicsResourceForObject(scalars, newScalarTexture, scalarsHash);
				model.scalarTextures[component] = newScalarTexture;
			} else model.scalarTextures[component] = tex.oglObject;
			if (hasUpdatedExtents) {
				actorProperty.setUpdatedExtents([]);
				const dims = imageData.getDimensions();
				model.scalarTextures[component].create3DFilterableFromDataArray({
					width: dims[0],
					height: dims[1],
					depth: dims[2],
					dataArray: scalars,
					updatedExtents
				});
			}
			replaceGraphicsResource(model._openGLRenderWindow, model._scalarTexturesCore[component], scalars);
			model._scalarTexturesCore[component] = scalars;
		});
		const firstActorProperty = actorProperties[model.currentValidInputs[0].inputIndex];
		if (!firstActorProperty) {
			vtkErrorMacro("Missing property for first input");
			return;
		}
		const iComps = firstActorProperty.getIndependentComponents();
		const numIComps = iComps ? model.numberOfComponents : 1;
		const textureHeight = iComps ? 2 * numIComps : 1;
		const colorTransferFunctions = [];
		for (let component = 0; component < numIComps; ++component) if (model.multiTexturePerVolumeEnabled) {
			const validInput = model.currentValidInputs[component];
			const prop = validInput ? actorProperties[validInput.inputIndex] : null;
			colorTransferFunctions.push(prop?.getRGBTransferFunction() || null);
		} else colorTransferFunctions.push(firstActorProperty.getRGBTransferFunction(component));
		const colorFuncHash = getTransferFunctionsHash(colorTransferFunctions, iComps, numIComps);
		const firstColorTransferFunc = firstActorProperty.getRGBTransferFunction();
		const cTex = model._openGLRenderWindow.getGraphicsResourceForObject(firstColorTransferFunc);
		if (!cTex?.oglObject?.getHandle() || cTex?.hash !== colorFuncHash) {
			let cWidth = model.renderable.getColorTextureWidth();
			if (cWidth <= 0) cWidth = model.context.getParameter(model.context.MAX_TEXTURE_SIZE);
			const cSize = cWidth * textureHeight * 3;
			const cTable = new Uint8ClampedArray(cSize);
			const newColorTexture = Texture_default.newInstance();
			newColorTexture.setOpenGLRenderWindow(model._openGLRenderWindow);
			if (firstColorTransferFunc) {
				const tmpTable = new Float32Array(cWidth * 3);
				for (let c = 0; c < numIComps; c++) {
					const cfun = colorTransferFunctions[c];
					if (cfun) {
						const cRange = cfun.getRange();
						cfun.getTable(cRange[0], cRange[1], cWidth, tmpTable, 1);
						if (iComps) for (let i = 0; i < cWidth * 3; i++) {
							cTable[c * cWidth * 6 + i] = 255 * tmpTable[i];
							cTable[c * cWidth * 6 + i + cWidth * 3] = 255 * tmpTable[i];
						}
						else for (let i = 0; i < cWidth * 3; i++) cTable[c * cWidth * 3 + i] = 255 * tmpTable[i];
					}
				}
				newColorTexture.resetFormatAndType();
				newColorTexture.create2DFromRaw({
					width: cWidth,
					height: textureHeight,
					numComps: 3,
					dataType: VtkDataTypes.UNSIGNED_CHAR,
					data: cTable
				});
			} else {
				for (let column = 0; column < cWidth * 3; ++column) {
					const opacity = 255 * column / ((cWidth - 1) * 3);
					for (let row = 0; row < textureHeight; ++row) {
						cTable[row * cWidth * 3 + column + 0] = opacity;
						cTable[row * cWidth * 3 + column + 1] = opacity;
						cTable[row * cWidth * 3 + column + 2] = opacity;
					}
				}
				newColorTexture.resetFormatAndType();
				newColorTexture.create2DFromRaw({
					width: cWidth,
					height: 1,
					numComps: 3,
					dataType: VtkDataTypes.UNSIGNED_CHAR,
					data: cTable
				});
			}
			if (firstColorTransferFunc) model._openGLRenderWindow.setGraphicsResourceForObject(firstColorTransferFunc, newColorTexture, colorFuncHash);
			model.colorTexture = newColorTexture;
		} else model.colorTexture = cTex.oglObject;
		replaceGraphicsResource(model._openGLRenderWindow, model._colorTextureCore, firstColorTransferFunc);
		model._colorTextureCore = firstColorTransferFunc;
		const opacityFunctions = [];
		for (let component = 0; component < numIComps; ++component) if (model.multiTexturePerVolumeEnabled) {
			const validInput = model.currentValidInputs[component];
			const prop = validInput ? actorProperties[validInput.inputIndex] : null;
			opacityFunctions.push(prop?.getPiecewiseFunction() || null);
		} else opacityFunctions.push(firstActorProperty.getPiecewiseFunction(component));
		const opacityFuncHash = getTransferFunctionsHash(opacityFunctions, iComps, numIComps);
		const firstPwFunc = firstActorProperty.getPiecewiseFunction();
		const pwfTex = model._openGLRenderWindow.getGraphicsResourceForObject(firstPwFunc);
		if (!pwfTex?.oglObject?.getHandle() || pwfTex?.hash !== opacityFuncHash) {
			let pwfWidth = model.renderable.getOpacityTextureWidth();
			if (pwfWidth <= 0) pwfWidth = model.context.getParameter(model.context.MAX_TEXTURE_SIZE);
			const pwfSize = pwfWidth * textureHeight;
			const pwfTable = new Uint8ClampedArray(pwfSize);
			const newOpacityTexture = Texture_default.newInstance();
			newOpacityTexture.setOpenGLRenderWindow(model._openGLRenderWindow);
			if (firstPwFunc) {
				const pwfFloatTable = new Float32Array(pwfSize);
				const tmpTable = new Float32Array(pwfWidth);
				for (let c = 0; c < numIComps; ++c) {
					const pwfun = opacityFunctions[c];
					if (pwfun === null) pwfFloatTable.fill(1);
					else {
						const pwfRange = pwfun.getRange();
						pwfun.getTable(pwfRange[0], pwfRange[1], pwfWidth, tmpTable, 1);
						if (iComps) for (let i = 0; i < pwfWidth; i++) {
							pwfFloatTable[c * pwfWidth * 2 + i] = tmpTable[i];
							pwfFloatTable[c * pwfWidth * 2 + i + pwfWidth] = tmpTable[i];
						}
						else for (let i = 0; i < pwfWidth; i++) pwfFloatTable[i] = tmpTable[i];
					}
				}
				newOpacityTexture.resetFormatAndType();
				newOpacityTexture.create2DFromRaw({
					width: pwfWidth,
					height: textureHeight,
					numComps: 1,
					dataType: VtkDataTypes.FLOAT,
					data: pwfFloatTable
				});
			} else {
				pwfTable.fill(255);
				newOpacityTexture.resetFormatAndType();
				newOpacityTexture.create2DFromRaw({
					width: pwfWidth,
					height: textureHeight,
					numComps: 1,
					dataType: VtkDataTypes.UNSIGNED_CHAR,
					data: pwfTable
				});
			}
			if (firstPwFunc) model._openGLRenderWindow.setGraphicsResourceForObject(firstPwFunc, newOpacityTexture, opacityFuncHash);
			model.pwfTexture = newOpacityTexture;
		} else model.pwfTexture = pwfTex.oglObject;
		replaceGraphicsResource(model._openGLRenderWindow, model._pwfTextureCore, firstPwFunc);
		model._pwfTextureCore = firstPwFunc;
		if (model.labelOutlineProperties.length > 0) {
			publicAPI.updateLabelOutlineThicknessTexture(model.labelOutlineProperties);
			publicAPI.updateLabelOutlineOpacityTexture(model.labelOutlineProperties);
		}
		const vboString = `${model.resliceGeom.getMTime()}A${model.renderable.getSlabThickness()}`;
		if (!model.tris.getCABO().getElementCount() || model.VBOBuildString !== vboString) {
			const points = DataArray_default.newInstance({
				numberOfComponents: 3,
				values: model.resliceGeom.getPoints().getData()
			});
			points.setName("points");
			const cells = DataArray_default.newInstance({
				numberOfComponents: 1,
				values: model.resliceGeom.getPolys().getData()
			});
			const options = {
				points,
				cellOffset: 0
			};
			if (model.renderable.getSlabThickness() > 0) {
				const n = model.resliceGeom.getPointData().getNormals();
				if (!n) vtkErrorMacro("Slab mode requested without normals");
				else options.normals = n;
			}
			model.tris.getCABO().createVBO(cells, "polys", Representation.SURFACE, options);
		}
		model.VBOBuildString = vboString;
		model.VBOBuildTime.modified();
	};
	publicAPI.updateShaders = (cellBO, ren, actor) => {
		model.lastBoundBO = cellBO;
		if (publicAPI.getNeedToRebuildShaders(cellBO, ren, actor)) {
			const shaders = {
				Vertex: null,
				Fragment: null,
				Geometry: null
			};
			publicAPI.buildShaders(shaders, ren, actor);
			const newShader = model._openGLRenderWindow.getShaderCache().readyShaderProgramArray(shaders.Vertex, shaders.Fragment, shaders.Geometry);
			if (newShader !== cellBO.getProgram()) {
				cellBO.setProgram(newShader);
				cellBO.getVAO().releaseGraphicsResources();
			}
			cellBO.getShaderSourceTime().modified();
		} else model._openGLRenderWindow.getShaderCache().readyShaderProgram(cellBO.getProgram());
		cellBO.getVAO().bind();
		publicAPI.setMapperShaderParameters(cellBO, ren, actor);
		publicAPI.setCameraShaderParameters(cellBO, ren, actor);
		publicAPI.setPropertyShaderParameters(cellBO, ren, actor);
	};
	publicAPI.setMapperShaderParameters = (cellBO, ren, actor) => {
		const program = cellBO.getProgram();
		const firstImageData = model.currentValidInputs[0].imageData;
		if (cellBO.getCABO().getElementCount() && (model.VBOBuildTime.getMTime() > cellBO.getAttributeUpdateTime().getMTime() || cellBO.getShaderSourceTime().getMTime() > cellBO.getAttributeUpdateTime().getMTime())) {
			model.scalarTextures.forEach((scalarTexture, component) => {
				program.setUniformi(`volumeTexture[${component}]`, scalarTexture.getTextureUnit());
			});
			if (program.isAttributeUsed("vertexWC")) {
				if (!cellBO.getVAO().addAttributeArray(program, cellBO.getCABO(), "vertexWC", cellBO.getCABO().getVertexOffset(), cellBO.getCABO().getStride(), model.context.FLOAT, 3, model.context.FALSE)) vtkErrorMacro("Error setting vertexWC in shader VAO.");
			}
			if (program.isAttributeUsed("normalWC")) {
				if (!cellBO.getVAO().addAttributeArray(program, cellBO.getCABO(), "normalWC", cellBO.getCABO().getNormalOffset(), cellBO.getCABO().getStride(), model.context.FLOAT, 3, model.context.FALSE)) vtkErrorMacro("Error setting normalWC in shader VAO.");
			}
			if (program.isUniformUsed("slabThickness")) program.setUniformf("slabThickness", model.renderable.getSlabThickness());
			if (program.isUniformUsed("spacing")) program.setUniform3fv("spacing", firstImageData.getSpacing());
			if (program.isUniformUsed("slabType")) program.setUniformi("slabType", model.renderable.getSlabType());
			if (program.isUniformUsed("slabTrapezoid")) program.setUniformi("slabTrapezoid", model.renderable.getSlabTrapezoidIntegration());
			const inverseShiftScaleMatrix = cellBO.getCABO().getCoordShiftAndScaleEnabled() ? cellBO.getCABO().getInverseShiftAndScaleMatrix() : null;
			for (let i = 0; i < model.currentValidInputs.length; i++) {
				const uniformName = `WCTCMatrix${i}`;
				if (program.isUniformUsed(uniformName)) {
					const imageData = model.currentValidInputs[i].imageData;
					const dim = imageData.getDimensions();
					mat4.copy(model.tmpMat4, imageData.getIndexToWorld());
					mat4.translate(model.tmpMat4, model.tmpMat4, [
						-.5,
						-.5,
						-.5
					]);
					mat4.scale(model.tmpMat4, model.tmpMat4, dim);
					mat4.invert(model.tmpMat4, model.tmpMat4);
					if (inverseShiftScaleMatrix) mat4.multiply(model.tmpMat4, model.tmpMat4, inverseShiftScaleMatrix);
					program.setUniformMatrix(uniformName, model.tmpMat4);
				}
			}
			if (program.isUniformUsed("vboScaling")) program.setUniform3fv("vboScaling", cellBO.getCABO().getCoordScale() ?? [
				1,
				1,
				1
			]);
			cellBO.getAttributeUpdateTime().modified();
		}
		if (model.haveSeenDepthRequest) cellBO.getProgram().setUniformi("depthRequest", model.renderDepth ? 1 : 0);
		if (cellBO.getProgram().isUniformUsed("coffset")) {
			const cp = publicAPI.getCoincidentParameters(ren, actor);
			cellBO.getProgram().setUniformf("coffset", cp.offset);
			if (cellBO.getProgram().isUniformUsed("cfactor")) cellBO.getProgram().setUniformf("cfactor", cp.factor);
		}
	};
	publicAPI.setCameraShaderParameters = (cellBO, ren, actor) => {
		const keyMats = model._openGLCamera.getKeyMatrices(ren);
		const actMats = model._openGLImageSlice.getKeyMatrices();
		const inverseShiftScaleMatrix = cellBO.getCABO().getCoordShiftAndScaleEnabled() ? cellBO.getCABO().getInverseShiftAndScaleMatrix() : null;
		const program = cellBO.getProgram();
		if (program.isUniformUsed("MCPCMatrix")) {
			mat4.identity(model.tmpMat4);
			program.setUniformMatrix("MCPCMatrix", safeMatrixMultiply([
				keyMats.wcpc,
				actMats.mcwc,
				inverseShiftScaleMatrix
			], mat4, model.tmpMat4));
		}
		if (program.isUniformUsed("MCVCMatrix")) {
			mat4.identity(model.tmpMat4);
			program.setUniformMatrix("MCVCMatrix", safeMatrixMultiply([
				keyMats.wcvc,
				actMats.mcwc,
				inverseShiftScaleMatrix
			], mat4, model.tmpMat4));
		}
	};
	publicAPI.setPropertyShaderParameters = (cellBO, ren, actor) => {
		const program = cellBO.getProgram();
		const firstPpty = actor.getProperty(model.currentValidInputs[0].inputIndex);
		const opacity = model.multiTexturePerVolumeEnabled ? 1 : firstPpty.getOpacity();
		program.setUniformf("opacity", opacity);
		const numComp = model.numberOfComponents;
		const iComps = firstPpty.getIndependentComponents();
		const useMultiTexture = model.multiTexturePerVolumeEnabled;
		const actorProperties = actor.getProperties();
		if (iComps) for (let i = 0; i < numComp; ++i) {
			const property = useMultiTexture ? actorProperties[model.currentValidInputs[i].inputIndex] : firstPpty;
			program.setUniformf(`mix${i}`, property.getComponentWeight(0));
		}
		for (let component = 0; component < numComp; component++) {
			const textureIndex = useMultiTexture ? component : 0;
			const volInfoIndex = useMultiTexture ? 0 : component;
			const volInfo = model.scalarTextures[textureIndex].getVolumeInfo();
			const volScale = volInfo.scale[volInfoIndex];
			const volOffset = volInfo.offset[volInfoIndex];
			const target = iComps ? component : 0;
			const property = useMultiTexture ? actorProperties[model.currentValidInputs[component].inputIndex] : firstPpty;
			const cw = property.getColorWindow();
			const cl = property.getColorLevel();
			const cfun = property.getRGBTransferFunction(useMultiTexture ? 0 : target);
			const { colorScale, colorShift } = computeColorShiftScale({
				colorWindow: cw,
				colorLevel: cl,
				useLookupTableScalarRange: property.getUseLookupTableScalarRange(),
				colorRange: cfun?.getRange?.(),
				volumeScale: volScale,
				volumeOffset: volOffset
			});
			program.setUniformf(`cshift${component}`, colorShift);
			program.setUniformf(`cscale${component}`, colorScale);
			const { opacityScale: pwfScale, opacityShift: pwfShift } = computeOpacityShiftScale({
				pwfRange: property.getPiecewiseFunction(useMultiTexture ? 0 : target)?.getRange?.(),
				volumeScale: volScale,
				volumeOffset: volOffset
			});
			program.setUniformf(`pwfshift${component}`, pwfShift);
			program.setUniformf(`pwfscale${component}`, pwfScale);
		}
		const texColorUnit = model.colorTexture.getTextureUnit();
		program.setUniformi("colorTexture1", texColorUnit);
		const texOpacityUnit = model.pwfTexture.getTextureUnit();
		program.setUniformi("pwfTexture1", texOpacityUnit);
		program.setUniform4fv("backgroundColor", model.renderable.getBackgroundColor());
		if (model.labelOutlineProperties.length > 0) {
			const outlineThicknessUnit = model.labelOutlineThicknessTexture.getTextureUnit();
			program.setUniformi("labelOutlineThicknessTexture", outlineThicknessUnit);
			const outlineOpacityUnit = model.labelOutlineOpacityTexture.getTextureUnit();
			program.setUniformi("labelOutlineOpacityTexture", outlineOpacityUnit);
			let textureWidth = model.renderable.getLabelOutlineTextureWidth();
			if (textureWidth <= 0) textureWidth = model.context.getParameter(model.context.MAX_TEXTURE_SIZE);
			program.setUniformf("labelOutlineTextureWidth", textureWidth);
			program.setUniformf("numLabelmaps", model.labelOutlineProperties.length);
			const { tangent1, tangent2 } = computeSliceTangents(model.renderable.getSlicePlane());
			for (let i = 0; i < model.currentValidInputs.length; i++) {
				const imageData = model.currentValidInputs[i].imageData;
				const { tangent1: inputT1, tangent2: inputT2 } = computeInputOutlineBasis(imageData, tangent1, tangent2, model._tmpTexelSize);
				const t1Name = `outlineTangent1_${i}`;
				const t2Name = `outlineTangent2_${i}`;
				if (program.isUniformUsed(t1Name)) program.setUniform3fv(t1Name, inputT1);
				if (program.isUniformUsed(t2Name)) program.setUniform3fv(t2Name, inputT2);
			}
			for (let i = 0; i < model.currentValidInputs.length; i++) {
				const uniformName = `texelSize${i}`;
				if (program.isUniformUsed(uniformName)) {
					const imageData = model.currentValidInputs[i].imageData;
					const { texelSize } = computeInputOutlineBasis(imageData, tangent1, tangent2, model._tmpTexelSize);
					program.setUniform3fv(uniformName, texelSize);
				}
			}
		}
	};
	publicAPI.getNeedToRebuildShaders = (cellBO, ren, actor) => {
		const iComp = actor.getProperty(model.currentValidInputs[0].inputIndex).getIndependentComponents();
		const useLabelOutline = model.labelOutlineProperties.length > 0;
		const slabTh = model.renderable.getSlabThickness();
		const slabType = model.renderable.getSlabType();
		const slabTrap = model.renderable.getSlabTrapezoidIntegration();
		let needRebuild = false;
		if (!model.currentRenderPass && model.lastRenderPassShaderReplacement || model.currentRenderPass && model.currentRenderPass.getShaderReplacement() !== model.lastRenderPassShaderReplacement) needRebuild = true;
		const numValidInputs = model.currentValidInputs?.length ?? 0;
		if (needRebuild || model.lastHaveSeenDepthRequest !== model.haveSeenDepthRequest || model.lastNumberOfComponents !== model.numberOfComponents || model.lastMultiTexturePerVolumeEnabled !== model.multiTexturePerVolumeEnabled || cellBO.getProgram()?.getHandle() === 0 || model.lastIndependentComponents !== iComp || model.lastUseLabelOutline !== useLabelOutline || model.lastNumValidInputs !== numValidInputs || model.lastSlabThickness !== slabTh || model.lastSlabType !== slabType || model.lastSlabTrapezoidIntegration !== slabTrap) {
			model.lastHaveSeenDepthRequest = model.haveSeenDepthRequest;
			model.lastNumberOfComponents = model.numberOfComponents;
			model.lastMultiTexturePerVolumeEnabled = model.multiTexturePerVolumeEnabled;
			model.lastIndependentComponents = iComp;
			model.lastUseLabelOutline = useLabelOutline;
			model.lastNumValidInputs = numValidInputs;
			model.lastSlabThickness = slabTh;
			model.lastSlabType = slabType;
			model.lastSlabTrapezoidIntegration = slabTrap;
			return true;
		}
		return false;
	};
	publicAPI.getShaderTemplate = (shaders, ren, actor) => {
		shaders.Vertex = vtkImageResliceMapperVS_default;
		shaders.Fragment = vtkImageResliceMapperFS_default;
		shaders.Geometry = "";
	};
	publicAPI.replaceShaderValues = (shaders, ren, actor) => {
		publicAPI.replaceShaderTCoord(shaders, ren, actor);
		publicAPI.replaceShaderPositionVC(shaders, ren, actor);
		if (model.haveSeenDepthRequest) {
			let FSSource = shaders.Fragment;
			FSSource = ShaderProgram_default.substitute(FSSource, "//VTK::ZBuffer::Dec", "uniform int depthRequest;").result;
			FSSource = ShaderProgram_default.substitute(FSSource, "//VTK::ZBuffer::Impl", [
				"if (depthRequest == 1) {",
				"float iz = floor(gl_FragCoord.z*65535.0 + 0.1);",
				"float rf = floor(iz/256.0)/255.0;",
				"float gf = mod(iz,256.0)/255.0;",
				"gl_FragData[0] = vec4(rf, gf, 0.0, 1.0); }"
			]).result;
			shaders.Fragment = FSSource;
		}
		publicAPI.replaceShaderCoincidentOffset(shaders, ren, actor);
	};
	function generateMultiInputCompositeShader(labelmapInputs, totalInputs) {
		const rgba = [
			"r",
			"g",
			"b",
			"a"
		];
		const backgroundInputs = Array.from({ length: totalInputs }, (_, i) => i).filter((i) => !labelmapInputs.includes(i));
		const texCoordLines = labelmapInputs.map((i) => `vec3 labelTexCoord${i} = (WCTCMatrix${i} * vec4(fragWorldPos, 1.0)).xyz;`).join("\n                ");
		const textureSampling = (() => {
			if (labelmapInputs.length === 0) return "";
			return `float neighborLabel = ${labelmapInputs.map((inputIdx, arrayIdx) => {
				if (arrayIdx === 0) return `(labelInputIdx == ${arrayIdx}) ? texture(volumeTexture[${inputIdx}], neighborTexCoord).r`;
				return ` : (labelInputIdx == ${arrayIdx}) ? texture(volumeTexture[${inputIdx}], neighborTexCoord).r`;
			}).join("")} : 0.0;`;
		})();
		const processInputs = [...backgroundInputs, ...labelmapInputs].map((inputIdx) => {
			const isLabelmap = labelmapInputs.includes(inputIdx);
			const labelArrayIdx = labelmapInputs.indexOf(inputIdx);
			if (isLabelmap) return `
        // Process input ${inputIdx} as labelmap
        {
          float labelValue = tvalue.${rgba[inputIdx]};
          int segmentIndex = int(labelValue * 255.0);

          if (segmentIndex > 0) {
            float textureCoordinate = float(segmentIndex - 1) / labelOutlineTextureWidth;
            float labelmapRow = (float(${labelArrayIdx}) + 0.5) / numLabelmaps;
            float thicknessValue = texture2D(labelOutlineThicknessTexture, vec2(textureCoordinate, labelmapRow)).r;
            float labelOutlineOpacityValue = texture2D(labelOutlineOpacityTexture, vec2(textureCoordinate, labelmapRow)).r;
            int actualThickness = int(thicknessValue * 255.0);

            vec3 currentLabelTC = labelTexCoord${inputIdx};
            vec3 currentTexelSize = texelSize${inputIdx};
            vec3 currentTangent1 = outlineTangent1_${inputIdx};
            vec3 currentTangent2 = outlineTangent2_${inputIdx};

            bool pixelOnBorder = false;
            int labelInputIdx = ${labelArrayIdx};
            for (int i = -actualThickness; i <= actualThickness; i++) {
              for (int j = -actualThickness; j <= actualThickness; j++) {
                if (i == 0 && j == 0) continue;
                vec3 neighborTexCoord = currentLabelTC + float(i) * currentTangent1 * currentTexelSize + float(j) * currentTangent2 * currentTexelSize;
                if (any(greaterThan(neighborTexCoord, vec3(1.0))) || any(lessThan(neighborTexCoord, vec3(0.0)))) {
                  pixelOnBorder = true;
                  break;
                }
                ${textureSampling}
                if (neighborLabel != labelValue) {
                  pixelOnBorder = true;
                  break;
                }
              }
              if (pixelOnBorder) break;
            }

            if (pixelOnBorder) {
              convergentColor.rgb = mix(convergentColor.rgb, tcolor${inputIdx}.rgb, labelOutlineOpacityValue);
              convergentColor.a = max(convergentColor.a, labelOutlineOpacityValue);
            } else if (compWeight${inputIdx} > 0.0) {
              float fillAlpha = compWeight${inputIdx} * opacity;
              convergentColor.rgb = mix(convergentColor.rgb, tcolor${inputIdx}.rgb, fillAlpha);
              convergentColor.a = max(convergentColor.a, fillAlpha);
            }
          }
        }`;
			return `
        // Process input ${inputIdx} as background image
        {
          float bgAlpha = compWeight${inputIdx} * opacity;
          convergentColor.rgb = mix(convergentColor.rgb, tcolor${inputIdx}.rgb, bgAlpha);
          convergentColor.a = max(convergentColor.a, bgAlpha);
        }`;
		}).join("\n        ");
		return splitStringOnEnter(`
      // Multi-texture mode: ${labelmapInputs.length > 0 ? `labelmaps at input${labelmapInputs.length > 1 ? "s" : ""} ${labelmapInputs.join(", ")}` : "no labelmaps"}, ${backgroundInputs.length > 0 ? `background at input${backgroundInputs.length > 1 ? "s" : ""} ${backgroundInputs.join(", ")}` : "no background"}
      vec4 convergentColor = vec4(0.0, 0.0, 0.0, 0.0);

      // Compute labelmap texture coordinates
      ${texCoordLines}

      // Process each input in order
      ${processInputs}

      gl_FragData[0] = convergentColor;
    `);
	}
	function getSlabCompositeDecLines() {
		return [
			"vec4 compositeValue(vec4 currVal, vec4 valToComp, int trapezoid)",
			"{",
			"  vec4 retVal = vec4(1.0);",
			"  if (slabType == 0) // min",
			"  {",
			"    retVal = min(currVal, valToComp);",
			"  }",
			"  else if (slabType == 1) // max",
			"  {",
			"    retVal = max(currVal, valToComp);",
			"  }",
			"  else if (slabType == 3) // sum",
			"  {",
			"    retVal = currVal + (trapezoid > 0 ? 0.5 * valToComp : valToComp); ",
			"  }",
			"  else // mean",
			"  {",
			"    retVal = currVal + (trapezoid > 0 ? 0.5 * valToComp : valToComp); ",
			"  }",
			"  return retVal;",
			"}"
		];
	}
	function getSlabSampleLoopLines() {
		return [
			"// Get the first and last samples",
			"int numSlices = 1;",
			"float scaling = min(min(spacing.x, spacing.y), spacing.z) * 0.5;",
			"vec3 slabNormal = normalize(normalWCVSOutput);",
			"vec3 normalxspacing = scaling * slabNormal;",
			"float distTraveled = length(normalxspacing);",
			"int trapezoid = 0;",
			"while (distTraveled < slabThickness * 0.5)",
			"{",
			"  distTraveled += length(normalxspacing);",
			"  float fnumSlices = float(numSlices);",
			"  if (distTraveled > slabThickness * 0.5)",
			"  {",
			"    // Before stepping outside the slab, sample at the boundaries",
			"    normalxspacing = slabNormal * slabThickness * 0.5 / fnumSlices;",
			"    trapezoid = slabTrapezoid;",
			"  }",
			"  vec3 worldPosNeg = vertexWCVSOutput.xyz - fnumSlices * normalxspacing * vboScaling;",
			"  vec3 fragTCoordNeg = (WCTCMatrix0 * vec4(worldPosNeg, 1.0)).xyz;",
			"  if (!any(greaterThan(fragTCoordNeg, vec3(1.0))) && !any(lessThan(fragTCoordNeg, vec3(0.0))))",
			"  {",
			"    vec4 newVal = rawSampleTexture(worldPosNeg);",
			"    tvalue = compositeValue(tvalue, newVal, trapezoid);",
			"    numSlices += 1;",
			"  }",
			"  vec3 worldPosPos = vertexWCVSOutput.xyz + fnumSlices * normalxspacing * vboScaling;",
			"  vec3 fragTCoordPos = (WCTCMatrix0 * vec4(worldPosPos, 1.0)).xyz;",
			"  if (!any(greaterThan(fragTCoordPos, vec3(1.0))) && !any(lessThan(fragTCoordPos, vec3(0.0))))",
			"  {",
			"    vec4 newVal = rawSampleTexture(worldPosPos);",
			"    tvalue = compositeValue(tvalue, newVal, trapezoid);",
			"    numSlices += 1;",
			"  }",
			"}",
			"// Finally, if slab type is *mean*, divide the sum by the numSlices",
			"if (slabType == 2)",
			"{",
			"  tvalue = tvalue / float(numSlices);",
			"}"
		];
	}
	function collectLabelmapInputIndices(numberOfComponents, isLabelmap) {
		const labelmapInputs = [];
		for (let i = 0; i < numberOfComponents; i++) if (isLabelmap(i)) labelmapInputs.push(i);
		return labelmapInputs;
	}
	publicAPI.replaceShaderTCoord = (shaders, ren, actor) => {
		let VSSource = shaders.Vertex;
		const GSSource = shaders.Geometry;
		let FSSource = shaders.Fragment;
		const useLabelOutline = model.labelOutlineProperties.length > 0;
		const slabThickness = model.renderable.getSlabThickness();
		VSSource = ShaderProgram_default.substitute(VSSource, "//VTK::TCoord::Dec", []).result;
		VSSource = ShaderProgram_default.substitute(VSSource, "//VTK::TCoord::Impl", []).result;
		const tNumComp = model.numberOfComponents;
		const iComps = actor.getProperty(model.currentValidInputs[0].inputIndex).getIndependentComponents();
		const numInputs = model.scalarTextures.length;
		let tcoordFSDec = [
			`uniform highp sampler3D volumeTexture[${numInputs}];`,
			"uniform float cshift0;",
			"uniform float cscale0;",
			"uniform float pwfshift0;",
			"uniform float pwfscale0;",
			"uniform sampler2D colorTexture1;",
			"uniform sampler2D pwfTexture1;",
			"uniform float opacity;",
			"uniform vec4 backgroundColor;"
		];
		for (let i = 0; i < numInputs; i++) tcoordFSDec.push(`uniform mat4 WCTCMatrix${i};`);
		if (useLabelOutline) {
			tcoordFSDec = tcoordFSDec.concat([
				"uniform sampler2D labelOutlineThicknessTexture;",
				"uniform sampler2D labelOutlineOpacityTexture;",
				"uniform float labelOutlineTextureWidth;",
				"uniform float numLabelmaps;"
			]);
			for (let i = 0; i < numInputs; i++) {
				tcoordFSDec.push(`uniform vec3 outlineTangent1_${i};`);
				tcoordFSDec.push(`uniform vec3 outlineTangent2_${i};`);
				tcoordFSDec.push(`uniform vec3 texelSize${i};`);
			}
		}
		tcoordFSDec.push("vec4 rawSampleTexture(vec3 worldPos) {");
		if (!model.multiTexturePerVolumeEnabled) tcoordFSDec.push("vec3 tc0 = (WCTCMatrix0 * vec4(worldPos, 1.0)).xyz;", "return texture(volumeTexture[0], tc0);", "}");
		else {
			tcoordFSDec.push("vec4 rawSample;");
			for (let component = 0; component < numInputs; ++component) tcoordFSDec.push(`vec3 tc${component} = (WCTCMatrix${component} * vec4(worldPos, 1.0)).xyz;`, `rawSample[${component}] = texture(volumeTexture[${component}], tc${component})[0];`);
			tcoordFSDec.push("return rawSample;", "}");
		}
		if (iComps) {
			for (let comp = 1; comp < tNumComp; comp++) tcoordFSDec = tcoordFSDec.concat([
				`uniform float cshift${comp};`,
				`uniform float cscale${comp};`,
				`uniform float pwfshift${comp};`,
				`uniform float pwfscale${comp};`
			]);
			switch (tNumComp) {
				case 1:
					tcoordFSDec = tcoordFSDec.concat(["uniform float mix0;", "#define height0 0.5"]);
					break;
				case 2:
					tcoordFSDec = tcoordFSDec.concat([
						"uniform float mix0;",
						"uniform float mix1;",
						"#define height0 0.25",
						"#define height1 0.75"
					]);
					break;
				case 3:
					tcoordFSDec = tcoordFSDec.concat([
						"uniform float mix0;",
						"uniform float mix1;",
						"uniform float mix2;",
						"#define height0 0.17",
						"#define height1 0.5",
						"#define height2 0.83"
					]);
					break;
				case 4:
					tcoordFSDec = tcoordFSDec.concat([
						"uniform float mix0;",
						"uniform float mix1;",
						"uniform float mix2;",
						"uniform float mix3;",
						"#define height0 0.125",
						"#define height1 0.375",
						"#define height2 0.625",
						"#define height3 0.875"
					]);
					break;
				default: vtkErrorMacro("Unsupported number of independent coordinates.");
			}
		}
		if (slabThickness > 0) {
			tcoordFSDec = tcoordFSDec.concat([
				"uniform vec3 spacing;",
				"uniform float slabThickness;",
				"uniform int slabType;",
				"uniform int slabTrapezoid;",
				"uniform vec3 vboScaling;"
			]);
			tcoordFSDec = tcoordFSDec.concat(getSlabCompositeDecLines());
		}
		FSSource = ShaderProgram_default.substitute(FSSource, "//VTK::TCoord::Dec", tcoordFSDec).result;
		let tcoordFSImpl = [
			"vec3 fragWorldPos = vertexWCVSOutput.xyz;",
			"vec3 fragTexCoord = (WCTCMatrix0 * vec4(fragWorldPos, 1.0)).xyz;",
			"if (any(greaterThan(fragTexCoord, vec3(1.0))) || any(lessThan(fragTexCoord, vec3(0.0))))",
			"{",
			"  // set the background color and exit",
			"  gl_FragData[0] = backgroundColor;",
			"  return;",
			"}",
			"vec4 tvalue = rawSampleTexture(fragWorldPos);"
		];
		if (slabThickness > 0) tcoordFSImpl = tcoordFSImpl.concat(getSlabSampleLoopLines());
		if (iComps) {
			const rgba = [
				"r",
				"g",
				"b",
				"a"
			];
			for (let comp = 0; comp < tNumComp; ++comp) tcoordFSImpl = tcoordFSImpl.concat([`vec3 tcolor${comp} = texture2D(colorTexture1, vec2(tvalue.${rgba[comp]} * cscale${comp} + cshift${comp}, height${comp})).rgb;`, `float compWeight${comp} = mix${comp} * texture2D(pwfTexture1, vec2(tvalue.${rgba[comp]} * pwfscale${comp} + pwfshift${comp}, height${comp})).r;`]);
			const labelmapInputs = useLabelOutline ? collectLabelmapInputIndices(tNumComp, (i) => actor.getProperty(model.currentValidInputs[i].inputIndex)?.getUseLabelOutline()) : [];
			const generateWeightedSumShader = (inputCount) => {
				if (inputCount === 1) return ["gl_FragData[0] = vec4(tcolor0.rgb, compWeight0 * opacity);"];
				const inputs = Array.from({ length: inputCount }, (_, i) => i);
				const weightSum = inputs.map((i) => `compWeight${i}`).join(" + ");
				const colorSum = inputs.map((i) => `(tcolor${i}.rgb * (compWeight${i} / weightSum))`).join(" + ");
				return [`float weightSum = ${weightSum};`, `gl_FragData[0] = vec4(vec3(${colorSum}), opacity);`];
			};
			if (labelmapInputs.length > 0) tcoordFSImpl = tcoordFSImpl.concat(generateMultiInputCompositeShader(labelmapInputs, tNumComp));
			else tcoordFSImpl = tcoordFSImpl.concat(generateWeightedSumShader(tNumComp));
		} else switch (tNumComp) {
			case 1:
				if (useLabelOutline) tcoordFSImpl = tcoordFSImpl.concat([...splitStringOnEnter(`
                // Label outline mode for single component
                float centerValue = tvalue.r;
                int segmentIndex = int(centerValue * 255.0);

                // Skip background (segment 0)
                if (segmentIndex == 0) {
                  gl_FragData[0] = vec4(0.0, 0.0, 0.0, 0.0);
                  return;
                }

                // Get outline parameters for this segment (row 0 for single labelmap)
                float textureCoordinate = float(segmentIndex - 1) / labelOutlineTextureWidth;
                float labelmapRow = 0.5 / numLabelmaps;
                float thicknessValue = texture2D(labelOutlineThicknessTexture, vec2(textureCoordinate, labelmapRow)).r;
                float outlineOpacity = texture2D(labelOutlineOpacityTexture, vec2(textureCoordinate, labelmapRow)).r;
                int actualThickness = int(thicknessValue * 255.0);

                // Get color for this segment
                vec3 tColor = texture2D(colorTexture1, vec2(centerValue * cscale0 + cshift0, 0.5)).rgb;
                float scalarOpacity = texture2D(pwfTexture1, vec2(centerValue * pwfscale0 + pwfshift0, 0.5)).r;
                float opacityToUse = scalarOpacity * opacity;

                // Check neighbors for border detection
                bool pixelOnBorder = false;
                for (int i = -actualThickness; i <= actualThickness; i++) {
                  for (int j = -actualThickness; j <= actualThickness; j++) {
                    if (i == 0 && j == 0) {
                      continue;
                    }
                    // Sample neighbor using tangent vectors in texture space
                    vec3 neighborTexCoord = fragTexCoord + float(i) * outlineTangent1_0 * texelSize0 + float(j) * outlineTangent2_0 * texelSize0;

                    // Skip if outside texture bounds
                    if (any(greaterThan(neighborTexCoord, vec3(1.0))) || any(lessThan(neighborTexCoord, vec3(0.0)))) {
                      pixelOnBorder = true;
                      break;
                    }

                    float neighborValue = texture(volumeTexture[0], neighborTexCoord).r;
                    if (neighborValue != centerValue) {
                      pixelOnBorder = true;
                      break;
                    }
                  }
                  if (pixelOnBorder) {
                    break;
                  }
                }

                if (pixelOnBorder) {
                  gl_FragData[0] = vec4(tColor, outlineOpacity);
                } else {
                  gl_FragData[0] = vec4(tColor, opacityToUse);
                }
              `)]);
				else tcoordFSImpl = tcoordFSImpl.concat([
					"// Dependent components",
					"float intensity = tvalue.r;",
					"vec3 tcolor = texture2D(colorTexture1, vec2(intensity * cscale0 + cshift0, 0.5)).rgb;",
					"float scalarOpacity = texture2D(pwfTexture1, vec2(intensity * pwfscale0 + pwfshift0, 0.5)).r;",
					"gl_FragData[0] = vec4(tcolor, scalarOpacity * opacity);"
				]);
				break;
			case 2:
				tcoordFSImpl = tcoordFSImpl.concat(["float intensity = tvalue.r*cscale0 + cshift0;", "gl_FragData[0] = vec4(texture2D(colorTexture1, vec2(intensity, 0.5)).rgb, pwfscale0*tvalue.g + pwfshift0);"]);
				break;
			case 3:
				tcoordFSImpl = tcoordFSImpl.concat([
					"vec4 tcolor = cscale0*tvalue + cshift0;",
					"gl_FragData[0] = vec4(texture2D(colorTexture1, vec2(tcolor.r,0.5)).r,",
					"  texture2D(colorTexture1, vec2(tcolor.g,0.5)).r,",
					"  texture2D(colorTexture1, vec2(tcolor.b,0.5)).r, opacity);"
				]);
				break;
			default: tcoordFSImpl = tcoordFSImpl.concat([
				"vec4 tcolor = cscale0*tvalue + cshift0;",
				"gl_FragData[0] = vec4(texture2D(colorTexture1, vec2(tcolor.r,0.5)).r,",
				"  texture2D(colorTexture1, vec2(tcolor.g,0.5)).r,",
				"  texture2D(colorTexture1, vec2(tcolor.b,0.5)).r, tcolor.a);"
			]);
		}
		FSSource = ShaderProgram_default.substitute(FSSource, "//VTK::TCoord::Impl", tcoordFSImpl).result;
		shaders.Vertex = VSSource;
		shaders.Fragment = FSSource;
		shaders.Geometry = GSSource;
	};
	publicAPI.replaceShaderPositionVC = (shaders, ren, actor) => {
		let VSSource = shaders.Vertex;
		const GSSource = shaders.Geometry;
		let FSSource = shaders.Fragment;
		const slabThickness = model.renderable.getSlabThickness();
		let posVCVSDec = ["attribute vec4 vertexWC;", "varying vec4 vertexWCVSOutput;"];
		posVCVSDec = posVCVSDec.concat([`//${publicAPI.getMTime()}${model.resliceGeomUpdateString}`]);
		if (slabThickness > 0) posVCVSDec = posVCVSDec.concat(["attribute vec3 normalWC;", "varying vec3 normalWCVSOutput;"]);
		VSSource = ShaderProgram_default.substitute(VSSource, "//VTK::PositionVC::Dec", posVCVSDec).result;
		let posVCVSImpl = ["gl_Position = MCPCMatrix * vertexWC;", "vertexWCVSOutput = vertexWC;"];
		if (slabThickness > 0) posVCVSImpl = posVCVSImpl.concat(["normalWCVSOutput = normalWC;"]);
		VSSource = ShaderProgram_default.substitute(VSSource, "//VTK::PositionVC::Impl", posVCVSImpl).result;
		VSSource = ShaderProgram_default.substitute(VSSource, "//VTK::Camera::Dec", ["uniform mat4 MCPCMatrix;", "uniform mat4 MCVCMatrix;"]).result;
		let posVCFSDec = ["varying vec4 vertexWCVSOutput;"];
		if (slabThickness > 0) posVCFSDec = posVCFSDec.concat(["varying vec3 normalWCVSOutput;"]);
		FSSource = ShaderProgram_default.substitute(FSSource, "//VTK::PositionVC::Dec", posVCFSDec).result;
		shaders.Vertex = VSSource;
		shaders.Geometry = GSSource;
		shaders.Fragment = FSSource;
	};
	publicAPI.updateResliceGeometry = () => {
		const firstImageData = model.currentValidInputs[0].imageData;
		const { resGeomString, slicePD, slicePlane, orthoSlicing, orthoAxis } = computeResliceGeometryState(model.renderable, firstImageData);
		if (!model.resliceGeom || model.resliceGeomUpdateString !== resGeomString) {
			if (slicePD) {
				if (!model.resliceGeom) model.resliceGeom = PolyData_default.newInstance();
				model.resliceGeom.getPoints().setData(slicePD.getPoints().getData(), 3);
				model.resliceGeom.getPolys().setData(slicePD.getPolys().getData(), 1);
				model.resliceGeom.getPointData().setNormals(slicePD.getPointData().getNormals());
			} else if (slicePlane) {
				if (!model.resliceGeom) model.resliceGeom = PolyData_default.newInstance();
				if (!orthoSlicing) {
					const { points, polys, normalsData } = computeObliqueSliceGeometryData(firstImageData, slicePlane, model.outlineFilter, model.cutter, model.lineToSurfaceFilter);
					model.resliceGeom.getPoints().setData(points, 3);
					model.resliceGeom.getPolys().setData(polys, 1);
					const normals = DataArray_default.newInstance({
						numberOfComponents: 3,
						values: normalsData,
						name: "Normals"
					});
					model.resliceGeom.getPointData().setNormals(normals);
				} else {
					const { points, polys, normalsData } = computeOrthoSliceGeometryData(firstImageData, slicePlane, orthoAxis, model.transform);
					model.resliceGeom.getPoints().setData(points, 3);
					model.resliceGeom.getPolys().setData(polys, 1);
					const normals = DataArray_default.newInstance({
						numberOfComponents: 3,
						values: normalsData,
						name: "Normals"
					});
					model.resliceGeom.getPointData().setNormals(normals);
				}
			} else vtkErrorMacro("Something went wrong.", "A default slice plane should have been created in the beginning of", "updateResliceGeometry.");
			model.resliceGeomUpdateString = resGeomString;
			model.resliceGeom?.modified();
		}
	};
	function buildLabelOutline2DTexture(dataArrays, ArrayType, vtkDataType) {
		const { width: maxDataWidth, height } = getLabelOutlineTextureParameters(dataArrays.map((dataArray) => ({ property: dataArray })), (property) => property);
		let width = model.renderable.getLabelOutlineTextureWidth();
		if (width <= 0) width = Math.max(maxDataWidth, model.context.getParameter(model.context.MAX_TEXTURE_SIZE));
		const table = new ArrayType(width * height);
		fillLabelOutlineTextureTable(table, dataArrays, width);
		const newTexture = Texture_default.newInstance({ resizable: false });
		newTexture.setOpenGLRenderWindow(model._openGLRenderWindow);
		newTexture.resetFormatAndType();
		newTexture.setMinificationFilter(Filter.NEAREST);
		newTexture.setMagnificationFilter(Filter.NEAREST);
		newTexture.create2DFromRaw({
			width,
			height,
			numComps: 1,
			dataType: vtkDataType,
			data: table
		});
		return newTexture;
	}
	function updateLabelOutlineTexture(dataArrays, ArrayType, vtkDataType, hashKey, textureKey) {
		const { hash } = getLabelOutlineTextureParameters(dataArrays.map((dataArray) => ({ property: dataArray })), (property) => property);
		if (hash === model[hashKey]) return;
		model[hashKey] = hash;
		if (model[textureKey]) model[textureKey].releaseGraphicsResources();
		model[textureKey] = buildLabelOutline2DTexture(dataArrays, ArrayType, vtkDataType);
	}
	publicAPI.updateLabelOutlineThicknessTexture = (labelOutlineProperties) => {
		updateLabelOutlineTexture(labelOutlineProperties.map(({ property }) => property.getLabelOutlineThicknessByReference()), Uint8Array, VtkDataTypes.UNSIGNED_CHAR, "_labelOutlineThicknessHash", "labelOutlineThicknessTexture");
	};
	publicAPI.updateLabelOutlineOpacityTexture = (labelOutlineProperties) => {
		updateLabelOutlineTexture(labelOutlineProperties.map(({ property }) => {
			let dataArray = property.getLabelOutlineOpacity();
			if (typeof dataArray === "number") dataArray = [dataArray];
			return dataArray;
		}), Float32Array, VtkDataTypes.FLOAT, "_labelOutlineOpacityHash", "labelOutlineOpacityTexture");
	};
	publicAPI.setScalarTextures = (scalarTextures) => {
		model.scalarTextures = [...scalarTextures];
		model._externalOpenGLTexture = true;
	};
	publicAPI.delete = chain(() => {
		if (model._openGLRenderWindow) unregisterGraphicsResources(model._openGLRenderWindow);
		if (model.labelOutlineThicknessTexture) {
			model.labelOutlineThicknessTexture.releaseGraphicsResources();
			model.labelOutlineThicknessTexture = null;
		}
		if (model.labelOutlineOpacityTexture) {
			model.labelOutlineOpacityTexture.releaseGraphicsResources();
			model.labelOutlineOpacityTexture = null;
		}
	}, publicAPI.delete);
}
var DEFAULT_VALUES = {
	VBOBuildTime: {},
	VBOBuildString: null,
	haveSeenDepthRequest: false,
	lastHaveSeenDepthRequest: false,
	lastIndependentComponents: false,
	lastUseLabelOutline: false,
	lastNumValidInputs: 0,
	lastNumberOfComponents: 0,
	lastMultiTexturePerVolumeEnabled: false,
	lastSlabThickness: 0,
	lastSlabTrapezoidIntegration: 0,
	lastSlabType: -1,
	scalarTextures: [],
	_scalarTexturesCore: [],
	colorTexture: null,
	_colorTextureCore: null,
	pwfTexture: null,
	_pwfTextureCore: null,
	labelOutlineProperties: [],
	labelOutlineThicknessTexture: null,
	_labelOutlineThicknessHash: null,
	labelOutlineOpacityTexture: null,
	_labelOutlineOpacityHash: null,
	_externalOpenGLTexture: false,
	resliceGeom: null,
	resliceGeomUpdateString: null,
	tris: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	ViewNode_default.extend(publicAPI, model, initialValues);
	ReplacementShaderMapper_default.implementReplaceShaderCoincidentOffset(publicAPI, model, initialValues);
	ReplacementShaderMapper_default.implementBuildShadersWithReplacements(publicAPI, model, initialValues);
	model.tris = Helper_default.newInstance();
	model.scalarTextures = [];
	model.colorTexture = null;
	model.pwfTexture = null;
	model.VBOBuildTime = {};
	obj(model.VBOBuildTime);
	model.tmpMat4 = mat4.identity(new Float64Array(16));
	model._tmpTexelSize = [
		0,
		0,
		0
	];
	model.outlineFilter = ImageDataOutlineFilter_default.newInstance();
	model.outlineFilter.setGenerateFaces(true);
	model.outlineFilter.setGenerateLines(false);
	model.cubePolyData = PolyData_default.newInstance();
	model.cutter = Cutter_default.newInstance();
	model.lineToSurfaceFilter = ClosedPolyLineToSurfaceFilter_default.newInstance();
	model.transform = Transform_default.newInstance();
	get(publicAPI, model, ["scalarTextures"]);
	vtkOpenGLImageResliceMapper(publicAPI, model);
}
var newInstance = newInstance$1(extend, "vtkOpenGLImageResliceMapper");
var ImageResliceMapper_default = {
	newInstance,
	extend
};
registerOverride("vtkImageResliceMapper", newInstance);
//#endregion
export { ImageResliceMapper_default as default, extend, newInstance };
