// @kitware/vtk.js/Rendering/OpenGL/OrderIndependentTranslucentPass.js
import macros_default from "../../macros.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import Property_default from "../Core/Property.js";
import Texture_default from "./Texture.js";
import Framebuffer_default from "./Framebuffer.js";
import RenderPass_default from "../SceneGraph/RenderPass.js";
import ShaderProgram_default from "./ShaderProgram.js";
import VertexArrayObject_default from "./VertexArrayObject.js";
import Helper_default from "./Helper.js";
//#region Sources/Rendering/OpenGL/OrderIndependentTranslucentPass/index.js
var { Representation } = Property_default;
var { vtkErrorMacro } = macros_default;
function translucentShaderReplacement(shaders) {
	shaders.Fragment = ShaderProgram_default.substitute(shaders.Fragment, "//VTK::RenderPassFragmentShader::Impl", `
      float weight = gl_FragData[0].a * pow(max(1.1 - gl_FragCoord.z, 0.0), 2.0);
      gl_FragData[0] = vec4(gl_FragData[0].rgb*weight, gl_FragData[0].a);
      gl_FragData[1].r = weight;
    `, false).result;
}
var oitpFragTemplate = `//VTK::System::Dec

in vec2 tcoord;

uniform sampler2D translucentRTexture;
uniform sampler2D translucentRGBATexture;

// the output of this shader
//VTK::Output::Dec

void main()
{
  vec4 t1Color = texture(translucentRGBATexture, tcoord);
  float t2Color = texture(translucentRTexture, tcoord).r;
  gl_FragData[0] = vec4(t1Color.rgb/max(t2Color,0.01), 1.0 - t1Color.a);
}
`;
function vtkOpenGLOrderIndependentTranslucentPass(publicAPI, model) {
	model.classHierarchy.push("vtkOpenGLOrderIndependentTranslucentPass");
	publicAPI.createVertexBuffer = () => {
		const ptsArray = new Float32Array([
			-1,
			-1,
			-1,
			1,
			-1,
			-1,
			-1,
			1,
			-1,
			1,
			1,
			-1
		]);
		const tcoordArray = new Float32Array([
			0,
			0,
			1,
			0,
			0,
			1,
			1,
			1
		]);
		const cellArray = new Uint16Array([
			4,
			0,
			1,
			3,
			2
		]);
		const points = DataArray_default.newInstance({
			numberOfComponents: 3,
			values: ptsArray
		});
		points.setName("points");
		const tcoords = DataArray_default.newInstance({
			numberOfComponents: 2,
			values: tcoordArray
		});
		tcoords.setName("tcoords");
		const cells = DataArray_default.newInstance({
			numberOfComponents: 1,
			values: cellArray
		});
		model.tris.getCABO().createVBO(cells, "polys", Representation.SURFACE, {
			points,
			tcoords,
			cellOffset: 0
		});
		model.VBOBuildTime.modified();
	};
	publicAPI.createFramebuffer = (viewNode) => {
		const size = viewNode.getSize();
		const gl = viewNode.getContext();
		model.framebuffer = Framebuffer_default.newInstance();
		model.framebuffer.setOpenGLRenderWindow(viewNode);
		model.framebuffer.create(...size);
		model.framebuffer.saveCurrentBindingsAndBuffers();
		model.framebuffer.bind();
		model.translucentRGBATexture = Texture_default.newInstance();
		model.translucentRGBATexture.setInternalFormat(gl.RGBA16F);
		model.translucentRGBATexture.setFormat(gl.RGBA);
		model.translucentRGBATexture.setOpenGLDataType(gl.HALF_FLOAT);
		model.translucentRGBATexture.setOpenGLRenderWindow(viewNode);
		model.translucentRGBATexture.create2DFromRaw({
			width: size[0],
			height: size[1],
			numComps: 4,
			dataType: "Float32Array",
			data: null
		});
		model.translucentRTexture = Texture_default.newInstance();
		model.translucentRTexture.setInternalFormat(gl.R16F);
		model.translucentRTexture.setFormat(gl.RED);
		model.translucentRTexture.setOpenGLDataType(gl.HALF_FLOAT);
		model.translucentRTexture.setOpenGLRenderWindow(viewNode);
		model.translucentRTexture.create2DFromRaw({
			width: size[0],
			height: size[1],
			numComps: 1,
			dataType: "Float32Array",
			data: null
		});
		model.translucentZTexture = Texture_default.newInstance();
		model.translucentZTexture.setOpenGLRenderWindow(viewNode);
		model.translucentZTexture.createDepthFromRaw({
			width: size[0],
			height: size[1],
			dataType: "Float32Array",
			data: null
		});
		model.framebuffer.setColorBuffer(model.translucentRGBATexture, 0);
		model.framebuffer.setColorBuffer(model.translucentRTexture, 1);
		model.framebuffer.setDepthBuffer(model.translucentZTexture);
	};
	publicAPI.createCopyShader = (viewNode) => {
		model.copyShader = viewNode.getShaderCache().readyShaderProgramArray([
			"//VTK::System::Dec",
			"attribute vec4 vertexDC;",
			"attribute vec2 tcoordTC;",
			"varying vec2 tcoord;",
			"void main() { tcoord = tcoordTC; gl_Position = vertexDC; }"
		].join("\n"), oitpFragTemplate, "");
	};
	publicAPI.createVBO = (viewNode) => {
		const gl = viewNode.getContext();
		model.tris.setOpenGLRenderWindow(viewNode);
		publicAPI.createVertexBuffer();
		const program = model.copyShader;
		model.tris.getCABO().bind();
		if (!model.copyVAO.addAttributeArray(program, model.tris.getCABO(), "vertexDC", model.tris.getCABO().getVertexOffset(), model.tris.getCABO().getStride(), gl.FLOAT, 3, gl.FALSE)) vtkErrorMacro("Error setting vertexDC in copy shader VAO.");
		if (!model.copyVAO.addAttributeArray(program, model.tris.getCABO(), "tcoordTC", model.tris.getCABO().getTCoordOffset(), model.tris.getCABO().getStride(), gl.FLOAT, 2, gl.FALSE)) vtkErrorMacro("Error setting vertexDC in copy shader VAO.");
	};
	publicAPI.traverse = (viewNode, renNode, forwardPass) => {
		if (model.deleted) return;
		const size = viewNode.getSize();
		const gl = viewNode.getContext();
		model._supported = false;
		if (renNode.getSelector() || !gl || !viewNode.getWebgl2() || !gl.getExtension("EXT_color_buffer_half_float") && !gl.getExtension("EXT_color_buffer_float")) {
			publicAPI.setCurrentOperation("translucentPass");
			renNode.traverse(publicAPI);
			return;
		}
		model._supported = true;
		if (model.framebuffer === null) publicAPI.createFramebuffer(viewNode);
		else {
			const fbSize = model.framebuffer.getSize();
			if (fbSize === null || fbSize[0] !== size[0] || fbSize[1] !== size[1]) {
				model.framebuffer.releaseGraphicsResources();
				model.translucentRGBATexture.releaseGraphicsResources(viewNode);
				model.translucentRTexture.releaseGraphicsResources(viewNode);
				model.translucentZTexture.releaseGraphicsResources(viewNode);
				publicAPI.createFramebuffer(viewNode);
			} else {
				model.framebuffer.saveCurrentBindingsAndBuffers();
				model.framebuffer.bind();
			}
		}
		gl.drawBuffers([gl.COLOR_ATTACHMENT0]);
		gl.clearBufferfv(gl.COLOR, 0, [
			0,
			0,
			0,
			0
		]);
		gl.clearBufferfv(gl.DEPTH, 0, [1]);
		gl.colorMask(false, false, false, false);
		if (forwardPass.getOpaqueActorCount() > 0) {
			forwardPass.setCurrentOperation("opaqueZBufferPass");
			renNode.traverse(forwardPass);
		}
		gl.colorMask(true, true, true, true);
		gl.drawBuffers([gl.COLOR_ATTACHMENT0, gl.COLOR_ATTACHMENT1]);
		gl.viewport(0, 0, size[0], size[1]);
		gl.scissor(0, 0, size[0], size[1]);
		gl.clearBufferfv(gl.COLOR, 0, [
			0,
			0,
			0,
			1
		]);
		gl.clearBufferfv(gl.COLOR, 1, [
			0,
			0,
			0,
			0
		]);
		gl.enable(gl.DEPTH_TEST);
		gl.enable(gl.BLEND);
		gl.blendFuncSeparate(gl.ONE, gl.ONE, gl.ZERO, gl.ONE_MINUS_SRC_ALPHA);
		publicAPI.setCurrentOperation("translucentPass");
		renNode.traverse(publicAPI);
		gl.drawBuffers([gl.NONE]);
		model.framebuffer.restorePreviousBindingsAndBuffers();
		if (model.copyShader === null) publicAPI.createCopyShader(viewNode);
		else viewNode.getShaderCache().readyShaderProgram(model.copyShader);
		if (!model.copyVAO) {
			model.copyVAO = VertexArrayObject_default.newInstance();
			model.copyVAO.setOpenGLRenderWindow(viewNode);
		}
		model.copyVAO.bind();
		if (model.VBOBuildTime.getMTime() < publicAPI.getMTime()) publicAPI.createVBO(viewNode);
		const cullFaceEnabled = gl.isEnabled(gl.CULL_FACE);
		const cullFaceMode = cullFaceEnabled ? gl.getParameter(gl.CULL_FACE_MODE) : null;
		viewNode.disableCullFace();
		gl.blendFuncSeparate(gl.SRC_ALPHA, gl.ONE_MINUS_SRC_ALPHA, gl.ONE, gl.ONE_MINUS_SRC_ALPHA);
		gl.depthMask(false);
		gl.depthFunc(gl.ALWAYS);
		gl.viewport(0, 0, size[0], size[1]);
		gl.scissor(0, 0, size[0], size[1]);
		model.translucentRGBATexture.activate();
		model.copyShader.setUniformi("translucentRGBATexture", model.translucentRGBATexture.getTextureUnit());
		model.translucentRTexture.activate();
		model.copyShader.setUniformi("translucentRTexture", model.translucentRTexture.getTextureUnit());
		gl.drawArrays(gl.TRIANGLES, 0, model.tris.getCABO().getElementCount());
		gl.depthMask(true);
		gl.depthFunc(gl.LEQUAL);
		if (cullFaceEnabled) {
			viewNode.enableCullFace();
			viewNode.setCullFaceMode(cullFaceMode);
		}
		model.translucentRGBATexture.deactivate();
		model.translucentRTexture.deactivate();
		const ts = renNode.getTiledSizeAndOrigin();
		gl.scissor(ts.lowerLeftU, ts.lowerLeftV, ts.usize, ts.vsize);
		gl.viewport(ts.lowerLeftU, ts.lowerLeftV, ts.usize, ts.vsize);
	};
	publicAPI.getShaderReplacement = () => {
		if (model._supported) return translucentShaderReplacement;
		return null;
	};
	publicAPI.releaseGraphicsResources = (viewNode) => {
		if (model.framebuffer) {
			model.framebuffer.releaseGraphicsResources(viewNode);
			model.framebuffer = null;
		}
		if (model.translucentRGBATexture) {
			model.translucentRGBATexture.releaseGraphicsResources(viewNode);
			model.translucentRGBATexture = null;
		}
		if (model.translucentRTexture) {
			model.translucentRTexture.releaseGraphicsResources(viewNode);
			model.translucentRTexture = null;
		}
		if (model.translucentZTexture) {
			model.translucentZTexture.releaseGraphicsResources(viewNode);
			model.translucentZTexture = null;
		}
		if (model.copyVAO) {
			model.copyVAO.releaseGraphicsResources(viewNode);
			model.copyVAO = null;
		}
		if (model.copyShader) {
			model.copyShader.releaseGraphicsResources(viewNode);
			model.copyShader = null;
		}
		if (model.tris) {
			model.tris.releaseGraphicsResources(viewNode);
			model.tris = null;
		}
		publicAPI.modified();
	};
}
var DEFAULT_VALUES = {
	framebuffer: null,
	copyShader: null,
	tris: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	RenderPass_default.extend(publicAPI, model, initialValues);
	model.VBOBuildTime = {};
	macros_default.obj(model.VBOBuildTime, { mtime: 0 });
	model.tris = Helper_default.newInstance();
	macros_default.get(publicAPI, model, ["framebuffer"]);
	vtkOpenGLOrderIndependentTranslucentPass(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkOpenGLOrderIndependentTranslucentPass");
var OrderIndependentTranslucentPass_default = {
	newInstance,
	extend
};
//#endregion
export { OrderIndependentTranslucentPass_default as default, extend, newInstance };
