// @kitware/vtk.js/Rendering/OpenGL/Shader.js
import macros_default from "../../macros.js";
//#region Sources/Rendering/OpenGL/Shader/index.js
var { vtkErrorMacro } = macros_default;
function vtkShader(publicAPI, model) {
	model.classHierarchy.push("vtkShader");
	publicAPI.compile = () => {
		let stype = model.context.VERTEX_SHADER;
		if (!model.source || !model.source.length || model.shaderType === "Unknown") return false;
		if (model.handle !== 0) {
			model.context.deleteShader(model.handle);
			model.handle = 0;
		}
		switch (model.shaderType) {
			case "Fragment":
				stype = model.context.FRAGMENT_SHADER;
				break;
			default:
				stype = model.context.VERTEX_SHADER;
				break;
		}
		model.handle = model.context.createShader(stype);
		model.context.shaderSource(model.handle, model.source);
		model.context.compileShader(model.handle);
		if (!model.context.getShaderParameter(model.handle, model.context.COMPILE_STATUS)) {
			const lastError = model.context.getShaderInfoLog(model.handle);
			vtkErrorMacro(`Error compiling shader '${model.source}': ${lastError}`);
			model.context.deleteShader(model.handle);
			model.handle = 0;
			return false;
		}
		return true;
	};
	publicAPI.cleanup = () => {
		if (model.shaderType === "Unknown" || model.handle === 0) return;
		model.context.deleteShader(model.handle);
		model.handle = 0;
		model.dirty = true;
	};
}
var DEFAULT_VALUES = {
	shaderType: "Unknown",
	source: "",
	error: "",
	handle: 0,
	dirty: false,
	context: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, [
		"shaderType",
		"source",
		"error",
		"handle",
		"context"
	]);
	vtkShader(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkShader");
var Shader_default = {
	newInstance,
	extend
};
//#endregion
export { Shader_default as default, extend, newInstance };
