// @kitware/vtk.js/Rendering/OpenGL/TextureUnitManager.js
import macros_default from "../../macros.js";
//#region Sources/Rendering/OpenGL/TextureUnitManager/index.js
var { vtkErrorMacro } = macros_default;
function vtkOpenGLTextureUnitManager(publicAPI, model) {
	model.classHierarchy.push("vtkOpenGLTextureUnitManager");
	publicAPI.deleteTable = () => {
		for (let i = 0; i < model.numberOfTextureUnits; ++i) if (model.textureUnits[i] === true) vtkErrorMacro("some texture units  were not properly released");
		model.textureUnits = [];
		model.numberOfTextureUnits = 0;
	};
	publicAPI.setContext = (ctx) => {
		if (model.context !== ctx) {
			if (model.context !== 0) publicAPI.deleteTable();
			model.context = ctx;
			if (model.context) {
				model.numberOfTextureUnits = ctx.getParameter(ctx.MAX_TEXTURE_IMAGE_UNITS);
				for (let i = 0; i < model.numberOfTextureUnits; ++i) model.textureUnits[i] = false;
			}
			publicAPI.modified();
		}
	};
	publicAPI.allocate = () => {
		for (let i = 0; i < model.numberOfTextureUnits; i++) if (!publicAPI.isAllocated(i)) {
			model.textureUnits[i] = true;
			return i;
		}
		return -1;
	};
	publicAPI.allocateUnit = (unit) => {
		if (publicAPI.isAllocated(unit)) return -1;
		model.textureUnits[unit] = true;
		return unit;
	};
	publicAPI.isAllocated = (textureUnitId) => model.textureUnits[textureUnitId];
	publicAPI.free = (val) => {
		model.textureUnits[val] = false;
	};
	publicAPI.freeAll = () => {
		for (let i = 0; i < model.numberOfTextureUnits; ++i) model.textureUnits[i] = false;
	};
}
var DEFAULT_VALUES = {
	context: null,
	numberOfTextureUnits: 0,
	textureUnits: 0
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	model.textureUnits = [];
	macros_default.get(publicAPI, model, ["numberOfTextureUnits"]);
	macros_default.setGet(publicAPI, model, ["context"]);
	vtkOpenGLTextureUnitManager(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkOpenGLTextureUnitManager");
var TextureUnitManager_default = {
	newInstance,
	extend
};
//#endregion
export { TextureUnitManager_default as default, extend, newInstance };
