// @kitware/vtk.js/Interaction/Style/InteractorStyleManipulator/Presets.js
import Manipulators_default from "../../Manipulators.js";
//#region Sources/Interaction/Style/InteractorStyleManipulator/Presets.js
var MANIPULTOR_TYPES = {
	slice: Manipulators_default.vtkMouseCameraSliceManipulator,
	multiRotate: Manipulators_default.vtkMouseCameraTrackballMultiRotateManipulator,
	pan: Manipulators_default.vtkMouseCameraTrackballPanManipulator,
	roll: Manipulators_default.vtkMouseCameraTrackballRollManipulator,
	rotate: Manipulators_default.vtkMouseCameraTrackballRotateManipulator,
	axisRotate: Manipulators_default.vtkMouseCameraAxisRotateManipulator,
	zoom: Manipulators_default.vtkMouseCameraTrackballZoomManipulator,
	zoomToMouse: Manipulators_default.vtkMouseCameraTrackballZoomToMouseManipulator,
	range: Manipulators_default.vtkMouseRangeManipulator,
	vrPan: Manipulators_default.vtkVRButtonPanManipulator,
	gestureCamera: Manipulators_default.vtkGestureCameraManipulator,
	movement: Manipulators_default.vtkKeyboardCameraManipulator,
	freeLook: Manipulators_default.vtkMouseCameraTrackballFirstPersonManipulator,
	unicam: Manipulators_default.vtkMouseCameraUnicamManipulator,
	unicamRotate: Manipulators_default.vtkMouseCameraUnicamRotateManipulator
};
var STYLES = {
	"3D": [
		{ type: "rotate" },
		{
			type: "pan",
			options: { shift: true }
		},
		{
			type: "zoom",
			options: { control: true }
		},
		{
			type: "zoom",
			options: { alt: true }
		},
		{
			type: "zoom",
			options: {
				dragEnabled: false,
				scrollEnabled: true
			}
		},
		{
			type: "zoom",
			options: { button: 3 }
		},
		{
			type: "roll",
			options: {
				shift: true,
				control: true
			}
		},
		{
			type: "roll",
			options: {
				shift: true,
				alt: true
			}
		},
		{
			type: "roll",
			options: {
				shift: true,
				button: 3
			}
		},
		{ type: "vrPan" },
		{ type: "gestureCamera" }
	],
	"2D": [
		{
			type: "pan",
			options: { shift: true }
		},
		{
			type: "zoom",
			options: { control: true }
		},
		{
			type: "zoom",
			options: { alt: true }
		},
		{
			type: "zoom",
			options: { button: 3 }
		},
		{
			type: "roll",
			options: {
				shift: true,
				alt: true
			}
		},
		{
			type: "roll",
			options: {
				shift: true,
				button: 3
			}
		},
		{
			type: "roll",
			options: { shift: true }
		},
		{ type: "vrPan" },
		{ type: "gestureCamera" }
	],
	FirstPerson: [{ type: "movement" }, { type: "freeLook" }],
	Unicam: [{ type: "unicam" }],
	zRotateTop: [
		{
			type: "pan",
			options: { shift: true }
		},
		{
			type: "axisRotate",
			options: {
				rotationAxis: [
					0,
					0,
					1
				],
				useHalfAxis: true
			}
		},
		{
			type: "zoom",
			options: { control: true }
		},
		{
			type: "zoom",
			options: { alt: true }
		},
		{
			type: "zoom",
			options: {
				dragEnabled: false,
				scrollEnabled: true
			}
		},
		{
			type: "zoom",
			options: { button: 3 }
		}
	],
	zRotateAll: [
		{
			type: "pan",
			options: { shift: true }
		},
		{
			type: "axisRotate",
			options: {
				rotationAxis: [
					0,
					0,
					1
				],
				useHalfAxis: false
			}
		},
		{
			type: "zoom",
			options: { control: true }
		},
		{
			type: "zoom",
			options: { alt: true }
		},
		{
			type: "zoom",
			options: {
				dragEnabled: false,
				scrollEnabled: true
			}
		},
		{
			type: "zoom",
			options: { button: 3 }
		}
	]
};
function applyDefinitions(definitions, manipulatorStyle) {
	manipulatorStyle.removeAllManipulators();
	for (let idx = 0; idx < definitions.length; idx++) {
		const definition = definitions[idx];
		const instance = MANIPULTOR_TYPES[definition.type].newInstance(definition.options);
		if (instance.isA("vtkCompositeVRManipulator")) manipulatorStyle.addVRManipulator(instance);
		else if (instance.isA("vtkCompositeGestureManipulator")) manipulatorStyle.addGestureManipulator(instance);
		else if (instance.isA("vtkCompositeKeyboardManipulator")) manipulatorStyle.addKeyboardManipulator(instance);
		else manipulatorStyle.addMouseManipulator(instance);
	}
	return true;
}
function applyPreset(name, manipulatorStyle) {
	return applyDefinitions(STYLES[name], manipulatorStyle);
}
function registerManipulatorType(type, classDef) {
	MANIPULTOR_TYPES[type] = classDef;
}
function registerStylePreset(name, definitions) {
	STYLES[name] = definitions;
}
var Presets_default = {
	applyDefinitions,
	applyPreset,
	registerManipulatorType,
	registerStylePreset
};
//#endregion
export { Presets_default as default };
