// @kitware/vtk.js/Common/Core/CellArray.js
import macros_default from "../../macros.js";
import { VtkDataTypes } from "./DataArray/Constants.js";
import DataArray_default from "./DataArray.js";
//#region Sources/Common/Core/CellArray/index.js
var { isVtkObject } = macros_default;
function extractCellSizes(cellArray) {
	const cellSizes = [];
	for (let i = 0; i < cellArray.length; i += cellArray[i] + 1) cellSizes.push(cellArray[i]);
	return cellSizes;
}
function getNumberOfCells(cellArray) {
	let cellId = 0;
	for (let cellArrayIndex = 0; cellArrayIndex < cellArray.length;) {
		cellArrayIndex += cellArray[cellArrayIndex] + 1;
		cellId++;
	}
	return cellId;
}
var STATIC = {
	extractCellSizes,
	getNumberOfCells
};
function vtkCellArray(publicAPI, model) {
	model.classHierarchy.push("vtkCellArray");
	const superClass = { ...publicAPI };
	publicAPI.getNumberOfCells = (recompute) => {
		if (model.numberOfCells !== void 0 && !recompute) return model.numberOfCells;
		if (model.cellSizes) model.numberOfCells = model.cellSizes.length;
		else model.numberOfCells = getNumberOfCells(publicAPI.getData());
		return model.numberOfCells;
	};
	publicAPI.getCellSizes = (recompute) => {
		if (model.cellSizes !== void 0 && !recompute) return model.cellSizes;
		model.cellSizes = extractCellSizes(publicAPI.getData());
		return model.cellSizes;
	};
	/**
	* When `resize()` is being used, you then MUST use `insertNextCell()`.
	* @see vtkCellArray#insertNextCell
	* @see vtkDataArray#allocate
	*/
	publicAPI.resize = (requestedNumTuples) => {
		const oldNumTuples = publicAPI.getNumberOfTuples();
		superClass.resize(requestedNumTuples);
		const newNumTuples = publicAPI.getNumberOfTuples();
		if (newNumTuples < oldNumTuples) if (newNumTuples === 0) {
			model.numberOfCells = 0;
			model.cellSizes = [];
		} else {
			model.numberOfCells = void 0;
			model.cellSizes = void 0;
		}
	};
	publicAPI.setData = (typedArray) => {
		superClass.setData(typedArray, 1);
		model.numberOfCells = void 0;
		model.cellSizes = void 0;
	};
	publicAPI.getCell = (loc) => {
		let cellLoc = loc;
		const numberOfPoints = model.values[cellLoc++];
		return model.values.subarray(cellLoc, cellLoc + numberOfPoints);
	};
	publicAPI.insertNextCell = (cell) => {
		let cellPointIds;
		if (isVtkObject(cell)) cellPointIds = cell.getPointsIds();
		else cellPointIds = cell;
		const cellId = publicAPI.getNumberOfCells();
		publicAPI.insertNextTuples([cellPointIds.length, ...cellPointIds]);
		++model.numberOfCells;
		if (model.cellSizes != null) model.cellSizes.push(cellPointIds.length);
		return cellId;
	};
	publicAPI.getMaxCellSize = () => publicAPI.getCellSizes().reduce((a, b) => Math.max(a, b), 0);
}
function defaultValues(initialValues) {
	return {
		empty: true,
		numberOfComponents: 1,
		dataType: VtkDataTypes.UNSIGNED_INT,
		...initialValues
	};
}
function extend(publicAPI, model, initialValues = {}) {
	DataArray_default.extend(publicAPI, model, defaultValues(initialValues));
	vtkCellArray(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkCellArray");
var CellArray_default = {
	newInstance,
	extend,
	...STATIC
};
//#endregion
export { STATIC, CellArray_default as default, extend, newInstance };
