// @kitware/vtk.js/Common/Core/Points.js
import macros_default from "../../macros.js";
import { VtkDataTypes } from "./DataArray/Constants.js";
import { uninitializeBounds } from "./Math.js";
import DataArray_default from "./DataArray.js";
//#region Sources/Common/Core/Points/index.js
var { vtkErrorMacro } = macros_default;
function vtkPoints(publicAPI, model) {
	let boundMTime = 0;
	model.classHierarchy.push("vtkPoints");
	publicAPI.getNumberOfPoints = publicAPI.getNumberOfTuples;
	publicAPI.setNumberOfPoints = (nbPoints, dimension = 3) => {
		if (publicAPI.getNumberOfPoints() !== nbPoints) {
			model.size = nbPoints * dimension;
			model.values = macros_default.newTypedArray(model.dataType, model.size);
			publicAPI.setNumberOfComponents(dimension);
			publicAPI.modified();
		}
	};
	publicAPI.setPoint = (idx, ...xyz) => {
		publicAPI.setTuple(idx, xyz);
	};
	publicAPI.getPoint = publicAPI.getTuple;
	publicAPI.findPoint = publicAPI.findTuple;
	publicAPI.insertNextPoint = (x, y, z) => publicAPI.insertNextTuple([
		x,
		y,
		z
	]);
	publicAPI.insertPoint = (ptId, point) => publicAPI.insertTuple(ptId, point);
	const superGetBounds = publicAPI.getBounds;
	publicAPI.getBounds = () => {
		if (boundMTime < model.mtime) publicAPI.computeBounds();
		return superGetBounds();
	};
	const superGetBoundsByReference = publicAPI.getBoundsByReference;
	publicAPI.getBoundsByReference = () => {
		if (boundMTime < model.mtime) publicAPI.computeBounds();
		return superGetBoundsByReference();
	};
	publicAPI.computeBounds = () => {
		if (publicAPI.getNumberOfComponents() === 3) {
			const xRange = publicAPI.getRange(0);
			model.bounds[0] = xRange[0];
			model.bounds[1] = xRange[1];
			const yRange = publicAPI.getRange(1);
			model.bounds[2] = yRange[0];
			model.bounds[3] = yRange[1];
			const zRange = publicAPI.getRange(2);
			model.bounds[4] = zRange[0];
			model.bounds[5] = zRange[1];
		} else if (publicAPI.getNumberOfComponents() === 2) {
			const xRange = publicAPI.getRange(0);
			model.bounds[0] = xRange[0];
			model.bounds[1] = xRange[1];
			const yRange = publicAPI.getRange(1);
			model.bounds[2] = yRange[0];
			model.bounds[3] = yRange[1];
			model.bounds[4] = 0;
			model.bounds[5] = 0;
		} else {
			vtkErrorMacro(`getBounds called on an array with components of ${publicAPI.getNumberOfComponents()}`);
			uninitializeBounds(model.bounds);
		}
		boundMTime = macros_default.getCurrentGlobalMTime();
	};
	publicAPI.setNumberOfComponents(model.numberOfComponents < 2 ? 3 : model.numberOfComponents);
}
var DEFAULT_VALUES = {
	empty: true,
	numberOfComponents: 3,
	dataType: VtkDataTypes.FLOAT,
	bounds: [
		1,
		-1,
		1,
		-1,
		1,
		-1
	]
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	DataArray_default.extend(publicAPI, model, initialValues);
	macros_default.getArray(publicAPI, model, ["bounds"], 6);
	vtkPoints(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkPoints");
var Points_default = {
	newInstance,
	extend
};
//#endregion
export { Points_default as default, extend, newInstance };
