// @kitware/vtk.js/Common/Core/VariantArray.js
import macros_default from "../../macros.js";
//#region Sources/Common/Core/VariantArray/index.js
function vtkVariantArray(publicAPI, model) {
	model.classHierarchy.push("vtkVariantArray");
	publicAPI.getComponent = (tupleIdx, compIdx = 0) => model.values[tupleIdx * model.numberOfComponents + compIdx];
	publicAPI.setComponent = (tupleIdx, compIdx, value) => {
		if (value !== model.values[tupleIdx * model.numberOfComponents + compIdx]) {
			model.values[tupleIdx * model.numberOfComponents + compIdx] = value;
			publicAPI.modified();
		}
	};
	publicAPI.getData = () => model.values;
	publicAPI.getTuple = (idx, tupleToFill = []) => {
		const numberOfComponents = model.numberOfComponents || 1;
		if (tupleToFill.length) tupleToFill.length = numberOfComponents;
		const offset = idx * numberOfComponents;
		for (let i = 0; i < numberOfComponents; i++) tupleToFill[i] = model.values[offset + i];
		return tupleToFill;
	};
	publicAPI.getTupleLocation = (idx = 1) => idx * model.numberOfComponents;
	publicAPI.getNumberOfComponents = () => model.numberOfComponents;
	publicAPI.getNumberOfValues = () => model.values.length;
	publicAPI.getNumberOfTuples = () => model.values.length / model.numberOfComponents;
	publicAPI.getDataType = () => model.dataType;
	publicAPI.newClone = () => newInstance({
		name: model.name,
		numberOfComponents: model.numberOfComponents
	});
	publicAPI.getName = () => {
		if (!model.name) {
			publicAPI.modified();
			model.name = `vtkVariantArray${publicAPI.getMTime()}`;
		}
		return model.name;
	};
	publicAPI.setData = (array, numberOfComponents) => {
		model.values = array;
		model.size = array.length;
		if (numberOfComponents) model.numberOfComponents = numberOfComponents;
		if (model.size % model.numberOfComponents !== 0) model.numberOfComponents = 1;
		publicAPI.modified();
	};
}
var DEFAULT_VALUES = {
	name: "",
	numberOfComponents: 1,
	size: 0,
	dataType: "JSON"
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	if (!model.empty && !model.values && !model.size) throw new TypeError("Cannot create vtkVariantArray object without: size > 0, values");
	if (!model.values) model.values = [];
	else if (Array.isArray(model.values)) model.values = [...model.values];
	if (model.values) model.size = model.values.length;
	macros_default.obj(publicAPI, model);
	macros_default.set(publicAPI, model, ["name"]);
	vtkVariantArray(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkVariantArray");
var VariantArray_default = {
	newInstance,
	extend
};
//#endregion
export { VariantArray_default as default, extend, newInstance };
