// @kitware/vtk.js/Proxy/Representations/GeometryRepresentationProxy.js
import macros_default from "../../macros.js";
import Actor_default from "../../Rendering/Core/Actor.js";
import Mapper_default from "../../Rendering/Core/Mapper.js";
import AbstractRepresentationProxy_default from "../Core/AbstractRepresentationProxy.js";
//#region Sources/Proxy/Representations/GeometryRepresentationProxy/index.js
var PROPERTIES_STATE = { representation: {
	"Surface with edges": { property: {
		edgeVisibility: true,
		representation: 2
	} },
	Surface: { property: {
		edgeVisibility: false,
		representation: 2
	} },
	Wireframe: { property: {
		edgeVisibility: false,
		representation: 1
	} },
	Points: { property: {
		edgeVisibility: false,
		representation: 0
	} }
} };
var PROPERTIES_DEFAULT = { representation: "Surface" };
function vtkGeometryRepresentationProxy(publicAPI, model) {
	model.classHierarchy.push("vtkGeometryRepresentationProxy");
	model.mapper = Mapper_default.newInstance({
		interpolateScalarsBeforeMapping: true,
		useLookupTableScalarRange: true,
		scalarVisibility: false
	});
	model.actor = Actor_default.newInstance();
	model.property = model.actor.getProperty();
	model.sourceDependencies.push(model.mapper);
	model.actor.setMapper(model.mapper);
	model.actors.push(model.actor);
}
var DEFAULT_VALUES = { representation: "Surface" };
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	AbstractRepresentationProxy_default.extend(publicAPI, model, initialValues);
	vtkGeometryRepresentationProxy(publicAPI, model);
	macros_default.proxyPropertyState(publicAPI, model, PROPERTIES_STATE, PROPERTIES_DEFAULT);
	macros_default.proxyPropertyMapping(publicAPI, model, {
		opacity: {
			modelKey: "property",
			property: "opacity"
		},
		visibility: {
			modelKey: "actor",
			property: "visibility"
		},
		color: {
			modelKey: "property",
			property: "diffuseColor"
		},
		interpolateScalarsBeforeMapping: {
			modelKey: "mapper",
			property: "interpolateScalarsBeforeMapping"
		},
		pointSize: {
			modelKey: "property",
			property: "pointSize"
		},
		useShadow: {
			modelKey: "property",
			property: "lighting"
		},
		lineWidth: {
			modelKey: "property",
			property: "lineWidth"
		},
		useBounds: {
			modelKey: "actor",
			property: "useBounds"
		}
	});
}
var newInstance = macros_default.newInstance(extend, "vtkGeometryRepresentationProxy");
var GeometryRepresentationProxy_default = {
	newInstance,
	extend
};
//#endregion
export { GeometryRepresentationProxy_default as default, extend, newInstance };
