// @kitware/vtk.js/Proxy/Representations/GlyphRepresentationProxy.js
import vtk from "../../vtk.js";
import macros_default from "../../macros.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
import "../../Filters/Sources.js";
import Property_default from "../../Rendering/Core/Property.js";
import Actor_default from "../../Rendering/Core/Actor.js";
import ColorTransferFunction_default from "../../Rendering/Core/ColorTransferFunction.js";
import Glyph3DMapper_default from "../../Rendering/Core/Glyph3DMapper.js";
import AbstractRepresentationProxy_default from "../Core/AbstractRepresentationProxy.js";
//#region Sources/Proxy/Representations/GlyphRepresentationProxy/index.js
function vtkGlyphRepresentationProxy(publicAPI, model) {
	model.classHierarchy.push("vtkGlyphRepresentationProxy");
	model.property = Property_default.newInstance();
	function processJSON(description) {
		model.actors.length = 0;
		const lookupTable = ColorTransferFunction_default.newInstance();
		lookupTable.applyColorMap({ RGBPoints: description.rgbPoints });
		model.glyph = {};
		let count = description.glyph.length;
		while (count--) {
			const glyph = description.glyph[count];
			model.glyph[glyph.id] = vtk(glyph);
		}
		count = description.mapping.length;
		while (count--) {
			const sourceDesc = description.mapping[count];
			const glyph = model.glyph[sourceDesc.glyphId];
			const source = PolyData_default.newInstance();
			source.getPoints().setData(Float32Array.from(sourceDesc.coordinates), 3);
			if (sourceDesc.scale) source.getPointData().addArray(DataArray_default.newInstance({
				name: "scaling",
				values: Float32Array.from(sourceDesc.scale),
				numberOfComponents: 3
			}));
			const mapper = Glyph3DMapper_default.newInstance({
				useLookupTableScalarRange: true,
				lookupTable,
				orient: false,
				scaling: !!sourceDesc.scale,
				scaleArray: "scaling",
				scaleMode: Glyph3DMapper_default.ScaleModes.SCALE_BY_COMPONENTS
			});
			const actor = Actor_default.newInstance();
			if (model.property) actor.setProperty(model.property);
			actor.setMapper(mapper);
			mapper.setInputData(source, 0);
			mapper.setInputConnection(glyph.getOutputPort(), 1);
			model.actors.push(actor);
		}
	}
	model.sourceDependencies.push({ setInputData: processJSON });
	publicAPI.setColorBy = () => {};
	publicAPI.getColorBy = () => [];
	publicAPI.listDataArrays = () => [];
}
var DEFAULT_VALUES = {};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	AbstractRepresentationProxy_default.extend(publicAPI, model, initialValues);
	vtkGlyphRepresentationProxy(publicAPI, model);
	macros_default.proxyPropertyMapping(publicAPI, model, { edgeVisibility: {
		modelKey: "property",
		property: "edgeVisibility"
	} });
}
var newInstance = macros_default.newInstance(extend, "vtkGlyphRepresentationProxy");
var GlyphRepresentationProxy_default = {
	newInstance,
	extend
};
//#endregion
export { GlyphRepresentationProxy_default as default, extend, newInstance };
