// @kitware/vtk.js/Proxy/Representations/MoleculeRepresentationProxy.js
import macros_default from "../../macros.js";
import MoleculeToRepresentation_default from "../../Filters/General/MoleculeToRepresentation.js";
import Actor_default from "../../Rendering/Core/Actor.js";
import SphereMapper_default from "../../Rendering/Core/SphereMapper.js";
import StickMapper_default from "../../Rendering/Core/StickMapper.js";
import AbstractRepresentationProxy_default from "../Core/AbstractRepresentationProxy.js";
//#region Sources/Proxy/Representations/MoleculeRepresentationProxy/index.js
function vtkMoleculeRepresentationProxy(publicAPI, model) {
	model.classHierarchy.push("vtkMoleculeRepresentationProxy");
	model.filter = MoleculeToRepresentation_default.newInstance();
	model.sphereMapper = SphereMapper_default.newInstance();
	model.stickMapper = StickMapper_default.newInstance();
	model.sphereActor = Actor_default.newInstance();
	model.stickActor = Actor_default.newInstance();
	model.sourceDependencies.push(model.filter);
	model.sphereMapper.setInputConnection(model.filter.getOutputPort(0));
	model.sphereMapper.setScaleArray(model.filter.getSphereScaleArrayName());
	model.sphereActor.setMapper(model.sphereMapper);
	model.stickMapper.setInputConnection(model.filter.getOutputPort(1));
	model.stickMapper.setScaleArray("stickScales");
	model.stickMapper.setOrientationArray("orientation");
	model.stickActor.setMapper(model.stickMapper);
	model.actors.push(model.sphereActor);
	model.actors.push(model.stickActor);
	publicAPI.setColorBy = () => {};
	publicAPI.getColorBy = () => [];
	publicAPI.listDataArrays = () => [];
}
var DEFAULT_VALUES = {};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	AbstractRepresentationProxy_default.extend(publicAPI, model, initialValues);
	vtkMoleculeRepresentationProxy(publicAPI, model);
	macros_default.proxyPropertyMapping(publicAPI, model, {
		tolerance: {
			modelKey: "filter",
			property: "tolerance"
		},
		atomicRadiusScaleFactor: {
			modelKey: "filter",
			property: "atomicRadiusScaleFactor"
		},
		bondRadius: {
			modelKey: "filter",
			property: "bondRadius"
		},
		deltaBondFactor: {
			modelKey: "filter",
			property: "deltaBondFactor"
		},
		radiusType: {
			modelKey: "filter",
			property: "radiusType"
		},
		hideElements: {
			modelKey: "filter",
			property: "hideElements"
		}
	});
}
var newInstance = macros_default.newInstance(extend, "vtkMoleculeRepresentationProxy");
var MoleculeRepresentationProxy_default = {
	newInstance,
	extend
};
//#endregion
export { MoleculeRepresentationProxy_default as default, extend, newInstance };
