// @kitware/vtk.js/Proxy/Representations/ResliceRepresentationProxy.js
import macros_default from "../../macros.js";
import Plane_default from "../../Common/DataModel/Plane.js";
import Cutter_default from "../../Filters/Core/Cutter.js";
import ImageDataOutlineFilter_default from "../../Filters/General/ImageDataOutlineFilter.js";
import Actor_default from "../../Rendering/Core/Actor.js";
import Mapper_default from "../../Rendering/Core/Mapper.js";
import { SlabTypes } from "../../Rendering/Core/ImageResliceMapper/Constants.js";
import ImageResliceMapper_default from "../../Rendering/Core/ImageResliceMapper.js";
import ImageSlice_default from "../../Rendering/Core/ImageSlice.js";
import AbstractRepresentationProxy_default from "../Core/AbstractRepresentationProxy.js";
//#region Sources/Proxy/Representations/ResliceRepresentationProxy/index.js
function sum(a, b) {
	return a + b;
}
function mean(...array) {
	return array.reduce(sum, 0) / array.length;
}
function updateDomains(dataset, dataArray, model, updateProp) {
	const dataRange = dataArray.getRange();
	const propToUpdate = {
		windowWidth: { domain: {
			min: 0,
			max: dataRange[1] - dataRange[0],
			step: "any"
		} },
		windowLevel: { domain: {
			min: dataRange[0],
			max: dataRange[1],
			step: "any"
		} }
	};
	updateProp("windowWidth", propToUpdate.windowWidth);
	updateProp("windowLevel", propToUpdate.windowLevel);
	return {
		windowWidth: propToUpdate.windowWidth.domain.max,
		windowLevel: Math.floor(mean(propToUpdate.windowLevel.domain.min, propToUpdate.windowLevel.domain.max))
	};
}
function vtkResliceRepresentationProxy(publicAPI, model) {
	model.classHierarchy.push("vtkResliceRepresentationProxy");
	model.mapper = ImageResliceMapper_default.newInstance();
	model.actor = ImageSlice_default.newInstance();
	model.property = model.actor.getProperty();
	model.slicePlane = Plane_default.newInstance();
	model.slicePlane.setNormal(0, 1, 0);
	model.mapper.setSlicePlane(model.slicePlane);
	model.mapper.setSlabType(SlabTypes.MAX);
	model.outline = {
		iof: ImageDataOutlineFilter_default.newInstance(),
		sof: Cutter_default.newInstance({ cutFunction: model.slicePlane }),
		mapper: Mapper_default.newInstance(),
		actor: Actor_default.newInstance({ visibility: false })
	};
	model.outline.sof.setInputConnection(model.outline.iof.getOutputPort());
	model.outline.mapper.setInputConnection(model.outline.sof.getOutputPort());
	model.actor.setMapper(model.mapper);
	model.actors.push(model.actor);
	model.outline.actor.setMapper(model.outline.mapper);
	model.actors.push(model.outline.actor);
	function setInputData(inputDataset) {
		const state = updateDomains(inputDataset, publicAPI.getDataArray(), model, publicAPI.updateProxyProperty);
		publicAPI.set(state);
		model.slicePlane.setOrigin(inputDataset.getCenter());
	}
	model.sourceDependencies.push(model.mapper);
	model.sourceDependencies.push({ setInputData });
	model.sourceDependencies.push(model.outline.iof);
	const parentSetColorBy = publicAPI.setColorBy;
	publicAPI.setColorBy = (arrayName, arrayLocation, componentIndex = -1) => {
		if (arrayName === null) {
			model.property.setRGBTransferFunction(null);
			model.property.setPiecewiseFunction(null);
		} else {
			parentSetColorBy(arrayName, arrayLocation, componentIndex);
			const lutProxy = publicAPI.getLookupTableProxy(arrayName);
			const pwfProxy = publicAPI.getPiecewiseFunctionProxy(arrayName);
			model.property.setRGBTransferFunction(lutProxy.getLookupTable());
			model.property.setPiecewiseFunction(pwfProxy.getPiecewiseFunction());
		}
	};
}
var DEFAULT_VALUES = {};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	AbstractRepresentationProxy_default.extend(publicAPI, model, initialValues);
	macros_default.get(publicAPI, model, ["slicePlane"]);
	vtkResliceRepresentationProxy(publicAPI, model);
	model.outlineActor = model.outline.actor;
	model.outlineProperty = model.outline.actor.getProperty();
	macros_default.proxyPropertyMapping(publicAPI, model, {
		visibility: {
			modelKey: "actor",
			property: "visibility"
		},
		windowWidth: {
			modelKey: "property",
			property: "colorWindow"
		},
		windowLevel: {
			modelKey: "property",
			property: "colorLevel"
		},
		interpolationType: {
			modelKey: "property",
			property: "interpolationType"
		},
		slicePlane: {
			modelKey: "mapper",
			property: "slicePlane"
		},
		slicePolyData: {
			modelKey: "mapper",
			property: "slicePolyData"
		},
		slabType: {
			modelKey: "mapper",
			property: "slabType"
		},
		slabThickness: {
			modelKey: "mapper",
			property: "slabThickness"
		},
		slabTrapezoidIntegration: {
			modelKey: "mapper",
			property: "slabTrapezoidIntegration"
		},
		outlineVisibility: {
			modelKey: "outlineActor",
			property: "visibility"
		},
		outlineColor: {
			modelKey: "outlineProperty",
			property: "color"
		},
		outlineLineWidth: {
			modelKey: "outlineProperty",
			property: "lineWidth"
		}
	});
}
var newInstance = macros_default.newInstance(extend, "vtkResliceRepresentationProxy");
var ResliceRepresentationProxy_default = {
	newInstance,
	extend
};
//#endregion
export { ResliceRepresentationProxy_default as default, extend, newInstance };
