// @kitware/vtk.js/Proxy/Representations/SlicedGeometryRepresentationProxy.js
import macros_default from "../../macros.js";
import Plane_default from "../../Common/DataModel/Plane.js";
import Cutter_default from "../../Filters/Core/Cutter.js";
import Actor_default from "../../Rendering/Core/Actor.js";
import Mapper_default from "../../Rendering/Core/Mapper.js";
import ImageMapper_default from "../../Rendering/Core/ImageMapper.js";
import AbstractRepresentationProxy_default from "../Core/AbstractRepresentationProxy.js";
//#region Sources/Proxy/Representations/SlicedGeometryRepresentationProxy/index.js
function vtkSlicedGeometryRepresentationProxy(publicAPI, model) {
	model.classHierarchy.push("vtkSlicedGeometryRepresentationProxy");
	model.plane = Plane_default.newInstance();
	model.cutter = Cutter_default.newInstance();
	model.cutter.setCutFunction(model.plane);
	model.mapper = Mapper_default.newInstance();
	model.actor = Actor_default.newInstance();
	model.property = model.actor.getProperty();
	model.property.setLighting(false);
	model.mapper.setInputConnection(model.cutter.getOutputPort());
	model.actor.setMapper(model.mapper);
	model.actors.push(model.actor);
	model.sourceDependencies.push(model.cutter);
	publicAPI.setSlice = (slice = 0) => {
		const stateModified = model.slice !== slice;
		model.slice = slice;
		const n = model.plane.getNormal();
		if (model.plane.setOrigin(n[0] * slice, n[1] * slice, n[2] * slice) || stateModified) {
			publicAPI.modified();
			return true;
		}
		return false;
	};
	publicAPI.setOffset = (offset = 0) => {
		const stateModified = model.offset !== offset;
		model.offset = offset;
		const normal = model.plane.getNormal();
		if (model.actor.setPosition(offset * normal[0], offset * normal[1], offset * normal[2]) || stateModified) {
			publicAPI.modified();
			return true;
		}
		return false;
	};
	publicAPI.setSlicingMode = (mode) => {
		if (model.slicingMode === mode || !mode) {
			console.log("skip setSlicingMode", mode);
			return;
		}
		model.slicingMode = mode;
		switch (ImageMapper_default.SlicingMode[mode]) {
			case ImageMapper_default.SlicingMode.X:
				model.plane.setNormal(1, 0, 0);
				break;
			case ImageMapper_default.SlicingMode.Y:
				model.plane.setNormal(0, 1, 0);
				break;
			case ImageMapper_default.SlicingMode.Z:
				model.plane.setNormal(0, 0, 1);
				break;
			default: return;
		}
		let alreadyModified = publicAPI.setSlice(model.slice);
		alreadyModified = publicAPI.setOffset(model.offset) || alreadyModified;
		if (!alreadyModified) publicAPI.modified();
	};
}
var DEFAULT_VALUES = {
	slicingMode: ImageMapper_default.SlicingMode.NONE,
	slice: 0,
	offset: 0
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	AbstractRepresentationProxy_default.extend(publicAPI, model, initialValues);
	macros_default.get(publicAPI, model, [
		"slicingMode",
		"slice",
		"offset"
	]);
	vtkSlicedGeometryRepresentationProxy(publicAPI, model);
	macros_default.proxyPropertyState(publicAPI, model);
	macros_default.proxyPropertyMapping(publicAPI, model, {
		opacity: {
			modelKey: "property",
			property: "opacity"
		},
		visibility: {
			modelKey: "actor",
			property: "visibility"
		},
		color: {
			modelKey: "property",
			property: "diffuseColor"
		},
		useShadow: {
			modelKey: "property",
			property: "lighting"
		},
		useBounds: {
			modelKey: "actor",
			property: "useBounds"
		}
	});
}
var newInstance = macros_default.newInstance(extend, "vtkSlicedGeometryRepresentationProxy");
var SlicedGeometryRepresentationProxy_default = {
	newInstance,
	extend
};
//#endregion
export { SlicedGeometryRepresentationProxy_default as default, extend, newInstance };
