// @kitware/vtk.js/Rendering/SceneGraph/RenderPass.js
import macros_default from "../../macros.js";
//#region Sources/Rendering/SceneGraph/RenderPass/index.js
function vtkRenderPass(publicAPI, model) {
	model.classHierarchy.push("vtkRenderPass");
	publicAPI.getOperation = () => model.currentOperation;
	publicAPI.setCurrentOperation = (val) => {
		model.currentOperation = val;
		model.currentTraverseOperation = `traverse${macros_default.capitalize(model.currentOperation)}`;
	};
	publicAPI.getTraverseOperation = () => model.currentTraverseOperation;
	publicAPI.traverse = (viewNode, parent = null) => {
		if (model.deleted) return;
		model._currentParent = parent;
		model.preDelegateOperations.forEach((val) => {
			publicAPI.setCurrentOperation(val);
			viewNode.traverse(publicAPI);
		});
		model.delegates.forEach((val) => {
			val.traverse(viewNode, publicAPI);
		});
		model.postDelegateOperations.forEach((val) => {
			publicAPI.setCurrentOperation(val);
			viewNode.traverse(publicAPI);
		});
	};
}
var DEFAULT_VALUES = {
	delegates: [],
	currentOperation: null,
	preDelegateOperations: [],
	postDelegateOperations: [],
	currentParent: null
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.get(publicAPI, model, ["currentOperation"]);
	macros_default.setGet(publicAPI, model, [
		"delegates",
		"_currentParent",
		"preDelegateOperations",
		"postDelegateOperations"
	]);
	macros_default.moveToProtected(publicAPI, model, ["currentParent"]);
	vtkRenderPass(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkRenderPass");
var RenderPass_default = {
	newInstance,
	extend
};
//#endregion
export { RenderPass_default as default, extend, newInstance };
