// @kitware/vtk.js/Rendering/SceneGraph/ViewNode.js
import macros_default from "../../macros.js";
//#region Sources/Rendering/SceneGraph/ViewNode/index.js
var { vtkErrorMacro } = macros_default;
var PASS_TYPES = ["Build", "Render"];
function vtkViewNode(publicAPI, model) {
	model.classHierarchy.push("vtkViewNode");
	/**
	* Convenient method to move a child to a specific index.
	* @param {*} child the child to move
	* @param {*} toIndex the index to move the child to
	* @returns true if the child was moved, false otherwise
	*/
	function moveChild(child, toIndex) {
		for (let i = 0; i < model.children.length; ++i) {
			const childIndex = (toIndex + 1 + i) % model.children.length;
			if (model.children[childIndex] === child) {
				model.children[childIndex] = model.children[toIndex];
				model.children[toIndex] = child;
				return true;
			}
		}
		return false;
	}
	publicAPI.build = (prepass) => {};
	publicAPI.render = (prepass) => {};
	publicAPI.traverse = (renderPass) => {
		const fn = publicAPI[renderPass.getTraverseOperation()];
		if (fn) {
			fn(renderPass);
			return;
		}
		publicAPI.apply(renderPass, true);
		for (let index = 0; index < model.children.length; index++) model.children[index].traverse(renderPass);
		publicAPI.apply(renderPass, false);
	};
	publicAPI.apply = (renderPass, prepass) => {
		const customRenderPass = publicAPI[renderPass.getOperation()];
		if (customRenderPass) customRenderPass(prepass, renderPass);
	};
	publicAPI.getViewNodeFor = (dataObject, hint = null) => {
		if (hint && hint.renderable === dataObject) return hint;
		if (model.renderable === dataObject) return publicAPI;
		for (let index = 0; index < model.children.length; ++index) {
			const vn = model.children[index].getViewNodeFor(dataObject);
			if (vn) return vn;
		}
	};
	publicAPI.getFirstAncestorOfType = (type) => {
		if (!model._parent) return null;
		if (model._parent.isA(type)) return model._parent;
		return model._parent.getFirstAncestorOfType(type);
	};
	publicAPI.getLastAncestorOfType = (type) => {
		if (!model._parent) return null;
		const lastAncestor = model._parent.getLastAncestorOfType(type);
		if (lastAncestor) return lastAncestor;
		if (model._parent.isA(type)) return model._parent;
		return null;
	};
	publicAPI.addMissingNode = (dobj) => {
		if (!dobj) return;
		const result = model._renderableChildMap.get(dobj);
		if (result !== void 0) {
			result.setVisited(true);
			return result;
		}
		const newNode = publicAPI.createViewNode(dobj);
		if (newNode) {
			newNode.setParent(publicAPI);
			newNode.setVisited(true);
			model._renderableChildMap.set(dobj, newNode);
			model.children.push(newNode);
			return newNode;
		}
	};
	publicAPI.addMissingNodes = (dataObjs, enforceOrder = false) => {
		if (!dataObjs || !dataObjs.length) return;
		let nextIndex;
		for (let index = 0; index < dataObjs.length; ++index) {
			const dobj = dataObjs[index];
			const node = publicAPI.addMissingNode(dobj);
			if (enforceOrder && node !== void 0) {
				if (nextIndex === void 0) nextIndex = model.children.lastIndexOf(node);
				else if (model.children[nextIndex] !== node) moveChild(model.children, node, nextIndex);
				nextIndex++;
			}
		}
	};
	publicAPI.addMissingChildren = (children) => {
		if (!children || !children.length) return;
		for (let index = 0; index < children.length; ++index) {
			const child = children[index];
			if (model.children.indexOf(child) === -1) {
				child.setParent(publicAPI);
				model.children.push(child);
				const childRenderable = child.getRenderable();
				if (childRenderable) model._renderableChildMap.set(childRenderable, child);
			}
			child.setVisited(true);
		}
	};
	publicAPI.removeNode = (child) => {
		const childIdx = model.children.findIndex((x) => x === child);
		if (childIdx < 0) return false;
		const renderable = child.getRenderable();
		if (renderable) model._renderableChildMap.delete(renderable);
		child.delete();
		model.children.splice(childIdx, 1);
		return true;
	};
	publicAPI.prepareNodes = () => {
		for (let index = 0; index < model.children.length; ++index) model.children[index].setVisited(false);
	};
	publicAPI.setVisited = (val) => {
		model.visited = val;
	};
	publicAPI.removeUnusedNodes = () => {
		let visitedCount = 0;
		for (let index = 0; index < model.children.length; ++index) {
			const child = model.children[index];
			if (child.getVisited()) {
				model.children[visitedCount++] = child;
				child.setVisited(false);
			} else {
				const renderable = child.getRenderable();
				if (renderable) model._renderableChildMap.delete(renderable);
				child.delete();
			}
		}
		model.children.length = visitedCount;
	};
	publicAPI.createViewNode = (dataObj) => {
		if (!model.myFactory) {
			vtkErrorMacro("Cannot create view nodes without my own factory");
			return null;
		}
		const ret = model.myFactory.createNode(dataObj);
		if (ret) ret.setRenderable(dataObj);
		return ret;
	};
	const parentDelete = publicAPI.delete;
	publicAPI.delete = () => {
		for (let i = 0; i < model.children.length; i++) model.children[i].delete();
		parentDelete();
	};
}
var DEFAULT_VALUES = {
	renderable: null,
	myFactory: null,
	children: [],
	visited: false
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.event(publicAPI, model, "event");
	model._renderableChildMap = /* @__PURE__ */ new Map();
	macros_default.get(publicAPI, model, ["visited"]);
	macros_default.setGet(publicAPI, model, [
		"_parent",
		"renderable",
		"myFactory"
	]);
	macros_default.getArray(publicAPI, model, ["children"]);
	macros_default.moveToProtected(publicAPI, model, ["parent"]);
	vtkViewNode(publicAPI, model);
}
var ViewNode_default = {
	newInstance: macros_default.newInstance(extend, "vtkViewNode"),
	extend,
	PASS_TYPES
};
//#endregion
export { ViewNode_default as default };
