// @kitware/vtk.js/Rendering/SceneGraph/ViewNodeFactory.js
import macros_default from "../../macros.js";
//#region Sources/Rendering/SceneGraph/ViewNodeFactory/index.js
var { vtkWarningMacro } = macros_default;
function listClassHierarchy(dataObject) {
	const classNames = [];
	let depth = 0;
	let className = dataObject.getClassName(depth++);
	while (className) {
		classNames.push(className);
		className = dataObject.getClassName(depth++);
	}
	return classNames;
}
function buildMissingImplementationMessage(factoryName, classNames) {
	const classList = classNames.join(" → ");
	return [
		`No ${factoryName} implementation found for ${classNames[0]}.`,
		`Class hierarchy: ${classList}.`,
		`A rendering Profile is likely missing for ${classNames[0]}.`,
		"Try importing '@kitware/vtk.js/Rendering/Profiles/All' or 'vtk.js/Sources/Rendering/Profiles/All',",
		"or import the specific rendering profile needed by this renderable if known.",
		"See https://kitware.github.io/vtk-js/docs/concepts_profile.html for details."
	].join("\n");
}
function isWidgetContainer(classNames) {
	return classNames.includes("vtkAbstractWidget") || classNames.includes("vtkWidgetRepresentation");
}
function vtkViewNodeFactory(publicAPI, model) {
	if (!model.overrides) model.overrides = {};
	model.classHierarchy.push("vtkViewNodeFactory");
	publicAPI.createNode = (dataObject) => {
		if (dataObject.isDeleted()) return null;
		const classNames = listClassHierarchy(dataObject);
		let cpt = 0;
		let className = classNames[cpt++];
		let isObject = false;
		const keys = Object.keys(model.overrides);
		while (className && !isObject) if (keys.indexOf(className) !== -1) isObject = true;
		else className = dataObject.getClassName(cpt++);
		if (!isObject) {
			if (isWidgetContainer(classNames)) return null;
			vtkWarningMacro(buildMissingImplementationMessage(publicAPI.getClassName(), classNames));
			return null;
		}
		const vn = model.overrides[className]();
		vn.setMyFactory(publicAPI);
		return vn;
	};
}
var DEFAULT_VALUES = {};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	vtkViewNodeFactory(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkViewNodeFactory");
var ViewNodeFactory_default = {
	newInstance,
	extend
};
//#endregion
export { ViewNodeFactory_default as default, extend, newInstance };
