// @kitware/vtk.js/Widgets/Representations.js
import WidgetRepresentation_default from "./Representations/WidgetRepresentation.js";
import HandleRepresentation_default from "./Representations/HandleRepresentation.js";
import ContextRepresentation_default from "./Representations/ContextRepresentation.js";
import GlyphRepresentation_default from "./Representations/GlyphRepresentation.js";
import CircleContextRepresentation_default from "./Representations/CircleContextRepresentation.js";
import ConvexFaceContextRepresentation_default from "./Representations/ConvexFaceContextRepresentation.js";
import CubeHandleRepresentation_default from "./Representations/CubeHandleRepresentation.js";
import ImplicitPlaneRepresentation_default from "./Representations/ImplicitPlaneRepresentation.js";
import LineHandleRepresentation_default from "./Representations/LineHandleRepresentation.js";
import OutlineContextRepresentation_default from "./Representations/OutlineContextRepresentation.js";
import PolyLineRepresentation_default from "./Representations/PolyLineRepresentation.js";
import RotateTransformHandleRepresentation_default from "./Representations/RotateTransformHandleRepresentation.js";
import ScaleTransformHandleRepresentation_default from "./Representations/ScaleTransformHandleRepresentation.js";
import SphereHandleRepresentation_default from "./Representations/SphereHandleRepresentation.js";
import SplineContextRepresentation_default from "./Representations/SplineContextRepresentation.js";
import TranslateTransformHandleRepresentation_default from "./Representations/TranslateTransformHandleRepresentation.js";
//#region Sources/Widgets/Representations/index.js
var Representations_default = {
	vtkCircleContextRepresentation: CircleContextRepresentation_default,
	vtkContextRepresentation: ContextRepresentation_default,
	vtkConvexFaceContextRepresentation: ConvexFaceContextRepresentation_default,
	vtkCubeHandleRepresentation: CubeHandleRepresentation_default,
	vtkGlyphRepresentation: GlyphRepresentation_default,
	vtkHandleRepresentation: HandleRepresentation_default,
	vtkImplicitPlaneRepresentation: ImplicitPlaneRepresentation_default,
	vtkLineHandleRepresentation: LineHandleRepresentation_default,
	vtkOutlineContextRepresentation: OutlineContextRepresentation_default,
	vtkPolyLineRepresentation: PolyLineRepresentation_default,
	vtkRotateTransformHandleRepresentation: RotateTransformHandleRepresentation_default,
	vtkScaleTransformHandleRepresentation: ScaleTransformHandleRepresentation_default,
	vtkSphereHandleRepresentation: SphereHandleRepresentation_default,
	vtkSplineContextRepresentation: SplineContextRepresentation_default,
	vtkTranslateTransformHandleRepresentation: TranslateTransformHandleRepresentation_default,
	vtkWidgetRepresentation: WidgetRepresentation_default
};
//#endregion
export { Representations_default as default };
