// @kitware/vtk.js/Widgets/Widgets3D.js
import AngleWidget_default from "./Widgets3D/AngleWidget.js";
import ShapeWidget_default from "./Widgets3D/ShapeWidget.js";
import EllipseWidget_default from "./Widgets3D/EllipseWidget.js";
import ImageCroppingWidget_default from "./Widgets3D/ImageCroppingWidget.js";
import ImplicitPlaneWidget_default from "./Widgets3D/ImplicitPlaneWidget.js";
import InteractiveOrientationWidget_default from "./Widgets3D/InteractiveOrientationWidget.js";
import LabelWidget_default from "./Widgets3D/LabelWidget.js";
import LineWidget_default from "./Widgets3D/LineWidget.js";
import PaintWidget_default from "./Widgets3D/PaintWidget.js";
import PolyLineWidget_default from "./Widgets3D/PolyLineWidget.js";
import RectangleWidget_default from "./Widgets3D/RectangleWidget.js";
import ResliceCursorWidget_default from "./Widgets3D/ResliceCursorWidget.js";
import SphereWidget_default from "./Widgets3D/SphereWidget.js";
import SplineWidget_default from "./Widgets3D/SplineWidget.js";
import TransformControlsWidget_default from "./Widgets3D/TransformControlsWidget.js";
//#region Sources/Widgets/Widgets3D/index.js
var Widgets3D_default = {
	vtkAngleWidget: AngleWidget_default,
	vtkEllipseWidget: EllipseWidget_default,
	vtkImageCroppingWidget: ImageCroppingWidget_default,
	vtkImplicitPlaneWidget: ImplicitPlaneWidget_default,
	vtkInteractiveOrientationWidget: InteractiveOrientationWidget_default,
	vtkLabelWidget: LabelWidget_default,
	vtkLineWidget: LineWidget_default,
	vtkPaintWidget: PaintWidget_default,
	vtkPolyLineWidget: PolyLineWidget_default,
	vtkRectangleWidget: RectangleWidget_default,
	vtkResliceCursorWidget: ResliceCursorWidget_default,
	vtkShapeWidget: ShapeWidget_default,
	vtkSphereWidget: SphereWidget_default,
	vtkSplineWidget: SplineWidget_default,
	vtkTransformControlsWidget: TransformControlsWidget_default
};
//#endregion
export { Widgets3D_default as default };
