// @kitware/vtk.js/Rendering/Misc/SynchronizableRenderWindow/ObjectManager.js
import { capitalize } from "../../../macros.js";
import DataArray_default from "../../../Common/Core/DataArray.js";
import CellArray_default from "../../../Common/Core/CellArray.js";
import ImageData_default from "../../../Common/DataModel/ImageData.js";
import LookupTable_default from "../../../Common/Core/LookupTable.js";
import Points_default from "../../../Common/Core/Points.js";
import PolyData_default from "../../../Common/DataModel/PolyData.js";
import PiecewiseFunction_default from "../../../Common/DataModel/PiecewiseFunction.js";
import Property_default from "../../Core/Property.js";
import Actor_default from "../../Core/Actor.js";
import Mapper_default from "../../Core/Mapper.js";
import Camera_default from "../../Core/Camera.js";
import Light_default from "../../Core/Light.js";
import Renderer_default from "../../Core/Renderer.js";
import ColorTransferFunction_default from "../../Core/ColorTransferFunction.js";
import VolumeProperty_default from "../../Core/VolumeProperty.js";
import Volume_default from "../../Core/Volume.js";
import Texture_default from "../../Core/Texture.js";
import VolumeMapper_default from "../../Core/VolumeMapper.js";
import AxesActor_default from "../../Core/AxesActor.js";
import CubeAxesActor_default from "../../Core/CubeAxesActor.js";
import Glyph3DMapper_default from "../../Core/Glyph3DMapper.js";
import ImageMapper_default from "../../Core/ImageMapper.js";
import ImageProperty_default from "../../Core/ImageProperty.js";
import ImageSlice_default from "../../Core/ImageSlice.js";
import RenderWindow_default from "../../Core/RenderWindow.js";
import ScalarBarActor_default from "../../Core/ScalarBarActor.js";
import BehaviorManager_default from "./BehaviorManager.js";
//#region Sources/Rendering/Misc/SynchronizableRenderWindow/ObjectManager/index.js
var TYPE_HANDLERS = {};
var WRAPPED_ID_RE = /instance:\${([^}]+)}/;
var WRAP_ID = (id) => `instance:$\{${id}}`;
var ONE_TIME_INSTANCE_TRACKERS = {};
var SKIPPED_INSTANCE_IDS = [];
var EXCLUDE_INSTANCE_MAP = {};
var DATA_ARRAY_MAPPER = {
	vtkPoints: Points_default,
	vtkCellArray: CellArray_default,
	vtkDataArray: DataArray_default
};
function extractCallArgs(synchronizerContext, argList) {
	return argList.map((arg) => {
		const m = WRAPPED_ID_RE.exec(arg);
		if (m) return synchronizerContext.getInstance(m[1]);
		return arg;
	});
}
function extractInstanceIds(argList) {
	return argList.map((arg) => WRAPPED_ID_RE.exec(arg)).filter((m) => m).map((m) => m[1]);
}
function extractDependencyIds(state, depList = []) {
	if (state.dependencies) state.dependencies.forEach((childState) => {
		depList.push(childState.id);
		extractDependencyIds(childState, depList);
	});
	return depList;
}
function bindArrays(arraysToBind) {
	while (arraysToBind.length) {
		const [fn, args] = arraysToBind.shift();
		fn(...args);
	}
}
function createNewArrayHandler(instance, arrayMetadata, arraysToBind) {
	return (values) => {
		const regMethod = arrayMetadata.registration ? arrayMetadata.registration : "addArray";
		const location = arrayMetadata.location ? instance.getReferenceByName(arrayMetadata.location) : instance;
		let previousArray = null;
		if (arrayMetadata.location) previousArray = instance.getReferenceByName(arrayMetadata.location).getArray(arrayMetadata.name);
		else previousArray = instance[`get${regMethod.substring(3)}`]();
		if (previousArray) {
			if (previousArray.getData() !== values) arraysToBind.push([previousArray.setData, [values, arrayMetadata.numberOfComponents]]);
			return previousArray;
		}
		const array = DATA_ARRAY_MAPPER[arrayMetadata.vtkClass ? arrayMetadata.vtkClass : "vtkDataArray"].newInstance({
			...arrayMetadata,
			values
		});
		arraysToBind.push([location[regMethod], [array]]);
		return array;
	};
}
function update(type, instance, props, context) {
	if (!instance) return Promise.reject(/* @__PURE__ */ new Error(`No instance provided.`));
	const handler = TYPE_HANDLERS[type];
	if (handler && handler.update) return handler.update(instance, props, context);
	return Promise.reject(/* @__PURE__ */ new Error(`No updater for ${type}`));
}
function build(type, initialProps = {}) {
	const handler = TYPE_HANDLERS[type];
	if (handler && handler.build) return handler.build(initialProps);
	console.log("No builder for", type);
	return null;
}
function excludeInstance(type, propertyName, propertyValue) {
	EXCLUDE_INSTANCE_MAP[type] = {
		key: propertyName,
		value: propertyValue
	};
}
function getSupportedTypes() {
	return Object.keys(TYPE_HANDLERS);
}
function clearTypeMapping() {
	Object.keys(TYPE_HANDLERS).forEach((key) => {
		delete TYPE_HANDLERS[key];
	});
}
function updateRenderWindow(instance, props, context) {
	return update("vtkRenderWindow", instance, props, context);
}
function clearAllOneTimeUpdaters() {
	Object.keys(ONE_TIME_INSTANCE_TRACKERS).forEach((key) => {
		delete ONE_TIME_INSTANCE_TRACKERS[key];
	});
}
function clearOneTimeUpdaters(...ids) {
	if (ids.length === 0) return clearAllOneTimeUpdaters();
	let array = ids;
	if (array.length === 1 && Array.isArray(array[0])) array = array[0];
	array.forEach((instanceId) => {
		delete ONE_TIME_INSTANCE_TRACKERS[instanceId];
	});
	return array;
}
function notSkippedInstance(call) {
	if (call[1].length === 1) return SKIPPED_INSTANCE_IDS.indexOf(call[1][0]) === -1;
	let keep = false;
	for (let i = 0; i < call[1].length; i++) keep = keep || SKIPPED_INSTANCE_IDS.indexOf(call[1][i]) === -1;
	return keep;
}
function genericUpdater(instance, state, context) {
	context.start();
	instance.set(state.properties);
	if (state.dependencies) state.dependencies.forEach((childState) => {
		const { id, type } = childState;
		if (EXCLUDE_INSTANCE_MAP[type]) {
			const { key, value } = EXCLUDE_INSTANCE_MAP[type];
			if (!key || childState.properties[key] === value) {
				SKIPPED_INSTANCE_IDS.push(WRAP_ID(id));
				return;
			}
		}
		let childInstance = context.getInstance(id);
		if (!childInstance) {
			childInstance = build(type, { managedInstanceId: id });
			context.registerInstance(id, childInstance);
		}
		update(type, childInstance, childState, context);
	});
	if (state.calls) state.calls.filter(notSkippedInstance).forEach((call) => {
		instance[call[0]].apply(null, extractCallArgs(context, call[1]));
	});
	const dependencies = [];
	if (state.arrays) {
		const arraysToBind = [];
		const promises = Object.values(state.arrays).map((arrayMetadata) => {
			context.start();
			return context.getArray(arrayMetadata.hash, arrayMetadata.dataType, context).then(createNewArrayHandler(instance, arrayMetadata, arraysToBind)).catch((error) => {
				console.log("Error fetching array", JSON.stringify(arrayMetadata), error);
			}).finally(context.end);
		});
		context.start();
		dependencies.push(Promise.all(promises).then(() => {
			if (arraysToBind.length) instance.modified();
			bindArrays(arraysToBind);
			return true;
		}).catch((error) => {
			console.error("Error in array handling for state", JSON.stringify(state), error);
		}).finally(context.end));
	}
	context.end();
	return Promise.all(dependencies);
}
function oneTimeGenericUpdater(instance, state, context) {
	if (!ONE_TIME_INSTANCE_TRACKERS[state.id]) genericUpdater(instance, state, context);
	ONE_TIME_INSTANCE_TRACKERS[state.id] = true;
}
function rendererUpdater(instance, state, context) {
	genericUpdater(instance, state, context);
	const allActorsDeps = /* @__PURE__ */ new Set();
	if (state.dependencies) state.dependencies.forEach((childState) => {
		const viewPropInstance = context.getInstance(childState.id);
		if (viewPropInstance) {
			const flattenedDepIds = extractDependencyIds(childState);
			viewPropInstance.set({ flattenedDepIds }, true);
			flattenedDepIds.forEach((depId) => allActorsDeps.add(depId));
		}
	});
	const unregisterCandidates = /* @__PURE__ */ new Set();
	if (state.calls) state.calls.filter(notSkippedInstance).filter((call) => call[0] === "removeViewProp").forEach((call) => {
		extractInstanceIds(call[1]).forEach((vpId) => {
			const deps = context.getInstance(vpId)?.get("flattenedDepIds").flattenedDepIds;
			if (deps) deps.forEach((depId) => unregisterCandidates.add(depId));
			unregisterCandidates.add(vpId);
		});
	});
	[...unregisterCandidates].filter((depId) => !allActorsDeps.has(depId)).forEach((depId) => context.unregisterInstance(depId));
}
function vtkRenderWindowUpdater(instance, state, context) {
	if (state.calls) state.calls.filter(notSkippedInstance).filter((call) => call[0] === "removeRenderer").forEach((call) => {
		extractInstanceIds(call[1]).forEach((renId) => {
			context.getInstance(renId).getViewProps().forEach((viewProp) => {
				const deps = viewProp.get("flattenedDepIds").flattenedDepIds;
				if (deps) deps.forEach((depId) => context.unregisterInstance(depId));
				context.unregisterInstance(context.getInstanceId(viewProp));
			});
		});
	});
	instance.render();
	genericUpdater(instance, state, context);
	BehaviorManager_default.applyBehaviors(instance, state, context);
}
function colorTransferFunctionUpdater(instance, state, context) {
	context.start();
	if (!state.properties.nodes) instance.set(state.properties);
	else {
		const nodes = state.properties.nodes.map(([x, r, g, b, midpoint, sharpness]) => ({
			x,
			r,
			g,
			b,
			midpoint,
			sharpness
		}));
		instance.set({
			...state.properties,
			nodes
		}, true);
	}
	context.end();
}
function piecewiseFunctionUpdater(instance, state, context) {
	context.start();
	if (!state.properties.nodes) instance.set(state.properties);
	else {
		const nodes = state.properties.nodes.map(([x, y, midpoint, sharpness]) => ({
			x,
			y,
			midpoint,
			sharpness
		}));
		instance.set({
			...state.properties,
			nodes
		}, true);
		instance.sortAndUpdateRange();
	}
	context.end();
}
function removeUnavailableArrays(fields, availableNames) {
	const namesToDelete = [];
	const size = fields.getNumberOfArrays();
	for (let i = 0; i < size; i++) {
		const name = fields.getArray(i).getName();
		if (!availableNames.has(name)) namesToDelete.push(name);
	}
	for (let i = 0; i < namesToDelete.length; i++) fields.removeArray(namesToDelete[i]);
}
/**
* Get a unique string suitable for use as state.arrays key.
* @param {object} arrayMeta
* @returns {string} array key
*/
function getArrayKey(arrayMeta) {
	const namePart = arrayMeta.name ? `_${arrayMeta.name}` : "";
	return `${arrayMeta.hash}_${arrayMeta.dataType}${namePart}`;
}
function createDataSetUpdate(piecesToFetch = []) {
	return (instance, state, context) => {
		context.start();
		const localProperties = { ...state.properties };
		if (!state.arrays) state.arrays = {};
		for (let i = 0; i < piecesToFetch.length; i++) {
			const key = piecesToFetch[i];
			if (state.properties[key]) {
				const arrayMeta = state.properties[key];
				arrayMeta.registration = `set${capitalize(key)}`;
				const arrayKey = getArrayKey(arrayMeta);
				state.arrays[arrayKey] = arrayMeta;
				delete localProperties[key];
			}
		}
		const fieldsArrays = state.properties.fields || [];
		for (let i = 0; i < fieldsArrays.length; i++) {
			const arrayMeta = fieldsArrays[i];
			const arrayKey = getArrayKey(arrayMeta);
			state.arrays[arrayKey] = arrayMeta;
		}
		delete localProperties.fields;
		const arrayToKeep = {
			pointData: /* @__PURE__ */ new Set(),
			cellData: /* @__PURE__ */ new Set(),
			fieldData: /* @__PURE__ */ new Set()
		};
		fieldsArrays.forEach(({ location, name }) => {
			arrayToKeep[location].add(name);
		});
		removeUnavailableArrays(instance.getPointData(), arrayToKeep.pointData);
		removeUnavailableArrays(instance.getCellData(), arrayToKeep.cellData);
		const cleanState = { ...state };
		cleanState.properties = localProperties;
		const res = genericUpdater(instance, cleanState, context);
		context.end();
		return res;
	};
}
var polydataUpdater = createDataSetUpdate([
	"points",
	"polys",
	"verts",
	"lines",
	"strips"
]);
var imageDataUpdater = createDataSetUpdate([]);
function setTypeMapping(type, buildFn = null, updateFn = genericUpdater) {
	if (!build && !update) {
		delete TYPE_HANDLERS[type];
		return;
	}
	TYPE_HANDLERS[type] = {
		build: buildFn,
		update: updateFn
	};
}
var DEFAULT_ALIASES = {
	vtkMapper: [
		"vtkOpenGLPolyDataMapper",
		"vtkCompositePolyDataMapper2",
		"vtkDataSetMapper"
	],
	vtkProperty: ["vtkOpenGLProperty"],
	vtkRenderer: ["vtkOpenGLRenderer"],
	vtkCamera: ["vtkOpenGLCamera"],
	vtkColorTransferFunction: ["vtkPVDiscretizableColorTransferFunction"],
	vtkActor: ["vtkOpenGLActor", "vtkPVLODActor"],
	vtkLight: ["vtkOpenGLLight", "vtkPVLight"],
	vtkTexture: ["vtkOpenGLTexture"],
	vtkImageMapper: ["vtkOpenGLImageSliceMapper"],
	vtkVolumeMapper: ["vtkFixedPointVolumeRayCastMapper"]
};
var DEFAULT_MAPPING = {
	vtkAxesActor: {
		build: AxesActor_default.newInstance,
		update: genericUpdater
	},
	vtkRenderWindow: {
		build: RenderWindow_default.newInstance,
		update: vtkRenderWindowUpdater
	},
	vtkRenderer: {
		build: Renderer_default.newInstance,
		update: rendererUpdater
	},
	vtkLookupTable: {
		build: LookupTable_default.newInstance,
		update: genericUpdater
	},
	vtkCamera: {
		build: Camera_default.newInstance,
		update: oneTimeGenericUpdater
	},
	vtkPolyData: {
		build: PolyData_default.newInstance,
		update: polydataUpdater
	},
	vtkImageData: {
		build: ImageData_default.newInstance,
		update: imageDataUpdater
	},
	vtkMapper: {
		build: Mapper_default.newInstance,
		update: genericUpdater
	},
	vtkGlyph3DMapper: {
		build: Glyph3DMapper_default.newInstance,
		update: genericUpdater
	},
	vtkProperty: {
		build: Property_default.newInstance,
		update: genericUpdater
	},
	vtkActor: {
		build: Actor_default.newInstance,
		update: genericUpdater
	},
	vtkLight: {
		build: Light_default.newInstance,
		update: genericUpdater
	},
	vtkColorTransferFunction: {
		build: ColorTransferFunction_default.newInstance,
		update: colorTransferFunctionUpdater
	},
	vtkTexture: {
		build: Texture_default.newInstance,
		update: genericUpdater
	},
	vtkVolume: {
		build: Volume_default.newInstance,
		update: genericUpdater
	},
	vtkVolumeMapper: {
		build: VolumeMapper_default.newInstance,
		update: genericUpdater
	},
	vtkVolumeProperty: {
		build: VolumeProperty_default.newInstance,
		update: genericUpdater
	},
	vtkImageSlice: {
		build: ImageSlice_default.newInstance,
		update: genericUpdater
	},
	vtkImageMapper: {
		build: ImageMapper_default.newInstance,
		update: genericUpdater
	},
	vtkImageProperty: {
		build: ImageProperty_default.newInstance,
		update: genericUpdater
	},
	vtkPiecewiseFunction: {
		build: PiecewiseFunction_default.newInstance,
		update: piecewiseFunctionUpdater
	},
	vtkCubeAxesActor: {
		build: CubeAxesActor_default.newInstance,
		update: genericUpdater
	},
	vtkScalarBarActor: {
		build: ScalarBarActor_default.newInstance,
		update: genericUpdater
	}
};
function setDefaultMapping(reset = true) {
	if (reset) clearTypeMapping();
	Object.keys(DEFAULT_MAPPING).forEach((type) => {
		const mapping = DEFAULT_MAPPING[type];
		setTypeMapping(type, mapping.build, mapping.update);
	});
}
function applyDefaultAliases() {
	Object.keys(DEFAULT_ALIASES).forEach((name) => {
		DEFAULT_ALIASES[name].forEach((alias) => {
			TYPE_HANDLERS[alias] = TYPE_HANDLERS[name];
		});
	});
}
function alwaysUpdateCamera() {
	setTypeMapping("vtkCamera", Camera_default.newInstance);
	applyDefaultAliases();
}
setDefaultMapping();
applyDefaultAliases();
EXCLUDE_INSTANCE_MAP.vtkPVLight = {};
var ObjectManager_default = {
	build,
	update,
	genericUpdater,
	oneTimeGenericUpdater,
	setTypeMapping,
	clearTypeMapping,
	getSupportedTypes,
	clearOneTimeUpdaters,
	updateRenderWindow,
	excludeInstance,
	setDefaultMapping,
	applyDefaultAliases,
	alwaysUpdateCamera
};
//#endregion
export { ObjectManager_default as default };
