// @kitware/vtk.js/Rendering/Core/Mapper/CoincidentTopologyHelper.js
import macros_default from "../../../macros.js";
import Static_default, { Resolve } from "./Static.js";
//#region Sources/Rendering/Core/Mapper/CoincidentTopologyHelper.js
function addCoincidentTopologyMethods(publicAPI, model, nameList) {
	nameList.forEach((item) => {
		publicAPI[`get${item.method}`] = () => model[item.key];
		publicAPI[`set${item.method}`] = macros_default.objectSetterMap.object(publicAPI, model, {
			name: item.key,
			params: ["factor", "offset"]
		});
	});
}
var CATEGORIES = [
	"Polygon",
	"Line",
	"Point"
];
var staticOffsetModel = {
	Polygon: {
		factor: 2,
		offset: 0
	},
	Line: {
		factor: 1,
		offset: -1
	},
	Point: {
		factor: 0,
		offset: -2
	}
};
var noOp = () => void 0;
var staticOffsetAPI = { modified: noOp };
addCoincidentTopologyMethods(staticOffsetAPI, staticOffsetModel, CATEGORIES.map((key) => ({
	key,
	method: `ResolveCoincidentTopology${key}OffsetParameters`
})));
function implementCoincidentTopologyMethods(publicAPI, model) {
	if (model.resolveCoincidentTopology === void 0) model.resolveCoincidentTopology = false;
	macros_default.setGet(publicAPI, model, ["resolveCoincidentTopology"]);
	model.topologyOffset = {
		Polygon: {
			factor: 0,
			offset: 0
		},
		Line: {
			factor: 0,
			offset: 0
		},
		Point: {
			factor: 0,
			offset: 0
		}
	};
	Object.keys(Static_default).forEach((methodName) => {
		publicAPI[methodName] = Static_default[methodName];
	});
	Object.keys(staticOffsetAPI).filter((methodName) => methodName !== "modified").forEach((methodName) => {
		publicAPI[methodName] = staticOffsetAPI[methodName];
	});
	addCoincidentTopologyMethods(publicAPI, model.topologyOffset, CATEGORIES.map((key) => ({
		key,
		method: `RelativeCoincidentTopology${key}OffsetParameters`
	})));
	publicAPI.getCoincidentTopologyPolygonOffsetParameters = () => {
		const globalValue = staticOffsetAPI.getResolveCoincidentTopologyPolygonOffsetParameters();
		const localValue = publicAPI.getRelativeCoincidentTopologyPolygonOffsetParameters();
		return {
			factor: globalValue.factor + localValue.factor,
			offset: globalValue.offset + localValue.offset
		};
	};
	publicAPI.getCoincidentTopologyLineOffsetParameters = () => {
		const globalValue = staticOffsetAPI.getResolveCoincidentTopologyLineOffsetParameters();
		const localValue = publicAPI.getRelativeCoincidentTopologyLineOffsetParameters();
		return {
			factor: globalValue.factor + localValue.factor,
			offset: globalValue.offset + localValue.offset
		};
	};
	publicAPI.getCoincidentTopologyPointOffsetParameter = () => {
		const globalValue = staticOffsetAPI.getResolveCoincidentTopologyPointOffsetParameters();
		const localValue = publicAPI.getRelativeCoincidentTopologyPointOffsetParameters();
		return {
			factor: globalValue.factor + localValue.factor,
			offset: globalValue.offset + localValue.offset
		};
	};
}
var CoincidentTopologyHelper_default = {
	implementCoincidentTopologyMethods,
	staticOffsetAPI,
	otherStaticMethods: Static_default,
	CATEGORIES,
	Resolve
};
//#endregion
export { CATEGORIES, Resolve, CoincidentTopologyHelper_default as default };
