// @kitware/vtk.js/Rendering/Core/Mapper/Static.js
//#region Sources/Rendering/Core/Mapper/Static.js
var Resolve = {
	Off: 0,
	PolygonOffset: 1
};
var resolveCoincidentTopologyPolygonOffsetFaces = Resolve.PolygonOffset;
var resolveCoincidentTopology = Resolve.Off;
var RESOLVE_COINCIDENT_TOPOLOGY_MODE = ["VTK_RESOLVE_OFF", "VTK_RESOLVE_POLYGON_OFFSET"];
function getResolveCoincidentTopologyPolygonOffsetFaces() {
	return resolveCoincidentTopologyPolygonOffsetFaces;
}
function setResolveCoincidentTopologyPolygonOffsetFaces(value) {
	const changed = resolveCoincidentTopologyPolygonOffsetFaces === value;
	resolveCoincidentTopologyPolygonOffsetFaces = value;
	return changed;
}
function getResolveCoincidentTopology() {
	return resolveCoincidentTopology;
}
function setResolveCoincidentTopology(mode = 0) {
	const changed = resolveCoincidentTopology === mode;
	resolveCoincidentTopology = mode;
	return changed;
}
function setResolveCoincidentTopologyToDefault() {
	return setResolveCoincidentTopology(Resolve.Off);
}
function setResolveCoincidentTopologyToOff() {
	return setResolveCoincidentTopology(Resolve.Off);
}
function setResolveCoincidentTopologyToPolygonOffset() {
	return setResolveCoincidentTopology(Resolve.PolygonOffset);
}
function getResolveCoincidentTopologyAsString() {
	return RESOLVE_COINCIDENT_TOPOLOGY_MODE[resolveCoincidentTopology];
}
var Static_default = {
	Resolve,
	getResolveCoincidentTopologyAsString,
	getResolveCoincidentTopologyPolygonOffsetFaces,
	getResolveCoincidentTopology,
	setResolveCoincidentTopology,
	setResolveCoincidentTopologyPolygonOffsetFaces,
	setResolveCoincidentTopologyToDefault,
	setResolveCoincidentTopologyToOff,
	setResolveCoincidentTopologyToPolygonOffset
};
//#endregion
export { RESOLVE_COINCIDENT_TOPOLOGY_MODE, Resolve, Static_default as default, getResolveCoincidentTopology, getResolveCoincidentTopologyAsString, getResolveCoincidentTopologyPolygonOffsetFaces, setResolveCoincidentTopology, setResolveCoincidentTopologyPolygonOffsetFaces, setResolveCoincidentTopologyToDefault, setResolveCoincidentTopologyToOff, setResolveCoincidentTopologyToPolygonOffset };
