// @kitware/vtk.js/IO/Geometry/GLTFImporter.js
import macros_default from "../../macros.js";
import DataAccessHelper_default from "../Core/DataAccessHelper.js";
import BinaryHelper_default from "../Core/BinaryHelper.js";
import DracoReader_default from "./DracoReader.js";
import "./GLTFImporter/Constants.js";
import { clearImageCaches } from "./GLTFImporter/Utils.js";
import { GLTFCameraToVTKCamera, applyTransformToCamera, createPropertyFromGLTFMaterial, createVTKObjects, parseGLTF } from "./GLTFImporter/Reader.js";
import parseGLB from "./GLTFImporter/Decoder.js";
import { createAnimationMixer } from "./GLTFImporter/Animations.js";
//#region Sources/IO/Geometry/GLTFImporter/index.js
var { vtkDebugMacro, vtkErrorMacro } = macros_default;
function vtkGLTFImporter(publicAPI, model) {
	model.classHierarchy.push("vtkGLTFImporter");
	if (!model.dataAccessHelper) model.dataAccessHelper = DataAccessHelper_default.get("http");
	function fetchData(url, option = {}) {
		const { compression, progressCallback } = model;
		if (option.binary) return model.dataAccessHelper.fetchBinary(url, {
			compression,
			progressCallback
		});
		return model.dataAccessHelper.fetchText(publicAPI, url, {
			compression,
			progressCallback
		});
	}
	publicAPI.setUrl = (url, option = { binary: true }) => {
		model.url = url;
		const path = url.split("/");
		path.pop();
		model.baseURL = path.join("/");
		model.compression = option.compression;
		model.sceneId = option.sceneId ? option.sceneId : 0;
		return publicAPI.loadData({
			progressCallback: option.progressCallback,
			binary: !!option.binary
		});
	};
	publicAPI.loadData = (option = {}) => {
		const promise = fetchData(model.url, option);
		promise.then(publicAPI.parse);
		return promise;
	};
	publicAPI.parse = (content) => {
		if (typeof content === "string") publicAPI.parseAsText(content);
		else publicAPI.parseAsBinary(content);
	};
	publicAPI.parseAsBinary = async (content) => {
		if (!content) return;
		if (content !== model.parseData) publicAPI.modified();
		else return;
		const glTF = {};
		const options = { baseUri: model.baseURL };
		if (BinaryHelper_default.arrayBufferToString(new Uint8Array(content, 0, 4)) === "glTF") {
			const { json, buffers } = parseGLB(content);
			vtkDebugMacro("Loaded GLB", json, buffers);
			glTF.glbBuffers = buffers;
			glTF.json = json;
		} else glTF.json = JSON.parse(BinaryHelper_default.arrayBufferToString(content));
		if (glTF.json.asset === void 0 || glTF.json.asset.version[0] < 2) {
			vtkErrorMacro("Unsupported asset. glTF versions >=2.0 are supported.");
			return;
		}
		model.glTFTree = await parseGLTF(glTF, options);
		model.actors = /* @__PURE__ */ new Map();
		model.cameras = /* @__PURE__ */ new Map();
		model.lights = /* @__PURE__ */ new Map();
		model.animations = [];
		model.variants = [];
		model.variantMappings = /* @__PURE__ */ new Map();
		await createVTKObjects(model);
		model.scenes = model.glTFTree.scenes;
		publicAPI.invokeReady();
	};
	publicAPI.parseAsText = (content) => {
		if (!content) return;
		if (content !== model.parseData) publicAPI.modified();
		else return;
		model.parseData = content;
	};
	publicAPI.requestData = (inData, outData) => {
		publicAPI.parse(model.parseData);
	};
	publicAPI.setDracoDecoder = async (dracoDecoder) => {
		await DracoReader_default.setDracoDecoder(dracoDecoder);
	};
	publicAPI.importActors = () => {
		model.actors.forEach((actor) => model.renderer.addActor(actor));
	};
	publicAPI.importCameras = () => {
		model.glTFTree.cameras?.forEach((glTFcamera) => {
			const camera = GLTFCameraToVTKCamera(glTFcamera);
			model.cameras.set(glTFcamera.id, camera);
		});
		model.scenes.forEach((scene) => {
			scene.nodes.forEach((node) => {
				const camera = model.cameras.get(node.camera?.id);
				if (camera) applyTransformToCamera(camera, node.transform);
			});
		});
	};
	publicAPI.importAnimations = () => {
		if (model.glTFTree.animations?.length > 0) {
			model.animationMixer = createAnimationMixer(model.actors, model.glTFTree.accessors);
			model.glTFTree.animations.forEach((animation) => {
				model.animationMixer.addAnimation(animation);
			});
		}
		model.animations = model.glTFTree.animations || [];
	};
	publicAPI.importLights = () => {
		model.lights?.forEach((light) => {
			vtkDebugMacro("Adding light", light);
			model.renderer.addLight(light);
		});
	};
	publicAPI.setCamera = (cameraId) => {
		const camera = model.cameras.get(cameraId);
		if (!camera) {
			vtkErrorMacro(`Camera ${cameraId} not found`);
			return;
		}
		vtkDebugMacro("Setting camera", camera);
		model.renderer.setActiveCamera(camera);
	};
	publicAPI.switchToVariant = async (variantIndex) => {
		const promises = Array.from(model.actors).map(async ([nodeId, actor]) => {
			vtkDebugMacro("Switching to variant", variantIndex, "for node", nodeId);
			const variantMappings = model.variantMappings.get(nodeId);
			if (variantMappings) {
				const mapping = variantMappings.find((m) => m.variants.includes(variantIndex));
				if (mapping) {
					const variantMaterial = model.glTFTree.materials[mapping.material];
					await createPropertyFromGLTFMaterial(model, variantMaterial, actor);
				}
			}
		});
		await Promise.all(promises);
	};
	publicAPI.clear = () => {
		model.actors?.clear?.();
		model.cameras?.clear?.();
		model.lights?.clear?.();
		model.nodeLights?.clear?.();
		model.variantMappings?.clear?.();
		model.nodeTransforms?.clear?.();
		model.nodeChildren?.clear?.();
		model.skins?.clear?.();
		model.morphTargets?.clear?.();
		model.materialProperties?.clear?.();
		model.pointerAnimations = [];
		model.nodeAnimations = [];
		model.animationClips = [];
		model.skeletons = [];
		model.animations = [];
		model.scenes = [];
		model.glTFTree = null;
		model.parseData = null;
		clearImageCaches();
	};
	publicAPI.delete = macros_default.chain(() => publicAPI.clear(), publicAPI.delete);
}
var DEFAULT_VALUES = {};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.get(publicAPI, model, [
		"url",
		"baseURL",
		"actors",
		"scenes",
		"cameras",
		"animations",
		"animationMixer",
		"variants",
		"variantMappings"
	]);
	macros_default.set(publicAPI, model, ["renderer", "dracoDecoder"]);
	macros_default.event(publicAPI, model, "ready");
	vtkGLTFImporter(publicAPI, model);
	if (!model.compression) model.compression = null;
	if (!model.progressCallback) model.progressCallback = null;
}
var newInstance = macros_default.newInstance(extend, "vtkGLTFImporter");
var GLTFImporter_default = {
	extend,
	newInstance
};
//#endregion
export { GLTFImporter_default as default, extend, newInstance };
