// @kitware/vtk.js/IO/Geometry/IFCImporter.js
import macros_default from "../../macros.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import CellArray_default from "../../Common/Core/CellArray.js";
import MatrixBuilder_default from "../../Common/Core/MatrixBuilder.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
import AppendPolyData_default from "../../Filters/General/AppendPolyData.js";
import DataAccessHelper_default from "../Core/DataAccessHelper.js";
import "../Core/DataAccessHelper/LiteHttpDataAccessHelper.js";
import Actor_default from "../../Rendering/Core/Actor.js";
import Mapper_default from "../../Rendering/Core/Mapper.js";
import { mat3 } from "gl-matrix";
//#region Sources/IO/Geometry/IFCImporter/index.js
var { vtkErrorMacro } = macros_default;
var WebIFC;
/**
* Set WebIFC api to be used by vtkIFCImporter
* @param {object} ifcApi
*/
function setIFCAPI(ifcApi) {
	WebIFC = ifcApi;
}
function vtkIFCImporter(publicAPI, model) {
	model.classHierarchy.push("vtkIFCImporter");
	const meshes = [];
	/**
	* Create a vtkPolyData from an IFC mesh object
	* @param {object} mesh the IFC web mesh object
	* @returns vtkPolyData
	*/
	function createPolyDataFromIFCMesh(mesh) {
		const { vertices, indices } = mesh;
		const pd = PolyData_default.newInstance();
		const cells = CellArray_default.newInstance();
		const pointValues = new Float32Array(vertices.length / 2);
		const normalsArray = new Float32Array(vertices.length / 2);
		for (let i = 0, p = 0; i < vertices.length; p += 3) {
			pointValues[p] = vertices[i++];
			pointValues[p + 1] = vertices[i++];
			pointValues[p + 2] = vertices[i++];
			normalsArray[p] = vertices[i++];
			normalsArray[p + 1] = vertices[i++];
			normalsArray[p + 2] = vertices[i++];
		}
		const nCells = indices.length;
		cells.resize(3 * nCells / 3);
		for (let cellId = 0; cellId < nCells; cellId += 3) {
			const cell = indices.slice(cellId, cellId + 3);
			cells.insertNextCell(cell);
		}
		pd.getPoints().setData(pointValues, 3);
		pd.setStrips(cells);
		pd.getPointData().setNormals(DataArray_default.newInstance({
			name: "Normals",
			values: normalsArray,
			numberOfComponents: 3
		}));
		return pd;
	}
	/**
	* Create a colored vtkPolyData from an IFC mesh object
	* @param {object} mesh the IFC mesh object
	* @returns vtkPolyData
	*/
	function createColoredPolyDataFromIFCMesh(mesh) {
		const { vertices, indices, color, userMatrix } = mesh;
		const pd = PolyData_default.newInstance();
		const cells = CellArray_default.newInstance();
		const pointValues = new Float32Array(vertices.length / 2);
		const normalsArray = new Float32Array(vertices.length / 2);
		const colorArray = new Float32Array(vertices.length / 2);
		if (userMatrix) {
			for (let i = 0, p = 0; i < vertices.length; p += 3) {
				pointValues[p] = vertices[i++];
				pointValues[p + 1] = vertices[i++];
				pointValues[p + 2] = vertices[i++];
				normalsArray[p] = vertices[i++];
				normalsArray[p + 1] = vertices[i++];
				normalsArray[p + 2] = vertices[i++];
				colorArray[p] = color.x;
				colorArray[p + 1] = color.y;
				colorArray[p + 2] = color.z;
			}
			MatrixBuilder_default.buildFromRadian().setMatrix(userMatrix).apply(pointValues);
			MatrixBuilder_default.buildFromRadian().multiply3x3(mat3.fromMat4(mat3.create(), userMatrix)).apply(normalsArray);
		} else for (let i = 0, p = 0; i < vertices.length; p += 3) {
			pointValues[p] = vertices[i++];
			pointValues[p + 1] = vertices[i++];
			pointValues[p + 2] = vertices[i++];
			normalsArray[p] = vertices[i++];
			normalsArray[p + 1] = vertices[i++];
			normalsArray[p + 2] = vertices[i++];
			colorArray[p] = color.x;
			colorArray[p + 1] = color.y;
			colorArray[p + 2] = color.z;
		}
		const nCells = indices.length;
		cells.resize(3 * nCells / 3);
		for (let cellId = 0; cellId < nCells; cellId += 3) {
			const cell = indices.slice(cellId, cellId + 3);
			cells.insertNextCell(cell);
		}
		pd.getPoints().setData(pointValues, 3);
		pd.setPolys(cells);
		pd.getPointData().setNormals(DataArray_default.newInstance({
			name: "Normals",
			values: normalsArray,
			numberOfComponents: 3
		}));
		pd.getPointData().setScalars(DataArray_default.newInstance({
			name: "Colors",
			values: colorArray,
			numberOfComponents: 3
		}));
		return pd;
	}
	function parseIfc(content) {
		const modelID = model._ifcApi.OpenModel(new Uint8Array(content), {
			COORDINATE_TO_ORIGIN: true,
			USE_FAST_BOOLS: true
		});
		model._ifcApi.StreamAllMeshes(modelID, (mesh) => {
			const placedGeometries = mesh.geometries;
			for (let i = 0; i < placedGeometries.size(); i++) {
				const placedGeometry = placedGeometries.get(i);
				const ifcGeometryData = model._ifcApi.GetGeometry(modelID, placedGeometry.geometryExpressID);
				const ifcVertices = model._ifcApi.GetVertexArray(ifcGeometryData.GetVertexData(), ifcGeometryData.GetVertexDataSize());
				const ifcIndices = model._ifcApi.GetIndexArray(ifcGeometryData.GetIndexData(), ifcGeometryData.GetIndexDataSize());
				meshes.push({
					vertices: ifcVertices,
					indices: ifcIndices,
					color: placedGeometry.color,
					userMatrix: placedGeometry.flatTransformation
				});
			}
		});
		model._ifcApi.CloseModel(modelID);
	}
	if (!model.dataAccessHelper) model.dataAccessHelper = DataAccessHelper_default.get("http");
	function fetchData(url, options = {}) {
		const { compression, progressCallback } = model;
		return model.dataAccessHelper.fetchBinary(url, {
			compression,
			progressCallback
		});
	}
	publicAPI.setUrl = (url, options = { binary: true }) => {
		model.url = url;
		model.baseURL = url.split("/").slice(0, -1).join("/");
		model.compression = options.compression;
		return publicAPI.loadData(options);
	};
	publicAPI.loadData = (options = {}) => fetchData(model.url, options).then(publicAPI.parse);
	publicAPI.parse = (content) => {
		publicAPI.parseAsArrayBuffer(content);
	};
	publicAPI.parseAsArrayBuffer = (content) => {
		if (!content) {
			vtkErrorMacro("No content to parse.");
			return;
		}
		if (!WebIFC) {
			vtkErrorMacro("vtkIFCImporter requires WebIFC API to be set.");
			return;
		}
		model._ifcApi = new WebIFC.IfcAPI();
		model._ifcApi.Init().then(() => {
			parseIfc(content);
			publicAPI.invokeReady();
		});
	};
	publicAPI.importActors = (renderer) => {
		if (model.mergeGeometries) {
			const opaqueMeshes = meshes.filter((mesh) => mesh.color.w === 1);
			const oapd = AppendPolyData_default.newInstance();
			opaqueMeshes.forEach((mesh) => {
				const pd = createColoredPolyDataFromIFCMesh(mesh);
				oapd.addInputData(pd);
			});
			let mapper = Mapper_default.newInstance();
			mapper.setColorModeToDirectScalars();
			mapper.setInputConnection(oapd.getOutputPort());
			let actor = Actor_default.newInstance();
			actor.setMapper(mapper);
			renderer.addActor(actor);
			const transparentMeshes = meshes.filter((mesh) => mesh.color.w < 1);
			const tapd = AppendPolyData_default.newInstance();
			transparentMeshes.forEach((mesh) => {
				const pd = createColoredPolyDataFromIFCMesh(mesh);
				tapd.addInputData(pd);
			});
			mapper = Mapper_default.newInstance();
			mapper.setColorModeToDirectScalars();
			mapper.setInputConnection(tapd.getOutputPort());
			actor = Actor_default.newInstance();
			actor.setMapper(mapper);
			actor.getProperty().setOpacity(.5);
			renderer.addActor(actor);
		} else meshes.forEach((mesh) => {
			const pd = createPolyDataFromIFCMesh(mesh);
			const mapper = Mapper_default.newInstance();
			mapper.setInputData(pd);
			const actor = Actor_default.newInstance();
			actor.setMapper(mapper);
			const { x, y, z, w } = mesh.color;
			actor.getProperty().setColor(x, y, z);
			actor.getProperty().setOpacity(w);
			actor.setUserMatrix(mesh.userMatrix);
			renderer.addActor(actor);
		});
	};
}
var DEFAULT_VALUES = { mergeGeometries: false };
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.get(publicAPI, model, ["url", "baseURL"]);
	macros_default.setGet(publicAPI, model, ["dataAccessHelper", "mergeGeometries"]);
	macros_default.event(publicAPI, model, "ready");
	macros_default.algo(publicAPI, model, 0, 1);
	vtkIFCImporter(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkIFCImporter");
var IFCImporter_default = {
	newInstance,
	extend,
	setIFCAPI
};
//#endregion
export { IFCImporter_default as default, extend, newInstance };
