// @kitware/vtk.js/Widgets/Widgets3D/SphereWidget/state.js
import StateBuilder_default from "../../Core/StateBuilder.js";
//#region Sources/Widgets/Widgets3D/SphereWidget/state.js
function stateGenerator() {
	return StateBuilder_default.createBuilder().addStateFromMixin({
		labels: ["moveHandle"],
		mixins: [
			"origin",
			"color",
			"scale1",
			"visible",
			"manipulator"
		],
		name: "moveHandle",
		initialValues: {
			scale1: 20,
			visible: true
		}
	}).addStateFromMixin({
		labels: ["centerHandle"],
		mixins: [
			"origin",
			"color",
			"scale1",
			"visible",
			"manipulator"
		],
		name: "centerHandle",
		initialValues: {
			scale1: 20,
			visible: true
		}
	}).addStateFromMixin({
		labels: ["borderHandle"],
		mixins: [
			"origin",
			"color",
			"scale1",
			"visible",
			"manipulator"
		],
		name: "borderHandle",
		initialValues: {
			scale1: 20,
			visible: true
		}
	}).addStateFromMixin({
		labels: ["sphereHandle"],
		mixins: [
			"origin",
			"color",
			"scale1",
			"visible",
			"orientation"
		],
		name: "sphereHandle",
		initialValues: {
			visible: true,
			radius: 1
		}
	}).build();
}
//#endregion
export { stateGenerator as default };
