// @kitware/vtk.js/Rendering/Core/ImageResliceMapper/OutlineBasisHelper.js
import "../../../Common/Core/Math/Constants.js";
import Math_default from "../../../Common/Core/Math.js";
import { mat3, vec3 } from "gl-matrix";
//#region Sources/Rendering/Core/ImageResliceMapper/OutlineBasisHelper.js
var tmpMat3 = new Float64Array(9);
var tmpTangent1 = [
	0,
	0,
	0
];
var tmpTangent2 = [
	0,
	0,
	0
];
var tmpVec3a = vec3.create();
var tmpVec3b = vec3.create();
function computeSliceTangents(slicePlane, planeNormalOut = [
	0,
	0,
	1
]) {
	planeNormalOut[0] = 0;
	planeNormalOut[1] = 0;
	planeNormalOut[2] = 1;
	if (slicePlane) {
		const n = slicePlane.getNormal();
		planeNormalOut[0] = n[0];
		planeNormalOut[1] = n[1];
		planeNormalOut[2] = n[2];
	}
	Math_default.normalize(planeNormalOut);
	tmpTangent1[0] = 0;
	tmpTangent1[1] = 0;
	tmpTangent1[2] = 0;
	tmpTangent2[0] = 0;
	tmpTangent2[1] = 0;
	tmpTangent2[2] = 0;
	if (slicePlane) {
		Math_default.perpendiculars(planeNormalOut, tmpTangent1, tmpTangent2, 0);
		if (Math_default.norm(tmpTangent1) < 1e-6 || Math_default.norm(tmpTangent2) < 1e-6) {
			tmpTangent1[0] = 1;
			tmpTangent1[1] = 0;
			tmpTangent1[2] = 0;
			tmpTangent2[0] = 0;
			tmpTangent2[1] = 1;
			tmpTangent2[2] = 0;
		}
	} else {
		tmpTangent1[0] = 1;
		tmpTangent2[1] = 1;
	}
	return {
		planeNormal: planeNormalOut,
		tangent1: tmpTangent1,
		tangent2: tmpTangent2
	};
}
function computeInputOutlineBasis(imageData, tangent1, tangent2, texelSizeOut) {
	mat3.set(tmpMat3, ...imageData.getDirection());
	mat3.invert(tmpMat3, tmpMat3);
	vec3.transformMat3(tmpVec3a, tangent1, tmpMat3);
	vec3.transformMat3(tmpVec3b, tangent2, tmpMat3);
	const dims = imageData.getDimensions();
	const spacing = imageData.getSpacing();
	const minSpacing = Math.min(Math.abs(spacing[0]), Math.abs(spacing[1]), Math.abs(spacing[2]));
	texelSizeOut[0] = minSpacing / (dims[0] * Math.abs(spacing[0]));
	texelSizeOut[1] = minSpacing / (dims[1] * Math.abs(spacing[1]));
	texelSizeOut[2] = minSpacing / (dims[2] * Math.abs(spacing[2]));
	return {
		tangent1: tmpVec3a,
		tangent2: tmpVec3b,
		texelSize: texelSizeOut
	};
}
//#endregion
export { computeInputOutlineBasis, computeSliceTangents };
