// @kitware/vtk.js/Rendering/Core/ImageResliceMapper/TransferFunctionHelper.js
//#region Sources/Rendering/Core/ImageResliceMapper/TransferFunctionHelper.js
function computeColorShiftScale({ colorWindow, colorLevel, useLookupTableScalarRange, colorRange, volumeScale = 1, volumeOffset = 0 }) {
	let cw = colorWindow;
	let cl = colorLevel;
	if (useLookupTableScalarRange && colorRange) {
		cw = colorRange[1] - colorRange[0];
		cl = .5 * (colorRange[1] + colorRange[0]);
	}
	return {
		colorScale: volumeScale / cw,
		colorShift: (volumeOffset - cl) / cw + .5
	};
}
function computeOpacityShiftScale({ pwfRange, volumeScale = 1, volumeOffset = 0 }) {
	if (!pwfRange) return {
		opacityScale: 1,
		opacityShift: 0
	};
	const length = pwfRange[1] - pwfRange[0];
	const mid = .5 * (pwfRange[0] + pwfRange[1]);
	return {
		opacityScale: volumeScale / length,
		opacityShift: (volumeOffset - mid) / length + .5
	};
}
function getTransferFunctionHash({ currentValidInputs, independentComponents, numberOfRows, kind, getInputProperty }) {
	if (!currentValidInputs.length) return "0";
	const fnName = kind === "color" ? "getRGBTransferFunction" : "getPiecewiseFunction";
	const rows = [];
	for (let i = 0; i < numberOfRows; i++) {
		const property = independentComponents ? getInputProperty(currentValidInputs[i].inputIndex) : getInputProperty(currentValidInputs[0].inputIndex);
		const fn = property?.[fnName]?.(independentComponents ? 0 : i);
		rows.push(`${property?.getMTime?.() ?? 0}:${fn?.getMTime?.() ?? 0}`);
	}
	return rows.join("|");
}
//#endregion
export { computeColorShiftScale, computeOpacityShiftScale, getTransferFunctionHash };
