// @kitware/vtk.js/Widgets/Widgets3D/ImageCroppingWidget/state.js
import StateBuilder_default from "../../Core/StateBuilder.js";
import { AXES, handleTypeFromName } from "./helpers.js";
//#region Sources/Widgets/Widgets3D/ImageCroppingWidget/state.js
function build() {
	const builder = StateBuilder_default.createBuilder();
	builder.addField({
		name: "indexToWorldT",
		initialValue: Array(16).fill(0)
	}).addField({
		name: "worldToIndexT",
		initialValue: Array(16).fill(0)
	});
	const croppingState = StateBuilder_default.createBuilder().addField({
		name: "planes",
		initialValue: [
			0,
			1,
			0,
			1,
			0,
			1
		]
	}).build();
	builder.addStateFromInstance({
		labels: ["croppingPlanes"],
		name: "croppingPlanes",
		instance: croppingState
	});
	for (let i = -1; i < 2; i++) for (let j = -1; j < 2; j++) for (let k = -1; k < 2; k++) if (i !== 0 || j !== 0 || k !== 0) {
		const name = AXES[i + 1] + AXES[j + 1] + AXES[k + 1];
		const type = handleTypeFromName(name);
		builder.addStateFromMixin({
			labels: [
				"handles",
				name,
				type
			],
			mixins: [
				"name",
				"origin",
				"color",
				"scale1",
				"visible",
				"manipulator"
			],
			name,
			initialValues: {
				scale1: 30,
				origin: [
					i,
					j,
					k
				],
				visible: true,
				name
			}
		});
	}
	return builder.build();
}
//#endregion
export { build as default };
