// @kitware/vtk.js/Filters/Sources/ArcSource.js
import macros_default from "../../macros.js";
import Math_default from "../../Common/Core/Math.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import CellArray_default from "../../Common/Core/CellArray.js";
import { DesiredOutputPrecision } from "../../Common/DataModel/DataSetAttributes/Constants.js";
import Points_default from "../../Common/Core/Points.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
//#region Sources/Filters/Sources/ArcSource/index.js
var { VtkDataTypes } = DataArray_default;
function vtkArcSource(publicAPI, model) {
	model.classHierarchy.push("vtkArcSource");
	publicAPI.requestData = (inData, outData) => {
		const numLines = model.resolution;
		const numPts = model.resolution + 1;
		const tc = [0, 0];
		const output = outData[0]?.initialize() || PolyData_default.newInstance();
		let angle = 0;
		let radius = .5;
		const perpendicular = [
			0,
			0,
			0
		];
		const v1 = [
			0,
			0,
			0
		];
		if (model.useNormalAndAngle) {
			angle = Math_default.radiansFromDegrees(model.angle);
			v1[0] = model.polarVector[0];
			v1[1] = model.polarVector[1];
			v1[2] = model.polarVector[2];
			Math_default.cross(model.normal, model.polarVector, perpendicular);
			radius = Math_default.normalize(v1);
		} else {
			Math_default.subtract(model.point1, model.center, v1);
			const v2 = [
				0,
				0,
				0
			];
			Math_default.subtract(model.point2, model.center, v2);
			const normal = [
				0,
				0,
				0
			];
			Math_default.cross(v1, v2, normal);
			Math_default.cross(normal, v1, perpendicular);
			const dotProduct = Math_default.dot(v1, v2) / (Math_default.norm(v1) * Math_default.norm(v2));
			angle = Math.acos(dotProduct);
			if (model.negative) angle -= 2 * Math.PI;
			radius = Math_default.normalize(v1);
		}
		const angleInc = angle / model.resolution;
		Math_default.normalize(perpendicular);
		let pointType = 0;
		if (model.outputPointsPrecision === DesiredOutputPrecision.SINGLE) pointType = VtkDataTypes.FLOAT;
		else if (model.outputPointsPrecision === DesiredOutputPrecision.DOUBLE) pointType = VtkDataTypes.DOUBLE;
		const points = Points_default.newInstance({ dataType: pointType });
		points.setNumberOfPoints(numPts);
		const tcoords = DataArray_default.newInstance({
			numberOfComponents: 2,
			size: numPts * 2,
			dataType: VtkDataTypes.FLOAT,
			name: "TextureCoordinates"
		});
		const lines = CellArray_default.newInstance();
		lines.allocate(numLines);
		let theta = 0;
		for (let i = 0; i <= model.resolution; ++i, theta += angleInc) {
			const cosine = Math.cos(theta);
			const sine = Math.sin(theta);
			const p = [
				model.center[0] + cosine * radius * v1[0] + sine * radius * perpendicular[0],
				model.center[1] + cosine * radius * v1[1] + sine * radius * perpendicular[1],
				model.center[2] + cosine * radius * v1[2] + sine * radius * perpendicular[2]
			];
			tc[0] = i / model.resolution;
			tc[1] = 0;
			points.setPoint(i, ...p);
			tcoords.setTuple(i, tc);
		}
		const pointIds = Array.from({ length: numPts }, (_, i) => i);
		lines.insertNextCell(pointIds);
		output.setPoints(points);
		output.getPointData().setTCoords(tcoords);
		output.setLines(lines);
		outData[0] = output;
	};
}
var DEFAULT_VALUES = {
	point1: [
		0,
		.5,
		0
	],
	point2: [
		.5,
		0,
		0
	],
	center: [
		0,
		0,
		0
	],
	normal: [
		0,
		0,
		1
	],
	polarVector: [
		1,
		0,
		0
	],
	angle: 90,
	resolution: 6,
	negative: false,
	useNormalAndAngle: false,
	outputPointsPrecision: DesiredOutputPrecision.SINGLE
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	if (model.resolution < 1) model.resolution = 1;
	macros_default.obj(publicAPI, model);
	macros_default.algo(publicAPI, model, 0, 1);
	macros_default.setGet(publicAPI, model, [
		"resolution",
		"angle",
		"negative",
		"useNormalAndAngle",
		"outputPointsPrecision"
	]);
	macros_default.setGetArray(publicAPI, model, [
		"point1",
		"point2",
		"center",
		"normal",
		"polarVector"
	], 3);
	vtkArcSource(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkArcSource");
var ArcSource_default = {
	newInstance,
	extend
};
//#endregion
export { ArcSource_default as default, extend, newInstance };
