// @kitware/vtk.js/Filters/Sources/ArrowSource.js
import macros_default from "../../macros.js";
import MatrixBuilder_default from "../../Common/Core/MatrixBuilder.js";
import AppendPolyData_default from "../General/AppendPolyData.js";
import ConeSource_default from "./ConeSource.js";
import CylinderSource_default from "./CylinderSource.js";
//#region Sources/Filters/Sources/ArrowSource/index.js
function vtkArrowSource(publicAPI, model) {
	model.classHierarchy.push("vtkArrowSource");
	publicAPI.requestData = (inData, outData) => {
		const cylinder = CylinderSource_default.newInstance({ capping: true });
		cylinder.setResolution(model.shaftResolution);
		cylinder.setRadius(model.shaftRadius);
		cylinder.setHeight(1 - model.tipLength);
		cylinder.setCenter(0, (1 - model.tipLength) * .5, 0);
		const cylinderPD = cylinder.getOutputData();
		const cylinderPts = cylinderPD.getPoints().getData();
		const cylinderNormals = cylinderPD.getPointData().getNormals().getData();
		MatrixBuilder_default.buildFromDegree().rotateZ(-90).apply(cylinderPts).apply(cylinderNormals);
		const cone = ConeSource_default.newInstance();
		cone.setResolution(model.tipResolution);
		cone.setHeight(model.tipLength);
		cone.setRadius(model.tipRadius);
		const conePD = cone.getOutputData();
		const conePts = conePD.getPoints().getData();
		MatrixBuilder_default.buildFromRadian().translate(1 - model.tipLength * .5, 0, 0).apply(conePts);
		const append = AppendPolyData_default.newInstance();
		append.setInputData(cylinderPD);
		append.addInputData(conePD);
		const appendPD = append.getOutputData();
		const appendPts = appendPD.getPoints().getData();
		MatrixBuilder_default.buildFromRadian().translate(-.5 + model.tipLength * .5, 0, 0).apply(appendPts);
		if (model.invert) {
			MatrixBuilder_default.buildFromRadian().rotateFromDirections([
				1,
				0,
				0
			], model.direction).scale(-1, -1, -1).apply(appendPts);
			outData[0] = appendPD;
		} else {
			MatrixBuilder_default.buildFromRadian().rotateFromDirections([
				1,
				0,
				0
			], model.direction).scale(1, 1, 1).apply(appendPts);
			outData[0] = append.getOutputData();
		}
	};
}
var DEFAULT_VALUES = {
	tipResolution: 6,
	tipRadius: .1,
	tipLength: .35,
	shaftResolution: 6,
	shaftRadius: .03,
	invert: false,
	direction: [
		1,
		0,
		0
	],
	pointType: "Float64Array"
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, [
		"tipResolution",
		"tipRadius",
		"tipLength",
		"shaftResolution",
		"shaftRadius",
		"invert"
	]);
	macros_default.setGetArray(publicAPI, model, ["direction"], 3);
	macros_default.algo(publicAPI, model, 0, 1);
	vtkArrowSource(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkArrowSource");
var ArrowSource_default = {
	newInstance,
	extend
};
//#endregion
export { ArrowSource_default as default, extend, newInstance };
