// @kitware/vtk.js/Filters/Sources/CircleSource.js
import macros_default from "../../macros.js";
import { multiplyScalar } from "../../Common/Core/Math.js";
import MatrixBuilder_default from "../../Common/Core/MatrixBuilder.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
//#region Sources/Filters/Sources/CircleSource/index.js
function vtkCircleSource(publicAPI, model) {
	model.classHierarchy.push("vtkCircleSource");
	publicAPI.requestData = (inData, outData) => {
		const dataset = outData[0]?.initialize() || PolyData_default.newInstance();
		const points = macros_default.newTypedArray(model.pointType, model.resolution * 3);
		const edges = new Uint32Array(model.resolution + 2);
		edges[0] = model.resolution + 1;
		const angle = 2 * Math.PI / model.resolution;
		for (let i = 0; i < model.resolution; i++) {
			const x = model.center[0];
			const y = model.radius * Math.cos(i * angle) + model.center[1];
			const z = model.radius * Math.sin(i * angle) + model.center[2];
			points.set([
				x,
				y,
				z
			], i * 3);
			edges[i + 1] = i;
		}
		edges[edges.length - 1] = edges[1];
		dataset.getPoints().setData(points, 3);
		if (model.lines) dataset.getLines().setData(edges, 1);
		if (model.face) dataset.getPolys().setData(edges, 1);
		MatrixBuilder_default.buildFromRadian().translate(...model.center).rotateFromDirections([
			1,
			0,
			0
		], model.direction).translate(...multiplyScalar([...model.center], -1)).apply(points);
		outData[0] = dataset;
	};
}
function defaultValues(initialValues) {
	return {
		face: true,
		center: [
			0,
			0,
			0
		],
		lines: false,
		direction: [
			1,
			0,
			0
		],
		pointType: "Float64Array",
		radius: 1,
		resolution: 6,
		...initialValues
	};
}
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, defaultValues(initialValues));
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, [
		"radius",
		"resolution",
		"lines",
		"face"
	]);
	macros_default.setGetArray(publicAPI, model, ["center", "direction"], 3);
	macros_default.algo(publicAPI, model, 0, 1);
	vtkCircleSource(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkCircleSource");
var CircleSource_default = {
	newInstance,
	extend
};
//#endregion
export { CircleSource_default as default, extend, newInstance };
