// @kitware/vtk.js/Filters/Sources/ConeSource.js
import macros_default from "../../macros.js";
import MatrixBuilder_default from "../../Common/Core/MatrixBuilder.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
//#region Sources/Filters/Sources/ConeSource/index.js
function vtkConeSource(publicAPI, model) {
	model.classHierarchy.push("vtkConeSource");
	publicAPI.requestData = (inData, outData) => {
		const angle = 2 * Math.PI / model.resolution;
		const xbot = -model.height / 2;
		const numberOfPoints = model.resolution + 1;
		const cellArraySize = 4 * model.resolution + 1 + model.resolution;
		let pointIdx = 0;
		const points = macros_default.newTypedArray(model.pointType, numberOfPoints * 3);
		let cellLocation = 0;
		const polys = new Uint32Array(cellArraySize);
		points[0] = model.height / 2;
		points[1] = 0;
		points[2] = 0;
		if (model.capping) polys[cellLocation++] = model.resolution;
		for (let i = 0; i < model.resolution; i++) {
			pointIdx++;
			points[pointIdx * 3 + 0] = xbot;
			points[pointIdx * 3 + 1] = model.radius * Math.cos(i * angle);
			points[pointIdx * 3 + 2] = model.radius * Math.sin(i * angle);
			if (model.capping) polys[model.resolution - cellLocation++ + 1] = pointIdx;
		}
		for (let i = 0; i < model.resolution; i++) {
			polys[cellLocation++] = 3;
			polys[cellLocation++] = 0;
			polys[cellLocation++] = i + 1;
			polys[cellLocation++] = i + 2 > model.resolution ? 1 : i + 2;
		}
		MatrixBuilder_default.buildFromRadian().translate(...model.center).rotateFromDirections([
			1,
			0,
			0
		], model.direction).apply(points);
		const dataset = outData[0]?.initialize() || PolyData_default.newInstance();
		dataset.getPoints().setData(points, 3);
		dataset.getPolys().setData(polys, 1);
		outData[0] = dataset;
	};
}
var DEFAULT_VALUES = {
	height: 1,
	radius: .5,
	resolution: 6,
	center: [
		0,
		0,
		0
	],
	direction: [
		1,
		0,
		0
	],
	capping: true,
	pointType: "Float64Array"
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, [
		"height",
		"radius",
		"resolution",
		"capping"
	]);
	macros_default.setGetArray(publicAPI, model, ["center", "direction"], 3);
	macros_default.algo(publicAPI, model, 0, 1);
	vtkConeSource(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkConeSource");
var ConeSource_default = {
	newInstance,
	extend
};
//#endregion
export { ConeSource_default as default, extend, newInstance };
