// @kitware/vtk.js/Filters/Sources/CylinderSource.js
import macros_default from "../../macros.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import MatrixBuilder_default from "../../Common/Core/MatrixBuilder.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
//#region Sources/Filters/Sources/CylinderSource/index.js
function vtkCylinderSource(publicAPI, model) {
	model.classHierarchy.push("vtkCylinderSource");
	publicAPI.requestData = (inData, outData) => {
		const angle = 2 * Math.PI / model.resolution;
		let numberOfPoints = 2 * model.resolution;
		let numberOfPolys = 5 * model.resolution;
		if (model.capping) {
			numberOfPoints = 4 * model.resolution;
			numberOfPolys = 7 * model.resolution + 2;
		}
		const points = macros_default.newTypedArray(model.pointType, numberOfPoints * 3);
		let cellLocation = 0;
		const polys = new Uint32Array(numberOfPolys);
		const normalsData = new Float32Array(numberOfPoints * 3);
		const normals = DataArray_default.newInstance({
			numberOfComponents: 3,
			values: normalsData,
			name: "Normals"
		});
		const tcData = new Float32Array(numberOfPoints * 2);
		const tcoords = DataArray_default.newInstance({
			numberOfComponents: 2,
			values: tcData,
			name: "TCoords"
		});
		const nbot = [
			0,
			0,
			0
		];
		const ntop = [
			0,
			0,
			0
		];
		const xbot = [
			0,
			0,
			0
		];
		const xtop = [
			0,
			0,
			0
		];
		const tcbot = [0, 0];
		const tctop = [0, 0];
		const otherRadius = model.otherRadius == null ? model.radius : model.otherRadius;
		for (let i = 0; i < model.resolution; i++) {
			nbot[0] = Math.cos(i * angle + model.initAngle);
			ntop[0] = nbot[0];
			xbot[0] = model.radius * nbot[0] + model.center[0];
			xtop[0] = xbot[0];
			tcbot[0] = Math.abs(2 * i / model.resolution - 1);
			tctop[0] = tcbot[0];
			xbot[1] = .5 * model.height + model.center[1];
			xtop[1] = -.5 * model.height + model.center[1];
			tcbot[1] = 0;
			tctop[1] = 1;
			nbot[2] = -Math.sin(i * angle + model.initAngle);
			ntop[2] = nbot[2];
			xbot[2] = otherRadius * nbot[2] + model.center[2];
			xtop[2] = xbot[2];
			const pointIdx = 2 * i;
			for (let j = 0; j < 3; j++) {
				normalsData[pointIdx * 3 + j] = nbot[j];
				normalsData[(pointIdx + 1) * 3 + j] = ntop[j];
				points[pointIdx * 3 + j] = xbot[j];
				points[(pointIdx + 1) * 3 + j] = xtop[j];
				if (j < 2) {
					tcData[pointIdx * 2 + j] = tcbot[j];
					tcData[(pointIdx + 1) * 2 + j] = tctop[j];
				}
			}
		}
		for (let i = 0; i < model.resolution; i++) {
			polys[cellLocation++] = 4;
			polys[cellLocation++] = 2 * i;
			polys[cellLocation++] = 2 * i + 1;
			const pt = (2 * i + 3) % (2 * model.resolution);
			polys[cellLocation++] = pt;
			polys[cellLocation++] = pt - 1;
		}
		if (model.capping) {
			for (let i = 0; i < model.resolution; i++) {
				xbot[0] = model.radius * Math.cos(i * angle + model.initAngle);
				xtop[0] = xbot[0];
				tcbot[0] = xbot[0];
				tctop[0] = xbot[0];
				xbot[0] += model.center[0];
				xtop[0] += model.center[0];
				nbot[1] = 1;
				ntop[1] = -1;
				xbot[1] = .5 * model.height + model.center[1];
				xtop[1] = -.5 * model.height + model.center[1];
				xbot[2] = -otherRadius * Math.sin(i * angle + model.initAngle);
				xtop[2] = xbot[2];
				tcbot[1] = xbot[2];
				tctop[1] = xbot[2];
				xbot[2] += model.center[2];
				xtop[2] += model.center[2];
				const botIdx = 2 * model.resolution + i;
				const topIdx = 3 * model.resolution + model.resolution - i - 1;
				for (let j = 0; j < 3; j++) {
					normalsData[3 * botIdx + j] = nbot[j];
					normalsData[3 * topIdx + j] = ntop[j];
					points[3 * botIdx + j] = xbot[j];
					points[3 * topIdx + j] = xtop[j];
					if (j < 2) {
						tcData[2 * botIdx + j] = tcbot[j];
						tcData[2 * topIdx + j] = tctop[j];
					}
				}
			}
			polys[cellLocation++] = model.resolution;
			for (let i = 0; i < model.resolution; i++) polys[cellLocation++] = 2 * model.resolution + i;
			polys[cellLocation++] = model.resolution;
			for (let i = 0; i < model.resolution; i++) polys[cellLocation++] = 3 * model.resolution + i;
		}
		MatrixBuilder_default.buildFromRadian().translate(...model.center).rotateFromDirections([
			0,
			1,
			0
		], model.direction).translate(...model.center.map((c) => c * -1)).apply(points);
		MatrixBuilder_default.buildFromRadian().rotateFromDirections([
			0,
			1,
			0
		], model.direction).apply(normalsData);
		const dataset = outData[0]?.initialize() || PolyData_default.newInstance();
		dataset.getPoints().setData(points, 3);
		dataset.getPolys().setData(polys, 1);
		dataset.getPointData().setNormals(normals);
		dataset.getPointData().setTCoords(tcoords);
		outData[0] = dataset;
	};
}
var DEFAULT_VALUES = {
	height: 1,
	initAngle: 0,
	radius: 1,
	resolution: 6,
	center: [
		0,
		0,
		0
	],
	direction: [
		0,
		1,
		0
	],
	capping: true,
	pointType: "Float64Array"
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, [
		"height",
		"initAngle",
		"otherRadius",
		"radius",
		"resolution",
		"capping"
	]);
	macros_default.setGetArray(publicAPI, model, ["center", "direction"], 3);
	macros_default.algo(publicAPI, model, 0, 1);
	vtkCylinderSource(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkCylinderSource");
var CylinderSource_default = {
	newInstance,
	extend
};
//#endregion
export { CylinderSource_default as default, extend, newInstance };
