// @kitware/vtk.js/Filters/Sources/DiskSource.js
import macros_default from "../../macros.js";
import CellArray_default from "../../Common/Core/CellArray.js";
import MatrixBuilder_default from "../../Common/Core/MatrixBuilder.js";
import Points_default from "../../Common/Core/Points.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
//#region Sources/Filters/Sources/DiskSource/index.js
var { vtkErrorMacro } = macros_default;
function vtkDiskSource(publicAPI, model) {
	model.classHierarchy.push("vtkDiskSource");
	publicAPI.requestData = (inData, outData) => {
		const { innerRadius, outerRadius, center, normal, radialResolution, circumferentialResolution, pointType } = model;
		const points = Points_default.newInstance({ dataType: pointType });
		const n = [
			normal[0],
			normal[1],
			normal[2]
		];
		const length = Math.hypot(n[0], n[1], n[2]);
		if (length === 0) {
			vtkErrorMacro("Normal vector cannot be zero-length");
			return;
		}
		n[0] /= length;
		n[1] /= length;
		n[2] /= length;
		const matrix = MatrixBuilder_default.buildFromDegree().identity().translate(-center[0], -center[1], -center[2]).rotateFromDirections([
			0,
			0,
			1
		], n).translate(center[0], center[1], center[2]).getMatrix();
		const polys = CellArray_default.newInstance();
		const deltaR = (outerRadius - innerRadius) / radialResolution;
		const thetaStep = 2 * Math.PI / circumferentialResolution;
		for (let i = 0; i < circumferentialResolution; i++) {
			const theta = i * thetaStep;
			const cosT = Math.cos(theta);
			const sinT = Math.sin(theta);
			for (let j = 0; j <= radialResolution; j++) {
				const r = innerRadius + j * deltaR;
				const localPoint = [
					center[0] + r * cosT,
					center[1] + r * sinT,
					center[2]
				];
				const x = matrix[0] * localPoint[0] + matrix[4] * localPoint[1] + matrix[8] * localPoint[2] + matrix[12];
				const y = matrix[1] * localPoint[0] + matrix[5] * localPoint[1] + matrix[9] * localPoint[2] + matrix[13];
				const z = matrix[2] * localPoint[0] + matrix[6] * localPoint[1] + matrix[10] * localPoint[2] + matrix[14];
				points.insertNextPoint(x, y, z);
			}
		}
		const cellCount = radialResolution * circumferentialResolution;
		const cellData = new Uint8Array(cellCount * 5);
		let offset = 0;
		for (let i = 0; i < model.circumferentialResolution; i++) for (let j = 0; j < model.radialResolution; j++) {
			const p0 = i * (model.radialResolution + 1) + j;
			const p1 = p0 + 1;
			const p2 = i < model.circumferentialResolution - 1 ? p1 + (model.radialResolution + 1) : j + 1;
			const p3 = p2 - 1;
			cellData[offset++] = 4;
			cellData[offset++] = p0;
			cellData[offset++] = p1;
			cellData[offset++] = p2;
			cellData[offset++] = p3;
		}
		polys.setData(cellData, cellCount, 1);
		const dataset = outData[0]?.initialize() || PolyData_default.newInstance();
		dataset.setPoints(points);
		dataset.setPolys(polys);
		outData[0] = dataset;
	};
}
function defaultValues(initialValues) {
	return {
		innerRadius: .25,
		outerRadius: .5,
		center: [
			0,
			0,
			0
		],
		normal: [
			0,
			0,
			1
		],
		radialResolution: 1,
		circumferentialResolution: 6,
		pointType: "Float64Array",
		...initialValues
	};
}
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, defaultValues(initialValues));
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, [
		"innerRadius",
		"outerRadius",
		"radialResolution",
		"circumferentialResolution",
		"pointType"
	]);
	macros_default.setGetArray(publicAPI, model, ["center", "normal"], 3);
	macros_default.algo(publicAPI, model, 0, 1);
	vtkDiskSource(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkDiskSource");
var DiskSource_default = {
	newInstance,
	extend
};
//#endregion
export { DiskSource_default as default, extend, newInstance };
