// @kitware/vtk.js/Filters/Sources/EllipseArcSource.js
import macros_default from "../../macros.js";
import Math_default from "../../Common/Core/Math.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import CellArray_default from "../../Common/Core/CellArray.js";
import { DesiredOutputPrecision } from "../../Common/DataModel/DataSetAttributes/Constants.js";
import Points_default from "../../Common/Core/Points.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
//#region Sources/Filters/Sources/EllipseArcSource/index.js
var { vtkErrorMacro } = macros_default;
var { VtkDataTypes } = DataArray_default;
function vtkEllipseArcSource(publicAPI, model) {
	model.classHierarchy.push("vtkEllipseArcSource");
	publicAPI.requestData = (inData, outData) => {
		const isClosedShape = Math.abs(model.segmentAngle - 360) < 1e-5;
		const resolution = model.close && !isClosedShape ? model.resolution + 1 : model.resolution;
		const numLines = resolution;
		const numPts = resolution + 1;
		const tc = [0, 0];
		const output = outData[0]?.initialize() || PolyData_default.newInstance();
		const normal = [...model.normal];
		Math_default.normalize(normal);
		const orthogonalVect = [
			0,
			0,
			0
		];
		Math_default.cross(normal, model.majorRadiusVector, orthogonalVect);
		if (Math_default.norm(orthogonalVect) < 1e-10) {
			vtkErrorMacro("Ellipse normal vector and major radius axis are collinear");
			return;
		}
		Math_default.normalize(orthogonalVect);
		const majorRadiusVect = [
			0,
			0,
			0
		];
		Math_default.cross(orthogonalVect, normal, majorRadiusVect);
		Math_default.normalize(majorRadiusVect);
		const a = Math_default.norm(model.majorRadiusVector);
		const b = a * model.ratio;
		let startAngleRad = Math_default.radiansFromDegrees(model.startAngle);
		if (startAngleRad < 0) startAngleRad += 2 * Math.PI;
		const angleIncRad = Math_default.radiansFromDegrees(model.segmentAngle) / model.resolution;
		let pointType = VtkDataTypes.FLOAT;
		if (model.outputPointsPrecision === DesiredOutputPrecision.SINGLE) pointType = VtkDataTypes.FLOAT;
		else if (model.outputPointsPrecision === DesiredOutputPrecision.DOUBLE) pointType = VtkDataTypes.DOUBLE;
		const points = Points_default.newInstance({ dataType: pointType });
		points.setNumberOfPoints(numPts);
		const tcoords = DataArray_default.newInstance({
			numberOfComponents: 2,
			size: numPts * 2,
			dataType: VtkDataTypes.FLOAT,
			name: "TextureCoordinates"
		});
		const lines = CellArray_default.newInstance();
		lines.allocate(numLines);
		const skipLastPoint = model.close && isClosedShape;
		let theta = startAngleRad;
		let pointIndex = 0;
		for (let i = 0; i <= resolution; ++i, theta += angleIncRad) {
			const quotient = Math.floor(theta / (2 * Math.PI));
			const normalizedTheta = theta - quotient * 2 * Math.PI;
			let thetaEllipse = Math.atan(Math.tan(normalizedTheta) * model.ratio);
			if (normalizedTheta > Math.PI / 2 && normalizedTheta <= Math.PI) thetaEllipse += Math.PI;
			else if (normalizedTheta > Math.PI && normalizedTheta <= 1.5 * Math.PI) thetaEllipse -= Math.PI;
			const cosTheta = Math.cos(thetaEllipse);
			const sinTheta = Math.sin(thetaEllipse);
			const p = [
				model.center[0] + a * cosTheta * majorRadiusVect[0] + b * sinTheta * orthogonalVect[0],
				model.center[1] + a * cosTheta * majorRadiusVect[1] + b * sinTheta * orthogonalVect[1],
				model.center[2] + a * cosTheta * majorRadiusVect[2] + b * sinTheta * orthogonalVect[2]
			];
			tc[0] = i / resolution;
			tc[1] = 0;
			if (i !== resolution || !skipLastPoint) {
				points.setPoint(pointIndex, ...p);
				tcoords.setTuple(pointIndex, tc);
				pointIndex++;
			}
		}
		const actualNumPts = pointIndex;
		const pointIds = [];
		for (let k = 0; k < actualNumPts - 1; ++k) pointIds.push(k);
		if (model.close) pointIds.push(0);
		else pointIds.push(actualNumPts - 1);
		lines.insertNextCell(pointIds);
		output.setPoints(points);
		output.getPointData().setTCoords(tcoords);
		output.setLines(lines);
		outData[0] = output;
	};
}
var DEFAULT_VALUES = {
	center: [
		0,
		0,
		0
	],
	normal: [
		0,
		0,
		1
	],
	majorRadiusVector: [
		1,
		0,
		0
	],
	startAngle: 0,
	segmentAngle: 90,
	resolution: 100,
	close: false,
	outputPointsPrecision: DesiredOutputPrecision.SINGLE,
	ratio: 1
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	if (model.resolution < 1) model.resolution = 1;
	macros_default.obj(publicAPI, model);
	macros_default.algo(publicAPI, model, 0, 1);
	macros_default.setGet(publicAPI, model, [
		"resolution",
		"startAngle",
		"segmentAngle",
		"close",
		"outputPointsPrecision",
		"ratio"
	]);
	macros_default.setGetArray(publicAPI, model, [
		"center",
		"normal",
		"majorRadiusVector"
	], 3);
	vtkEllipseArcSource(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkEllipseArcSource");
var EllipseArcSource_default = {
	newInstance,
	extend
};
//#endregion
export { EllipseArcSource_default as default, extend, newInstance };
