// @kitware/vtk.js/Filters/Sources/LineSource.js
import macros_default from "../../macros.js";
import { norm, subtract } from "../../Common/Core/Math.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
//#region Sources/Filters/Sources/LineSource/index.js
var { vtkWarningMacro } = macros_default;
function vtkLineSource(publicAPI, model) {
	model.classHierarchy.push("vtkLineSource");
	publicAPI.requestData = (inData, outData) => {
		const dataset = outData[0];
		const pointDataType = dataset ? dataset.getPoints().getDataType() : model.pointType;
		const pd = dataset?.initialize() || PolyData_default.newInstance();
		const v21 = [];
		subtract(model.point2, model.point1, v21);
		if (norm(v21) <= 0) {
			vtkWarningMacro("Zero-length line definition");
			return;
		}
		const res = model.resolution;
		const numPts = res + 1;
		const points = macros_default.newTypedArray(pointDataType, numPts * 3);
		pd.getPoints().setData(points, 3);
		const lines = new Uint32Array(numPts + 1);
		pd.getLines().setData(lines, 1);
		let idx = 0;
		let t = 0;
		for (let i = 0; i < res + 1; i++) {
			t = i / res;
			points[idx * 3] = model.point1[0] + t * v21[0];
			points[idx * 3 + 1] = model.point1[1] + t * v21[1];
			points[idx * 3 + 2] = model.point1[2] + t * v21[2];
			idx++;
		}
		idx = 0;
		lines[0] = numPts;
		for (let i = 0; i < numPts; i++) lines[i + 1] = i;
		outData[0] = pd;
	};
}
var DEFAULT_VALUES = {
	resolution: 10,
	point1: [
		-1,
		0,
		0
	],
	point2: [
		1,
		0,
		0
	],
	pointType: "Float64Array"
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, ["resolution"]);
	macros_default.setGetArray(publicAPI, model, ["point1", "point2"], 3);
	macros_default.algo(publicAPI, model, 0, 1);
	vtkLineSource(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkLineSource");
var LineSource_default = {
	newInstance,
	extend
};
//#endregion
export { LineSource_default as default, extend, newInstance };
