// @kitware/vtk.js/Filters/Sources/PlaneSource.js
import macros_default from "../../macros.js";
import { cross, dot, normalize, radiansFromDegrees, subtract } from "../../Common/Core/Math.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import MatrixBuilder_default from "../../Common/Core/MatrixBuilder.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
import { mat4, vec3 } from "gl-matrix";
//#region Sources/Filters/Sources/PlaneSource/index.js
var { vtkWarningMacro } = macros_default;
var EPSILON = 1e-6;
function vtkPlaneSource(publicAPI, model) {
	model.classHierarchy.push("vtkPlaneSource");
	publicAPI.requestData = (inData, outData) => {
		const dataset = outData[0];
		const pointDataType = dataset ? dataset.getPoints().getDataType() : model.pointType;
		const pd = dataset?.initialize() || PolyData_default.newInstance();
		const v10 = [];
		const v20 = [];
		subtract(model.point1, model.origin, v10);
		subtract(model.point2, model.origin, v20);
		if (!publicAPI.updatePlane(v10, v20)) {
			vtkWarningMacro("Bad plane definition");
			return;
		}
		const xres = model.xResolution;
		const yres = model.yResolution;
		const numPts = (xres + 1) * (yres + 1);
		const numPolys = xres * yres;
		const points = macros_default.newTypedArray(pointDataType, numPts * 3);
		pd.getPoints().setData(points, 3);
		const polys = new Uint32Array(5 * numPolys);
		pd.getPolys().setData(polys, 1);
		const normalsData = new Float32Array(numPts * 3);
		const normals = DataArray_default.newInstance({
			numberOfComponents: 3,
			values: normalsData,
			name: "Normals"
		});
		pd.getPointData().setNormals(normals);
		const tcData = new Float32Array(numPts * 2);
		const tcoords = DataArray_default.newInstance({
			numberOfComponents: 2,
			values: tcData,
			name: "TextureCoordinates"
		});
		pd.getPointData().setTCoords(tcoords);
		const tc = new Float32Array(2);
		let idx = 0;
		for (let j = 0; j < yres + 1; j++) {
			tc[1] = j / yres;
			for (let i = 0; i < xres + 1; i++) {
				tc[0] = i / xres;
				points[idx * 3] = model.origin[0] + tc[0] * v10[0] + tc[1] * v20[0];
				points[idx * 3 + 1] = model.origin[1] + tc[0] * v10[1] + tc[1] * v20[1];
				points[idx * 3 + 2] = model.origin[2] + tc[0] * v10[2] + tc[1] * v20[2];
				tcData[idx * 2] = tc[0];
				tcData[idx * 2 + 1] = tc[1];
				normalsData[idx * 3] = model.normal[0];
				normalsData[idx * 3 + 1] = model.normal[1];
				normalsData[idx * 3 + 2] = model.normal[2];
				idx++;
			}
		}
		idx = 0;
		for (let j = 0; j < yres; j++) for (let i = 0; i < xres; i++) {
			polys[idx * 5 + 0] = 4;
			polys[idx * 5 + 1] = i + j * (xres + 1);
			polys[idx * 5 + 2] = polys[idx * 5 + 1] + 1;
			polys[idx * 5 + 3] = polys[idx * 5 + 1] + xres + 2;
			polys[idx * 5 + 4] = polys[idx * 5 + 1] + xres + 1;
			idx++;
		}
		outData[0] = pd;
	};
	publicAPI.setNormal = (...normal) => {
		let n = [];
		if (normal.length === 1 || Array.isArray(normal[0])) n = [...normal[0]];
		else if (normal.length === 3) n = [
			normal[0],
			normal[1],
			normal[2]
		];
		if (normalize(n) !== 0) {
			const dp = dot(model.normal, n);
			let theta = 0;
			const rotationVector = [];
			if (dp < 1) {
				if (dp <= -1) {
					theta = radiansFromDegrees(180);
					subtract(model.point1, model.origin, rotationVector);
				} else {
					cross(model.normal, n, rotationVector);
					theta = Math.acos(dp);
				}
				publicAPI.rotate(theta, rotationVector);
			}
		}
	};
	/**
	* Rotate plane around a given axis
	* @param {float} theta Angle (radian) to rotate about
	* @param {vec3} rotationAxis Axis to rotate around
	*/
	publicAPI.rotate = (angle, rotationAxis) => {
		if (Math.abs(angle) < EPSILON) return;
		const transform = mat4.identity(new Float64Array(16));
		const negCenter = [];
		vec3.negate(negCenter, model.center);
		mat4.translate(transform, transform, model.center);
		mat4.rotate(transform, transform, angle, rotationAxis);
		mat4.translate(transform, transform, negCenter);
		vec3.transformMat4(model.origin, model.origin, transform);
		vec3.transformMat4(model.point1, model.point1, transform);
		vec3.transformMat4(model.point2, model.point2, transform);
		MatrixBuilder_default.buildFromRadian().rotate(angle, rotationAxis).apply(model.normal);
		publicAPI.modified();
	};
	publicAPI.setCenter = (...center) => {
		let c = [];
		if (center.length === 1 || Array.isArray(center[0])) c = [...center[0]];
		else if (center.length === 3) c = [
			center[0],
			center[1],
			center[2]
		];
		if (!vec3.exactEquals(c, model.center)) {
			const v1 = [];
			subtract(model.point1, model.origin, v1);
			const v2 = [];
			subtract(model.point2, model.origin, v2);
			for (let i = 0; i < 3; i++) {
				model.center[i] = c[i];
				model.origin[i] = model.center[i] - .5 * (v1[i] + v2[i]);
				model.point1[i] = model.origin[i] + v1[i];
				model.point2[i] = model.origin[i] + v2[i];
			}
			publicAPI.modified();
		}
	};
	publicAPI.setPoint1 = (...point) => {
		let point1 = [];
		if (point.length === 1 || Array.isArray(point[0])) point1 = [...point[0]];
		else if (point.length === 3) point1 = [
			point[0],
			point[1],
			point[2]
		];
		if (!vec3.exactEquals(point1, model.point1)) {
			const v1 = [];
			const v2 = [];
			model.point1 = [...point1];
			subtract(model.point1, model.origin, v1);
			subtract(model.point2, model.origin, v2);
			publicAPI.updatePlane(v1, v2);
			publicAPI.modified();
		}
	};
	publicAPI.setPoint2 = (...point) => {
		let point2 = [];
		if (point.length === 1 || Array.isArray(point[0])) point2 = [...point[0]];
		else if (point.length === 3) point2 = [
			point[0],
			point[1],
			point[2]
		];
		if (!vec3.exactEquals(point2, model.point2)) {
			const v1 = [];
			const v2 = [];
			model.point2 = [...point2];
			subtract(model.point1, model.origin, v1);
			subtract(model.point2, model.origin, v2);
			publicAPI.updatePlane(v1, v2);
			publicAPI.modified();
		}
	};
	publicAPI.updatePlane = (v1, v2) => {
		for (let i = 0; i < 3; i++) model.center[i] = model.origin[i] + .5 * (v1[i] + v2[i]);
		cross(v1, v2, model.normal);
		return normalize(model.normal) !== 0;
	};
}
var DEFAULT_VALUES = {
	xResolution: 10,
	yResolution: 10,
	origin: [
		0,
		0,
		0
	],
	point1: [
		1,
		0,
		0
	],
	point2: [
		0,
		1,
		0
	],
	pointType: "Float64Array"
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	model.normal = [
		0,
		0,
		1
	];
	model.center = [
		0,
		0,
		0
	];
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, ["xResolution", "yResolution"]);
	macros_default.setGetArray(publicAPI, model, ["origin"], 3);
	macros_default.getArray(publicAPI, model, [
		"point1",
		"point2",
		"normal",
		"center"
	], 3);
	macros_default.algo(publicAPI, model, 0, 1);
	vtkPlaneSource(publicAPI, model);
	publicAPI.setPoint1(model.point1);
	publicAPI.setPoint2(model.point2);
}
var newInstance = macros_default.newInstance(extend, "vtkPlaneSource");
var PlaneSource_default = {
	newInstance,
	extend
};
//#endregion
export { PlaneSource_default as default, extend, newInstance };
