// @kitware/vtk.js/Filters/Sources/PointSource.js
import macros_default from "../../macros.js";
import { random } from "../../Common/Core/Math.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
//#region Sources/Filters/Sources/PointSource/index.js
function vtkPointSource(publicAPI, model) {
	model.classHierarchy.push("vtkPointSource");
	publicAPI.requestData = (inData, outData) => {
		const dataset = outData[0];
		const pointDataType = dataset ? dataset.getPoints().getDataType() : model.pointType;
		const pd = dataset?.initialize() || PolyData_default.newInstance();
		const numPts = model.numberOfPoints;
		const points = macros_default.newTypedArray(pointDataType, numPts * 3);
		pd.getPoints().setData(points, 3);
		const verts = new Uint32Array(numPts + 1);
		pd.getVerts().setData(verts, 1);
		let cosphi;
		let sinphi;
		let rho;
		let radius;
		let theta;
		for (let i = 0; i < numPts; i++) {
			cosphi = 1 - 2 * random();
			sinphi = Math.sqrt(1 - cosphi * cosphi);
			rho = model.radius * random() ** .33333333;
			radius = rho * sinphi;
			theta = 2 * Math.PI * random();
			points[i * 3] = model.center[0] + radius * Math.cos(theta);
			points[i * 3 + 1] = model.center[1] + radius * Math.sin(theta);
			points[i * 3 + 2] = model.center[2] + rho * cosphi;
		}
		verts[0] = numPts;
		for (let i = 0; i < numPts; i++) verts[i + 1] = i;
		outData[0] = pd;
	};
}
var DEFAULT_VALUES = {
	numberOfPoints: 10,
	center: [
		0,
		0,
		0
	],
	radius: .5,
	pointType: "Float64Array"
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, ["numberOfPoints", "radius"]);
	macros_default.setGetArray(publicAPI, model, ["center"], 3);
	macros_default.algo(publicAPI, model, 0, 1);
	vtkPointSource(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkPointSource");
var PointSource_default = {
	newInstance,
	extend
};
//#endregion
export { PointSource_default as default, extend, newInstance };
