// @kitware/vtk.js/Filters/Sources/RegularPolygonSource.js
import macros_default from "../../macros.js";
import { VtkDataTypes } from "../../Common/Core/DataArray/Constants.js";
import Math_default from "../../Common/Core/Math.js";
import CellArray_default from "../../Common/Core/CellArray.js";
import { DesiredOutputPrecision } from "../../Common/DataModel/DataSetAttributes/Constants.js";
import Points_default from "../../Common/Core/Points.js";
import PolyData_default from "../../Common/DataModel/PolyData.js";
//#region Sources/Filters/Sources/RegularPolygonSource/index.js
function vtkRegularPolygonSource(publicAPI, model) {
	model.classHierarchy.push("vtkRegularPolygonSource");
	publicAPI.requestData = (inData, outData) => {
		const output = outData[0]?.initialize() || PolyData_default.newInstance();
		const numPts = model.numberOfSides;
		const newPoints = Points_default.newInstance({ dataType: model.outputPointsPrecision === DesiredOutputPrecision.DOUBLE ? VtkDataTypes.DOUBLE : VtkDataTypes.FLOAT });
		if (model.generatePolyline) {
			const newLine = CellArray_default.newInstance();
			const linePoints = [];
			for (let i = 0; i < numPts; i++) linePoints.push(i);
			linePoints.push(0);
			newLine.insertNextCell(linePoints);
			output.setLines(newLine);
		}
		if (model.generatePolygon) {
			const newPoly = CellArray_default.newInstance();
			const polyPoints = [];
			for (let i = 0; i < numPts; i++) polyPoints.push(i);
			newPoly.insertNextCell(polyPoints);
			output.setPolys(newPoly);
		}
		const n = [...model.normal];
		if (Math_default.normalize(n) === 0) {
			n[0] = 0;
			n[1] = 0;
			n[2] = 1;
		}
		const px = [
			0,
			0,
			0
		];
		const py = [
			0,
			0,
			0
		];
		let foundPlaneVector = false;
		const axis = [
			1,
			0,
			0
		];
		Math_default.cross(n, axis, px);
		if (Math_default.normalize(px) > .001) foundPlaneVector = true;
		if (!foundPlaneVector) {
			axis[0] = 0;
			axis[1] = 1;
			axis[2] = 0;
			Math_default.cross(n, axis, px);
			if (Math_default.normalize(px) > .001) foundPlaneVector = true;
		}
		if (!foundPlaneVector) {
			axis[0] = 0;
			axis[1] = 0;
			axis[2] = 1;
			Math_default.cross(n, axis, px);
			Math_default.normalize(px);
		}
		Math_default.cross(px, n, py);
		const theta = 2 * Math.PI / numPts;
		const points = [];
		const r = [
			0,
			0,
			0
		];
		const x = [
			0,
			0,
			0
		];
		for (let j = 0; j < numPts; j++) {
			const cosTheta = Math.cos(j * theta);
			const sinTheta = Math.sin(j * theta);
			r[0] = px[0] * cosTheta + py[0] * sinTheta;
			r[1] = px[1] * cosTheta + py[1] * sinTheta;
			r[2] = px[2] * cosTheta + py[2] * sinTheta;
			x[0] = model.center[0] + model.radius * r[0];
			x[1] = model.center[1] + model.radius * r[1];
			x[2] = model.center[2] + model.radius * r[2];
			points.push(x[0], x[1], x[2]);
		}
		newPoints.setData(points);
		output.setPoints(newPoints);
		outData[0] = output;
	};
}
var DEFAULT_VALUES = {
	numberOfSides: 6,
	center: [
		0,
		0,
		0
	],
	normal: [
		0,
		0,
		1
	],
	radius: .5,
	generatePolygon: true,
	generatePolyline: true,
	outputPointsPrecision: DesiredOutputPrecision.FLOAT
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.algo(publicAPI, model, 0, 1);
	macros_default.setGet(publicAPI, model, [
		"numberOfSides",
		"radius",
		"generatePolygon",
		"generatePolyline",
		"outputPointsPrecision"
	]);
	macros_default.setGetArray(publicAPI, model, ["center", "normal"], 3);
	vtkRegularPolygonSource(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkRegularPolygonSource");
var RegularPolygonSource_default = {
	newInstance,
	extend
};
//#endregion
export { RegularPolygonSource_default as default, extend, newInstance };
