// @kitware/vtk.js/Filters/Sources/RTAnalyticSource.js
import macros_default from "../../macros.js";
import DataArray_default from "../../Common/Core/DataArray.js";
import ImageData_default from "../../Common/DataModel/ImageData.js";
//#region Sources/Filters/Sources/RTAnalyticSource/index.js
function vtkRTAnalyticSource(publicAPI, model) {
	model.classHierarchy.push("vtkRTAnalyticSource");
	publicAPI.requestData = (inData, outData) => {
		const state = {};
		const dataset = {
			type: "vtkImageData",
			mtime: model.mtime,
			metadata: {
				source: "vtkRTAnalyticSource",
				state
			}
		};
		[
			"standardDeviation",
			"center",
			"frequency",
			"magnitude",
			"maximum"
		].forEach((field) => {
			state[field] = [].concat(model[field]);
		});
		const id = ImageData_default.newInstance(dataset);
		id.setSpacing(1, 1, 1);
		id.setExtent.apply(this, model.wholeExtent);
		id.setOrigin(0, 0, 0);
		id.setDirection(model.dataDirection);
		let dims = [
			0,
			0,
			0
		];
		dims = dims.map((_, i) => model.wholeExtent[i * 2 + 1] - model.wholeExtent[i * 2] + 1);
		const newArray = new Uint8Array(dims[0] * dims[1] * dims[2]);
		const temp2 = 1 / (2 * model.standardDeviation * model.standardDeviation);
		let xval = 0;
		let yval = 0;
		let zval = 0;
		const scale = [
			1 / (model.wholeExtent[1] - model.wholeExtent[0]),
			1 / (model.wholeExtent[3] - model.wholeExtent[2]),
			1 / (model.wholeExtent[5] - model.wholeExtent[4])
		];
		let i = 0;
		for (let z = model.wholeExtent[4]; z <= model.wholeExtent[5]; z++) {
			zval = (model.center[2] - z) * scale[2];
			const zfactor = model.magnitude[2] * Math.cos(model.frequency[2] * zval);
			zval *= zval;
			for (let y = model.wholeExtent[2]; y <= model.wholeExtent[3]; y++) {
				yval = (model.center[1] - y) * scale[1];
				const yfactor = model.magnitude[1] * Math.sin(model.frequency[1] * yval);
				yval *= yval;
				for (let x = model.wholeExtent[0]; x <= model.wholeExtent[1]; x++) {
					xval = (model.center[0] - x) * scale[0];
					const sum = zval + yval + xval * xval;
					const xfactor = model.magnitude[0] * Math.sin(model.frequency[0] * xval);
					newArray[i] = model.maximum * Math.exp(-sum * temp2) + xfactor + yfactor + zfactor + model.offset;
					i++;
				}
			}
		}
		const da = DataArray_default.newInstance({
			numberOfComponents: 1,
			values: newArray
		});
		da.setName("scalars");
		id.getPointData().setScalars(da);
		outData[0] = id;
	};
}
var DEFAULT_VALUES = {
	offset: 40,
	maximum: 120,
	center: [
		0,
		0,
		0
	],
	frequency: [
		60,
		30,
		40
	],
	magnitude: [
		10,
		18,
		5
	],
	standardDeviation: .5,
	wholeExtent: [
		-10,
		10,
		-10,
		10,
		-10,
		10
	],
	dataDirection: [
		1,
		0,
		0,
		0,
		1,
		0,
		0,
		0,
		1
	]
};
function extend(publicAPI, model, initialValues = {}) {
	Object.assign(model, DEFAULT_VALUES, initialValues);
	macros_default.obj(publicAPI, model);
	macros_default.setGet(publicAPI, model, [
		"offset",
		"maximum",
		"standardDeviation"
	]);
	macros_default.setGetArray(publicAPI, model, [
		"center",
		"frequency",
		"magnitude"
	], 3);
	macros_default.setGetArray(publicAPI, model, ["wholeExtent"], 6);
	macros_default.setGetArray(publicAPI, model, ["dataDirection"], 9);
	macros_default.algo(publicAPI, model, 0, 1);
	vtkRTAnalyticSource(publicAPI, model);
}
var newInstance = macros_default.newInstance(extend, "vtkRTAnalyticSource");
var RTAnalyticSource_default = {
	newInstance,
	extend
};
//#endregion
export { RTAnalyticSource_default as default, extend, newInstance };
